<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>R语言学习笔记 - 21&nbsp; Normalization and regressing out unwanted variation</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../single_cell/scRNA-seq_online/06_integration.html" rel="next">
<link href="../../single_cell/scRNA-seq_online/05_theory_of_PCA.html" rel="prev">
<link href="../../images/logo.jpg" rel="icon" type="image/jpeg">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark"><script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><script type="application/json" class="js-hypothesis-config">
{
  "theme": "classic",
  "openSidebar": false,
  "showHighlights": "always",
  "enableExperimentalNewNoteButton": true
}
</script><script async="" src="https://hypothes.is/embed.js"></script>
<meta name="twitter:title" content="R语言学习笔记 - 21&nbsp; Normalization and regressing out unwanted variation">
<meta name="twitter:description" content="">
<meta name="twitter:image" content="https://djhcod.github.io/r-notes/single_cell/scRNA-seq_online/images/sc_workflow_2022-01.jpg">
<meta name="twitter:card" content="summary_large_image">
</head>
<body class="nav-sidebar docked nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="navbar navbar-expand-lg navbar-dark "><div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a href="../../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../../images/logo.jpg" alt="" class="navbar-logo"></a>
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">R语言学习笔记</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
<li class="nav-item">
    <a class="nav-link active" href="../../index.html" rel="" target="" aria-current="page">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../intro.html" rel="" target="">
 <span class="menu-text">Introduction</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-parts" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Parts</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-parts">
<li>
    <a class="dropdown-item" href="../../r_basic/r_basics.html" rel="" target="">
 <span class="dropdown-text">R语言基础</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../single_cell/seurat/seurat.html" rel="" target="">
 <span class="dropdown-text">Seurat v5官方文档</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../single_cell/scRNA-seq_online/00_intro.html" rel="" target="">
 <span class="dropdown-text">scRNA-seq_online学习材料</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../quarto_foundation/quarto_foundation.html" rel="" target="">
 <span class="dropdown-text">Quarto基础</span></a>
  </li>  
    </ul>
</li>
  <li class="nav-item">
    <a class="nav-link" href="../../references.html" rel="" target="">
 <span class="menu-text">References</span></a>
  </li>  
</ul>
<ul class="navbar-nav navbar-nav-scroll ms-auto">
<li class="nav-item">
    <a class="nav-link" href="https://bookdown.org" rel="" target=""><i class="bi bi-book-fill" role="img">
</i> 
 <span class="menu-text">Bookdown</span></a>
  </li>  
</ul>
<div class="quarto-navbar-tools tools-wide">
    <a href="https://github.com/djhcod/r-notes" rel="" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
    <div class="dropdown">
      <a href="" title="Share" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Share"><i class="bi bi-share"></i></a>
      <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="quarto-navigation-tool-dropdown-0">
<li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://twitter.com/intent/tweet?url=%7Curl%7C">
              <i class="bi bi-bi-twitter pe-1"></i>
            Twitter
            </a>
          </li>
          <li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://www.facebook.com/sharer/sharer.php?u=%7Curl%7C">
              <i class="bi bi-bi-facebook pe-1"></i>
            Facebook
            </a>
          </li>
      </ul>
</div>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../single_cell/scRNA-seq_online/00_intro.html">scRNA-seq_online学习材料</a></li><li class="breadcrumb-item"><a href="../../single_cell/scRNA-seq_online/06_SC_SCT_normalization.html"><span class="chapter-number">21</span>&nbsp; <span class="chapter-title">Normalization and regressing out unwanted variation</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto"><div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">主页</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">前言</span></a>
  </div>
</li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../r_basic/r_basics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">R语言基础</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/environment_configuration.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Rstudio环境配置</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/data_input_output.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">数据的读取与输出</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/basic_data_function.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">数据处理基本函数</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/character.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">字符的处理</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../single_cell/seurat/seurat.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Seurat v5官方文档</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/seurat_command_list.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Seurat常用函数清单</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/seurat_tutorial.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Seurat细胞分群官方教程</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/data_visualization_methods_in_seurat.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Seurat中的数据可视化方法</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/sctransform.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">基于SCTransform的单细胞数据标准化</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/integration.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">单细胞数据整合（integration）</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/integrative_analysis_in_seurat_v5.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Seurat v5单细胞数据整合分析</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/mapping_and_annotating_query_datasets.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">基于参考集的细胞注释</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/cell_cycle_regression.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">消除细胞周期的影响</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/de_vignette.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">差异表达分析</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../single_cell/scRNA-seq_online/00_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">scRNA-seq_online学习材料</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/01_intro_to_scRNA-seq.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">单细胞测序技术介绍</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/02_SC_generation_of_count_matrix.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">单细胞RNA测序—从raw data到count matrix</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/03_SC_quality_control-setup.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">数据导入与Seurat对象构建</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/04_SC_quality_control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">质控</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/postQC_workflow.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">Single-cell RNA-seq Clustering Workflow</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/05_theory_of_PCA.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">PCA原理</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/06_SC_SCT_normalization.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">21</span>&nbsp; <span class="chapter-title">Normalization and regressing out unwanted variation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/06_integration.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">22</span>&nbsp; <span class="chapter-title">单细胞数据整合（Integration）</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/07_SC_clustering_cells_SCT.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">23</span>&nbsp; <span class="chapter-title">细胞聚类（clustering analysis）</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/08_SC_clustering_quality_control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">24</span>&nbsp; <span class="chapter-title">细胞分群质量评估</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../quarto_foundation/quarto_foundation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Quarto基础</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/yaml_settings.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">25</span>&nbsp; <span class="chapter-title">YAML设置</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/images_settings.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">26</span>&nbsp; <span class="chapter-title">图片设置</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/cross_references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">27</span>&nbsp; <span class="chapter-title">交叉引用</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/insert_other_content.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">28</span>&nbsp; <span class="chapter-title">插入其他内容</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/quarto_books.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">29</span>&nbsp; <span class="chapter-title">Quarto Books</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/github_pages.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">30</span>&nbsp; <span class="chapter-title">发布到GitHub Pages</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="false">
 <span class="menu-text">参考文献</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active" data-toc-expanded="1"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li>
<a href="#%E6%A0%87%E5%87%86%E5%8C%96normalization%E7%9A%84%E5%8E%9F%E7%90%86%E5%92%8C%E7%B1%BB%E5%9E%8B" id="toc-标准化normalization的原理和类型" class="nav-link active" data-scroll-target="#%E6%A0%87%E5%87%86%E5%8C%96normalization%E7%9A%84%E5%8E%9F%E7%90%86%E5%92%8C%E7%B1%BB%E5%9E%8B"><span class="header-section-number">21.1</span> 标准化（Normalization）的原理和类型</a>
  <ul class="collapse">
<li><a href="#methods-for-scrna-seq-normalization" id="toc-methods-for-scrna-seq-normalization" class="nav-link" data-scroll-target="#methods-for-scrna-seq-normalization">Methods for scRNA-seq normalization</a></li>
  </ul>
</li>
  <li>
<a href="#%E5%9F%BA%E4%BA%8E%E4%BC%A0%E7%BB%9F%E6%A0%87%E5%87%86%E5%8C%96%E6%B5%81%E7%A8%8B%E5%AF%B9%E9%9D%9E%E6%9C%9F%E6%9C%9B%E5%8F%98%E5%BC%82%E6%9D%A5%E6%BA%90%E8%BF%9B%E8%A1%8C%E8%AF%84%E4%BC%B0" id="toc-基于传统标准化流程对非期望变异来源进行评估" class="nav-link" data-scroll-target="#%E5%9F%BA%E4%BA%8E%E4%BC%A0%E7%BB%9F%E6%A0%87%E5%87%86%E5%8C%96%E6%B5%81%E7%A8%8B%E5%AF%B9%E9%9D%9E%E6%9C%9F%E6%9C%9B%E5%8F%98%E5%BC%82%E6%9D%A5%E6%BA%90%E8%BF%9B%E8%A1%8C%E8%AF%84%E4%BC%B0"><span class="header-section-number">21.2</span> 基于传统标准化流程对非期望变异来源进行评估</a>
  <ul class="collapse">
<li><a href="#set-up" id="toc-set-up" class="nav-link" data-scroll-target="#set-up">Set-up</a></li>
  <li><a href="#sec-evaluating_effects_of_cell_cycle" id="toc-sec-evaluating_effects_of_cell_cycle" class="nav-link" data-scroll-target="#sec-evaluating_effects_of_cell_cycle">评估细胞周期的影响（Evaluating effects of cell cycle）</a></li>
  <li><a href="#%E8%AF%84%E4%BC%B0%E7%BA%BF%E7%B2%92%E4%BD%93%E5%9F%BA%E5%9B%A0%E8%A1%A8%E8%BE%BE%E7%9A%84%E5%BD%B1%E5%93%8D" id="toc-评估线粒体基因表达的影响" class="nav-link" data-scroll-target="#%E8%AF%84%E4%BC%B0%E7%BA%BF%E7%B2%92%E4%BD%93%E5%9F%BA%E5%9B%A0%E8%A1%A8%E8%BE%BE%E7%9A%84%E5%BD%B1%E5%93%8D">评估线粒体基因表达的影响</a></li>
  </ul>
</li>
  <li>
<a href="#sctransform-based-normalization-and-regressing-out-sources-of-unwanted-variation" id="toc-sctransform-based-normalization-and-regressing-out-sources-of-unwanted-variation" class="nav-link" data-scroll-target="#sctransform-based-normalization-and-regressing-out-sources-of-unwanted-variation"><span class="header-section-number">21.3</span> <code>SCTransform</code>-based Normalization and regressing out sources of unwanted variation</a>
  <ul class="collapse">
<li><a href="#%E6%95%B0%E6%8D%AE%E5%AF%BC%E5%85%A5" id="toc-数据导入" class="nav-link" data-scroll-target="#%E6%95%B0%E6%8D%AE%E5%AF%BC%E5%85%A5">数据导入</a></li>
  <li><a href="#%E6%89%A7%E8%A1%8Csctransform" id="toc-执行sctransform" class="nav-link" data-scroll-target="#%E6%89%A7%E8%A1%8Csctransform">执行<code>SCTransform</code></a></li>
  <li><a href="#%E8%AF%84%E4%BC%B0%E7%BB%86%E8%83%9E%E5%91%A8%E6%9C%9F%E7%9A%84%E5%BD%B1%E5%93%8D" id="toc-评估细胞周期的影响" class="nav-link" data-scroll-target="#%E8%AF%84%E4%BC%B0%E7%BB%86%E8%83%9E%E5%91%A8%E6%9C%9F%E7%9A%84%E5%BD%B1%E5%93%8D">评估细胞周期的影响</a></li>
  <li><a href="#%E8%AF%84%E4%BC%B0%E7%BA%BF%E7%B2%92%E4%BD%93%E5%9F%BA%E5%9B%A0%E7%9A%84%E5%BD%B1%E5%93%8D" id="toc-评估线粒体基因的影响" class="nav-link" data-scroll-target="#%E8%AF%84%E4%BC%B0%E7%BA%BF%E7%B2%92%E4%BD%93%E5%9F%BA%E5%9B%A0%E7%9A%84%E5%BD%B1%E5%93%8D">评估线粒体基因的影响</a></li>
  <li><a href="#%E5%88%86%E5%89%B2layer%E5%86%8D%E6%AC%A1%E6%89%A7%E8%A1%8Csctranform" id="toc-分割layer再次执行sctranform" class="nav-link" data-scroll-target="#%E5%88%86%E5%89%B2layer%E5%86%8D%E6%AC%A1%E6%89%A7%E8%A1%8Csctranform">分割layer，再次执行<code>SCTranform</code></a></li>
  <li><a href="#save-the-object" id="toc-save-the-object" class="nav-link" data-scroll-target="#save-the-object">Save the object!</a></li>
  </ul>
</li>
  </ul><div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/djhcod/r-notes/blob/main/single_cell/scRNA-seq_online/06_SC_SCT_normalization.qmd" class="toc-action">View source</a></p><p><a href="https://github.com/djhcod/r-notes/issues/new" class="toc-action">Report an issue</a></p><p><a href="https://github.com/djhcod/r-notes/edit/main/single_cell/scRNA-seq_online/06_SC_SCT_normalization.qmd" class="toc-action">Edit this page</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title"><span id="sec-normalization-and-regressing-out-unwanted-variation" class="quarto-section-identifier"><span class="chapter-number">21</span>&nbsp; <span class="chapter-title">Normalization and regressing out unwanted variation</span></span></h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<strong>Learning Objectives:</strong>
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>Discuss why normalizing counts is necessary for accurate comparison between cells</li>
<li>Describe different normalization approaches</li>
<li>Evaluate the effects from any unwanted sources of variation and correct for them</li>
</ul>
</div>
</div>
<p>Now that we have our high quality cells, we can explore our data and see if we are able to identify any <strong>sources of unwanted variation</strong>. Depending on what we observe, we will utilize that information when performing variance stabilization using <code>SCTransform</code> but also to regress out the effects of any covariates that have an effect on our data.</p>
<p><img src="images/sc_workflow_2022-01.jpg" class="img-fluid" width="545"></p>
<hr>
<p><strong>Goals:</strong></p>
<ul>
<li>To accurately <strong>normalize the gene expression values</strong> to account for differences in <strong>sequencing depth</strong> and overdispersed count values.</li>
<li>To <strong>identify the most variant genes</strong> likely to be indicative of the different cell types present.</li>
</ul>
<p><strong>Challenges:</strong></p>
<ul>
<li>
<strong>Checking and removing unwanted variation</strong> so that we do not have cells clustering by artifacts downstream</li>
</ul>
<p><strong>Recommendations:</strong></p>
<ul>
<li>Have a good idea of your expectations for the <strong>cell types to be present</strong> prior to performing the clustering. Know whether you expect cell types of low complexity or higher mitochondrial content AND whether the cells are differentiating</li>
<li>
<strong>Regress out</strong> number of <strong>UMIs</strong> (default using <code>SCTransform</code>), <strong>mitochondrial content</strong>, and <strong>cell cycle</strong>, if needed and appropriate for experiment, so not to drive clustering downstream</li>
</ul>
<hr>
<section id="标准化normalization的原理和类型" class="level2" data-number="21.1"><h2 data-number="21.1" class="anchored" data-anchor-id="标准化normalization的原理和类型">
<span class="header-section-number">21.1</span> 标准化（Normalization）的原理和类型</h2>
<p>An essential first step in the majority of mRNA expression analyses is normalization, whereby systematic variations are adjusted for to <strong>make expression counts comparable across genes and/or samples</strong>. The counts of mapped reads for each gene is proportional to the expression of RNA (“interesting”) in addition to many other factors (“uninteresting”). Normalization is the process of adjusting raw count values to account for the “uninteresting” factors.</p>
<p>The main factors often considered during normalization are:</p>
<ul>
<li>
<strong>Sequencing depth:</strong> Accounting for sequencing depth is necessary for comparison of gene expression between cells. In the example below, each gene appears to have doubled in expression in cell 2, however this is a consequence of cell 2 having twice the sequencing depth.</li>
</ul>
<p><img src="images/sequencing_depth.png" class="img-fluid" width="509"></p>
<blockquote class="blockquote">
<p>Each cell in scRNA-seq will have a differing number of reads associated with it. So to accurately compare expression between cells, it is necessary to normalize for sequencing depth.</p>
</blockquote>
<ul>
<li>
<strong>Gene length:</strong> Accounting for gene length is necessary for comparing expression between different genes within the same cell. The number of reads mapped to a longer gene can appear to have equal count/expression as a shorter gene that is more highly expressed.</li>
</ul>
<p><img src="images/length_of_gene.png" class="img-fluid" width="464"></p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>If using a <strong>3’ or 5’ droplet-based method (10X Genomics, CEL-seq2, Drop-seq, inDrops)</strong>, the length of the gene will <strong>not</strong> affect the analysis because only the 5’ or 3’ end of the transcript is sequenced. However, if using <strong>full-length sequencing (Smart-seq)</strong>, the transcript length should be accounted for.</p>
</div>
</div>
<section id="methods-for-scrna-seq-normalization" class="level3"><h3 class="anchored" data-anchor-id="methods-for-scrna-seq-normalization">Methods for scRNA-seq normalization</h3>
<p>Various methods have been developed specifically for scRNA-seq normalization. Some simpler methods resemble what we have seen with bulk RNA-seq; the application of <strong>global scale factors</strong> adjusting for a count-depth relationship that is assumed common across all genes. However, if those assumptions are not true then this basic normalization can lead to over-correction for lowly and moderately expressed genes and, in some cases, under-normalization of highly expressed genes <span class="citation" data-cites="bacher2017">(<a href="../../references.html#ref-bacher2017" role="doc-biblioref">Bacher et al. 2017</a>)</span>. More complex methods will apply correction on a per-gene basis<strong>.</strong> In this lesson we will explore both approaches.</p>
<p>Regardless of which method is used for normalization, it can be helpful to think of it as a <strong>two-step process</strong> (even though it is often described as a single step in most papers). The first is a scaling step and the second is a transformation.</p>
<p><strong>1. Scaling</strong></p>
<p>The first step in normalization is to <strong>multiply each UMI count by a <u>cell specific factor</u> to get <u>all cells to have the same UMI counts</u></strong>. Why would we want to do this? Different cells have different amounts of mRNA; this could be due to differences between cell types or variation within the same cell type depending on how well the chemistry worked in one drop versus another. In either case, we are not interested in comparing these <strong>absolute counts</strong> between cells. Instead we are interested in comparing <strong>concentrations</strong>, and scaling helps achieve this.</p>
<p><strong>2. Transformation</strong></p>
<p>The next step is a transformation, and it is at this step where we can distinguish the simpler versus complex methods as mentioned above.</p>
<p><strong>Simple transformations</strong> are those which apply the same function to each individual measurement. Common examples include a <strong>log transform</strong> (which is applied in the original Seurat workflow), or a square root transform (less commonly used).</p>
<p>In the <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1874-1">Hafemeister and Satija, 2019 paper</a> the authors explored the issues with simple transformations. Specifically they evaluated the standard log normalization approach and found that genes with different abundances are affected differently and that effective normalization (using the log transform) is only observed with low/medium abundance genes (Figure 1D, below). Additionally, substantial imbalances in variance were observed with the log-normalized data (Figure 1E, below). In particular, cells with low total UMI counts exhibited disproportionately higher variance for high-abundance genes, dampening the variance contribution from other gene abundances.&nbsp;</p>
<p><img src="images/SCT_Fig1.png" class="img-fluid"></p>
<blockquote class="blockquote">
<p>Image credit: Hafemeister C and Satija R. Normalization and variance stabilization of single-cell RNA-seq data using regularized negative binomial regression, Genom Biology 2019 (<a href="https://doi.org/10.1186/s13059-019-1874-1" class="uri">https://doi.org/10.1186/s13059-019-1874-1</a>)</p>
</blockquote>
<p>The conclusion is, <strong>we cannot treat all genes the same.</strong></p>
<p>The proposed solution was the use of <strong>Pearson residuals for transformation</strong>, as implemented in Seurat’s <code>SCTransform</code> function. With this approach:</p>
<ul>
<li><p>Measurements are multiplied by a <strong>gene-specific weight</strong></p></li>
<li><p>Each gene is weighted based on how much evidence there is that it is non-uniformly expressed across cells. More evidence == more of a weight</p></li>
<li><p>Genes that are expressed in only a small fraction of cells will be favored (useful for finding rare cell populations)</p></li>
<li><p>Not just a consideration of the expression level is, but also the distribution of expression</p></li>
</ul>
<p>In this workshop we will demonstrate the use of both transformations at different steps in the workflow.</p>
</section></section><section id="基于传统标准化流程对非期望变异来源进行评估" class="level2" data-number="21.2"><h2 data-number="21.2" class="anchored" data-anchor-id="基于传统标准化流程对非期望变异来源进行评估">
<span class="header-section-number">21.2</span> 基于传统标准化流程对非期望变异来源进行评估</h2>
<p>The most common biological data correction (or source of “uninteresting” variation) in single cell RNA-seq is the effects of the <strong>cell cycle</strong> on the transcriptome. We need to explore the data and see if we observe any effects in our data.</p>
<section id="set-up" class="level3"><h3 class="anchored" data-anchor-id="set-up">Set-up</h3>
<p>Before we make any comparisons across cells, we will apply a <strong>simple normalization.</strong> <strong>This is solely for the purpose of exploring the sources of variation in our data.</strong></p>
<p>The input for this analysis is a <code>seurat</code> object. We will use the one that we created in <a href="04_SC_quality_control.html"><span>Chapter&nbsp;18</span></a> called <code>filtered_seurat</code>.</p>
<div class="cell" data-hash="06_SC_SCT_normalization_cache/html/unnamed-chunk-1_2b3d41d03675b09e7a0556314825e8de">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span></span>
<span><span class="va">filtered_seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="st">"output/scRNA-seq_online/seurat_filtered.rds"</span><span class="op">)</span></span>
<span><span class="va">filtered_seurat</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
14065 features across 29629 samples within 1 assay 
Active assay: RNA (14065 features, 0 variable features)
 1 layer present: counts</code></pre>
</div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">filtered_seurat</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                                      orig.ident nCount_RNA nFeature_RNA
ctrl_AAACATACAATGCC-1 ctrl_raw_feature_bc_matrix       2344          874
ctrl_AAACATACATTTCC-1 ctrl_raw_feature_bc_matrix       3125          896
ctrl_AAACATACCAGAAA-1 ctrl_raw_feature_bc_matrix       2578          725
ctrl_AAACATACCAGCTA-1 ctrl_raw_feature_bc_matrix       3261          979
ctrl_AAACATACCATGCA-1 ctrl_raw_feature_bc_matrix        746          362
ctrl_AAACATACCTCGCT-1 ctrl_raw_feature_bc_matrix       3519          866
                      log10GenesPerUMI  mitoRatio                 cells sample
ctrl_AAACATACAATGCC-1        0.8728630 0.01962457 ctrl_AAACATACAATGCC-1   ctrl
ctrl_AAACATACATTTCC-1        0.8447596 0.01792000 ctrl_AAACATACATTTCC-1   ctrl
ctrl_AAACATACCAGAAA-1        0.8384933 0.01551590 ctrl_AAACATACCAGAAA-1   ctrl
ctrl_AAACATACCAGCTA-1        0.8512622 0.01379945 ctrl_AAACATACCAGCTA-1   ctrl
ctrl_AAACATACCATGCA-1        0.8906861 0.02144772 ctrl_AAACATACCATGCA-1   ctrl
ctrl_AAACATACCTCGCT-1        0.8283053 0.01392441 ctrl_AAACATACCTCGCT-1   ctrl</code></pre>
</div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Normalize the counts</span></span>
<span><span class="va">seurat_phase</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html">NormalizeData</a></span><span class="op">(</span><span class="va">filtered_seurat</span><span class="op">)</span></span>
<span><span class="va">seurat_phase</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
14065 features across 29629 samples within 1 assay 
Active assay: RNA (14065 features, 0 variable features)
 2 layers present: counts, data</code></pre>
</div>
</div>
<p>可以发现，运行<code>NormalizeDataNext</code>后的数据多出了新的layer：data, 里面即储存了标准化后的数据。Next, we take this normalized data and check to see if data correction methods are necessary.</p>
</section><section id="sec-evaluating_effects_of_cell_cycle" class="level3"><h3 class="anchored" data-anchor-id="sec-evaluating_effects_of_cell_cycle">评估细胞周期的影响（Evaluating effects of cell cycle）</h3>
<section id="计算细胞周期评分" class="level4"><h4 class="anchored" data-anchor-id="计算细胞周期评分">计算细胞周期评分</h4>
<p>To assign each cell a score based on its expression of G2/M and S phase markers, we can use the Seuart function <code><a href="https://satijalab.org/seurat/reference/CellCycleScoring.html">CellCycleScoring()</a></code>. This function calculates cell cycle phase scores based on canonical markers that required as input.</p>
<p>A list of cell cycle markers, from Tirosh et al, 2015, is loaded with Seurat. We can segregate this list into markers of G2/M phase and markers of S phase. However, if you are not working with human data we have <a href="https://hbctraining.github.io/scRNA-seq_online/lessons/cell_cycle_scoring.html">additional materials</a> detailing how to acquire cell cycle markers for other organisms of interest.</p>
<div class="cell" data-hash="06_SC_SCT_normalization_cache/html/unnamed-chunk-2_31b1dc53b562428dee2d82e6172d1f6e">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Load cell cycle markers</span></span>
<span><span class="va">s.genes</span> <span class="op">&lt;-</span> <span class="va">cc.genes</span><span class="op">$</span><span class="va">s.genes</span></span>
<span><span class="va">g2m.genes</span> <span class="op">&lt;-</span> <span class="va">cc.genes</span><span class="op">$</span><span class="va">g2m.genes</span></span>
<span></span>
<span><span class="co"># Score cells for cell cycle</span></span>
<span><span class="va">seurat_phase</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/CellCycleScoring.html">CellCycleScoring</a></span><span class="op">(</span><span class="va">seurat_phase</span>, </span>
<span>                                 g2m.features <span class="op">=</span> <span class="va">g2m.genes</span>, </span>
<span>                                 s.features <span class="op">=</span> <span class="va">s.genes</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 现在的meta.data中多出了细胞周期评分“S.Score”和“G2M.Score”，以及推断的细胞所处的周期“Phase”</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">seurat_phase</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                                      orig.ident nCount_RNA nFeature_RNA
ctrl_AAACATACAATGCC-1 ctrl_raw_feature_bc_matrix       2344          874
ctrl_AAACATACATTTCC-1 ctrl_raw_feature_bc_matrix       3125          896
ctrl_AAACATACCAGAAA-1 ctrl_raw_feature_bc_matrix       2578          725
ctrl_AAACATACCAGCTA-1 ctrl_raw_feature_bc_matrix       3261          979
ctrl_AAACATACCATGCA-1 ctrl_raw_feature_bc_matrix        746          362
ctrl_AAACATACCTCGCT-1 ctrl_raw_feature_bc_matrix       3519          866
                      log10GenesPerUMI  mitoRatio                 cells sample
ctrl_AAACATACAATGCC-1        0.8728630 0.01962457 ctrl_AAACATACAATGCC-1   ctrl
ctrl_AAACATACATTTCC-1        0.8447596 0.01792000 ctrl_AAACATACATTTCC-1   ctrl
ctrl_AAACATACCAGAAA-1        0.8384933 0.01551590 ctrl_AAACATACCAGAAA-1   ctrl
ctrl_AAACATACCAGCTA-1        0.8512622 0.01379945 ctrl_AAACATACCAGCTA-1   ctrl
ctrl_AAACATACCATGCA-1        0.8906861 0.02144772 ctrl_AAACATACCATGCA-1   ctrl
ctrl_AAACATACCTCGCT-1        0.8283053 0.01392441 ctrl_AAACATACCTCGCT-1   ctrl
                          S.Score   G2M.Score Phase
ctrl_AAACATACAATGCC-1  0.02713602  0.04344302   G2M
ctrl_AAACATACATTTCC-1  0.01519129  0.01846409   G2M
ctrl_AAACATACCAGAAA-1 -0.05272781 -0.05038367    G1
ctrl_AAACATACCAGCTA-1 -0.05194312  0.04583528   G2M
ctrl_AAACATACCATGCA-1  0.04406978 -0.03445262     S
ctrl_AAACATACCTCGCT-1  0.03421052  0.02033139     S</code></pre>
</div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 查看一下细胞周期的分布情况</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">seurat_phase</span><span class="op">$</span><span class="va">Phase</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
   G1   G2M     S 
10387  9547  9695 </code></pre>
</div>
</div>
</section><section id="使用pca确定细胞周期是否是我们数据集中的主要变异来源" class="level4"><h4 class="anchored" data-anchor-id="使用pca确定细胞周期是否是我们数据集中的主要变异来源">使用PCA确定细胞周期是否是我们数据集中的主要变异来源</h4>
<p>After scoring the cells for cell cycle, we would like to determine <strong>whether cell cycle is a major source of variation in our dataset using PCA（PCA的原理详见：</strong> <a href="05_theory_of_PCA.html"><span>Chapter&nbsp;20</span></a> <strong>）</strong>.</p>
<p>To perform PCA, we need to <strong>first choose the most variable features, then scale the data</strong>. Since highly expressed genes exhibit the highest amount of variation and we don’t want our ‘highly variable genes’ only to reflect high expression, we need to scale the data to scale variation with expression level. The Seurat <code><a href="https://satijalab.org/seurat/reference/ScaleData.html">ScaleData()</a></code> function will scale the data by:</p>
<ul>
<li>adjusting the expression of each gene <strong>to give a mean expression across cells to be 0</strong>
</li>
<li>scaling expression of each gene <strong>to give a variance across cells to be 1</strong>
</li>
</ul>
<div class="cell" data-hash="06_SC_SCT_normalization_cache/html/unnamed-chunk-3_c3a9547eb839f0adade5b81c6b41d47f">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Identify the most variable genes</span></span>
<span><span class="va">seurat_phase</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html">FindVariableFeatures</a></span><span class="op">(</span><span class="va">seurat_phase</span>, </span>
<span>                                     selection.method <span class="op">=</span> <span class="st">"vst"</span>, <span class="co"># 默认值</span></span>
<span>                                     nfeatures <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span> <span class="co"># 默认值</span></span>
<span>             </span>
<span><span class="co"># Scale the counts</span></span>
<span><span class="va">seurat_phase</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html">ScaleData</a></span><span class="op">(</span><span class="va">seurat_phase</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Layers.html">Layers</a></span><span class="op">(</span><span class="va">seurat_phase</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "counts"     "data"       "scale.data"</code></pre>
</div>
</div>
<p>可以发现，运行<code>ScaleData</code>后的数据多出了新的layer：scale.data, 里面即储存了归一化后的数据。Now, we can perform the PCA analysis and plot the first two principal components against each other. We also split the figure by cell cycle phase, to evaluate similarities and/or differences.</p>
<div class="cell" data-hash="06_SC_SCT_normalization_cache/html/unnamed-chunk-4_e22a3cfa42da4347bdd1b12c773fdf18">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Perform PCA。如果没有指定features，RunPCA默认使用FindVariableFeatures找到的高变基因作为PCA输入.</span></span>
<span><span class="va">seurat_phase</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html">RunPCA</a></span><span class="op">(</span><span class="va">seurat_phase</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot the PCA colored by cell cycle phase</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">seurat_phase</span>,</span>
<span>              reduction <span class="op">=</span> <span class="st">"pca"</span>,</span>
<span>              group.by<span class="op">=</span> <span class="st">"Phase"</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">seurat_phase</span>,</span>
<span>              reduction <span class="op">=</span> <span class="st">"pca"</span>,</span>
<span>              group.by<span class="op">=</span> <span class="st">"Phase"</span>,</span>
<span>              split.by <span class="op">=</span> <span class="st">"Phase"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/">cowplot</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">plot_grid</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span>, ncol <span class="op">=</span> <span class="fl">2</span>, labels <span class="op">=</span> <span class="st">"AUTO"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="06_SC_SCT_normalization_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid" width="960"></p>
</div>
</div>
<p><strong>We do not see large differences due to cell cycle phase. Based on this plot, we would not regress out the variation due to cell cycle.</strong></p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-3-contents" aria-controls="callout-3" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
When should cell cycle phase be regressed out?
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-3" class="callout-3-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Below are two PCA plots taken from <a href="../seurat/cell_cycle_regression.html"><span>Chapter&nbsp;13</span></a> 。This first plot is similar to what we plotted above, it is a PCA prior to regression to evaluate if the cell cycle is playing a big role in driving PC1 and PC2. Clearly, the cells are separating by cell type in this case, so it suggests regressing out these effects.</p>
<p><img src="images/unnamed-chunk-6-1.png" class="img-fluid" width="427"></p>
<p>This second PCA plot is post-regression, and displays how effective the regression was in removing the effect we observed.</p>
<p><img src="images/unnamed-chunk-10-1.png" class="img-fluid" width="430"></p>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>在需要消除细胞周期的影响时，如何通过<code>ScaleData</code>回归掉（regress out）细胞周期的影响，以及如何在消除细胞周期的影响同时保留增殖细胞与静止细胞的区分，参考 <a href="../seurat/cell_cycle_regression.html"><span>Chapter&nbsp;13</span></a> 。</p>
</div>
</div>
<hr></section></section><section id="评估线粒体基因表达的影响" class="level3"><h3 class="anchored" data-anchor-id="评估线粒体基因表达的影响">评估线粒体基因表达的影响</h3>
<p>Mitochondrial expression is another factor which can greatly influence clustering. Oftentimes, it is useful to regress out variation due to mitochondrial expression. However, <strong>if the differences in mitochondrial gene expression represent a biological phenomenon that may help to distinguish cell clusters, then we advise not regressing this out</strong>. In this exercise, we can perform a quick check similar to looking at cell cycle and decide whether or not we want to regress it out.</p>
<ol type="1">
<li>
<p>First, turn the mitochondrial ratio variable into a new categorical variable based on quartiles. 根据各细胞线粒体基因的比例的四分位数将所有细胞分为低线粒体基因比例细胞、中线粒体基因比例细胞···</p>
<div class="cell" data-hash="06_SC_SCT_normalization_cache/html/unnamed-chunk-5_db73b946a4ab25b9a1401917f1d1c611">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Check quartile values</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">seurat_phase</span><span class="op">$</span><span class="va">mitoRatio</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
0.00000 0.01438 0.01993 0.02139 0.02669 0.14464 </code></pre>
</div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Turn mitoRatio into categorical factor vector based on quartile values</span></span>
<span><span class="va">seurat_phase</span><span class="op">$</span><span class="va">mitoFr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cut.html">cut</a></span><span class="op">(</span><span class="va">seurat_phase</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">mitoRatio</span>, </span>
<span>                           breaks<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="cn">Inf</span>, <span class="fl">0.0144</span>, <span class="fl">0.0199</span>, <span class="fl">0.0267</span>, <span class="cn">Inf</span><span class="op">)</span>, </span>
<span>                           labels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Low"</span>, <span class="st">"Medium"</span>, <span class="st">"Medium high"</span>, <span class="st">"High"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">seurat_phase</span><span class="op">$</span><span class="va">mitoFr</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
        Low      Medium Medium high        High 
       7443        7325        7459        7402 </code></pre>
</div>
</div>
</li>
<li>
<p>Next, plot the PCA similar to how we did with cell cycle regression. <em>Hint: use the new <code>mitoFr</code> variable to split cells and color them accordingly.</em></p>
<div class="cell" data-hash="06_SC_SCT_normalization_cache/html/unnamed-chunk-6_16eda56e8930a5a802d5fd81c034deb8">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 根据各细胞线粒体基因的比例信息绘制PCA</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">seurat_phase</span>,</span>
<span>              reduction <span class="op">=</span> <span class="st">"pca"</span>,</span>
<span>              group.by<span class="op">=</span> <span class="st">"mitoFr"</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">seurat_phase</span>,</span>
<span>              reduction <span class="op">=</span> <span class="st">"pca"</span>,</span>
<span>              group.by<span class="op">=</span> <span class="st">"mitoFr"</span>,</span>
<span>              split.by <span class="op">=</span> <span class="st">"mitoFr"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">plot_grid</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span>, ncol <span class="op">=</span> <span class="fl">2</span>, labels <span class="op">=</span> <span class="st">"AUTO"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="06_SC_SCT_normalization_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid" width="1920"></p>
</div>
</div>
</li>
<li>
<p>Evaluate the PCA plot.</p>
<p>We do not see large differences due to mitochondrial expression. Based on this plot, we would not regress out the variation due to mitochondrial expression.</p>
</li>
</ol>
<hr></section></section><section id="sctransform-based-normalization-and-regressing-out-sources-of-unwanted-variation" class="level2" data-number="21.3"><h2 data-number="21.3" class="anchored" data-anchor-id="sctransform-based-normalization-and-regressing-out-sources-of-unwanted-variation">
<span class="header-section-number">21.3</span> <code>SCTransform</code>-based Normalization and regressing out sources of unwanted variation</h2>
<p>Now that we have established which effects are observed in our data, we can use the <code>SCTransform</code> method to regress out these effects. The <code>SCTransform</code> method was proposed as a better alternative to the log transform normalization method that we used for exploring sources of unwanted variation. The method not only <strong>normalizes data, but it also performs a variance stabilization and allows for additional covariates to be regressed out</strong>.</p>
<p>As described earlier, all genes cannot be treated the same. As such, the <code>SCTransform</code> <strong>method constructs a generalized linear model (GLM) for each gene</strong> with UMI counts as the response and sequencing depth as the explanatory variable. Information is pooled across genes with similar abundances, to regularize parameter estimates and <strong>obtain residuals which represent effectively normalized data values</strong> which are no longer correlated with sequencing depth.</p>
<p><img src="images/SCT_UMAP.png" class="img-fluid" width="532"></p>
<blockquote class="blockquote">
<p>Image credit: Hafemeister C and Satija R. Normalization and variance stabilization of single-cell RNA-seq data using regularized negative binomial regression, Genom Biology 2019 (<a href="https://doi.org/10.1101/576827)_" class="uri">https://doi.org/10.1101/576827)_</a></p>
</blockquote>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Since the UMI counts are part of the GLM, the effects are automatically regressed out. The user can include any additional covariates (<code>vars.to.regress</code>) that may have an effect on expression and will be included in the model.</p>
</div>
</div>
<section id="数据导入" class="level3"><h3 class="anchored" data-anchor-id="数据导入">数据导入</h3>
<p>前面的流程和此前一样</p>
<div class="cell" data-hash="06_SC_SCT_normalization_cache/html/unnamed-chunk-7_398ce14c850444e158eec96121e2bae9">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/rm.html">rm</a></span><span class="op">(</span>list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ls.html">ls</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span></span>
<span><span class="va">filtered_seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="st">"output/seurat_filtered.rds"</span><span class="op">)</span></span>
<span><span class="va">filtered_seurat</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
14065 features across 29629 samples within 1 assay 
Active assay: RNA (14065 features, 0 variable features)
 1 layer present: counts</code></pre>
</div>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">filtered_seurat</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                                      orig.ident nCount_RNA nFeature_RNA
ctrl_AAACATACAATGCC-1 ctrl_raw_feature_bc_matrix       2344          874
ctrl_AAACATACATTTCC-1 ctrl_raw_feature_bc_matrix       3125          896
ctrl_AAACATACCAGAAA-1 ctrl_raw_feature_bc_matrix       2578          725
ctrl_AAACATACCAGCTA-1 ctrl_raw_feature_bc_matrix       3261          979
ctrl_AAACATACCATGCA-1 ctrl_raw_feature_bc_matrix        746          362
ctrl_AAACATACCTCGCT-1 ctrl_raw_feature_bc_matrix       3519          866
                      log10GenesPerUMI  mitoRatio                 cells sample
ctrl_AAACATACAATGCC-1        0.8728630 0.01962457 ctrl_AAACATACAATGCC-1   ctrl
ctrl_AAACATACATTTCC-1        0.8447596 0.01792000 ctrl_AAACATACATTTCC-1   ctrl
ctrl_AAACATACCAGAAA-1        0.8384933 0.01551590 ctrl_AAACATACCAGAAA-1   ctrl
ctrl_AAACATACCAGCTA-1        0.8512622 0.01379945 ctrl_AAACATACCAGCTA-1   ctrl
ctrl_AAACATACCATGCA-1        0.8906861 0.02144772 ctrl_AAACATACCATGCA-1   ctrl
ctrl_AAACATACCTCGCT-1        0.8283053 0.01392441 ctrl_AAACATACCTCGCT-1   ctrl</code></pre>
</div>
</div>
</section><section id="执行sctransform" class="level3"><h3 class="anchored" data-anchor-id="执行sctransform">执行<code>SCTransform</code>
</h3>
<ul>
<li><p>这里先运行一次<code>SCTransform</code>以便后面评估细胞周期、线粒体基因等非期望变异来源（<strong>This is solely for the purpose of exploring the sources of variation in our data</strong>）</p></li>
<li><p><code>SCTransform</code>替代了传统单细胞数据分析流程中的<code><a href="https://satijalab.org/seurat/reference/NormalizeData.html">NormalizeData()</a></code>、<code><a href="https://satijalab.org/seurat/reference/ScaleData.html">ScaleData()</a></code>和<code><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html">FindVariableFeatures()</a></code>函数的功能，因此不再需要运行这些函数。</p></li>
<li><p>In Seurat v5, SCT v2 is applied by default. You can revert to v1 by setting&nbsp;<code>vst.flavor = 'v1'</code>。</p></li>
<li><p><code>SCTransform</code>的运算调用了<a href="https://bioconductor.org/packages/release/bioc/html/glmGamPoi.html"><code>glmGamPoi</code></a>包以显著提升运算速度。所以事先需要通过<code>BiocManager</code>安装该包。</p></li>
</ul>
<div class="cell" data-hash="06_SC_SCT_normalization_cache/html/unnamed-chunk-8_5accd98d48adbb19f68c128a45ac2756">
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># SCTranform</span></span>
<span><span class="co"># BiocManager::install("glmGamPoi")</span></span>
<span><span class="va">seurat_phase</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SCTransform.html">SCTransform</a></span><span class="op">(</span><span class="va">filtered_seurat</span><span class="op">)</span></span>
<span><span class="va">seurat_phase</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
28130 features across 29629 samples within 2 assays 
Active assay: SCT (14065 features, 3000 variable features)
 3 layers present: counts, data, scale.data
 1 other assay present: RNA</code></pre>
</div>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Check which assays are stored in objects</span></span>
<span><span class="va">seurat_phase</span><span class="op">@</span><span class="va">assays</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$RNA
Assay (v5) data with 14065 features for 29629 cells
First 10 features:
 AL627309.1, AL669831.5, LINC00115, FAM41C, NOC2L, KLHL17, PLEKHN1,
HES4, ISG15, AGRN 
Layers:
 counts 

$SCT
SCTAssay data with 14065 features for 29629 cells, and 1 SCTModel(s) 
Top 10 variable features:
 CCL8, IGKC, CXCL10, FTL, CCL2, CCL7, ISG15, GNLY, IGLC2, CCL4 </code></pre>
</div>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 查看目前默认的assay</span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/DefaultAssay.html">DefaultAssay</a></span><span class="op">(</span><span class="va">seurat_phase</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "SCT"</code></pre>
</div>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 查看默认assay的layers</span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Layers.html">Layers</a></span><span class="op">(</span><span class="va">seurat_phase</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "counts"     "data"       "scale.data"</code></pre>
</div>
</div>
<p>Note, the last line of output specifies <strong>“Set default assay to SCT”</strong>. 表明运行<code>SCTransform</code>之后，会将默认的assay指定为<code>SCTransform</code>之后的数据。This specifies that moving forward we would like to use the data after SCT was implemented. We can view the different assays that we have stored in our seurat object.</p>
</section><section id="评估细胞周期的影响" class="level3"><h3 class="anchored" data-anchor-id="评估细胞周期的影响">评估细胞周期的影响</h3>
<div class="cell" data-hash="06_SC_SCT_normalization_cache/html/unnamed-chunk-9_e05e1e11b267df72cf64222e0d2c06ad">
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Load cell cycle markers</span></span>
<span><span class="va">s.genes</span> <span class="op">&lt;-</span> <span class="va">cc.genes</span><span class="op">$</span><span class="va">s.genes</span></span>
<span><span class="va">g2m.genes</span> <span class="op">&lt;-</span> <span class="va">cc.genes</span><span class="op">$</span><span class="va">g2m.genes</span></span>
<span></span>
<span><span class="co"># Score cells for cell cycle</span></span>
<span><span class="va">seurat_phase</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/CellCycleScoring.html">CellCycleScoring</a></span><span class="op">(</span><span class="va">seurat_phase</span>, </span>
<span>                                 g2m.features <span class="op">=</span> <span class="va">g2m.genes</span>, </span>
<span>                                 s.features <span class="op">=</span> <span class="va">s.genes</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 现在的meta.data中多出了细胞周期评分“S.Score”和“G2M.Score”，以及推断的细胞所处的周期“Phase”</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">seurat_phase</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                                      orig.ident nCount_RNA nFeature_RNA
ctrl_AAACATACAATGCC-1 ctrl_raw_feature_bc_matrix       2344          874
ctrl_AAACATACATTTCC-1 ctrl_raw_feature_bc_matrix       3125          896
ctrl_AAACATACCAGAAA-1 ctrl_raw_feature_bc_matrix       2578          725
ctrl_AAACATACCAGCTA-1 ctrl_raw_feature_bc_matrix       3261          979
ctrl_AAACATACCATGCA-1 ctrl_raw_feature_bc_matrix        746          362
ctrl_AAACATACCTCGCT-1 ctrl_raw_feature_bc_matrix       3519          866
                      log10GenesPerUMI  mitoRatio                 cells sample
ctrl_AAACATACAATGCC-1        0.8728630 0.01962457 ctrl_AAACATACAATGCC-1   ctrl
ctrl_AAACATACATTTCC-1        0.8447596 0.01792000 ctrl_AAACATACATTTCC-1   ctrl
ctrl_AAACATACCAGAAA-1        0.8384933 0.01551590 ctrl_AAACATACCAGAAA-1   ctrl
ctrl_AAACATACCAGCTA-1        0.8512622 0.01379945 ctrl_AAACATACCAGCTA-1   ctrl
ctrl_AAACATACCATGCA-1        0.8906861 0.02144772 ctrl_AAACATACCATGCA-1   ctrl
ctrl_AAACATACCTCGCT-1        0.8283053 0.01392441 ctrl_AAACATACCTCGCT-1   ctrl
                      nCount_SCT nFeature_SCT      S.Score    G2M.Score Phase
ctrl_AAACATACAATGCC-1       1591          863  0.010526369  0.011803814   G2M
ctrl_AAACATACATTTCC-1       1553          724  0.010251663  0.015119823   G2M
ctrl_AAACATACCAGAAA-1       1549          668 -0.019803499 -0.015779795    G1
ctrl_AAACATACCAGCTA-1       1579          777 -0.032093208  0.013380044   G2M
ctrl_AAACATACCATGCA-1       1096          371  0.008301833 -0.008402066     S
ctrl_AAACATACCTCGCT-1       1493          632  0.018235066  0.018993438   G2M</code></pre>
</div>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 查看一下细胞周期的分布情况</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">seurat_phase</span><span class="op">$</span><span class="va">Phase</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
   G1   G2M     S 
10554  9586  9489 </code></pre>
</div>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 执行PCA</span></span>
<span><span class="va">seurat_phase</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html">RunPCA</a></span><span class="op">(</span><span class="va">seurat_phase</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot the PCA colored by cell cycle phase</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">seurat_phase</span>,</span>
<span>              reduction <span class="op">=</span> <span class="st">"pca"</span>,</span>
<span>              group.by<span class="op">=</span> <span class="st">"Phase"</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">seurat_phase</span>,</span>
<span>              reduction <span class="op">=</span> <span class="st">"pca"</span>,</span>
<span>              group.by<span class="op">=</span> <span class="st">"Phase"</span>,</span>
<span>              split.by <span class="op">=</span> <span class="st">"Phase"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/">cowplot</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">plot_grid</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span>, ncol <span class="op">=</span> <span class="fl">2</span>, labels <span class="op">=</span> <span class="st">"AUTO"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="06_SC_SCT_normalization_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid" width="960"></p>
</div>
</div>
</section><section id="评估线粒体基因的影响" class="level3"><h3 class="anchored" data-anchor-id="评估线粒体基因的影响">评估线粒体基因的影响</h3>
<div class="cell" data-hash="06_SC_SCT_normalization_cache/html/unnamed-chunk-10_0178b134a976c3c798eec2543778add7">
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Check quartile values</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">seurat_phase</span><span class="op">$</span><span class="va">mitoRatio</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
0.00000 0.01438 0.01993 0.02139 0.02669 0.14464 </code></pre>
</div>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Turn mitoRatio into categorical factor vector based on quartile values</span></span>
<span><span class="va">seurat_phase</span><span class="op">$</span><span class="va">mitoFr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cut.html">cut</a></span><span class="op">(</span><span class="va">seurat_phase</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">mitoRatio</span>, </span>
<span>                           breaks<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="cn">Inf</span>, <span class="fl">0.0144</span>, <span class="fl">0.0199</span>, <span class="fl">0.0267</span>, <span class="cn">Inf</span><span class="op">)</span>, </span>
<span>                           labels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Low"</span>, <span class="st">"Medium"</span>, <span class="st">"Medium high"</span>, <span class="st">"High"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">seurat_phase</span><span class="op">$</span><span class="va">mitoFr</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
        Low      Medium Medium high        High 
       7443        7325        7459        7402 </code></pre>
</div>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Plot the PCA colored by cell cycle phase</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">seurat_phase</span>,</span>
<span>              reduction <span class="op">=</span> <span class="st">"pca"</span>,</span>
<span>              group.by<span class="op">=</span> <span class="st">"mitoFr"</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">seurat_phase</span>,</span>
<span>              reduction <span class="op">=</span> <span class="st">"pca"</span>,</span>
<span>              group.by<span class="op">=</span> <span class="st">"mitoFr"</span>,</span>
<span>              split.by <span class="op">=</span> <span class="st">"mitoFr"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">plot_grid</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span>, ncol <span class="op">=</span> <span class="fl">2</span>, labels <span class="op">=</span> <span class="st">"AUTO"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="06_SC_SCT_normalization_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid" width="1920"></p>
</div>
</div>
</section><section id="分割layer再次执行sctranform" class="level3"><h3 class="anchored" data-anchor-id="分割layer再次执行sctranform">分割layer，再次执行<code>SCTranform</code>
</h3>
<p>Since we have two samples in our dataset (from two conditions), we want to keep them as separate layers and transform them as that is <strong>what is required for integration</strong>.</p>
<div class="cell" data-hash="06_SC_SCT_normalization_cache/html/unnamed-chunk-11_5aa581d1cbc0e05ff5055fc7d675d5ac">
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Split RNA assay by condition to perform cell cycle scoring and SCT on all samples</span></span>
<span><span class="va">seurat_phase</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html">split</a></span><span class="op">(</span><span class="va">seurat_phase</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span>, </span>
<span>                               f <span class="op">=</span> <span class="va">seurat_phase</span><span class="op">$</span><span class="va">sample</span><span class="op">)</span> <span class="co"># 按照meta.data中的“sample”列进行分割</span></span>
<span><span class="va">seurat_phase</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
28130 features across 29629 samples within 2 assays 
Active assay: SCT (14065 features, 3000 variable features)
 3 layers present: counts, data, scale.data
 1 other assay present: RNA
 1 dimensional reduction calculated: pca</code></pre>
</div>
</div>
<p>现在可以发现RNA assay的<code>counts</code>和<code>data</code>按照”seurat_phase$sample”（ctrl vs.&nbsp;stim）被分别分割成了2个layer：</p>
<p><img src="images/截屏2023-12-19 10.46.18.png" class="img-fluid"></p>
<p>Now we will run the <code><a href="https://satijalab.org/seurat/reference/SCTransform.html">SCTransform()</a></code> on each sample, and regress out mitochondrial expression by specifying in the <code>vars.to.regress</code> argument of the <code><a href="https://satijalab.org/seurat/reference/SCTransform.html">SCTransform()</a></code> function.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>The output of <code><a href="https://satijalab.org/seurat/reference/SCTransform.html">SCTransform()</a></code> can generate large R objects/variables in terms of memory. If we have a large dataset, then we might need to <strong>adjust the limit for allowable object sizes within R</strong> (<em>Default is 500</em> 1024 ^ 2 = 500 Mb*) using the following code:</p>
<div class="cell" data-hash="06_SC_SCT_normalization_cache/html/unnamed-chunk-12_7bd61303ffdbfc0b44e7f68f2b748ac1">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">future.globals.maxSize =</span> <span class="dv">4000</span> <span class="sc">*</span> <span class="dv">1024</span><span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<section id="sec-perform_sctranform" class="level4"><h4 class="anchored" data-anchor-id="sec-perform_sctranform">执行<code>SCTranform</code>
</h4>
<div class="cell" data-hash="06_SC_SCT_normalization_cache/html/unnamed-chunk-13_7feac7ffc25b40f0737ac041c266e227">
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># SCTranform</span></span>
<span><span class="va">seurat_phase</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SCTransform.html">SCTransform</a></span><span class="op">(</span><span class="va">seurat_phase</span>, vars.to.regress <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"mitoRatio"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>By default, after normalizing, adjusting the variance, and regressing out uninteresting sources of variation, <code>SCTransform</code> will rank the genes by residual variance and output the <strong>3000</strong> most variant genes. If the dataset has larger cell numbers, then it may be beneficial to adjust this parameter higher using the <code>variable.features.n</code> argument.</p>
<p>Now we can see that in addition to the raw RNA counts, we now have a SCT component in our <code>assays</code> slot. The most variable features will be the only genes stored inside the SCT assay. As we move through the scRNA-seq analysis, we will choose the most appropriate assay to use for the different steps in the analysis.</p>
<hr></section></section><section id="save-the-object" class="level3"><h3 class="anchored" data-anchor-id="save-the-object">Save the object!</h3>
<div class="cell" data-hash="06_SC_SCT_normalization_cache/html/unnamed-chunk-14_6a927843fc3b14ceb78a65dea011d685">
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">seurat_phase</span>, <span class="st">"output/scRNA-seq_online/split_seurat.rds"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-7-contents" aria-controls="callout-7" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Session Info
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-7" class="callout-7-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-hash="06_SC_SCT_normalization_cache/html/unnamed-chunk-15_7c92f6dad57b5352def4548afbf3d82f">
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.3.2 (2023-10-31)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Sonoma 14.2

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: Asia/Shanghai
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] cowplot_1.1.1      Seurat_5.0.1       SeuratObject_5.0.1 sp_2.1-2          

loaded via a namespace (and not attached):
  [1] RColorBrewer_1.1-3          rstudioapi_0.15.0          
  [3] jsonlite_1.8.7              magrittr_2.0.3             
  [5] spatstat.utils_3.0-4        farver_2.1.1               
  [7] rmarkdown_2.25              zlibbioc_1.48.0            
  [9] vctrs_0.6.5                 ROCR_1.0-11                
 [11] DelayedMatrixStats_1.24.0   spatstat.explore_3.2-5     
 [13] RCurl_1.98-1.13             S4Arrays_1.2.0             
 [15] htmltools_0.5.7             SparseArray_1.2.2          
 [17] sctransform_0.4.1           parallelly_1.36.0          
 [19] KernSmooth_2.23-22          htmlwidgets_1.6.3          
 [21] ica_1.0-3                   plyr_1.8.9                 
 [23] plotly_4.10.3               zoo_1.8-12                 
 [25] igraph_1.5.1                mime_0.12                  
 [27] lifecycle_1.0.4             pkgconfig_2.0.3            
 [29] Matrix_1.6-4                R6_2.5.1                   
 [31] fastmap_1.1.1               GenomeInfoDbData_1.2.11    
 [33] MatrixGenerics_1.14.0       fitdistrplus_1.1-11        
 [35] future_1.33.0               shiny_1.8.0                
 [37] digest_0.6.33               colorspace_2.1-0           
 [39] patchwork_1.1.3             S4Vectors_0.40.2           
 [41] tensor_1.5                  RSpectra_0.16-1            
 [43] irlba_2.3.5.1               GenomicRanges_1.54.1       
 [45] labeling_0.4.3              progressr_0.14.0           
 [47] fansi_1.0.5                 spatstat.sparse_3.0-3      
 [49] httr_1.4.7                  polyclip_1.10-6            
 [51] abind_1.4-5                 compiler_4.3.2             
 [53] withr_2.5.2                 fastDummies_1.7.3          
 [55] MASS_7.3-60                 DelayedArray_0.28.0        
 [57] tools_4.3.2                 lmtest_0.9-40              
 [59] httpuv_1.6.12               future.apply_1.11.0        
 [61] goftest_1.2-3               glmGamPoi_1.14.0           
 [63] glue_1.6.2                  nlme_3.1-164               
 [65] promises_1.2.1              grid_4.3.2                 
 [67] Rtsne_0.16                  cluster_2.1.6              
 [69] reshape2_1.4.4              generics_0.1.3             
 [71] gtable_0.3.4                spatstat.data_3.0-3        
 [73] tidyr_1.3.0                 data.table_1.14.8          
 [75] XVector_0.42.0              utf8_1.2.4                 
 [77] BiocGenerics_0.48.1         spatstat.geom_3.2-7        
 [79] RcppAnnoy_0.0.21            ggrepel_0.9.4              
 [81] RANN_2.6.1                  pillar_1.9.0               
 [83] stringr_1.5.1               spam_2.10-0                
 [85] RcppHNSW_0.5.0              later_1.3.1                
 [87] splines_4.3.2               dplyr_1.1.4                
 [89] lattice_0.22-5              survival_3.5-7             
 [91] deldir_2.0-2                tidyselect_1.2.0           
 [93] miniUI_0.1.1.1              pbapply_1.7-2              
 [95] knitr_1.45                  gridExtra_2.3              
 [97] IRanges_2.36.0              SummarizedExperiment_1.32.0
 [99] scattermore_1.2             stats4_4.3.2               
[101] xfun_0.41                   Biobase_2.62.0             
[103] matrixStats_1.1.0           stringi_1.8.2              
[105] lazyeval_0.2.2              yaml_2.3.7                 
[107] evaluate_0.23               codetools_0.2-19           
[109] tibble_3.2.1                cli_3.6.1                  
[111] uwot_0.1.16                 xtable_1.8-4               
[113] reticulate_1.34.0           munsell_0.5.0              
[115] Rcpp_1.0.11                 GenomeInfoDb_1.38.1        
[117] globals_0.16.2              spatstat.random_3.2-2      
[119] png_0.1-8                   parallel_4.3.2             
[121] ellipsis_0.3.2              ggplot2_3.4.4              
[123] dotCall64_1.1-1             sparseMatrixStats_1.14.0   
[125] bitops_1.0-7                listenv_0.9.0              
[127] viridisLite_0.4.2           scales_1.3.0               
[129] ggridges_0.5.4              crayon_1.5.2               
[131] leiden_0.4.3.1              purrr_1.0.2                
[133] rlang_1.1.2                </code></pre>
</div>
</div>
</div>
</div>
</div>


<!-- -->

<div id="refs" class="references csl-bib-body hanging-indent" role="list" style="display: none">
<div id="ref-bacher2017" class="csl-entry" role="listitem">
Bacher, Rhonda, Li-Fang Chu, Ning Leng, Audrey P Gasch, James A Thomson, Ron M Stewart, Michael Newton, and Christina Kendziorski. 2017. <span>“SCnorm: Robust Normalization of Single-Cell RNA-Seq Data.”</span> <em>Nature Methods</em> 14 (6): 584–86. <a href="https://doi.org/10.1038/nmeth.4263">https://doi.org/10.1038/nmeth.4263</a>.
</div>
</div>
</section></section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  let localAlternateSentinel = 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var filterRegex = new RegExp("https:\/\/djhcod\.github\.io\/r-notes\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
          // target, if specified
          link.setAttribute("target", "_blank");
      }
    }
});
</script><script src="https://utteranc.es/client.js" repo="https://github.com/djhcod/r-notes" issue-term="pathname" theme="github-light" crossorigin="anonymous" async="">
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="../../single_cell/scRNA-seq_online/05_theory_of_PCA.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">PCA原理</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../../single_cell/scRNA-seq_online/06_integration.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">22</span>&nbsp; <span class="chapter-title">单细胞数据整合（Integration）</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb47" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a><span class="an">bibliography:</span><span class="co"> references.bib</span></span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a><span class="fu"># Normalization and regressing out unwanted variation {#sec-normalization-and-regressing-out-unwanted-variation}</span></span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-7"><a href="#cb47-7" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb47-8"><a href="#cb47-8" aria-hidden="true" tabindex="-1"></a><span class="fu">###### **Learning Objectives:**</span></span>
<span id="cb47-9"><a href="#cb47-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-10"><a href="#cb47-10" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Discuss why normalizing counts is necessary for accurate comparison between cells</span>
<span id="cb47-11"><a href="#cb47-11" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Describe different normalization approaches</span>
<span id="cb47-12"><a href="#cb47-12" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Evaluate the effects from any unwanted sources of variation and correct for them</span>
<span id="cb47-13"><a href="#cb47-13" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb47-14"><a href="#cb47-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-15"><a href="#cb47-15" aria-hidden="true" tabindex="-1"></a>Now that we have our high quality cells, we can explore our data and see if we are able to identify any **sources of unwanted variation**. Depending on what we observe, we will utilize that information when performing variance stabilization using <span class="in">`SCTransform`</span> but also to regress out the effects of any covariates that have an effect on our data.</span>
<span id="cb47-16"><a href="#cb47-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-17"><a href="#cb47-17" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/sc_workflow_2022-01.jpg)</span>{width="545"}</span>
<span id="cb47-18"><a href="#cb47-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-19"><a href="#cb47-19" aria-hidden="true" tabindex="-1"></a>------------------------------------------------------------------------</span>
<span id="cb47-20"><a href="#cb47-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-21"><a href="#cb47-21" aria-hidden="true" tabindex="-1"></a>**Goals:**</span>
<span id="cb47-22"><a href="#cb47-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-23"><a href="#cb47-23" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>To accurately **normalize the gene expression values** to account for differences in **sequencing depth** and overdispersed count values.</span>
<span id="cb47-24"><a href="#cb47-24" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>To **identify the most variant genes** likely to be indicative of the different cell types present.</span>
<span id="cb47-25"><a href="#cb47-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-26"><a href="#cb47-26" aria-hidden="true" tabindex="-1"></a>**Challenges:**</span>
<span id="cb47-27"><a href="#cb47-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-28"><a href="#cb47-28" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>**Checking and removing unwanted variation** so that we do not have cells clustering by artifacts downstream</span>
<span id="cb47-29"><a href="#cb47-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-30"><a href="#cb47-30" aria-hidden="true" tabindex="-1"></a>**Recommendations:**</span>
<span id="cb47-31"><a href="#cb47-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-32"><a href="#cb47-32" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Have a good idea of your expectations for the **cell types to be present** prior to performing the clustering. Know whether you expect cell types of low complexity or higher mitochondrial content AND whether the cells are differentiating</span>
<span id="cb47-33"><a href="#cb47-33" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>**Regress out** number of **UMIs** (default using `SCTransform`), **mitochondrial content**, and **cell cycle**, if needed and appropriate for experiment, so not to drive clustering downstream</span>
<span id="cb47-34"><a href="#cb47-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-35"><a href="#cb47-35" aria-hidden="true" tabindex="-1"></a>------------------------------------------------------------------------</span>
<span id="cb47-36"><a href="#cb47-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-37"><a href="#cb47-37" aria-hidden="true" tabindex="-1"></a><span class="fu">## 标准化（Normalization）的原理和类型</span></span>
<span id="cb47-38"><a href="#cb47-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-39"><a href="#cb47-39" aria-hidden="true" tabindex="-1"></a>An essential first step in the majority of mRNA expression analyses is normalization, whereby systematic variations are adjusted for to **make expression counts comparable across genes and/or samples**. The counts of mapped reads for each gene is proportional to the expression of RNA ("interesting") in addition to many other factors ("uninteresting"). Normalization is the process of adjusting raw count values to account for the "uninteresting" factors.</span>
<span id="cb47-40"><a href="#cb47-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-41"><a href="#cb47-41" aria-hidden="true" tabindex="-1"></a>The main factors often considered during normalization are:</span>
<span id="cb47-42"><a href="#cb47-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-43"><a href="#cb47-43" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>**Sequencing depth:** Accounting for sequencing depth is necessary for comparison of gene expression between cells. In the example below, each gene appears to have doubled in expression in cell 2, however this is a consequence of cell 2 having twice the sequencing depth.</span>
<span id="cb47-44"><a href="#cb47-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-45"><a href="#cb47-45" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/sequencing_depth.png)</span>{width="509"}</span>
<span id="cb47-46"><a href="#cb47-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-47"><a href="#cb47-47" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; Each cell in scRNA-seq will have a differing number of reads associated with it. So to accurately compare expression between cells, it is necessary to normalize for sequencing depth.</span></span>
<span id="cb47-48"><a href="#cb47-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-49"><a href="#cb47-49" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>**Gene length:** Accounting for gene length is necessary for comparing expression between different genes within the same cell. The number of reads mapped to a longer gene can appear to have equal count/expression as a shorter gene that is more highly expressed.</span>
<span id="cb47-50"><a href="#cb47-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-51"><a href="#cb47-51" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/length_of_gene.png)</span>{width="464"}</span>
<span id="cb47-52"><a href="#cb47-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-53"><a href="#cb47-53" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb47-54"><a href="#cb47-54" aria-hidden="true" tabindex="-1"></a>If using a **3' or 5' droplet-based method (10X Genomics, CEL-seq2, Drop-seq, inDrops)**, the length of the gene will **not** affect the analysis because only the 5' or 3' end of the transcript is sequenced. However, if using **full-length sequencing (Smart-seq)**, the transcript length should be accounted for.</span>
<span id="cb47-55"><a href="#cb47-55" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb47-56"><a href="#cb47-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-57"><a href="#cb47-57" aria-hidden="true" tabindex="-1"></a><span class="fu">### Methods for scRNA-seq normalization</span></span>
<span id="cb47-58"><a href="#cb47-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-59"><a href="#cb47-59" aria-hidden="true" tabindex="-1"></a>Various methods have been developed specifically for scRNA-seq normalization. Some simpler methods resemble what we have seen with bulk RNA-seq; the application of **global scale factors** adjusting for a count-depth relationship that is assumed common across all genes. However, if those assumptions are not true then this basic normalization can lead to over-correction for lowly and moderately expressed genes and, in some cases, under-normalization of highly expressed genes [@bacher2017]. More complex methods will apply correction on a per-gene basis**.** In this lesson we will explore both approaches.</span>
<span id="cb47-60"><a href="#cb47-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-61"><a href="#cb47-61" aria-hidden="true" tabindex="-1"></a>Regardless of which method is used for normalization, it can be helpful to think of it as a **two-step process** (even though it is often described as a single step in most papers). The first is a scaling step and the second is a transformation.</span>
<span id="cb47-62"><a href="#cb47-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-63"><a href="#cb47-63" aria-hidden="true" tabindex="-1"></a>**1. Scaling**</span>
<span id="cb47-64"><a href="#cb47-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-65"><a href="#cb47-65" aria-hidden="true" tabindex="-1"></a>The first step in normalization is to **multiply each UMI count by a [cell specific factor]{.underline} to get [all cells to have the same UMI counts]{.underline}**. Why would we want to do this? Different cells have different amounts of mRNA; this could be due to differences between cell types or variation within the same cell type depending on how well the chemistry worked in one drop versus another. In either case, we are not interested in comparing these **absolute counts** between cells. Instead we are interested in comparing **concentrations**, and scaling helps achieve this.</span>
<span id="cb47-66"><a href="#cb47-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-67"><a href="#cb47-67" aria-hidden="true" tabindex="-1"></a>**2. Transformation**</span>
<span id="cb47-68"><a href="#cb47-68" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-69"><a href="#cb47-69" aria-hidden="true" tabindex="-1"></a>The next step is a transformation, and it is at this step where we can distinguish the simpler versus complex methods as mentioned above.</span>
<span id="cb47-70"><a href="#cb47-70" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-71"><a href="#cb47-71" aria-hidden="true" tabindex="-1"></a>**Simple transformations** are those which apply the same function to each individual measurement. Common examples include a **log transform** (which is applied in the original Seurat workflow), or a square root transform (less commonly used).</span>
<span id="cb47-72"><a href="#cb47-72" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-73"><a href="#cb47-73" aria-hidden="true" tabindex="-1"></a>In the <span class="co">[</span><span class="ot">Hafemeister and Satija, 2019 paper</span><span class="co">](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1874-1)</span> the authors explored the issues with simple transformations. Specifically they evaluated the standard log normalization approach and found that genes with different abundances are affected differently and that effective normalization (using the log transform) is only observed with low/medium abundance genes (Figure 1D, below). Additionally, substantial imbalances in variance were observed with the log-normalized data (Figure 1E, below). In particular, cells with low total UMI counts exhibited disproportionately higher variance for high-abundance genes, dampening the variance contribution from other gene abundances.&nbsp;</span>
<span id="cb47-74"><a href="#cb47-74" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-75"><a href="#cb47-75" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/SCT_Fig1.png)</span></span>
<span id="cb47-76"><a href="#cb47-76" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-77"><a href="#cb47-77" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; Image credit: Hafemeister C and Satija R. Normalization and variance stabilization of single-cell RNA-seq data using regularized negative binomial regression, Genom Biology 2019 (</span><span class="ot">&lt;https://doi.org/10.1186/s13059-019-1874-1&gt;</span><span class="at">)</span></span>
<span id="cb47-78"><a href="#cb47-78" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-79"><a href="#cb47-79" aria-hidden="true" tabindex="-1"></a>The conclusion is, **we cannot treat all genes the same.**</span>
<span id="cb47-80"><a href="#cb47-80" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-81"><a href="#cb47-81" aria-hidden="true" tabindex="-1"></a>The proposed solution was the use of **Pearson residuals for transformation**, as implemented in Seurat's <span class="in">`SCTransform`</span> function. With this approach:</span>
<span id="cb47-82"><a href="#cb47-82" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-83"><a href="#cb47-83" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Measurements are multiplied by a **gene-specific weight**</span>
<span id="cb47-84"><a href="#cb47-84" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-85"><a href="#cb47-85" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Each gene is weighted based on how much evidence there is that it is non-uniformly expressed across cells. More evidence == more of a weight</span>
<span id="cb47-86"><a href="#cb47-86" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-87"><a href="#cb47-87" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Genes that are expressed in only a small fraction of cells will be favored (useful for finding rare cell populations)</span>
<span id="cb47-88"><a href="#cb47-88" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-89"><a href="#cb47-89" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Not just a consideration of the expression level is, but also the distribution of expression</span>
<span id="cb47-90"><a href="#cb47-90" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-91"><a href="#cb47-91" aria-hidden="true" tabindex="-1"></a>In this workshop we will demonstrate the use of both transformations at different steps in the workflow.</span>
<span id="cb47-92"><a href="#cb47-92" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-93"><a href="#cb47-93" aria-hidden="true" tabindex="-1"></a><span class="fu">## 基于传统标准化流程对非期望变异来源进行评估</span></span>
<span id="cb47-94"><a href="#cb47-94" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-95"><a href="#cb47-95" aria-hidden="true" tabindex="-1"></a>The most common biological data correction (or source of "uninteresting" variation) in single cell RNA-seq is the effects of the **cell cycle** on the transcriptome. We need to explore the data and see if we observe any effects in our data.</span>
<span id="cb47-96"><a href="#cb47-96" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-97"><a href="#cb47-97" aria-hidden="true" tabindex="-1"></a><span class="fu">### Set-up</span></span>
<span id="cb47-98"><a href="#cb47-98" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-99"><a href="#cb47-99" aria-hidden="true" tabindex="-1"></a>Before we make any comparisons across cells, we will apply a **simple normalization.** **This is solely for the purpose of exploring the sources of variation in our data.**</span>
<span id="cb47-100"><a href="#cb47-100" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-101"><a href="#cb47-101" aria-hidden="true" tabindex="-1"></a>The input for this analysis is a <span class="in">`seurat`</span> object. We will use the one that we created in @sec-qc called <span class="in">`filtered_seurat`</span>.</span>
<span id="cb47-102"><a href="#cb47-102" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-105"><a href="#cb47-105" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb47-106"><a href="#cb47-106" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb47-107"><a href="#cb47-107" aria-hidden="true" tabindex="-1"></a>filtered_seurat <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"output/scRNA-seq_online/seurat_filtered.rds"</span>)</span>
<span id="cb47-108"><a href="#cb47-108" aria-hidden="true" tabindex="-1"></a>filtered_seurat</span>
<span id="cb47-109"><a href="#cb47-109" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(filtered_seurat<span class="sc">@</span>meta.data)</span>
<span id="cb47-110"><a href="#cb47-110" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-111"><a href="#cb47-111" aria-hidden="true" tabindex="-1"></a><span class="co"># Normalize the counts</span></span>
<span id="cb47-112"><a href="#cb47-112" aria-hidden="true" tabindex="-1"></a>seurat_phase <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(filtered_seurat)</span>
<span id="cb47-113"><a href="#cb47-113" aria-hidden="true" tabindex="-1"></a>seurat_phase</span>
<span id="cb47-114"><a href="#cb47-114" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb47-115"><a href="#cb47-115" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-116"><a href="#cb47-116" aria-hidden="true" tabindex="-1"></a>可以发现，运行<span class="in">`NormalizeDataNext`</span>后的数据多出了新的layer：data, 里面即储存了标准化后的数据。Next, we take this normalized data and check to see if data correction methods are necessary.</span>
<span id="cb47-117"><a href="#cb47-117" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-118"><a href="#cb47-118" aria-hidden="true" tabindex="-1"></a><span class="fu">### 评估细胞周期的影响（Evaluating effects of cell cycle） {#sec-evaluating_effects_of_cell_cycle}</span></span>
<span id="cb47-119"><a href="#cb47-119" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-120"><a href="#cb47-120" aria-hidden="true" tabindex="-1"></a><span class="fu">#### 计算细胞周期评分</span></span>
<span id="cb47-121"><a href="#cb47-121" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-122"><a href="#cb47-122" aria-hidden="true" tabindex="-1"></a>To assign each cell a score based on its expression of G2/M and S phase markers, we can use the Seuart function <span class="in">`CellCycleScoring()`</span>. This function calculates cell cycle phase scores based on canonical markers that required as input.</span>
<span id="cb47-123"><a href="#cb47-123" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-124"><a href="#cb47-124" aria-hidden="true" tabindex="-1"></a>A list of cell cycle markers, from Tirosh et al, 2015, is loaded with Seurat. We can segregate this list into markers of G2/M phase and markers of S phase. However, if you are not working with human data we have <span class="co">[</span><span class="ot">additional materials</span><span class="co">](https://hbctraining.github.io/scRNA-seq_online/lessons/cell_cycle_scoring.html)</span> detailing how to acquire cell cycle markers for other organisms of interest.</span>
<span id="cb47-125"><a href="#cb47-125" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-128"><a href="#cb47-128" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb47-129"><a href="#cb47-129" aria-hidden="true" tabindex="-1"></a><span class="co"># Load cell cycle markers</span></span>
<span id="cb47-130"><a href="#cb47-130" aria-hidden="true" tabindex="-1"></a>s.genes <span class="ot">&lt;-</span> cc.genes<span class="sc">$</span>s.genes</span>
<span id="cb47-131"><a href="#cb47-131" aria-hidden="true" tabindex="-1"></a>g2m.genes <span class="ot">&lt;-</span> cc.genes<span class="sc">$</span>g2m.genes</span>
<span id="cb47-132"><a href="#cb47-132" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-133"><a href="#cb47-133" aria-hidden="true" tabindex="-1"></a><span class="co"># Score cells for cell cycle</span></span>
<span id="cb47-134"><a href="#cb47-134" aria-hidden="true" tabindex="-1"></a>seurat_phase <span class="ot">&lt;-</span> <span class="fu">CellCycleScoring</span>(seurat_phase, </span>
<span id="cb47-135"><a href="#cb47-135" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">g2m.features =</span> g2m.genes, </span>
<span id="cb47-136"><a href="#cb47-136" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">s.features =</span> s.genes)</span>
<span id="cb47-137"><a href="#cb47-137" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-138"><a href="#cb47-138" aria-hidden="true" tabindex="-1"></a><span class="co"># 现在的meta.data中多出了细胞周期评分“S.Score”和“G2M.Score”，以及推断的细胞所处的周期“Phase”</span></span>
<span id="cb47-139"><a href="#cb47-139" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(seurat_phase<span class="sc">@</span>meta.data)</span>
<span id="cb47-140"><a href="#cb47-140" aria-hidden="true" tabindex="-1"></a><span class="co"># 查看一下细胞周期的分布情况</span></span>
<span id="cb47-141"><a href="#cb47-141" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(seurat_phase<span class="sc">$</span>Phase)</span>
<span id="cb47-142"><a href="#cb47-142" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb47-143"><a href="#cb47-143" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-144"><a href="#cb47-144" aria-hidden="true" tabindex="-1"></a><span class="fu">#### 使用PCA确定细胞周期是否是我们数据集中的主要变异来源</span></span>
<span id="cb47-145"><a href="#cb47-145" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-146"><a href="#cb47-146" aria-hidden="true" tabindex="-1"></a>After scoring the cells for cell cycle, we would like to determine **whether cell cycle is a major source of variation in our dataset using PCA（PCA的原理详见：** @sec-pca_theory **）**.</span>
<span id="cb47-147"><a href="#cb47-147" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-148"><a href="#cb47-148" aria-hidden="true" tabindex="-1"></a>To perform PCA, we need to **first choose the most variable features, then scale the data**. Since highly expressed genes exhibit the highest amount of variation and we don't want our 'highly variable genes' only to reflect high expression, we need to scale the data to scale variation with expression level. The Seurat <span class="in">`ScaleData()`</span> function will scale the data by:</span>
<span id="cb47-149"><a href="#cb47-149" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-150"><a href="#cb47-150" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>adjusting the expression of each gene **to give a mean expression across cells to be 0**</span>
<span id="cb47-151"><a href="#cb47-151" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>scaling expression of each gene **to give a variance across cells to be 1**</span>
<span id="cb47-152"><a href="#cb47-152" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-155"><a href="#cb47-155" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb47-156"><a href="#cb47-156" aria-hidden="true" tabindex="-1"></a><span class="co"># Identify the most variable genes</span></span>
<span id="cb47-157"><a href="#cb47-157" aria-hidden="true" tabindex="-1"></a>seurat_phase <span class="ot">&lt;-</span> <span class="fu">FindVariableFeatures</span>(seurat_phase, </span>
<span id="cb47-158"><a href="#cb47-158" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">selection.method =</span> <span class="st">"vst"</span>, <span class="co"># 默认值</span></span>
<span id="cb47-159"><a href="#cb47-159" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">nfeatures =</span> <span class="dv">2000</span>) <span class="co"># 默认值</span></span>
<span id="cb47-160"><a href="#cb47-160" aria-hidden="true" tabindex="-1"></a>             </span>
<span id="cb47-161"><a href="#cb47-161" aria-hidden="true" tabindex="-1"></a><span class="co"># Scale the counts</span></span>
<span id="cb47-162"><a href="#cb47-162" aria-hidden="true" tabindex="-1"></a>seurat_phase <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(seurat_phase)</span>
<span id="cb47-163"><a href="#cb47-163" aria-hidden="true" tabindex="-1"></a><span class="fu">Layers</span>(seurat_phase)</span>
<span id="cb47-164"><a href="#cb47-164" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb47-165"><a href="#cb47-165" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-166"><a href="#cb47-166" aria-hidden="true" tabindex="-1"></a>可以发现，运行<span class="in">`ScaleData`</span>后的数据多出了新的layer：scale.data, 里面即储存了归一化后的数据。Now, we can perform the PCA analysis and plot the first two principal components against each other. We also split the figure by cell cycle phase, to evaluate similarities and/or differences.</span>
<span id="cb47-167"><a href="#cb47-167" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-170"><a href="#cb47-170" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb47-171"><a href="#cb47-171" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 10</span></span>
<span id="cb47-172"><a href="#cb47-172" aria-hidden="true" tabindex="-1"></a><span class="co"># Perform PCA。如果没有指定features，RunPCA默认使用FindVariableFeatures找到的高变基因作为PCA输入.</span></span>
<span id="cb47-173"><a href="#cb47-173" aria-hidden="true" tabindex="-1"></a>seurat_phase <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(seurat_phase)</span>
<span id="cb47-174"><a href="#cb47-174" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-175"><a href="#cb47-175" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the PCA colored by cell cycle phase</span></span>
<span id="cb47-176"><a href="#cb47-176" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(seurat_phase,</span>
<span id="cb47-177"><a href="#cb47-177" aria-hidden="true" tabindex="-1"></a>              <span class="at">reduction =</span> <span class="st">"pca"</span>,</span>
<span id="cb47-178"><a href="#cb47-178" aria-hidden="true" tabindex="-1"></a>              <span class="at">group.by=</span> <span class="st">"Phase"</span>)</span>
<span id="cb47-179"><a href="#cb47-179" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(seurat_phase,</span>
<span id="cb47-180"><a href="#cb47-180" aria-hidden="true" tabindex="-1"></a>              <span class="at">reduction =</span> <span class="st">"pca"</span>,</span>
<span id="cb47-181"><a href="#cb47-181" aria-hidden="true" tabindex="-1"></a>              <span class="at">group.by=</span> <span class="st">"Phase"</span>,</span>
<span id="cb47-182"><a href="#cb47-182" aria-hidden="true" tabindex="-1"></a>              <span class="at">split.by =</span> <span class="st">"Phase"</span>)</span>
<span id="cb47-183"><a href="#cb47-183" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(cowplot)</span>
<span id="cb47-184"><a href="#cb47-184" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_grid</span>(p1, p2, <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">labels =</span> <span class="st">"AUTO"</span>)</span>
<span id="cb47-185"><a href="#cb47-185" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb47-186"><a href="#cb47-186" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-187"><a href="#cb47-187" aria-hidden="true" tabindex="-1"></a>**We do not see large differences due to cell cycle phase. Based on this plot, we would not regress out the variation due to cell cycle.**</span>
<span id="cb47-188"><a href="#cb47-188" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-189"><a href="#cb47-189" aria-hidden="true" tabindex="-1"></a>::: {.callout-tip collapse="true"}</span>
<span id="cb47-190"><a href="#cb47-190" aria-hidden="true" tabindex="-1"></a><span class="fu">###### When should cell cycle phase be regressed out?</span></span>
<span id="cb47-191"><a href="#cb47-191" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-192"><a href="#cb47-192" aria-hidden="true" tabindex="-1"></a>Below are two PCA plots taken from @sec-Elimination_of_cell_cycle_effects 。This first plot is similar to what we plotted above, it is a PCA prior to regression to evaluate if the cell cycle is playing a big role in driving PC1 and PC2. Clearly, the cells are separating by cell type in this case, so it suggests regressing out these effects.</span>
<span id="cb47-193"><a href="#cb47-193" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-194"><a href="#cb47-194" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/unnamed-chunk-6-1.png)</span>{width="427"}</span>
<span id="cb47-195"><a href="#cb47-195" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-196"><a href="#cb47-196" aria-hidden="true" tabindex="-1"></a>This second PCA plot is post-regression, and displays how effective the regression was in removing the effect we observed.</span>
<span id="cb47-197"><a href="#cb47-197" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-198"><a href="#cb47-198" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/unnamed-chunk-10-1.png)</span>{width="430"}</span>
<span id="cb47-199"><a href="#cb47-199" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb47-200"><a href="#cb47-200" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-201"><a href="#cb47-201" aria-hidden="true" tabindex="-1"></a>::: callout-tip</span>
<span id="cb47-202"><a href="#cb47-202" aria-hidden="true" tabindex="-1"></a>在需要消除细胞周期的影响时，如何通过<span class="in">`ScaleData`</span>回归掉（regress out）细胞周期的影响，以及如何在消除细胞周期的影响同时保留增殖细胞与静止细胞的区分，参考 @sec-Elimination_of_cell_cycle_effects 。</span>
<span id="cb47-203"><a href="#cb47-203" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb47-204"><a href="#cb47-204" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-205"><a href="#cb47-205" aria-hidden="true" tabindex="-1"></a>------------------------------------------------------------------------</span>
<span id="cb47-206"><a href="#cb47-206" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-207"><a href="#cb47-207" aria-hidden="true" tabindex="-1"></a><span class="fu">### 评估线粒体基因表达的影响</span></span>
<span id="cb47-208"><a href="#cb47-208" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-209"><a href="#cb47-209" aria-hidden="true" tabindex="-1"></a>Mitochondrial expression is another factor which can greatly influence clustering. Oftentimes, it is useful to regress out variation due to mitochondrial expression. However, **if the differences in mitochondrial gene expression represent a biological phenomenon that may help to distinguish cell clusters, then we advise not regressing this out**. In this exercise, we can perform a quick check similar to looking at cell cycle and decide whether or not we want to regress it out.</span>
<span id="cb47-210"><a href="#cb47-210" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-211"><a href="#cb47-211" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>First, turn the mitochondrial ratio variable into a new categorical variable based on quartiles. 根据各细胞线粒体基因的比例的四分位数将所有细胞分为低线粒体基因比例细胞、中线粒体基因比例细胞···</span>
<span id="cb47-212"><a href="#cb47-212" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-213"><a href="#cb47-213" aria-hidden="true" tabindex="-1"></a>    <span class="in">```{r}</span></span>
<span id="cb47-214"><a href="#cb47-214" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Check quartile values</span></span>
<span id="cb47-215"><a href="#cb47-215" aria-hidden="true" tabindex="-1"></a>    <span class="fu">summary</span>(seurat_phase<span class="sc">$</span>mitoRatio)</span>
<span id="cb47-216"><a href="#cb47-216" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-217"><a href="#cb47-217" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Turn mitoRatio into categorical factor vector based on quartile values</span></span>
<span id="cb47-218"><a href="#cb47-218" aria-hidden="true" tabindex="-1"></a>    seurat_phase<span class="sc">$</span>mitoFr <span class="ot">&lt;-</span> <span class="fu">cut</span>(seurat_phase<span class="sc">@</span>meta.data<span class="sc">$</span>mitoRatio, </span>
<span id="cb47-219"><a href="#cb47-219" aria-hidden="true" tabindex="-1"></a>                               <span class="at">breaks=</span><span class="fu">c</span>(<span class="sc">-</span><span class="cn">Inf</span>, <span class="fl">0.0144</span>, <span class="fl">0.0199</span>, <span class="fl">0.0267</span>, <span class="cn">Inf</span>), </span>
<span id="cb47-220"><a href="#cb47-220" aria-hidden="true" tabindex="-1"></a>                               <span class="at">labels=</span><span class="fu">c</span>(<span class="st">"Low"</span>, <span class="st">"Medium"</span>, <span class="st">"Medium high"</span>, <span class="st">"High"</span>))</span>
<span id="cb47-221"><a href="#cb47-221" aria-hidden="true" tabindex="-1"></a>    <span class="fu">table</span>(seurat_phase<span class="sc">$</span>mitoFr)</span>
<span id="cb47-222"><a href="#cb47-222" aria-hidden="true" tabindex="-1"></a>    <span class="in">```</span></span>
<span id="cb47-223"><a href="#cb47-223" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-224"><a href="#cb47-224" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>Next, plot the PCA similar to how we did with cell cycle regression. *Hint: use the new `mitoFr` variable to split cells and color them accordingly.*</span>
<span id="cb47-225"><a href="#cb47-225" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-226"><a href="#cb47-226" aria-hidden="true" tabindex="-1"></a>    <span class="in">```{r}</span></span>
<span id="cb47-227"><a href="#cb47-227" aria-hidden="true" tabindex="-1"></a>    <span class="co">#| fig-width: 20</span></span>
<span id="cb47-228"><a href="#cb47-228" aria-hidden="true" tabindex="-1"></a>    <span class="co"># 根据各细胞线粒体基因的比例信息绘制PCA</span></span>
<span id="cb47-229"><a href="#cb47-229" aria-hidden="true" tabindex="-1"></a>    p1 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(seurat_phase,</span>
<span id="cb47-230"><a href="#cb47-230" aria-hidden="true" tabindex="-1"></a>                  <span class="at">reduction =</span> <span class="st">"pca"</span>,</span>
<span id="cb47-231"><a href="#cb47-231" aria-hidden="true" tabindex="-1"></a>                  <span class="at">group.by=</span> <span class="st">"mitoFr"</span>)</span>
<span id="cb47-232"><a href="#cb47-232" aria-hidden="true" tabindex="-1"></a>    p2 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(seurat_phase,</span>
<span id="cb47-233"><a href="#cb47-233" aria-hidden="true" tabindex="-1"></a>                  <span class="at">reduction =</span> <span class="st">"pca"</span>,</span>
<span id="cb47-234"><a href="#cb47-234" aria-hidden="true" tabindex="-1"></a>                  <span class="at">group.by=</span> <span class="st">"mitoFr"</span>,</span>
<span id="cb47-235"><a href="#cb47-235" aria-hidden="true" tabindex="-1"></a>                  <span class="at">split.by =</span> <span class="st">"mitoFr"</span>)</span>
<span id="cb47-236"><a href="#cb47-236" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot_grid</span>(p1, p2, <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">labels =</span> <span class="st">"AUTO"</span>)</span>
<span id="cb47-237"><a href="#cb47-237" aria-hidden="true" tabindex="-1"></a>    <span class="in">```</span></span>
<span id="cb47-238"><a href="#cb47-238" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-239"><a href="#cb47-239" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>Evaluate the PCA plot.</span>
<span id="cb47-240"><a href="#cb47-240" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-241"><a href="#cb47-241" aria-hidden="true" tabindex="-1"></a>    We do not see large differences due to mitochondrial expression. Based on this plot, we would not regress out the variation due to mitochondrial expression.</span>
<span id="cb47-242"><a href="#cb47-242" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-243"><a href="#cb47-243" aria-hidden="true" tabindex="-1"></a>------------------------------------------------------------------------</span>
<span id="cb47-244"><a href="#cb47-244" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-245"><a href="#cb47-245" aria-hidden="true" tabindex="-1"></a><span class="fu">## `SCTransform`-based Normalization and regressing out sources of unwanted variation</span></span>
<span id="cb47-246"><a href="#cb47-246" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-247"><a href="#cb47-247" aria-hidden="true" tabindex="-1"></a>Now that we have established which effects are observed in our data, we can use the <span class="in">`SCTransform`</span> method to regress out these effects. The <span class="in">`SCTransform`</span> method was proposed as a better alternative to the log transform normalization method that we used for exploring sources of unwanted variation. The method not only **normalizes data, but it also performs a variance stabilization and allows for additional covariates to be regressed out**.</span>
<span id="cb47-248"><a href="#cb47-248" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-249"><a href="#cb47-249" aria-hidden="true" tabindex="-1"></a>As described earlier, all genes cannot be treated the same. As such, the <span class="in">`SCTransform`</span> **method constructs a generalized linear model (GLM) for each gene** with UMI counts as the response and sequencing depth as the explanatory variable. Information is pooled across genes with similar abundances, to regularize parameter estimates and **obtain residuals which represent effectively normalized data values** which are no longer correlated with sequencing depth.</span>
<span id="cb47-250"><a href="#cb47-250" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-251"><a href="#cb47-251" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/SCT_UMAP.png)</span>{width="532"}</span>
<span id="cb47-252"><a href="#cb47-252" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-253"><a href="#cb47-253" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; Image credit: Hafemeister C and Satija R. Normalization and variance stabilization of single-cell RNA-seq data using regularized negative binomial regression, Genom Biology 2019 (</span><span class="ot">&lt;https://doi.org/10.1101/576827)_&gt;</span></span>
<span id="cb47-254"><a href="#cb47-254" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-255"><a href="#cb47-255" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb47-256"><a href="#cb47-256" aria-hidden="true" tabindex="-1"></a>Since the UMI counts are part of the GLM, the effects are automatically regressed out. The user can include any additional covariates (<span class="in">`vars.to.regress`</span>) that may have an effect on expression and will be included in the model.</span>
<span id="cb47-257"><a href="#cb47-257" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb47-258"><a href="#cb47-258" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-259"><a href="#cb47-259" aria-hidden="true" tabindex="-1"></a><span class="fu">### 数据导入</span></span>
<span id="cb47-260"><a href="#cb47-260" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-261"><a href="#cb47-261" aria-hidden="true" tabindex="-1"></a>前面的流程和此前一样</span>
<span id="cb47-262"><a href="#cb47-262" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-265"><a href="#cb47-265" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb47-266"><a href="#cb47-266" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(<span class="at">list =</span> <span class="fu">ls</span>())</span>
<span id="cb47-267"><a href="#cb47-267" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-268"><a href="#cb47-268" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb47-269"><a href="#cb47-269" aria-hidden="true" tabindex="-1"></a>filtered_seurat <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"output/seurat_filtered.rds"</span>)</span>
<span id="cb47-270"><a href="#cb47-270" aria-hidden="true" tabindex="-1"></a>filtered_seurat</span>
<span id="cb47-271"><a href="#cb47-271" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(filtered_seurat<span class="sc">@</span>meta.data)</span>
<span id="cb47-272"><a href="#cb47-272" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb47-273"><a href="#cb47-273" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-274"><a href="#cb47-274" aria-hidden="true" tabindex="-1"></a><span class="fu">### 执行`SCTransform`</span></span>
<span id="cb47-275"><a href="#cb47-275" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-276"><a href="#cb47-276" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>这里先运行一次<span class="in">`SCTransform`</span>以便后面评估细胞周期、线粒体基因等非期望变异来源（**This is solely for the purpose of exploring the sources of variation in our data**）</span>
<span id="cb47-277"><a href="#cb47-277" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-278"><a href="#cb47-278" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`SCTransform`</span>替代了传统单细胞数据分析流程中的<span class="in">`NormalizeData()`</span>、<span class="in">`ScaleData()`</span>和<span class="in">`FindVariableFeatures()`</span>函数的功能，因此不再需要运行这些函数。</span>
<span id="cb47-279"><a href="#cb47-279" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-280"><a href="#cb47-280" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>In Seurat v5, SCT v2 is applied by default. You can revert to v1 by setting&nbsp;<span class="in">`vst.flavor = 'v1'`</span>。</span>
<span id="cb47-281"><a href="#cb47-281" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-282"><a href="#cb47-282" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`SCTransform`</span>的运算调用了<span class="co">[</span><span class="ot">`glmGamPoi`</span><span class="co">](https://bioconductor.org/packages/release/bioc/html/glmGamPoi.html)</span>包以显著提升运算速度。所以事先需要通过<span class="in">`BiocManager`</span>安装该包。</span>
<span id="cb47-283"><a href="#cb47-283" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-286"><a href="#cb47-286" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb47-287"><a href="#cb47-287" aria-hidden="true" tabindex="-1"></a><span class="co"># SCTranform</span></span>
<span id="cb47-288"><a href="#cb47-288" aria-hidden="true" tabindex="-1"></a><span class="co"># BiocManager::install("glmGamPoi")</span></span>
<span id="cb47-289"><a href="#cb47-289" aria-hidden="true" tabindex="-1"></a>seurat_phase <span class="ot">&lt;-</span> <span class="fu">SCTransform</span>(filtered_seurat)</span>
<span id="cb47-290"><a href="#cb47-290" aria-hidden="true" tabindex="-1"></a>seurat_phase</span>
<span id="cb47-291"><a href="#cb47-291" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-292"><a href="#cb47-292" aria-hidden="true" tabindex="-1"></a><span class="co"># Check which assays are stored in objects</span></span>
<span id="cb47-293"><a href="#cb47-293" aria-hidden="true" tabindex="-1"></a>seurat_phase<span class="sc">@</span>assays</span>
<span id="cb47-294"><a href="#cb47-294" aria-hidden="true" tabindex="-1"></a><span class="co"># 查看目前默认的assay</span></span>
<span id="cb47-295"><a href="#cb47-295" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(seurat_phase)</span>
<span id="cb47-296"><a href="#cb47-296" aria-hidden="true" tabindex="-1"></a><span class="co"># 查看默认assay的layers</span></span>
<span id="cb47-297"><a href="#cb47-297" aria-hidden="true" tabindex="-1"></a><span class="fu">Layers</span>(seurat_phase)</span>
<span id="cb47-298"><a href="#cb47-298" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb47-299"><a href="#cb47-299" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-300"><a href="#cb47-300" aria-hidden="true" tabindex="-1"></a>Note, the last line of output specifies **"Set default assay to SCT"**. 表明运行<span class="in">`SCTransform`</span>之后，会将默认的assay指定为<span class="in">`SCTransform`</span>之后的数据。This specifies that moving forward we would like to use the data after SCT was implemented. We can view the different assays that we have stored in our seurat object.</span>
<span id="cb47-301"><a href="#cb47-301" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-302"><a href="#cb47-302" aria-hidden="true" tabindex="-1"></a><span class="fu">### 评估细胞周期的影响</span></span>
<span id="cb47-303"><a href="#cb47-303" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-306"><a href="#cb47-306" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb47-307"><a href="#cb47-307" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 10</span></span>
<span id="cb47-308"><a href="#cb47-308" aria-hidden="true" tabindex="-1"></a><span class="co"># Load cell cycle markers</span></span>
<span id="cb47-309"><a href="#cb47-309" aria-hidden="true" tabindex="-1"></a>s.genes <span class="ot">&lt;-</span> cc.genes<span class="sc">$</span>s.genes</span>
<span id="cb47-310"><a href="#cb47-310" aria-hidden="true" tabindex="-1"></a>g2m.genes <span class="ot">&lt;-</span> cc.genes<span class="sc">$</span>g2m.genes</span>
<span id="cb47-311"><a href="#cb47-311" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-312"><a href="#cb47-312" aria-hidden="true" tabindex="-1"></a><span class="co"># Score cells for cell cycle</span></span>
<span id="cb47-313"><a href="#cb47-313" aria-hidden="true" tabindex="-1"></a>seurat_phase <span class="ot">&lt;-</span> <span class="fu">CellCycleScoring</span>(seurat_phase, </span>
<span id="cb47-314"><a href="#cb47-314" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">g2m.features =</span> g2m.genes, </span>
<span id="cb47-315"><a href="#cb47-315" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">s.features =</span> s.genes)</span>
<span id="cb47-316"><a href="#cb47-316" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-317"><a href="#cb47-317" aria-hidden="true" tabindex="-1"></a><span class="co"># 现在的meta.data中多出了细胞周期评分“S.Score”和“G2M.Score”，以及推断的细胞所处的周期“Phase”</span></span>
<span id="cb47-318"><a href="#cb47-318" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(seurat_phase<span class="sc">@</span>meta.data)</span>
<span id="cb47-319"><a href="#cb47-319" aria-hidden="true" tabindex="-1"></a><span class="co"># 查看一下细胞周期的分布情况</span></span>
<span id="cb47-320"><a href="#cb47-320" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(seurat_phase<span class="sc">$</span>Phase)</span>
<span id="cb47-321"><a href="#cb47-321" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-322"><a href="#cb47-322" aria-hidden="true" tabindex="-1"></a><span class="co"># 执行PCA</span></span>
<span id="cb47-323"><a href="#cb47-323" aria-hidden="true" tabindex="-1"></a>seurat_phase <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(seurat_phase)</span>
<span id="cb47-324"><a href="#cb47-324" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-325"><a href="#cb47-325" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the PCA colored by cell cycle phase</span></span>
<span id="cb47-326"><a href="#cb47-326" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(seurat_phase,</span>
<span id="cb47-327"><a href="#cb47-327" aria-hidden="true" tabindex="-1"></a>              <span class="at">reduction =</span> <span class="st">"pca"</span>,</span>
<span id="cb47-328"><a href="#cb47-328" aria-hidden="true" tabindex="-1"></a>              <span class="at">group.by=</span> <span class="st">"Phase"</span>)</span>
<span id="cb47-329"><a href="#cb47-329" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(seurat_phase,</span>
<span id="cb47-330"><a href="#cb47-330" aria-hidden="true" tabindex="-1"></a>              <span class="at">reduction =</span> <span class="st">"pca"</span>,</span>
<span id="cb47-331"><a href="#cb47-331" aria-hidden="true" tabindex="-1"></a>              <span class="at">group.by=</span> <span class="st">"Phase"</span>,</span>
<span id="cb47-332"><a href="#cb47-332" aria-hidden="true" tabindex="-1"></a>              <span class="at">split.by =</span> <span class="st">"Phase"</span>)</span>
<span id="cb47-333"><a href="#cb47-333" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(cowplot)</span>
<span id="cb47-334"><a href="#cb47-334" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_grid</span>(p1, p2, <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">labels =</span> <span class="st">"AUTO"</span>)</span>
<span id="cb47-335"><a href="#cb47-335" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb47-336"><a href="#cb47-336" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-337"><a href="#cb47-337" aria-hidden="true" tabindex="-1"></a><span class="fu">### 评估线粒体基因的影响</span></span>
<span id="cb47-338"><a href="#cb47-338" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-341"><a href="#cb47-341" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb47-342"><a href="#cb47-342" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 20</span></span>
<span id="cb47-343"><a href="#cb47-343" aria-hidden="true" tabindex="-1"></a><span class="co"># Check quartile values</span></span>
<span id="cb47-344"><a href="#cb47-344" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(seurat_phase<span class="sc">$</span>mitoRatio)</span>
<span id="cb47-345"><a href="#cb47-345" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-346"><a href="#cb47-346" aria-hidden="true" tabindex="-1"></a><span class="co"># Turn mitoRatio into categorical factor vector based on quartile values</span></span>
<span id="cb47-347"><a href="#cb47-347" aria-hidden="true" tabindex="-1"></a>seurat_phase<span class="sc">$</span>mitoFr <span class="ot">&lt;-</span> <span class="fu">cut</span>(seurat_phase<span class="sc">@</span>meta.data<span class="sc">$</span>mitoRatio, </span>
<span id="cb47-348"><a href="#cb47-348" aria-hidden="true" tabindex="-1"></a>                           <span class="at">breaks=</span><span class="fu">c</span>(<span class="sc">-</span><span class="cn">Inf</span>, <span class="fl">0.0144</span>, <span class="fl">0.0199</span>, <span class="fl">0.0267</span>, <span class="cn">Inf</span>), </span>
<span id="cb47-349"><a href="#cb47-349" aria-hidden="true" tabindex="-1"></a>                           <span class="at">labels=</span><span class="fu">c</span>(<span class="st">"Low"</span>, <span class="st">"Medium"</span>, <span class="st">"Medium high"</span>, <span class="st">"High"</span>))</span>
<span id="cb47-350"><a href="#cb47-350" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(seurat_phase<span class="sc">$</span>mitoFr)</span>
<span id="cb47-351"><a href="#cb47-351" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-352"><a href="#cb47-352" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-353"><a href="#cb47-353" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the PCA colored by cell cycle phase</span></span>
<span id="cb47-354"><a href="#cb47-354" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(seurat_phase,</span>
<span id="cb47-355"><a href="#cb47-355" aria-hidden="true" tabindex="-1"></a>              <span class="at">reduction =</span> <span class="st">"pca"</span>,</span>
<span id="cb47-356"><a href="#cb47-356" aria-hidden="true" tabindex="-1"></a>              <span class="at">group.by=</span> <span class="st">"mitoFr"</span>)</span>
<span id="cb47-357"><a href="#cb47-357" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(seurat_phase,</span>
<span id="cb47-358"><a href="#cb47-358" aria-hidden="true" tabindex="-1"></a>              <span class="at">reduction =</span> <span class="st">"pca"</span>,</span>
<span id="cb47-359"><a href="#cb47-359" aria-hidden="true" tabindex="-1"></a>              <span class="at">group.by=</span> <span class="st">"mitoFr"</span>,</span>
<span id="cb47-360"><a href="#cb47-360" aria-hidden="true" tabindex="-1"></a>              <span class="at">split.by =</span> <span class="st">"mitoFr"</span>)</span>
<span id="cb47-361"><a href="#cb47-361" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_grid</span>(p1, p2, <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">labels =</span> <span class="st">"AUTO"</span>)</span>
<span id="cb47-362"><a href="#cb47-362" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb47-363"><a href="#cb47-363" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-364"><a href="#cb47-364" aria-hidden="true" tabindex="-1"></a><span class="fu">### 分割layer，再次执行`SCTranform`</span></span>
<span id="cb47-365"><a href="#cb47-365" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-366"><a href="#cb47-366" aria-hidden="true" tabindex="-1"></a>Since we have two samples in our dataset (from two conditions), we want to keep them as separate layers and transform them as that is **what is required for integration**.</span>
<span id="cb47-367"><a href="#cb47-367" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-370"><a href="#cb47-370" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb47-371"><a href="#cb47-371" aria-hidden="true" tabindex="-1"></a><span class="co"># Split RNA assay by condition to perform cell cycle scoring and SCT on all samples</span></span>
<span id="cb47-372"><a href="#cb47-372" aria-hidden="true" tabindex="-1"></a>seurat_phase[[<span class="st">"RNA"</span>]] <span class="ot">&lt;-</span> <span class="fu">split</span>(seurat_phase[[<span class="st">"RNA"</span>]], </span>
<span id="cb47-373"><a href="#cb47-373" aria-hidden="true" tabindex="-1"></a>                               <span class="at">f =</span> seurat_phase<span class="sc">$</span>sample) <span class="co"># 按照meta.data中的“sample”列进行分割</span></span>
<span id="cb47-374"><a href="#cb47-374" aria-hidden="true" tabindex="-1"></a>seurat_phase</span>
<span id="cb47-375"><a href="#cb47-375" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb47-376"><a href="#cb47-376" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-377"><a href="#cb47-377" aria-hidden="true" tabindex="-1"></a>现在可以发现RNA assay的<span class="in">`counts`</span>和<span class="in">`data`</span>按照"seurat_phase\$sample"（ctrl vs. stim）被分别分割成了2个layer：</span>
<span id="cb47-378"><a href="#cb47-378" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-379"><a href="#cb47-379" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/截屏2023-12-19%2010.46.18.png)</span></span>
<span id="cb47-380"><a href="#cb47-380" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-381"><a href="#cb47-381" aria-hidden="true" tabindex="-1"></a>Now we will run the <span class="in">`SCTransform()`</span> on each sample, and regress out mitochondrial expression by specifying in the <span class="in">`vars.to.regress`</span> argument of the <span class="in">`SCTransform()`</span> function.</span>
<span id="cb47-382"><a href="#cb47-382" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-383"><a href="#cb47-383" aria-hidden="true" tabindex="-1"></a>::: callout-tip</span>
<span id="cb47-384"><a href="#cb47-384" aria-hidden="true" tabindex="-1"></a>The output of <span class="in">`SCTransform()`</span> can generate large R objects/variables in terms of memory. If we have a large dataset, then we might need to **adjust the limit for allowable object sizes within R** (*Default is 500* 1024 \^ 2 = 500 Mb<span class="sc">\*</span>) using the following code:</span>
<span id="cb47-385"><a href="#cb47-385" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-388"><a href="#cb47-388" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb47-389"><a href="#cb47-389" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb47-390"><a href="#cb47-390" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: fenced</span></span>
<span id="cb47-391"><a href="#cb47-391" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">future.globals.maxSize =</span> <span class="dv">4000</span> <span class="sc">*</span> <span class="dv">1024</span><span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb47-392"><a href="#cb47-392" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb47-393"><a href="#cb47-393" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb47-394"><a href="#cb47-394" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-395"><a href="#cb47-395" aria-hidden="true" tabindex="-1"></a><span class="fu">#### 执行`SCTranform` {#sec-perform_sctranform}</span></span>
<span id="cb47-396"><a href="#cb47-396" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-399"><a href="#cb47-399" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb47-400"><a href="#cb47-400" aria-hidden="true" tabindex="-1"></a><span class="co"># SCTranform</span></span>
<span id="cb47-401"><a href="#cb47-401" aria-hidden="true" tabindex="-1"></a>seurat_phase <span class="ot">&lt;-</span> <span class="fu">SCTransform</span>(seurat_phase, <span class="at">vars.to.regress =</span> <span class="fu">c</span>(<span class="st">"mitoRatio"</span>))</span>
<span id="cb47-402"><a href="#cb47-402" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb47-403"><a href="#cb47-403" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-404"><a href="#cb47-404" aria-hidden="true" tabindex="-1"></a>By default, after normalizing, adjusting the variance, and regressing out uninteresting sources of variation, <span class="in">`SCTransform`</span> will rank the genes by residual variance and output the **3000** most variant genes. If the dataset has larger cell numbers, then it may be beneficial to adjust this parameter higher using the <span class="in">`variable.features.n`</span> argument.</span>
<span id="cb47-405"><a href="#cb47-405" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-406"><a href="#cb47-406" aria-hidden="true" tabindex="-1"></a>Now we can see that in addition to the raw RNA counts, we now have a SCT component in our <span class="in">`assays`</span> slot. The most variable features will be the only genes stored inside the SCT assay. As we move through the scRNA-seq analysis, we will choose the most appropriate assay to use for the different steps in the analysis.</span>
<span id="cb47-407"><a href="#cb47-407" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-408"><a href="#cb47-408" aria-hidden="true" tabindex="-1"></a>------------------------------------------------------------------------</span>
<span id="cb47-409"><a href="#cb47-409" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-410"><a href="#cb47-410" aria-hidden="true" tabindex="-1"></a><span class="fu">### Save the object!</span></span>
<span id="cb47-411"><a href="#cb47-411" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-414"><a href="#cb47-414" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb47-415"><a href="#cb47-415" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(seurat_phase, <span class="st">"output/scRNA-seq_online/split_seurat.rds"</span>)</span>
<span id="cb47-416"><a href="#cb47-416" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb47-417"><a href="#cb47-417" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-418"><a href="#cb47-418" aria-hidden="true" tabindex="-1"></a>------------------------------------------------------------------------</span>
<span id="cb47-419"><a href="#cb47-419" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-420"><a href="#cb47-420" aria-hidden="true" tabindex="-1"></a>::: {.callout-note collapse="true" icon="false"}</span>
<span id="cb47-421"><a href="#cb47-421" aria-hidden="true" tabindex="-1"></a><span class="fu">## Session Info</span></span>
<span id="cb47-422"><a href="#cb47-422" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-425"><a href="#cb47-425" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb47-426"><a href="#cb47-426" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb47-427"><a href="#cb47-427" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span>
<span id="cb47-428"><a href="#cb47-428" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb47-429"><a href="#cb47-429" aria-hidden="true" tabindex="-1"></a>:::</span>
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">
      <ul class="footer-items list-unstyled">
<li class="nav-item">
 This book was built with Quarto
  </li>  
</ul>
</div>   
    <div class="nav-footer-center">Copyright 2023, Du Junhong</div>
    <div class="nav-footer-right">
      <ul class="footer-items list-unstyled">
<li class="nav-item compact">
    <a class="nav-link" href="https://github.com/djhcod">
      <i class="bi bi-github" role="img">
</i> 
    </a>
  </li>  
</ul>
</div>
  </div>
</footer>


</body></html>