<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>R语言学习笔记 - 24&nbsp; 细胞分群质量评估</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../quarto_foundation/quarto_foundation.html" rel="next">
<link href="../../single_cell/scRNA-seq_online/07_SC_clustering_cells_SCT.html" rel="prev">
<link href="../../images/logo.jpg" rel="icon" type="image/jpeg">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark"><script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><script type="application/json" class="js-hypothesis-config">
{
  "theme": "classic",
  "openSidebar": false,
  "showHighlights": "always",
  "enableExperimentalNewNoteButton": true
}
</script><script async="" src="https://hypothes.is/embed.js"></script>
<meta name="twitter:title" content="R语言学习笔记 - 24&nbsp; 细胞分群质量评估">
<meta name="twitter:description" content="">
<meta name="twitter:image" content="https://djhcod.github.io/r-notes/single_cell/scRNA-seq_online/08_SC_clustering_quality_control_files/figure-html/unnamed-chunk-2-1.png">
<meta name="twitter:image-height" content="960">
<meta name="twitter:image-width" content="1344">
<meta name="twitter:card" content="summary_large_image">
</head>
<body class="nav-sidebar docked nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="navbar navbar-expand-lg navbar-dark "><div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a href="../../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../../images/logo.jpg" alt="" class="navbar-logo"></a>
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">R语言学习笔记</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
<li class="nav-item">
    <a class="nav-link active" href="../../index.html" rel="" target="" aria-current="page">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../intro.html" rel="" target="">
 <span class="menu-text">Introduction</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-parts" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Parts</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-parts">
<li>
    <a class="dropdown-item" href="../../r_basic/r_basics.html" rel="" target="">
 <span class="dropdown-text">R语言基础</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../single_cell/seurat/seurat.html" rel="" target="">
 <span class="dropdown-text">Seurat v5官方文档</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../single_cell/scRNA-seq_online/00_intro.html" rel="" target="">
 <span class="dropdown-text">scRNA-seq_online学习材料</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../quarto_foundation/quarto_foundation.html" rel="" target="">
 <span class="dropdown-text">Quarto基础</span></a>
  </li>  
    </ul>
</li>
  <li class="nav-item">
    <a class="nav-link" href="../../references.html" rel="" target="">
 <span class="menu-text">References</span></a>
  </li>  
</ul>
<ul class="navbar-nav navbar-nav-scroll ms-auto">
<li class="nav-item">
    <a class="nav-link" href="https://bookdown.org" rel="" target=""><i class="bi bi-book-fill" role="img">
</i> 
 <span class="menu-text">Bookdown</span></a>
  </li>  
</ul>
<div class="quarto-navbar-tools tools-wide">
    <a href="https://github.com/djhcod/r-notes" rel="" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
    <div class="dropdown">
      <a href="" title="Share" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Share"><i class="bi bi-share"></i></a>
      <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="quarto-navigation-tool-dropdown-0">
<li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://twitter.com/intent/tweet?url=%7Curl%7C">
              <i class="bi bi-bi-twitter pe-1"></i>
            Twitter
            </a>
          </li>
          <li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://www.facebook.com/sharer/sharer.php?u=%7Curl%7C">
              <i class="bi bi-bi-facebook pe-1"></i>
            Facebook
            </a>
          </li>
      </ul>
</div>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../single_cell/scRNA-seq_online/00_intro.html">scRNA-seq_online学习材料</a></li><li class="breadcrumb-item"><a href="../../single_cell/scRNA-seq_online/08_SC_clustering_quality_control.html"><span class="chapter-number">24</span>&nbsp; <span class="chapter-title">细胞分群质量评估</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto"><div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">主页</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">前言</span></a>
  </div>
</li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../r_basic/r_basics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">R语言基础</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/environment_configuration.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Rstudio环境配置</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/data_input_output.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">数据的读取与输出</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/basic_data_function.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">数据处理基本函数</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/character.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">字符的处理</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../single_cell/seurat/seurat.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Seurat v5官方文档</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/seurat_command_list.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Seurat常用函数清单</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/seurat_tutorial.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Seurat细胞分群官方教程</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/data_visualization_methods_in_seurat.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Seurat中的数据可视化方法</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/sctransform.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">基于SCTransform的单细胞数据标准化</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/integration.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">单细胞数据整合（integration）</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/integrative_analysis_in_seurat_v5.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Seurat v5单细胞数据整合分析</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/mapping_and_annotating_query_datasets.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">基于参考集的细胞注释</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/cell_cycle_regression.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">消除细胞周期的影响</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/de_vignette.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">差异表达分析</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../single_cell/scRNA-seq_online/00_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">scRNA-seq_online学习材料</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/01_intro_to_scRNA-seq.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">单细胞测序技术介绍</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/02_SC_generation_of_count_matrix.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">单细胞RNA测序—从raw data到count matrix</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/03_SC_quality_control-setup.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">数据导入与Seurat对象构建</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/04_SC_quality_control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">质控</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/postQC_workflow.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">Single-cell RNA-seq Clustering Workflow</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/05_theory_of_PCA.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">PCA原理</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/06_SC_SCT_normalization.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">21</span>&nbsp; <span class="chapter-title">Normalization and regressing out unwanted variation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/06_integration.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">22</span>&nbsp; <span class="chapter-title">单细胞数据整合（Integration）</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/07_SC_clustering_cells_SCT.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">23</span>&nbsp; <span class="chapter-title">细胞聚类（clustering analysis）</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/08_SC_clustering_quality_control.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">24</span>&nbsp; <span class="chapter-title">细胞分群质量评估</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../quarto_foundation/quarto_foundation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Quarto基础</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/yaml_settings.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">25</span>&nbsp; <span class="chapter-title">YAML设置</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/images_settings.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">26</span>&nbsp; <span class="chapter-title">图片设置</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/cross_references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">27</span>&nbsp; <span class="chapter-title">交叉引用</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/insert_other_content.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">28</span>&nbsp; <span class="chapter-title">插入其他内容</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/quarto_books.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">29</span>&nbsp; <span class="chapter-title">Quarto Books</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/github_pages.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">30</span>&nbsp; <span class="chapter-title">发布到GitHub Pages</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="false">
 <span class="menu-text">参考文献</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active" data-toc-expanded="1"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#exploration-of-quality-control-metrics" id="toc-exploration-of-quality-control-metrics" class="nav-link active" data-scroll-target="#exploration-of-quality-control-metrics"><span class="header-section-number">24.1</span> Exploration of quality control metrics</a></li>
  <li><a href="#%E6%95%B0%E6%8D%AE%E8%AF%BB%E5%8F%96" id="toc-数据读取" class="nav-link" data-scroll-target="#%E6%95%B0%E6%8D%AE%E8%AF%BB%E5%8F%96"><span class="header-section-number">24.2</span> 数据读取</a></li>
  <li><a href="#%E5%88%86%E6%9E%90%E6%A0%B7%E6%9C%AC%E7%B1%BB%E5%9E%8B%E6%98%AF%E5%90%A6%E5%BD%B1%E5%93%8D%E7%BB%86%E8%83%9E%E5%88%86%E7%BE%A4" id="toc-分析样本类型是否影响细胞分群" class="nav-link" data-scroll-target="#%E5%88%86%E6%9E%90%E6%A0%B7%E6%9C%AC%E7%B1%BB%E5%9E%8B%E6%98%AF%E5%90%A6%E5%BD%B1%E5%93%8D%E7%BB%86%E8%83%9E%E5%88%86%E7%BE%A4"><span class="header-section-number">24.3</span> 分析样本类型是否影响细胞分群</a></li>
  <li><a href="#%E5%88%86%E6%9E%90%E7%BB%86%E8%83%9E%E5%91%A8%E6%9C%9F%E6%98%AF%E5%90%A6%E5%BD%B1%E5%93%8D%E7%BB%86%E8%83%9E%E5%88%86%E7%BE%A4" id="toc-分析细胞周期是否影响细胞分群" class="nav-link" data-scroll-target="#%E5%88%86%E6%9E%90%E7%BB%86%E8%83%9E%E5%91%A8%E6%9C%9F%E6%98%AF%E5%90%A6%E5%BD%B1%E5%93%8D%E7%BB%86%E8%83%9E%E5%88%86%E7%BE%A4"><span class="header-section-number">24.4</span> 分析细胞周期是否影响细胞分群</a></li>
  <li><a href="#%E5%88%86%E6%9E%90%E5%85%B6%E4%BB%96%E9%9D%9E%E6%9C%9F%E6%9C%9B%E5%8F%98%E5%BC%82%E6%9D%A5%E6%BA%90%E6%98%AF%E5%90%A6%E4%BC%9A%E5%BD%B1%E5%93%8D%E7%BB%86%E8%83%9E%E5%88%86%E7%BE%A4" id="toc-分析其他非期望变异来源是否会影响细胞分群" class="nav-link" data-scroll-target="#%E5%88%86%E6%9E%90%E5%85%B6%E4%BB%96%E9%9D%9E%E6%9C%9F%E6%9C%9B%E5%8F%98%E5%BC%82%E6%9D%A5%E6%BA%90%E6%98%AF%E5%90%A6%E4%BC%9A%E5%BD%B1%E5%93%8D%E7%BB%86%E8%83%9E%E5%88%86%E7%BE%A4"><span class="header-section-number">24.5</span> 分析其他非期望变异来源是否会影响细胞分群</a></li>
  <li><a href="#%E5%88%86%E6%9E%90%E4%B8%BB%E6%88%90%E5%88%86pcs%E5%AF%B9%E7%BB%86%E8%83%9E%E5%88%86%E7%BE%A4%E7%9A%84%E5%BD%B1%E5%93%8D" id="toc-分析主成分pcs对细胞分群的影响" class="nav-link" data-scroll-target="#%E5%88%86%E6%9E%90%E4%B8%BB%E6%88%90%E5%88%86pcs%E5%AF%B9%E7%BB%86%E8%83%9E%E5%88%86%E7%BE%A4%E7%9A%84%E5%BD%B1%E5%93%8D"><span class="header-section-number">24.6</span> 分析主成分（PCs）对细胞分群的影响</a></li>
  <li>
<a href="#%E6%8E%A2%E7%B4%A2%E5%B7%B2%E7%9F%A5%E7%9A%84cell-type-markers%E7%9A%84%E8%A1%A8%E8%BE%BE" id="toc-探索已知的cell-type-markers的表达" class="nav-link" data-scroll-target="#%E6%8E%A2%E7%B4%A2%E5%B7%B2%E7%9F%A5%E7%9A%84cell-type-markers%E7%9A%84%E8%A1%A8%E8%BE%BE"><span class="header-section-number">24.7</span> 探索已知的cell type markers的表达</a>
  <ul class="collapse">
<li><a href="#cd14-monocyte-markers" id="toc-cd14-monocyte-markers" class="nav-link" data-scroll-target="#cd14-monocyte-markers">CD14+ monocyte markers</a></li>
  <li><a href="#fcgr3a-monocyte-markers" id="toc-fcgr3a-monocyte-markers" class="nav-link" data-scroll-target="#fcgr3a-monocyte-markers">FCGR3A+ monocyte markers</a></li>
  <li><a href="#macrophages" id="toc-macrophages" class="nav-link" data-scroll-target="#macrophages">Macrophages</a></li>
  <li><a href="#conventional-dendritic-cell-markers" id="toc-conventional-dendritic-cell-markers" class="nav-link" data-scroll-target="#conventional-dendritic-cell-markers">Conventional dendritic cell markers</a></li>
  <li><a href="#plasmacytoid-dendritic-cell-markers" id="toc-plasmacytoid-dendritic-cell-markers" class="nav-link" data-scroll-target="#plasmacytoid-dendritic-cell-markers">Plasmacytoid dendritic cell markers</a></li>
  </ul>
</li>
  </ul><div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/djhcod/r-notes/blob/main/single_cell/scRNA-seq_online/08_SC_clustering_quality_control.qmd" class="toc-action">View source</a></p><p><a href="https://github.com/djhcod/r-notes/issues/new" class="toc-action">Report an issue</a></p><p><a href="https://github.com/djhcod/r-notes/edit/main/single_cell/scRNA-seq_online/08_SC_clustering_quality_control.qmd" class="toc-action">Edit this page</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title">
<span class="chapter-number">24</span>&nbsp; <span class="chapter-title">细胞分群质量评估</span>
</h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Learning Objectives:
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>Evaluate whether clustering artifacts are present</li>
<li>Determine the quality of clustering with PCA and UMAP plots, and decide when to re-cluster</li>
<li>Assess known cell type markers to hypothesize cell type identities of clusters</li>
</ul>
</div>
</div>
<p>After separating cells into clusters, it is crtical to evaluate whether they are <strong>biologically meaningful</strong> or not. At this point we can also decide <strong>if we need to re-cluster and/or potentialy go back to a previous QC step</strong>.<br>
In this lesson you will:</p>
<ul>
<li><p>Check to see that clusters are not influenced by uninteresting sources of variation</p></li>
<li><p>Check to see whether the major principal components are driving the different clusters</p></li>
<li><p>Explore the cell type identities by looking at the expression for known markers across the clusters.</p></li>
</ul>
<section id="exploration-of-quality-control-metrics" class="level2" data-number="24.1"><h2 data-number="24.1" class="anchored" data-anchor-id="exploration-of-quality-control-metrics">
<span class="header-section-number">24.1</span> Exploration of quality control metrics</h2>
<p>To determine whether our clusters might be due to artifacts such as cell cycle phase or mitochondrial expression, it can be useful to explore these metrics visually to see if any clusters exhibit enrichment or are different from the other clusters. However, if enrichment or differences are observed for particular clusters it may not be worrisome if it can be explained by the cell type.</p>
<p>To explore and visualize the various quality metrics, we will use the versatile <code><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot()</a></code> and <code><a href="https://satijalab.org/seurat/reference/FeaturePlot.html">FeaturePlot()</a></code> functions from Seurat.</p>
</section><section id="数据读取" class="level2" data-number="24.2"><h2 data-number="24.2" class="anchored" data-anchor-id="数据读取">
<span class="header-section-number">24.2</span> 数据读取</h2>
<p>载入上一节中完成细胞分群的数据<code>seurat_integrated</code>。</p>
<div class="cell" data-hash="08_SC_clustering_quality_control_cache/html/unnamed-chunk-1_5fbda7c35c54d9ff18594da5112d2cd9">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span></span>
<span><span class="va">seurat_integrated</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="st">"output/scRNA-seq_online/seurat_clustered.rds"</span><span class="op">)</span></span>
<span><span class="va">seurat_integrated</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
31130 features across 29629 samples within 3 assays 
Active assay: integrated (3000 features, 3000 variable features)
 2 layers present: data, scale.data
 2 other assays present: RNA, SCT
 2 dimensional reductions calculated: pca, umap</code></pre>
</div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">seurat_integrated</span>, <span class="fl">5</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                      orig.ident nCount_RNA nFeature_RNA
ctrl_AAACATACAATGCC-1       ctrl       2344          874
ctrl_AAACATACATTTCC-1       ctrl       3124          895
ctrl_AAACATACCAGAAA-1       ctrl       2578          725
ctrl_AAACATACCAGCTA-1       ctrl       3260          978
ctrl_AAACATACCATGCA-1       ctrl        746          362
                                      seq_folder nUMI nGene log10GenesPerUMI
ctrl_AAACATACAATGCC-1 ctrl_raw_feature_bc_matrix 2344   874        0.8728630
ctrl_AAACATACATTTCC-1 ctrl_raw_feature_bc_matrix 3125   896        0.8447596
ctrl_AAACATACCAGAAA-1 ctrl_raw_feature_bc_matrix 2578   725        0.8384933
ctrl_AAACATACCAGCTA-1 ctrl_raw_feature_bc_matrix 3261   979        0.8512622
ctrl_AAACATACCATGCA-1 ctrl_raw_feature_bc_matrix  746   362        0.8906861
                       mitoRatio                 cells sample     S.Score
ctrl_AAACATACAATGCC-1 0.01962457 ctrl_AAACATACAATGCC-1   ctrl  0.04330502
ctrl_AAACATACATTTCC-1 0.01792000 ctrl_AAACATACATTTCC-1   ctrl  0.02661900
ctrl_AAACATACCAGAAA-1 0.01551590 ctrl_AAACATACCAGAAA-1   ctrl -0.04670650
ctrl_AAACATACCAGCTA-1 0.01379945 ctrl_AAACATACCAGCTA-1   ctrl -0.05832833
ctrl_AAACATACCATGCA-1 0.02144772 ctrl_AAACATACCATGCA-1   ctrl  0.03929605
                        G2M.Score Phase      mitoFr nCount_SCT nFeature_SCT
ctrl_AAACATACAATGCC-1  0.05422631   G2M      Medium       1572          829
ctrl_AAACATACATTTCC-1  0.05159679   G2M      Medium       1572          718
ctrl_AAACATACCAGAAA-1 -0.04841661    G1      Medium       1553          648
ctrl_AAACATACCAGCTA-1  0.05045960   G2M         Low       1576          756
ctrl_AAACATACCATGCA-1 -0.02995512     S Medium high       1075          363
                      integrated_snn_res.0.4 integrated_snn_res.0.6
ctrl_AAACATACAATGCC-1                      2                      1
ctrl_AAACATACATTTCC-1                      0                      2
ctrl_AAACATACCAGAAA-1                      0                      3
ctrl_AAACATACCAGCTA-1                      0                      3
ctrl_AAACATACCATGCA-1                      5                      6
                      integrated_snn_res.0.8 integrated_snn_res.1
ctrl_AAACATACAATGCC-1                      2                    2
ctrl_AAACATACATTTCC-1                      1                    0
ctrl_AAACATACCAGAAA-1                      3                   15
ctrl_AAACATACCAGCTA-1                      3                    3
ctrl_AAACATACCATGCA-1                      4                   12
                      integrated_snn_res.1.4 seurat_clusters
ctrl_AAACATACAATGCC-1                      5               5
ctrl_AAACATACATTTCC-1                      0               0
ctrl_AAACATACCAGAAA-1                     19              19
ctrl_AAACATACCAGCTA-1                      3               3
ctrl_AAACATACCATGCA-1                     13              13</code></pre>
</div>
</div>
</section><section id="分析样本类型是否影响细胞分群" class="level2" data-number="24.3"><h2 data-number="24.3" class="anchored" data-anchor-id="分析样本类型是否影响细胞分群">
<span class="header-section-number">24.3</span> 分析样本类型是否影响细胞分群</h2>
<p>首先通过UMAP图，直观查看不同样本类型的细胞分群情况：</p>
<div class="cell" data-hash="08_SC_clustering_quality_control_cache/html/unnamed-chunk-2_b36cc45dc4e69663bc8108122c6cd527">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 先简单查看一下分群情况及不同cluster的细胞数</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">seurat_integrated</span><span class="op">@</span><span class="va">active.ident</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
   0    1    2    3    4    5    6    7    8    9   10   11   12   13   14   15 
4220 3718 3649 3004 2164 1959 1810 1646 1504 1375 1208 1174  858  468  459  289 
  16 
 124 </code></pre>
</div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># UMAP of cells in each cluster by sample</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">seurat_integrated</span>, </span>
<span>        label <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>        split.by <span class="op">=</span> <span class="st">"sample"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html">NoLegend</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="08_SC_clustering_quality_control_files/figure-html/unnamed-chunk-2-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<blockquote class="blockquote">
<p>从UMAP图上可以直观的发现不同样本（ctrl vs.&nbsp;stim）的细胞分群非常一致，这是符合预期的。</p>
</blockquote>
<p>接下来通过提取不同样本类型中各cluster的细胞数来以数据的形式验证这种一致性。下面是实现方法：</p>
<ol type="1">
<li>
<p>通过<code>Seurat</code>包的<code><a href="https://mojaveazure.github.io/seurat-object/reference/FetchData.html">FetchData()</a></code>函数可以从Seurat对象中提取指定的变量并形成行为细胞列为变量的数据框。这里通过关键词“ident”提取“active.ident”（即每个细胞所在的cluster的编号）；并提取”orig.ident”变量，即每个细胞对应的样本类型（ctrl vs.&nbsp;stim）（也可以提取“sample”）</p>
<p><img src="images/截屏2023-12-20 10.10.43.png" class="img-fluid" width="313"></p>
</li>
<li>
<p>然后通过<code>dplyr</code>包的<code><a href="https://dplyr.tidyverse.org/reference/count.html">count()</a></code>函数统计每个样本类型内每个cluster的细胞数量</p>
<p><img src="images/截屏2023-12-20 10.10.22.png" class="img-fluid" width="278"></p>
</li>
<li>
<p>最后，通过<code>tidyr</code>包的<code><a href="https://tidyr.tidyverse.org/reference/pivot_wider.html">pivot_wider()</a></code>函数将长数据转换成宽数据。其中的<code>names_from</code>参数用于指定原数据中要拆分的那一列的名字（会在转换后的数据中变成列名）；<code>values_from</code>参数用于指定新数据集中单元格的值由旧数据的哪个（或哪些）变量的值填充</p>
<p><img src="images/截屏2023-12-20 10.43.50.png" class="img-fluid"></p>
</li>
</ol>
<div class="cell" data-hash="08_SC_clustering_quality_control_cache/html/unnamed-chunk-3_849294b3a3f45a2aa65cc17696860772">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org">tidyr</a></span><span class="op">)</span></span>
<span><span class="va">n_cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/FetchData.html">FetchData</a></span><span class="op">(</span><span class="va">seurat_integrated</span>, </span>
<span>                     vars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ident"</span>, <span class="st">"orig.ident"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>        <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html">count</a></span><span class="op">(</span><span class="va">ident</span>, <span class="va">orig.ident</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>        <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_wider.html">pivot_wider</a></span><span class="op">(</span>names_from <span class="op">=</span> <span class="va">ident</span>,</span>
<span>                    values_from <span class="op">=</span> <span class="va">n</span><span class="op">)</span></span>
<span><span class="va">n_cells</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 18
  orig.ident   `0`   `1`   `2`   `3`   `4`   `5`   `6`   `7`   `8`   `9`  `10`
  &lt;chr&gt;      &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;
1 ctrl        2017  2398  1840  1076  1112   976   952   830   775   661   600
2 stim        2203  1320  1809  1928  1052   983   858   816   729   714   608
# ℹ 6 more variables: `11` &lt;int&gt;, `12` &lt;int&gt;, `13` &lt;int&gt;, `14` &lt;int&gt;,
#   `15` &lt;int&gt;, `16` &lt;int&gt;</code></pre>
</div>
</div>
<blockquote class="blockquote">
<p>从该表中可以看出，不同样本类型下的各细胞群的细胞数量基本一致。These clusters look pretty similar between conditions, which is good since we expected similar cell types to be present in both control and stimulated conditions.</p>
</blockquote>
<div class="callout callout-style-default callout-caution callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Caution
</div>
</div>
<div class="callout-body-container callout-body">
<p>Generally, we expect to see the majority of the cell type clusters to be present in all conditions; however, depending on the experiment we might expect to see some <strong>condition-specific cell types</strong> present.</p>
</div>
</div>
</section><section id="分析细胞周期是否影响细胞分群" class="level2" data-number="24.4"><h2 data-number="24.4" class="anchored" data-anchor-id="分析细胞周期是否影响细胞分群">
<span class="header-section-number">24.4</span> 分析细胞周期是否影响细胞分群</h2>
<p>Next, we can explore whether the <strong>cells cluster influenced by the different cell cycle phases</strong>. We did not regress out variation due to cell cycle phase when we performed the <code>SCTransform</code> normalization ( <a href="06_SC_SCT_normalization.html#sec-evaluating_effects_of_cell_cycle"><span>Section&nbsp;21.2.2</span></a> ). <strong>If our cell clusters showed large differences in cell cycle expression, this would be an indication we would want to re-run the <code>SCTransform</code> and add the <code>S.Score</code> and <code>G2M.Score</code> to our variables to regress, then re-run the rest of the steps.</strong></p>
<div class="cell" data-hash="08_SC_clustering_quality_control_cache/html/unnamed-chunk-4_c62fdbe75fc87881ca4c6269ed800f32">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Explore whether clusters segregate by cell cycle phase</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">seurat_integrated</span>,</span>
<span>        label <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>        split.by <span class="op">=</span> <span class="st">"Phase"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html">NoLegend</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="08_SC_clustering_quality_control_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid" width="960"></p>
</div>
</div>
<blockquote class="blockquote">
<p>We do not see much clustering by cell cycle score, so we can proceed with the QC.</p>
</blockquote>
</section><section id="分析其他非期望变异来源是否会影响细胞分群" class="level2" data-number="24.5"><h2 data-number="24.5" class="anchored" data-anchor-id="分析其他非期望变异来源是否会影响细胞分群">
<span class="header-section-number">24.5</span> 分析其他非期望变异来源是否会影响细胞分群</h2>
<p>Next we will explore additional metrics, such as the <strong>number of UMIs</strong> and <strong>genes per cell</strong>, <strong>S-phase and G2M-phase markers</strong>, and <strong>mitochondrial gene expression</strong> by UMAP. Looking at the individual S and G2M scores can give us additional information to checking the phase as we did previously.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>The <code>order</code> argument will plot the positive cells above the negative cells, while the <code>min.cutoff</code> argument will determine the threshold for shading. A <code>min.cutoff</code> of <code>q10</code> translates to the 10% of cells with the lowest expression of the gene will not exhibit any purple shading (completely gray) (<code>min.cutoff</code>定义着色阈值。这里指定值最低的10%的细胞不着色).</p>
</div>
</div>
<div class="cell" data-hash="08_SC_clustering_quality_control_cache/html/unnamed-chunk-5_cabe798ee4657c746cc5f80ba0fbb775">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Determine metrics to plot present in seurat_integrated@meta.data</span></span>
<span><span class="va">metrics</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"nUMI"</span>, <span class="st">"nGene"</span>, <span class="st">"S.Score"</span>, <span class="st">"G2M.Score"</span>, <span class="st">"mitoRatio"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html">FeaturePlot</a></span><span class="op">(</span><span class="va">seurat_integrated</span>, </span>
<span>            reduction <span class="op">=</span> <span class="st">"umap"</span>, </span>
<span>            features <span class="op">=</span> <span class="va">metrics</span>,</span>
<span>            pt.size <span class="op">=</span> <span class="fl">0.4</span>, </span>
<span>            order <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            min.cutoff <span class="op">=</span> <span class="st">'q10'</span>, </span>
<span>            label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="08_SC_clustering_quality_control_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid" width="960"></p>
</div>
</div>
<blockquote class="blockquote">
<p>The metrics seem to be relatively even across the clusters, with the exception of <code>nGene</code> exhibiting slightly higher values in clusters to the left of the plot. We can keep an eye on these clusters to see whether the cell types may explain the increase.</p>
<p>If we see differences corresponding to any of these metrics at this point in time, then we will often note them and then decide after identifying the cell type identities whether to take any further action.</p>
</blockquote>
</section><section id="分析主成分pcs对细胞分群的影响" class="level2" data-number="24.6"><h2 data-number="24.6" class="anchored" data-anchor-id="分析主成分pcs对细胞分群的影响">
<span class="header-section-number">24.6</span> 分析主成分（PCs）对细胞分群的影响</h2>
<p>We can also explore how well our clusters separate by the different PCs; <strong>we hope that the defined PCs separate the cell types well</strong>. To visualize this information, we need to extract the UMAP coordinate information for the cells along with their corresponding scores for each of the PCs to view by UMAP.</p>
<p>First, we identify the information we would like to extract from the Seurat object, then, we can use the <code><a href="https://mojaveazure.github.io/seurat-object/reference/FetchData.html">FetchData()</a></code> function to extract it.</p>
<div class="cell" data-hash="08_SC_clustering_quality_control_cache/html/unnamed-chunk-6_81c58ec7d97ad695795e6833cffcb3e6">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Defining the information in the seurat object of interest</span></span>
<span><span class="va">columns</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ident"</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"PC_"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">16</span><span class="op">)</span>, <span class="st">"UMAP_1"</span>, <span class="st">"UMAP_2"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extracting this data from the seurat object</span></span>
<span><span class="va">pc_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/FetchData.html">FetchData</a></span><span class="op">(</span><span class="va">seurat_integrated</span>, </span>
<span>                     vars <span class="op">=</span> <span class="va">columns</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">pc_data</span>, <span class="fl">3</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                      ident      PC_1      PC_2      PC_3      PC_4       PC_5
ctrl_AAACATACAATGCC-1     2 -14.97782 -2.879193 -5.059351 -1.602766  0.8728779
ctrl_AAACATACATTTCC-1     1  22.39233 -5.296913  4.951958  3.112632  0.3379874
ctrl_AAACATACCAGAAA-1     3  28.98473  1.203408 -5.947993 -1.042701 -7.5690434
                           PC_6        PC_7       PC_8       PC_9     PC_10
ctrl_AAACATACAATGCC-1 -1.501420   0.5327841 -0.5855545  0.8866479 0.8223524
ctrl_AAACATACATTTCC-1 -7.873002   2.2740054 -5.8362430 -0.9666741 0.2170141
ctrl_AAACATACCAGAAA-1  5.477813 -10.7953185 19.5052392 -1.8312647 2.1803084
                          PC_11      PC_12      PC_13       PC_14     PC_15
ctrl_AAACATACAATGCC-1 -1.722880 -0.1832132  0.2985759 -0.04539058  1.260942
ctrl_AAACATACATTTCC-1  3.105132 -0.7964707  2.8982516  0.04362601 -3.189937
ctrl_AAACATACCAGAAA-1 -7.796582 -1.3324607 -2.3117140  3.06677862  1.663382
                           PC_16     UMAP_1    UMAP_2
ctrl_AAACATACAATGCC-1  0.1993319   7.270473 0.9072988
ctrl_AAACATACATTTCC-1  5.3805322  -8.742020 1.5622634
ctrl_AAACATACCAGAAA-1 -3.0593032 -10.032904 4.7139827</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>How did we know in the <code><a href="https://mojaveazure.github.io/seurat-object/reference/FetchData.html">FetchData()</a></code> function to include <code>UMAP_1</code> to obtain the UMAP coordinates? <a href="../seurat/seurat_command_list.html"><span>Chapter&nbsp;6</span></a> describes the function as being able to pull any data from the expression matrices, cell embeddings, or metadata.</p>
<p>For instance, if you explore the <code>seurat_integrated@reductions</code> list object, the first component is for PCA, and includes a slot for <code>cell.embeddings</code> (坐标数据). We can use the column names (<code>PC_1</code>, <code>PC_2</code>, <code>PC_3</code>, etc.) to pull out the coordinates or PC scores corresponding to each cell for each of the PCs (提取每个细胞在指定主成分上的坐标).</p>
<p>We could do the same thing for UMAP:</p>
<div class="cell" data-hash="08_SC_clustering_quality_control_cache/html/unnamed-chunk-7_16e25bad6f1e53e137ca554ae7e2595d">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 提取前5个细胞在前两个UMAP主成分上的坐标</span></span>
<span><span class="va">seurat_integrated</span><span class="op">@</span><span class="va">reductions</span><span class="op">[[</span><span class="st">"umap"</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">cell.embeddings</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                          UMAP_1     UMAP_2
ctrl_AAACATACAATGCC-1   7.270473  0.9072988
ctrl_AAACATACATTTCC-1  -8.742020  1.5622634
ctrl_AAACATACCAGAAA-1 -10.032904  4.7139827
ctrl_AAACATACCAGCTA-1  -8.363044  5.0377137
ctrl_AAACATACCATGCA-1   6.875784 -4.6442526</code></pre>
</div>
</div>
<p>等价于：</p>
<div class="cell" data-hash="08_SC_clustering_quality_control_cache/html/unnamed-chunk-8_6a41564164db91467ebf3418ae7f6ca9">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/FetchData.html">FetchData</a></span><span class="op">(</span><span class="va">seurat_integrated</span>, vars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"UMAP_1"</span>, <span class="st">"UMAP_2"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fl">5</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <code><a href="https://mojaveazure.github.io/seurat-object/reference/FetchData.html">FetchData()</a></code> function just allows us to extract the data more easily.</p>
</div>
</div>
<div class="callout callout-style-default callout-caution callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Caution
</div>
</div>
<div class="callout-body-container callout-body">
<p>The pre-existing <code>seurat_integrated</code> loaded in previously was created using an older version of Seurat. As such the columns we <code>Fetch()</code> are in upper case (i.e <code>UMAP_1</code>). <strong>If you are using your own seurat object using a newer version of Seurat you will need to change the column names as shown below.</strong> Alternatively, explore your Seurat object to see how they have been stored.</p>
<div class="cell" data-hash="08_SC_clustering_quality_control_cache/html/unnamed-chunk-9_d3000e458ca21f8e23b3f3a9e089d654">
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span> <span class="co"># Defining the information in the seurat object of interest</span></span>
<span> <span class="va">columns</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ident"</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"PC_"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">16</span><span class="op">)</span>, <span class="st">"umap_1"</span>, <span class="st">"umap_2"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<p>In the UMAP plots below, the cells are colored by their PC score for each respective principal component.</p>
<p>Let’s take a quick look at the top 16 PCs:</p>
<div class="cell" data-hash="08_SC_clustering_quality_control_cache/html/unnamed-chunk-10_31bfa68f7ce748009036dc2edd48a13a">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Adding cluster label to center of cluster on UMAP</span></span>
<span><span class="va">umap_label</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/FetchData.html">FetchData</a></span><span class="op">(</span><span class="va">seurat_integrated</span>, </span>
<span>                        vars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ident"</span>, <span class="st">"UMAP_1"</span>, <span class="st">"UMAP_2"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">ident</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="co"># 指定分组计算依据</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise</a></span><span class="op">(</span>UMAP1_mean <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">UMAP_1</span><span class="op">)</span>, UMAP2_mean <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">UMAP_2</span><span class="op">)</span><span class="op">)</span> <span class="co"># 根据指定的分组依据进行分组计算</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">umap_label</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 3
  ident UMAP1_mean UMAP2_mean
  &lt;fct&gt;      &lt;dbl&gt;      &lt;dbl&gt;
1 0          8.42       1.99 
2 1         -9.18       1.84 
3 2          5.51       1.68 
4 3         -8.49       3.87 
5 4          6.67      -2.58 
6 5          0.359      0.115</code></pre>
</div>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Plotting a UMAP plot for each of the PCs</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://purrr.tidyverse.org/">purrr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/">cowplot</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"PC_"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">16</span><span class="op">)</span>, </span>
<span>    <span class="kw">function</span><span class="op">(</span><span class="va">pc</span><span class="op">)</span> <span class="op">{</span></span>
<span>      <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">pc_data</span>, </span>
<span>             <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">UMAP_1</span>, <span class="va">UMAP_2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes_.html">aes_string</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">pc</span><span class="op">)</span>, </span>
<span>                   alpha <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html">scale_color_gradient</a></span><span class="op">(</span>low <span class="op">=</span> <span class="st">"grey90"</span>, </span>
<span>                             high <span class="op">=</span> <span class="st">"blue"</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html">geom_text</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">umap_label</span>, </span>
<span>                  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">ident</span>, <span class="va">UMAP1_mean</span>, <span class="va">UMAP2_mean</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="va">pc</span><span class="op">)</span></span>
<span>      <span class="op">}</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">plot_grid</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="va">.</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="08_SC_clustering_quality_control_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid" width="1920"></p>
</div>
</div>
<p>We can see how the clusters are represented by the different PCs. For instance, the genes driving <code>PC_2</code> exhibit higher expression in clusters 8 and 12. We could look back at our genes driving this PC to get an idea of what the cell types might be:</p>
<div class="cell" data-hash="08_SC_clustering_quality_control_cache/html/unnamed-chunk-11_d3faa8e052704035b36b64c7d8b5f171">
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 提取PCA信息中的第二主成分，并展示对该主成分影响最大的前5个基因名</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">seurat_integrated</span><span class="op">[[</span><span class="st">"pca"</span><span class="op">]</span><span class="op">]</span>, dims <span class="op">=</span> <span class="fl">2</span>, nfeatures <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>PC_ 2 
Positive:  GNLY, CCL5, NKG7, GZMB, FGFBP2 
Negative:  CD74, IGHM, IGKC, HLA-DRA, CD79A </code></pre>
</div>
</div>
<p>With the GNLY and NKG7 genes as positive markers of <code>PC_2</code>, <strong>we can hypothesize that clusters 8 and 12 correspond to NK cells</strong>. This just hints at what the clusters identity could be, with the identities of the clusters being determined through a combination of the PCs.</p>
<p>To truly determine the identity of the clusters and whether the <code>resolution</code> is appropriate, it is helpful to explore a handful of known gene markers for the cell types expected.</p>
</section><section id="探索已知的cell-type-markers的表达" class="level2" data-number="24.7"><h2 data-number="24.7" class="anchored" data-anchor-id="探索已知的cell-type-markers的表达">
<span class="header-section-number">24.7</span> 探索已知的cell type markers的表达</h2>
<p>With the cells clustered, we can explore the cell type identities by looking for known markers. 下面给出了本案例的cell type markers及其对应的细胞类型：</p>
<table class="table">
<thead><tr class="header">
<th style="text-align: center;">Cell Type</th>
<th style="text-align: center;">Marker</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">CD14+ monocytes</td>
<td style="text-align: center;">CD14, LYZ</td>
</tr>
<tr class="even">
<td style="text-align: center;">FCGR3A+ monocytes</td>
<td style="text-align: center;">FCGR3A, MS4A7</td>
</tr>
<tr class="odd">
<td style="text-align: center;">Conventional dendritic cells</td>
<td style="text-align: center;">FCER1A, CST3</td>
</tr>
<tr class="even">
<td style="text-align: center;">Plasmacytoid dendritic cells</td>
<td style="text-align: center;">IL3RA, GZMB, SERPINF1, ITM2C</td>
</tr>
<tr class="odd">
<td style="text-align: center;">B cells</td>
<td style="text-align: center;">CD79A, MS4A1</td>
</tr>
<tr class="even">
<td style="text-align: center;">T cells</td>
<td style="text-align: center;">CD3D</td>
</tr>
<tr class="odd">
<td style="text-align: center;">CD4+ T cells</td>
<td style="text-align: center;">CD3D, IL7R, CCR7</td>
</tr>
<tr class="even">
<td style="text-align: center;">CD8+ T cells</td>
<td style="text-align: center;">CD3D, CD8A</td>
</tr>
<tr class="odd">
<td style="text-align: center;">NK cells</td>
<td style="text-align: center;">GNLY, NKG7</td>
</tr>
<tr class="even">
<td style="text-align: center;">Megakaryocytes</td>
<td style="text-align: center;">PPBP</td>
</tr>
<tr class="odd">
<td style="text-align: center;">Erythrocytes</td>
<td style="text-align: center;">HBB, HBA2</td>
</tr>
</tbody>
</table>
<p>The <code><a href="https://satijalab.org/seurat/reference/FeaturePlot.html">FeaturePlot()</a></code> function from Seurat makes it easy to visualize a handful of genes using the gene IDs stored in the Seurat object. We can easily explore the expression of known gene markers on top of our UMAP visualizations. Let’s go through and determine the identities of the clusters. To access the normalized expression levels of all genes, we can use the normalized count data stored in the <code>RNA</code> assay slot.</p>
<div class="callout callout-style-default callout-caution callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Caution
</div>
</div>
<div class="callout-body-container callout-body">
<p>The <code>SCTransform</code> normalization was performed only on the <strong>3000</strong> most variable genes, so many of our genes of interest may not be present in this data ( <a href="06_SC_SCT_normalization.html#sec-perform_sctranform"><span>Section&nbsp;21.3.5.1</span></a> ).</p>
</div>
</div>
<p>这里为了准确可视化marker基因的表达情况，我们将默认的assay改回“RNA”，然后运行<code>NormalizeData</code>：</p>
<div class="cell" data-hash="08_SC_clustering_quality_control_cache/html/unnamed-chunk-12_de44c925fc4c262f8185478605eca2d4">
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/DefaultAssay.html">DefaultAssay</a></span><span class="op">(</span><span class="va">seurat_integrated</span><span class="op">)</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "integrated"</code></pre>
</div>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Select the RNA counts slot to be the default assay</span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/DefaultAssay.html">DefaultAssay</a></span><span class="op">(</span><span class="va">seurat_integrated</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"RNA"</span></span>
<span></span>
<span><span class="co"># Normalize RNA data for visualization purposes</span></span>
<span><span class="va">seurat_integrated</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html">NormalizeData</a></span><span class="op">(</span><span class="va">seurat_integrated</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>Assay is a slot defined in the Seurat object, it has multiple slots within it. In a given assay, the <code>counts</code> layer stores non-normalized raw counts, and the <code>data</code> layer stores normalized expression data. Therefore, when we run the <code><a href="https://satijalab.org/seurat/reference/NormalizeData.html">NormalizeData()</a></code> function in the above code, the normalized data will be stored in the <code>data</code> layer of the RNA assay while the <code>counts</code> layer will remain unaltered.</p>
</div>
</div>
<p>Depending on our markers of interest, they could be positive or negative markers for a particular cell type. The combined expression of our chosen handful of markers should give us an idea on whether a cluster corresponds to that particular cell type.</p>
<p>For the markers used here, we are looking for positive markers and consistency of expression of the markers across the clusters. For example, if there are two markers for a cell type and only one of them is expressed in a cluster - then we cannot reliably assign that cluster to the cell type.</p>
<section id="cd14-monocyte-markers" class="level3"><h3 class="anchored" data-anchor-id="cd14-monocyte-markers">CD14+ monocyte markers</h3>
<div class="cell" data-hash="08_SC_clustering_quality_control_cache/html/unnamed-chunk-13_ed99c2aa03142c9e239190951380e2d6">
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html">FeaturePlot</a></span><span class="op">(</span><span class="va">seurat_integrated</span>, </span>
<span>            reduction <span class="op">=</span> <span class="st">"umap"</span>, </span>
<span>            features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CD14"</span>, <span class="st">"LYZ"</span><span class="op">)</span>, </span>
<span>            order <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            min.cutoff <span class="op">=</span> <span class="st">'q10'</span>, </span>
<span>            label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="08_SC_clustering_quality_control_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid" width="960"></p>
</div>
</div>
<p>CD14+ monocytes appear to correspond to clusters 1, and 3. We wouldn’t include clusters 14 and 10 because they do not highly express both of these markers.</p>
</section><section id="fcgr3a-monocyte-markers" class="level3"><h3 class="anchored" data-anchor-id="fcgr3a-monocyte-markers">FCGR3A+ monocyte markers</h3>
<div class="cell" data-hash="08_SC_clustering_quality_control_cache/html/unnamed-chunk-14_3b33353ffc55e8e849c2215801f91e9a">
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html">FeaturePlot</a></span><span class="op">(</span><span class="va">seurat_integrated</span>, </span>
<span>            reduction <span class="op">=</span> <span class="st">"umap"</span>, </span>
<span>            features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"FCGR3A"</span>, <span class="st">"MS4A7"</span><span class="op">)</span>, </span>
<span>            order <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            min.cutoff <span class="op">=</span> <span class="st">'q10'</span>, </span>
<span>            label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="08_SC_clustering_quality_control_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid" width="960"></p>
</div>
</div>
<p>FCGR3A+ monocytes markers distinctly highlight cluster 10, although we do see some decent expression in clusters 1 and 3. We would like to see additional markers for FCGR3A+ cells show up when we perform the marker identification.</p>
</section><section id="macrophages" class="level3"><h3 class="anchored" data-anchor-id="macrophages">Macrophages</h3>
<div class="cell" data-hash="08_SC_clustering_quality_control_cache/html/unnamed-chunk-15_aba4978611d36eadb464d2256f6225eb">
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html">FeaturePlot</a></span><span class="op">(</span><span class="va">seurat_integrated</span>, </span>
<span>            reduction <span class="op">=</span> <span class="st">"umap"</span>, </span>
<span>            features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"MARCO"</span>, <span class="st">"ITGAM"</span>, <span class="st">"ADGRE1"</span><span class="op">)</span>, </span>
<span>            order <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            min.cutoff <span class="op">=</span> <span class="st">'q10'</span>, </span>
<span>            label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="08_SC_clustering_quality_control_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid" width="960"></p>
</div>
</div>
<p>We don’t see much overlap of our markers, so no clusters appear to correspond to macrophages; perhaps cell culture conditions negatively selected for macrophages (more highly adherent).</p>
</section><section id="conventional-dendritic-cell-markers" class="level3"><h3 class="anchored" data-anchor-id="conventional-dendritic-cell-markers">Conventional dendritic cell markers</h3>
<div class="cell" data-hash="08_SC_clustering_quality_control_cache/html/unnamed-chunk-16_a7b971ea77a0a50edddfcceac3a404ca">
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html">FeaturePlot</a></span><span class="op">(</span><span class="va">seurat_integrated</span>, </span>
<span>            reduction <span class="op">=</span> <span class="st">"umap"</span>, </span>
<span>            features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"FCER1A"</span>, <span class="st">"CST3"</span><span class="op">)</span>, </span>
<span>            order <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            min.cutoff <span class="op">=</span> <span class="st">'q10'</span>, </span>
<span>            label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="08_SC_clustering_quality_control_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid" width="960"></p>
</div>
</div>
<p>The markers corresponding to conventional dendritic cells identify cluster 14 (both markers consistently show expression).</p>
</section><section id="plasmacytoid-dendritic-cell-markers" class="level3"><h3 class="anchored" data-anchor-id="plasmacytoid-dendritic-cell-markers">Plasmacytoid dendritic cell markers</h3>
<div class="cell" data-hash="08_SC_clustering_quality_control_cache/html/unnamed-chunk-17_75813b828dff7d5f693005d9bd0983f9">
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html">FeaturePlot</a></span><span class="op">(</span><span class="va">seurat_integrated</span>, </span>
<span>            reduction <span class="op">=</span> <span class="st">"umap"</span>, </span>
<span>            features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"IL3RA"</span>, <span class="st">"GZMB"</span>, <span class="st">"SERPINF1"</span>, <span class="st">"ITM2C"</span><span class="op">)</span>, </span>
<span>            order <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            min.cutoff <span class="op">=</span> <span class="st">'q10'</span>, </span>
<span>            label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="08_SC_clustering_quality_control_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid" width="960"></p>
</div>
</div>
<p>Plasmacytoid dendritic cells represent cluster 16. While there are a lot of differences in the expression of these markers, we see cluster 16 (though small) is consistently strongly expressed.</p>
<hr>
<p><strong>Exercise</strong></p>
<p>Hypothesize the clusters corresponding to each of the different clusters in the table:</p>
<table class="table">
<thead><tr class="header">
<th style="text-align: center;">Cell Type</th>
<th style="text-align: center;">Clusters</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">CD14+ monocytes</td>
<td style="text-align: center;">1, 3</td>
</tr>
<tr class="even">
<td style="text-align: center;">FCGR3A+ monocytes</td>
<td style="text-align: center;">10</td>
</tr>
<tr class="odd">
<td style="text-align: center;">Conventional dendritic cells</td>
<td style="text-align: center;">14</td>
</tr>
<tr class="even">
<td style="text-align: center;">Plasmacytoid dendritic cells</td>
<td style="text-align: center;">16</td>
</tr>
<tr class="odd">
<td style="text-align: center;">Marcrophages</td>
<td style="text-align: center;">-</td>
</tr>
<tr class="even">
<td style="text-align: center;">B cells</td>
<td style="text-align: center;">?</td>
</tr>
<tr class="odd">
<td style="text-align: center;">T cells</td>
<td style="text-align: center;">?</td>
</tr>
<tr class="even">
<td style="text-align: center;">CD4+ T cells</td>
<td style="text-align: center;">?</td>
</tr>
<tr class="odd">
<td style="text-align: center;">CD8+ T cells</td>
<td style="text-align: center;">?</td>
</tr>
<tr class="even">
<td style="text-align: center;">NK cells</td>
<td style="text-align: center;">?</td>
</tr>
<tr class="odd">
<td style="text-align: center;">Megakaryocytes</td>
<td style="text-align: center;">?</td>
</tr>
<tr class="even">
<td style="text-align: center;">Erythrocytes</td>
<td style="text-align: center;">?</td>
</tr>
<tr class="odd">
<td style="text-align: center;">Unknown</td>
<td style="text-align: center;">?</td>
</tr>
</tbody>
</table>
<hr>
<blockquote class="blockquote">
<p><strong>NOTE:</strong> If any cluster appears to contain two separate cell types, it’s helpful to increase our clustering resolution to properly subset the clusters. Alternatively, if we still can’t separate out the clusters using increased resolution, then it’s possible that we had used too few principal components such that we are just not separating out these cell types of interest. To inform our choice of PCs, we could look at our PC gene expression overlapping the UMAP plots and determine whether our cell populations are separating by the PCs included.</p>
</blockquote>
<p>Now we have a decent idea as to the cell types corresponding to the majority of the clusters, but some questions remain:</p>
<ol type="1">
<li><em>T cell markers appear to be highly expressed in may clusters. How can we differentiate and subset the larger group into smaller subset of cells?</em></li>
<li><em>Do the clusters corresponding to the same cell types have biologically meaningful differences? Are there subpopulations of these cell types?</em></li>
<li><em>Can we acquire higher confidence in these cell type identities by identifying other marker genes for these clusters?</em></li>
</ol>
<p>Marker identification analysis can help us address all of these questions!!</p>
<p>The next step will be to perform marker identification analysis, which will output the genes that significantly differ in expression between clusters. Using these genes we can determine or improve confidence in the identities of the clusters/subclusters.</p>
<hr>


<!-- -->

</section></section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  let localAlternateSentinel = 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var filterRegex = new RegExp("https:\/\/djhcod\.github\.io\/r-notes\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
          // target, if specified
          link.setAttribute("target", "_blank");
      }
    }
});
</script><script src="https://utteranc.es/client.js" repo="https://github.com/djhcod/r-notes" issue-term="pathname" theme="github-light" crossorigin="anonymous" async="">
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="../../single_cell/scRNA-seq_online/07_SC_clustering_cells_SCT.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">23</span>&nbsp; <span class="chapter-title">细胞聚类（clustering analysis）</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../../quarto_foundation/quarto_foundation.html" class="pagination-link">
        <span class="nav-page-text">Quarto基础</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb31" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu"># 细胞分群质量评估</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a><span class="fu">###### Learning Objectives:</span></span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Evaluate whether clustering artifacts are present</span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Determine the quality of clustering with PCA and UMAP plots, and decide when to re-cluster</span>
<span id="cb31-8"><a href="#cb31-8" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Assess known cell type markers to hypothesize cell type identities of clusters</span>
<span id="cb31-9"><a href="#cb31-9" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb31-10"><a href="#cb31-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-11"><a href="#cb31-11" aria-hidden="true" tabindex="-1"></a>After separating cells into clusters, it is crtical to evaluate whether they are **biologically meaningful** or not. At this point we can also decide **if we need to re-cluster and/or potentialy go back to a previous QC step**.\</span>
<span id="cb31-12"><a href="#cb31-12" aria-hidden="true" tabindex="-1"></a>In this lesson you will:</span>
<span id="cb31-13"><a href="#cb31-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-14"><a href="#cb31-14" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Check to see that clusters are not influenced by uninteresting sources of variation</span>
<span id="cb31-15"><a href="#cb31-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-16"><a href="#cb31-16" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Check to see whether the major principal components are driving the different clusters</span>
<span id="cb31-17"><a href="#cb31-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-18"><a href="#cb31-18" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Explore the cell type identities by looking at the expression for known markers across the clusters.</span>
<span id="cb31-19"><a href="#cb31-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-20"><a href="#cb31-20" aria-hidden="true" tabindex="-1"></a><span class="fu">## Exploration of quality control metrics</span></span>
<span id="cb31-21"><a href="#cb31-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-22"><a href="#cb31-22" aria-hidden="true" tabindex="-1"></a>To determine whether our clusters might be due to artifacts such as cell cycle phase or mitochondrial expression, it can be useful to explore these metrics visually to see if any clusters exhibit enrichment or are different from the other clusters. However, if enrichment or differences are observed for particular clusters it may not be worrisome if it can be explained by the cell type.</span>
<span id="cb31-23"><a href="#cb31-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-24"><a href="#cb31-24" aria-hidden="true" tabindex="-1"></a>To explore and visualize the various quality metrics, we will use the versatile <span class="in">`DimPlot()`</span> and <span class="in">`FeaturePlot()`</span> functions from Seurat.</span>
<span id="cb31-25"><a href="#cb31-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-26"><a href="#cb31-26" aria-hidden="true" tabindex="-1"></a><span class="fu">## 数据读取</span></span>
<span id="cb31-27"><a href="#cb31-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-28"><a href="#cb31-28" aria-hidden="true" tabindex="-1"></a>载入上一节中完成细胞分群的数据<span class="in">`seurat_integrated`</span>。</span>
<span id="cb31-29"><a href="#cb31-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-32"><a href="#cb31-32" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb31-33"><a href="#cb31-33" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb31-34"><a href="#cb31-34" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb31-35"><a href="#cb31-35" aria-hidden="true" tabindex="-1"></a>seurat_integrated <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"output/scRNA-seq_online/seurat_clustered.rds"</span>)</span>
<span id="cb31-36"><a href="#cb31-36" aria-hidden="true" tabindex="-1"></a>seurat_integrated</span>
<span id="cb31-37"><a href="#cb31-37" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(seurat_integrated, <span class="dv">5</span>)</span>
<span id="cb31-38"><a href="#cb31-38" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-39"><a href="#cb31-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-40"><a href="#cb31-40" aria-hidden="true" tabindex="-1"></a><span class="fu">## 分析样本类型是否影响细胞分群</span></span>
<span id="cb31-41"><a href="#cb31-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-42"><a href="#cb31-42" aria-hidden="true" tabindex="-1"></a>首先通过UMAP图，直观查看不同样本类型的细胞分群情况：</span>
<span id="cb31-43"><a href="#cb31-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-46"><a href="#cb31-46" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb31-47"><a href="#cb31-47" aria-hidden="true" tabindex="-1"></a><span class="co"># 先简单查看一下分群情况及不同cluster的细胞数</span></span>
<span id="cb31-48"><a href="#cb31-48" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(seurat_integrated<span class="sc">@</span>active.ident)</span>
<span id="cb31-49"><a href="#cb31-49" aria-hidden="true" tabindex="-1"></a><span class="co"># UMAP of cells in each cluster by sample</span></span>
<span id="cb31-50"><a href="#cb31-50" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(seurat_integrated, </span>
<span id="cb31-51"><a href="#cb31-51" aria-hidden="true" tabindex="-1"></a>        <span class="at">label =</span> <span class="cn">TRUE</span>, </span>
<span id="cb31-52"><a href="#cb31-52" aria-hidden="true" tabindex="-1"></a>        <span class="at">split.by =</span> <span class="st">"sample"</span>) <span class="sc">+</span> </span>
<span id="cb31-53"><a href="#cb31-53" aria-hidden="true" tabindex="-1"></a>  <span class="fu">NoLegend</span>()</span>
<span id="cb31-54"><a href="#cb31-54" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-55"><a href="#cb31-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-56"><a href="#cb31-56" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; 从UMAP图上可以直观的发现不同样本（ctrl vs. stim）的细胞分群非常一致，这是符合预期的。</span></span>
<span id="cb31-57"><a href="#cb31-57" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-58"><a href="#cb31-58" aria-hidden="true" tabindex="-1"></a>接下来通过提取不同样本类型中各cluster的细胞数来以数据的形式验证这种一致性。下面是实现方法：</span>
<span id="cb31-59"><a href="#cb31-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-60"><a href="#cb31-60" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>通过<span class="in">`Seurat`</span>包的<span class="in">`FetchData()`</span>函数可以从Seurat对象中提取指定的变量并形成行为细胞列为变量的数据框。这里通过关键词“ident”提取“active.ident”（即每个细胞所在的cluster的编号）；并提取"orig.ident"变量，即每个细胞对应的样本类型（ctrl vs. stim）（也可以提取“sample”）</span>
<span id="cb31-61"><a href="#cb31-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-62"><a href="#cb31-62" aria-hidden="true" tabindex="-1"></a>    <span class="al">![](images/截屏2023-12-20%2010.10.43.png)</span>{width="313"}</span>
<span id="cb31-63"><a href="#cb31-63" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-64"><a href="#cb31-64" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>然后通过<span class="in">`dplyr`</span>包的<span class="in">`count()`</span>函数统计每个样本类型内每个cluster的细胞数量</span>
<span id="cb31-65"><a href="#cb31-65" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-66"><a href="#cb31-66" aria-hidden="true" tabindex="-1"></a>    <span class="al">![](images/截屏2023-12-20%2010.10.22.png)</span>{width="278"}</span>
<span id="cb31-67"><a href="#cb31-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-68"><a href="#cb31-68" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>最后，通过<span class="in">`tidyr`</span>包的<span class="in">`pivot_wider()`</span>函数将长数据转换成宽数据。其中的<span class="in">`names_from`</span>参数用于指定原数据中要拆分的那一列的名字（会在转换后的数据中变成列名）；<span class="in">`values_from`</span>参数用于指定新数据集中单元格的值由旧数据的哪个（或哪些）变量的值填充</span>
<span id="cb31-69"><a href="#cb31-69" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-70"><a href="#cb31-70" aria-hidden="true" tabindex="-1"></a>    <span class="al">![](images/截屏2023-12-20%2010.43.50.png)</span></span>
<span id="cb31-71"><a href="#cb31-71" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-74"><a href="#cb31-74" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb31-75"><a href="#cb31-75" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb31-76"><a href="#cb31-76" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyr)</span>
<span id="cb31-77"><a href="#cb31-77" aria-hidden="true" tabindex="-1"></a>n_cells <span class="ot">&lt;-</span> <span class="fu">FetchData</span>(seurat_integrated, </span>
<span id="cb31-78"><a href="#cb31-78" aria-hidden="true" tabindex="-1"></a>                     <span class="at">vars =</span> <span class="fu">c</span>(<span class="st">"ident"</span>, <span class="st">"orig.ident"</span>)) <span class="sc">|&gt;</span></span>
<span id="cb31-79"><a href="#cb31-79" aria-hidden="true" tabindex="-1"></a>        <span class="fu">count</span>(ident, orig.ident) <span class="sc">|&gt;</span></span>
<span id="cb31-80"><a href="#cb31-80" aria-hidden="true" tabindex="-1"></a>        <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> ident,</span>
<span id="cb31-81"><a href="#cb31-81" aria-hidden="true" tabindex="-1"></a>                    <span class="at">values_from =</span> n)</span>
<span id="cb31-82"><a href="#cb31-82" aria-hidden="true" tabindex="-1"></a>n_cells</span>
<span id="cb31-83"><a href="#cb31-83" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-84"><a href="#cb31-84" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-85"><a href="#cb31-85" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; 从该表中可以看出，不同样本类型下的各细胞群的细胞数量基本一致。These clusters look pretty similar between conditions, which is good since we expected similar cell types to be present in both control and stimulated conditions.</span></span>
<span id="cb31-86"><a href="#cb31-86" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-87"><a href="#cb31-87" aria-hidden="true" tabindex="-1"></a>::: callout-caution</span>
<span id="cb31-88"><a href="#cb31-88" aria-hidden="true" tabindex="-1"></a>Generally, we expect to see the majority of the cell type clusters to be present in all conditions; however, depending on the experiment we might expect to see some **condition-specific cell types** present.</span>
<span id="cb31-89"><a href="#cb31-89" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb31-90"><a href="#cb31-90" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-91"><a href="#cb31-91" aria-hidden="true" tabindex="-1"></a><span class="fu">## 分析细胞周期是否影响细胞分群</span></span>
<span id="cb31-92"><a href="#cb31-92" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-93"><a href="#cb31-93" aria-hidden="true" tabindex="-1"></a>Next, we can explore whether the **cells cluster influenced by the different cell cycle phases**. We did not regress out variation due to cell cycle phase when we performed the `SCTransform` normalization ( @sec-evaluating_effects_of_cell_cycle ). **If our cell clusters showed large differences in cell cycle expression, this would be an indication we would want to re-run the `SCTransform` and add the `S.Score` and `G2M.Score` to our variables to regress, then re-run the rest of the steps.**</span>
<span id="cb31-94"><a href="#cb31-94" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-97"><a href="#cb31-97" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb31-98"><a href="#cb31-98" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 10</span></span>
<span id="cb31-99"><a href="#cb31-99" aria-hidden="true" tabindex="-1"></a><span class="co"># Explore whether clusters segregate by cell cycle phase</span></span>
<span id="cb31-100"><a href="#cb31-100" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(seurat_integrated,</span>
<span id="cb31-101"><a href="#cb31-101" aria-hidden="true" tabindex="-1"></a>        <span class="at">label =</span> <span class="cn">TRUE</span>, </span>
<span id="cb31-102"><a href="#cb31-102" aria-hidden="true" tabindex="-1"></a>        <span class="at">split.by =</span> <span class="st">"Phase"</span>) <span class="sc">+</span> </span>
<span id="cb31-103"><a href="#cb31-103" aria-hidden="true" tabindex="-1"></a>  <span class="fu">NoLegend</span>()</span>
<span id="cb31-104"><a href="#cb31-104" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-105"><a href="#cb31-105" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-106"><a href="#cb31-106" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; We do not see much clustering by cell cycle score, so we can proceed with the QC.</span></span>
<span id="cb31-107"><a href="#cb31-107" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-108"><a href="#cb31-108" aria-hidden="true" tabindex="-1"></a><span class="fu">## 分析其他非期望变异来源是否会影响细胞分群</span></span>
<span id="cb31-109"><a href="#cb31-109" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-110"><a href="#cb31-110" aria-hidden="true" tabindex="-1"></a>Next we will explore additional metrics, such as the **number of UMIs** and **genes per cell**, **S-phase and G2M-phase markers**, and **mitochondrial gene expression** by UMAP. Looking at the individual S and G2M scores can give us additional information to checking the phase as we did previously.</span>
<span id="cb31-111"><a href="#cb31-111" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-112"><a href="#cb31-112" aria-hidden="true" tabindex="-1"></a>::: callout-tip</span>
<span id="cb31-113"><a href="#cb31-113" aria-hidden="true" tabindex="-1"></a>The <span class="in">`order`</span> argument will plot the positive cells above the negative cells, while the <span class="in">`min.cutoff`</span> argument will determine the threshold for shading. A <span class="in">`min.cutoff`</span> of <span class="in">`q10`</span> translates to the 10% of cells with the lowest expression of the gene will not exhibit any purple shading (completely gray) (<span class="in">`min.cutoff`</span>定义着色阈值。这里指定值最低的10%的细胞不着色).</span>
<span id="cb31-114"><a href="#cb31-114" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb31-115"><a href="#cb31-115" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-118"><a href="#cb31-118" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb31-119"><a href="#cb31-119" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 10</span></span>
<span id="cb31-120"><a href="#cb31-120" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 8</span></span>
<span id="cb31-121"><a href="#cb31-121" aria-hidden="true" tabindex="-1"></a><span class="co"># Determine metrics to plot present in seurat_integrated@meta.data</span></span>
<span id="cb31-122"><a href="#cb31-122" aria-hidden="true" tabindex="-1"></a>metrics <span class="ot">&lt;-</span>  <span class="fu">c</span>(<span class="st">"nUMI"</span>, <span class="st">"nGene"</span>, <span class="st">"S.Score"</span>, <span class="st">"G2M.Score"</span>, <span class="st">"mitoRatio"</span>)</span>
<span id="cb31-123"><a href="#cb31-123" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-124"><a href="#cb31-124" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>(seurat_integrated, </span>
<span id="cb31-125"><a href="#cb31-125" aria-hidden="true" tabindex="-1"></a>            <span class="at">reduction =</span> <span class="st">"umap"</span>, </span>
<span id="cb31-126"><a href="#cb31-126" aria-hidden="true" tabindex="-1"></a>            <span class="at">features =</span> metrics,</span>
<span id="cb31-127"><a href="#cb31-127" aria-hidden="true" tabindex="-1"></a>            <span class="at">pt.size =</span> <span class="fl">0.4</span>, </span>
<span id="cb31-128"><a href="#cb31-128" aria-hidden="true" tabindex="-1"></a>            <span class="at">order =</span> <span class="cn">TRUE</span>,</span>
<span id="cb31-129"><a href="#cb31-129" aria-hidden="true" tabindex="-1"></a>            <span class="at">min.cutoff =</span> <span class="st">'q10'</span>, </span>
<span id="cb31-130"><a href="#cb31-130" aria-hidden="true" tabindex="-1"></a>            <span class="at">label =</span> <span class="cn">TRUE</span>)</span>
<span id="cb31-131"><a href="#cb31-131" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-132"><a href="#cb31-132" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-133"><a href="#cb31-133" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; The metrics seem to be relatively even across the clusters, with the exception of </span><span class="in">`nGene`</span><span class="at"> exhibiting slightly higher values in clusters to the left of the plot. We can keep an eye on these clusters to see whether the cell types may explain the increase.</span></span>
<span id="cb31-134"><a href="#cb31-134" aria-hidden="true" tabindex="-1"></a><span class="at">&gt;</span></span>
<span id="cb31-135"><a href="#cb31-135" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; If we see differences corresponding to any of these metrics at this point in time, then we will often note them and then decide after identifying the cell type identities whether to take any further action.</span></span>
<span id="cb31-136"><a href="#cb31-136" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-137"><a href="#cb31-137" aria-hidden="true" tabindex="-1"></a><span class="fu">## 分析主成分（PCs）对细胞分群的影响</span></span>
<span id="cb31-138"><a href="#cb31-138" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-139"><a href="#cb31-139" aria-hidden="true" tabindex="-1"></a>We can also explore how well our clusters separate by the different PCs; **we hope that the defined PCs separate the cell types well**. To visualize this information, we need to extract the UMAP coordinate information for the cells along with their corresponding scores for each of the PCs to view by UMAP.</span>
<span id="cb31-140"><a href="#cb31-140" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-141"><a href="#cb31-141" aria-hidden="true" tabindex="-1"></a>First, we identify the information we would like to extract from the Seurat object, then, we can use the <span class="in">`FetchData()`</span> function to extract it.</span>
<span id="cb31-142"><a href="#cb31-142" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-145"><a href="#cb31-145" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb31-146"><a href="#cb31-146" aria-hidden="true" tabindex="-1"></a><span class="co"># Defining the information in the seurat object of interest</span></span>
<span id="cb31-147"><a href="#cb31-147" aria-hidden="true" tabindex="-1"></a>columns <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"ident"</span>, <span class="fu">paste0</span>(<span class="st">"PC_"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">16</span>), <span class="st">"UMAP_1"</span>, <span class="st">"UMAP_2"</span>)</span>
<span id="cb31-148"><a href="#cb31-148" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-149"><a href="#cb31-149" aria-hidden="true" tabindex="-1"></a><span class="co"># Extracting this data from the seurat object</span></span>
<span id="cb31-150"><a href="#cb31-150" aria-hidden="true" tabindex="-1"></a>pc_data <span class="ot">&lt;-</span> <span class="fu">FetchData</span>(seurat_integrated, </span>
<span id="cb31-151"><a href="#cb31-151" aria-hidden="true" tabindex="-1"></a>                     <span class="at">vars =</span> columns)</span>
<span id="cb31-152"><a href="#cb31-152" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(pc_data, <span class="dv">3</span>)</span>
<span id="cb31-153"><a href="#cb31-153" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-154"><a href="#cb31-154" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-155"><a href="#cb31-155" aria-hidden="true" tabindex="-1"></a>::: callout-tip</span>
<span id="cb31-156"><a href="#cb31-156" aria-hidden="true" tabindex="-1"></a>How did we know in the <span class="in">`FetchData()`</span> function to include <span class="in">`UMAP_1`</span> to obtain the UMAP coordinates? @sec-seurat_command_list describes the function as being able to pull any data from the expression matrices, cell embeddings, or metadata.</span>
<span id="cb31-157"><a href="#cb31-157" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-158"><a href="#cb31-158" aria-hidden="true" tabindex="-1"></a>For instance, if you explore the <span class="in">`seurat_integrated@reductions`</span> list object, the first component is for PCA, and includes a slot for <span class="in">`cell.embeddings`</span> (坐标数据). We can use the column names (<span class="in">`PC_1`</span>, <span class="in">`PC_2`</span>, <span class="in">`PC_3`</span>, etc.) to pull out the coordinates or PC scores corresponding to each cell for each of the PCs (提取每个细胞在指定主成分上的坐标).</span>
<span id="cb31-159"><a href="#cb31-159" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-160"><a href="#cb31-160" aria-hidden="true" tabindex="-1"></a>We could do the same thing for UMAP:</span>
<span id="cb31-161"><a href="#cb31-161" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-164"><a href="#cb31-164" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb31-165"><a href="#cb31-165" aria-hidden="true" tabindex="-1"></a><span class="co"># 提取前5个细胞在前两个UMAP主成分上的坐标</span></span>
<span id="cb31-166"><a href="#cb31-166" aria-hidden="true" tabindex="-1"></a>seurat_integrated<span class="sc">@</span>reductions[[<span class="st">"umap"</span>]]<span class="sc">@</span>cell.embeddings[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]</span>
<span id="cb31-167"><a href="#cb31-167" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-168"><a href="#cb31-168" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-169"><a href="#cb31-169" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-170"><a href="#cb31-170" aria-hidden="true" tabindex="-1"></a>等价于：</span>
<span id="cb31-171"><a href="#cb31-171" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-174"><a href="#cb31-174" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb31-175"><a href="#cb31-175" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb31-176"><a href="#cb31-176" aria-hidden="true" tabindex="-1"></a><span class="fu">FetchData</span>(seurat_integrated, <span class="at">vars =</span> <span class="fu">c</span>(<span class="st">"UMAP_1"</span>, <span class="st">"UMAP_2"</span>)) <span class="sc">|&gt;</span> <span class="fu">head</span>(<span class="dv">5</span>)</span>
<span id="cb31-177"><a href="#cb31-177" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-178"><a href="#cb31-178" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-179"><a href="#cb31-179" aria-hidden="true" tabindex="-1"></a>The <span class="in">`FetchData()`</span> function just allows us to extract the data more easily.</span>
<span id="cb31-180"><a href="#cb31-180" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb31-181"><a href="#cb31-181" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-182"><a href="#cb31-182" aria-hidden="true" tabindex="-1"></a>::: callout-caution</span>
<span id="cb31-183"><a href="#cb31-183" aria-hidden="true" tabindex="-1"></a>The pre-existing <span class="in">`seurat_integrated`</span> loaded in previously was created using an older version of Seurat. As such the columns we <span class="in">`Fetch()`</span> are in upper case (i.e <span class="in">`UMAP_1`</span>). **If you are using your own seurat object using a newer version of Seurat you will need to change the column names as shown below.** Alternatively, explore your Seurat object to see how they have been stored.</span>
<span id="cb31-184"><a href="#cb31-184" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-187"><a href="#cb31-187" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb31-188"><a href="#cb31-188" aria-hidden="true" tabindex="-1"></a> <span class="co"># Defining the information in the seurat object of interest</span></span>
<span id="cb31-189"><a href="#cb31-189" aria-hidden="true" tabindex="-1"></a> columns <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"ident"</span>, <span class="fu">paste0</span>(<span class="st">"PC_"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">16</span>), <span class="st">"umap_1"</span>, <span class="st">"umap_2"</span>)</span>
<span id="cb31-190"><a href="#cb31-190" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-191"><a href="#cb31-191" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb31-192"><a href="#cb31-192" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-193"><a href="#cb31-193" aria-hidden="true" tabindex="-1"></a>In the UMAP plots below, the cells are colored by their PC score for each respective principal component.</span>
<span id="cb31-194"><a href="#cb31-194" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-195"><a href="#cb31-195" aria-hidden="true" tabindex="-1"></a>Let's take a quick look at the top 16 PCs:</span>
<span id="cb31-196"><a href="#cb31-196" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-199"><a href="#cb31-199" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb31-200"><a href="#cb31-200" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 20</span></span>
<span id="cb31-201"><a href="#cb31-201" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 15</span></span>
<span id="cb31-202"><a href="#cb31-202" aria-hidden="true" tabindex="-1"></a><span class="co"># Adding cluster label to center of cluster on UMAP</span></span>
<span id="cb31-203"><a href="#cb31-203" aria-hidden="true" tabindex="-1"></a>umap_label <span class="ot">&lt;-</span> <span class="fu">FetchData</span>(seurat_integrated, </span>
<span id="cb31-204"><a href="#cb31-204" aria-hidden="true" tabindex="-1"></a>                        <span class="at">vars =</span> <span class="fu">c</span>(<span class="st">"ident"</span>, <span class="st">"UMAP_1"</span>, <span class="st">"UMAP_2"</span>)) <span class="sc">|&gt;</span></span>
<span id="cb31-205"><a href="#cb31-205" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(ident) <span class="sc">|&gt;</span> <span class="co"># 指定分组计算依据</span></span>
<span id="cb31-206"><a href="#cb31-206" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">UMAP1_mean =</span> <span class="fu">mean</span>(UMAP_1), <span class="at">UMAP2_mean =</span> <span class="fu">mean</span>(UMAP_2)) <span class="co"># 根据指定的分组依据进行分组计算</span></span>
<span id="cb31-207"><a href="#cb31-207" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(umap_label)</span>
<span id="cb31-208"><a href="#cb31-208" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb31-209"><a href="#cb31-209" aria-hidden="true" tabindex="-1"></a><span class="co"># Plotting a UMAP plot for each of the PCs</span></span>
<span id="cb31-210"><a href="#cb31-210" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(purrr)</span>
<span id="cb31-211"><a href="#cb31-211" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb31-212"><a href="#cb31-212" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(cowplot)</span>
<span id="cb31-213"><a href="#cb31-213" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb31-214"><a href="#cb31-214" aria-hidden="true" tabindex="-1"></a><span class="fu">map</span>(<span class="fu">paste0</span>(<span class="st">"PC_"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">16</span>), </span>
<span id="cb31-215"><a href="#cb31-215" aria-hidden="true" tabindex="-1"></a>    <span class="cf">function</span>(pc) {</span>
<span id="cb31-216"><a href="#cb31-216" aria-hidden="true" tabindex="-1"></a>      <span class="fu">ggplot</span>(pc_data, </span>
<span id="cb31-217"><a href="#cb31-217" aria-hidden="true" tabindex="-1"></a>             <span class="fu">aes</span>(UMAP_1, UMAP_2)) <span class="sc">+</span></span>
<span id="cb31-218"><a href="#cb31-218" aria-hidden="true" tabindex="-1"></a>        <span class="fu">geom_point</span>(<span class="fu">aes_string</span>(<span class="at">color =</span> pc), </span>
<span id="cb31-219"><a href="#cb31-219" aria-hidden="true" tabindex="-1"></a>                   <span class="at">alpha =</span> <span class="fl">0.7</span>) <span class="sc">+</span></span>
<span id="cb31-220"><a href="#cb31-220" aria-hidden="true" tabindex="-1"></a>        <span class="fu">scale_color_gradient</span>(<span class="at">low =</span> <span class="st">"grey90"</span>, </span>
<span id="cb31-221"><a href="#cb31-221" aria-hidden="true" tabindex="-1"></a>                             <span class="at">high =</span> <span class="st">"blue"</span>) <span class="sc">+</span></span>
<span id="cb31-222"><a href="#cb31-222" aria-hidden="true" tabindex="-1"></a>        <span class="fu">geom_text</span>(<span class="at">data =</span> umap_label, </span>
<span id="cb31-223"><a href="#cb31-223" aria-hidden="true" tabindex="-1"></a>                  <span class="fu">aes</span>(<span class="at">label =</span> ident, UMAP1_mean, UMAP2_mean)) <span class="sc">+</span></span>
<span id="cb31-224"><a href="#cb31-224" aria-hidden="true" tabindex="-1"></a>        <span class="fu">ggtitle</span>(pc)</span>
<span id="cb31-225"><a href="#cb31-225" aria-hidden="true" tabindex="-1"></a>      }) <span class="sc">%&gt;%</span> </span>
<span id="cb31-226"><a href="#cb31-226" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot_grid</span>(<span class="at">plotlist =</span> .)</span>
<span id="cb31-227"><a href="#cb31-227" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-228"><a href="#cb31-228" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-229"><a href="#cb31-229" aria-hidden="true" tabindex="-1"></a>We can see how the clusters are represented by the different PCs. For instance, the genes driving <span class="in">`PC_2`</span> exhibit higher expression in clusters 8 and 12. We could look back at our genes driving this PC to get an idea of what the cell types might be:</span>
<span id="cb31-230"><a href="#cb31-230" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-233"><a href="#cb31-233" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb31-234"><a href="#cb31-234" aria-hidden="true" tabindex="-1"></a><span class="co"># 提取PCA信息中的第二主成分，并展示对该主成分影响最大的前5个基因名</span></span>
<span id="cb31-235"><a href="#cb31-235" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(seurat_integrated[[<span class="st">"pca"</span>]], <span class="at">dims =</span> <span class="dv">2</span>, <span class="at">nfeatures =</span> <span class="dv">5</span>)</span>
<span id="cb31-236"><a href="#cb31-236" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-237"><a href="#cb31-237" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-238"><a href="#cb31-238" aria-hidden="true" tabindex="-1"></a>With the GNLY and NKG7 genes as positive markers of <span class="in">`PC_2`</span>, **we can hypothesize that clusters 8 and 12 correspond to NK cells**. This just hints at what the clusters identity could be, with the identities of the clusters being determined through a combination of the PCs.</span>
<span id="cb31-239"><a href="#cb31-239" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-240"><a href="#cb31-240" aria-hidden="true" tabindex="-1"></a>To truly determine the identity of the clusters and whether the <span class="in">`resolution`</span> is appropriate, it is helpful to explore a handful of known gene markers for the cell types expected.</span>
<span id="cb31-241"><a href="#cb31-241" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-242"><a href="#cb31-242" aria-hidden="true" tabindex="-1"></a><span class="fu">## 探索已知的cell type markers的表达</span></span>
<span id="cb31-243"><a href="#cb31-243" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-244"><a href="#cb31-244" aria-hidden="true" tabindex="-1"></a>With the cells clustered, we can explore the cell type identities by looking for known markers. 下面给出了本案例的cell type markers及其对应的细胞类型：</span>
<span id="cb31-245"><a href="#cb31-245" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-246"><a href="#cb31-246" aria-hidden="true" tabindex="-1"></a>|          Cell Type           |            Marker            |</span>
<span id="cb31-247"><a href="#cb31-247" aria-hidden="true" tabindex="-1"></a>|:----------------------------:|:----------------------------:|</span>
<span id="cb31-248"><a href="#cb31-248" aria-hidden="true" tabindex="-1"></a>|       CD14+ monocytes        |          CD14, LYZ           |</span>
<span id="cb31-249"><a href="#cb31-249" aria-hidden="true" tabindex="-1"></a>|      FCGR3A+ monocytes       |        FCGR3A, MS4A7         |</span>
<span id="cb31-250"><a href="#cb31-250" aria-hidden="true" tabindex="-1"></a>| Conventional dendritic cells |         FCER1A, CST3         |</span>
<span id="cb31-251"><a href="#cb31-251" aria-hidden="true" tabindex="-1"></a>| Plasmacytoid dendritic cells | IL3RA, GZMB, SERPINF1, ITM2C |</span>
<span id="cb31-252"><a href="#cb31-252" aria-hidden="true" tabindex="-1"></a>|           B cells            |         CD79A, MS4A1         |</span>
<span id="cb31-253"><a href="#cb31-253" aria-hidden="true" tabindex="-1"></a>|           T cells            |             CD3D             |</span>
<span id="cb31-254"><a href="#cb31-254" aria-hidden="true" tabindex="-1"></a>|         CD4+ T cells         |       CD3D, IL7R, CCR7       |</span>
<span id="cb31-255"><a href="#cb31-255" aria-hidden="true" tabindex="-1"></a>|         CD8+ T cells         |          CD3D, CD8A          |</span>
<span id="cb31-256"><a href="#cb31-256" aria-hidden="true" tabindex="-1"></a>|           NK cells           |          GNLY, NKG7          |</span>
<span id="cb31-257"><a href="#cb31-257" aria-hidden="true" tabindex="-1"></a>|        Megakaryocytes        |             PPBP             |</span>
<span id="cb31-258"><a href="#cb31-258" aria-hidden="true" tabindex="-1"></a>|         Erythrocytes         |          HBB, HBA2           |</span>
<span id="cb31-259"><a href="#cb31-259" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-260"><a href="#cb31-260" aria-hidden="true" tabindex="-1"></a>The <span class="in">`FeaturePlot()`</span> function from Seurat makes it easy to visualize a handful of genes using the gene IDs stored in the Seurat object. We can easily explore the expression of known gene markers on top of our UMAP visualizations. Let's go through and determine the identities of the clusters. To access the normalized expression levels of all genes, we can use the normalized count data stored in the <span class="in">`RNA`</span> assay slot.</span>
<span id="cb31-261"><a href="#cb31-261" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-262"><a href="#cb31-262" aria-hidden="true" tabindex="-1"></a>::: callout-caution</span>
<span id="cb31-263"><a href="#cb31-263" aria-hidden="true" tabindex="-1"></a>The <span class="in">`SCTransform`</span> normalization was performed only on the **3000** most variable genes, so many of our genes of interest may not be present in this data ( @sec-perform_sctranform ).</span>
<span id="cb31-264"><a href="#cb31-264" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb31-265"><a href="#cb31-265" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-266"><a href="#cb31-266" aria-hidden="true" tabindex="-1"></a>这里为了准确可视化marker基因的表达情况，我们将默认的assay改回“RNA”，然后运行<span class="in">`NormalizeData`</span>：</span>
<span id="cb31-267"><a href="#cb31-267" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-270"><a href="#cb31-270" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb31-271"><a href="#cb31-271" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb31-272"><a href="#cb31-272" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(seurat_integrated) </span>
<span id="cb31-273"><a href="#cb31-273" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-274"><a href="#cb31-274" aria-hidden="true" tabindex="-1"></a><span class="co"># Select the RNA counts slot to be the default assay</span></span>
<span id="cb31-275"><a href="#cb31-275" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(seurat_integrated) <span class="ot">&lt;-</span> <span class="st">"RNA"</span></span>
<span id="cb31-276"><a href="#cb31-276" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-277"><a href="#cb31-277" aria-hidden="true" tabindex="-1"></a><span class="co"># Normalize RNA data for visualization purposes</span></span>
<span id="cb31-278"><a href="#cb31-278" aria-hidden="true" tabindex="-1"></a>seurat_integrated <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(seurat_integrated, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb31-279"><a href="#cb31-279" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-280"><a href="#cb31-280" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-281"><a href="#cb31-281" aria-hidden="true" tabindex="-1"></a>::: callout-tip</span>
<span id="cb31-282"><a href="#cb31-282" aria-hidden="true" tabindex="-1"></a>Assay is a slot defined in the Seurat object, it has multiple slots within it. In a given assay, the <span class="in">`counts`</span> layer stores non-normalized raw counts, and the <span class="in">`data`</span> layer stores normalized expression data. Therefore, when we run the <span class="in">`NormalizeData()`</span> function in the above code, the normalized data will be stored in the <span class="in">`data`</span> layer of the RNA assay while the <span class="in">`counts`</span> layer will remain unaltered.</span>
<span id="cb31-283"><a href="#cb31-283" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb31-284"><a href="#cb31-284" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-285"><a href="#cb31-285" aria-hidden="true" tabindex="-1"></a>Depending on our markers of interest, they could be positive or negative markers for a particular cell type. The combined expression of our chosen handful of markers should give us an idea on whether a cluster corresponds to that particular cell type.</span>
<span id="cb31-286"><a href="#cb31-286" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-287"><a href="#cb31-287" aria-hidden="true" tabindex="-1"></a>For the markers used here, we are looking for positive markers and consistency of expression of the markers across the clusters. For example, if there are two markers for a cell type and only one of them is expressed in a cluster - then we cannot reliably assign that cluster to the cell type.</span>
<span id="cb31-288"><a href="#cb31-288" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-289"><a href="#cb31-289" aria-hidden="true" tabindex="-1"></a><span class="fu">### CD14+ monocyte markers</span></span>
<span id="cb31-290"><a href="#cb31-290" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-293"><a href="#cb31-293" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb31-294"><a href="#cb31-294" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 10</span></span>
<span id="cb31-295"><a href="#cb31-295" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>(seurat_integrated, </span>
<span id="cb31-296"><a href="#cb31-296" aria-hidden="true" tabindex="-1"></a>            <span class="at">reduction =</span> <span class="st">"umap"</span>, </span>
<span id="cb31-297"><a href="#cb31-297" aria-hidden="true" tabindex="-1"></a>            <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"CD14"</span>, <span class="st">"LYZ"</span>), </span>
<span id="cb31-298"><a href="#cb31-298" aria-hidden="true" tabindex="-1"></a>            <span class="at">order =</span> <span class="cn">TRUE</span>,</span>
<span id="cb31-299"><a href="#cb31-299" aria-hidden="true" tabindex="-1"></a>            <span class="at">min.cutoff =</span> <span class="st">'q10'</span>, </span>
<span id="cb31-300"><a href="#cb31-300" aria-hidden="true" tabindex="-1"></a>            <span class="at">label =</span> <span class="cn">TRUE</span>)</span>
<span id="cb31-301"><a href="#cb31-301" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-302"><a href="#cb31-302" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-303"><a href="#cb31-303" aria-hidden="true" tabindex="-1"></a>CD14+ monocytes appear to correspond to clusters 1, and 3. We wouldn't include clusters 14 and 10 because they do not highly express both of these markers.</span>
<span id="cb31-304"><a href="#cb31-304" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-305"><a href="#cb31-305" aria-hidden="true" tabindex="-1"></a><span class="fu">### FCGR3A+ monocyte markers</span></span>
<span id="cb31-306"><a href="#cb31-306" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-309"><a href="#cb31-309" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb31-310"><a href="#cb31-310" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 10</span></span>
<span id="cb31-311"><a href="#cb31-311" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>(seurat_integrated, </span>
<span id="cb31-312"><a href="#cb31-312" aria-hidden="true" tabindex="-1"></a>            <span class="at">reduction =</span> <span class="st">"umap"</span>, </span>
<span id="cb31-313"><a href="#cb31-313" aria-hidden="true" tabindex="-1"></a>            <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"FCGR3A"</span>, <span class="st">"MS4A7"</span>), </span>
<span id="cb31-314"><a href="#cb31-314" aria-hidden="true" tabindex="-1"></a>            <span class="at">order =</span> <span class="cn">TRUE</span>,</span>
<span id="cb31-315"><a href="#cb31-315" aria-hidden="true" tabindex="-1"></a>            <span class="at">min.cutoff =</span> <span class="st">'q10'</span>, </span>
<span id="cb31-316"><a href="#cb31-316" aria-hidden="true" tabindex="-1"></a>            <span class="at">label =</span> <span class="cn">TRUE</span>)</span>
<span id="cb31-317"><a href="#cb31-317" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-318"><a href="#cb31-318" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-319"><a href="#cb31-319" aria-hidden="true" tabindex="-1"></a>FCGR3A+ monocytes markers distinctly highlight cluster 10, although we do see some decent expression in clusters 1 and 3. We would like to see additional markers for FCGR3A+ cells show up when we perform the marker identification.</span>
<span id="cb31-320"><a href="#cb31-320" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-321"><a href="#cb31-321" aria-hidden="true" tabindex="-1"></a><span class="fu">### Macrophages</span></span>
<span id="cb31-322"><a href="#cb31-322" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-325"><a href="#cb31-325" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb31-326"><a href="#cb31-326" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 10</span></span>
<span id="cb31-327"><a href="#cb31-327" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>(seurat_integrated, </span>
<span id="cb31-328"><a href="#cb31-328" aria-hidden="true" tabindex="-1"></a>            <span class="at">reduction =</span> <span class="st">"umap"</span>, </span>
<span id="cb31-329"><a href="#cb31-329" aria-hidden="true" tabindex="-1"></a>            <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"MARCO"</span>, <span class="st">"ITGAM"</span>, <span class="st">"ADGRE1"</span>), </span>
<span id="cb31-330"><a href="#cb31-330" aria-hidden="true" tabindex="-1"></a>            <span class="at">order =</span> <span class="cn">TRUE</span>,</span>
<span id="cb31-331"><a href="#cb31-331" aria-hidden="true" tabindex="-1"></a>            <span class="at">min.cutoff =</span> <span class="st">'q10'</span>, </span>
<span id="cb31-332"><a href="#cb31-332" aria-hidden="true" tabindex="-1"></a>            <span class="at">label =</span> <span class="cn">TRUE</span>)</span>
<span id="cb31-333"><a href="#cb31-333" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-334"><a href="#cb31-334" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-335"><a href="#cb31-335" aria-hidden="true" tabindex="-1"></a>We don't see much overlap of our markers, so no clusters appear to correspond to macrophages; perhaps cell culture conditions negatively selected for macrophages (more highly adherent).</span>
<span id="cb31-336"><a href="#cb31-336" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-337"><a href="#cb31-337" aria-hidden="true" tabindex="-1"></a><span class="fu">### Conventional dendritic cell markers</span></span>
<span id="cb31-338"><a href="#cb31-338" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-341"><a href="#cb31-341" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb31-342"><a href="#cb31-342" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 10</span></span>
<span id="cb31-343"><a href="#cb31-343" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>(seurat_integrated, </span>
<span id="cb31-344"><a href="#cb31-344" aria-hidden="true" tabindex="-1"></a>            <span class="at">reduction =</span> <span class="st">"umap"</span>, </span>
<span id="cb31-345"><a href="#cb31-345" aria-hidden="true" tabindex="-1"></a>            <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"FCER1A"</span>, <span class="st">"CST3"</span>), </span>
<span id="cb31-346"><a href="#cb31-346" aria-hidden="true" tabindex="-1"></a>            <span class="at">order =</span> <span class="cn">TRUE</span>,</span>
<span id="cb31-347"><a href="#cb31-347" aria-hidden="true" tabindex="-1"></a>            <span class="at">min.cutoff =</span> <span class="st">'q10'</span>, </span>
<span id="cb31-348"><a href="#cb31-348" aria-hidden="true" tabindex="-1"></a>            <span class="at">label =</span> <span class="cn">TRUE</span>)</span>
<span id="cb31-349"><a href="#cb31-349" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-350"><a href="#cb31-350" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-351"><a href="#cb31-351" aria-hidden="true" tabindex="-1"></a>The markers corresponding to conventional dendritic cells identify cluster 14 (both markers consistently show expression).</span>
<span id="cb31-352"><a href="#cb31-352" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-353"><a href="#cb31-353" aria-hidden="true" tabindex="-1"></a><span class="fu">### Plasmacytoid dendritic cell markers</span></span>
<span id="cb31-354"><a href="#cb31-354" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-357"><a href="#cb31-357" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb31-358"><a href="#cb31-358" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 10</span></span>
<span id="cb31-359"><a href="#cb31-359" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>(seurat_integrated, </span>
<span id="cb31-360"><a href="#cb31-360" aria-hidden="true" tabindex="-1"></a>            <span class="at">reduction =</span> <span class="st">"umap"</span>, </span>
<span id="cb31-361"><a href="#cb31-361" aria-hidden="true" tabindex="-1"></a>            <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"IL3RA"</span>, <span class="st">"GZMB"</span>, <span class="st">"SERPINF1"</span>, <span class="st">"ITM2C"</span>), </span>
<span id="cb31-362"><a href="#cb31-362" aria-hidden="true" tabindex="-1"></a>            <span class="at">order =</span> <span class="cn">TRUE</span>,</span>
<span id="cb31-363"><a href="#cb31-363" aria-hidden="true" tabindex="-1"></a>            <span class="at">min.cutoff =</span> <span class="st">'q10'</span>, </span>
<span id="cb31-364"><a href="#cb31-364" aria-hidden="true" tabindex="-1"></a>            <span class="at">label =</span> <span class="cn">TRUE</span>)</span>
<span id="cb31-365"><a href="#cb31-365" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb31-366"><a href="#cb31-366" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-367"><a href="#cb31-367" aria-hidden="true" tabindex="-1"></a>Plasmacytoid dendritic cells represent cluster 16. While there are a lot of differences in the expression of these markers, we see cluster 16 (though small) is consistently strongly expressed.</span>
<span id="cb31-368"><a href="#cb31-368" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-369"><a href="#cb31-369" aria-hidden="true" tabindex="-1"></a>------------------------------------------------------------------------</span>
<span id="cb31-370"><a href="#cb31-370" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-371"><a href="#cb31-371" aria-hidden="true" tabindex="-1"></a>**Exercise**</span>
<span id="cb31-372"><a href="#cb31-372" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-373"><a href="#cb31-373" aria-hidden="true" tabindex="-1"></a>Hypothesize the clusters corresponding to each of the different clusters in the table:</span>
<span id="cb31-374"><a href="#cb31-374" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-375"><a href="#cb31-375" aria-hidden="true" tabindex="-1"></a>|          Cell Type           | Clusters |</span>
<span id="cb31-376"><a href="#cb31-376" aria-hidden="true" tabindex="-1"></a>|:----------------------------:|:--------:|</span>
<span id="cb31-377"><a href="#cb31-377" aria-hidden="true" tabindex="-1"></a>|       CD14+ monocytes        |   1, 3   |</span>
<span id="cb31-378"><a href="#cb31-378" aria-hidden="true" tabindex="-1"></a>|      FCGR3A+ monocytes       |    10    |</span>
<span id="cb31-379"><a href="#cb31-379" aria-hidden="true" tabindex="-1"></a>| Conventional dendritic cells |    14    |</span>
<span id="cb31-380"><a href="#cb31-380" aria-hidden="true" tabindex="-1"></a>| Plasmacytoid dendritic cells |    16    |</span>
<span id="cb31-381"><a href="#cb31-381" aria-hidden="true" tabindex="-1"></a>|         Marcrophages         |    <span class="sc">\-</span>    |</span>
<span id="cb31-382"><a href="#cb31-382" aria-hidden="true" tabindex="-1"></a>|           B cells            |    ?     |</span>
<span id="cb31-383"><a href="#cb31-383" aria-hidden="true" tabindex="-1"></a>|           T cells            |    ?     |</span>
<span id="cb31-384"><a href="#cb31-384" aria-hidden="true" tabindex="-1"></a>|         CD4+ T cells         |    ?     |</span>
<span id="cb31-385"><a href="#cb31-385" aria-hidden="true" tabindex="-1"></a>|         CD8+ T cells         |    ?     |</span>
<span id="cb31-386"><a href="#cb31-386" aria-hidden="true" tabindex="-1"></a>|           NK cells           |    ?     |</span>
<span id="cb31-387"><a href="#cb31-387" aria-hidden="true" tabindex="-1"></a>|        Megakaryocytes        |    ?     |</span>
<span id="cb31-388"><a href="#cb31-388" aria-hidden="true" tabindex="-1"></a>|         Erythrocytes         |    ?     |</span>
<span id="cb31-389"><a href="#cb31-389" aria-hidden="true" tabindex="-1"></a>|           Unknown            |    ?     |</span>
<span id="cb31-390"><a href="#cb31-390" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-391"><a href="#cb31-391" aria-hidden="true" tabindex="-1"></a>------------------------------------------------------------------------</span>
<span id="cb31-392"><a href="#cb31-392" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-393"><a href="#cb31-393" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; **NOTE:** If any cluster appears to contain two separate cell types, it's helpful to increase our clustering resolution to properly subset the clusters. Alternatively, if we still can't separate out the clusters using increased resolution, then it's possible that we had used too few principal components such that we are just not separating out these cell types of interest. To inform our choice of PCs, we could look at our PC gene expression overlapping the UMAP plots and determine whether our cell populations are separating by the PCs included.</span></span>
<span id="cb31-394"><a href="#cb31-394" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-395"><a href="#cb31-395" aria-hidden="true" tabindex="-1"></a>Now we have a decent idea as to the cell types corresponding to the majority of the clusters, but some questions remain:</span>
<span id="cb31-396"><a href="#cb31-396" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-397"><a href="#cb31-397" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>*T cell markers appear to be highly expressed in may clusters. How can we differentiate and subset the larger group into smaller subset of cells?*</span>
<span id="cb31-398"><a href="#cb31-398" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>*Do the clusters corresponding to the same cell types have biologically meaningful differences? Are there subpopulations of these cell types?*</span>
<span id="cb31-399"><a href="#cb31-399" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>*Can we acquire higher confidence in these cell type identities by identifying other marker genes for these clusters?*</span>
<span id="cb31-400"><a href="#cb31-400" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-401"><a href="#cb31-401" aria-hidden="true" tabindex="-1"></a>Marker identification analysis can help us address all of these questions!!</span>
<span id="cb31-402"><a href="#cb31-402" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-403"><a href="#cb31-403" aria-hidden="true" tabindex="-1"></a>The next step will be to perform marker identification analysis, which will output the genes that significantly differ in expression between clusters. Using these genes we can determine or improve confidence in the identities of the clusters/subclusters.</span>
<span id="cb31-404"><a href="#cb31-404" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-405"><a href="#cb31-405" aria-hidden="true" tabindex="-1"></a>------------------------------------------------------------------------</span>
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">
      <ul class="footer-items list-unstyled">
<li class="nav-item">
 This book was built with Quarto
  </li>  
</ul>
</div>   
    <div class="nav-footer-center">Copyright 2023, Du Junhong</div>
    <div class="nav-footer-right">
      <ul class="footer-items list-unstyled">
<li class="nav-item compact">
    <a class="nav-link" href="https://github.com/djhcod">
      <i class="bi bi-github" role="img">
</i> 
    </a>
  </li>  
</ul>
</div>
  </div>
</footer>


</body></html>