<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>R语言学习笔记：从数据清洗到高级统计学和生物信息学 - 寻找marker基因+细胞注释</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../single_cell/sc_supplementary/sc_supplementary_intro.html" rel="next">
<link href="../../single_cell/scRNA-seq_online/08_SC_clustering_quality_control.html" rel="prev">
<link href="../../images/logo.jpg" rel="icon" type="image/jpeg">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark"><script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><script async="" src="https://www.googletagmanager.com/gtag/js?id=G-G2HZEXHNCE"></script><script type="text/javascript">

window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', 'G-G2HZEXHNCE', { 'anonymize_ip': true});
</script><script type="application/json" class="js-hypothesis-config">
{
  "theme": "classic",
  "openSidebar": false,
  "showHighlights": "always",
  "enableExperimentalNewNoteButton": true
}
</script><script async="" src="https://hypothes.is/embed.js"></script>
<meta property="og:title" content="R语言学习笔记：从数据清洗到高级统计学和生物信息学 - 寻找marker基因+细胞注释">
<meta property="og:description" content="">
<meta property="og:image" content="https://djhcod.github.io/r-notes/single_cell/scRNA-seq_online/images/sc_workflow_2022-01.jpg">
<meta property="og:site-name" content="R语言学习笔记：从数据清洗到高级统计学和生物信息学">
<meta name="twitter:title" content="R语言学习笔记：从数据清洗到高级统计学和生物信息学 - 寻找marker基因+细胞注释">
<meta name="twitter:description" content="">
<meta name="twitter:image" content="https://djhcod.github.io/r-notes/single_cell/scRNA-seq_online/images/sc_workflow_2022-01.jpg">
<meta name="twitter:card" content="summary_large_image">
</head>
<body class="nav-sidebar docked nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="navbar navbar-expand-lg navbar-dark "><div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a href="../../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../../images/logo.jpg" alt="" class="navbar-logo"></a>
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">R语言学习笔记</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
<li class="nav-item">
    <a class="nav-link active" href="../../index.html" rel="" target="" aria-current="page">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../r_basic/r_basics.html" rel="" target="">
 <span class="menu-text">R语言基础</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">单细胞数据分析</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-">
<li>
    <a class="dropdown-item" href="../../single_cell/single_cell_intro.html" rel="" target="">
 <span class="dropdown-text">单细胞数据分析介绍</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../single_cell/seurat/seurat_intro.html" rel="" target="">
 <span class="dropdown-text">Seurat v5官方文档</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../single_cell/scRNA-seq_online/00_intro.html" rel="" target="">
 <span class="dropdown-text">scRNA-seq_online学习材料</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../single_cell/sc_supplementary/sc_supplementary_intro.html" rel="" target="">
 <span class="dropdown-text">单细胞补充内容</span></a>
  </li>  
    </ul>
</li>
  <li class="nav-item">
    <a class="nav-link" href="../../quarto_foundation/quarto_foundation.html" rel="" target="">
 <span class="menu-text">Quarto基础</span></a>
  </li>  
</ul>
<ul class="navbar-nav navbar-nav-scroll ms-auto">
<li class="nav-item">
    <a class="nav-link" href="https://bookdown.org" rel="" target=""><i class="bi bi-book-fill" role="img">
</i> 
 <span class="menu-text">Bookdown</span></a>
  </li>  
</ul>
<div class="quarto-navbar-tools tools-wide">
    <a href="https://github.com/djhcod/r-notes" rel="" title="" class="quarto-navigation-tool px-1" aria-label=""><i class="bi bi-github"></i></a>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
  <a href="" class="quarto-reader-toggle quarto-navigation-tool px-1" onclick="window.quartoToggleReader(); return false;" title="Toggle reader mode">
  <div class="quarto-reader-toggle-btn">
  <i class="bi"></i>
  </div>
</a>
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../single_cell/single_cell_intro.html">单细胞数据分析</a></li><li class="breadcrumb-item"><a href="../../single_cell/scRNA-seq_online/00_intro.html">scRNA-seq_online学习材料</a></li><li class="breadcrumb-item"><a href="../../single_cell/scRNA-seq_online/09_merged_SC_marker_identification.html">寻找marker基因+细胞注释</a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto"><div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">主页</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../r_basic/r_basics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">R语言基础</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/environment_configuration.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Rstudio环境配置</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/data_input_output.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">数据的读取与输出</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/basic_data_function.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">数据处理基本函数</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/dplyr.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">基于dplyr包的数据处理</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/character.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">字符的处理</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/loop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">循环</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../single_cell/single_cell_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">单细胞数据分析</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">单细胞测序技术介绍</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/generation_of_count_matrix.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">单细胞RNA测序—从raw data到count matrix</span></a>
  </div>
</li>
          <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../single_cell/seurat/seurat_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Seurat v5官方文档</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth2 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/seurat_command_list.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Seurat常用函数清单</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/seurat_tutorial.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Seurat细胞分群官方教程</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/data_visualization_methods_in_seurat.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Seurat中的数据可视化方法</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/sctransform.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">基于SCTransform的单细胞数据标准化</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/cell_cycle_regression.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">消除细胞周期的影响</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/integration.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">整合（integration）</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/integrative_analysis_in_seurat_v5.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Seurat v5单细胞数据整合分析</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/marker_gene_identification.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">寻找marker gene</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/de_vignette.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">寻找不同样本类型间同一细胞类型内的差异基因</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/mapping_and_annotating_query_datasets.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">基于参考集的细胞注释</span></a>
  </div>
</li>
      </ul>
</li>
          <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../single_cell/scRNA-seq_online/00_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">scRNA-seq_online学习材料</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth2 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/03_SC_quality_control-setup.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">数据导入与Seurat对象构建</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/04_SC_quality_control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">质控</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/postQC_workflow.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Single-cell RNA-seq Clustering Workflow</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/05_theory_of_PCA.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">PCA原理</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/06_SC_SCT_normalization.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Normalization and regressing out unwanted variation</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/06_integration.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">单细胞数据整合（Integration）</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/07_SC_clustering_cells_SCT.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">细胞聚类（clustering analysis）</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/08_SC_clustering_quality_control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">细胞分群质量评估</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/09_merged_SC_marker_identification.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">寻找marker基因+细胞注释</span></a>
  </div>
</li>
      </ul>
</li>
          <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../single_cell/sc_supplementary/sc_supplementary_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">单细胞补充内容</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth2 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/sc_supplementary/read_sc_data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">读取非标准格式的单细胞数据</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/sc_supplementary/universal_marker.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">细胞分群通用marker</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/sc_supplementary/integrated_analysis_multiple_single_cell_datasets_1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">多个单细胞数据集整合分析（上）</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/sc_supplementary/integrated_analysis_multiple_single_cell_datasets_2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">多个单细胞数据集整合分析（下）</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/sc_supplementary/DecontX.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">预测和去除单细胞转录组的环境游离RNA污染</span></a>
  </div>
</li>
      </ul>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../quarto_foundation/quarto_foundation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Quarto基础</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/yaml_settings.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">YAML设置</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/images_settings.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">图片设置</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/cross_references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">交叉引用</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/insert_other_content.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">插入其他内容</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/quarto_shiny.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">在Quarto中嵌入Shiny应用程序</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/quarto_books.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Quarto Books</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/github_pages.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">发布到GitHub Pages</span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active" data-toc-expanded="1"><h2 id="toc-title">On this page</h2>
   
  <ul>
<li><a href="#%E6%95%B0%E6%8D%AE%E5%AF%BC%E5%85%A5" id="toc-数据导入" class="nav-link active" data-scroll-target="#%E6%95%B0%E6%8D%AE%E5%AF%BC%E5%85%A5"><span class="header-section-number">1</span> 数据导入</a></li>
  <li><a href="#findallmarkers-identification-of-all-markers-for-each-cluster" id="toc-findallmarkers-identification-of-all-markers-for-each-cluster" class="nav-link" data-scroll-target="#findallmarkers-identification-of-all-markers-for-each-cluster"><span class="header-section-number">2</span> <code>FindAllMarkers</code>-Identification of all markers for each cluster</a></li>
  <li>
<a href="#findconservedmarkers-identification-of-conserved-markers-in-all-conditions" id="toc-findconservedmarkers-identification-of-conserved-markers-in-all-conditions" class="nav-link" data-scroll-target="#findconservedmarkers-identification-of-conserved-markers-in-all-conditions"><span class="header-section-number">3</span> <code>FindConservedMarkers</code>-Identification of conserved markers in all conditions</a>
  <ul class="collapse">
<li>
<a href="#%E5%AF%BB%E6%89%BEcluster-0%E5%92%8Ccluster-10%E7%9A%84conserved-markers" id="toc-寻找cluster-0和cluster-10的conserved-markers" class="nav-link" data-scroll-target="#%E5%AF%BB%E6%89%BEcluster-0%E5%92%8Ccluster-10%E7%9A%84conserved-markers"><span class="header-section-number">3.1</span> 寻找cluster 0和cluster 10的conserved markers</a>
  <ul class="collapse">
<li><a href="#%E5%AF%BB%E6%89%BEcluster-0%E7%9A%84conserved-markers" id="toc-寻找cluster-0的conserved-markers" class="nav-link" data-scroll-target="#%E5%AF%BB%E6%89%BEcluster-0%E7%9A%84conserved-markers">寻找cluster 0的conserved markers</a></li>
  <li><a href="#%E6%B7%BB%E5%8A%A0%E5%9F%BA%E5%9B%A0%E6%B3%A8%E9%87%8A%E4%BF%A1%E6%81%AF" id="toc-添加基因注释信息" class="nav-link" data-scroll-target="#%E6%B7%BB%E5%8A%A0%E5%9F%BA%E5%9B%A0%E6%B3%A8%E9%87%8A%E4%BF%A1%E6%81%AF">添加基因注释信息</a></li>
  <li><a href="#%E5%AF%BB%E6%89%BEcluster-10%E7%9A%84conserved-markers" id="toc-寻找cluster-10的conserved-markers" class="nav-link" data-scroll-target="#%E5%AF%BB%E6%89%BEcluster-10%E7%9A%84conserved-markers">寻找cluster 10的conserved markers</a></li>
  </ul>
</li>
  <li>
<a href="#sec-function_to_find_markers" id="toc-sec-function_to_find_markers" class="nav-link" data-scroll-target="#sec-function_to_find_markers"><span class="header-section-number">3.2</span> 批量寻找多个clusters的conserved markers</a>
  <ul class="collapse">
<li><a href="#sec-top_conserved_markers" id="toc-sec-top_conserved_markers" class="nav-link" data-scroll-target="#sec-top_conserved_markers">获取top marker基因</a></li>
  <li><a href="#%E6%A0%B9%E6%8D%AEtop-marker%E5%9F%BA%E5%9B%A0%E9%87%8D%E6%96%B0%E8%AF%84%E4%BC%B0%E7%BB%86%E8%83%9E%E7%BE%A4%E7%9A%84%E6%B3%A8%E9%87%8A%E7%BB%93%E6%9E%9C" id="toc-根据top-marker基因重新评估细胞群的注释结果" class="nav-link" data-scroll-target="#%E6%A0%B9%E6%8D%AEtop-marker%E5%9F%BA%E5%9B%A0%E9%87%8D%E6%96%B0%E8%AF%84%E4%BC%B0%E7%BB%86%E8%83%9E%E7%BE%A4%E7%9A%84%E6%B3%A8%E9%87%8A%E7%BB%93%E6%9E%9C">根据top marker基因重新评估细胞群的注释结果</a></li>
  </ul>
</li>
  </ul>
</li>
  <li><a href="#findmarkers-marker-identification-between-specific-clusters" id="toc-findmarkers-marker-identification-between-specific-clusters" class="nav-link" data-scroll-target="#findmarkers-marker-identification-between-specific-clusters"><span class="header-section-number">4</span> <code>FindMarkers</code>-<strong>Marker identification between specific clusters</strong></a></li>
  <li><a href="#%E6%B3%A8%E9%87%8A%E7%BB%86%E8%83%9Ecluster" id="toc-注释细胞cluster" class="nav-link" data-scroll-target="#%E6%B3%A8%E9%87%8A%E7%BB%86%E8%83%9Ecluster"><span class="header-section-number">5</span> 注释细胞cluster</a></li>
  <li>
<a href="#%E5%8F%AF%E9%80%89%E7%9A%84%E5%90%8E%E7%BB%AD%E5%88%86%E6%9E%90" id="toc-可选的后续分析" class="nav-link" data-scroll-target="#%E5%8F%AF%E9%80%89%E7%9A%84%E5%90%8E%E7%BB%AD%E5%88%86%E6%9E%90"><span class="header-section-number">6</span> 可选的后续分析:</a>
  <ul class="collapse">
<li><a href="#%E6%8F%90%E5%8F%96%E6%84%9F%E5%85%B4%E8%B6%A3%E7%BB%86%E8%83%9E%E4%BA%9A%E7%BE%A4%E8%BF%9B%E4%B8%80%E6%AD%A5%E7%BB%86%E5%88%86explore-a-subset-of-the-cell-types-to-discover-subclusters-of-cells" id="toc-提取感兴趣细胞亚群进一步细分explore-a-subset-of-the-cell-types-to-discover-subclusters-of-cells" class="nav-link" data-scroll-target="#%E6%8F%90%E5%8F%96%E6%84%9F%E5%85%B4%E8%B6%A3%E7%BB%86%E8%83%9E%E4%BA%9A%E7%BE%A4%E8%BF%9B%E4%B8%80%E6%AD%A5%E7%BB%86%E5%88%86explore-a-subset-of-the-cell-types-to-discover-subclusters-of-cells"><span class="header-section-number">6.1</span> 提取感兴趣细胞亚群进一步细分（Explore a subset of the cell types to discover subclusters of cells）</a></li>
  <li><a href="#%E5%8D%95%E7%BB%86%E8%83%9E%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90perform-differential-expression-analysis-between-conditions" id="toc-单细胞差异分析perform-differential-expression-analysis-between-conditions" class="nav-link" data-scroll-target="#%E5%8D%95%E7%BB%86%E8%83%9E%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90perform-differential-expression-analysis-between-conditions"><span class="header-section-number">6.2</span> <strong>单细胞差异分析（</strong>Perform <strong>differential expression analysis</strong> <strong>between conditions）</strong></a></li>
  <li><a href="#%E5%85%8D%E7%96%AB%E7%B1%BB%E7%BE%A4%E5%88%86%E6%9E%90" id="toc-免疫类群分析" class="nav-link" data-scroll-target="#%E5%85%8D%E7%96%AB%E7%B1%BB%E7%BE%A4%E5%88%86%E6%9E%90"><span class="header-section-number">6.3</span> 免疫类群分析</a></li>
  <li><a href="#%E5%85%B3%E9%94%AE%E7%BB%86%E8%83%9E%E7%9A%84%E5%88%86%E5%8C%96%E8%BD%A8%E8%BF%B9" id="toc-关键细胞的分化轨迹" class="nav-link" data-scroll-target="#%E5%85%B3%E9%94%AE%E7%BB%86%E8%83%9E%E7%9A%84%E5%88%86%E5%8C%96%E8%BD%A8%E8%BF%B9"><span class="header-section-number">6.4</span> 关键细胞的分化轨迹</a></li>
  <li><a href="#%E6%8E%A2%E7%B4%A2%E7%BB%86%E8%83%9E-%E7%BB%86%E8%83%9E%E4%BA%92%E4%BD%9C" id="toc-探索细胞-细胞互作" class="nav-link" data-scroll-target="#%E6%8E%A2%E7%B4%A2%E7%BB%86%E8%83%9E-%E7%BB%86%E8%83%9E%E4%BA%92%E4%BD%9C"><span class="header-section-number">6.5</span> 探索细胞-细胞互作</a></li>
  <li><a href="#%E6%8C%96%E6%8E%98%E5%85%B3%E9%94%AE%E8%B0%83%E6%8E%A7%E8%BD%AC%E5%BD%95%E5%9B%A0%E5%AD%90" id="toc-挖掘关键调控转录因子" class="nav-link" data-scroll-target="#%E6%8C%96%E6%8E%98%E5%85%B3%E9%94%AE%E8%B0%83%E6%8E%A7%E8%BD%AC%E5%BD%95%E5%9B%A0%E5%AD%90"><span class="header-section-number">6.6</span> 挖掘关键调控转录因子</a></li>
  <li><a href="#%E6%B9%BF%E5%AE%9E%E9%AA%8C%E9%AA%8C%E8%AF%81%E6%8E%A2%E7%B4%A2%E6%BD%9C%E5%9C%A8%E6%9C%BA%E5%88%B6experimentally-validate-intriguing-markers-for-our-identified-cell-types" id="toc-湿实验验证探索潜在机制experimentally-validate-intriguing-markers-for-our-identified-cell-types" class="nav-link" data-scroll-target="#%E6%B9%BF%E5%AE%9E%E9%AA%8C%E9%AA%8C%E8%AF%81%E6%8E%A2%E7%B4%A2%E6%BD%9C%E5%9C%A8%E6%9C%BA%E5%88%B6experimentally-validate-intriguing-markers-for-our-identified-cell-types"><span class="header-section-number">6.7</span> 湿实验验证，探索潜在机制（Experimentally validate intriguing markers for our identified cell types）</a></li>
  </ul>
</li>
  </ul><div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/djhcod/r-notes/blob/main/single_cell/scRNA-seq_online/09_merged_SC_marker_identification.qmd" class="toc-action">View source</a></p><p><a href="https://github.com/djhcod/r-notes/issues/new" class="toc-action">Report an issue</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title">寻找marker基因+细胞注释</h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Learning Objectives:
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>Describe how to determine markers of individual clusters</li>
<li>Discuss the iterative processes of clustering and marker identification</li>
</ul>
</div>
</div>
<p>Now that we have identified our desired clusters, we can move on to marker identification, which will allow us to <strong>verify the identity of certain clusters</strong> and help surmise the identity of any unknown clusters.</p>
<p><img src="images/sc_workflow_2022-01.jpg" class="img-fluid" width="545"></p>
<hr>
<p><strong>Goals:</strong></p>
<ul>
<li>To <strong>determine the gene markers</strong> for each of the clusters</li>
<li>To <strong>identify cell types</strong> of each cluster using markers</li>
<li>To determine whether there’s a need to <strong>re-cluster based on cell type markers</strong>, perhaps clusters need to be merged or split</li>
</ul>
<p><strong>Challenges:</strong></p>
<ul>
<li>Over-interpretation of the results</li>
<li>Combining different types of marker identification</li>
</ul>
<p><strong>Recommendations:</strong></p>
<ul>
<li>Think of the results as hypotheses that need verification. Inflated p-values can lead to over-interpretation of results (essentially each cell is used as a replicate). Top markers are most trustworthy.</li>
<li>Identify all markers conserved between conditions for each cluster</li>
<li>Identify markers that are differentially expressed between specific clusters</li>
</ul>
<hr>
<section id="数据导入" class="level1" data-number="1"><h1 data-number="1">
<span class="header-section-number">1</span> 数据导入</h1>
<p>载入<a href="../../single_cell/scRNA-seq_online/08_SC_clustering_quality_control.html">此前</a>完成细胞分群质量评估的数据<code>seurat_clustered_qc</code>。</p>
<div class="cell" data-hash="09_merged_SC_marker_identification_cache/html/unnamed-chunk-1_3d7bc534bb72b76b6ab4a9ef62237b62">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span></span>
<span><span class="va">seurat_clustered_qc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="st">"output/scRNA-seq_online/seurat_clustered_qc.rds"</span><span class="op">)</span></span>
<span><span class="va">seurat_clustered_qc</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
31130 features across 29629 samples within 3 assays 
Active assay: RNA (14065 features, 0 variable features)
 2 layers present: counts, data
 2 other assays present: SCT, integrated
 2 dimensional reductions calculated: pca, umap</code></pre>
</div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">seurat_clustered_qc</span>, <span class="fl">3</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                      orig.ident nCount_RNA nFeature_RNA
ctrl_AAACATACAATGCC-1       ctrl       2344          874
ctrl_AAACATACATTTCC-1       ctrl       3124          895
ctrl_AAACATACCAGAAA-1       ctrl       2578          725
                                      seq_folder nUMI nGene log10GenesPerUMI
ctrl_AAACATACAATGCC-1 ctrl_raw_feature_bc_matrix 2344   874        0.8728630
ctrl_AAACATACATTTCC-1 ctrl_raw_feature_bc_matrix 3125   896        0.8447596
ctrl_AAACATACCAGAAA-1 ctrl_raw_feature_bc_matrix 2578   725        0.8384933
                       mitoRatio                 cells sample     S.Score
ctrl_AAACATACAATGCC-1 0.01962457 ctrl_AAACATACAATGCC-1   ctrl  0.04330502
ctrl_AAACATACATTTCC-1 0.01792000 ctrl_AAACATACATTTCC-1   ctrl  0.02661900
ctrl_AAACATACCAGAAA-1 0.01551590 ctrl_AAACATACCAGAAA-1   ctrl -0.04670650
                        G2M.Score Phase mitoFr nCount_SCT nFeature_SCT
ctrl_AAACATACAATGCC-1  0.05422631   G2M Medium       1572          829
ctrl_AAACATACATTTCC-1  0.05159679   G2M Medium       1572          718
ctrl_AAACATACCAGAAA-1 -0.04841661    G1 Medium       1553          648
                      integrated_snn_res.0.4 integrated_snn_res.0.6
ctrl_AAACATACAATGCC-1                      2                      1
ctrl_AAACATACATTTCC-1                      0                      2
ctrl_AAACATACCAGAAA-1                      0                      3
                      integrated_snn_res.0.8 integrated_snn_res.1
ctrl_AAACATACAATGCC-1                      2                    2
ctrl_AAACATACATTTCC-1                      1                    0
ctrl_AAACATACCAGAAA-1                      3                   15
                      integrated_snn_res.1.4 seurat_clusters
ctrl_AAACATACAATGCC-1                      5               5
ctrl_AAACATACATTTCC-1                      0               0
ctrl_AAACATACCAGAAA-1                     19              19</code></pre>
</div>
</div>
<p>Before we start our marker identification we will explicitly set our default assay, we want to use the <strong>normalized data, but not the integrated data</strong>.</p>
<div class="cell" data-hash="09_merged_SC_marker_identification_cache/html/unnamed-chunk-2_93666f32bbd6c706505123f8bad301ad">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/DefaultAssay.html">DefaultAssay</a></span><span class="op">(</span><span class="va">seurat_clustered_qc</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"RNA"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Important
</div>
</div>
<div class="callout-body-container callout-body">
<p><a href="https://hbctraining.github.io/scRNA-seq_online/lessons/09_merged_SC_marker_identification.html">原教程</a>将默认的assay设置为了”RNA”，其解释如下：The default assay should have already been <code>RNA</code>, because we set it up in the <a href="../../single_cell/scRNA-seq_online/08_SC_clustering_quality_control.html#sec-explore_known_markers">previous clustering quality control lesson</a>. But we encourage you to run this line of code above to be absolutely sure in case the active slot was changed somewhere upstream in your analysis. Note that the raw and normalized counts are stored in the <code>counts</code> and <code>data</code> slots of <code>RNA</code> assay. By default, the functions for finding markers will use normalized data. （关于<code>FindMarkers</code>为什么要使用”RNA” assay的更多解释，参阅<a href="https://github.com/hbctraining/scRNA-seq_online/issues/58">此链接</a>）</p>
<p>而在<a href="../../single_cell/seurat/marker_gene_identification.html#sec-findmarkers_function">Seurat V5的官方教程</a>中，对于经过<code>SCTransform</code>归一化处理后的单细胞数据，在进行<code>FindMarkers</code>差异分析之前，需要先运行<code>seurat_clustered_qc &lt;- PrepSCTFindMarkers(seurat_clustered_qc)</code>，来预处理SCT assay。详细解释见<a href="https://www.jianshu.com/p/fb2e43905559">此链接</a>。</p>
<p>如果是基于<code>NormalizeData</code>标准化的单细胞数据，和这里一样，需要使用”RNA” assay进行差异分析，如果不是，需要通过<code>DefaultAssay(seurat_clustered_qc) &lt;- "RNA"</code>进行设定。</p>
<p>这里为了和原教程保持一致，默认的将assay设置为”RNA”。</p>
</div>
</div>
<p>Our clustering analysis resulted in the following clusters:</p>
<div class="cell" data-hash="09_merged_SC_marker_identification_cache/html/unnamed-chunk-3_fd9c976be0abe58a929675600e7de2e4">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">seurat_clustered_qc</span>, label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="09_merged_SC_marker_identification_files/figure-html/unnamed-chunk-3-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Remember that we had the following questions from the <a href="../../single_cell/scRNA-seq_online/08_SC_clustering_quality_control.html#sec-Preliminary_cell_type_identification_results">clustering analysis</a>:</p>
<blockquote class="blockquote">
<ol type="1">
<li><p>Do the clusters corresponding to the same cell types have biologically meaningful differences? Are there subpopulations of these cell types?</p></li>
<li><p>Can we acquire higher confidence in these cell type identities by identifying other marker genes for these clusters?</p></li>
</ol>
</blockquote>
<p>There are a few different types of marker identification that we can explore using Seurat to get to the answer of these questions. Each with their own benefits and drawbacks:</p>
<ol type="1">
<li>
<p><strong>Identification of all markers for each cluster</strong></p>
<p>This analysis compares each cluster against all others and outputs the genes that are differentially expressed/present.</p>
<ul>
<li><em>Useful for identifying <strong>unknown clusters</strong> and i<strong>mproving confidence</strong> in hypothesized cell types.</em></li>
</ul>
</li>
<li>
<p><strong>Identification of conserved markers for each cluster</strong></p>
<p>This analysis looks for genes that are differentially expressed/present within each condition first, and then reports those <strong>genes that are conserved in the cluster across all conditions</strong>. These genes can help to figure out the identity for the cluster.</p>
<ul>
<li><em>Useful with <strong>more than one condition</strong> to identify cell type markers that are conserved across conditions.</em></li>
</ul>
</li>
<li>
<p><strong>Marker identification between specific clusters</strong></p>
<p>This analysis explores differentially expressed genes between specific clusters.</p>
<ul>
<li><em>Useful for determining differences in gene expression between clusters that appear to be representing the same celltype (i.e with markers that are similar) from the above analyses.</em></li>
</ul>
</li>
</ol>
<hr></section><section id="findallmarkers-identification-of-all-markers-for-each-cluster" class="level1" data-number="2"><h1 data-number="2">
<span class="header-section-number">2</span> <code>FindAllMarkers</code>-Identification of all markers for each cluster</h1>
<p>This type of analysis is typically recommended for when <strong>evaluating a single sample group/condition</strong>. With the <code><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html">FindAllMarkers()</a></code> function we are comparing each cluster against all other clusters to identify potential marker genes. The cells in each cluster are treated as replicates, and essentially a <strong>differential expression analysis</strong> is performed with some statistical test.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>The default is a <strong>Wilcoxon Rank Sum test</strong>, but there are other options available.</p>
</div>
</div>
<p><img src="images/marker_ident_function1.png" class="img-fluid" width="398"></p>
<p>The <code><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html">FindAllMarkers()</a></code> function has <strong>three important arguments</strong> which provide thresholds for determining whether a gene is a marker:</p>
<ul>
<li>
<code>logfc.threshold</code>: minimum log2 fold change for average expression of gene in cluster relative to the average expression in all other clusters combined. Default is <strong>0.25</strong>.
<ul>
<li>
<strong>Cons:</strong>
<ul>
<li>could miss those cell <strong>markers that are expressed in a small fraction of cells</strong> within the cluster of interest, but not in the other clusters, if the average logfc doesn’t meet the threshold</li>
<li>could return a lot of metabolic/ribosomal genes due to slight differences in metabolic output by different cell types, which are not as useful to distinguish cell type identities</li>
</ul>
</li>
</ul>
</li>
<li>
<code>min.diff.pct</code>: minimum percent difference between the percent of cells expressing the gene in the cluster and the percent of cells expressing gene in all other clusters combined.
<ul>
<li>
<strong>Cons:</strong> could miss those cell markers that are expressed in all cells, but are highly up-regulated in this specific cell type</li>
</ul>
</li>
<li>
<code>min.pct</code>: only test genes that are detected in a minimum fraction of cells in either of the two populations. Meant to speed up the function by not testing genes that are very infrequently expressed. Default is 0.1.
<ul>
<li>
<strong>Cons:</strong> if set to a very high value could incur many false negatives due to the fact that not all genes are detected in all cells (even if it is expressed)</li>
</ul>
</li>
</ul>
<p>You could use any combination of these arguments depending on how stringent/lenient you want to be. Also, by default this function will return to you genes that exhibit both positive and negative expression changes. Typically, we add an argument <code>only.pos</code> to opt for keeping only the positive changes. The code to find markers for each cluster is shown below.</p>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Important
</div>
</div>
<div class="callout-body-container callout-body">
<p>In Seurat v5, we use the&nbsp;<strong><code>presto</code></strong> package&nbsp;(as described&nbsp;<a href="https://www.biorxiv.org/content/10.1101/653253v1">here</a>&nbsp;and available for installation&nbsp;<a href="https://github.com/immunogenomics/presto">here</a>), to dramatically improve the speed of DE analysis, particularly for large datasets (见<a href="../../single_cell/seurat/marker_gene_identification.html#sec-findmarkers_function">Seurat-FindMarkers</a>).&nbsp;因此，需要先安装<code>presto</code>包：</p>
<div class="cell" data-hash="09_merged_SC_marker_identification_cache/html/unnamed-chunk-4_4d4f8ca01666f2d172b2d18ee104dd1a">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="co"># install.packages("devtools")</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">"immunogenomics/presto"</span>)</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<div class="cell" data-hash="09_merged_SC_marker_identification_cache/html/unnamed-chunk-5_58d26009456eb1e14ac53709b144d816">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Find markers for every cluster compared to all remaining cells, report only the positive ones</span></span>
<span><span class="va">markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html">FindAllMarkers</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">seurat_clustered_qc</span>, </span>
<span>                          only.pos <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                          logfc.threshold <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span> </span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">markers</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       p_val avg_log2FC pct.1 pct.2 p_val_adj cluster   gene
CCR7       0   1.494342 0.885 0.399         0       0   CCR7
SELL       0   1.847509 0.726 0.277         0       0   SELL
GIMAP7     0   1.482245 0.872 0.443         0       0 GIMAP7
LTB        0   1.561875 0.735 0.311         0       0    LTB
LDHB       0   1.569828 0.733 0.330         0       0   LDHB
TRAC       0   1.285588 0.746 0.356         0       0   TRAC</code></pre>
</div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">tail</a></span><span class="op">(</span><span class="va">markers</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>               p_val avg_log2FC pct.1 pct.2 p_val_adj cluster    gene
DNAJC213 0.009747737  0.5018165 0.056 0.022         1      16 DNAJC21
POGK     0.009763789  1.4868037 0.032 0.010         1      16    POGK
ZNF5443  0.009781767  0.6873261 0.024 0.006         1      16  ZNF544
MRPS313  0.009912914  0.4306063 0.105 0.051         1      16  MRPS31
UCKL11   0.009958018  0.5284076 0.040 0.013         1      16   UCKL1
QTRT22   0.009998863  0.3087035 0.056 0.022         1      16   QTRT2</code></pre>
</div>
</div>
<p>The results data frame has the following columns :</p>
<ul>
<li><p><code>p_val</code>&nbsp;: p-value (unadjusted)</p></li>
<li><p><code>avg_log2FC</code>&nbsp;: log fold-change of the average expression between the two groups. Positive values indicate that the feature is more highly expressed in the first group.</p></li>
<li><p><code>pct.1</code>&nbsp;: The percentage of cells where the feature is detected in the first group</p></li>
<li><p><code>pct.2</code>&nbsp;: The percentage of cells where the feature is detected in the second group</p></li>
<li><p><code>p_val_adj</code>&nbsp;: Adjusted p-value, based on&nbsp;<strong>Bonferroni correction</strong>&nbsp;using all features in the dataset.</p></li>
</ul>
<p>获取每个cluster的前10个marker基因 ( 详细解释见下面的 <a href="#sec-top_conserved_markers">Section&nbsp;3.2.1</a> ):</p>
<div class="cell" data-hash="09_merged_SC_marker_identification_cache/html/unnamed-chunk-6_2ad3bad2f901ab2e4e718d8a781c566a">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></span>
<span><span class="co"># 获取每个cluster的前10个marker基因</span></span>
<span><span class="va">top_markers</span> <span class="op">&lt;-</span> <span class="va">markers</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html">slice_max</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">10</span>, order_by <span class="op">=</span> <span class="va">avg_log2FC</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="findconservedmarkers-identification-of-conserved-markers-in-all-conditions" class="level1" data-number="3"><h1 data-number="3">
<span class="header-section-number">3</span> <code>FindConservedMarkers</code>-Identification of conserved markers in all conditions</h1>
<p>Since we have samples representing different conditions in our dataset, <strong>our best option is to find conserved markers</strong>. This function internally <strong>separates out cells by sample group/condition</strong>, and then performs differential gene expression testing for a single specified cluster against all other clusters (or a second cluster, if specified). Gene-level p-values are computed for each condition and then combined across groups using meta-analysis methods from the MetaDE R package.</p>
<p><img src="images/marker_ident_function2.png" class="img-fluid" width="360"></p>
<section id="寻找cluster-0和cluster-10的conserved-markers" class="level2" data-number="3.1"><h2 data-number="3.1" class="anchored" data-anchor-id="寻找cluster-0和cluster-10的conserved-markers">
<span class="header-section-number">3.1</span> 寻找cluster 0和cluster 10的conserved markers</h2>
<p>我们首先通过寻找cluster 0和cluster 10的conserved markers来初步学习<code>FindConservedMarkers</code>函数的用法。</p>
<p>for <code>FindConservedMarkers</code>, you will recognize some of the arguments we described previously for the <code><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html">FindAllMarkers()</a></code> function; this is because internally it is using that function to first find markers within each group. Here, we list some additional arguments which provide for when using <code><a href="https://satijalab.org/seurat/reference/FindConservedMarkers.html">FindConservedMarkers()</a></code>:</p>
<ul>
<li>
<code>ident.1</code>: this function only evaluates one cluster at a time; here you would specify the cluster of interest.</li>
<li>
<code>grouping.var</code>: the variable (column header) in your metadata which specifies the separation of cells into groups</li>
</ul>
<p><code>FindConservedMarkers</code>函数会调用<code>metap</code>包，<code>metap</code>包需要<code>multtest</code>包，所以需要先安装这两个依赖包：</p>
<div class="cell" data-hash="09_merged_SC_marker_identification_cache/html/unnamed-chunk-7_e88e0aef433b0bebb747a2c266bb6756">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="st">'multtest'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">'metap'</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="寻找cluster-0的conserved-markers" class="level3"><h3 class="anchored" data-anchor-id="寻找cluster-0的conserved-markers">寻找cluster 0的conserved markers</h3>
<p>For our analysis we will be fairly lenient and <strong>use only the log fold change threshold greater than 0.25</strong>. We will also specify to return only the positive markers for each cluster.</p>
<p>Let’s <strong>test it out on cluster 0</strong> to see how it works:</p>
<div class="cell" data-hash="09_merged_SC_marker_identification_cache/html/unnamed-chunk-8_15837dd4149c3d7f056e793913580801">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">cluster0_conserved_markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindConservedMarkers.html">FindConservedMarkers</a></span><span class="op">(</span><span class="va">seurat_clustered_qc</span>,</span>
<span>                                                   ident.1 <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                                                   grouping.var <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>                                                   only.pos <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                                   logfc.threshold <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">cluster0_conserved_markers</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       stim_p_val stim_avg_log2FC stim_pct.1 stim_pct.2 stim_p_val_adj
CCR7            0        1.481347      0.927      0.430              0
SELL            0        1.724888      0.832      0.370              0
LDHB            0        1.720061      0.732      0.304              0
GIMAP7          0        1.463588      0.934      0.507              0
LTB             0        1.538317      0.702      0.295              0
RPL10A          0        1.216299      0.966      0.664              0
          ctrl_p_val ctrl_avg_log2FC ctrl_pct.1 ctrl_pct.2 ctrl_p_val_adj
CCR7    0.000000e+00       1.4694462      0.839      0.368   0.000000e+00
SELL    0.000000e+00       2.0273581      0.610      0.186   0.000000e+00
LDHB   2.056415e-293       1.4530062      0.734      0.356  2.892347e-289
GIMAP7  0.000000e+00       1.4612133      0.804      0.380   0.000000e+00
LTB     0.000000e+00       1.5963402      0.770      0.327   0.000000e+00
RPL10A  0.000000e+00       0.9640366      0.969      0.809   0.000000e+00
            max_pval minimump_p_val
CCR7    0.000000e+00              0
SELL    0.000000e+00              0
LDHB   2.056415e-293              0
GIMAP7  0.000000e+00              0
LTB     0.000000e+00              0
RPL10A  0.000000e+00              0</code></pre>
</div>
</div>
<p>The output from the <code>FindConservedMarkers</code> function, is a matrix containing a ranked list of putative markers listed by gene ID for the cluster we specified, and associated statistics. Note that the same set of statistics are computed for each group (in our case, Ctrl and Stim) and the last two columns (<code>max_pval</code>和<code>minimump_p_val</code>) correspond to the combined p-value across the two groups. We describe some of these columns below:</p>
<ul>
<li>
<code>gene</code>: gene symbol</li>
<li>
<code>condition_p_val</code>(即本例中的“stim_p_val”和“ctrl_p_val”列，后面同理): p-value not adjusted for multiple test correction for condition</li>
<li>
<code>condition_avg_log2FC</code>: average log fold change for condition. Positive values indicate that the gene is more highly expressed in the cluster.</li>
<li>
<code>condition_pct.1</code>: percentage of cells where the gene is detected in the cluster for condition</li>
<li>
<code>condition_pct.2</code>: percentage of cells where the gene is detected on average in the other clusters for condition</li>
<li>
<code>condition_p_val_adj</code>: adjusted p-value for condition, based on <strong>bonferroni correction</strong> using all genes in the dataset, used to determine significance</li>
<li>
<code>max_pval</code>: largest p value of p value calculated by each group/condition</li>
<li>
<code>minimump_p_val</code>: combined p value</li>
</ul>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>The&nbsp;<code>condition_p_val</code>,&nbsp;<code>condition_avg_log2FC</code>,&nbsp;<code>condition_pct.1</code>,&nbsp;<code>condition_pct.2</code>, and&nbsp;<code>condition_p_val_adj</code>&nbsp;mean the same thing as they do in&nbsp;<code>FindMarkers</code>, just restricted to only the cells present in group X. The&nbsp;<code>max_pval</code>&nbsp;is the maximum p-value across all groups. The&nbsp;<code>mimimump_p_val</code>&nbsp;represents one way of doing a meta-analysis of significance values (combining p-values across different tests).</p>
<p>(来自：<a href="https://github.com/satijalab/seurat/issues/1164" class="uri">https://github.com/satijalab/seurat/issues/1164</a>)</p>
</div>
</div>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Warning
</div>
</div>
<div class="callout-body-container callout-body">
<p>Since each cell is being treated as a replicate this will result in inflated p-values within each group! A gene may have an incredibly low p-value &lt; 1e-50 but that doesn’t translate as a highly reliable marker gene.</p>
</div>
</div>
<p>When looking at the output, <u><strong>we suggest looking for markers with large differences in expression between <code>pct.1</code> and <code>pct.2</code> and larger fold changes</strong></u>. For instance if <code>pct.1</code> = 0.90 and <code>pct.2</code> = 0.80, it may not be as exciting of a marker. However, if <code>pct.2</code> = 0.1 instead, the bigger difference would be more convincing. Also, of interest is if the majority of cells expressing the marker is in my cluster of interest. If <code>pct.1</code> is low, such as 0.3, it may not be as interesting. Both of these are also possible parameters to include when running the function, as described above.</p>
</section><section id="添加基因注释信息" class="level3"><h3 class="anchored" data-anchor-id="添加基因注释信息">添加基因注释信息</h3>
<p>It can be helpful to add columns with gene annotation information. In order to do that we will load in an annotation file located in your <code>data</code> folder, using the code provided below:</p>
<div class="cell" data-hash="09_merged_SC_marker_identification_cache/html/unnamed-chunk-9_1745189a28708bd04ef72eedaf3afc91">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">annotations</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="st">"data/scRNA-seq_online/annotations.rds"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">annotations</span>, <span class="fl">3</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          gene_id gene_name seq_name                       gene_biotype
1 ENSG00000290825   DDX11L2        1                             lncRNA
6 ENSG00000223972   DDX11L1        1 transcribed_unprocessed_pseudogene
7 ENSG00000227232    WASH7P        1             unprocessed_pseudogene
                                                                                    description
1 DEAD/H-box helicase 11 like 2 (pseudogene) [Source:NCBI gene (formerly Entrezgene);Acc:84771]
6                DEAD/H-box helicase 11 like 1 (pseudogene) [Source:HGNC Symbol;Acc:HGNC:37102]
7                         WASP family homolog 7, pseudogene [Source:HGNC Symbol;Acc:HGNC:38034]</code></pre>
</div>
</div>
<p>该数据的”description”列即对基因的注释，下面我们把这一列通过匹配基因名将其添加到cluster0_conserved_markers数据框中。</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-7-contents" aria-controls="callout-7" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
获取基因注释信息的方法
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-7" class="callout-7-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>首先从<code>BiocManager</code>安装<a href="https://bioconductor.org/packages/release/bioc/html/AnnotationHub.html"><code>AnnotationHub</code>包</a>和<a href="https://bioconductor.org/packages/release/bioc/html/ensembldb.html"><code>ensembldb</code>包</a>：</p>
<div class="cell" data-hash="09_merged_SC_marker_identification_cache/html/unnamed-chunk-10_318d29c9f538f7cc1c96a636c8ad8834">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="st">"AnnotationHub"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="st">"ensembldb"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>从AnnotationHub下载并提取所需的注释信息数据库:</strong></p>
<p>To access the various annotations available from Ensembl for human, we need to first connect to <code>AnnotationHub</code>, then specify the organism and database we are interested in.</p>
<div class="cell" data-hash="09_merged_SC_marker_identification_cache/html/unnamed-chunk-11_e5dedf6c374f2b073572bc7dd21c356b">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 从AnnotationHub下载注释信息数据库</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">AnnotationHub</span><span class="op">)</span></span>
<span><span class="va">ah</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html">AnnotationHub</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">ah</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre><code>AnnotationHub with 70762 records
# snapshotDate(): 2023-10-20
# $dataprovider: Ensembl, BroadInstitute, UCSC, ftp://ftp.ncbi.nlm.nih.gov/g...
# $species: Homo sapiens, Mus musculus, Drosophila melanogaster, Bos taurus,...
# $rdataclass: GRanges, TwoBitFile, BigWigFile, EnsDb, Rle, OrgDb, SQLiteFil...
# additional mcols(): taxonomyid, genome, description,
#   coordinate_1_based, maintainer, rdatadateadded, preparerclass, tags,
#   rdatapath, sourceurl, sourcetype 
# retrieve records with, e.g., 'object[["AH5012"]]' 

             title                              
  AH5012   | Chromosome Band                    
  AH5013   | STS Markers                        
  AH5014   | FISH Clones                        
  AH5015   | Recomb Rate                        
  AH5016   | ENCODE Pilot                       
  ...        ...                                
  AH116159 | org.Aegialitis_vocifera.eg.sqlite  
  AH116160 | org.Charadrius_vociferous.eg.sqlite
  AH116161 | org.Charadrius_vociferus.eg.sqlite 
  AH116162 | org.Oxyechus_vociferus.eg.sqlite   
  AH116163 | org.Drosophila_erecta.eg.sqlite    </code></pre>
<div class="cell" data-hash="09_merged_SC_marker_identification_cache/html/unnamed-chunk-12_1b8d056e6d440ec7ad57cf3358e6bb48">
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html">unique</a></span><span class="op">(</span><span class="va">ah</span><span class="op">$</span><span class="va">species</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre><code>[1] "Homo sapiens"         "Vicugna pacos"        "Dasypus novemcinctus"
[4] "Otolemur garnettii"   "Papio hamadryas"      "Papio anubis"     </code></pre>
<div class="cell" data-hash="09_merged_SC_marker_identification_cache/html/unnamed-chunk-13_ed8bfbce5284fc736ddaa6952fa11f67">
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Access the Ensembl database for organism</span></span>
<span><span class="va">ahDb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html">query</a></span><span class="op">(</span><span class="va">ah</span>, </span>
<span>              pattern <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Homo sapiens"</span>, <span class="st">"EnsDb"</span><span class="op">)</span>, </span>
<span>              ignore.case <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">ahDb</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre><code>AnnotationHub with 25 records
# snapshotDate(): 2023-10-20
# $dataprovider: Ensembl
# $species: Homo sapiens
# $rdataclass: EnsDb
# additional mcols(): taxonomyid, genome, description,
#   coordinate_1_based, maintainer, rdatadateadded, preparerclass, tags,
#   rdatapath, sourceurl, sourcetype 
# retrieve records with, e.g., 'object[["AH53211"]]' 

             title                             
  AH53211  | Ensembl 87 EnsDb for Homo Sapiens 
  AH53715  | Ensembl 88 EnsDb for Homo Sapiens 
  AH56681  | Ensembl 89 EnsDb for Homo Sapiens 
  AH57757  | Ensembl 90 EnsDb for Homo Sapiens 
  AH60773  | Ensembl 91 EnsDb for Homo Sapiens 
  ...        ...                               
  AH100643 | Ensembl 106 EnsDb for Homo sapiens
  AH104864 | Ensembl 107 EnsDb for Homo sapiens
  AH109336 | Ensembl 108 EnsDb for Homo sapiens
  AH109606 | Ensembl 109 EnsDb for Homo sapiens
  AH113665 | Ensembl 110 EnsDb for Homo sapiens</code></pre>
<p>Next, we acquire the latest annotation files from this Ensembl database.</p>
<p>We can first check which annotation versions are available. Since we want the most recent, we will return the AnnotationHub ID for this database:</p>
<div class="cell" data-hash="09_merged_SC_marker_identification_cache/html/unnamed-chunk-14_72245e6e4f33df3781313c33dad8ffa6">
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Acquire the latest annotation files</span></span>
<span><span class="va">id</span> <span class="op">&lt;-</span> <span class="va">ahDb</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">mcols</span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html">rownames</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html">tail</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">id</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre><code>[1] "AH113665"</code></pre>
<p>Finally, we can use the AnnotationHub connection to download the appropriate Ensembl database.</p>
<div class="cell" data-hash="09_merged_SC_marker_identification_cache/html/unnamed-chunk-15_a91ddfcfacf369019a83812317928b5e">
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Download the appropriate Ensembldb database</span></span>
<span><span class="co"># 需要开启全局代理</span></span>
<span><span class="va">edb</span> <span class="op">&lt;-</span> <span class="va">ah</span><span class="op">[[</span><span class="va">id</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">edb</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre><code>EnsDb for Ensembl:
|Backend: SQLite
|Db type: EnsDb
|Type of Gene ID: Ensembl Gene ID
|Supporting package: ensembldb
|Db created by: ensembldb package from Bioconductor
|script_version: 0.3.10
|Creation time: Mon Aug  7 09:02:07 2023
|ensembl_version: 110
|ensembl_host: 127.0.0.1
|Organism: Homo sapiens
|taxonomy_id: 9606
|genome_build: GRCh38
|DBSCHEMAVERSION: 2.2
|common_name: human
|species: homo_sapiens
| No. of genes: 71440.
| No. of transcripts: 278545.
|Protein data available.</code></pre>
<p><strong>提取并保存注释信息：</strong></p>
<p>And to extract gene-level information we can use the Ensembldb function <code>genes()</code> to return a data frame of annotations.</p>
<div class="cell" data-hash="09_merged_SC_marker_identification_cache/html/unnamed-chunk-16_27126be5040a81e7b4e647529fb5c048">
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Extract gene-level information from database</span></span>
<span><span class="va">annotations</span> <span class="op">&lt;-</span> <span class="fu">genes</span><span class="op">(</span><span class="va">edb</span>, return.type <span class="op">=</span> <span class="st">"data.frame"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html">colnames</a></span><span class="op">(</span><span class="va">annotations</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre><code>[1] "gene_id"              "gene_name"            "gene_biotype"        
[4] "gene_seq_start"       "gene_seq_end"         "seq_name"            
[7] "seq_strand"           "seq_coord_system"     "description"         
[10] "gene_id_version"      "canonical_transcript" "symbol"              
[13] "entrezid"   </code></pre>
<p>We aren’t interested in all of the information present in this <code>annotations</code> file, so we are going to extract that which is useful to us.</p>
<div class="cell" data-hash="09_merged_SC_marker_identification_cache/html/unnamed-chunk-17_49757d7ca0f50a4b032545ae740f4576">
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Select annotations of interest</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></span>
<span><span class="va">annotations</span> <span class="op">&lt;-</span> <span class="va">annotations</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">gene_id</span>, <span class="va">gene_name</span>, <span class="va">seq_name</span>, <span class="va">gene_biotype</span>, <span class="va">description</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>保存到本地:</strong></p>
<div class="cell" data-hash="09_merged_SC_marker_identification_cache/html/unnamed-chunk-18_6153fb1dab883b11aee9e5f62fd8407f">
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">annotations</span>, file <span class="op">=</span> <span class="st">"data/scRNA-seq_online/annotations.rds"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<p>First, we will turn the row names with gene identifiers into its own columns. Then we will merge this annotation file with our results from the <code><a href="https://satijalab.org/seurat/reference/FindConservedMarkers.html">FindConservedMarkers()</a></code>:</p>
<ol type="1">
<li><p>首先通过<code>tibble</code>包的<code>rownames_to_column</code>函数将“cluster0_conserved_markers”数据框的行名（基因symbol）转换成新的一列“gene”</p></li>
<li><p>然后，通过<code>dplyr</code>包的<code>left_join</code>函数合并“cluster0_conserved_markers”数据（<code>x</code>）和“annotations”数据框的 “description”列（<code>y</code>）。通过匹配“cluster0_conserved_markers”数据中的“gene”列（上一步生成）和“annotations”数据的“gene_name”列来进行合并。同时，<code>left_join</code>会保留<code>x</code>（这里即上一步添加了“gene列”的“cluster0_conserved_markers”数据框）中的所有值；而删除y（“annotations”数据框）中用于匹配的那一列（即“gene_name”）</p></li>
</ol>
<div class="cell" data-hash="09_merged_SC_marker_identification_cache/html/unnamed-chunk-19_3a2b41f500082b29af400f09c37c9275">
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Combine markers with gene descriptions </span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tibble.tidyverse.org/">tibble</a></span><span class="op">)</span> <span class="co"># 调用rownames_to_column函数</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span> <span class="co"># 调用left_join函数</span></span>
<span><span class="va">cluster0_ann_markers</span> <span class="op">&lt;-</span> <span class="va">cluster0_conserved_markers</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames_to_column</a></span><span class="op">(</span>var <span class="op">=</span> <span class="st">"gene"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="co"># left_join保留x中的所有观测</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join</a></span><span class="op">(</span>y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html">unique</a></span><span class="op">(</span><span class="va">annotations</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"gene_name"</span>, <span class="st">"description"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>,</span>
<span>            by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"gene"</span> <span class="op">=</span> <span class="st">"gene_name"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">cluster0_ann_markers</span>, <span class="fl">3</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  gene stim_p_val stim_avg_log2FC stim_pct.1 stim_pct.2 stim_p_val_adj
1 CCR7          0        1.481347      0.927      0.430              0
2 SELL          0        1.724888      0.832      0.370              0
3 LDHB          0        1.720061      0.732      0.304              0
     ctrl_p_val ctrl_avg_log2FC ctrl_pct.1 ctrl_pct.2 ctrl_p_val_adj
1  0.000000e+00        1.469446      0.839      0.368   0.000000e+00
2  0.000000e+00        2.027358      0.610      0.186   0.000000e+00
3 2.056415e-293        1.453006      0.734      0.356  2.892347e-289
       max_pval minimump_p_val
1  0.000000e+00              0
2  0.000000e+00              0
3 2.056415e-293              0
                                                        description
1 C-C motif chemokine receptor 7 [Source:HGNC Symbol;Acc:HGNC:1608]
2                    selectin L [Source:HGNC Symbol;Acc:HGNC:10720]
3        lactate dehydrogenase B [Source:HGNC Symbol;Acc:HGNC:6541]</code></pre>
</div>
</div>
</section><section id="寻找cluster-10的conserved-markers" class="level3"><h3 class="anchored" data-anchor-id="寻找cluster-10的conserved-markers">寻找cluster 10的conserved markers</h3>
<p>In the <a href="../../single_cell/scRNA-seq_online/08_SC_clustering_quality_control.html#sec-fcgr3a_monocyte_markers">previous lesson</a>, we identified <strong>cluster 10</strong> as FCGR3A+ monocytes by inspecting the expression of known cell markers FCGR3A and MS4A7.</p>
<div class="cell" data-hash="09_merged_SC_marker_identification_cache/html/unnamed-chunk-20_6c27a810fc9ebe4c091a4b2c2d0ac48f">
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html">FeaturePlot</a></span><span class="op">(</span><span class="va">seurat_clustered_qc</span>, </span>
<span>            reduction <span class="op">=</span> <span class="st">"umap"</span>, </span>
<span>            features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"FCGR3A"</span>, <span class="st">"MS4A7"</span><span class="op">)</span>, </span>
<span>            order <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            min.cutoff <span class="op">=</span> <span class="st">'q10'</span>, </span>
<span>            label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="09_merged_SC_marker_identification_files/figure-html/unnamed-chunk-20-1.png" class="img-fluid" width="960"></p>
</div>
</div>
<p>Now, we use <code><a href="https://satijalab.org/seurat/reference/FindConservedMarkers.html">FindConservedMarkers()</a></code> function to find conserved markers for cluster 10.</p>
<div class="cell" data-hash="09_merged_SC_marker_identification_cache/html/unnamed-chunk-21_d3759b42e20968a72bcbd94d4f369f93">
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">cluster10_conserved_markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindConservedMarkers.html">FindConservedMarkers</a></span><span class="op">(</span><span class="va">seurat_clustered_qc</span>,</span>
<span>                                                    ident.1 <span class="op">=</span> <span class="fl">10</span>,</span>
<span>                                                    grouping.var <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>                                                    only.pos <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                                    logfc.threshold <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cluster10_ann_markers</span> <span class="op">&lt;-</span> <span class="va">cluster10_conserved_markers</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames_to_column</a></span><span class="op">(</span>var<span class="op">=</span><span class="st">"gene"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join</a></span><span class="op">(</span>y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html">unique</a></span><span class="op">(</span><span class="va">annotations</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"gene_name"</span>, <span class="st">"description"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>,</span>
<span>            by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"gene"</span> <span class="op">=</span> <span class="st">"gene_name"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">cluster10_ann_markers</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    gene stim_p_val stim_avg_log2FC stim_pct.1 stim_pct.2 stim_p_val_adj
1 FCGR3A          0        5.208580      0.988      0.101              0
2 MS4A4A          0        5.106558      0.893      0.051              0
3  MS4A7          0        4.313442      0.995      0.159              0
4 CXCL16          0        4.097853      0.924      0.108              0
5   VMO1          0        7.808405      0.758      0.016              0
6   LST1          0        3.195206      0.873      0.148              0
  ctrl_p_val ctrl_avg_log2FC ctrl_pct.1 ctrl_pct.2 ctrl_p_val_adj max_pval
1          0        4.391965      0.980      0.141              0        0
2          0        5.623548      0.577      0.016              0        0
3          0        4.307014      0.962      0.122              0        0
4          0        3.467690      0.945      0.148              0        0
5          0        6.478637      0.850      0.039              0        0
6          0        3.288439      0.932      0.165              0        0
  minimump_p_val
1              0
2              0
3              0
4              0
5              0
6              0
                                                                   description
1                    Fc gamma receptor IIIa [Source:HGNC Symbol;Acc:HGNC:3619]
2          membrane spanning 4-domains A4A [Source:HGNC Symbol;Acc:HGNC:13371]
3           membrane spanning 4-domains A7 [Source:HGNC Symbol;Acc:HGNC:13378]
4          C-X-C motif chemokine ligand 16 [Source:HGNC Symbol;Acc:HGNC:16642]
5 vitelline membrane outer layer 1 homolog [Source:HGNC Symbol;Acc:HGNC:30387]
6          leukocyte specific transcript 1 [Source:HGNC Symbol;Acc:HGNC:14189]</code></pre>
</div>
</div>
<blockquote class="blockquote">
<p>可以发现cluster10的FCGR3A和MS4A7的表达比例显著高于其他cluster。符合此前的判断。</p>
</blockquote>
<hr></section></section><section id="sec-function_to_find_markers" class="level2" data-number="3.2"><h2 data-number="3.2" class="anchored" data-anchor-id="sec-function_to_find_markers">
<span class="header-section-number">3.2</span> 批量寻找多个clusters的conserved markers</h2>
<p>The function <code><a href="https://satijalab.org/seurat/reference/FindConservedMarkers.html">FindConservedMarkers()</a></code> <strong>accepts a single cluster at a time</strong>, and we could run this function as many times as we have clusters. However, this is not very efficient. Instead we will first create a function to find the conserved markers including all the parameters we want to include. We will also <strong>add a few lines of code to modify the output</strong>. Our function will:</p>
<ol type="1">
<li>Run the <code><a href="https://satijalab.org/seurat/reference/FindConservedMarkers.html">FindConservedMarkers()</a></code> function</li>
<li>Transfer row names to a column using <code><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames_to_column()</a></code> function</li>
<li>Merge in annotations</li>
<li>Create the column of cluster IDs using the <code><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html">cbind()</a></code> function</li>
</ol>
<div class="cell" data-hash="09_merged_SC_marker_identification_cache/html/unnamed-chunk-22_67d1a4383c2722ec5b7cfd8d9ae91b65">
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Create function to get conserved markers for any given cluster</span></span>
<span><span class="va">get_conserved</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/FindConservedMarkers.html">FindConservedMarkers</a></span><span class="op">(</span><span class="va">seurat_clustered_qc</span>,</span>
<span>                       ident.1 <span class="op">=</span> <span class="va">cluster</span>,</span>
<span>                       grouping.var <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>                       only.pos <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames_to_column</a></span><span class="op">(</span>var <span class="op">=</span> <span class="st">"gene"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join</a></span><span class="op">(</span>y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html">unique</a></span><span class="op">(</span><span class="va">annotations</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"gene_name"</span>, <span class="st">"description"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>,</span>
<span>               by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"gene"</span> <span class="op">=</span> <span class="st">"gene_name"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html">cbind</a></span><span class="op">(</span>cluster_id <span class="op">=</span> <span class="va">cluster</span>, <span class="va">.</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now that we have this function created we can use it as an argument to the appropriate <code>map</code> function. We want the output of the <code>map</code> family of functions to be a <strong>dataframe with each cluster output bound together by rows</strong>. （<code>map</code>函数输出的为一个list，通过<code>list_rbind</code>函数按照行组合列表中的每一个对象，并输出为数据框）</p>
<p>Now, let’s try this function to find the conserved markers for the clusters that were identified as <strong>CD4+ T cells (4, 0, 6, 2)</strong> from our use of <a href="../../single_cell/scRNA-seq_online/08_SC_clustering_quality_control.html#sec-cd4_t_cells">known marker genes</a>.</p>
<div class="cell" data-hash="09_merged_SC_marker_identification_cache/html/unnamed-chunk-23_42ebfc9e833079dec6b92acd38d42d77">
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html">FeaturePlot</a></span><span class="op">(</span><span class="va">seurat_clustered_qc</span>, </span>
<span>            reduction <span class="op">=</span> <span class="st">"umap"</span>, </span>
<span>            features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CD3D"</span>, <span class="st">"IL7R"</span>, <span class="st">"CCR7"</span><span class="op">)</span>, </span>
<span>            order <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            min.cutoff <span class="op">=</span> <span class="st">'q10'</span>, </span>
<span>            label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="09_merged_SC_marker_identification_files/figure-html/unnamed-chunk-23-1.png" class="img-fluid" width="960"></p>
</div>
</div>
<p>Let’s see what genes we identify and of there are overlaps or obvious differences that can help us tease this apart a bit more.</p>
<div class="cell" data-hash="09_merged_SC_marker_identification_cache/html/unnamed-chunk-24_7ac0d6c9cd1012a18d481eba50b12d55">
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://purrr.tidyverse.org/">purrr</a></span><span class="op">)</span></span>
<span><span class="va">conserved_markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">0</span>, <span class="fl">6</span>, <span class="fl">2</span><span class="op">)</span>, <span class="va">get_conserved</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/list_c.html">list_rbind</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Finding markers for all clusters
</div>
</div>
<div class="callout-body-container callout-body">
<p>For your data, you may want to run this function on all clusters, in which case you could input <code>0:20</code> instead of <code>c(4,0,6,2)</code>. Also, it is possible that when you run this function on all clusters, in <strong>some cases you will have clusters that do not have enough cells for a particular group</strong> - and your function will fail. For these clusters you will need to use <code><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html">FindAllMarkers()</a></code>.</p>
</div>
</div>
<section id="sec-top_conserved_markers" class="level3"><h3 class="anchored" data-anchor-id="sec-top_conserved_markers">获取top marker基因</h3>
<p>We would like to use these gene lists to see of we can <strong>identify which celltypes these clusters identify with.</strong> Let’s take a look at the top genes for each of the clusters and see if that gives us any hints. We can view the top 10 markers by <strong>average fold change</strong> across the two groups, for each cluster for a quick perusal:</p>
<ol type="1">
<li>首先通过<code>dplyr</code>包的<code>mutate</code>函数计算新的变量“avg_fc”，计算依据为：avg_fc = (ctrl_avg_log2FC + stim_avg_log2FC) /2</li>
<li>然后通过<code>dplyr</code>包的<code>group_by</code>函数以“cluster_id”列为依据进行分组计算</li>
<li>最后，通过<code>dplyr</code>包的<code>slice_max</code>函数取“avg_fc”（<code>order_by = avg_fc</code>）最大的前10行数据（<code>n = 10</code>）。由于<code>group_by</code>定义了分组计算，所以会输出每个cluster的前10个marker基因</li>
</ol>
<div class="cell" data-hash="09_merged_SC_marker_identification_cache/html/unnamed-chunk-25_894e43fe09e2adef3139aa16e2a4d97e">
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 获取每个cluster的前10个marker基因</span></span>
<span><span class="va">top_conserved_markers</span> <span class="op">&lt;-</span> <span class="va">conserved_markers</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>avg_fc <span class="op">=</span> <span class="op">(</span><span class="va">ctrl_avg_log2FC</span> <span class="op">+</span> <span class="va">stim_avg_log2FC</span><span class="op">)</span> <span class="op">/</span><span class="fl">2</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">cluster_id</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html">slice_max</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">10</span>, order_by <span class="op">=</span> <span class="va">avg_fc</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
待解决的问题
</div>
</div>
<div class="callout-body-container callout-body">
<p>这里基于Seurat V5的运行结果和<a href="https://hbctraining.github.io/scRNA-seq_online/lessons/09_merged_SC_marker_identification.html">原教程</a>的top markers结果不一致。原教程的top_conserved_markers如下：</p>
<p><img src="images/tcell_marker_table_SCTv2.png" class="img-fluid"></p>
<p>可能的原因：</p>
<blockquote class="blockquote">
<p>“In addition, in Seurat v5 we implement a&nbsp;<strong>pseudocount</strong>&nbsp;(when calculating log-FC) at the <strong>group level</strong> instead of the cell level. As a result, users will observe <strong>higher logFC estimates</strong> in v5 - but should note that these estimates may be more unstable - particularly for genes that are very lowly expressed in one of the two groups” ——<a href="../../single_cell/seurat/seurat_intro.html#sec-changes_in_seurat_v5">Changes in Seurat v5</a></p>
</blockquote>
<p>这一小节的后续内容暂时以原教程的marker基因结果为准。</p>
</div>
</div>
</section><section id="根据top-marker基因重新评估细胞群的注释结果" class="level3"><h3 class="anchored" data-anchor-id="根据top-marker基因重新评估细胞群的注释结果">根据top marker基因重新评估细胞群的注释结果</h3>
<p>When we look at the entire list, we see <strong>clusters 0</strong> <strong>and 6</strong> have some overlapping genes, like <strong>CCR7</strong> and <strong>SELL</strong> which correspond to <strong>markers of memory T cells</strong>.</p>
<p>It is possible that these two clusters are more similar to one another and could be merged together as <strong>naive T cells</strong>. On the other hand, with <strong>cluster 2</strong> we observe <strong>CREM</strong> as one of our top genes; a marker gene of activation. This suggests that perhaps cluster 2 represents activated T cells.</p>
<table class="table">
<thead><tr class="header">
<th style="text-align: center;">Cell State</th>
<th style="text-align: center;">Marker</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">Naive T cells</td>
<td style="text-align: center;">CCR7, SELL</td>
</tr>
<tr class="even">
<td style="text-align: center;">Activated T cells</td>
<td style="text-align: center;">CREM, CD69</td>
</tr>
</tbody>
</table>
<p>For <strong>cluster 4</strong>, we see a lot of heat shock and DNA damage genes appear in the top gene list. Based on these markers, it is likely that these are <strong>stressed or dying cells</strong>. However, if we explore the quality metrics for these cells in more detail (i.e.&nbsp;mitoRatio and nUMI overlayed on the cluster) we don’t really support for this argument：</p>
<div class="cell" data-hash="09_merged_SC_marker_identification_cache/html/unnamed-chunk-26_c808d63bf833df841f2410a8f199b999">
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Visualize the distribution of mitochondrial gene expression detected per cell</span></span>
<span><span class="co"># cluster 4中每个细胞检测到的线粒体基因表达分布情况</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span><span class="va">seurat_clustered_qc</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html">subset</a></span><span class="op">(</span>idents <span class="op">=</span> <span class="fl">4</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="va">.</span><span class="op">@</span><span class="va">meta.data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">sample</span>, x <span class="op">=</span> <span class="va">mitoRatio</span>, fill <span class="op">=</span> <span class="va">sample</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="09_merged_SC_marker_identification_files/figure-html/unnamed-chunk-26-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<blockquote class="blockquote">
<p>There is a breadth of research supporting the association of heat shock proteins with reactive T cells in the induction of anti‐inflammatory cytokines in chronic inflammation. This is a cluster for which we would need a deeper understanding of immune cells to really tease apart the results and make a final conclusion.</p>
</blockquote>
<p>To get a better idea of cell type identity for cluster 4 we can explore the expression of different identified markers by cluster using the <code><a href="https://satijalab.org/seurat/reference/FeaturePlot.html">FeaturePlot()</a></code> function.</p>
<div class="cell" data-hash="09_merged_SC_marker_identification_cache/html/unnamed-chunk-27_bfbf1e046704ed1cced17fe08a3e5728">
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Plot interesting marker gene expression for cluster 4</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html">FeaturePlot</a></span><span class="op">(</span><span class="va">seurat_clustered_qc</span>, </span>
<span>            features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"HSPH1"</span>, <span class="st">"HSPE1"</span>, <span class="st">"DNAJB1"</span><span class="op">)</span>,</span>
<span>            order <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            min.cutoff <span class="op">=</span> <span class="st">'q10'</span>, </span>
<span>            label <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            repel <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="09_merged_SC_marker_identification_files/figure-html/unnamed-chunk-27-1.png" class="img-fluid" width="960"></p>
</div>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 提取cluster 4，并单独查看interesting marker gene在其中的表达情况</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html">subset</a></span><span class="op">(</span><span class="va">seurat_clustered_qc</span>, idents <span class="op">=</span> <span class="fl">4</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html">FeaturePlot</a></span><span class="op">(</span></span>
<span>            features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"HSPH1"</span>, <span class="st">"HSPE1"</span>, <span class="st">"DNAJB1"</span><span class="op">)</span>,</span>
<span>            order <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            min.cutoff <span class="op">=</span> <span class="st">'q10'</span>, </span>
<span>            label <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            repel <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="09_merged_SC_marker_identification_files/figure-html/unnamed-chunk-27-2.png" class="img-fluid" width="960"></p>
</div>
</div>
<blockquote class="blockquote">
<p>We see that <strong>only a subset of cluster 4 are highly expressing these genes</strong>.</p>
</blockquote>
<p>We can also explore the range in expression of specific markers by using <strong>violin plots</strong>:</p>
<div class="cell" data-hash="09_merged_SC_marker_identification_cache/html/unnamed-chunk-28_df258cf8bab249137721441b151ceb49">
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Vln plot - cluster 4</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html">VlnPlot</a></span><span class="op">(</span><span class="va">seurat_clustered_qc</span>, </span>
<span>        features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"HSPH1"</span>, <span class="st">"HSPE1"</span>, <span class="st">"DNAJB1"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="09_merged_SC_marker_identification_files/figure-html/unnamed-chunk-28-1.png" class="img-fluid" width="960"></p>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-10-contents" aria-controls="callout-10" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Violin plots
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-10" class="callout-10-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Violin plots are similar to box plots, except that they also show the <strong>probability density</strong> of the data at different values, usually smoothed by a <strong>kernel density estimator</strong>. A violin plot is more informative than a plain box plot. While a box plot only shows summary statistics such as mean/median and interquartile ranges, the <strong>violin plot shows the full distribution of the data</strong>. The difference is particularly useful when the data distribution is multimodal (more than one peak). In this case a violin plot shows the presence of different peaks, their position and relative amplitude.</p>
</div>
</div>
</div>
</section></section></section><section id="findmarkers-marker-identification-between-specific-clusters" class="level1" data-number="4"><h1 data-number="4">
<span class="header-section-number">4</span> <code>FindMarkers</code>-<strong>Marker identification between specific clusters</strong>
</h1>
<p>Sometimes the list of markers returned don’t sufficiently separate some of the clusters. For instance, we had previously identified clusters 0, 4, 6 and 2 as CD4+ T cells, but when looking at marker gene lists we identfied markers to help us further subset cells. We were lucky and the signal observed from <code><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html">FindAllMarkers()</a></code> helped us differentiate between naive and activated cells.</p>
<p>Another option to identify biologically meaningful differences would be to use the <strong><code><a href="https://satijalab.org/seurat/reference/FindMarkers.html">FindMarkers()</a></code> function to determine the genes that are differentially expressed between two specific clusters</strong>.</p>
<p><img src="images/marker_ident_function3.png" class="img-fluid" width="379"></p>
<p>We can try all combinations of comparisons, but we’ll start with <strong>cluster 2</strong> versus all other CD4+ T cell clusters:</p>
<div class="cell" data-hash="09_merged_SC_marker_identification_cache/html/unnamed-chunk-29_612ceab8995c54c164fafe381b93f64a">
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Determine differentiating markers for CD4+ T cell</span></span>
<span><span class="va">cd4_tcells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindMarkers.html">FindMarkers</a></span><span class="op">(</span><span class="va">seurat_clustered_qc</span>,</span>
<span>                          ident.1 <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                          ident.2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">4</span>, <span class="fl">6</span><span class="op">)</span><span class="op">)</span>  </span>
<span></span>
<span><span class="co"># Add gene symbols to the DE table</span></span>
<span><span class="va">cd4_tcells</span> <span class="op">&lt;-</span> <span class="va">cd4_tcells</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames_to_column</a></span><span class="op">(</span>var <span class="op">=</span> <span class="st">"gene"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join</a></span><span class="op">(</span>y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html">unique</a></span><span class="op">(</span><span class="va">annotations</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"gene_name"</span>, <span class="st">"description"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>,</span>
<span>            by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"gene"</span> <span class="op">=</span> <span class="st">"gene_name"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html">colnames</a></span><span class="op">(</span><span class="va">cd4_tcells</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "gene"        "p_val"       "avg_log2FC"  "pct.1"       "pct.2"      
[6] "p_val_adj"   "description"</code></pre>
</div>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Reorder columns and sort by "p_val_adj"</span></span>
<span><span class="va">cd4_tcells</span> <span class="op">&lt;-</span> <span class="va">cd4_tcells</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">2</span>, <span class="fl">6</span><span class="op">:</span><span class="fl">7</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">cd4_tcells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">cd4_tcells</span>, <span class="va">p_val_adj</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cd4_tcells</span><span class="op">$</span><span class="va">gene</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "GAPDH"   "SRGN"    "CYBA"    "ALOX5AP" "CCR7"    "FTH1"    "TMSB4X" 
 [8] "CREM"    "SELL"    "ANXA1"  </code></pre>
</div>
</div>
<p><img src="images/截屏2023-12-22 11.37.02.png" class="img-fluid"></p>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Warning
</div>
</div>
<div class="callout-body-container callout-body">
<p>这个表格和<a href="https://hbctraining.github.io/scRNA-seq_online/lessons/09_merged_SC_marker_identification.html">原教程</a>的仍然有差异。</p>
</div>
</div>
<blockquote class="blockquote">
<p>Of these top genes the <strong>CREM gene</strong> stands out as a marker of activation with a positive fold change. We also see markers of naive or memory cells include the <strong>SELL</strong> and <strong>CCR7</strong> genes with negative fold changes, which is in line with previous results.</p>
<table class="table">
<thead><tr class="header">
<th style="text-align: center;">Cell State</th>
<th style="text-align: center;">Marker</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">Naive T cells</td>
<td style="text-align: center;">CCR7, SELL</td>
</tr>
<tr class="even">
<td style="text-align: center;">Activated T cells</td>
<td style="text-align: center;">CREM, CD69</td>
</tr>
</tbody>
</table>
</blockquote>
<p>进一步通过气泡图来可视化上述基因的表达.</p>
<div class="cell" data-hash="09_merged_SC_marker_identification_cache/html/unnamed-chunk-30_5d5cec6477b4c040ff00c359ec928e18">
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DotPlot.html">DotPlot</a></span><span class="op">(</span><span class="va">seurat_clustered_qc</span>, </span>
<span>        features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CCR7"</span>, <span class="st">"SELL"</span>, <span class="st">"CREM"</span>, <span class="st">"CD69"</span><span class="op">)</span>, </span>
<span>        idents <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">2</span>, <span class="fl">4</span>, <span class="fl">6</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html">RotatedAxis</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="09_merged_SC_marker_identification_files/figure-html/unnamed-chunk-30-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<blockquote class="blockquote">
<p>可以看到，和<code>FindMarkers</code>找到的top genes的fold change一致，cluster 2中的CREM基因相较于cluster 0, 4, 6高表达，而cluster 2中的CCR7和SELL相对低表达。</p>
<p>Based on these plots it seems as though <strong>clusters 2</strong> are <strong>Activated T cells, cluster 0 and 6 are</strong> Naive or memory CD4+ T cells. However, for <strong>clusters 4</strong> it is hard to tell.</p>
</blockquote>
<hr></section><section id="注释细胞cluster" class="level1" data-number="5"><h1 data-number="5">
<span class="header-section-number">5</span> 注释细胞cluster</h1>
<p>Now taking all of this information, we can surmise the cell types of the different clusters and plot the cells with cell type labels.</p>
<div id="tbl-细胞注释结果" class="anchored">
<table class="table">
<caption>Table&nbsp;1: 细胞注释结果</caption>
<thead><tr class="header">
<th style="text-align: center;">Cluster ID</th>
<th style="text-align: center;">Cell Type</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">0</td>
<td style="text-align: center;">Naive or memory CD4+ T cells</td>
</tr>
<tr class="even">
<td style="text-align: center;">1</td>
<td style="text-align: center;">CD14+ monocytes</td>
</tr>
<tr class="odd">
<td style="text-align: center;">2</td>
<td style="text-align: center;">Activated T cells</td>
</tr>
<tr class="even">
<td style="text-align: center;">3</td>
<td style="text-align: center;">CD14+ monocytes</td>
</tr>
<tr class="odd">
<td style="text-align: center;">4</td>
<td style="text-align: center;">Stressed cells / Unknown</td>
</tr>
<tr class="even">
<td style="text-align: center;">5</td>
<td style="text-align: center;">CD8+ T cells</td>
</tr>
<tr class="odd">
<td style="text-align: center;">6</td>
<td style="text-align: center;">Naive or memory CD4+ T cells</td>
</tr>
<tr class="even">
<td style="text-align: center;">7</td>
<td style="text-align: center;">B cells</td>
</tr>
<tr class="odd">
<td style="text-align: center;">8</td>
<td style="text-align: center;">NK cells</td>
</tr>
<tr class="even">
<td style="text-align: center;">9</td>
<td style="text-align: center;">CD8+ T cells</td>
</tr>
<tr class="odd">
<td style="text-align: center;">10</td>
<td style="text-align: center;">FCGR3A+ monocytes</td>
</tr>
<tr class="even">
<td style="text-align: center;">11</td>
<td style="text-align: center;">B cells</td>
</tr>
<tr class="odd">
<td style="text-align: center;">12</td>
<td style="text-align: center;">NK cells</td>
</tr>
<tr class="even">
<td style="text-align: center;">13</td>
<td style="text-align: center;">B cells</td>
</tr>
<tr class="odd">
<td style="text-align: center;">14</td>
<td style="text-align: center;">Conventional dendritic cells</td>
</tr>
<tr class="even">
<td style="text-align: center;">15</td>
<td style="text-align: center;">Megakaryocytes</td>
</tr>
<tr class="odd">
<td style="text-align: center;">16</td>
<td style="text-align: center;">Plasmacytoid dendritic cells</td>
</tr>
</tbody>
</table>
</div>
<p>We can then reassign the identity of the clusters to these cell types:</p>
<div class="cell" data-hash="09_merged_SC_marker_identification_cache/html/unnamed-chunk-31_e40dd283514074996f800384d9732eec">
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Rename all identities</span></span>
<span><span class="va">seurat_clustered_qc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/Idents.html">RenameIdents</a></span><span class="op">(</span><span class="va">seurat_clustered_qc</span>, </span>
<span>                                    <span class="st">"0"</span> <span class="op">=</span> <span class="st">"Naive or memory CD4+ T cells"</span>,</span>
<span>                                    <span class="st">"1"</span> <span class="op">=</span> <span class="st">"CD14+ monocytes"</span>,</span>
<span>                                    <span class="st">"2"</span> <span class="op">=</span> <span class="st">"Activated T cells"</span>,</span>
<span>                                    <span class="st">"3"</span> <span class="op">=</span> <span class="st">"CD14+ monocytes"</span>,</span>
<span>                                    <span class="st">"4"</span> <span class="op">=</span> <span class="st">"Stressed cells / Unknown"</span>,</span>
<span>                                    <span class="st">"5"</span> <span class="op">=</span> <span class="st">"CD8+ T cells"</span>,</span>
<span>                                    <span class="st">"6"</span> <span class="op">=</span> <span class="st">"Naive or memory CD4+ T cells"</span>,</span>
<span>                                    <span class="st">"7"</span> <span class="op">=</span> <span class="st">"B cells"</span>,</span>
<span>                                    <span class="st">"8"</span> <span class="op">=</span> <span class="st">"NK cells"</span>,</span>
<span>                                    <span class="st">"9"</span> <span class="op">=</span> <span class="st">"CD8+ T cells"</span>,</span>
<span>                                    <span class="st">"10"</span> <span class="op">=</span> <span class="st">"FCGR3A+ monocytes"</span>,</span>
<span>                                    <span class="st">"11"</span> <span class="op">=</span> <span class="st">"B cells"</span>,</span>
<span>                                    <span class="st">"12"</span> <span class="op">=</span> <span class="st">"NK cells"</span>,</span>
<span>                                    <span class="st">"13"</span> <span class="op">=</span> <span class="st">"B cells"</span>,</span>
<span>                                    <span class="st">"14"</span> <span class="op">=</span> <span class="st">"Conventional dendritic cells"</span>,</span>
<span>                                    <span class="st">"15"</span> <span class="op">=</span> <span class="st">"Megakaryocytes"</span>,</span>
<span>                                    <span class="st">"16"</span> <span class="op">=</span> <span class="st">"Plasmacytoid dendritic cells"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-hash="09_merged_SC_marker_identification_cache/html/unnamed-chunk-32_d39954f0e09fc9e50ee1642b2052c121">
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Plot the UMAP</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">seurat_clustered_qc</span>, </span>
<span>        reduction <span class="op">=</span> <span class="st">"umap"</span>, </span>
<span>        label <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="09_merged_SC_marker_identification_files/figure-html/unnamed-chunk-32-1.png" class="img-fluid" width="960"></p>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>If we wanted to remove the potentially stressed cells, we could use the <code><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html">subset()</a></code> function:</p>
<div class="cell" data-hash="09_merged_SC_marker_identification_cache/html/unnamed-chunk-33_1422392eba466ec7f17d9a0310430bb2">
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Remove the stressed or dying cells</span></span>
<span><span class="va">seurat_subset_labeled</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html">subset</a></span><span class="op">(</span><span class="va">seurat_clustered_qc</span>,</span>
<span>                               idents <span class="op">=</span> <span class="st">"Stressed cells / Unknown"</span>, <span class="co"># 取子集的标准</span></span>
<span>                               invert <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="co"># 反选</span></span>
<span></span>
<span><span class="co"># Re-visualize the clusters</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">seurat_subset_labeled</span>, </span>
<span>        reduction <span class="op">=</span> <span class="st">"umap"</span>, </span>
<span>        label <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="09_merged_SC_marker_identification_files/figure-html/unnamed-chunk-33-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</div>
</div>
<p>Now we would want to save our final labelled Seurat object:</p>
<div class="cell" data-hash="09_merged_SC_marker_identification_cache/html/unnamed-chunk-34_5ac16e083a80fed3a46c13139fb6f057">
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Save final R object</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">seurat_clustered_qc</span>, file <span class="op">=</span> <span class="st">"output/scRNA-seq_online/seurat_labelled.rds"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>最后，我们把sessionInfo也导出来：</p>
<div class="cell" data-hash="09_merged_SC_marker_identification_cache/html/unnamed-chunk-35_04f7950739480884abf82f9b94e8422f">
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Create and save a text file with sessionInfo</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sink.html">sink</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"output/scRNA-seq_online/sessionInfo_scrnaseq.txt"</span>, </span>
<span>     append <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>     split <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sink.html">sink</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-13-contents" aria-controls="callout-13" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
关于<code>sink</code>函数
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-13" class="callout-13-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p><code>sink</code>函数能够将R脚本运行的结果输出到文本文件中，输出的内容为两个<code><a href="https://rdrr.io/r/base/sink.html">sink()</a></code>命令之间的所有内容。</p>
<ul>
<li><p><code>file</code>：输出目录</p></li>
<li><p><code>append</code>：取<code>TRUE</code>表示若输出目录下有与结果文件同名的文件，则计算结果将追加到原文件内容的后面；取<code>FALSE</code>（默认）表示将本次的计算结果覆盖原文件的内容</p></li>
<li><p><code>split</code>：取<code>TRUE</code>表示在计算结果输出到指定文件中的同时，还输出到控制台上；取<code>FALSE</code>（默认）表示计算结果仅输出到指定文件中。</p></li>
</ul>
</div>
</div>
</div>
<hr></section><section id="可选的后续分析" class="level1" data-number="6"><h1 data-number="6">
<span class="header-section-number">6</span> 可选的后续分析:</h1>
<div class="quarto-figure quarto-figure-left">
<figure class="figure"><p><a href="https://mp.weixin.qq.com/s?__biz=Mzg2MTExNTkwNA==&amp;mid=2247532151&amp;idx=1&amp;sn=55ff4362b5a572d53072c1fe85c593d6&amp;chksm=ce1e3d18f969b40e71503729f38a8e01cd80e313446da1571d7d42c61e094d94b3519cdb85be&amp;mpshare=1&amp;scene=1&amp;srcid=0910pFfJBQOE8XVDU2opxc6B&amp;sharer_shareinfo=98ce7b038453fa671b5e609905ff82f0&amp;sharer_shareinfo_first=98ce7b038453fa671b5e609905ff82f0#rd"><img src="images/640.png" class="img-fluid figure-img"></a></p>
</figure>
</div>
<div class="quarto-figure quarto-figure-left">
<figure class="figure"><p><a href="https://mp.weixin.qq.com/s?__biz=Mzg2MTExNTkwNA==&amp;mid=2247532151&amp;idx=1&amp;sn=55ff4362b5a572d53072c1fe85c593d6&amp;chksm=ce1e3d18f969b40e71503729f38a8e01cd80e313446da1571d7d42c61e094d94b3519cdb85be&amp;mpshare=1&amp;scene=1&amp;srcid=0910pFfJBQOE8XVDU2opxc6B&amp;sharer_shareinfo=98ce7b038453fa671b5e609905ff82f0&amp;sharer_shareinfo_first=98ce7b038453fa671b5e609905ff82f0#rd"><img src="images/640-2.png" class="img-fluid figure-img"></a></p>
</figure>
</div>
<section id="提取感兴趣细胞亚群进一步细分explore-a-subset-of-the-cell-types-to-discover-subclusters-of-cells" class="level2" data-number="6.1"><h2 data-number="6.1" class="anchored" data-anchor-id="提取感兴趣细胞亚群进一步细分explore-a-subset-of-the-cell-types-to-discover-subclusters-of-cells">
<span class="header-section-number">6.1</span> 提取感兴趣细胞亚群进一步细分（Explore a subset of the cell types to discover subclusters of cells）</h2>
<p>鉴定完细胞类型后，接下来就是对鉴定的目标细胞群进行深入的研究。一般是提取目标亚群，然后在进一步细分。细胞再分群的意义在于进一步分析细胞的组成成分。见<a href="../../single_cell/sc_supplementary/integrated_analysis_multiple_single_cell_datasets_2.html">多个单细胞数据集整合分析（下）</a>。</p>
</section><section id="单细胞差异分析perform-differential-expression-analysis-between-conditions" class="level2" data-number="6.2"><h2 data-number="6.2" class="anchored" data-anchor-id="单细胞差异分析perform-differential-expression-analysis-between-conditions">
<span class="header-section-number">6.2</span> <strong>单细胞差异分析（</strong>Perform <strong>differential expression analysis</strong> <strong>between conditions）</strong>
</h2>
<p>在实际研究中，往往需要研究两类细胞之间的差异，或者是某类细胞在不同处理样本间的差异基因情况，因此需要将待比较的细胞群单独拿出来进行分析。后面，我们在<a href="../../single_cell/seurat/de_vignette.html">此前的章节</a>中对如何<strong>寻找不同样本类型/条件间同一细胞类型内的差异基因</strong>进行了探索。</p>
<p>除了分析细胞内基因表达的差异，也可以采用最常见的GO和KEGG富集分析，还可以进行GSEA、GSVA、SCENIC等分析，从功能、通路、基因调控网络等方面进行探究。</p>
</section><section id="免疫类群分析" class="level2" data-number="6.3"><h2 data-number="6.3" class="anchored" data-anchor-id="免疫类群分析">
<span class="header-section-number">6.3</span> 免疫类群分析</h2>
<p>很多疾病的发生、发展都和免疫细胞有关，一般病灶组织内或者附近的免疫细胞较正常的组织会更丰富。为了将疾病中免疫反应描绘出来，可以采用以下分析策略：①对免疫细胞再分群，得到更多功能更加精细化的免疫细胞亚群；②免疫细胞类群在不同分组中的分布变化；③免疫细胞类群拟时序分析，探究不同疾病程度、不同发育阶段或者不同生理状态下的免疫反应机制。</p>
</section><section id="关键细胞的分化轨迹" class="level2" data-number="6.4"><h2 data-number="6.4" class="anchored" data-anchor-id="关键细胞的分化轨迹">
<span class="header-section-number">6.4</span> 关键细胞的分化轨迹</h2>
<p>最开始拿到的单细胞转录组测序数据，并未直接告诉我们每个细胞处在什么状态。因此，需要借助一些分析方法来实现轨迹上的排序，比如Monocle2拟时序分析/轨迹推断，ScVelo RNA速度分析等，推断潜在的细胞分化方向性，挖掘一些稀少的中间状态细胞，解析细胞分化过程中起调控作用的关键基因；也可以比较实验组/对照组中分化的差异。</p>
</section><section id="探索细胞-细胞互作" class="level2" data-number="6.5"><h2 data-number="6.5" class="anchored" data-anchor-id="探索细胞-细胞互作">
<span class="header-section-number">6.5</span> 探索细胞-细胞互作</h2>
<p>每个细胞都能分泌细胞因子或者激素，这些细胞因子能够被周围细胞上的受体接收并用于调节相应的生理活动，细胞与细胞之间有相互联系，致病细胞可能是由于大类中某一种亚类细胞分化或者演化过来的；那么这些演化过程可能是由于其他细胞的细胞因子导致的。</p>
<p>在锁定目标细胞类群之后，通过CellphoneDB细胞通讯分析、受体-配体分析等，找到与目标细胞相互作用的其他细胞类型，找到“直接”和“间接”的细胞调控网络。当然，对于已经筛选出目标基因了，也可以通过PPI分析找到基因/蛋白间的互作关系。</p>
</section><section id="挖掘关键调控转录因子" class="level2" data-number="6.6"><h2 data-number="6.6" class="anchored" data-anchor-id="挖掘关键调控转录因子">
<span class="header-section-number">6.6</span> 挖掘关键调控转录因子</h2>
<p>细胞异质性以及这种异质性是如何发展和维持的，在很大程度上是由潜在的基因调控网络决定的，特定转录因子（transcription factor，TF）集合的协同表达驱动各自靶标基因的表达，从而建立特定的基因表达谱；SCENIC是用来研究和破译基因调控的工具，能从单细胞转录组数据中推断TF、基因调控网络和细胞类型。其基本原理是基于共表达和DNA调控保守序列（motif）分析推断基因调控网络，然后在每个细胞中分析网络活性以鉴定细胞状态。</p>
</section><section id="湿实验验证探索潜在机制experimentally-validate-intriguing-markers-for-our-identified-cell-types" class="level2" data-number="6.7"><h2 data-number="6.7" class="anchored" data-anchor-id="湿实验验证探索潜在机制experimentally-validate-intriguing-markers-for-our-identified-cell-types">
<span class="header-section-number">6.7</span> 湿实验验证，探索潜在机制（Experimentally validate intriguing markers for our identified cell types）</h2>
<p>除了以上单细胞测序数据相关分析以外，湿实验验证已逐渐成为单细胞分析的有力补充。比如用RT-qPCR/FISH等在RNA层面进行验证，用WB/免疫荧光/免疫组化/流式分析等在蛋白层面进行验证，还可以利用组织芯片或者TCGA数据库信息，进行临床水平的验证；也可以用细胞谱系示踪技术标记细胞 , 对关键细胞类群及其后代所有细胞的增殖、分化和迁移等活动进行追踪观察等；甚至可以利用流式分选出感兴趣的细胞群，然后对目的细胞群进行基因操作（上调/下调/敲除等），进行更深入的机制研究。总之，根据具体的实验目的，进行更完善的实验设计。</p>
<hr>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-14-contents" aria-controls="callout-14" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Session Info
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-14" class="callout-14-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-hash="09_merged_SC_marker_identification_cache/html/unnamed-chunk-36_2ce80fa5655d3e01a892cb77ac662916">
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.3.2 (2023-10-31)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Sonoma 14.3

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: Asia/Shanghai
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] ggplot2_3.4.4      purrr_1.0.2        tibble_3.2.1       dplyr_1.1.4       
[5] Seurat_5.0.1       SeuratObject_5.0.1 sp_2.1-2          

loaded via a namespace (and not attached):
  [1] mathjaxr_1.6-0         RColorBrewer_1.1-3     rstudioapi_0.15.0     
  [4] jsonlite_1.8.8         magrittr_2.0.3         ggbeeswarm_0.7.2      
  [7] TH.data_1.1-2          spatstat.utils_3.0-4   farver_2.1.1          
 [10] rmarkdown_2.25         vctrs_0.6.5            multtest_2.58.0       
 [13] ROCR_1.0-11            spatstat.explore_3.2-5 htmltools_0.5.7       
 [16] plotrix_3.8-4          sctransform_0.4.1      parallelly_1.36.0     
 [19] KernSmooth_2.23-22     htmlwidgets_1.6.4      ica_1.0-3             
 [22] sandwich_3.1-0         plyr_1.8.9             plotly_4.10.4         
 [25] zoo_1.8-12             igraph_1.6.0           mime_0.12             
 [28] lifecycle_1.0.4        pkgconfig_2.0.3        Matrix_1.6-5          
 [31] R6_2.5.1               fastmap_1.1.1          rbibutils_2.2.16      
 [34] fitdistrplus_1.1-11    future_1.33.1          shiny_1.8.0           
 [37] numDeriv_2016.8-1.1    digest_0.6.34          colorspace_2.1-0      
 [40] patchwork_1.2.0        tensor_1.5             RSpectra_0.16-1       
 [43] irlba_2.3.5.1          labeling_0.4.3         progressr_0.14.0      
 [46] fansi_1.0.6            spatstat.sparse_3.0-3  httr_1.4.7            
 [49] TFisher_0.2.0          polyclip_1.10-6        abind_1.4-5           
 [52] compiler_4.3.2         withr_3.0.0            mutoss_0.1-13         
 [55] fastDummies_1.7.3      MASS_7.3-60.0.1        tools_4.3.2           
 [58] vipor_0.4.7            lmtest_0.9-40          beeswarm_0.4.0        
 [61] metap_1.9              httpuv_1.6.13          future.apply_1.11.1   
 [64] qqconf_1.3.2           goftest_1.2-3          glue_1.7.0            
 [67] nlme_3.1-164           promises_1.2.1         grid_4.3.2            
 [70] Rtsne_0.17             cluster_2.1.6          reshape2_1.4.4        
 [73] generics_0.1.3         gtable_0.3.4           spatstat.data_3.0-4   
 [76] tidyr_1.3.0            sn_2.1.1               data.table_1.14.10    
 [79] utf8_1.2.4             BiocGenerics_0.48.1    spatstat.geom_3.2-7   
 [82] RcppAnnoy_0.0.21       ggrepel_0.9.5          RANN_2.6.1            
 [85] pillar_1.9.0           stringr_1.5.1          spam_2.10-0           
 [88] RcppHNSW_0.5.0         limma_3.58.1           later_1.3.2           
 [91] splines_4.3.2          lattice_0.22-5         survival_3.5-7        
 [94] deldir_2.0-2           tidyselect_1.2.0       miniUI_0.1.1.1        
 [97] pbapply_1.7-2          knitr_1.45             gridExtra_2.3         
[100] scattermore_1.2        stats4_4.3.2           xfun_0.41             
[103] Biobase_2.62.0         statmod_1.5.0          matrixStats_1.2.0     
[106] stringi_1.8.3          lazyeval_0.2.2         yaml_2.3.8            
[109] evaluate_0.23          codetools_0.2-19       cli_3.6.2             
[112] uwot_0.1.16            xtable_1.8-4           reticulate_1.34.0     
[115] Rdpack_2.6             munsell_0.5.0          Rcpp_1.0.12           
[118] globals_0.16.2         spatstat.random_3.2-2  png_0.1-8             
[121] ggrastr_1.0.2          parallel_4.3.2         ellipsis_0.3.2        
[124] presto_1.0.0           dotCall64_1.1-1        listenv_0.9.0         
[127] viridisLite_0.4.2      mvtnorm_1.2-4          scales_1.3.0          
[130] ggridges_0.5.5         leiden_0.4.3.1         rlang_1.1.3           
[133] multcomp_1.4-25        mnormt_2.1.1           cowplot_1.1.2         </code></pre>
</div>
</div>
</div>
</div>
</div>


<!-- -->

</section></section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  let localAlternateSentinel = 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var filterRegex = new RegExp("https:\/\/djhcod\.github\.io\/r-notes\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
          // target, if specified
          link.setAttribute("target", "_blank");
      }
    }
});
</script><script src="https://giscus.app/client.js" data-repo="djhcod/r-notes" data-repo-id="R_kgDOKw7GMg" data-category="General" data-category-id="DIC_kwDOKw7GMs4Cbz4B" data-mapping="title" data-reactions-enabled="1" data-emit-metadata="0" data-input-position="top" data-theme="light" data-lang="en" crossorigin="anonymous" async="">
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="../../single_cell/scRNA-seq_online/08_SC_clustering_quality_control.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">细胞分群质量评估</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../../single_cell/sc_supplementary/sc_supplementary_intro.html" class="pagination-link">
        <span class="nav-page-text">单细胞补充内容</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb57" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> "寻找marker基因+细胞注释"</span></span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb57-4"><a href="#cb57-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-5"><a href="#cb57-5" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb57-6"><a href="#cb57-6" aria-hidden="true" tabindex="-1"></a><span class="fu">###### Learning Objectives:</span></span>
<span id="cb57-7"><a href="#cb57-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-8"><a href="#cb57-8" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Describe how to determine markers of individual clusters</span>
<span id="cb57-9"><a href="#cb57-9" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Discuss the iterative processes of clustering and marker identification</span>
<span id="cb57-10"><a href="#cb57-10" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb57-11"><a href="#cb57-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-12"><a href="#cb57-12" aria-hidden="true" tabindex="-1"></a>Now that we have identified our desired clusters, we can move on to marker identification, which will allow us to **verify the identity of certain clusters** and help surmise the identity of any unknown clusters.</span>
<span id="cb57-13"><a href="#cb57-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-14"><a href="#cb57-14" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/sc_workflow_2022-01.jpg)</span>{width="545"}</span>
<span id="cb57-15"><a href="#cb57-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-16"><a href="#cb57-16" aria-hidden="true" tabindex="-1"></a>------------------------------------------------------------------------</span>
<span id="cb57-17"><a href="#cb57-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-18"><a href="#cb57-18" aria-hidden="true" tabindex="-1"></a>**Goals:**</span>
<span id="cb57-19"><a href="#cb57-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-20"><a href="#cb57-20" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>To **determine the gene markers** for each of the clusters</span>
<span id="cb57-21"><a href="#cb57-21" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>To **identify cell types** of each cluster using markers</span>
<span id="cb57-22"><a href="#cb57-22" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>To determine whether there's a need to **re-cluster based on cell type markers**, perhaps clusters need to be merged or split</span>
<span id="cb57-23"><a href="#cb57-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-24"><a href="#cb57-24" aria-hidden="true" tabindex="-1"></a>**Challenges:**</span>
<span id="cb57-25"><a href="#cb57-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-26"><a href="#cb57-26" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Over-interpretation of the results</span>
<span id="cb57-27"><a href="#cb57-27" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Combining different types of marker identification</span>
<span id="cb57-28"><a href="#cb57-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-29"><a href="#cb57-29" aria-hidden="true" tabindex="-1"></a>**Recommendations:**</span>
<span id="cb57-30"><a href="#cb57-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-31"><a href="#cb57-31" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Think of the results as hypotheses that need verification. Inflated p-values can lead to over-interpretation of results (essentially each cell is used as a replicate). Top markers are most trustworthy.</span>
<span id="cb57-32"><a href="#cb57-32" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Identify all markers conserved between conditions for each cluster</span>
<span id="cb57-33"><a href="#cb57-33" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Identify markers that are differentially expressed between specific clusters</span>
<span id="cb57-34"><a href="#cb57-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-35"><a href="#cb57-35" aria-hidden="true" tabindex="-1"></a>------------------------------------------------------------------------</span>
<span id="cb57-36"><a href="#cb57-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-37"><a href="#cb57-37" aria-hidden="true" tabindex="-1"></a><span class="fu"># 数据导入</span></span>
<span id="cb57-38"><a href="#cb57-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-39"><a href="#cb57-39" aria-hidden="true" tabindex="-1"></a>载入<span class="co">[</span><span class="ot">此前</span><span class="co">](/single_cell/scRNA-seq_online/08_SC_clustering_quality_control.qmd)</span>完成细胞分群质量评估的数据<span class="in">`seurat_clustered_qc`</span>。</span>
<span id="cb57-40"><a href="#cb57-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-43"><a href="#cb57-43" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-44"><a href="#cb57-44" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb57-45"><a href="#cb57-45" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb57-46"><a href="#cb57-46" aria-hidden="true" tabindex="-1"></a>seurat_clustered_qc <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"output/scRNA-seq_online/seurat_clustered_qc.rds"</span>)</span>
<span id="cb57-47"><a href="#cb57-47" aria-hidden="true" tabindex="-1"></a>seurat_clustered_qc</span>
<span id="cb57-48"><a href="#cb57-48" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(seurat_clustered_qc, <span class="dv">3</span>)</span>
<span id="cb57-49"><a href="#cb57-49" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-50"><a href="#cb57-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-51"><a href="#cb57-51" aria-hidden="true" tabindex="-1"></a>Before we start our marker identification we will explicitly set our default assay, we want to use the **normalized data, but not the integrated data**.</span>
<span id="cb57-52"><a href="#cb57-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-55"><a href="#cb57-55" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-56"><a href="#cb57-56" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb57-57"><a href="#cb57-57" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(seurat_clustered_qc) <span class="ot">&lt;-</span> <span class="st">"RNA"</span></span>
<span id="cb57-58"><a href="#cb57-58" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-59"><a href="#cb57-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-60"><a href="#cb57-60" aria-hidden="true" tabindex="-1"></a>::: callout-important</span>
<span id="cb57-61"><a href="#cb57-61" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">原教程</span><span class="co">](https://hbctraining.github.io/scRNA-seq_online/lessons/09_merged_SC_marker_identification.html)</span>将默认的assay设置为了"RNA"，其解释如下：The default assay should have already been <span class="in">`RNA`</span>, because we set it up in the <span class="co">[</span><span class="ot">previous clustering quality control lesson</span><span class="co">](/single_cell/scRNA-seq_online/08_SC_clustering_quality_control.qmd#sec-explore_known_markers)</span>. But we encourage you to run this line of code above to be absolutely sure in case the active slot was changed somewhere upstream in your analysis. Note that the raw and normalized counts are stored in the <span class="in">`counts`</span> and <span class="in">`data`</span> slots of <span class="in">`RNA`</span> assay. By default, the functions for finding markers will use normalized data. （关于<span class="in">`FindMarkers`</span>为什么要使用"RNA" assay的更多解释，参阅<span class="co">[</span><span class="ot">此链接</span><span class="co">](https://github.com/hbctraining/scRNA-seq_online/issues/58)</span>）</span>
<span id="cb57-62"><a href="#cb57-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-63"><a href="#cb57-63" aria-hidden="true" tabindex="-1"></a>而在<span class="co">[</span><span class="ot">Seurat V5的官方教程</span><span class="co">](/single_cell/seurat/marker_gene_identification.qmd#sec-findmarkers_function)</span>中，对于经过<span class="in">`SCTransform`</span>归一化处理后的单细胞数据，在进行<span class="in">`FindMarkers`</span>差异分析之前，需要先运行<span class="in">`seurat_clustered_qc &lt;- PrepSCTFindMarkers(seurat_clustered_qc)`</span>，来预处理SCT assay。详细解释见<span class="co">[</span><span class="ot">此链接</span><span class="co">](https://www.jianshu.com/p/fb2e43905559)</span>。</span>
<span id="cb57-64"><a href="#cb57-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-65"><a href="#cb57-65" aria-hidden="true" tabindex="-1"></a>如果是基于<span class="in">`NormalizeData`</span>标准化的单细胞数据，和这里一样，需要使用"RNA" assay进行差异分析，如果不是，需要通过<span class="in">`DefaultAssay(seurat_clustered_qc) &lt;- "RNA"`</span>进行设定。</span>
<span id="cb57-66"><a href="#cb57-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-67"><a href="#cb57-67" aria-hidden="true" tabindex="-1"></a>这里为了和原教程保持一致，默认的将assay设置为"RNA"。</span>
<span id="cb57-68"><a href="#cb57-68" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb57-69"><a href="#cb57-69" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-70"><a href="#cb57-70" aria-hidden="true" tabindex="-1"></a>Our clustering analysis resulted in the following clusters:</span>
<span id="cb57-71"><a href="#cb57-71" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-74"><a href="#cb57-74" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-75"><a href="#cb57-75" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(seurat_clustered_qc, <span class="at">label =</span> <span class="cn">TRUE</span>)</span>
<span id="cb57-76"><a href="#cb57-76" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-77"><a href="#cb57-77" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-78"><a href="#cb57-78" aria-hidden="true" tabindex="-1"></a>Remember that we had the following questions from the <span class="co">[</span><span class="ot">clustering analysis</span><span class="co">](/single_cell/scRNA-seq_online/08_SC_clustering_quality_control.qmd#sec-Preliminary_cell_type_identification_results)</span>:</span>
<span id="cb57-79"><a href="#cb57-79" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-80"><a href="#cb57-80" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; 1.  Do the clusters corresponding to the same cell types have biologically meaningful differences? Are there subpopulations of these cell types?</span></span>
<span id="cb57-81"><a href="#cb57-81" aria-hidden="true" tabindex="-1"></a><span class="at">&gt;</span></span>
<span id="cb57-82"><a href="#cb57-82" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; 2.  Can we acquire higher confidence in these cell type identities by identifying other marker genes for these clusters?</span></span>
<span id="cb57-83"><a href="#cb57-83" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-84"><a href="#cb57-84" aria-hidden="true" tabindex="-1"></a>There are a few different types of marker identification that we can explore using Seurat to get to the answer of these questions. Each with their own benefits and drawbacks:</span>
<span id="cb57-85"><a href="#cb57-85" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-86"><a href="#cb57-86" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>**Identification of all markers for each cluster**</span>
<span id="cb57-87"><a href="#cb57-87" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-88"><a href="#cb57-88" aria-hidden="true" tabindex="-1"></a>    This analysis compares each cluster against all others and outputs the genes that are differentially expressed/present.</span>
<span id="cb57-89"><a href="#cb57-89" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-90"><a href="#cb57-90" aria-hidden="true" tabindex="-1"></a><span class="ss">    -   </span>*Useful for identifying **unknown clusters** and i**mproving confidence** in hypothesized cell types.*</span>
<span id="cb57-91"><a href="#cb57-91" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-92"><a href="#cb57-92" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>**Identification of conserved markers for each cluster**</span>
<span id="cb57-93"><a href="#cb57-93" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-94"><a href="#cb57-94" aria-hidden="true" tabindex="-1"></a>    This analysis looks for genes that are differentially expressed/present within each condition first, and then reports those **genes that are conserved in the cluster across all conditions**. These genes can help to figure out the identity for the cluster.</span>
<span id="cb57-95"><a href="#cb57-95" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-96"><a href="#cb57-96" aria-hidden="true" tabindex="-1"></a><span class="ss">    -   </span>*Useful with **more than one condition** to identify cell type markers that are conserved across conditions.*</span>
<span id="cb57-97"><a href="#cb57-97" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-98"><a href="#cb57-98" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>**Marker identification between specific clusters**</span>
<span id="cb57-99"><a href="#cb57-99" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-100"><a href="#cb57-100" aria-hidden="true" tabindex="-1"></a>    This analysis explores differentially expressed genes between specific clusters.</span>
<span id="cb57-101"><a href="#cb57-101" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-102"><a href="#cb57-102" aria-hidden="true" tabindex="-1"></a><span class="ss">    -   </span>*Useful for determining differences in gene expression between clusters that appear to be representing the same celltype (i.e with markers that are similar) from the above analyses.*</span>
<span id="cb57-103"><a href="#cb57-103" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-104"><a href="#cb57-104" aria-hidden="true" tabindex="-1"></a>------------------------------------------------------------------------</span>
<span id="cb57-105"><a href="#cb57-105" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-106"><a href="#cb57-106" aria-hidden="true" tabindex="-1"></a><span class="fu"># `FindAllMarkers`-Identification of all markers for each cluster</span></span>
<span id="cb57-107"><a href="#cb57-107" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-108"><a href="#cb57-108" aria-hidden="true" tabindex="-1"></a>This type of analysis is typically recommended for when **evaluating a single sample group/condition**. With the `FindAllMarkers()` function we are comparing each cluster against all other clusters to identify potential marker genes. The cells in each cluster are treated as replicates, and essentially a **differential expression analysis** is performed with some statistical test.</span>
<span id="cb57-109"><a href="#cb57-109" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-110"><a href="#cb57-110" aria-hidden="true" tabindex="-1"></a>::: callout-tip</span>
<span id="cb57-111"><a href="#cb57-111" aria-hidden="true" tabindex="-1"></a>The default is a **Wilcoxon Rank Sum test**, but there are other options available.</span>
<span id="cb57-112"><a href="#cb57-112" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb57-113"><a href="#cb57-113" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-114"><a href="#cb57-114" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/marker_ident_function1.png)</span>{width="398"}</span>
<span id="cb57-115"><a href="#cb57-115" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-116"><a href="#cb57-116" aria-hidden="true" tabindex="-1"></a>The <span class="in">`FindAllMarkers()`</span> function has **three important arguments** which provide thresholds for determining whether a gene is a marker:</span>
<span id="cb57-117"><a href="#cb57-117" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-118"><a href="#cb57-118" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`logfc.threshold`</span>: minimum log2 fold change for average expression of gene in cluster relative to the average expression in all other clusters combined. Default is **0.25**.</span>
<span id="cb57-119"><a href="#cb57-119" aria-hidden="true" tabindex="-1"></a><span class="ss">    -   </span>**Cons:**</span>
<span id="cb57-120"><a href="#cb57-120" aria-hidden="true" tabindex="-1"></a><span class="ss">        -   </span>could miss those cell **markers that are expressed in a small fraction of cells** within the cluster of interest, but not in the other clusters, if the average logfc doesn't meet the threshold</span>
<span id="cb57-121"><a href="#cb57-121" aria-hidden="true" tabindex="-1"></a><span class="ss">        -   </span>could return a lot of metabolic/ribosomal genes due to slight differences in metabolic output by different cell types, which are not as useful to distinguish cell type identities</span>
<span id="cb57-122"><a href="#cb57-122" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`min.diff.pct`</span>: minimum percent difference between the percent of cells expressing the gene in the cluster and the percent of cells expressing gene in all other clusters combined.</span>
<span id="cb57-123"><a href="#cb57-123" aria-hidden="true" tabindex="-1"></a><span class="ss">    -   </span>**Cons:** could miss those cell markers that are expressed in all cells, but are highly up-regulated in this specific cell type</span>
<span id="cb57-124"><a href="#cb57-124" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`min.pct`</span>: only test genes that are detected in a minimum fraction of cells in either of the two populations. Meant to speed up the function by not testing genes that are very infrequently expressed. Default is 0.1.</span>
<span id="cb57-125"><a href="#cb57-125" aria-hidden="true" tabindex="-1"></a><span class="ss">    -   </span>**Cons:** if set to a very high value could incur many false negatives due to the fact that not all genes are detected in all cells (even if it is expressed)</span>
<span id="cb57-126"><a href="#cb57-126" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-127"><a href="#cb57-127" aria-hidden="true" tabindex="-1"></a>You could use any combination of these arguments depending on how stringent/lenient you want to be. Also, by default this function will return to you genes that exhibit both positive and negative expression changes. Typically, we add an argument <span class="in">`only.pos`</span> to opt for keeping only the positive changes. The code to find markers for each cluster is shown below.</span>
<span id="cb57-128"><a href="#cb57-128" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-129"><a href="#cb57-129" aria-hidden="true" tabindex="-1"></a>::: callout-important</span>
<span id="cb57-130"><a href="#cb57-130" aria-hidden="true" tabindex="-1"></a>In Seurat v5, we use the&nbsp;**`presto`** package&nbsp;(as described&nbsp;<span class="co">[</span><span class="ot">here</span><span class="co">](https://www.biorxiv.org/content/10.1101/653253v1)</span>&nbsp;and available for installation&nbsp;<span class="co">[</span><span class="ot">here</span><span class="co">](https://github.com/immunogenomics/presto)</span>), to dramatically improve the speed of DE analysis, particularly for large datasets (见<span class="co">[</span><span class="ot">Seurat-FindMarkers</span><span class="co">](/single_cell/seurat/marker_gene_identification.qmd#sec-findmarkers_function)</span>).&nbsp;因此，需要先安装<span class="in">`presto`</span>包：</span>
<span id="cb57-131"><a href="#cb57-131" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-134"><a href="#cb57-134" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-135"><a href="#cb57-135" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb57-136"><a href="#cb57-136" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: fenced</span></span>
<span id="cb57-137"><a href="#cb57-137" aria-hidden="true" tabindex="-1"></a><span class="co"># install.packages("devtools")</span></span>
<span id="cb57-138"><a href="#cb57-138" aria-hidden="true" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">"immunogenomics/presto"</span>)</span>
<span id="cb57-139"><a href="#cb57-139" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-140"><a href="#cb57-140" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb57-141"><a href="#cb57-141" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-144"><a href="#cb57-144" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-145"><a href="#cb57-145" aria-hidden="true" tabindex="-1"></a><span class="co"># Find markers for every cluster compared to all remaining cells, report only the positive ones</span></span>
<span id="cb57-146"><a href="#cb57-146" aria-hidden="true" tabindex="-1"></a>markers <span class="ot">&lt;-</span> <span class="fu">FindAllMarkers</span>(<span class="at">object =</span> seurat_clustered_qc, </span>
<span id="cb57-147"><a href="#cb57-147" aria-hidden="true" tabindex="-1"></a>                          <span class="at">only.pos =</span> <span class="cn">TRUE</span>,</span>
<span id="cb57-148"><a href="#cb57-148" aria-hidden="true" tabindex="-1"></a>                          <span class="at">logfc.threshold =</span> <span class="fl">0.25</span>) </span>
<span id="cb57-149"><a href="#cb57-149" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(markers)</span>
<span id="cb57-150"><a href="#cb57-150" aria-hidden="true" tabindex="-1"></a><span class="fu">tail</span>(markers)</span>
<span id="cb57-151"><a href="#cb57-151" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-152"><a href="#cb57-152" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-153"><a href="#cb57-153" aria-hidden="true" tabindex="-1"></a>The results data frame has the following columns :</span>
<span id="cb57-154"><a href="#cb57-154" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-155"><a href="#cb57-155" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`p_val`</span>&nbsp;: p-value (unadjusted)</span>
<span id="cb57-156"><a href="#cb57-156" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-157"><a href="#cb57-157" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`avg_log2FC`</span>&nbsp;: log fold-change of the average expression between the two groups. Positive values indicate that the feature is more highly expressed in the first group.</span>
<span id="cb57-158"><a href="#cb57-158" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-159"><a href="#cb57-159" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`pct.1`</span>&nbsp;: The percentage of cells where the feature is detected in the first group</span>
<span id="cb57-160"><a href="#cb57-160" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-161"><a href="#cb57-161" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`pct.2`</span>&nbsp;: The percentage of cells where the feature is detected in the second group</span>
<span id="cb57-162"><a href="#cb57-162" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-163"><a href="#cb57-163" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`p_val_adj`</span>&nbsp;: Adjusted p-value, based on&nbsp;**Bonferroni correction**&nbsp;using all features in the dataset.</span>
<span id="cb57-164"><a href="#cb57-164" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-165"><a href="#cb57-165" aria-hidden="true" tabindex="-1"></a>获取每个cluster的前10个marker基因 ( 详细解释见下面的 @sec-top_conserved_markers ):</span>
<span id="cb57-166"><a href="#cb57-166" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-169"><a href="#cb57-169" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-170"><a href="#cb57-170" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb57-171"><a href="#cb57-171" aria-hidden="true" tabindex="-1"></a><span class="co"># 获取每个cluster的前10个marker基因</span></span>
<span id="cb57-172"><a href="#cb57-172" aria-hidden="true" tabindex="-1"></a>top_markers <span class="ot">&lt;-</span> markers <span class="sc">|&gt;</span></span>
<span id="cb57-173"><a href="#cb57-173" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(cluster) <span class="sc">|&gt;</span> </span>
<span id="cb57-174"><a href="#cb57-174" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_max</span>(<span class="at">n =</span> <span class="dv">10</span>, <span class="at">order_by =</span> avg_log2FC)</span>
<span id="cb57-175"><a href="#cb57-175" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-176"><a href="#cb57-176" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-177"><a href="#cb57-177" aria-hidden="true" tabindex="-1"></a><span class="fu"># `FindConservedMarkers`-Identification of conserved markers in all conditions</span></span>
<span id="cb57-178"><a href="#cb57-178" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-179"><a href="#cb57-179" aria-hidden="true" tabindex="-1"></a>Since we have samples representing different conditions in our dataset, **our best option is to find conserved markers**. This function internally **separates out cells by sample group/condition**, and then performs differential gene expression testing for a single specified cluster against all other clusters (or a second cluster, if specified). Gene-level p-values are computed for each condition and then combined across groups using meta-analysis methods from the MetaDE R package.</span>
<span id="cb57-180"><a href="#cb57-180" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-181"><a href="#cb57-181" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/marker_ident_function2.png)</span>{width="360"}</span>
<span id="cb57-182"><a href="#cb57-182" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-183"><a href="#cb57-183" aria-hidden="true" tabindex="-1"></a><span class="fu">## 寻找cluster 0和cluster 10的conserved markers</span></span>
<span id="cb57-184"><a href="#cb57-184" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-185"><a href="#cb57-185" aria-hidden="true" tabindex="-1"></a>我们首先通过寻找cluster 0和cluster 10的conserved markers来初步学习<span class="in">`FindConservedMarkers`</span>函数的用法。</span>
<span id="cb57-186"><a href="#cb57-186" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-187"><a href="#cb57-187" aria-hidden="true" tabindex="-1"></a>for <span class="in">`FindConservedMarkers`</span>, you will recognize some of the arguments we described previously for the <span class="in">`FindAllMarkers()`</span> function; this is because internally it is using that function to first find markers within each group. Here, we list some additional arguments which provide for when using <span class="in">`FindConservedMarkers()`</span>:</span>
<span id="cb57-188"><a href="#cb57-188" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-189"><a href="#cb57-189" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`ident.1`</span>: this function only evaluates one cluster at a time; here you would specify the cluster of interest.</span>
<span id="cb57-190"><a href="#cb57-190" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`grouping.var`</span>: the variable (column header) in your metadata which specifies the separation of cells into groups</span>
<span id="cb57-191"><a href="#cb57-191" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-192"><a href="#cb57-192" aria-hidden="true" tabindex="-1"></a><span class="in">`FindConservedMarkers`</span>函数会调用<span class="in">`metap`</span>包，<span class="in">`metap`</span>包需要<span class="in">`multtest`</span>包，所以需要先安装这两个依赖包：</span>
<span id="cb57-193"><a href="#cb57-193" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-196"><a href="#cb57-196" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-197"><a href="#cb57-197" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb57-198"><a href="#cb57-198" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">'multtest'</span>)</span>
<span id="cb57-199"><a href="#cb57-199" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">'metap'</span>)</span>
<span id="cb57-200"><a href="#cb57-200" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-201"><a href="#cb57-201" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-202"><a href="#cb57-202" aria-hidden="true" tabindex="-1"></a><span class="fu">### 寻找cluster 0的conserved markers</span></span>
<span id="cb57-203"><a href="#cb57-203" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-204"><a href="#cb57-204" aria-hidden="true" tabindex="-1"></a>For our analysis we will be fairly lenient and **use only the log fold change threshold greater than 0.25**. We will also specify to return only the positive markers for each cluster.</span>
<span id="cb57-205"><a href="#cb57-205" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-206"><a href="#cb57-206" aria-hidden="true" tabindex="-1"></a>Let's **test it out on cluster 0** to see how it works:</span>
<span id="cb57-207"><a href="#cb57-207" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-210"><a href="#cb57-210" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-211"><a href="#cb57-211" aria-hidden="true" tabindex="-1"></a>cluster0_conserved_markers <span class="ot">&lt;-</span> <span class="fu">FindConservedMarkers</span>(seurat_clustered_qc,</span>
<span id="cb57-212"><a href="#cb57-212" aria-hidden="true" tabindex="-1"></a>                                                   <span class="at">ident.1 =</span> <span class="dv">0</span>,</span>
<span id="cb57-213"><a href="#cb57-213" aria-hidden="true" tabindex="-1"></a>                                                   <span class="at">grouping.var =</span> <span class="st">"sample"</span>,</span>
<span id="cb57-214"><a href="#cb57-214" aria-hidden="true" tabindex="-1"></a>                                                   <span class="at">only.pos =</span> <span class="cn">TRUE</span>,</span>
<span id="cb57-215"><a href="#cb57-215" aria-hidden="true" tabindex="-1"></a>                                                   <span class="at">logfc.threshold =</span> <span class="fl">0.25</span>)</span>
<span id="cb57-216"><a href="#cb57-216" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(cluster0_conserved_markers)</span>
<span id="cb57-217"><a href="#cb57-217" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-218"><a href="#cb57-218" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-219"><a href="#cb57-219" aria-hidden="true" tabindex="-1"></a>The output from the <span class="in">`FindConservedMarkers`</span> function, is a matrix containing a ranked list of putative markers listed by gene ID for the cluster we specified, and associated statistics. Note that the same set of statistics are computed for each group (in our case, Ctrl and Stim) and the last two columns (<span class="in">`max_pval`</span>和<span class="in">`minimump_p_val`</span>) correspond to the combined p-value across the two groups. We describe some of these columns below:</span>
<span id="cb57-220"><a href="#cb57-220" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-221"><a href="#cb57-221" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`gene`</span>: gene symbol</span>
<span id="cb57-222"><a href="#cb57-222" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`condition_p_val`</span>(即本例中的“stim_p_val”和“ctrl_p_val”列，后面同理): p-value not adjusted for multiple test correction for condition</span>
<span id="cb57-223"><a href="#cb57-223" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`condition_avg_log2FC`</span>: average log fold change for condition. Positive values indicate that the gene is more highly expressed in the cluster.</span>
<span id="cb57-224"><a href="#cb57-224" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`condition_pct.1`</span>: percentage of cells where the gene is detected in the cluster for condition</span>
<span id="cb57-225"><a href="#cb57-225" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`condition_pct.2`</span>: percentage of cells where the gene is detected on average in the other clusters for condition</span>
<span id="cb57-226"><a href="#cb57-226" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`condition_p_val_adj`</span>: adjusted p-value for condition, based on **bonferroni correction** using all genes in the dataset, used to determine significance</span>
<span id="cb57-227"><a href="#cb57-227" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`max_pval`</span>: largest p value of p value calculated by each group/condition</span>
<span id="cb57-228"><a href="#cb57-228" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`minimump_p_val`</span>: combined p value</span>
<span id="cb57-229"><a href="#cb57-229" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-230"><a href="#cb57-230" aria-hidden="true" tabindex="-1"></a>::: callout-tip</span>
<span id="cb57-231"><a href="#cb57-231" aria-hidden="true" tabindex="-1"></a>The&nbsp;<span class="in">`condition_p_val`</span>,&nbsp;<span class="in">`condition_avg_log2FC`</span>,&nbsp;<span class="in">`condition_pct.1`</span>,&nbsp;<span class="in">`condition_pct.2`</span>, and&nbsp;<span class="in">`condition_p_val_adj`</span>&nbsp;mean the same thing as they do in&nbsp;<span class="in">`FindMarkers`</span>, just restricted to only the cells present in group X. The&nbsp;<span class="in">`max_pval`</span>&nbsp;is the maximum p-value across all groups. The&nbsp;<span class="in">`mimimump_p_val`</span>&nbsp;represents one way of doing a meta-analysis of significance values (combining p-values across different tests).</span>
<span id="cb57-232"><a href="#cb57-232" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-233"><a href="#cb57-233" aria-hidden="true" tabindex="-1"></a>(来自：<span class="ot">&lt;https://github.com/satijalab/seurat/issues/1164&gt;</span>)</span>
<span id="cb57-234"><a href="#cb57-234" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb57-235"><a href="#cb57-235" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-236"><a href="#cb57-236" aria-hidden="true" tabindex="-1"></a>::: callout-warning</span>
<span id="cb57-237"><a href="#cb57-237" aria-hidden="true" tabindex="-1"></a>Since each cell is being treated as a replicate this will result in inflated p-values within each group! A gene may have an incredibly low p-value <span class="sc">\&lt;</span> 1e-50 but that doesn't translate as a highly reliable marker gene.</span>
<span id="cb57-238"><a href="#cb57-238" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb57-239"><a href="#cb57-239" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-240"><a href="#cb57-240" aria-hidden="true" tabindex="-1"></a>When looking at the output, <span class="co">[</span><span class="ot">**we suggest looking for markers with large differences in expression between `pct.1` and `pct.2` and larger fold changes**</span><span class="co">]</span>{.underline}. For instance if <span class="in">`pct.1`</span> = 0.90 and <span class="in">`pct.2`</span> = 0.80, it may not be as exciting of a marker. However, if <span class="in">`pct.2`</span> = 0.1 instead, the bigger difference would be more convincing. Also, of interest is if the majority of cells expressing the marker is in my cluster of interest. If <span class="in">`pct.1`</span> is low, such as 0.3, it may not be as interesting. Both of these are also possible parameters to include when running the function, as described above.</span>
<span id="cb57-241"><a href="#cb57-241" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-242"><a href="#cb57-242" aria-hidden="true" tabindex="-1"></a><span class="fu">### 添加基因注释信息</span></span>
<span id="cb57-243"><a href="#cb57-243" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-244"><a href="#cb57-244" aria-hidden="true" tabindex="-1"></a>It can be helpful to add columns with gene annotation information. In order to do that we will load in an annotation file located in your <span class="in">`data`</span> folder, using the code provided below:</span>
<span id="cb57-245"><a href="#cb57-245" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-248"><a href="#cb57-248" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-249"><a href="#cb57-249" aria-hidden="true" tabindex="-1"></a>annotations <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"data/scRNA-seq_online/annotations.rds"</span>)</span>
<span id="cb57-250"><a href="#cb57-250" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(annotations, <span class="dv">3</span>)</span>
<span id="cb57-251"><a href="#cb57-251" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-252"><a href="#cb57-252" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-253"><a href="#cb57-253" aria-hidden="true" tabindex="-1"></a>该数据的"description"列即对基因的注释，下面我们把这一列通过匹配基因名将其添加到cluster0_conserved_markers数据框中。</span>
<span id="cb57-254"><a href="#cb57-254" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-255"><a href="#cb57-255" aria-hidden="true" tabindex="-1"></a>::: {.callout-tip collapse="true"}</span>
<span id="cb57-256"><a href="#cb57-256" aria-hidden="true" tabindex="-1"></a><span class="fu">###### 获取基因注释信息的方法</span></span>
<span id="cb57-257"><a href="#cb57-257" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-258"><a href="#cb57-258" aria-hidden="true" tabindex="-1"></a>首先从<span class="in">`BiocManager`</span>安装<span class="co">[</span><span class="ot">`AnnotationHub`包</span><span class="co">](https://bioconductor.org/packages/release/bioc/html/AnnotationHub.html)</span>和<span class="co">[</span><span class="ot">`ensembldb`包</span><span class="co">](https://bioconductor.org/packages/release/bioc/html/ensembldb.html)</span>：</span>
<span id="cb57-259"><a href="#cb57-259" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-262"><a href="#cb57-262" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-263"><a href="#cb57-263" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb57-264"><a href="#cb57-264" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"AnnotationHub"</span>)</span>
<span id="cb57-265"><a href="#cb57-265" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"ensembldb"</span>)</span>
<span id="cb57-266"><a href="#cb57-266" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-267"><a href="#cb57-267" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-268"><a href="#cb57-268" aria-hidden="true" tabindex="-1"></a>**从AnnotationHub下载并提取所需的注释信息数据库:**</span>
<span id="cb57-269"><a href="#cb57-269" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-270"><a href="#cb57-270" aria-hidden="true" tabindex="-1"></a>To access the various annotations available from Ensembl for human, we need to first connect to <span class="in">`AnnotationHub`</span>, then specify the organism and database we are interested in.</span>
<span id="cb57-271"><a href="#cb57-271" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-274"><a href="#cb57-274" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-275"><a href="#cb57-275" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb57-276"><a href="#cb57-276" aria-hidden="true" tabindex="-1"></a><span class="co"># 从AnnotationHub下载注释信息数据库</span></span>
<span id="cb57-277"><a href="#cb57-277" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(AnnotationHub)</span>
<span id="cb57-278"><a href="#cb57-278" aria-hidden="true" tabindex="-1"></a>ah <span class="ot">&lt;-</span> <span class="fu">AnnotationHub</span>()</span>
<span id="cb57-279"><a href="#cb57-279" aria-hidden="true" tabindex="-1"></a>ah</span>
<span id="cb57-280"><a href="#cb57-280" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-281"><a href="#cb57-281" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-282"><a href="#cb57-282" aria-hidden="true" tabindex="-1"></a><span class="in">```         </span></span>
<span id="cb57-283"><a href="#cb57-283" aria-hidden="true" tabindex="-1"></a><span class="in">AnnotationHub with 70762 records</span></span>
<span id="cb57-284"><a href="#cb57-284" aria-hidden="true" tabindex="-1"></a><span class="in"># snapshotDate(): 2023-10-20</span></span>
<span id="cb57-285"><a href="#cb57-285" aria-hidden="true" tabindex="-1"></a><span class="in"># $dataprovider: Ensembl, BroadInstitute, UCSC, ftp://ftp.ncbi.nlm.nih.gov/g...</span></span>
<span id="cb57-286"><a href="#cb57-286" aria-hidden="true" tabindex="-1"></a><span class="in"># $species: Homo sapiens, Mus musculus, Drosophila melanogaster, Bos taurus,...</span></span>
<span id="cb57-287"><a href="#cb57-287" aria-hidden="true" tabindex="-1"></a><span class="in"># $rdataclass: GRanges, TwoBitFile, BigWigFile, EnsDb, Rle, OrgDb, SQLiteFil...</span></span>
<span id="cb57-288"><a href="#cb57-288" aria-hidden="true" tabindex="-1"></a><span class="in"># additional mcols(): taxonomyid, genome, description,</span></span>
<span id="cb57-289"><a href="#cb57-289" aria-hidden="true" tabindex="-1"></a><span class="in">#   coordinate_1_based, maintainer, rdatadateadded, preparerclass, tags,</span></span>
<span id="cb57-290"><a href="#cb57-290" aria-hidden="true" tabindex="-1"></a><span class="in">#   rdatapath, sourceurl, sourcetype </span></span>
<span id="cb57-291"><a href="#cb57-291" aria-hidden="true" tabindex="-1"></a><span class="in"># retrieve records with, e.g., 'object[["AH5012"]]' </span></span>
<span id="cb57-292"><a href="#cb57-292" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-293"><a href="#cb57-293" aria-hidden="true" tabindex="-1"></a><span class="in">             title                              </span></span>
<span id="cb57-294"><a href="#cb57-294" aria-hidden="true" tabindex="-1"></a><span class="in">  AH5012   | Chromosome Band                    </span></span>
<span id="cb57-295"><a href="#cb57-295" aria-hidden="true" tabindex="-1"></a><span class="in">  AH5013   | STS Markers                        </span></span>
<span id="cb57-296"><a href="#cb57-296" aria-hidden="true" tabindex="-1"></a><span class="in">  AH5014   | FISH Clones                        </span></span>
<span id="cb57-297"><a href="#cb57-297" aria-hidden="true" tabindex="-1"></a><span class="in">  AH5015   | Recomb Rate                        </span></span>
<span id="cb57-298"><a href="#cb57-298" aria-hidden="true" tabindex="-1"></a><span class="in">  AH5016   | ENCODE Pilot                       </span></span>
<span id="cb57-299"><a href="#cb57-299" aria-hidden="true" tabindex="-1"></a><span class="in">  ...        ...                                </span></span>
<span id="cb57-300"><a href="#cb57-300" aria-hidden="true" tabindex="-1"></a><span class="in">  AH116159 | org.Aegialitis_vocifera.eg.sqlite  </span></span>
<span id="cb57-301"><a href="#cb57-301" aria-hidden="true" tabindex="-1"></a><span class="in">  AH116160 | org.Charadrius_vociferous.eg.sqlite</span></span>
<span id="cb57-302"><a href="#cb57-302" aria-hidden="true" tabindex="-1"></a><span class="in">  AH116161 | org.Charadrius_vociferus.eg.sqlite </span></span>
<span id="cb57-303"><a href="#cb57-303" aria-hidden="true" tabindex="-1"></a><span class="in">  AH116162 | org.Oxyechus_vociferus.eg.sqlite   </span></span>
<span id="cb57-304"><a href="#cb57-304" aria-hidden="true" tabindex="-1"></a><span class="in">  AH116163 | org.Drosophila_erecta.eg.sqlite    </span></span>
<span id="cb57-305"><a href="#cb57-305" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-306"><a href="#cb57-306" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-309"><a href="#cb57-309" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-310"><a href="#cb57-310" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb57-311"><a href="#cb57-311" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">unique</span>(ah<span class="sc">$</span>species))</span>
<span id="cb57-312"><a href="#cb57-312" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-313"><a href="#cb57-313" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-314"><a href="#cb57-314" aria-hidden="true" tabindex="-1"></a><span class="in">```         </span></span>
<span id="cb57-315"><a href="#cb57-315" aria-hidden="true" tabindex="-1"></a><span class="in">[1] "Homo sapiens"         "Vicugna pacos"        "Dasypus novemcinctus"</span></span>
<span id="cb57-316"><a href="#cb57-316" aria-hidden="true" tabindex="-1"></a><span class="in">[4] "Otolemur garnettii"   "Papio hamadryas"      "Papio anubis"     </span></span>
<span id="cb57-317"><a href="#cb57-317" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-318"><a href="#cb57-318" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-321"><a href="#cb57-321" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-322"><a href="#cb57-322" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb57-323"><a href="#cb57-323" aria-hidden="true" tabindex="-1"></a><span class="co"># Access the Ensembl database for organism</span></span>
<span id="cb57-324"><a href="#cb57-324" aria-hidden="true" tabindex="-1"></a>ahDb <span class="ot">&lt;-</span> <span class="fu">query</span>(ah, </span>
<span id="cb57-325"><a href="#cb57-325" aria-hidden="true" tabindex="-1"></a>              <span class="at">pattern =</span> <span class="fu">c</span>(<span class="st">"Homo sapiens"</span>, <span class="st">"EnsDb"</span>), </span>
<span id="cb57-326"><a href="#cb57-326" aria-hidden="true" tabindex="-1"></a>              <span class="at">ignore.case =</span> <span class="cn">TRUE</span>)</span>
<span id="cb57-327"><a href="#cb57-327" aria-hidden="true" tabindex="-1"></a>ahDb</span>
<span id="cb57-328"><a href="#cb57-328" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-329"><a href="#cb57-329" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-330"><a href="#cb57-330" aria-hidden="true" tabindex="-1"></a><span class="in">```         </span></span>
<span id="cb57-331"><a href="#cb57-331" aria-hidden="true" tabindex="-1"></a><span class="in">AnnotationHub with 25 records</span></span>
<span id="cb57-332"><a href="#cb57-332" aria-hidden="true" tabindex="-1"></a><span class="in"># snapshotDate(): 2023-10-20</span></span>
<span id="cb57-333"><a href="#cb57-333" aria-hidden="true" tabindex="-1"></a><span class="in"># $dataprovider: Ensembl</span></span>
<span id="cb57-334"><a href="#cb57-334" aria-hidden="true" tabindex="-1"></a><span class="in"># $species: Homo sapiens</span></span>
<span id="cb57-335"><a href="#cb57-335" aria-hidden="true" tabindex="-1"></a><span class="in"># $rdataclass: EnsDb</span></span>
<span id="cb57-336"><a href="#cb57-336" aria-hidden="true" tabindex="-1"></a><span class="in"># additional mcols(): taxonomyid, genome, description,</span></span>
<span id="cb57-337"><a href="#cb57-337" aria-hidden="true" tabindex="-1"></a><span class="in">#   coordinate_1_based, maintainer, rdatadateadded, preparerclass, tags,</span></span>
<span id="cb57-338"><a href="#cb57-338" aria-hidden="true" tabindex="-1"></a><span class="in">#   rdatapath, sourceurl, sourcetype </span></span>
<span id="cb57-339"><a href="#cb57-339" aria-hidden="true" tabindex="-1"></a><span class="in"># retrieve records with, e.g., 'object[["AH53211"]]' </span></span>
<span id="cb57-340"><a href="#cb57-340" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-341"><a href="#cb57-341" aria-hidden="true" tabindex="-1"></a><span class="in">             title                             </span></span>
<span id="cb57-342"><a href="#cb57-342" aria-hidden="true" tabindex="-1"></a><span class="in">  AH53211  | Ensembl 87 EnsDb for Homo Sapiens </span></span>
<span id="cb57-343"><a href="#cb57-343" aria-hidden="true" tabindex="-1"></a><span class="in">  AH53715  | Ensembl 88 EnsDb for Homo Sapiens </span></span>
<span id="cb57-344"><a href="#cb57-344" aria-hidden="true" tabindex="-1"></a><span class="in">  AH56681  | Ensembl 89 EnsDb for Homo Sapiens </span></span>
<span id="cb57-345"><a href="#cb57-345" aria-hidden="true" tabindex="-1"></a><span class="in">  AH57757  | Ensembl 90 EnsDb for Homo Sapiens </span></span>
<span id="cb57-346"><a href="#cb57-346" aria-hidden="true" tabindex="-1"></a><span class="in">  AH60773  | Ensembl 91 EnsDb for Homo Sapiens </span></span>
<span id="cb57-347"><a href="#cb57-347" aria-hidden="true" tabindex="-1"></a><span class="in">  ...        ...                               </span></span>
<span id="cb57-348"><a href="#cb57-348" aria-hidden="true" tabindex="-1"></a><span class="in">  AH100643 | Ensembl 106 EnsDb for Homo sapiens</span></span>
<span id="cb57-349"><a href="#cb57-349" aria-hidden="true" tabindex="-1"></a><span class="in">  AH104864 | Ensembl 107 EnsDb for Homo sapiens</span></span>
<span id="cb57-350"><a href="#cb57-350" aria-hidden="true" tabindex="-1"></a><span class="in">  AH109336 | Ensembl 108 EnsDb for Homo sapiens</span></span>
<span id="cb57-351"><a href="#cb57-351" aria-hidden="true" tabindex="-1"></a><span class="in">  AH109606 | Ensembl 109 EnsDb for Homo sapiens</span></span>
<span id="cb57-352"><a href="#cb57-352" aria-hidden="true" tabindex="-1"></a><span class="in">  AH113665 | Ensembl 110 EnsDb for Homo sapiens</span></span>
<span id="cb57-353"><a href="#cb57-353" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-354"><a href="#cb57-354" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-355"><a href="#cb57-355" aria-hidden="true" tabindex="-1"></a>Next, we acquire the latest annotation files from this Ensembl database.</span>
<span id="cb57-356"><a href="#cb57-356" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-357"><a href="#cb57-357" aria-hidden="true" tabindex="-1"></a>We can first check which annotation versions are available. Since we want the most recent, we will return the AnnotationHub ID for this database:</span>
<span id="cb57-358"><a href="#cb57-358" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-361"><a href="#cb57-361" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-362"><a href="#cb57-362" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb57-363"><a href="#cb57-363" aria-hidden="true" tabindex="-1"></a><span class="co"># Acquire the latest annotation files</span></span>
<span id="cb57-364"><a href="#cb57-364" aria-hidden="true" tabindex="-1"></a>id <span class="ot">&lt;-</span> ahDb <span class="sc">|&gt;</span></span>
<span id="cb57-365"><a href="#cb57-365" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mcols</span>() <span class="sc">|&gt;</span></span>
<span id="cb57-366"><a href="#cb57-366" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames</span>() <span class="sc">|&gt;</span></span>
<span id="cb57-367"><a href="#cb57-367" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tail</span>(<span class="at">n =</span> <span class="dv">1</span>)</span>
<span id="cb57-368"><a href="#cb57-368" aria-hidden="true" tabindex="-1"></a>id</span>
<span id="cb57-369"><a href="#cb57-369" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-370"><a href="#cb57-370" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-371"><a href="#cb57-371" aria-hidden="true" tabindex="-1"></a><span class="in">```         </span></span>
<span id="cb57-372"><a href="#cb57-372" aria-hidden="true" tabindex="-1"></a><span class="in">[1] "AH113665"</span></span>
<span id="cb57-373"><a href="#cb57-373" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-374"><a href="#cb57-374" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-375"><a href="#cb57-375" aria-hidden="true" tabindex="-1"></a>Finally, we can use the AnnotationHub connection to download the appropriate Ensembl database.</span>
<span id="cb57-376"><a href="#cb57-376" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-379"><a href="#cb57-379" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-380"><a href="#cb57-380" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb57-381"><a href="#cb57-381" aria-hidden="true" tabindex="-1"></a><span class="co"># Download the appropriate Ensembldb database</span></span>
<span id="cb57-382"><a href="#cb57-382" aria-hidden="true" tabindex="-1"></a><span class="co"># 需要开启全局代理</span></span>
<span id="cb57-383"><a href="#cb57-383" aria-hidden="true" tabindex="-1"></a>edb <span class="ot">&lt;-</span> ah[[id]]</span>
<span id="cb57-384"><a href="#cb57-384" aria-hidden="true" tabindex="-1"></a>edb</span>
<span id="cb57-385"><a href="#cb57-385" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-386"><a href="#cb57-386" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-387"><a href="#cb57-387" aria-hidden="true" tabindex="-1"></a><span class="in">```         </span></span>
<span id="cb57-388"><a href="#cb57-388" aria-hidden="true" tabindex="-1"></a><span class="in">EnsDb for Ensembl:</span></span>
<span id="cb57-389"><a href="#cb57-389" aria-hidden="true" tabindex="-1"></a><span class="in">|Backend: SQLite</span></span>
<span id="cb57-390"><a href="#cb57-390" aria-hidden="true" tabindex="-1"></a><span class="in">|Db type: EnsDb</span></span>
<span id="cb57-391"><a href="#cb57-391" aria-hidden="true" tabindex="-1"></a><span class="in">|Type of Gene ID: Ensembl Gene ID</span></span>
<span id="cb57-392"><a href="#cb57-392" aria-hidden="true" tabindex="-1"></a><span class="in">|Supporting package: ensembldb</span></span>
<span id="cb57-393"><a href="#cb57-393" aria-hidden="true" tabindex="-1"></a><span class="in">|Db created by: ensembldb package from Bioconductor</span></span>
<span id="cb57-394"><a href="#cb57-394" aria-hidden="true" tabindex="-1"></a><span class="in">|script_version: 0.3.10</span></span>
<span id="cb57-395"><a href="#cb57-395" aria-hidden="true" tabindex="-1"></a><span class="in">|Creation time: Mon Aug  7 09:02:07 2023</span></span>
<span id="cb57-396"><a href="#cb57-396" aria-hidden="true" tabindex="-1"></a><span class="in">|ensembl_version: 110</span></span>
<span id="cb57-397"><a href="#cb57-397" aria-hidden="true" tabindex="-1"></a><span class="in">|ensembl_host: 127.0.0.1</span></span>
<span id="cb57-398"><a href="#cb57-398" aria-hidden="true" tabindex="-1"></a><span class="in">|Organism: Homo sapiens</span></span>
<span id="cb57-399"><a href="#cb57-399" aria-hidden="true" tabindex="-1"></a><span class="in">|taxonomy_id: 9606</span></span>
<span id="cb57-400"><a href="#cb57-400" aria-hidden="true" tabindex="-1"></a><span class="in">|genome_build: GRCh38</span></span>
<span id="cb57-401"><a href="#cb57-401" aria-hidden="true" tabindex="-1"></a><span class="in">|DBSCHEMAVERSION: 2.2</span></span>
<span id="cb57-402"><a href="#cb57-402" aria-hidden="true" tabindex="-1"></a><span class="in">|common_name: human</span></span>
<span id="cb57-403"><a href="#cb57-403" aria-hidden="true" tabindex="-1"></a><span class="in">|species: homo_sapiens</span></span>
<span id="cb57-404"><a href="#cb57-404" aria-hidden="true" tabindex="-1"></a><span class="in">| No. of genes: 71440.</span></span>
<span id="cb57-405"><a href="#cb57-405" aria-hidden="true" tabindex="-1"></a><span class="in">| No. of transcripts: 278545.</span></span>
<span id="cb57-406"><a href="#cb57-406" aria-hidden="true" tabindex="-1"></a><span class="in">|Protein data available.</span></span>
<span id="cb57-407"><a href="#cb57-407" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-408"><a href="#cb57-408" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-409"><a href="#cb57-409" aria-hidden="true" tabindex="-1"></a>**提取并保存注释信息：**</span>
<span id="cb57-410"><a href="#cb57-410" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-411"><a href="#cb57-411" aria-hidden="true" tabindex="-1"></a>And to extract gene-level information we can use the Ensembldb function <span class="in">`genes()`</span> to return a data frame of annotations.</span>
<span id="cb57-412"><a href="#cb57-412" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-415"><a href="#cb57-415" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-416"><a href="#cb57-416" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb57-417"><a href="#cb57-417" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract gene-level information from database</span></span>
<span id="cb57-418"><a href="#cb57-418" aria-hidden="true" tabindex="-1"></a>annotations <span class="ot">&lt;-</span> <span class="fu">genes</span>(edb, <span class="at">return.type =</span> <span class="st">"data.frame"</span>)</span>
<span id="cb57-419"><a href="#cb57-419" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(annotations)</span>
<span id="cb57-420"><a href="#cb57-420" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-421"><a href="#cb57-421" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-422"><a href="#cb57-422" aria-hidden="true" tabindex="-1"></a><span class="in">```         </span></span>
<span id="cb57-423"><a href="#cb57-423" aria-hidden="true" tabindex="-1"></a><span class="in">[1] "gene_id"              "gene_name"            "gene_biotype"        </span></span>
<span id="cb57-424"><a href="#cb57-424" aria-hidden="true" tabindex="-1"></a><span class="in">[4] "gene_seq_start"       "gene_seq_end"         "seq_name"            </span></span>
<span id="cb57-425"><a href="#cb57-425" aria-hidden="true" tabindex="-1"></a><span class="in">[7] "seq_strand"           "seq_coord_system"     "description"         </span></span>
<span id="cb57-426"><a href="#cb57-426" aria-hidden="true" tabindex="-1"></a><span class="in">[10] "gene_id_version"      "canonical_transcript" "symbol"              </span></span>
<span id="cb57-427"><a href="#cb57-427" aria-hidden="true" tabindex="-1"></a><span class="in">[13] "entrezid"   </span></span>
<span id="cb57-428"><a href="#cb57-428" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-429"><a href="#cb57-429" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-430"><a href="#cb57-430" aria-hidden="true" tabindex="-1"></a>We aren't interested in all of the information present in this <span class="in">`annotations`</span> file, so we are going to extract that which is useful to us.</span>
<span id="cb57-431"><a href="#cb57-431" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-434"><a href="#cb57-434" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-435"><a href="#cb57-435" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb57-436"><a href="#cb57-436" aria-hidden="true" tabindex="-1"></a><span class="co"># Select annotations of interest</span></span>
<span id="cb57-437"><a href="#cb57-437" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb57-438"><a href="#cb57-438" aria-hidden="true" tabindex="-1"></a>annotations <span class="ot">&lt;-</span> annotations <span class="sc">|&gt;</span></span>
<span id="cb57-439"><a href="#cb57-439" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(gene_id, gene_name, seq_name, gene_biotype, description)</span>
<span id="cb57-440"><a href="#cb57-440" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-441"><a href="#cb57-441" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-442"><a href="#cb57-442" aria-hidden="true" tabindex="-1"></a>**保存到本地:**</span>
<span id="cb57-443"><a href="#cb57-443" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-446"><a href="#cb57-446" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-447"><a href="#cb57-447" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb57-448"><a href="#cb57-448" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(annotations, <span class="at">file =</span> <span class="st">"data/scRNA-seq_online/annotations.rds"</span>)</span>
<span id="cb57-449"><a href="#cb57-449" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-450"><a href="#cb57-450" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb57-451"><a href="#cb57-451" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-452"><a href="#cb57-452" aria-hidden="true" tabindex="-1"></a>First, we will turn the row names with gene identifiers into its own columns. Then we will merge this annotation file with our results from the <span class="in">`FindConservedMarkers()`</span>:</span>
<span id="cb57-453"><a href="#cb57-453" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-454"><a href="#cb57-454" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>首先通过<span class="in">`tibble`</span>包的<span class="in">`rownames_to_column`</span>函数将“cluster0_conserved_markers”数据框的行名（基因symbol）转换成新的一列“gene”</span>
<span id="cb57-455"><a href="#cb57-455" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-456"><a href="#cb57-456" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>然后，通过<span class="in">`dplyr`</span>包的<span class="in">`left_join`</span>函数合并“cluster0_conserved_markers”数据（<span class="in">`x`</span>）和“annotations”数据框的 "description"列（<span class="in">`y`</span>）。通过匹配“cluster0_conserved_markers”数据中的“gene”列（上一步生成）和“annotations”数据的“gene_name”列来进行合并。同时，<span class="in">`left_join`</span>会保留<span class="in">`x`</span>（这里即上一步添加了“gene列”的“cluster0_conserved_markers”数据框）中的所有值；而删除y（“annotations”数据框）中用于匹配的那一列（即“gene_name”）</span>
<span id="cb57-457"><a href="#cb57-457" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-460"><a href="#cb57-460" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-461"><a href="#cb57-461" aria-hidden="true" tabindex="-1"></a><span class="co"># Combine markers with gene descriptions </span></span>
<span id="cb57-462"><a href="#cb57-462" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tibble) <span class="co"># 调用rownames_to_column函数</span></span>
<span id="cb57-463"><a href="#cb57-463" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr) <span class="co"># 调用left_join函数</span></span>
<span id="cb57-464"><a href="#cb57-464" aria-hidden="true" tabindex="-1"></a>cluster0_ann_markers <span class="ot">&lt;-</span> cluster0_conserved_markers <span class="sc">|&gt;</span></span>
<span id="cb57-465"><a href="#cb57-465" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="at">var =</span> <span class="st">"gene"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb57-466"><a href="#cb57-466" aria-hidden="true" tabindex="-1"></a>  <span class="co"># left_join保留x中的所有观测</span></span>
<span id="cb57-467"><a href="#cb57-467" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(<span class="at">y =</span> <span class="fu">unique</span>(annotations[, <span class="fu">c</span>(<span class="st">"gene_name"</span>, <span class="st">"description"</span>)]),</span>
<span id="cb57-468"><a href="#cb57-468" aria-hidden="true" tabindex="-1"></a>            <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"gene"</span> <span class="ot">=</span> <span class="st">"gene_name"</span>))</span>
<span id="cb57-469"><a href="#cb57-469" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(cluster0_ann_markers, <span class="dv">3</span>)</span>
<span id="cb57-470"><a href="#cb57-470" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-471"><a href="#cb57-471" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-472"><a href="#cb57-472" aria-hidden="true" tabindex="-1"></a><span class="fu">### 寻找cluster 10的conserved markers</span></span>
<span id="cb57-473"><a href="#cb57-473" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-474"><a href="#cb57-474" aria-hidden="true" tabindex="-1"></a>In the <span class="co">[</span><span class="ot">previous lesson</span><span class="co">](/single_cell/scRNA-seq_online/08_SC_clustering_quality_control.qmd#sec-fcgr3a_monocyte_markers)</span>, we identified **cluster 10** as FCGR3A+ monocytes by inspecting the expression of known cell markers FCGR3A and MS4A7.</span>
<span id="cb57-475"><a href="#cb57-475" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-478"><a href="#cb57-478" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-479"><a href="#cb57-479" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 10</span></span>
<span id="cb57-480"><a href="#cb57-480" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>(seurat_clustered_qc, </span>
<span id="cb57-481"><a href="#cb57-481" aria-hidden="true" tabindex="-1"></a>            <span class="at">reduction =</span> <span class="st">"umap"</span>, </span>
<span id="cb57-482"><a href="#cb57-482" aria-hidden="true" tabindex="-1"></a>            <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"FCGR3A"</span>, <span class="st">"MS4A7"</span>), </span>
<span id="cb57-483"><a href="#cb57-483" aria-hidden="true" tabindex="-1"></a>            <span class="at">order =</span> <span class="cn">TRUE</span>,</span>
<span id="cb57-484"><a href="#cb57-484" aria-hidden="true" tabindex="-1"></a>            <span class="at">min.cutoff =</span> <span class="st">'q10'</span>, </span>
<span id="cb57-485"><a href="#cb57-485" aria-hidden="true" tabindex="-1"></a>            <span class="at">label =</span> <span class="cn">TRUE</span>)</span>
<span id="cb57-486"><a href="#cb57-486" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-487"><a href="#cb57-487" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-488"><a href="#cb57-488" aria-hidden="true" tabindex="-1"></a>Now, we use <span class="in">`FindConservedMarkers()`</span> function to find conserved markers for cluster 10.</span>
<span id="cb57-489"><a href="#cb57-489" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-492"><a href="#cb57-492" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-493"><a href="#cb57-493" aria-hidden="true" tabindex="-1"></a>cluster10_conserved_markers <span class="ot">&lt;-</span> <span class="fu">FindConservedMarkers</span>(seurat_clustered_qc,</span>
<span id="cb57-494"><a href="#cb57-494" aria-hidden="true" tabindex="-1"></a>                                                    <span class="at">ident.1 =</span> <span class="dv">10</span>,</span>
<span id="cb57-495"><a href="#cb57-495" aria-hidden="true" tabindex="-1"></a>                                                    <span class="at">grouping.var =</span> <span class="st">"sample"</span>,</span>
<span id="cb57-496"><a href="#cb57-496" aria-hidden="true" tabindex="-1"></a>                                                    <span class="at">only.pos =</span> <span class="cn">TRUE</span>,</span>
<span id="cb57-497"><a href="#cb57-497" aria-hidden="true" tabindex="-1"></a>                                                    <span class="at">logfc.threshold =</span> <span class="fl">0.25</span>)</span>
<span id="cb57-498"><a href="#cb57-498" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-499"><a href="#cb57-499" aria-hidden="true" tabindex="-1"></a>cluster10_ann_markers <span class="ot">&lt;-</span> cluster10_conserved_markers <span class="sc">|&gt;</span> </span>
<span id="cb57-500"><a href="#cb57-500" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="at">var=</span><span class="st">"gene"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb57-501"><a href="#cb57-501" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(<span class="at">y =</span> <span class="fu">unique</span>(annotations[, <span class="fu">c</span>(<span class="st">"gene_name"</span>, <span class="st">"description"</span>)]),</span>
<span id="cb57-502"><a href="#cb57-502" aria-hidden="true" tabindex="-1"></a>            <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"gene"</span> <span class="ot">=</span> <span class="st">"gene_name"</span>))</span>
<span id="cb57-503"><a href="#cb57-503" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(cluster10_ann_markers)</span>
<span id="cb57-504"><a href="#cb57-504" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-505"><a href="#cb57-505" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-506"><a href="#cb57-506" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; 可以发现cluster10的FCGR3A和MS4A7的表达比例显著高于其他cluster。符合此前的判断。</span></span>
<span id="cb57-507"><a href="#cb57-507" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-508"><a href="#cb57-508" aria-hidden="true" tabindex="-1"></a>------------------------------------------------------------------------</span>
<span id="cb57-509"><a href="#cb57-509" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-510"><a href="#cb57-510" aria-hidden="true" tabindex="-1"></a><span class="fu">## 批量寻找多个clusters的conserved markers {#sec-function_to_find_markers}</span></span>
<span id="cb57-511"><a href="#cb57-511" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-512"><a href="#cb57-512" aria-hidden="true" tabindex="-1"></a>The function <span class="in">`FindConservedMarkers()`</span> **accepts a single cluster at a time**, and we could run this function as many times as we have clusters. However, this is not very efficient. Instead we will first create a function to find the conserved markers including all the parameters we want to include. We will also **add a few lines of code to modify the output**. Our function will:</span>
<span id="cb57-513"><a href="#cb57-513" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-514"><a href="#cb57-514" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>Run the <span class="in">`FindConservedMarkers()`</span> function</span>
<span id="cb57-515"><a href="#cb57-515" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>Transfer row names to a column using <span class="in">`rownames_to_column()`</span> function</span>
<span id="cb57-516"><a href="#cb57-516" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>Merge in annotations</span>
<span id="cb57-517"><a href="#cb57-517" aria-hidden="true" tabindex="-1"></a><span class="ss">4.  </span>Create the column of cluster IDs using the <span class="in">`cbind()`</span> function</span>
<span id="cb57-518"><a href="#cb57-518" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-521"><a href="#cb57-521" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-522"><a href="#cb57-522" aria-hidden="true" tabindex="-1"></a><span class="co"># Create function to get conserved markers for any given cluster</span></span>
<span id="cb57-523"><a href="#cb57-523" aria-hidden="true" tabindex="-1"></a>get_conserved <span class="ot">&lt;-</span> <span class="cf">function</span>(cluster) {</span>
<span id="cb57-524"><a href="#cb57-524" aria-hidden="true" tabindex="-1"></a>  <span class="fu">FindConservedMarkers</span>(seurat_clustered_qc,</span>
<span id="cb57-525"><a href="#cb57-525" aria-hidden="true" tabindex="-1"></a>                       <span class="at">ident.1 =</span> cluster,</span>
<span id="cb57-526"><a href="#cb57-526" aria-hidden="true" tabindex="-1"></a>                       <span class="at">grouping.var =</span> <span class="st">"sample"</span>,</span>
<span id="cb57-527"><a href="#cb57-527" aria-hidden="true" tabindex="-1"></a>                       <span class="at">only.pos =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb57-528"><a href="#cb57-528" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rownames_to_column</span>(<span class="at">var =</span> <span class="st">"gene"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb57-529"><a href="#cb57-529" aria-hidden="true" tabindex="-1"></a>    <span class="fu">left_join</span>(<span class="at">y =</span> <span class="fu">unique</span>(annotations[, <span class="fu">c</span>(<span class="st">"gene_name"</span>, <span class="st">"description"</span>)]),</span>
<span id="cb57-530"><a href="#cb57-530" aria-hidden="true" tabindex="-1"></a>               <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"gene"</span> <span class="ot">=</span> <span class="st">"gene_name"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb57-531"><a href="#cb57-531" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cbind</span>(<span class="at">cluster_id =</span> cluster, .)</span>
<span id="cb57-532"><a href="#cb57-532" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb57-533"><a href="#cb57-533" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-534"><a href="#cb57-534" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-535"><a href="#cb57-535" aria-hidden="true" tabindex="-1"></a>Now that we have this function created we can use it as an argument to the appropriate <span class="in">`map`</span> function. We want the output of the <span class="in">`map`</span> family of functions to be a **dataframe with each cluster output bound together by rows**. （<span class="in">`map`</span>函数输出的为一个list，通过<span class="in">`list_rbind`</span>函数按照行组合列表中的每一个对象，并输出为数据框）</span>
<span id="cb57-536"><a href="#cb57-536" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-537"><a href="#cb57-537" aria-hidden="true" tabindex="-1"></a>Now, let's try this function to find the conserved markers for the clusters that were identified as **CD4+ T cells (4, 0, 6, 2)** from our use of <span class="co">[</span><span class="ot">known marker genes</span><span class="co">](/single_cell/scRNA-seq_online/08_SC_clustering_quality_control.qmd#sec-cd4_t_cells)</span>.</span>
<span id="cb57-538"><a href="#cb57-538" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-541"><a href="#cb57-541" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-542"><a href="#cb57-542" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 10</span></span>
<span id="cb57-543"><a href="#cb57-543" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 8</span></span>
<span id="cb57-544"><a href="#cb57-544" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>(seurat_clustered_qc, </span>
<span id="cb57-545"><a href="#cb57-545" aria-hidden="true" tabindex="-1"></a>            <span class="at">reduction =</span> <span class="st">"umap"</span>, </span>
<span id="cb57-546"><a href="#cb57-546" aria-hidden="true" tabindex="-1"></a>            <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"CD3D"</span>, <span class="st">"IL7R"</span>, <span class="st">"CCR7"</span>), </span>
<span id="cb57-547"><a href="#cb57-547" aria-hidden="true" tabindex="-1"></a>            <span class="at">order =</span> <span class="cn">TRUE</span>,</span>
<span id="cb57-548"><a href="#cb57-548" aria-hidden="true" tabindex="-1"></a>            <span class="at">min.cutoff =</span> <span class="st">'q10'</span>, </span>
<span id="cb57-549"><a href="#cb57-549" aria-hidden="true" tabindex="-1"></a>            <span class="at">label =</span> <span class="cn">TRUE</span>)</span>
<span id="cb57-550"><a href="#cb57-550" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-551"><a href="#cb57-551" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-552"><a href="#cb57-552" aria-hidden="true" tabindex="-1"></a>Let's see what genes we identify and of there are overlaps or obvious differences that can help us tease this apart a bit more.</span>
<span id="cb57-553"><a href="#cb57-553" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-556"><a href="#cb57-556" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-557"><a href="#cb57-557" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(purrr)</span>
<span id="cb57-558"><a href="#cb57-558" aria-hidden="true" tabindex="-1"></a>conserved_markers <span class="ot">&lt;-</span> <span class="fu">map</span>(<span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">0</span>, <span class="dv">6</span>, <span class="dv">2</span>), get_conserved) <span class="sc">|&gt;</span> <span class="fu">list_rbind</span>()</span>
<span id="cb57-559"><a href="#cb57-559" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-560"><a href="#cb57-560" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-561"><a href="#cb57-561" aria-hidden="true" tabindex="-1"></a>::: callout-tip</span>
<span id="cb57-562"><a href="#cb57-562" aria-hidden="true" tabindex="-1"></a><span class="fu">###### Finding markers for all clusters</span></span>
<span id="cb57-563"><a href="#cb57-563" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-564"><a href="#cb57-564" aria-hidden="true" tabindex="-1"></a>For your data, you may want to run this function on all clusters, in which case you could input <span class="in">`0:20`</span> instead of <span class="in">`c(4,0,6,2)`</span>. Also, it is possible that when you run this function on all clusters, in **some cases you will have clusters that do not have enough cells for a particular group** - and your function will fail. For these clusters you will need to use <span class="in">`FindAllMarkers()`</span>.</span>
<span id="cb57-565"><a href="#cb57-565" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb57-566"><a href="#cb57-566" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-567"><a href="#cb57-567" aria-hidden="true" tabindex="-1"></a><span class="fu">### 获取top marker基因 {#sec-top_conserved_markers}</span></span>
<span id="cb57-568"><a href="#cb57-568" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-569"><a href="#cb57-569" aria-hidden="true" tabindex="-1"></a>We would like to use these gene lists to see of we can **identify which celltypes these clusters identify with.** Let's take a look at the top genes for each of the clusters and see if that gives us any hints. We can view the top 10 markers by **average fold change** across the two groups, for each cluster for a quick perusal:</span>
<span id="cb57-570"><a href="#cb57-570" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-571"><a href="#cb57-571" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>首先通过<span class="in">`dplyr`</span>包的<span class="in">`mutate`</span>函数计算新的变量“avg_fc”，计算依据为：avg_fc = (ctrl_avg_log2FC + stim_avg_log2FC) /2</span>
<span id="cb57-572"><a href="#cb57-572" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>然后通过<span class="in">`dplyr`</span>包的<span class="in">`group_by`</span>函数以“cluster_id”列为依据进行分组计算</span>
<span id="cb57-573"><a href="#cb57-573" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>最后，通过<span class="in">`dplyr`</span>包的<span class="in">`slice_max`</span>函数取“avg_fc”（<span class="in">`order_by = avg_fc`</span>）最大的前10行数据（<span class="in">`n = 10`</span>）。由于<span class="in">`group_by`</span>定义了分组计算，所以会输出每个cluster的前10个marker基因</span>
<span id="cb57-574"><a href="#cb57-574" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-577"><a href="#cb57-577" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-578"><a href="#cb57-578" aria-hidden="true" tabindex="-1"></a><span class="co"># 获取每个cluster的前10个marker基因</span></span>
<span id="cb57-579"><a href="#cb57-579" aria-hidden="true" tabindex="-1"></a>top_conserved_markers <span class="ot">&lt;-</span> conserved_markers <span class="sc">%&gt;%</span> </span>
<span id="cb57-580"><a href="#cb57-580" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">avg_fc =</span> (ctrl_avg_log2FC <span class="sc">+</span> stim_avg_log2FC) <span class="sc">/</span><span class="dv">2</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb57-581"><a href="#cb57-581" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(cluster_id) <span class="sc">%&gt;%</span> </span>
<span id="cb57-582"><a href="#cb57-582" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_max</span>(<span class="at">n =</span> <span class="dv">10</span>, <span class="at">order_by =</span> avg_fc)</span>
<span id="cb57-583"><a href="#cb57-583" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-584"><a href="#cb57-584" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-585"><a href="#cb57-585" aria-hidden="true" tabindex="-1"></a>::: callout-warning</span>
<span id="cb57-586"><a href="#cb57-586" aria-hidden="true" tabindex="-1"></a><span class="fu">###### 待解决的问题</span></span>
<span id="cb57-587"><a href="#cb57-587" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-588"><a href="#cb57-588" aria-hidden="true" tabindex="-1"></a>这里基于Seurat V5的运行结果和<span class="co">[</span><span class="ot">原教程</span><span class="co">](https://hbctraining.github.io/scRNA-seq_online/lessons/09_merged_SC_marker_identification.html)</span>的top markers结果不一致。原教程的top_conserved_markers如下：</span>
<span id="cb57-589"><a href="#cb57-589" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-590"><a href="#cb57-590" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/tcell_marker_table_SCTv2.png)</span></span>
<span id="cb57-591"><a href="#cb57-591" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-592"><a href="#cb57-592" aria-hidden="true" tabindex="-1"></a>可能的原因：</span>
<span id="cb57-593"><a href="#cb57-593" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-594"><a href="#cb57-594" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; “In addition, in Seurat v5 we implement a&nbsp;**pseudocount**&nbsp;(when calculating log-FC) at the **group level** instead of the cell level. As a result, users will observe **higher logFC estimates** in v5 - but should note that these estimates may be more unstable - particularly for genes that are very lowly expressed in one of the two groups” ——</span><span class="co">[</span><span class="ot">Changes in Seurat v5</span><span class="co">](/single_cell/seurat/seurat_intro.qmd#sec-changes_in_seurat_v5)</span></span>
<span id="cb57-595"><a href="#cb57-595" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-596"><a href="#cb57-596" aria-hidden="true" tabindex="-1"></a>这一小节的后续内容暂时以原教程的marker基因结果为准。</span>
<span id="cb57-597"><a href="#cb57-597" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb57-598"><a href="#cb57-598" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-599"><a href="#cb57-599" aria-hidden="true" tabindex="-1"></a><span class="fu">### 根据top marker基因重新评估细胞群的注释结果</span></span>
<span id="cb57-600"><a href="#cb57-600" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-601"><a href="#cb57-601" aria-hidden="true" tabindex="-1"></a>When we look at the entire list, we see **clusters 0** **and 6** have some overlapping genes, like **CCR7** and **SELL** which correspond to **markers of memory T cells**.</span>
<span id="cb57-602"><a href="#cb57-602" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-603"><a href="#cb57-603" aria-hidden="true" tabindex="-1"></a>It is possible that these two clusters are more similar to one another and could be merged together as **naive T cells**. On the other hand, with **cluster 2** we observe **CREM** as one of our top genes; a marker gene of activation. This suggests that perhaps cluster 2 represents activated T cells.</span>
<span id="cb57-604"><a href="#cb57-604" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-605"><a href="#cb57-605" aria-hidden="true" tabindex="-1"></a>|    Cell State     |   Marker   |</span>
<span id="cb57-606"><a href="#cb57-606" aria-hidden="true" tabindex="-1"></a>|:-----------------:|:----------:|</span>
<span id="cb57-607"><a href="#cb57-607" aria-hidden="true" tabindex="-1"></a>|   Naive T cells   | CCR7, SELL |</span>
<span id="cb57-608"><a href="#cb57-608" aria-hidden="true" tabindex="-1"></a>| Activated T cells | CREM, CD69 |</span>
<span id="cb57-609"><a href="#cb57-609" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-610"><a href="#cb57-610" aria-hidden="true" tabindex="-1"></a>For **cluster 4**, we see a lot of heat shock and DNA damage genes appear in the top gene list. Based on these markers, it is likely that these are **stressed or dying cells**. However, if we explore the quality metrics for these cells in more detail (i.e. mitoRatio and nUMI overlayed on the cluster) we don't really support for this argument：</span>
<span id="cb57-611"><a href="#cb57-611" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-614"><a href="#cb57-614" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-615"><a href="#cb57-615" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualize the distribution of mitochondrial gene expression detected per cell</span></span>
<span id="cb57-616"><a href="#cb57-616" aria-hidden="true" tabindex="-1"></a><span class="co"># cluster 4中每个细胞检测到的线粒体基因表达分布情况</span></span>
<span id="cb57-617"><a href="#cb57-617" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb57-618"><a href="#cb57-618" aria-hidden="true" tabindex="-1"></a>seurat_clustered_qc <span class="sc">%&gt;%</span> </span>
<span id="cb57-619"><a href="#cb57-619" aria-hidden="true" tabindex="-1"></a>  <span class="fu">subset</span>(<span class="at">idents =</span> <span class="dv">4</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb57-620"><a href="#cb57-620" aria-hidden="true" tabindex="-1"></a>  .<span class="sc">@</span>meta.data <span class="sc">%&gt;%</span> </span>
<span id="cb57-621"><a href="#cb57-621" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">color =</span> sample, <span class="at">x =</span> mitoRatio, <span class="at">fill =</span> sample)) <span class="sc">+</span> </span>
<span id="cb57-622"><a href="#cb57-622" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">alpha =</span> <span class="fl">0.2</span>) <span class="sc">+</span> </span>
<span id="cb57-623"><a href="#cb57-623" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_log10</span>() <span class="sc">+</span> </span>
<span id="cb57-624"><a href="#cb57-624" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb57-625"><a href="#cb57-625" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fl">0.2</span>)</span>
<span id="cb57-626"><a href="#cb57-626" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-627"><a href="#cb57-627" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-628"><a href="#cb57-628" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; There is a breadth of research supporting the association of heat shock proteins with reactive T cells in the induction of anti‐inflammatory cytokines in chronic inflammation. This is a cluster for which we would need a deeper understanding of immune cells to really tease apart the results and make a final conclusion.</span></span>
<span id="cb57-629"><a href="#cb57-629" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-630"><a href="#cb57-630" aria-hidden="true" tabindex="-1"></a>To get a better idea of cell type identity for cluster 4 we can explore the expression of different identified markers by cluster using the <span class="in">`FeaturePlot()`</span> function.</span>
<span id="cb57-631"><a href="#cb57-631" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-634"><a href="#cb57-634" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-635"><a href="#cb57-635" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 10</span></span>
<span id="cb57-636"><a href="#cb57-636" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 8</span></span>
<span id="cb57-637"><a href="#cb57-637" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot interesting marker gene expression for cluster 4</span></span>
<span id="cb57-638"><a href="#cb57-638" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>(seurat_clustered_qc, </span>
<span id="cb57-639"><a href="#cb57-639" aria-hidden="true" tabindex="-1"></a>            <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"HSPH1"</span>, <span class="st">"HSPE1"</span>, <span class="st">"DNAJB1"</span>),</span>
<span id="cb57-640"><a href="#cb57-640" aria-hidden="true" tabindex="-1"></a>            <span class="at">order =</span> <span class="cn">TRUE</span>,</span>
<span id="cb57-641"><a href="#cb57-641" aria-hidden="true" tabindex="-1"></a>            <span class="at">min.cutoff =</span> <span class="st">'q10'</span>, </span>
<span id="cb57-642"><a href="#cb57-642" aria-hidden="true" tabindex="-1"></a>            <span class="at">label =</span> <span class="cn">TRUE</span>,</span>
<span id="cb57-643"><a href="#cb57-643" aria-hidden="true" tabindex="-1"></a>            <span class="at">repel =</span> <span class="cn">TRUE</span>)</span>
<span id="cb57-644"><a href="#cb57-644" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-645"><a href="#cb57-645" aria-hidden="true" tabindex="-1"></a><span class="co"># 提取cluster 4，并单独查看interesting marker gene在其中的表达情况</span></span>
<span id="cb57-646"><a href="#cb57-646" aria-hidden="true" tabindex="-1"></a><span class="fu">subset</span>(seurat_clustered_qc, <span class="at">idents =</span> <span class="dv">4</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb57-647"><a href="#cb57-647" aria-hidden="true" tabindex="-1"></a>  <span class="fu">FeaturePlot</span>(</span>
<span id="cb57-648"><a href="#cb57-648" aria-hidden="true" tabindex="-1"></a>            <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"HSPH1"</span>, <span class="st">"HSPE1"</span>, <span class="st">"DNAJB1"</span>),</span>
<span id="cb57-649"><a href="#cb57-649" aria-hidden="true" tabindex="-1"></a>            <span class="at">order =</span> <span class="cn">TRUE</span>,</span>
<span id="cb57-650"><a href="#cb57-650" aria-hidden="true" tabindex="-1"></a>            <span class="at">min.cutoff =</span> <span class="st">'q10'</span>, </span>
<span id="cb57-651"><a href="#cb57-651" aria-hidden="true" tabindex="-1"></a>            <span class="at">label =</span> <span class="cn">TRUE</span>,</span>
<span id="cb57-652"><a href="#cb57-652" aria-hidden="true" tabindex="-1"></a>            <span class="at">repel =</span> <span class="cn">TRUE</span>)</span>
<span id="cb57-653"><a href="#cb57-653" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-654"><a href="#cb57-654" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-655"><a href="#cb57-655" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; We see that **only a subset of cluster 4 are highly expressing these genes**.</span></span>
<span id="cb57-656"><a href="#cb57-656" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-657"><a href="#cb57-657" aria-hidden="true" tabindex="-1"></a>We can also explore the range in expression of specific markers by using **violin plots**:</span>
<span id="cb57-658"><a href="#cb57-658" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-661"><a href="#cb57-661" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-662"><a href="#cb57-662" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 10</span></span>
<span id="cb57-663"><a href="#cb57-663" aria-hidden="true" tabindex="-1"></a><span class="co"># Vln plot - cluster 4</span></span>
<span id="cb57-664"><a href="#cb57-664" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(seurat_clustered_qc, </span>
<span id="cb57-665"><a href="#cb57-665" aria-hidden="true" tabindex="-1"></a>        <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"HSPH1"</span>, <span class="st">"HSPE1"</span>, <span class="st">"DNAJB1"</span>))</span>
<span id="cb57-666"><a href="#cb57-666" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-667"><a href="#cb57-667" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-668"><a href="#cb57-668" aria-hidden="true" tabindex="-1"></a>::: {.callout-tip collapse="true"}</span>
<span id="cb57-669"><a href="#cb57-669" aria-hidden="true" tabindex="-1"></a><span class="fu">###### Violin plots</span></span>
<span id="cb57-670"><a href="#cb57-670" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-671"><a href="#cb57-671" aria-hidden="true" tabindex="-1"></a>Violin plots are similar to box plots, except that they also show the **probability density** of the data at different values, usually smoothed by a **kernel density estimator**. A violin plot is more informative than a plain box plot. While a box plot only shows summary statistics such as mean/median and interquartile ranges, the **violin plot shows the full distribution of the data**. The difference is particularly useful when the data distribution is multimodal (more than one peak). In this case a violin plot shows the presence of different peaks, their position and relative amplitude.</span>
<span id="cb57-672"><a href="#cb57-672" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb57-673"><a href="#cb57-673" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-674"><a href="#cb57-674" aria-hidden="true" tabindex="-1"></a><span class="fu"># `FindMarkers`-**Marker identification between specific clusters**</span></span>
<span id="cb57-675"><a href="#cb57-675" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-676"><a href="#cb57-676" aria-hidden="true" tabindex="-1"></a>Sometimes the list of markers returned don't sufficiently separate some of the clusters. For instance, we had previously identified clusters 0, 4, 6 and 2 as CD4+ T cells, but when looking at marker gene lists we identfied markers to help us further subset cells. We were lucky and the signal observed from <span class="in">`FindAllMarkers()`</span> helped us differentiate between naive and activated cells.</span>
<span id="cb57-677"><a href="#cb57-677" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-678"><a href="#cb57-678" aria-hidden="true" tabindex="-1"></a>Another option to identify biologically meaningful differences would be to use the **`FindMarkers()` function to determine the genes that are differentially expressed between two specific clusters**.</span>
<span id="cb57-679"><a href="#cb57-679" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-680"><a href="#cb57-680" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/marker_ident_function3.png)</span>{width="379"}</span>
<span id="cb57-681"><a href="#cb57-681" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-682"><a href="#cb57-682" aria-hidden="true" tabindex="-1"></a>We can try all combinations of comparisons, but we'll start with **cluster 2** versus all other CD4+ T cell clusters:</span>
<span id="cb57-683"><a href="#cb57-683" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-686"><a href="#cb57-686" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-687"><a href="#cb57-687" aria-hidden="true" tabindex="-1"></a><span class="co"># Determine differentiating markers for CD4+ T cell</span></span>
<span id="cb57-688"><a href="#cb57-688" aria-hidden="true" tabindex="-1"></a>cd4_tcells <span class="ot">&lt;-</span> <span class="fu">FindMarkers</span>(seurat_clustered_qc,</span>
<span id="cb57-689"><a href="#cb57-689" aria-hidden="true" tabindex="-1"></a>                          <span class="at">ident.1 =</span> <span class="dv">2</span>,</span>
<span id="cb57-690"><a href="#cb57-690" aria-hidden="true" tabindex="-1"></a>                          <span class="at">ident.2 =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">4</span>, <span class="dv">6</span>))  </span>
<span id="cb57-691"><a href="#cb57-691" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-692"><a href="#cb57-692" aria-hidden="true" tabindex="-1"></a><span class="co"># Add gene symbols to the DE table</span></span>
<span id="cb57-693"><a href="#cb57-693" aria-hidden="true" tabindex="-1"></a>cd4_tcells <span class="ot">&lt;-</span> cd4_tcells <span class="sc">%&gt;%</span></span>
<span id="cb57-694"><a href="#cb57-694" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="at">var =</span> <span class="st">"gene"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb57-695"><a href="#cb57-695" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(<span class="at">y =</span> <span class="fu">unique</span>(annotations[, <span class="fu">c</span>(<span class="st">"gene_name"</span>, <span class="st">"description"</span>)]),</span>
<span id="cb57-696"><a href="#cb57-696" aria-hidden="true" tabindex="-1"></a>            <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"gene"</span> <span class="ot">=</span> <span class="st">"gene_name"</span>))</span>
<span id="cb57-697"><a href="#cb57-697" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(cd4_tcells)</span>
<span id="cb57-698"><a href="#cb57-698" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-699"><a href="#cb57-699" aria-hidden="true" tabindex="-1"></a><span class="co"># Reorder columns and sort by "p_val_adj"</span></span>
<span id="cb57-700"><a href="#cb57-700" aria-hidden="true" tabindex="-1"></a>cd4_tcells <span class="ot">&lt;-</span> cd4_tcells[, <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">2</span>, <span class="dv">6</span><span class="sc">:</span><span class="dv">7</span>)]</span>
<span id="cb57-701"><a href="#cb57-701" aria-hidden="true" tabindex="-1"></a>cd4_tcells <span class="ot">&lt;-</span> <span class="fu">arrange</span>(cd4_tcells, p_val_adj)</span>
<span id="cb57-702"><a href="#cb57-702" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-703"><a href="#cb57-703" aria-hidden="true" tabindex="-1"></a>cd4_tcells<span class="sc">$</span>gene[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span>
<span id="cb57-704"><a href="#cb57-704" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-705"><a href="#cb57-705" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-706"><a href="#cb57-706" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/截屏2023-12-22%2011.37.02.png)</span></span>
<span id="cb57-707"><a href="#cb57-707" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-708"><a href="#cb57-708" aria-hidden="true" tabindex="-1"></a>::: callout-warning</span>
<span id="cb57-709"><a href="#cb57-709" aria-hidden="true" tabindex="-1"></a>这个表格和<span class="co">[</span><span class="ot">原教程</span><span class="co">](https://hbctraining.github.io/scRNA-seq_online/lessons/09_merged_SC_marker_identification.html)</span>的仍然有差异。</span>
<span id="cb57-710"><a href="#cb57-710" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb57-711"><a href="#cb57-711" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-712"><a href="#cb57-712" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; Of these top genes the **CREM gene** stands out as a marker of activation with a positive fold change. We also see markers of naive or memory cells include the **SELL** and **CCR7** genes with negative fold changes, which is in line with previous results.</span></span>
<span id="cb57-713"><a href="#cb57-713" aria-hidden="true" tabindex="-1"></a><span class="at">&gt;</span></span>
<span id="cb57-714"><a href="#cb57-714" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; |    Cell State     |   Marker   |</span></span>
<span id="cb57-715"><a href="#cb57-715" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; |:-----------------:|:----------:|</span></span>
<span id="cb57-716"><a href="#cb57-716" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; |   Naive T cells   | CCR7, SELL |</span></span>
<span id="cb57-717"><a href="#cb57-717" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; | Activated T cells | CREM, CD69 |</span></span>
<span id="cb57-718"><a href="#cb57-718" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-719"><a href="#cb57-719" aria-hidden="true" tabindex="-1"></a>进一步通过气泡图来可视化上述基因的表达.</span>
<span id="cb57-720"><a href="#cb57-720" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-723"><a href="#cb57-723" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-724"><a href="#cb57-724" aria-hidden="true" tabindex="-1"></a><span class="fu">DotPlot</span>(seurat_clustered_qc, </span>
<span id="cb57-725"><a href="#cb57-725" aria-hidden="true" tabindex="-1"></a>        <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"CCR7"</span>, <span class="st">"SELL"</span>, <span class="st">"CREM"</span>, <span class="st">"CD69"</span>), </span>
<span id="cb57-726"><a href="#cb57-726" aria-hidden="true" tabindex="-1"></a>        <span class="at">idents =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">2</span>, <span class="dv">4</span>, <span class="dv">6</span>)) <span class="sc">+</span> </span>
<span id="cb57-727"><a href="#cb57-727" aria-hidden="true" tabindex="-1"></a>  <span class="fu">RotatedAxis</span>()</span>
<span id="cb57-728"><a href="#cb57-728" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-729"><a href="#cb57-729" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-730"><a href="#cb57-730" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; 可以看到，和</span><span class="in">`FindMarkers`</span><span class="at">找到的top genes的fold change一致，cluster 2中的CREM基因相较于cluster 0, 4, 6高表达，而cluster 2中的CCR7和SELL相对低表达。</span></span>
<span id="cb57-731"><a href="#cb57-731" aria-hidden="true" tabindex="-1"></a><span class="at">&gt;</span></span>
<span id="cb57-732"><a href="#cb57-732" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; Based on these plots it seems as though **clusters 2** are **Activated T cells, cluster 0 and 6 are** Naive or memory CD4+ T cells. However, for **clusters 4** it is hard to tell.</span></span>
<span id="cb57-733"><a href="#cb57-733" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-734"><a href="#cb57-734" aria-hidden="true" tabindex="-1"></a>------------------------------------------------------------------------</span>
<span id="cb57-735"><a href="#cb57-735" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-736"><a href="#cb57-736" aria-hidden="true" tabindex="-1"></a><span class="fu"># 注释细胞cluster</span></span>
<span id="cb57-737"><a href="#cb57-737" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-738"><a href="#cb57-738" aria-hidden="true" tabindex="-1"></a>Now taking all of this information, we can surmise the cell types of the different clusters and plot the cells with cell type labels.</span>
<span id="cb57-739"><a href="#cb57-739" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-740"><a href="#cb57-740" aria-hidden="true" tabindex="-1"></a>| Cluster ID |          Cell Type           |</span>
<span id="cb57-741"><a href="#cb57-741" aria-hidden="true" tabindex="-1"></a>|:----------:|:----------------------------:|</span>
<span id="cb57-742"><a href="#cb57-742" aria-hidden="true" tabindex="-1"></a>|     0      | Naive or memory CD4+ T cells |</span>
<span id="cb57-743"><a href="#cb57-743" aria-hidden="true" tabindex="-1"></a>|     1      |       CD14+ monocytes        |</span>
<span id="cb57-744"><a href="#cb57-744" aria-hidden="true" tabindex="-1"></a>|     2      |      Activated T cells       |</span>
<span id="cb57-745"><a href="#cb57-745" aria-hidden="true" tabindex="-1"></a>|     3      |       CD14+ monocytes        |</span>
<span id="cb57-746"><a href="#cb57-746" aria-hidden="true" tabindex="-1"></a>|     4      |   Stressed cells / Unknown   |</span>
<span id="cb57-747"><a href="#cb57-747" aria-hidden="true" tabindex="-1"></a>|     5      |         CD8+ T cells         |</span>
<span id="cb57-748"><a href="#cb57-748" aria-hidden="true" tabindex="-1"></a>|     6      | Naive or memory CD4+ T cells |</span>
<span id="cb57-749"><a href="#cb57-749" aria-hidden="true" tabindex="-1"></a>|     7      |           B cells            |</span>
<span id="cb57-750"><a href="#cb57-750" aria-hidden="true" tabindex="-1"></a>|     8      |           NK cells           |</span>
<span id="cb57-751"><a href="#cb57-751" aria-hidden="true" tabindex="-1"></a>|     9      |         CD8+ T cells         |</span>
<span id="cb57-752"><a href="#cb57-752" aria-hidden="true" tabindex="-1"></a>|     10     |      FCGR3A+ monocytes       |</span>
<span id="cb57-753"><a href="#cb57-753" aria-hidden="true" tabindex="-1"></a>|     11     |           B cells            |</span>
<span id="cb57-754"><a href="#cb57-754" aria-hidden="true" tabindex="-1"></a>|     12     |           NK cells           |</span>
<span id="cb57-755"><a href="#cb57-755" aria-hidden="true" tabindex="-1"></a>|     13     |           B cells            |</span>
<span id="cb57-756"><a href="#cb57-756" aria-hidden="true" tabindex="-1"></a>|     14     | Conventional dendritic cells |</span>
<span id="cb57-757"><a href="#cb57-757" aria-hidden="true" tabindex="-1"></a>|     15     |        Megakaryocytes        |</span>
<span id="cb57-758"><a href="#cb57-758" aria-hidden="true" tabindex="-1"></a>|     16     | Plasmacytoid dendritic cells |</span>
<span id="cb57-759"><a href="#cb57-759" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-760"><a href="#cb57-760" aria-hidden="true" tabindex="-1"></a>: 细胞注释结果 {#tbl-细胞注释结果}</span>
<span id="cb57-761"><a href="#cb57-761" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-762"><a href="#cb57-762" aria-hidden="true" tabindex="-1"></a>We can then reassign the identity of the clusters to these cell types:</span>
<span id="cb57-763"><a href="#cb57-763" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-766"><a href="#cb57-766" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-767"><a href="#cb57-767" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb57-768"><a href="#cb57-768" aria-hidden="true" tabindex="-1"></a><span class="co"># Rename all identities</span></span>
<span id="cb57-769"><a href="#cb57-769" aria-hidden="true" tabindex="-1"></a>seurat_clustered_qc <span class="ot">&lt;-</span> <span class="fu">RenameIdents</span>(seurat_clustered_qc, </span>
<span id="cb57-770"><a href="#cb57-770" aria-hidden="true" tabindex="-1"></a>                                    <span class="st">"0"</span> <span class="ot">=</span> <span class="st">"Naive or memory CD4+ T cells"</span>,</span>
<span id="cb57-771"><a href="#cb57-771" aria-hidden="true" tabindex="-1"></a>                                    <span class="st">"1"</span> <span class="ot">=</span> <span class="st">"CD14+ monocytes"</span>,</span>
<span id="cb57-772"><a href="#cb57-772" aria-hidden="true" tabindex="-1"></a>                                    <span class="st">"2"</span> <span class="ot">=</span> <span class="st">"Activated T cells"</span>,</span>
<span id="cb57-773"><a href="#cb57-773" aria-hidden="true" tabindex="-1"></a>                                    <span class="st">"3"</span> <span class="ot">=</span> <span class="st">"CD14+ monocytes"</span>,</span>
<span id="cb57-774"><a href="#cb57-774" aria-hidden="true" tabindex="-1"></a>                                    <span class="st">"4"</span> <span class="ot">=</span> <span class="st">"Stressed cells / Unknown"</span>,</span>
<span id="cb57-775"><a href="#cb57-775" aria-hidden="true" tabindex="-1"></a>                                    <span class="st">"5"</span> <span class="ot">=</span> <span class="st">"CD8+ T cells"</span>,</span>
<span id="cb57-776"><a href="#cb57-776" aria-hidden="true" tabindex="-1"></a>                                    <span class="st">"6"</span> <span class="ot">=</span> <span class="st">"Naive or memory CD4+ T cells"</span>,</span>
<span id="cb57-777"><a href="#cb57-777" aria-hidden="true" tabindex="-1"></a>                                    <span class="st">"7"</span> <span class="ot">=</span> <span class="st">"B cells"</span>,</span>
<span id="cb57-778"><a href="#cb57-778" aria-hidden="true" tabindex="-1"></a>                                    <span class="st">"8"</span> <span class="ot">=</span> <span class="st">"NK cells"</span>,</span>
<span id="cb57-779"><a href="#cb57-779" aria-hidden="true" tabindex="-1"></a>                                    <span class="st">"9"</span> <span class="ot">=</span> <span class="st">"CD8+ T cells"</span>,</span>
<span id="cb57-780"><a href="#cb57-780" aria-hidden="true" tabindex="-1"></a>                                    <span class="st">"10"</span> <span class="ot">=</span> <span class="st">"FCGR3A+ monocytes"</span>,</span>
<span id="cb57-781"><a href="#cb57-781" aria-hidden="true" tabindex="-1"></a>                                    <span class="st">"11"</span> <span class="ot">=</span> <span class="st">"B cells"</span>,</span>
<span id="cb57-782"><a href="#cb57-782" aria-hidden="true" tabindex="-1"></a>                                    <span class="st">"12"</span> <span class="ot">=</span> <span class="st">"NK cells"</span>,</span>
<span id="cb57-783"><a href="#cb57-783" aria-hidden="true" tabindex="-1"></a>                                    <span class="st">"13"</span> <span class="ot">=</span> <span class="st">"B cells"</span>,</span>
<span id="cb57-784"><a href="#cb57-784" aria-hidden="true" tabindex="-1"></a>                                    <span class="st">"14"</span> <span class="ot">=</span> <span class="st">"Conventional dendritic cells"</span>,</span>
<span id="cb57-785"><a href="#cb57-785" aria-hidden="true" tabindex="-1"></a>                                    <span class="st">"15"</span> <span class="ot">=</span> <span class="st">"Megakaryocytes"</span>,</span>
<span id="cb57-786"><a href="#cb57-786" aria-hidden="true" tabindex="-1"></a>                                    <span class="st">"16"</span> <span class="ot">=</span> <span class="st">"Plasmacytoid dendritic cells"</span>)</span>
<span id="cb57-787"><a href="#cb57-787" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-788"><a href="#cb57-788" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-791"><a href="#cb57-791" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-792"><a href="#cb57-792" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 10</span></span>
<span id="cb57-793"><a href="#cb57-793" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the UMAP</span></span>
<span id="cb57-794"><a href="#cb57-794" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(seurat_clustered_qc, </span>
<span id="cb57-795"><a href="#cb57-795" aria-hidden="true" tabindex="-1"></a>        <span class="at">reduction =</span> <span class="st">"umap"</span>, </span>
<span id="cb57-796"><a href="#cb57-796" aria-hidden="true" tabindex="-1"></a>        <span class="at">label =</span> <span class="cn">FALSE</span>)</span>
<span id="cb57-797"><a href="#cb57-797" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-798"><a href="#cb57-798" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-799"><a href="#cb57-799" aria-hidden="true" tabindex="-1"></a>::: callout-tip</span>
<span id="cb57-800"><a href="#cb57-800" aria-hidden="true" tabindex="-1"></a>If we wanted to remove the potentially stressed cells, we could use the <span class="in">`subset()`</span> function:</span>
<span id="cb57-801"><a href="#cb57-801" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-804"><a href="#cb57-804" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-805"><a href="#cb57-805" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb57-806"><a href="#cb57-806" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove the stressed or dying cells</span></span>
<span id="cb57-807"><a href="#cb57-807" aria-hidden="true" tabindex="-1"></a>seurat_subset_labeled <span class="ot">&lt;-</span> <span class="fu">subset</span>(seurat_clustered_qc,</span>
<span id="cb57-808"><a href="#cb57-808" aria-hidden="true" tabindex="-1"></a>                               <span class="at">idents =</span> <span class="st">"Stressed cells / Unknown"</span>, <span class="co"># 取子集的标准</span></span>
<span id="cb57-809"><a href="#cb57-809" aria-hidden="true" tabindex="-1"></a>                               <span class="at">invert =</span> <span class="cn">TRUE</span>) <span class="co"># 反选</span></span>
<span id="cb57-810"><a href="#cb57-810" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-811"><a href="#cb57-811" aria-hidden="true" tabindex="-1"></a><span class="co"># Re-visualize the clusters</span></span>
<span id="cb57-812"><a href="#cb57-812" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(seurat_subset_labeled, </span>
<span id="cb57-813"><a href="#cb57-813" aria-hidden="true" tabindex="-1"></a>        <span class="at">reduction =</span> <span class="st">"umap"</span>, </span>
<span id="cb57-814"><a href="#cb57-814" aria-hidden="true" tabindex="-1"></a>        <span class="at">label =</span> <span class="cn">FALSE</span>)</span>
<span id="cb57-815"><a href="#cb57-815" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-816"><a href="#cb57-816" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb57-817"><a href="#cb57-817" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-818"><a href="#cb57-818" aria-hidden="true" tabindex="-1"></a>Now we would want to save our final labelled Seurat object:</span>
<span id="cb57-819"><a href="#cb57-819" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-822"><a href="#cb57-822" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-823"><a href="#cb57-823" aria-hidden="true" tabindex="-1"></a><span class="co"># Save final R object</span></span>
<span id="cb57-824"><a href="#cb57-824" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(seurat_clustered_qc, <span class="at">file =</span> <span class="st">"output/scRNA-seq_online/seurat_labelled.rds"</span>)</span>
<span id="cb57-825"><a href="#cb57-825" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-826"><a href="#cb57-826" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-827"><a href="#cb57-827" aria-hidden="true" tabindex="-1"></a>最后，我们把sessionInfo也导出来：</span>
<span id="cb57-828"><a href="#cb57-828" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-831"><a href="#cb57-831" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-832"><a href="#cb57-832" aria-hidden="true" tabindex="-1"></a><span class="co">#| output: false</span></span>
<span id="cb57-833"><a href="#cb57-833" aria-hidden="true" tabindex="-1"></a><span class="co"># Create and save a text file with sessionInfo</span></span>
<span id="cb57-834"><a href="#cb57-834" aria-hidden="true" tabindex="-1"></a><span class="fu">sink</span>(<span class="at">file =</span> <span class="st">"output/scRNA-seq_online/sessionInfo_scrnaseq.txt"</span>, </span>
<span id="cb57-835"><a href="#cb57-835" aria-hidden="true" tabindex="-1"></a>     <span class="at">append =</span> <span class="cn">FALSE</span>, </span>
<span id="cb57-836"><a href="#cb57-836" aria-hidden="true" tabindex="-1"></a>     <span class="at">split =</span> <span class="cn">FALSE</span>)</span>
<span id="cb57-837"><a href="#cb57-837" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span>
<span id="cb57-838"><a href="#cb57-838" aria-hidden="true" tabindex="-1"></a><span class="fu">sink</span>()</span>
<span id="cb57-839"><a href="#cb57-839" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-840"><a href="#cb57-840" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-841"><a href="#cb57-841" aria-hidden="true" tabindex="-1"></a>::: {.callout-tip collapse="true"}</span>
<span id="cb57-842"><a href="#cb57-842" aria-hidden="true" tabindex="-1"></a><span class="fu">###### 关于`sink`函数</span></span>
<span id="cb57-843"><a href="#cb57-843" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-844"><a href="#cb57-844" aria-hidden="true" tabindex="-1"></a><span class="in">`sink`</span>函数能够将R脚本运行的结果输出到文本文件中，输出的内容为两个<span class="in">`sink()`</span>命令之间的所有内容。</span>
<span id="cb57-845"><a href="#cb57-845" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-846"><a href="#cb57-846" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`file`</span>：输出目录</span>
<span id="cb57-847"><a href="#cb57-847" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-848"><a href="#cb57-848" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`append`</span>：取<span class="in">`TRUE`</span>表示若输出目录下有与结果文件同名的文件，则计算结果将追加到原文件内容的后面；取<span class="in">`FALSE`</span>（默认）表示将本次的计算结果覆盖原文件的内容</span>
<span id="cb57-849"><a href="#cb57-849" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-850"><a href="#cb57-850" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`split`</span>：取<span class="in">`TRUE`</span>表示在计算结果输出到指定文件中的同时，还输出到控制台上；取<span class="in">`FALSE`</span>（默认）表示计算结果仅输出到指定文件中。</span>
<span id="cb57-851"><a href="#cb57-851" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb57-852"><a href="#cb57-852" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-853"><a href="#cb57-853" aria-hidden="true" tabindex="-1"></a>------------------------------------------------------------------------</span>
<span id="cb57-854"><a href="#cb57-854" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-855"><a href="#cb57-855" aria-hidden="true" tabindex="-1"></a><span class="fu"># 可选的后续分析:</span></span>
<span id="cb57-856"><a href="#cb57-856" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-857"><a href="#cb57-857" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="al">![](images/640.png)</span><span class="co">](https://mp.weixin.qq.com/s?__biz=Mzg2MTExNTkwNA==&amp;mid=2247532151&amp;idx=1&amp;sn=55ff4362b5a572d53072c1fe85c593d6&amp;chksm=ce1e3d18f969b40e71503729f38a8e01cd80e313446da1571d7d42c61e094d94b3519cdb85be&amp;mpshare=1&amp;scene=1&amp;srcid=0910pFfJBQOE8XVDU2opxc6B&amp;sharer_shareinfo=98ce7b038453fa671b5e609905ff82f0&amp;sharer_shareinfo_first=98ce7b038453fa671b5e609905ff82f0#rd)</span></span>
<span id="cb57-858"><a href="#cb57-858" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-859"><a href="#cb57-859" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="al">![](images/640-2.png)</span><span class="co">](https://mp.weixin.qq.com/s?__biz=Mzg2MTExNTkwNA==&amp;mid=2247532151&amp;idx=1&amp;sn=55ff4362b5a572d53072c1fe85c593d6&amp;chksm=ce1e3d18f969b40e71503729f38a8e01cd80e313446da1571d7d42c61e094d94b3519cdb85be&amp;mpshare=1&amp;scene=1&amp;srcid=0910pFfJBQOE8XVDU2opxc6B&amp;sharer_shareinfo=98ce7b038453fa671b5e609905ff82f0&amp;sharer_shareinfo_first=98ce7b038453fa671b5e609905ff82f0#rd)</span></span>
<span id="cb57-860"><a href="#cb57-860" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-861"><a href="#cb57-861" aria-hidden="true" tabindex="-1"></a><span class="fu">## 提取感兴趣细胞亚群进一步细分（Explore a subset of the cell types to discover subclusters of cells）</span></span>
<span id="cb57-862"><a href="#cb57-862" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-863"><a href="#cb57-863" aria-hidden="true" tabindex="-1"></a>鉴定完细胞类型后，接下来就是对鉴定的目标细胞群进行深入的研究。一般是提取目标亚群，然后在进一步细分。细胞再分群的意义在于进一步分析细胞的组成成分。见<span class="co">[</span><span class="ot">多个单细胞数据集整合分析（下）</span><span class="co">](/single_cell/sc_supplementary/integrated_analysis_multiple_single_cell_datasets_2.qmd)</span>。</span>
<span id="cb57-864"><a href="#cb57-864" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-865"><a href="#cb57-865" aria-hidden="true" tabindex="-1"></a><span class="fu">## **单细胞差异分析（**Perform **differential expression analysis** **between conditions）**</span></span>
<span id="cb57-866"><a href="#cb57-866" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-867"><a href="#cb57-867" aria-hidden="true" tabindex="-1"></a>在实际研究中，往往需要研究两类细胞之间的差异，或者是某类细胞在不同处理样本间的差异基因情况，因此需要将待比较的细胞群单独拿出来进行分析。后面，我们在<span class="co">[</span><span class="ot">此前的章节</span><span class="co">](/single_cell/seurat/de_vignette.qmd)</span>中对如何**寻找不同样本类型/条件间同一细胞类型内的差异基因**进行了探索。</span>
<span id="cb57-868"><a href="#cb57-868" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-869"><a href="#cb57-869" aria-hidden="true" tabindex="-1"></a>除了分析细胞内基因表达的差异，也可以采用最常见的GO和KEGG富集分析，还可以进行GSEA、GSVA、SCENIC等分析，从功能、通路、基因调控网络等方面进行探究。</span>
<span id="cb57-870"><a href="#cb57-870" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-871"><a href="#cb57-871" aria-hidden="true" tabindex="-1"></a><span class="fu">## 免疫类群分析</span></span>
<span id="cb57-872"><a href="#cb57-872" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-873"><a href="#cb57-873" aria-hidden="true" tabindex="-1"></a>很多疾病的发生、发展都和免疫细胞有关，一般病灶组织内或者附近的免疫细胞较正常的组织会更丰富。为了将疾病中免疫反应描绘出来，可以采用以下分析策略：①对免疫细胞再分群，得到更多功能更加精细化的免疫细胞亚群；②免疫细胞类群在不同分组中的分布变化；③免疫细胞类群拟时序分析，探究不同疾病程度、不同发育阶段或者不同生理状态下的免疫反应机制。</span>
<span id="cb57-874"><a href="#cb57-874" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-875"><a href="#cb57-875" aria-hidden="true" tabindex="-1"></a><span class="fu">## 关键细胞的分化轨迹</span></span>
<span id="cb57-876"><a href="#cb57-876" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-877"><a href="#cb57-877" aria-hidden="true" tabindex="-1"></a>最开始拿到的单细胞转录组测序数据，并未直接告诉我们每个细胞处在什么状态。因此，需要借助一些分析方法来实现轨迹上的排序，比如Monocle2拟时序分析/轨迹推断，ScVelo RNA速度分析等，推断潜在的细胞分化方向性，挖掘一些稀少的中间状态细胞，解析细胞分化过程中起调控作用的关键基因；也可以比较实验组/对照组中分化的差异。</span>
<span id="cb57-878"><a href="#cb57-878" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-879"><a href="#cb57-879" aria-hidden="true" tabindex="-1"></a><span class="fu">## 探索细胞-细胞互作</span></span>
<span id="cb57-880"><a href="#cb57-880" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-881"><a href="#cb57-881" aria-hidden="true" tabindex="-1"></a>每个细胞都能分泌细胞因子或者激素，这些细胞因子能够被周围细胞上的受体接收并用于调节相应的生理活动，细胞与细胞之间有相互联系，致病细胞可能是由于大类中某一种亚类细胞分化或者演化过来的；那么这些演化过程可能是由于其他细胞的细胞因子导致的。</span>
<span id="cb57-882"><a href="#cb57-882" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-883"><a href="#cb57-883" aria-hidden="true" tabindex="-1"></a>在锁定目标细胞类群之后，通过CellphoneDB细胞通讯分析、受体-配体分析等，找到与目标细胞相互作用的其他细胞类型，找到“直接”和“间接”的细胞调控网络。当然，对于已经筛选出目标基因了，也可以通过PPI分析找到基因/蛋白间的互作关系。</span>
<span id="cb57-884"><a href="#cb57-884" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-885"><a href="#cb57-885" aria-hidden="true" tabindex="-1"></a><span class="fu">## 挖掘关键调控转录因子</span></span>
<span id="cb57-886"><a href="#cb57-886" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-887"><a href="#cb57-887" aria-hidden="true" tabindex="-1"></a>细胞异质性以及这种异质性是如何发展和维持的，在很大程度上是由潜在的基因调控网络决定的，特定转录因子（transcription factor，TF）集合的协同表达驱动各自靶标基因的表达，从而建立特定的基因表达谱；SCENIC是用来研究和破译基因调控的工具，能从单细胞转录组数据中推断TF、基因调控网络和细胞类型。其基本原理是基于共表达和DNA调控保守序列（motif）分析推断基因调控网络，然后在每个细胞中分析网络活性以鉴定细胞状态。</span>
<span id="cb57-888"><a href="#cb57-888" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-889"><a href="#cb57-889" aria-hidden="true" tabindex="-1"></a><span class="fu">## 湿实验验证，探索潜在机制（Experimentally validate intriguing markers for our identified cell types）</span></span>
<span id="cb57-890"><a href="#cb57-890" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-891"><a href="#cb57-891" aria-hidden="true" tabindex="-1"></a>除了以上单细胞测序数据相关分析以外，湿实验验证已逐渐成为单细胞分析的有力补充。比如用RT-qPCR/FISH等在RNA层面进行验证，用WB/免疫荧光/免疫组化/流式分析等在蛋白层面进行验证，还可以利用组织芯片或者TCGA数据库信息，进行临床水平的验证；也可以用细胞谱系示踪技术标记细胞 , 对关键细胞类群及其后代所有细胞的增殖、分化和迁移等活动进行追踪观察等；甚至可以利用流式分选出感兴趣的细胞群，然后对目的细胞群进行基因操作（上调/下调/敲除等），进行更深入的机制研究。总之，根据具体的实验目的，进行更完善的实验设计。</span>
<span id="cb57-892"><a href="#cb57-892" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-893"><a href="#cb57-893" aria-hidden="true" tabindex="-1"></a>------------------------------------------------------------------------</span>
<span id="cb57-894"><a href="#cb57-894" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-895"><a href="#cb57-895" aria-hidden="true" tabindex="-1"></a>::: {.callout-note collapse="true" icon="false"}</span>
<span id="cb57-896"><a href="#cb57-896" aria-hidden="true" tabindex="-1"></a><span class="fu">## Session Info</span></span>
<span id="cb57-897"><a href="#cb57-897" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-900"><a href="#cb57-900" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-901"><a href="#cb57-901" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb57-902"><a href="#cb57-902" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span>
<span id="cb57-903"><a href="#cb57-903" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-904"><a href="#cb57-904" aria-hidden="true" tabindex="-1"></a>:::</span>
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">
      <ul class="footer-items list-unstyled">
<li class="nav-item">
 This website was built with Quarto
  </li>  
</ul>
</div>   
    <div class="nav-footer-center">Copyright 2024, Du Junhong</div>
    <div class="nav-footer-right">
      <ul class="footer-items list-unstyled">
<li class="nav-item compact">
    <a class="nav-link" href="https://github.com/djhcod">
      <i class="bi bi-github" role="img">
</i> 
    </a>
  </li>  
</ul>
</div>
  </div>
</footer>


</body></html>