<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.433">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>R语言学习笔记 - 14&nbsp; 差异表达分析</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../single_cell/scRNA-seq_online/00_intro.html" rel="next">
<link href="../../single_cell/seurat/cell_cycle_regression.html" rel="prev">
<link href="../../images/logo.jpg" rel="icon" type="image/jpeg">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark"><script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><script type="application/json" class="js-hypothesis-config">
{
  "theme": "classic",
  "openSidebar": false,
  "showHighlights": "always",
  "enableExperimentalNewNoteButton": true
}
</script><script async="" src="https://hypothes.is/embed.js"></script>
<meta name="twitter:title" content="R语言学习笔记 - 14&nbsp; 差异表达分析">
<meta name="twitter:description" content="">
<meta name="twitter:image" content="https://djhcod.github.io/r-notes/single_cell/seurat/de_vignette_files/figure-html/unnamed-chunk-17-1.png">
<meta name="twitter:image-height" content="960">
<meta name="twitter:image-width" content="1344">
<meta name="twitter:card" content="summary_large_image">
</head>
<body class="nav-sidebar docked nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="navbar navbar-expand-lg navbar-dark "><div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a href="../../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../../images/logo.jpg" alt="" class="navbar-logo"></a>
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">R语言学习笔记</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
<li class="nav-item">
    <a class="nav-link active" href="../../index.html" rel="" target="" aria-current="page">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../intro.html" rel="" target="">
 <span class="menu-text">Introduction</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-parts" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Parts</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-parts">
<li>
    <a class="dropdown-item" href="../../r_basic/r_basics.html" rel="" target="">
 <span class="dropdown-text">R语言基础</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../single_cell/seurat/seurat.html" rel="" target="">
 <span class="dropdown-text">Seurat v5官方文档</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../single_cell/scRNA-seq_online/00_intro.html" rel="" target="">
 <span class="dropdown-text">scRNA-seq_online学习材料</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../quarto_foundation/quarto_foundation.html" rel="" target="">
 <span class="dropdown-text">Quarto基础</span></a>
  </li>  
    </ul>
</li>
  <li class="nav-item">
    <a class="nav-link" href="../../references.html" rel="" target="">
 <span class="menu-text">References</span></a>
  </li>  
</ul>
<ul class="navbar-nav navbar-nav-scroll ms-auto">
<li class="nav-item">
    <a class="nav-link" href="https://bookdown.org" rel="" target=""><i class="bi bi-book-fill" role="img">
</i> 
 <span class="menu-text">Bookdown</span></a>
  </li>  
</ul>
<div class="quarto-navbar-tools tools-wide">
    <a href="https://github.com/djhcod/r-notes" rel="" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
    <div class="dropdown">
      <a href="" title="Share" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Share"><i class="bi bi-share"></i></a>
      <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="quarto-navigation-tool-dropdown-0">
<li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://twitter.com/intent/tweet?url=%7Curl%7C">
              <i class="bi bi-bi-twitter pe-1"></i>
            Twitter
            </a>
          </li>
          <li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://www.facebook.com/sharer/sharer.php?u=%7Curl%7C">
              <i class="bi bi-bi-facebook pe-1"></i>
            Facebook
            </a>
          </li>
      </ul>
</div>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../single_cell/seurat/seurat.html">Seurat v5官方文档</a></li><li class="breadcrumb-item"><a href="../../single_cell/seurat/de_vignette.html"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">差异表达分析</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto"><div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">主页</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">前言</span></a>
  </div>
</li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../r_basic/r_basics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">R语言基础</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/environment_configuration.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Rstudio环境配置</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/data_input_output.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">数据的读取与输出</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/basic_data_function.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">数据处理基本函数</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/character.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">字符的处理</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../single_cell/seurat/seurat.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Seurat v5官方文档</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/seurat_command_list.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Seurat常用函数清单</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/seurat_tutorial.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Seurat细胞分群官方教程</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/data_visualization_methods_in_seurat.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Seurat中的数据可视化方法</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/sctransform.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">基于SCTransform的单细胞数据标准化</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/integration.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">单细胞数据整合（integration）</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/integrative_analysis_in_seurat_v5.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Seurat v5单细胞数据整合分析</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/mapping_and_annotating_query_datasets.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">基于参考集的细胞注释</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/cell_cycle_regression.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">消除细胞周期的影响</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/de_vignette.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">差异表达分析</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../single_cell/scRNA-seq_online/00_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">scRNA-seq_online学习材料</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/01_intro_to_scRNA-seq.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">单细胞测序技术介绍</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/02_SC_generation_of_count_matrix.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">单细胞RNA测序—从raw data到count matrix</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/03_SC_quality_control-setup.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">数据导入与Seurat对象构建</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/04_SC_quality_control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title"><strong>质控</strong></span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../quarto_foundation/quarto_foundation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Quarto基础</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/yaml_settings.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">YAML设置</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/images_settings.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">图片设置</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/cross_references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">21</span>&nbsp; <span class="chapter-title">交叉引用</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/insert_other_content.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">22</span>&nbsp; <span class="chapter-title">插入其他内容</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/quarto_books.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">23</span>&nbsp; <span class="chapter-title">Quarto Books</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/github_pages.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">24</span>&nbsp; <span class="chapter-title">发布到GitHub Pages</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="false">
 <span class="menu-text">参考文献</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active" data-toc-expanded="1"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#%E6%95%B0%E6%8D%AE%E8%AF%BB%E5%8F%96%E5%92%8C%E9%A2%84%E5%A4%84%E7%90%86" id="toc-数据读取和预处理" class="nav-link active" data-scroll-target="#%E6%95%B0%E6%8D%AE%E8%AF%BB%E5%8F%96%E5%92%8C%E9%A2%84%E5%A4%84%E7%90%86"><span class="header-section-number">14.1</span> 数据读取和预处理</a></li>
  <li><a href="#%E5%AF%BB%E6%89%BE%E7%BB%86%E8%83%9E%E7%B1%BB%E5%9E%8B%E4%B9%8B%E9%97%B4%E7%9A%84%E5%B7%AE%E5%BC%82%E5%9F%BA%E5%9B%A0" id="toc-寻找细胞类型之间的差异基因" class="nav-link" data-scroll-target="#%E5%AF%BB%E6%89%BE%E7%BB%86%E8%83%9E%E7%B1%BB%E5%9E%8B%E4%B9%8B%E9%97%B4%E7%9A%84%E5%B7%AE%E5%BC%82%E5%9F%BA%E5%9B%A0"><span class="header-section-number">14.2</span> 寻找细胞类型之间的差异基因</a></li>
  <li><a href="#sec-%E5%AF%BB%E6%89%BE%E4%B8%8D%E5%90%8C%E6%A0%B7%E6%9C%AC%E7%B1%BB%E5%9E%8B%E9%97%B4%E5%90%8C%E4%B8%80%E7%BB%86%E8%83%9E%E7%B1%BB%E5%9E%8B%E5%86%85%E7%9A%84%E5%B7%AE%E5%BC%82%E5%9F%BA%E5%9B%A0" id="toc-sec-寻找不同样本类型间同一细胞类型内的差异基因" class="nav-link" data-scroll-target="#sec-%E5%AF%BB%E6%89%BE%E4%B8%8D%E5%90%8C%E6%A0%B7%E6%9C%AC%E7%B1%BB%E5%9E%8B%E9%97%B4%E5%90%8C%E4%B8%80%E7%BB%86%E8%83%9E%E7%B1%BB%E5%9E%8B%E5%86%85%E7%9A%84%E5%B7%AE%E5%BC%82%E5%9F%BA%E5%9B%A0"><span class="header-section-number">14.3</span> 寻找不同样本类型间同一细胞类型内的差异基因</a></li>
  <li>
<a href="#pseudobulking%E5%90%8E%E7%9A%84%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90" id="toc-pseudobulking后的差异分析" class="nav-link" data-scroll-target="#pseudobulking%E5%90%8E%E7%9A%84%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90"><span class="header-section-number">14.4</span> pseudobulking后的差异分析</a>
  <ul class="collapse">
<li><a href="#%E5%87%86%E5%A4%87%E6%A0%B7%E6%9C%AC%E4%BF%A1%E6%81%AF" id="toc-准备样本信息" class="nav-link" data-scroll-target="#%E5%87%86%E5%A4%87%E6%A0%B7%E6%9C%AC%E4%BF%A1%E6%81%AF">准备样本信息</a></li>
  <li><a href="#%E6%89%A7%E8%A1%8Cpseudobulking" id="toc-执行pseudobulking" class="nav-link" data-scroll-target="#%E6%89%A7%E8%A1%8Cpseudobulking">执行pseudobulking</a></li>
  <li><a href="#%E6%89%A7%E8%A1%8C%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90" id="toc-执行差异分析" class="nav-link" data-scroll-target="#%E6%89%A7%E8%A1%8C%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90">执行差异分析</a></li>
  <li><a href="#%E6%AF%94%E8%BE%83%E5%8D%95%E7%BB%86%E8%83%9E%E6%B0%B4%E5%B9%B3%E5%92%8Cpseudobulk%E6%B0%B4%E5%B9%B3%E7%9A%84%E5%B7%AE%E5%BC%82%E8%A1%A8%E8%BE%BE%E5%88%86%E6%9E%90" id="toc-比较单细胞水平和pseudobulk水平的差异表达分析" class="nav-link" data-scroll-target="#%E6%AF%94%E8%BE%83%E5%8D%95%E7%BB%86%E8%83%9E%E6%B0%B4%E5%B9%B3%E5%92%8Cpseudobulk%E6%B0%B4%E5%B9%B3%E7%9A%84%E5%B7%AE%E5%BC%82%E8%A1%A8%E8%BE%BE%E5%88%86%E6%9E%90">比较单细胞水平和pseudobulk水平的差异表达分析</a></li>
  </ul>
</li>
  </ul><div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/djhcod/r-notes/blob/main/single_cell/seurat/de_vignette.qmd" class="toc-action">View source</a></p><p><a href="https://github.com/djhcod/r-notes/issues/new" class="toc-action">Report an issue</a></p><p><a href="https://github.com/djhcod/r-notes/edit/main/single_cell/seurat/de_vignette.qmd" class="toc-action">Edit this page</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title"><span id="sec-差异表达分析" class="quarto-section-identifier"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">差异表达分析</span></span></h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><p>原文：<a href="https://satijalab.org/seurat/articles/de_vignette"><em>Differential expression testing</em></a></p>
<p>原文发布日期：2023年10月31日</p>
<section id="数据读取和预处理" class="level2" data-number="14.1"><h2 data-number="14.1" class="anchored" data-anchor-id="数据读取和预处理">
<span class="header-section-number">14.1</span> 数据读取和预处理</h2>
<p>This vignette highlights some example workflows for performing differential expression in Seurat. For demonstration purposes, we will be using the interferon-beta stimulated human PBMCs dataset <span class="citation" data-cites="kang2017">(<a href="../../references.html#ref-kang2017" role="doc-biblioref">Kang et al. 2017</a>)</span> that is available via the&nbsp;SeuratData&nbsp;package.</p>
<div class="callout callout-style-simple callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
在线读取（可能需要全局代理）
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-hash="de_vignette_cache/html/unnamed-chunk-1_15605a9f02196faa4d2d3c840ad28b02">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://www.satijalab.org/seurat">SeuratData</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/InstallData.html">InstallData</a></span><span class="op">(</span><span class="st">"ifnb"</span><span class="op">)</span></span>
<span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/LoadData.html">LoadData</a></span><span class="op">(</span><span class="st">"ifnb"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<p>从本地下载好的数据读取：</p>
<div class="cell" data-hash="de_vignette_cache/html/unnamed-chunk-2_9cf7bb8d5faca24a0733bb0b04fef1f0">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span></span>
<span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="st">"data/pbmc_ifnb.rds"</span><span class="op">)</span></span>
<span><span class="va">ifnb</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
14053 features across 13999 samples within 1 assay 
Active assay: RNA (14053 features, 0 variable features)
 2 layers present: counts, data</code></pre>
</div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">ifnb</span><span class="op">@</span><span class="va">meta.data</span>, <span class="fl">5</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                  orig.ident nCount_RNA nFeature_RNA stim seurat_annotations
AAACATACATTTCC.1 IMMUNE_CTRL       3017          877 CTRL          CD14 Mono
AAACATACCAGAAA.1 IMMUNE_CTRL       2481          713 CTRL          CD14 Mono
AAACATACCTCGCT.1 IMMUNE_CTRL       3420          850 CTRL          CD14 Mono
AAACATACCTGGTA.1 IMMUNE_CTRL       3156         1109 CTRL                pDC
AAACATACGATGAA.1 IMMUNE_CTRL       1868          634 CTRL       CD4 Memory T</code></pre>
</div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">ifnb</span><span class="op">$</span><span class="va">seurat_annotations</span><span class="op">)</span> <span class="co"># 这里的数据已经提前注释好了细胞类型</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
   CD14 Mono  CD4 Naive T CD4 Memory T    CD16 Mono            B        CD8 T 
        4362         2504         1762         1044          978          814 
 T activated           NK           DC  B Activated           Mk          pDC 
         633          619          472          388          236          132 
       Eryth 
          55 </code></pre>
</div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 标准化</span></span>
<span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html">NormalizeData</a></span><span class="op">(</span><span class="va">ifnb</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="寻找细胞类型之间的差异基因" class="level2" data-number="14.2"><h2 data-number="14.2" class="anchored" data-anchor-id="寻找细胞类型之间的差异基因">
<span class="header-section-number">14.2</span> 寻找细胞类型之间的差异基因</h2>
<div class="callout callout-style-default callout-caution callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Caution
</div>
</div>
<div class="callout-body-container callout-body">
<p>For a <strong>much</strong> faster implementation of the Wilcoxon Rank Sum Test,(default method for <code>FindMarkers</code>) please install the <a href="https://github.com/immunogenomics/presto"><code>presto</code></a> package:</p>
<div class="cell" data-hash="de_vignette_cache/html/unnamed-chunk-3_695f538ab197575f0aec39a44495efd0">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html">install_github</a></span><span class="op">(</span><span class="st">'immunogenomics/presto'</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<p>Find DE features between CD16 Mono and CD1 Mono：</p>
<div class="cell" data-hash="de_vignette_cache/html/unnamed-chunk-4_ed910db548a0df4f356b3ac071137d0f">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html">Idents</a></span><span class="op">(</span><span class="va">ifnb</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"seurat_annotations"</span></span>
<span><span class="va">monocyte.de.markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindMarkers.html">FindMarkers</a></span><span class="op">(</span><span class="va">ifnb</span>, </span>
<span>                                   ident.1 <span class="op">=</span> <span class="st">"CD16 Mono"</span>, </span>
<span>                                   ident.2 <span class="op">=</span> <span class="st">"CD14 Mono"</span><span class="op">)</span></span>
<span><span class="co"># view results</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">monocyte.de.markers</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 6835</code></pre>
</div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">monocyte.de.markers</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       p_val avg_log2FC pct.1 pct.2 p_val_adj
VMO1       0   5.700274 0.778 0.084         0
MS4A4A     0   3.349751 0.748 0.143         0
FCGR3A     0   3.281942 0.982 0.418         0
PLAC8      0   3.268470 0.636 0.124         0
CXCL16     0   2.014758 0.938 0.475         0
MS4A7      0   2.386436 0.978 0.558         0</code></pre>
</div>
</div>
<p>The results data frame has the following columns :</p>
<ul>
<li><p><code>p_val</code> : p-value (unadjusted)</p></li>
<li><p><code>avg_log2FC</code> : log fold-change of the average expression between the two groups. Positive values indicate that the feature is more highly expressed in the first group.</p></li>
<li><p><code>pct.1</code> : The percentage of cells where the feature is detected in the first group</p></li>
<li><p><code>pct.2</code> : The percentage of cells where the feature is detected in the second group</p></li>
<li><p><code>p_val_adj</code> : Adjusted p-value, based on <strong>Bonferroni correction</strong> using all features in the dataset.</p></li>
</ul>
<p>If the&nbsp;<code>ident.2</code>&nbsp;parameter is omitted or set to&nbsp;<code>NULL</code>,&nbsp;<a href="https://satijalab.org/seurat/reference/findmarkers"><code>FindMarkers()</code></a>&nbsp;will test for differentially expressed features <strong>between the group specified by&nbsp;<code>ident.1</code>&nbsp;and all other cells</strong>. Additionally, the parameter&nbsp;<code>only.pos</code>&nbsp;can be set to&nbsp;<code>TRUE</code>&nbsp;to only search for positive markers, i.e.&nbsp;features that are more highly expressed in the&nbsp;<code>ident.1</code>&nbsp;group.</p>
<div class="cell" data-hash="de_vignette_cache/html/unnamed-chunk-5_992141784cfbe5a3d569db1f70642b2f">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">monocyte.de.markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindMarkers.html">FindMarkers</a></span><span class="op">(</span><span class="va">ifnb</span>, </span>
<span>                                   ident.1 <span class="op">=</span> <span class="st">"CD16 Mono"</span>, </span>
<span>                                   ident.2 <span class="op">=</span> <span class="cn">NULL</span>, </span>
<span>                                   only.pos <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">monocyte.de.markers</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2354</code></pre>
</div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">monocyte.de.markers</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       p_val avg_log2FC pct.1 pct.2 p_val_adj
FCGR3A     0   4.532656 0.982 0.168         0
MS4A7      0   3.806350 0.978 0.216         0
CXCL16     0   3.274267 0.938 0.196         0
VMO1       0   6.254651 0.778 0.044         0
MS4A4A     0   4.747731 0.748 0.055         0
LST1       0   2.927351 0.912 0.228         0</code></pre>
</div>
</div>
</section><section id="sec-寻找不同样本类型间同一细胞类型内的差异基因" class="level2" data-number="14.3"><h2 data-number="14.3" class="anchored" data-anchor-id="sec-寻找不同样本类型间同一细胞类型内的差异基因">
<span class="header-section-number">14.3</span> 寻找不同样本类型间同一细胞类型内的差异基因</h2>
<p>Since this dataset contains treatment information (control versus stimulated with interferon-beta), we can also ask <strong>what genes change in different conditions for cells of the same type</strong>.</p>
<ul>
<li>First, we create a column in the&nbsp;<code>meta.data</code>&nbsp;slot to hold both the cell type and treatment information and switch the current&nbsp;<code>Idents</code>&nbsp;to that column.</li>
</ul>
<div class="cell" data-hash="de_vignette_cache/html/unnamed-chunk-6_f01fe38db30a7a30eaeeaa6eee636e84">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">ifnb</span><span class="op">$</span><span class="va">celltype.stim</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="va">ifnb</span><span class="op">$</span><span class="va">seurat_annotations</span>, <span class="va">ifnb</span><span class="op">$</span><span class="va">stim</span>, sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">ifnb</span><span class="op">$</span><span class="va">celltype.stim</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
 B Activated_CTRL  B Activated_STIM            B_CTRL            B_STIM 
              185               203               407               571 
   CD14 Mono_CTRL    CD14 Mono_STIM    CD16 Mono_CTRL    CD16 Mono_STIM 
             2215              2147               507               537 
CD4 Memory T_CTRL CD4 Memory T_STIM  CD4 Naive T_CTRL  CD4 Naive T_STIM 
              859               903               978              1526 
       CD8 T_CTRL        CD8 T_STIM           DC_CTRL           DC_STIM 
              352               462               258               214 
       Eryth_CTRL        Eryth_STIM           Mk_CTRL           Mk_STIM 
               23                32               115               121 
          NK_CTRL           NK_STIM          pDC_CTRL          pDC_STIM 
              298               321                51                81 
 T activated_CTRL  T activated_STIM 
              300               333 </code></pre>
</div>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html">Idents</a></span><span class="op">(</span><span class="va">ifnb</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"celltype.stim"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>Then we use&nbsp;<code><a href="https://satijalab.org/seurat/reference/FindMarkers.html">FindMarkers()</a></code>&nbsp;to find the genes that are different between control and stimulated CD14 monocytes.</li>
</ul>
<div class="cell" data-hash="de_vignette_cache/html/unnamed-chunk-7_f4bd60fba7621f60d9fe64c02df9ddb7">
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">mono.de</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindMarkers.html">FindMarkers</a></span><span class="op">(</span><span class="va">ifnb</span>, </span>
<span>                       ident.1 <span class="op">=</span> <span class="st">"CD14 Mono_STIM"</span>, </span>
<span>                       ident.2 <span class="op">=</span> <span class="st">"CD14 Mono_CTRL"</span>, </span>
<span>                       verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">mono.de</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 6956</code></pre>
</div>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">mono.de</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        p_val avg_log2FC pct.1 pct.2 p_val_adj
IFIT1       0   7.319139 0.985 0.033         0
CXCL10      0   8.036564 0.984 0.035         0
RSAD2       0   6.741673 0.988 0.045         0
TNFSF10     0   6.991279 0.989 0.047         0
IFIT3       0   6.883785 0.992 0.056         0
IFIT2       0   7.179929 0.961 0.039         0
CXCL11      0   8.624208 0.932 0.012         0
CCL8        0   9.134191 0.918 0.017         0
IDO1        0   5.455898 0.965 0.089         0
MX1         0   5.059052 0.960 0.093         0</code></pre>
</div>
</div>
<p>However, the p-values obtained from this analysis should be interpreted with caution, because <strong>these tests treat each cell as an independent replicate and ignore inherent correlations between cells originating from the same sample</strong>. Such analyses have been shown to find a large number of <strong>false positive associations</strong>, as has been demonstrated by&nbsp;<span class="citation" data-cites="squair2021">(<a href="../../references.html#ref-squair2021" role="doc-biblioref">Squair et al. 2021</a>)</span>,&nbsp;<span class="citation" data-cites="zimmerman2021">(<a href="../../references.html#ref-zimmerman2021" role="doc-biblioref">Zimmerman, Espeland, and Langefeld 2021</a>)</span>,&nbsp;<span class="citation" data-cites="junttila2022">(<a href="../../references.html#ref-junttila2022" role="doc-biblioref">Junttila, Smolander, and Elo 2022</a>)</span>, and others. Below, we show how <strong>pseudobulking</strong> can be used to account for such within-sample correlation.</p>
</section><section id="pseudobulking后的差异分析" class="level2" data-number="14.4"><h2 data-number="14.4" class="anchored" data-anchor-id="pseudobulking后的差异分析">
<span class="header-section-number">14.4</span> pseudobulking后的差异分析</h2>
<p>To pseudobulk, we will use&nbsp;<code><a href="https://satijalab.org/seurat/reference/AggregateExpression.html">AggregateExpression()</a></code>&nbsp;to sum together <strong>gene counts of all the cells from the same sample</strong> for each cell type. This results in <strong>one gene expression profile per sample</strong> <strong>and cell type</strong>. We can then perform DE analysis using&nbsp;<a href="https://bioconductor.org/packages/release/bioc/html/DESeq2.html"><code>DESeq2</code></a>&nbsp;<strong>on the sample level</strong>. This treats the samples, rather than the individual cells, as independent observations. 参考前面的 <a href="#sec-%E8%AF%86%E5%88%AB%E4%B8%8D%E5%90%8C%E6%A0%B7%E6%9C%AC%E7%B1%BB%E5%9E%8B%E9%97%B4%E7%9A%84%E5%B7%AE%E5%BC%82%E5%9F%BA%E5%9B%A0"><span>Section&nbsp;10.5</span></a> 。</p>
<section id="准备样本信息" class="level3"><h3 class="anchored" data-anchor-id="准备样本信息">准备样本信息</h3>
<p>First, we need to retrieve the sample information for each cell. This is not loaded in the metadata, so we will load it from the&nbsp;<a href="https://github.com/yelabucsf/demuxlet_paper_code/tree/master/">Github repo</a>&nbsp;of the source data for the original paper.</p>
<div class="callout callout-style-simple callout-tip no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-3-contents" aria-controls="callout-3" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Add sample information to the dataset
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-3" class="callout-3-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-hash="de_vignette_cache/html/unnamed-chunk-8_aa781827929a6bc0ecac931f9ccecf36">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a><span class="co"># 从GitHub仓库读取（可能需要代理）</span></span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a><span class="co"># load the inferred sample IDs of each cell</span></span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>ctrl <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="fu">url</span>(<span class="st">"https://raw.githubusercontent.com/yelabucsf/demuxlet_paper_code/master/fig3/ye1.ctrl.8.10.sm.best"</span>), <span class="at">head =</span> T, <span class="at">stringsAsFactors =</span> F)</span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>stim <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="fu">url</span>(<span class="st">"https://raw.githubusercontent.com/yelabucsf/demuxlet_paper_code/master/fig3/ye2.stim.8.10.sm.best"</span>), <span class="at">head =</span> T, <span class="at">stringsAsFactors =</span> F)</span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-hash="de_vignette_cache/html/unnamed-chunk-9_f5873510b5883a5f90c5c23b033121cd">
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 这里提前下载好了两个样本信息文件，所以直接从本地读取</span></span>
<span><span class="va">ctrl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="st">"data/inferred_sample_ids_ctrl.rds"</span><span class="op">)</span></span>
<span><span class="va">stim</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="st">"data/inferred_sample_ids_stim.rds"</span><span class="op">)</span></span>
<span><span class="va">info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span><span class="op">(</span><span class="va">ctrl</span>, <span class="va">stim</span><span class="op">)</span></span>
<span><span class="va">info</span><span class="op">$</span><span class="va">BARCODE</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "AAACATACAATGCC-1" "AAACATACATTTCC-1" "AAACATACCAGAAA-1" "AAACATACCAGCTA-1"
[5] "AAACATACCATGCA-1"</code></pre>
</div>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">ifnb</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "AAACATACATTTCC.1" "AAACATACCAGAAA.1" "AAACATACCTCGCT.1" "AAACATACCTGGTA.1"
[5] "AAACATACGATGAA.1"</code></pre>
</div>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 可以看到两者的barcode形式不一致</span></span>
<span><span class="co"># rename the cell IDs by substituting the '-' into '.'</span></span>
<span><span class="va">info</span><span class="op">$</span><span class="va">BARCODE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"\\-"</span>, replacement <span class="op">=</span> <span class="st">"\\."</span>, <span class="va">info</span><span class="op">$</span><span class="va">BARCODE</span><span class="op">)</span></span>
<span><span class="va">info</span><span class="op">$</span><span class="va">BARCODE</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "AAACATACAATGCC.1" "AAACATACATTTCC.1" "AAACATACCAGAAA.1" "AAACATACCAGCTA.1"
[5] "AAACATACCATGCA.1"</code></pre>
</div>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># only keep the cells with high-confidence sample ID</span></span>
<span><span class="va">info</span> <span class="op">&lt;-</span> <span class="va">info</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"SNG"</span>, x <span class="op">=</span> <span class="va">info</span><span class="op">$</span><span class="va">BEST</span><span class="op">)</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># remove cells with duplicated IDs in both ctrl and stim groups</span></span>
<span><span class="va">info</span> <span class="op">&lt;-</span> <span class="va">info</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/duplicated.html">duplicated</a></span><span class="op">(</span><span class="va">info</span><span class="op">$</span><span class="va">BARCODE</span><span class="op">)</span> <span class="op">&amp;</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/duplicated.html">duplicated</a></span><span class="op">(</span><span class="va">info</span><span class="op">$</span><span class="va">BARCODE</span>, fromLast <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># now add the sample IDs to ifnb </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">info</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">info</span><span class="op">$</span><span class="va">BARCODE</span></span>
<span><span class="va">info</span> <span class="op">&lt;-</span> <span class="va">info</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"BEST"</span><span class="op">)</span>, drop <span class="op">=</span> <span class="cn">F</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">info</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"donor_id"</span><span class="op">)</span></span>
<span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/AddMetaData.html">AddMetaData</a></span><span class="op">(</span><span class="va">ifnb</span>, metadata <span class="op">=</span> <span class="va">info</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># remove cells without donor IDs</span></span>
<span><span class="va">ifnb</span><span class="op">$</span><span class="va">donor_id</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">ifnb</span><span class="op">$</span><span class="va">donor_id</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"unknown"</span></span>
<span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="va">ifnb</span>, subset <span class="op">=</span> <span class="va">donor_id</span> <span class="op">!=</span> <span class="st">"unknown"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<p>可以看到，现在的meta.dat中多了样本信息列（<code>donor_id</code>），记录了每个细胞来自哪个患者：</p>
<div class="cell" data-hash="de_vignette_cache/html/unnamed-chunk-10_38653ae6984511bb9c3169c3dc1d369b">
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">ifnb</span><span class="op">@</span><span class="va">meta.data</span>, <span class="fl">5</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                  orig.ident nCount_RNA nFeature_RNA stim seurat_annotations
AAACATACATTTCC.1 IMMUNE_CTRL       3017          877 CTRL          CD14 Mono
AAACATACCAGAAA.1 IMMUNE_CTRL       2481          713 CTRL          CD14 Mono
AAACATACCTCGCT.1 IMMUNE_CTRL       3420          850 CTRL          CD14 Mono
AAACATACCTGGTA.1 IMMUNE_CTRL       3156         1109 CTRL                pDC
AAACATACGATGAA.1 IMMUNE_CTRL       1868          634 CTRL       CD4 Memory T
                     celltype.stim donor_id
AAACATACATTTCC.1    CD14 Mono_CTRL SNG-1016
AAACATACCAGAAA.1    CD14 Mono_CTRL SNG-1256
AAACATACCTCGCT.1    CD14 Mono_CTRL SNG-1256
AAACATACCTGGTA.1          pDC_CTRL SNG-1039
AAACATACGATGAA.1 CD4 Memory T_CTRL SNG-1488</code></pre>
</div>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">ifnb</span><span class="op">$</span><span class="va">donor_id</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
 SNG-101 SNG-1015 SNG-1016 SNG-1039  SNG-107 SNG-1244 SNG-1256 SNG-1488 
    1197     3116     1438      663      652     1998     2363     2241 </code></pre>
</div>
</div>
</section><section id="执行pseudobulking" class="level3"><h3 class="anchored" data-anchor-id="执行pseudobulking">执行pseudobulking</h3>
<p>按照治疗分组（STIM vs.&nbsp;CTRL）、患者IDs、细胞类型（seurat_annotations）3个条件，执行pseudobulking (<code>AggregateExpression</code>)。</p>
<div class="cell" data-hash="de_vignette_cache/html/unnamed-chunk-11_48e854387f3d6b3d7db0d4246bdc7123">
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pseudo_ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/AggregateExpression.html">AggregateExpression</a></span><span class="op">(</span><span class="va">ifnb</span>, </span>
<span>                                   assays <span class="op">=</span> <span class="st">"RNA"</span>, </span>
<span>                                   return.seurat <span class="op">=</span> <span class="cn">T</span>, </span>
<span>                                   group.by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"stim"</span>, <span class="st">"donor_id"</span>, <span class="st">"seurat_annotations"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">pseudo_ifnb</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
14053 features across 206 samples within 1 assay 
Active assay: RNA (14053 features, 0 variable features)
 3 layers present: counts, data, scale.data</code></pre>
</div>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">pseudo_ifnb</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span> <span class="co"># 可以看到现在的表达矩阵的列（即样本）为治疗分组+患者IDs+细胞类型</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                                         orig.ident stim donor_id
CTRL_SNG-101_CD14 Mono       CTRL_SNG-101_CD14 Mono CTRL  SNG-101
CTRL_SNG-101_CD4 Naive T   CTRL_SNG-101_CD4 Naive T CTRL  SNG-101
CTRL_SNG-101_CD4 Memory T CTRL_SNG-101_CD4 Memory T CTRL  SNG-101
CTRL_SNG-101_CD16 Mono       CTRL_SNG-101_CD16 Mono CTRL  SNG-101
CTRL_SNG-101_B                       CTRL_SNG-101_B CTRL  SNG-101
CTRL_SNG-101_CD8 T               CTRL_SNG-101_CD8 T CTRL  SNG-101
                          seurat_annotations
CTRL_SNG-101_CD14 Mono             CD14 Mono
CTRL_SNG-101_CD4 Naive T         CD4 Naive T
CTRL_SNG-101_CD4 Memory T       CD4 Memory T
CTRL_SNG-101_CD16 Mono             CD16 Mono
CTRL_SNG-101_B                             B
CTRL_SNG-101_CD8 T                     CD8 T</code></pre>
</div>
</div>
<p>然后我们在meta.data中增加一列，记录治疗分组（STIM vs.&nbsp;CTRL）+ 细胞类型，这是用于差异分析的分组依据。这和 <a href="#sec-%E5%AF%BB%E6%89%BE%E4%B8%8D%E5%90%8C%E6%A0%B7%E6%9C%AC%E7%B1%BB%E5%9E%8B%E9%97%B4%E5%90%8C%E4%B8%80%E7%BB%86%E8%83%9E%E7%B1%BB%E5%9E%8B%E5%86%85%E7%9A%84%E5%B7%AE%E5%BC%82%E5%9F%BA%E5%9B%A0"><span>Section&nbsp;14.3</span></a> 的第一步是一样的。</p>
<div class="cell" data-hash="de_vignette_cache/html/unnamed-chunk-12_6bb6be1c1852c746bb5ff76061f1742a">
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pseudo_ifnb</span><span class="op">$</span><span class="va">celltype.stim</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="va">pseudo_ifnb</span><span class="op">$</span><span class="va">seurat_annotations</span>, </span>
<span>                                   <span class="va">pseudo_ifnb</span><span class="op">$</span><span class="va">stim</span>, </span>
<span>                                   sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span></span>
<span><span class="va">pseudo_ifnb</span><span class="op">$</span><span class="va">celltype.stim</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   CTRL_SNG-101_CD14 Mono  CTRL_SNG-101_CD4 Naive T CTRL_SNG-101_CD4 Memory T 
         "CD14 Mono_CTRL"        "CD4 Naive T_CTRL"       "CD4 Memory T_CTRL" 
   CTRL_SNG-101_CD16 Mono            CTRL_SNG-101_B 
         "CD16 Mono_CTRL"                  "B_CTRL" </code></pre>
</div>
</div>
</section><section id="执行差异分析" class="level3"><h3 class="anchored" data-anchor-id="执行差异分析">执行差异分析</h3>
<p>Next, we perform DE testing on the pseudobulk level for CD14 monocytes, and compare it against the previous single-cell-level DE results.</p>
<div class="callout callout-style-default callout-caution callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-4-contents" aria-controls="callout-4" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
安装<code>DESeq2</code>包
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-4" class="callout-4-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>由于pseudobulking后的<code>FindMarkers</code>差异分析需要采用<code>DESeq2</code>包提供的方法，所以需要提前安装<code>DESeq2</code>包：</p>
<div class="cell" data-hash="de_vignette_cache/html/unnamed-chunk-13_891700e4f572c497f6f14590d6f67ac5">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"DESeq2"</span>)</span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<div class="cell" data-hash="de_vignette_cache/html/unnamed-chunk-14_102bba7d78f56a2cc98d65c7206d2d8d">
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html">Idents</a></span><span class="op">(</span><span class="va">pseudo_ifnb</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"celltype.stim"</span></span>
<span></span>
<span><span class="va">bulk.mono.de</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindMarkers.html">FindMarkers</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pseudo_ifnb</span>, </span>
<span>                            ident.1 <span class="op">=</span> <span class="st">"CD14 Mono_STIM"</span>, </span>
<span>                            ident.2 <span class="op">=</span> <span class="st">"CD14 Mono_CTRL"</span>,</span>
<span>                            test.use <span class="op">=</span> <span class="st">"DESeq2"</span><span class="op">)</span> <span class="co"># 指定差异分析方法为"DESeq2"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">bulk.mono.de</span>, n <span class="op">=</span> <span class="fl">15</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                 p_val avg_log2FC pct.1 pct.2     p_val_adj
IL1RN    3.701542e-275   6.160156     1     1 5.201777e-271
IFITM2   1.955626e-250   4.318976     1     1 2.748242e-246
SSB      2.699554e-203   3.066647     1     1 3.793683e-199
NT5C3A   2.239898e-198   5.412972     1     1 3.147729e-194
RTCB     5.700554e-162   3.133362     1     1 8.010989e-158
RABGAP1L 4.743010e-161   5.562364     1     1 6.665352e-157
DYNLT1   9.735640e-159   2.402726     1     1 1.368150e-154
PLSCR1   3.191691e-146   2.676047     1     1 4.485284e-142
ISG20    9.664488e-145   5.443114     1     1 1.358151e-140
NAPA     2.858013e-144   1.977719     1     1 4.016365e-140
DDX58    5.957026e-142   4.640111     1     1 8.371409e-138
HERC5    6.333722e-133   5.266515     1     1 8.900780e-129
OASL     3.892853e-130   3.946745     1     1 5.470627e-126
EIF2AK2  6.636434e-128   3.940167     1     1 9.326180e-124
TMEM50A  6.731955e-117   1.355947     1     1 9.460416e-113</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-5-contents" aria-controls="callout-5" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<code>FindMarkers</code>支持的差异分析方法
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-5" class="callout-5-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>We also support many other DE tests using other methods. For completeness, the following tests are currently supported:</p>
<ul>
<li><p>“wilcox” : Wilcoxon rank sum test (default, using ‘<a href="https://github.com/immunogenomics/presto">presto</a>’ package)</p></li>
<li><p>“wilcox_limma” : Wilcoxon rank sum test (using ‘<a href="https://bioconductor.org/packages/release/bioc/html/limma.html">limma</a>’ package)</p></li>
<li><p>“bimod” : Likelihood-ratio test for single cell feature expression, (McDavid et al., Bioinformatics, 2013)</p></li>
<li><p>“roc” : Standard AUC classifier</p></li>
<li><p>“t” : Student’s t-test</p></li>
<li><p>“poisson” : Likelihood ratio test assuming an underlying negative binomial distribution. Use only for UMI-based datasets</p></li>
<li><p>“negbinom” : Likelihood ratio test assuming an underlying negative binomial distribution. Use only for UMI-based datasets</p></li>
<li><p>“LR” : Uses a logistic regression framework to determine differentially expressed genes. Constructs a logistic regression model predicting group membership based on each feature individually and compares this to a null model with a likelihood ratio test.</p></li>
<li><p>“MAST” : GLM-framework that treates cellular detection rate as a covariate (Finak et al, Genome Biology, 2015) (Installation instructions)</p></li>
<li><p>“DESeq2” : DE based on a model using the negative binomial distribution (Love et al, Genome Biology, 2014) (Installation instructions) For MAST and DESeq2, please ensure that these packages are installed separately in order to use them as part of Seurat. Once installed, use the test.use parameter can be used to specify which DE test to use.</p></li>
</ul>
<div class="cell" data-hash="de_vignette_cache/html/unnamed-chunk-15_6a752d3f50f95229e5900878faef1f1c">
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Test for DE features using the MAST package</span></span>
<span><span class="co"># BiocManager::install('limma')</span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html">Idents</a></span><span class="op">(</span><span class="va">ifnb</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"seurat_annotations"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://satijalab.org/seurat/reference/FindMarkers.html">FindMarkers</a></span><span class="op">(</span><span class="va">ifnb</span>, </span>
<span>                 ident.1 <span class="op">=</span> <span class="st">"CD14 Mono"</span>, </span>
<span>                 ident.2 <span class="op">=</span> <span class="st">"CD16 Mono"</span>, </span>
<span>                 test.use <span class="op">=</span> <span class="st">"wilcox_limma"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>               p_val avg_log2FC pct.1 pct.2     p_val_adj
VMO1    0.000000e+00  -5.689802 0.084 0.777  0.000000e+00
MS4A4A  0.000000e+00  -3.356037 0.141 0.747  0.000000e+00
FCGR3A  0.000000e+00  -3.279465 0.418 0.982  0.000000e+00
MS4A7   0.000000e+00  -2.390652 0.557 0.978  0.000000e+00
RPS19   0.000000e+00  -1.321132 0.965 0.999  0.000000e+00
FTL    1.636254e-307   1.318127 1.000 1.000 2.299427e-303</code></pre>
</div>
</div>
</div>
</div>
</div>
</section><section id="比较单细胞水平和pseudobulk水平的差异表达分析" class="level3"><h3 class="anchored" data-anchor-id="比较单细胞水平和pseudobulk水平的差异表达分析">比较单细胞水平和pseudobulk水平的差异表达分析</h3>
<p>接下来，我们可以比较一下单细胞水平的差异表达分析的P值和pseudobulk水平的P值：</p>
<div class="cell" data-hash="de_vignette_cache/html/unnamed-chunk-16_6984869433dffeddc6129a5a1c84416b">
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">bulk.mono.de</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">bulk.mono.de</span><span class="op">)</span>, <span class="st">".bulk"</span><span class="op">)</span> <span class="co"># 重命名列</span></span>
<span><span class="va">bulk.mono.de</span><span class="op">$</span><span class="va">gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">bulk.mono.de</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">mono.de</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">mono.de</span><span class="op">)</span>, <span class="st">".sc"</span><span class="op">)</span></span>
<span><span class="va">mono.de</span><span class="op">$</span><span class="va">gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">mono.de</span><span class="op">)</span></span>
<span></span>
<span><span class="va">merge_dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html">merge</a></span><span class="op">(</span><span class="va">mono.de</span>, <span class="va">bulk.mono.de</span>, by <span class="op">=</span> <span class="st">"gene"</span><span class="op">)</span></span>
<span><span class="va">merge_dat</span> <span class="op">&lt;-</span> <span class="va">merge_dat</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="va">merge_dat</span><span class="op">$</span><span class="va">p_val.bulk</span><span class="op">)</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># 查看在两种差异分析方法中P值都有意义的基因名</span></span>
<span><span class="va">common</span> <span class="op">&lt;-</span> <span class="va">merge_dat</span><span class="op">$</span><span class="va">gene</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">merge_dat</span><span class="op">$</span><span class="va">p_val.bulk</span> <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">&amp;</span> </span>
<span>                                <span class="va">merge_dat</span><span class="op">$</span><span class="va">p_val.sc</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co"># 查看在pseudobulk水平P&gt;0.05但是在单细胞水平P&lt;0.05的基因名：</span></span>
<span><span class="va">only_sc</span> <span class="op">&lt;-</span> <span class="va">merge_dat</span><span class="op">$</span><span class="va">gene</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">merge_dat</span><span class="op">$</span><span class="va">p_val.bulk</span> <span class="op">&gt;</span> <span class="fl">0.05</span> <span class="op">&amp;</span> </span>
<span>                                  <span class="va">merge_dat</span><span class="op">$</span><span class="va">p_val.sc</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co"># 查看在pseudobulk水平P&lt;0.05但是在单细胞水平P&gt;0.05的基因名：</span></span>
<span><span class="va">only_bulk</span> <span class="op">&lt;-</span> <span class="va">merge_dat</span><span class="op">$</span><span class="va">gene</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">merge_dat</span><span class="op">$</span><span class="va">p_val.bulk</span> <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">&amp;</span> </span>
<span>                                    <span class="va">merge_dat</span><span class="op">$</span><span class="va">p_val.sc</span> <span class="op">&gt;</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">'# 在两种差异分析方法中P值都&lt;0.05的基因有: '</span>,<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">common</span><span class="op">)</span>, <span class="st">"个"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "# 在两种差异分析方法中P值都&lt;0.05的基因有: 3519个"</code></pre>
</div>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">'# 仅在单细胞水平差异分析中P值&lt;0.05的基因有: '</span>,<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">only_sc</span><span class="op">)</span>, <span class="st">"个"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "# 仅在单细胞水平差异分析中P值&lt;0.05的基因有: 1649个"</code></pre>
</div>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">'# 仅在pseudobulk水平差异分析中P值&lt;0.05的基因有: '</span>,<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">only_bulk</span><span class="op">)</span>, <span class="st">"个"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "# 仅在pseudobulk水平差异分析中P值&lt;0.05的基因有: 204个"</code></pre>
</div>
</div>
<blockquote class="blockquote">
<p>We can see that while the p-values are correlated between the single-cell and pseudobulk data, the <u><strong>single-cell p-values are often smaller</strong></u> and suggest higher levels of significance. In particular, there are 3,519 genes with evidence of differential expression (prior to multiple hypothesis testing) in both analyses, 1,649 genes that only appear to be differentially expressed in the single-cell analysis, and just 204 genes that only appear to be differentially expressed in the bulk analysis. We can investigate these discrepancies using&nbsp;<code>VlnPlot</code>.</p>
</blockquote>
<p>接下来，我们通过小提琴图来检查两种方法中的Top共同差异基因在在刺激组和对照组的表达水平：</p>
<div class="cell" data-hash="de_vignette_cache/html/unnamed-chunk-17_d9321252c5c273f974d2ccfe2c1c31e1">
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># create a new column to annotate sample-condition-celltype in the single-cell dataset</span></span>
<span><span class="va">ifnb</span><span class="op">$</span><span class="va">donor_id.stim</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">ifnb</span><span class="op">$</span><span class="va">stim</span>, <span class="st">"-"</span>, <span class="va">ifnb</span><span class="op">$</span><span class="va">donor_id</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">ifnb</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                  orig.ident nCount_RNA nFeature_RNA stim seurat_annotations
AAACATACATTTCC.1 IMMUNE_CTRL       3017          877 CTRL          CD14 Mono
AAACATACCAGAAA.1 IMMUNE_CTRL       2481          713 CTRL          CD14 Mono
AAACATACCTCGCT.1 IMMUNE_CTRL       3420          850 CTRL          CD14 Mono
AAACATACCTGGTA.1 IMMUNE_CTRL       3156         1109 CTRL                pDC
AAACATACGATGAA.1 IMMUNE_CTRL       1868          634 CTRL       CD4 Memory T
AAACATACGGCATT.1 IMMUNE_CTRL       1581          557 CTRL          CD14 Mono
                     celltype.stim donor_id donor_id.stim
AAACATACATTTCC.1    CD14 Mono_CTRL SNG-1016 CTRL-SNG-1016
AAACATACCAGAAA.1    CD14 Mono_CTRL SNG-1256 CTRL-SNG-1256
AAACATACCTCGCT.1    CD14 Mono_CTRL SNG-1256 CTRL-SNG-1256
AAACATACCTGGTA.1          pDC_CTRL SNG-1039 CTRL-SNG-1039
AAACATACGATGAA.1 CD4 Memory T_CTRL SNG-1488 CTRL-SNG-1488
AAACATACGGCATT.1    CD14 Mono_CTRL SNG-1015 CTRL-SNG-1015</code></pre>
</div>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">ifnb</span><span class="op">$</span><span class="va">celltype.stim</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
 B Activated_CTRL  B Activated_STIM            B_CTRL            B_STIM 
              176               198               403               554 
   CD14 Mono_CTRL    CD14 Mono_STIM    CD16 Mono_CTRL    CD16 Mono_STIM 
             2167              2086               494               521 
CD4 Memory T_CTRL CD4 Memory T_STIM  CD4 Naive T_CTRL  CD4 Naive T_STIM 
              849               876               960              1497 
       CD8 T_CTRL        CD8 T_STIM           DC_CTRL           DC_STIM 
              349               451               255               208 
       Eryth_CTRL        Eryth_STIM           Mk_CTRL           Mk_STIM 
               22                32               110               114 
          NK_CTRL           NK_STIM          pDC_CTRL          pDC_STIM 
              295               310                49                79 
 T activated_CTRL  T activated_STIM 
              291               322 </code></pre>
</div>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html">Idents</a></span><span class="op">(</span><span class="va">ifnb</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"celltype.stim"</span></span>
<span><span class="co"># 这里我们检查p_val.bulk最小的前两个Top差异基因</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">merge_dat</span><span class="op">[</span><span class="va">merge_dat</span><span class="op">$</span><span class="va">gene</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="va">common</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'gene'</span>,<span class="st">'p_val.sc'</span>,<span class="st">'p_val.bulk'</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       gene p_val.sc    p_val.bulk
2785  IL1RN        0 3.701542e-275
2739 IFITM2        0 1.955626e-250</code></pre>
</div>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 在细胞类型水平（CD14 Mono）查看这两个Top差异基因在刺激组和对照组的表达水平</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html">VlnPlot</a></span><span class="op">(</span><span class="va">ifnb</span>, </span>
<span>        features <span class="op">=</span> <span class="va">common</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>, </span>
<span>        idents <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CD14 Mono_CTRL"</span>, <span class="st">"CD14 Mono_STIM"</span><span class="op">)</span>, </span>
<span>        group.by <span class="op">=</span> <span class="st">"stim"</span><span class="op">)</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="de_vignette_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 在样本（患者）水平查看这两个Top差异基因在刺激组和对照组的表达水平</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html">VlnPlot</a></span><span class="op">(</span><span class="va">ifnb</span>, </span>
<span>        features <span class="op">=</span> <span class="va">common</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>, </span>
<span>        idents <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CD14 Mono_CTRL"</span>, <span class="st">"CD14 Mono_STIM"</span><span class="op">)</span>, </span>
<span>        group.by <span class="op">=</span> <span class="st">"donor_id.stim"</span>, </span>
<span>        ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="de_vignette_files/figure-html/unnamed-chunk-17-2.png" class="img-fluid" width="672"></p>
</div>
</div>
<blockquote class="blockquote">
<p>In both the pseudobulk and single-cell analyses, the p-values for these two genes are astronomically small. For both of these genes, when just comparing all stimulated CD4 monocytes to all control CD4 monocytes across samples, we see much higher expression in the stimulated cells. When breaking down these cells by sample, we continue to see consistently higher expression levels in the stimulated samples compared to the control samples; in other words, this finding is not driven by just one or two samples. Because of this consistency, we find this signal in both analyses.</p>
</blockquote>
<p>By contrast, we can examine examples of genes that are only DE under the single-cell analysis.</p>
<div class="cell" data-hash="de_vignette_cache/html/unnamed-chunk-18_3096561ec9ee14a19b33d38680247a6b">
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">merge_dat</span><span class="op">[</span><span class="va">merge_dat</span><span class="op">$</span><span class="va">gene</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'SRGN'</span>,<span class="st">'HLA-DRA'</span><span class="op">)</span>, </span>
<span>                <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'gene'</span>,<span class="st">'p_val.sc'</span>,<span class="st">'p_val.bulk'</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        gene     p_val.sc p_val.bulk
5710    SRGN 4.025076e-21  0.1823188
2603 HLA-DRA 4.989863e-09  0.1851302</code></pre>
</div>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html">VlnPlot</a></span><span class="op">(</span><span class="va">ifnb</span>, </span>
<span>        features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'SRGN'</span>,<span class="st">'HLA-DRA'</span><span class="op">)</span>, </span>
<span>        idents <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CD14 Mono_CTRL"</span>, <span class="st">"CD14 Mono_STIM"</span><span class="op">)</span>, </span>
<span>        group.by <span class="op">=</span> <span class="st">"stim"</span><span class="op">)</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="de_vignette_files/figure-html/unnamed-chunk-18-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html">VlnPlot</a></span><span class="op">(</span><span class="va">ifnb</span>, </span>
<span>        features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'SRGN'</span>,<span class="st">'HLA-DRA'</span><span class="op">)</span>, </span>
<span>        idents <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CD14 Mono_CTRL"</span>, <span class="st">"CD14 Mono_STIM"</span><span class="op">)</span>, </span>
<span>        group.by <span class="op">=</span> <span class="st">"donor_id.stim"</span>, </span>
<span>        ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="de_vignette_files/figure-html/unnamed-chunk-18-2.png" class="img-fluid" width="672"></p>
</div>
</div>
<blockquote class="blockquote">
<p>Here, SRGN and HLA-DRA both have very small p-values in the single-cell analysis (on the orders of&nbsp;10<sup>−21</sup>&nbsp;and&nbsp;10<sup>−9</sup>), but much larger p-values around 0.18 in the pseudobulk analysis. While there appears to be a difference between control and simulated cells when ignoring sample information, the signal is much weaker on the sample level, and we can see notable variability from sample to sample.</p>
</blockquote>
<p>所以，从这个例子中可以看出<strong>pseudobulk后的差异分析的结果更加准确</strong>。</p>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-6-contents" aria-controls="callout-6" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Session Info
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-6" class="callout-6-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-hash="de_vignette_cache/html/unnamed-chunk-19_e872b82874773bb28f5d38afc06d8f35">
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.3.2 (2023-10-31)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Sonoma 14.1.2

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: Asia/Shanghai
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] Seurat_5.0.1       SeuratObject_5.0.1 sp_2.1-2          

loaded via a namespace (and not attached):
  [1] deldir_2.0-2           pbapply_1.7-2          gridExtra_2.3         
  [4] rlang_1.1.2            magrittr_2.0.3         RcppAnnoy_0.0.21      
  [7] spatstat.geom_3.2-7    matrixStats_1.1.0      ggridges_0.5.4        
 [10] compiler_4.3.2         png_0.1-8              vctrs_0.6.5           
 [13] reshape2_1.4.4         stringr_1.5.1          pkgconfig_2.0.3       
 [16] fastmap_1.1.1          ellipsis_0.3.2         labeling_0.4.3        
 [19] utf8_1.2.4             promises_1.2.1         rmarkdown_2.25        
 [22] purrr_1.0.2            xfun_0.41              jsonlite_1.8.7        
 [25] goftest_1.2-3          later_1.3.1            spatstat.utils_3.0-4  
 [28] irlba_2.3.5.1          parallel_4.3.2         cluster_2.1.6         
 [31] R6_2.5.1               ica_1.0-3              stringi_1.8.2         
 [34] RColorBrewer_1.1-3     spatstat.data_3.0-3    limma_3.58.1          
 [37] reticulate_1.34.0      parallelly_1.36.0      lmtest_0.9-40         
 [40] scattermore_1.2        Rcpp_1.0.11            knitr_1.45            
 [43] tensor_1.5             future.apply_1.11.0    zoo_1.8-12            
 [46] sctransform_0.4.1      httpuv_1.6.12          Matrix_1.6-4          
 [49] splines_4.3.2          igraph_1.5.1           tidyselect_1.2.0      
 [52] abind_1.4-5            rstudioapi_0.15.0      yaml_2.3.7            
 [55] spatstat.random_3.2-2  codetools_0.2-19       miniUI_0.1.1.1        
 [58] spatstat.explore_3.2-5 listenv_0.9.0          lattice_0.22-5        
 [61] tibble_3.2.1           plyr_1.8.9             withr_2.5.2           
 [64] shiny_1.8.0            ROCR_1.0-11            evaluate_0.23         
 [67] Rtsne_0.16             future_1.33.0          fastDummies_1.7.3     
 [70] survival_3.5-7         polyclip_1.10-6        fitdistrplus_1.1-11   
 [73] pillar_1.9.0           KernSmooth_2.23-22     plotly_4.10.3         
 [76] generics_0.1.3         RcppHNSW_0.5.0         ggplot2_3.4.4         
 [79] munsell_0.5.0          scales_1.3.0           globals_0.16.2        
 [82] xtable_1.8-4           glue_1.6.2             lazyeval_0.2.2        
 [85] tools_4.3.2            data.table_1.14.8      RSpectra_0.16-1       
 [88] RANN_2.6.1             leiden_0.4.3.1         dotCall64_1.1-1       
 [91] cowplot_1.1.1          grid_4.3.2             tidyr_1.3.0           
 [94] colorspace_2.1-0       nlme_3.1-164           patchwork_1.1.3       
 [97] presto_1.0.0           cli_3.6.1              spatstat.sparse_3.0-3 
[100] spam_2.10-0            fansi_1.0.5            viridisLite_0.4.2     
[103] dplyr_1.1.4            uwot_0.1.16            gtable_0.3.4          
[106] digest_0.6.33          progressr_0.14.0       ggrepel_0.9.4         
[109] farver_2.1.1           htmlwidgets_1.6.3      htmltools_0.5.7       
[112] lifecycle_1.0.4        httr_1.4.7             statmod_1.5.0         
[115] mime_0.12              MASS_7.3-60           </code></pre>
</div>
</div>
</div>
</div>
</div>


<!-- -->

<div id="refs" class="references csl-bib-body hanging-indent" role="list" style="display: none">
<div id="ref-junttila2022" class="csl-entry" role="listitem">
Junttila, Sini, Johannes Smolander, and Laura L Elo. 2022. <span>“Benchmarking Methods for Detecting Differential States Between Conditions from Multi-Subject Single-Cell RNA-Seq Data.”</span> <em>Briefings in Bioinformatics</em> 23 (5). <a href="https://doi.org/10.1093/bib/bbac286">https://doi.org/10.1093/bib/bbac286</a>.
</div>
<div id="ref-kang2017" class="csl-entry" role="listitem">
Kang, Hyun Min, Meena Subramaniam, Sasha Targ, Michelle Nguyen, Lenka Maliskova, Elizabeth McCarthy, Eunice Wan, et al. 2017. <span>“Multiplexed Droplet Single-Cell RNA-Sequencing Using Natural Genetic Variation.”</span> <em>Nature Biotechnology</em> 36 (1): 89–94. <a href="https://doi.org/10.1038/nbt.4042">https://doi.org/10.1038/nbt.4042</a>.
</div>
<div id="ref-squair2021" class="csl-entry" role="listitem">
Squair, Jordan W., Matthieu Gautier, Claudia Kathe, Mark A. Anderson, Nicholas D. James, Thomas H. Hutson, Rémi Hudelle, et al. 2021. <span>“Confronting False Discoveries in Single-Cell Differential Expression.”</span> <em>Nature Communications</em> 12 (1). <a href="https://doi.org/10.1038/s41467-021-25960-2">https://doi.org/10.1038/s41467-021-25960-2</a>.
</div>
<div id="ref-zimmerman2021" class="csl-entry" role="listitem">
Zimmerman, Kip D., Mark A. Espeland, and Carl D. Langefeld. 2021. <span>“A Practical Solution to Pseudoreplication Bias in Single-Cell Studies.”</span> <em>Nature Communications</em> 12 (1). <a href="https://doi.org/10.1038/s41467-021-21038-1">https://doi.org/10.1038/s41467-021-21038-1</a>.
</div>
</div>
</section></section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  let localAlternateSentinel = 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var filterRegex = new RegExp("https:\/\/djhcod\.github\.io\/r-notes\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
          // target, if specified
          link.setAttribute("target", "_blank");
      }
    }
});
</script><script src="https://utteranc.es/client.js" repo="https://github.com/djhcod/r-notes" issue-term="pathname" theme="github-light" crossorigin="anonymous" async="">
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="../../single_cell/seurat/cell_cycle_regression.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">消除细胞周期的影响</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../../single_cell/scRNA-seq_online/00_intro.html" class="pagination-link">
        <span class="nav-page-text">scRNA-seq_online学习材料</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb67" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="fu"># 差异表达分析 {#sec-差异表达分析}</span></span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a>原文：<span class="co">[</span><span class="ot">*Differential expression testing*</span><span class="co">](https://satijalab.org/seurat/articles/de_vignette)</span></span>
<span id="cb67-4"><a href="#cb67-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-5"><a href="#cb67-5" aria-hidden="true" tabindex="-1"></a>原文发布日期：2023年10月31日</span>
<span id="cb67-6"><a href="#cb67-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-7"><a href="#cb67-7" aria-hidden="true" tabindex="-1"></a><span class="fu">## 数据读取和预处理</span></span>
<span id="cb67-8"><a href="#cb67-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-9"><a href="#cb67-9" aria-hidden="true" tabindex="-1"></a>This vignette highlights some example workflows for performing differential expression in Seurat. For demonstration purposes, we will be using the interferon-beta stimulated human PBMCs dataset <span class="co">[</span><span class="ot">@kang2017</span><span class="co">]</span> that is available via the&nbsp;SeuratData&nbsp;package.</span>
<span id="cb67-10"><a href="#cb67-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-11"><a href="#cb67-11" aria-hidden="true" tabindex="-1"></a>::: {.callout-note collapse="true" appearance="minimal" icon="false"}</span>
<span id="cb67-12"><a href="#cb67-12" aria-hidden="true" tabindex="-1"></a><span class="fu">###### 在线读取（可能需要全局代理）</span></span>
<span id="cb67-13"><a href="#cb67-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-16"><a href="#cb67-16" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb67-17"><a href="#cb67-17" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb67-18"><a href="#cb67-18" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb67-19"><a href="#cb67-19" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SeuratData)</span>
<span id="cb67-20"><a href="#cb67-20" aria-hidden="true" tabindex="-1"></a><span class="fu">InstallData</span>(<span class="st">"ifnb"</span>)</span>
<span id="cb67-21"><a href="#cb67-21" aria-hidden="true" tabindex="-1"></a>ifnb <span class="ot">&lt;-</span> <span class="fu">LoadData</span>(<span class="st">"ifnb"</span>)</span>
<span id="cb67-22"><a href="#cb67-22" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb67-23"><a href="#cb67-23" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb67-24"><a href="#cb67-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-25"><a href="#cb67-25" aria-hidden="true" tabindex="-1"></a>从本地下载好的数据读取：</span>
<span id="cb67-26"><a href="#cb67-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-29"><a href="#cb67-29" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb67-30"><a href="#cb67-30" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb67-31"><a href="#cb67-31" aria-hidden="true" tabindex="-1"></a>ifnb <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"data/pbmc_ifnb.rds"</span>)</span>
<span id="cb67-32"><a href="#cb67-32" aria-hidden="true" tabindex="-1"></a>ifnb</span>
<span id="cb67-33"><a href="#cb67-33" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(ifnb<span class="sc">@</span>meta.data, <span class="dv">5</span>)</span>
<span id="cb67-34"><a href="#cb67-34" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(ifnb<span class="sc">$</span>seurat_annotations) <span class="co"># 这里的数据已经提前注释好了细胞类型</span></span>
<span id="cb67-35"><a href="#cb67-35" aria-hidden="true" tabindex="-1"></a><span class="co"># 标准化</span></span>
<span id="cb67-36"><a href="#cb67-36" aria-hidden="true" tabindex="-1"></a>ifnb <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(ifnb)</span>
<span id="cb67-37"><a href="#cb67-37" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb67-38"><a href="#cb67-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-39"><a href="#cb67-39" aria-hidden="true" tabindex="-1"></a><span class="fu">## 寻找细胞类型之间的差异基因</span></span>
<span id="cb67-40"><a href="#cb67-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-41"><a href="#cb67-41" aria-hidden="true" tabindex="-1"></a>::: callout-caution</span>
<span id="cb67-42"><a href="#cb67-42" aria-hidden="true" tabindex="-1"></a>For a **much** faster implementation of the Wilcoxon Rank Sum Test,(default method for <span class="in">`FindMarkers`</span>) please install the <span class="co">[</span><span class="ot">`presto`</span><span class="co">](https://github.com/immunogenomics/presto)</span> package:</span>
<span id="cb67-43"><a href="#cb67-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-46"><a href="#cb67-46" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb67-47"><a href="#cb67-47" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb67-48"><a href="#cb67-48" aria-hidden="true" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">'immunogenomics/presto'</span>)</span>
<span id="cb67-49"><a href="#cb67-49" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb67-50"><a href="#cb67-50" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb67-51"><a href="#cb67-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-52"><a href="#cb67-52" aria-hidden="true" tabindex="-1"></a>Find DE features between CD16 Mono and CD1 Mono：</span>
<span id="cb67-53"><a href="#cb67-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-56"><a href="#cb67-56" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb67-57"><a href="#cb67-57" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(ifnb) <span class="ot">&lt;-</span> <span class="st">"seurat_annotations"</span></span>
<span id="cb67-58"><a href="#cb67-58" aria-hidden="true" tabindex="-1"></a>monocyte.de.markers <span class="ot">&lt;-</span> <span class="fu">FindMarkers</span>(ifnb, </span>
<span id="cb67-59"><a href="#cb67-59" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">ident.1 =</span> <span class="st">"CD16 Mono"</span>, </span>
<span id="cb67-60"><a href="#cb67-60" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">ident.2 =</span> <span class="st">"CD14 Mono"</span>)</span>
<span id="cb67-61"><a href="#cb67-61" aria-hidden="true" tabindex="-1"></a><span class="co"># view results</span></span>
<span id="cb67-62"><a href="#cb67-62" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(monocyte.de.markers)</span>
<span id="cb67-63"><a href="#cb67-63" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(monocyte.de.markers)</span>
<span id="cb67-64"><a href="#cb67-64" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb67-65"><a href="#cb67-65" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-66"><a href="#cb67-66" aria-hidden="true" tabindex="-1"></a>The results data frame has the following columns :</span>
<span id="cb67-67"><a href="#cb67-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-68"><a href="#cb67-68" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`p_val`</span> : p-value (unadjusted)</span>
<span id="cb67-69"><a href="#cb67-69" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-70"><a href="#cb67-70" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`avg_log2FC`</span> : log fold-change of the average expression between the two groups. Positive values indicate that the feature is more highly expressed in the first group.</span>
<span id="cb67-71"><a href="#cb67-71" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-72"><a href="#cb67-72" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`pct.1`</span> : The percentage of cells where the feature is detected in the first group</span>
<span id="cb67-73"><a href="#cb67-73" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-74"><a href="#cb67-74" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`pct.2`</span> : The percentage of cells where the feature is detected in the second group</span>
<span id="cb67-75"><a href="#cb67-75" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-76"><a href="#cb67-76" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`p_val_adj`</span> : Adjusted p-value, based on **Bonferroni correction** using all features in the dataset.</span>
<span id="cb67-77"><a href="#cb67-77" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-78"><a href="#cb67-78" aria-hidden="true" tabindex="-1"></a>If the&nbsp;<span class="in">`ident.2`</span>&nbsp;parameter is omitted or set to&nbsp;<span class="in">`NULL`</span>,&nbsp;<span class="co">[</span><span class="ot">`FindMarkers()`</span><span class="co">](https://satijalab.org/seurat/reference/findmarkers)</span>&nbsp;will test for differentially expressed features **between the group specified by&nbsp;`ident.1`&nbsp;and all other cells**. Additionally, the parameter&nbsp;<span class="in">`only.pos`</span>&nbsp;can be set to&nbsp;<span class="in">`TRUE`</span>&nbsp;to only search for positive markers, i.e.&nbsp;features that are more highly expressed in the&nbsp;<span class="in">`ident.1`</span>&nbsp;group.</span>
<span id="cb67-79"><a href="#cb67-79" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-82"><a href="#cb67-82" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb67-83"><a href="#cb67-83" aria-hidden="true" tabindex="-1"></a>monocyte.de.markers <span class="ot">&lt;-</span> <span class="fu">FindMarkers</span>(ifnb, </span>
<span id="cb67-84"><a href="#cb67-84" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">ident.1 =</span> <span class="st">"CD16 Mono"</span>, </span>
<span id="cb67-85"><a href="#cb67-85" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">ident.2 =</span> <span class="cn">NULL</span>, </span>
<span id="cb67-86"><a href="#cb67-86" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">only.pos =</span> <span class="cn">TRUE</span>)</span>
<span id="cb67-87"><a href="#cb67-87" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(monocyte.de.markers)</span>
<span id="cb67-88"><a href="#cb67-88" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(monocyte.de.markers)</span>
<span id="cb67-89"><a href="#cb67-89" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb67-90"><a href="#cb67-90" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-91"><a href="#cb67-91" aria-hidden="true" tabindex="-1"></a><span class="fu">## 寻找不同样本类型间同一细胞类型内的差异基因 {#sec-寻找不同样本类型间同一细胞类型内的差异基因}</span></span>
<span id="cb67-92"><a href="#cb67-92" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-93"><a href="#cb67-93" aria-hidden="true" tabindex="-1"></a>Since this dataset contains treatment information (control versus stimulated with interferon-beta), we can also ask **what genes change in different conditions for cells of the same type**.</span>
<span id="cb67-94"><a href="#cb67-94" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-95"><a href="#cb67-95" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>First, we create a column in the&nbsp;<span class="in">`meta.data`</span>&nbsp;slot to hold both the cell type and treatment information and switch the current&nbsp;<span class="in">`Idents`</span>&nbsp;to that column.</span>
<span id="cb67-96"><a href="#cb67-96" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-99"><a href="#cb67-99" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb67-100"><a href="#cb67-100" aria-hidden="true" tabindex="-1"></a>ifnb<span class="sc">$</span>celltype.stim <span class="ot">&lt;-</span> <span class="fu">paste</span>(ifnb<span class="sc">$</span>seurat_annotations, ifnb<span class="sc">$</span>stim, <span class="at">sep =</span> <span class="st">"_"</span>)</span>
<span id="cb67-101"><a href="#cb67-101" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(ifnb<span class="sc">$</span>celltype.stim)</span>
<span id="cb67-102"><a href="#cb67-102" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(ifnb) <span class="ot">&lt;-</span> <span class="st">"celltype.stim"</span></span>
<span id="cb67-103"><a href="#cb67-103" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb67-104"><a href="#cb67-104" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-105"><a href="#cb67-105" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Then we use&nbsp;<span class="in">`FindMarkers()`</span>&nbsp;to find the genes that are different between control and stimulated CD14 monocytes.</span>
<span id="cb67-106"><a href="#cb67-106" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-109"><a href="#cb67-109" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb67-110"><a href="#cb67-110" aria-hidden="true" tabindex="-1"></a>mono.de <span class="ot">&lt;-</span> <span class="fu">FindMarkers</span>(ifnb, </span>
<span id="cb67-111"><a href="#cb67-111" aria-hidden="true" tabindex="-1"></a>                       <span class="at">ident.1 =</span> <span class="st">"CD14 Mono_STIM"</span>, </span>
<span id="cb67-112"><a href="#cb67-112" aria-hidden="true" tabindex="-1"></a>                       <span class="at">ident.2 =</span> <span class="st">"CD14 Mono_CTRL"</span>, </span>
<span id="cb67-113"><a href="#cb67-113" aria-hidden="true" tabindex="-1"></a>                       <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb67-114"><a href="#cb67-114" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(mono.de)</span>
<span id="cb67-115"><a href="#cb67-115" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(mono.de, <span class="at">n =</span> <span class="dv">10</span>)</span>
<span id="cb67-116"><a href="#cb67-116" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb67-117"><a href="#cb67-117" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-118"><a href="#cb67-118" aria-hidden="true" tabindex="-1"></a>However, the p-values obtained from this analysis should be interpreted with caution, because **these tests treat each cell as an independent replicate and ignore inherent correlations between cells originating from the same sample**. Such analyses have been shown to find a large number of **false positive associations**, as has been demonstrated by&nbsp;[@squair2021],&nbsp;[@zimmerman2021],&nbsp;[@junttila2022], and others. Below, we show how **pseudobulking** can be used to account for such within-sample correlation.</span>
<span id="cb67-119"><a href="#cb67-119" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-120"><a href="#cb67-120" aria-hidden="true" tabindex="-1"></a><span class="fu">## pseudobulking后的差异分析</span></span>
<span id="cb67-121"><a href="#cb67-121" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-122"><a href="#cb67-122" aria-hidden="true" tabindex="-1"></a>To pseudobulk, we will use&nbsp;<span class="in">`AggregateExpression()`</span>&nbsp;to sum together **gene counts of all the cells from the same sample** for each cell type. This results in **one gene expression profile per sample** **and cell type**. We can then perform DE analysis using&nbsp;[`DESeq2`](https://bioconductor.org/packages/release/bioc/html/DESeq2.html)&nbsp;**on the sample level**. This treats the samples, rather than the individual cells, as independent observations. 参考前面的 @sec-识别不同样本类型间的差异基因 。</span>
<span id="cb67-123"><a href="#cb67-123" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-124"><a href="#cb67-124" aria-hidden="true" tabindex="-1"></a><span class="fu">### 准备样本信息</span></span>
<span id="cb67-125"><a href="#cb67-125" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-126"><a href="#cb67-126" aria-hidden="true" tabindex="-1"></a>First, we need to retrieve the sample information for each cell. This is not loaded in the metadata, so we will load it from the&nbsp;<span class="co">[</span><span class="ot">Github repo</span><span class="co">](https://github.com/yelabucsf/demuxlet_paper_code/tree/master/)</span>&nbsp;of the source data for the original paper.</span>
<span id="cb67-127"><a href="#cb67-127" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-128"><a href="#cb67-128" aria-hidden="true" tabindex="-1"></a>::: {.callout-tip collapse="true" appearance="minimal" icon="false"}</span>
<span id="cb67-129"><a href="#cb67-129" aria-hidden="true" tabindex="-1"></a><span class="fu">###### Add sample information to the dataset</span></span>
<span id="cb67-130"><a href="#cb67-130" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-133"><a href="#cb67-133" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb67-134"><a href="#cb67-134" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb67-135"><a href="#cb67-135" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: fenced</span></span>
<span id="cb67-136"><a href="#cb67-136" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-137"><a href="#cb67-137" aria-hidden="true" tabindex="-1"></a><span class="co"># 从GitHub仓库读取（可能需要代理）</span></span>
<span id="cb67-138"><a href="#cb67-138" aria-hidden="true" tabindex="-1"></a><span class="co"># load the inferred sample IDs of each cell</span></span>
<span id="cb67-139"><a href="#cb67-139" aria-hidden="true" tabindex="-1"></a>ctrl <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="fu">url</span>(<span class="st">"https://raw.githubusercontent.com/yelabucsf/demuxlet_paper_code/master/fig3/ye1.ctrl.8.10.sm.best"</span>), <span class="at">head =</span> T, <span class="at">stringsAsFactors =</span> F)</span>
<span id="cb67-140"><a href="#cb67-140" aria-hidden="true" tabindex="-1"></a>stim <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="fu">url</span>(<span class="st">"https://raw.githubusercontent.com/yelabucsf/demuxlet_paper_code/master/fig3/ye2.stim.8.10.sm.best"</span>), <span class="at">head =</span> T, <span class="at">stringsAsFactors =</span> F)</span>
<span id="cb67-141"><a href="#cb67-141" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb67-142"><a href="#cb67-142" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-145"><a href="#cb67-145" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb67-146"><a href="#cb67-146" aria-hidden="true" tabindex="-1"></a><span class="co"># 这里提前下载好了两个样本信息文件，所以直接从本地读取</span></span>
<span id="cb67-147"><a href="#cb67-147" aria-hidden="true" tabindex="-1"></a>ctrl <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"data/inferred_sample_ids_ctrl.rds"</span>)</span>
<span id="cb67-148"><a href="#cb67-148" aria-hidden="true" tabindex="-1"></a>stim <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"data/inferred_sample_ids_stim.rds"</span>)</span>
<span id="cb67-149"><a href="#cb67-149" aria-hidden="true" tabindex="-1"></a>info <span class="ot">&lt;-</span> <span class="fu">rbind</span>(ctrl, stim)</span>
<span id="cb67-150"><a href="#cb67-150" aria-hidden="true" tabindex="-1"></a>info<span class="sc">$</span>BARCODE[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span>
<span id="cb67-151"><a href="#cb67-151" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(ifnb)[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span>
<span id="cb67-152"><a href="#cb67-152" aria-hidden="true" tabindex="-1"></a><span class="co"># 可以看到两者的barcode形式不一致</span></span>
<span id="cb67-153"><a href="#cb67-153" aria-hidden="true" tabindex="-1"></a><span class="co"># rename the cell IDs by substituting the '-' into '.'</span></span>
<span id="cb67-154"><a href="#cb67-154" aria-hidden="true" tabindex="-1"></a>info<span class="sc">$</span>BARCODE <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">-"</span>, <span class="at">replacement =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">."</span>, info<span class="sc">$</span>BARCODE)</span>
<span id="cb67-155"><a href="#cb67-155" aria-hidden="true" tabindex="-1"></a>info<span class="sc">$</span>BARCODE[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span>
<span id="cb67-156"><a href="#cb67-156" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-157"><a href="#cb67-157" aria-hidden="true" tabindex="-1"></a><span class="co"># only keep the cells with high-confidence sample ID</span></span>
<span id="cb67-158"><a href="#cb67-158" aria-hidden="true" tabindex="-1"></a>info <span class="ot">&lt;-</span> info[<span class="fu">grep</span>(<span class="at">pattern =</span> <span class="st">"SNG"</span>, <span class="at">x =</span> info<span class="sc">$</span>BEST), ]</span>
<span id="cb67-159"><a href="#cb67-159" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-160"><a href="#cb67-160" aria-hidden="true" tabindex="-1"></a><span class="co"># remove cells with duplicated IDs in both ctrl and stim groups</span></span>
<span id="cb67-161"><a href="#cb67-161" aria-hidden="true" tabindex="-1"></a>info <span class="ot">&lt;-</span> info[<span class="sc">!</span><span class="fu">duplicated</span>(info<span class="sc">$</span>BARCODE) <span class="sc">&amp;</span> <span class="sc">!</span><span class="fu">duplicated</span>(info<span class="sc">$</span>BARCODE, <span class="at">fromLast =</span> T), ]</span>
<span id="cb67-162"><a href="#cb67-162" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-163"><a href="#cb67-163" aria-hidden="true" tabindex="-1"></a><span class="co"># now add the sample IDs to ifnb </span></span>
<span id="cb67-164"><a href="#cb67-164" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(info) <span class="ot">&lt;-</span> info<span class="sc">$</span>BARCODE</span>
<span id="cb67-165"><a href="#cb67-165" aria-hidden="true" tabindex="-1"></a>info <span class="ot">&lt;-</span> info[, <span class="fu">c</span>(<span class="st">"BEST"</span>), drop <span class="ot">=</span> F]</span>
<span id="cb67-166"><a href="#cb67-166" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(info) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"donor_id"</span>)</span>
<span id="cb67-167"><a href="#cb67-167" aria-hidden="true" tabindex="-1"></a>ifnb <span class="ot">&lt;-</span> <span class="fu">AddMetaData</span>(ifnb, <span class="at">metadata =</span> info)</span>
<span id="cb67-168"><a href="#cb67-168" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-169"><a href="#cb67-169" aria-hidden="true" tabindex="-1"></a><span class="co"># remove cells without donor IDs</span></span>
<span id="cb67-170"><a href="#cb67-170" aria-hidden="true" tabindex="-1"></a>ifnb<span class="sc">$</span>donor_id[<span class="fu">is.na</span>(ifnb<span class="sc">$</span>donor_id)] <span class="ot">&lt;-</span> <span class="st">"unknown"</span></span>
<span id="cb67-171"><a href="#cb67-171" aria-hidden="true" tabindex="-1"></a>ifnb <span class="ot">&lt;-</span> <span class="fu">subset</span>(ifnb, <span class="at">subset =</span> donor_id <span class="sc">!=</span> <span class="st">"unknown"</span>)</span>
<span id="cb67-172"><a href="#cb67-172" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb67-173"><a href="#cb67-173" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb67-174"><a href="#cb67-174" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-175"><a href="#cb67-175" aria-hidden="true" tabindex="-1"></a>可以看到，现在的meta.dat中多了样本信息列（<span class="in">`donor_id`</span>），记录了每个细胞来自哪个患者：</span>
<span id="cb67-176"><a href="#cb67-176" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-179"><a href="#cb67-179" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb67-180"><a href="#cb67-180" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(ifnb<span class="sc">@</span>meta.data, <span class="dv">5</span>)</span>
<span id="cb67-181"><a href="#cb67-181" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(ifnb<span class="sc">$</span>donor_id)</span>
<span id="cb67-182"><a href="#cb67-182" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb67-183"><a href="#cb67-183" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-184"><a href="#cb67-184" aria-hidden="true" tabindex="-1"></a><span class="fu">### 执行pseudobulking</span></span>
<span id="cb67-185"><a href="#cb67-185" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-186"><a href="#cb67-186" aria-hidden="true" tabindex="-1"></a>按照治疗分组（STIM vs. CTRL）、患者IDs、细胞类型（seurat_annotations）3个条件，执行pseudobulking (<span class="in">`AggregateExpression`</span>)。</span>
<span id="cb67-187"><a href="#cb67-187" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-190"><a href="#cb67-190" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb67-191"><a href="#cb67-191" aria-hidden="true" tabindex="-1"></a>pseudo_ifnb <span class="ot">&lt;-</span> <span class="fu">AggregateExpression</span>(ifnb, </span>
<span id="cb67-192"><a href="#cb67-192" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">assays =</span> <span class="st">"RNA"</span>, </span>
<span id="cb67-193"><a href="#cb67-193" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">return.seurat =</span> T, </span>
<span id="cb67-194"><a href="#cb67-194" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">group.by =</span> <span class="fu">c</span>(<span class="st">"stim"</span>, <span class="st">"donor_id"</span>, <span class="st">"seurat_annotations"</span>))</span>
<span id="cb67-195"><a href="#cb67-195" aria-hidden="true" tabindex="-1"></a>pseudo_ifnb</span>
<span id="cb67-196"><a href="#cb67-196" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(pseudo_ifnb<span class="sc">@</span>meta.data) <span class="co"># 可以看到现在的表达矩阵的列（即样本）为治疗分组+患者IDs+细胞类型</span></span>
<span id="cb67-197"><a href="#cb67-197" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb67-198"><a href="#cb67-198" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-199"><a href="#cb67-199" aria-hidden="true" tabindex="-1"></a>然后我们在meta.data中增加一列，记录治疗分组（STIM vs. CTRL）+ 细胞类型，这是用于差异分析的分组依据。这和 @sec-寻找不同样本类型间同一细胞类型内的差异基因 的第一步是一样的。</span>
<span id="cb67-200"><a href="#cb67-200" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-203"><a href="#cb67-203" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb67-204"><a href="#cb67-204" aria-hidden="true" tabindex="-1"></a>pseudo_ifnb<span class="sc">$</span>celltype.stim <span class="ot">&lt;-</span> <span class="fu">paste</span>(pseudo_ifnb<span class="sc">$</span>seurat_annotations, </span>
<span id="cb67-205"><a href="#cb67-205" aria-hidden="true" tabindex="-1"></a>                                   pseudo_ifnb<span class="sc">$</span>stim, </span>
<span id="cb67-206"><a href="#cb67-206" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">sep =</span> <span class="st">"_"</span>)</span>
<span id="cb67-207"><a href="#cb67-207" aria-hidden="true" tabindex="-1"></a>pseudo_ifnb<span class="sc">$</span>celltype.stim[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span>
<span id="cb67-208"><a href="#cb67-208" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb67-209"><a href="#cb67-209" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-210"><a href="#cb67-210" aria-hidden="true" tabindex="-1"></a><span class="fu">### 执行差异分析</span></span>
<span id="cb67-211"><a href="#cb67-211" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-212"><a href="#cb67-212" aria-hidden="true" tabindex="-1"></a>Next, we perform DE testing on the pseudobulk level for CD14 monocytes, and compare it against the previous single-cell-level DE results.</span>
<span id="cb67-213"><a href="#cb67-213" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-214"><a href="#cb67-214" aria-hidden="true" tabindex="-1"></a>::: {.callout-caution collapse="true"}</span>
<span id="cb67-215"><a href="#cb67-215" aria-hidden="true" tabindex="-1"></a><span class="fu">###### 安装`DESeq2`包</span></span>
<span id="cb67-216"><a href="#cb67-216" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-217"><a href="#cb67-217" aria-hidden="true" tabindex="-1"></a>由于pseudobulking后的<span class="in">`FindMarkers`</span>差异分析需要采用<span class="in">`DESeq2`</span>包提供的方法，所以需要提前安装<span class="in">`DESeq2`</span>包：</span>
<span id="cb67-218"><a href="#cb67-218" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-221"><a href="#cb67-221" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb67-222"><a href="#cb67-222" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb67-223"><a href="#cb67-223" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: fenced</span></span>
<span id="cb67-224"><a href="#cb67-224" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"DESeq2"</span>)</span>
<span id="cb67-225"><a href="#cb67-225" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb67-226"><a href="#cb67-226" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb67-227"><a href="#cb67-227" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-230"><a href="#cb67-230" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb67-231"><a href="#cb67-231" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(pseudo_ifnb) <span class="ot">&lt;-</span> <span class="st">"celltype.stim"</span></span>
<span id="cb67-232"><a href="#cb67-232" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-233"><a href="#cb67-233" aria-hidden="true" tabindex="-1"></a>bulk.mono.de <span class="ot">&lt;-</span> <span class="fu">FindMarkers</span>(<span class="at">object =</span> pseudo_ifnb, </span>
<span id="cb67-234"><a href="#cb67-234" aria-hidden="true" tabindex="-1"></a>                            <span class="at">ident.1 =</span> <span class="st">"CD14 Mono_STIM"</span>, </span>
<span id="cb67-235"><a href="#cb67-235" aria-hidden="true" tabindex="-1"></a>                            <span class="at">ident.2 =</span> <span class="st">"CD14 Mono_CTRL"</span>,</span>
<span id="cb67-236"><a href="#cb67-236" aria-hidden="true" tabindex="-1"></a>                            <span class="at">test.use =</span> <span class="st">"DESeq2"</span>) <span class="co"># 指定差异分析方法为"DESeq2"</span></span>
<span id="cb67-237"><a href="#cb67-237" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(bulk.mono.de, <span class="at">n =</span> <span class="dv">15</span>)</span>
<span id="cb67-238"><a href="#cb67-238" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb67-239"><a href="#cb67-239" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-240"><a href="#cb67-240" aria-hidden="true" tabindex="-1"></a>::: {.callout-tip collapse="true"}</span>
<span id="cb67-241"><a href="#cb67-241" aria-hidden="true" tabindex="-1"></a><span class="fu">###### `FindMarkers`支持的差异分析方法</span></span>
<span id="cb67-242"><a href="#cb67-242" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-243"><a href="#cb67-243" aria-hidden="true" tabindex="-1"></a>We also support many other DE tests using other methods. For completeness, the following tests are currently supported:</span>
<span id="cb67-244"><a href="#cb67-244" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-245"><a href="#cb67-245" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>"wilcox" : Wilcoxon rank sum test (default, using '<span class="co">[</span><span class="ot">presto</span><span class="co">](https://github.com/immunogenomics/presto)</span>' package)</span>
<span id="cb67-246"><a href="#cb67-246" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-247"><a href="#cb67-247" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>"wilcox_limma" : Wilcoxon rank sum test (using '<span class="co">[</span><span class="ot">limma</span><span class="co">](https://bioconductor.org/packages/release/bioc/html/limma.html)</span>' package)</span>
<span id="cb67-248"><a href="#cb67-248" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-249"><a href="#cb67-249" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>"bimod" : Likelihood-ratio test for single cell feature expression, (McDavid et al., Bioinformatics, 2013)</span>
<span id="cb67-250"><a href="#cb67-250" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-251"><a href="#cb67-251" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>"roc" : Standard AUC classifier</span>
<span id="cb67-252"><a href="#cb67-252" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-253"><a href="#cb67-253" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>"t" : Student's t-test</span>
<span id="cb67-254"><a href="#cb67-254" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-255"><a href="#cb67-255" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>"poisson" : Likelihood ratio test assuming an underlying negative binomial distribution. Use only for UMI-based datasets</span>
<span id="cb67-256"><a href="#cb67-256" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-257"><a href="#cb67-257" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>"negbinom" : Likelihood ratio test assuming an underlying negative binomial distribution. Use only for UMI-based datasets</span>
<span id="cb67-258"><a href="#cb67-258" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-259"><a href="#cb67-259" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>"LR" : Uses a logistic regression framework to determine differentially expressed genes. Constructs a logistic regression model predicting group membership based on each feature individually and compares this to a null model with a likelihood ratio test.</span>
<span id="cb67-260"><a href="#cb67-260" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-261"><a href="#cb67-261" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>"MAST" : GLM-framework that treates cellular detection rate as a covariate (Finak et al, Genome Biology, 2015) (Installation instructions)</span>
<span id="cb67-262"><a href="#cb67-262" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-263"><a href="#cb67-263" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>"DESeq2" : DE based on a model using the negative binomial distribution (Love et al, Genome Biology, 2014) (Installation instructions) For MAST and DESeq2, please ensure that these packages are installed separately in order to use them as part of Seurat. Once installed, use the test.use parameter can be used to specify which DE test to use.</span>
<span id="cb67-264"><a href="#cb67-264" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-267"><a href="#cb67-267" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb67-268"><a href="#cb67-268" aria-hidden="true" tabindex="-1"></a><span class="co"># Test for DE features using the MAST package</span></span>
<span id="cb67-269"><a href="#cb67-269" aria-hidden="true" tabindex="-1"></a><span class="co"># BiocManager::install('limma')</span></span>
<span id="cb67-270"><a href="#cb67-270" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(ifnb) <span class="ot">&lt;-</span> <span class="st">"seurat_annotations"</span></span>
<span id="cb67-271"><a href="#cb67-271" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">FindMarkers</span>(ifnb, </span>
<span id="cb67-272"><a href="#cb67-272" aria-hidden="true" tabindex="-1"></a>                 <span class="at">ident.1 =</span> <span class="st">"CD14 Mono"</span>, </span>
<span id="cb67-273"><a href="#cb67-273" aria-hidden="true" tabindex="-1"></a>                 <span class="at">ident.2 =</span> <span class="st">"CD16 Mono"</span>, </span>
<span id="cb67-274"><a href="#cb67-274" aria-hidden="true" tabindex="-1"></a>                 <span class="at">test.use =</span> <span class="st">"wilcox_limma"</span>))</span>
<span id="cb67-275"><a href="#cb67-275" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb67-276"><a href="#cb67-276" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb67-277"><a href="#cb67-277" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-278"><a href="#cb67-278" aria-hidden="true" tabindex="-1"></a><span class="fu">### 比较单细胞水平和pseudobulk水平的差异表达分析</span></span>
<span id="cb67-279"><a href="#cb67-279" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-280"><a href="#cb67-280" aria-hidden="true" tabindex="-1"></a>接下来，我们可以比较一下单细胞水平的差异表达分析的P值和pseudobulk水平的P值：</span>
<span id="cb67-281"><a href="#cb67-281" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-284"><a href="#cb67-284" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb67-285"><a href="#cb67-285" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(bulk.mono.de) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="fu">names</span>(bulk.mono.de), <span class="st">".bulk"</span>) <span class="co"># 重命名列</span></span>
<span id="cb67-286"><a href="#cb67-286" aria-hidden="true" tabindex="-1"></a>bulk.mono.de<span class="sc">$</span>gene <span class="ot">&lt;-</span> <span class="fu">rownames</span>(bulk.mono.de)</span>
<span id="cb67-287"><a href="#cb67-287" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-288"><a href="#cb67-288" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(mono.de) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="fu">names</span>(mono.de), <span class="st">".sc"</span>)</span>
<span id="cb67-289"><a href="#cb67-289" aria-hidden="true" tabindex="-1"></a>mono.de<span class="sc">$</span>gene <span class="ot">&lt;-</span> <span class="fu">rownames</span>(mono.de)</span>
<span id="cb67-290"><a href="#cb67-290" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-291"><a href="#cb67-291" aria-hidden="true" tabindex="-1"></a>merge_dat <span class="ot">&lt;-</span> <span class="fu">merge</span>(mono.de, bulk.mono.de, <span class="at">by =</span> <span class="st">"gene"</span>)</span>
<span id="cb67-292"><a href="#cb67-292" aria-hidden="true" tabindex="-1"></a>merge_dat <span class="ot">&lt;-</span> merge_dat[<span class="fu">order</span>(merge_dat<span class="sc">$</span>p_val.bulk), ]</span>
<span id="cb67-293"><a href="#cb67-293" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-294"><a href="#cb67-294" aria-hidden="true" tabindex="-1"></a><span class="co"># 查看在两种差异分析方法中P值都有意义的基因名</span></span>
<span id="cb67-295"><a href="#cb67-295" aria-hidden="true" tabindex="-1"></a>common <span class="ot">&lt;-</span> merge_dat<span class="sc">$</span>gene[<span class="fu">which</span>(merge_dat<span class="sc">$</span>p_val.bulk <span class="sc">&lt;</span> <span class="fl">0.05</span> <span class="sc">&amp;</span> </span>
<span id="cb67-296"><a href="#cb67-296" aria-hidden="true" tabindex="-1"></a>                                merge_dat<span class="sc">$</span>p_val.sc <span class="sc">&lt;</span> <span class="fl">0.05</span>)]</span>
<span id="cb67-297"><a href="#cb67-297" aria-hidden="true" tabindex="-1"></a><span class="co"># 查看在pseudobulk水平P&gt;0.05但是在单细胞水平P&lt;0.05的基因名：</span></span>
<span id="cb67-298"><a href="#cb67-298" aria-hidden="true" tabindex="-1"></a>only_sc <span class="ot">&lt;-</span> merge_dat<span class="sc">$</span>gene[<span class="fu">which</span>(merge_dat<span class="sc">$</span>p_val.bulk <span class="sc">&gt;</span> <span class="fl">0.05</span> <span class="sc">&amp;</span> </span>
<span id="cb67-299"><a href="#cb67-299" aria-hidden="true" tabindex="-1"></a>                                  merge_dat<span class="sc">$</span>p_val.sc <span class="sc">&lt;</span> <span class="fl">0.05</span>)]</span>
<span id="cb67-300"><a href="#cb67-300" aria-hidden="true" tabindex="-1"></a><span class="co"># 查看在pseudobulk水平P&lt;0.05但是在单细胞水平P&gt;0.05的基因名：</span></span>
<span id="cb67-301"><a href="#cb67-301" aria-hidden="true" tabindex="-1"></a>only_bulk <span class="ot">&lt;-</span> merge_dat<span class="sc">$</span>gene[<span class="fu">which</span>(merge_dat<span class="sc">$</span>p_val.bulk <span class="sc">&lt;</span> <span class="fl">0.05</span> <span class="sc">&amp;</span> </span>
<span id="cb67-302"><a href="#cb67-302" aria-hidden="true" tabindex="-1"></a>                                    merge_dat<span class="sc">$</span>p_val.sc <span class="sc">&gt;</span> <span class="fl">0.05</span>)]</span>
<span id="cb67-303"><a href="#cb67-303" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste0</span>(<span class="st">'# 在两种差异分析方法中P值都&lt;0.05的基因有: '</span>,<span class="fu">length</span>(common), <span class="st">"个"</span>))</span>
<span id="cb67-304"><a href="#cb67-304" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste0</span>(<span class="st">'# 仅在单细胞水平差异分析中P值&lt;0.05的基因有: '</span>,<span class="fu">length</span>(only_sc), <span class="st">"个"</span>))</span>
<span id="cb67-305"><a href="#cb67-305" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste0</span>(<span class="st">'# 仅在pseudobulk水平差异分析中P值&lt;0.05的基因有: '</span>,<span class="fu">length</span>(only_bulk), <span class="st">"个"</span>))</span>
<span id="cb67-306"><a href="#cb67-306" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb67-307"><a href="#cb67-307" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-308"><a href="#cb67-308" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; We can see that while the p-values are correlated between the single-cell and pseudobulk data, the </span><span class="co">[</span><span class="ot">**single-cell p-values are often smaller**</span><span class="co">]</span><span class="at">{.underline} and suggest higher levels of significance. In particular, there are 3,519 genes with evidence of differential expression (prior to multiple hypothesis testing) in both analyses, 1,649 genes that only appear to be differentially expressed in the single-cell analysis, and just 204 genes that only appear to be differentially expressed in the bulk analysis. We can investigate these discrepancies using&nbsp;</span><span class="in">`VlnPlot`</span><span class="at">.</span></span>
<span id="cb67-309"><a href="#cb67-309" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-310"><a href="#cb67-310" aria-hidden="true" tabindex="-1"></a>接下来，我们通过小提琴图来检查两种方法中的Top共同差异基因在在刺激组和对照组的表达水平：</span>
<span id="cb67-311"><a href="#cb67-311" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-314"><a href="#cb67-314" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb67-315"><a href="#cb67-315" aria-hidden="true" tabindex="-1"></a><span class="co"># create a new column to annotate sample-condition-celltype in the single-cell dataset</span></span>
<span id="cb67-316"><a href="#cb67-316" aria-hidden="true" tabindex="-1"></a>ifnb<span class="sc">$</span>donor_id.stim <span class="ot">&lt;-</span> <span class="fu">paste0</span>(ifnb<span class="sc">$</span>stim, <span class="st">"-"</span>, ifnb<span class="sc">$</span>donor_id)</span>
<span id="cb67-317"><a href="#cb67-317" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(ifnb<span class="sc">@</span>meta.data)</span>
<span id="cb67-318"><a href="#cb67-318" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(ifnb<span class="sc">$</span>celltype.stim)</span>
<span id="cb67-319"><a href="#cb67-319" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-320"><a href="#cb67-320" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(ifnb) <span class="ot">&lt;-</span> <span class="st">"celltype.stim"</span></span>
<span id="cb67-321"><a href="#cb67-321" aria-hidden="true" tabindex="-1"></a><span class="co"># 这里我们检查p_val.bulk最小的前两个Top差异基因</span></span>
<span id="cb67-322"><a href="#cb67-322" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(merge_dat[merge_dat<span class="sc">$</span>gene <span class="sc">%in%</span> common[<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>], <span class="fu">c</span>(<span class="st">'gene'</span>,<span class="st">'p_val.sc'</span>,<span class="st">'p_val.bulk'</span>)])</span>
<span id="cb67-323"><a href="#cb67-323" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-324"><a href="#cb67-324" aria-hidden="true" tabindex="-1"></a><span class="co"># 在细胞类型水平（CD14 Mono）查看这两个Top差异基因在刺激组和对照组的表达水平</span></span>
<span id="cb67-325"><a href="#cb67-325" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(ifnb, </span>
<span id="cb67-326"><a href="#cb67-326" aria-hidden="true" tabindex="-1"></a>        <span class="at">features =</span> common[<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>], </span>
<span id="cb67-327"><a href="#cb67-327" aria-hidden="true" tabindex="-1"></a>        <span class="at">idents =</span> <span class="fu">c</span>(<span class="st">"CD14 Mono_CTRL"</span>, <span class="st">"CD14 Mono_STIM"</span>), </span>
<span id="cb67-328"><a href="#cb67-328" aria-hidden="true" tabindex="-1"></a>        <span class="at">group.by =</span> <span class="st">"stim"</span>) </span>
<span id="cb67-329"><a href="#cb67-329" aria-hidden="true" tabindex="-1"></a><span class="co"># 在样本（患者）水平查看这两个Top差异基因在刺激组和对照组的表达水平</span></span>
<span id="cb67-330"><a href="#cb67-330" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(ifnb, </span>
<span id="cb67-331"><a href="#cb67-331" aria-hidden="true" tabindex="-1"></a>        <span class="at">features =</span> common[<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>], </span>
<span id="cb67-332"><a href="#cb67-332" aria-hidden="true" tabindex="-1"></a>        <span class="at">idents =</span> <span class="fu">c</span>(<span class="st">"CD14 Mono_CTRL"</span>, <span class="st">"CD14 Mono_STIM"</span>), </span>
<span id="cb67-333"><a href="#cb67-333" aria-hidden="true" tabindex="-1"></a>        <span class="at">group.by =</span> <span class="st">"donor_id.stim"</span>, </span>
<span id="cb67-334"><a href="#cb67-334" aria-hidden="true" tabindex="-1"></a>        <span class="at">ncol =</span> <span class="dv">1</span>) </span>
<span id="cb67-335"><a href="#cb67-335" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb67-336"><a href="#cb67-336" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-337"><a href="#cb67-337" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; In both the pseudobulk and single-cell analyses, the p-values for these two genes are astronomically small. For both of these genes, when just comparing all stimulated CD4 monocytes to all control CD4 monocytes across samples, we see much higher expression in the stimulated cells. When breaking down these cells by sample, we continue to see consistently higher expression levels in the stimulated samples compared to the control samples; in other words, this finding is not driven by just one or two samples. Because of this consistency, we find this signal in both analyses.</span></span>
<span id="cb67-338"><a href="#cb67-338" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-339"><a href="#cb67-339" aria-hidden="true" tabindex="-1"></a>By contrast, we can examine examples of genes that are only DE under the single-cell analysis.</span>
<span id="cb67-340"><a href="#cb67-340" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-343"><a href="#cb67-343" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb67-344"><a href="#cb67-344" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(merge_dat[merge_dat<span class="sc">$</span>gene <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">'SRGN'</span>,<span class="st">'HLA-DRA'</span>), </span>
<span id="cb67-345"><a href="#cb67-345" aria-hidden="true" tabindex="-1"></a>                <span class="fu">c</span>(<span class="st">'gene'</span>,<span class="st">'p_val.sc'</span>,<span class="st">'p_val.bulk'</span>)])</span>
<span id="cb67-346"><a href="#cb67-346" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(ifnb, </span>
<span id="cb67-347"><a href="#cb67-347" aria-hidden="true" tabindex="-1"></a>        <span class="at">features =</span> <span class="fu">c</span>(<span class="st">'SRGN'</span>,<span class="st">'HLA-DRA'</span>), </span>
<span id="cb67-348"><a href="#cb67-348" aria-hidden="true" tabindex="-1"></a>        <span class="at">idents =</span> <span class="fu">c</span>(<span class="st">"CD14 Mono_CTRL"</span>, <span class="st">"CD14 Mono_STIM"</span>), </span>
<span id="cb67-349"><a href="#cb67-349" aria-hidden="true" tabindex="-1"></a>        <span class="at">group.by =</span> <span class="st">"stim"</span>) </span>
<span id="cb67-350"><a href="#cb67-350" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(ifnb, </span>
<span id="cb67-351"><a href="#cb67-351" aria-hidden="true" tabindex="-1"></a>        <span class="at">features =</span> <span class="fu">c</span>(<span class="st">'SRGN'</span>,<span class="st">'HLA-DRA'</span>), </span>
<span id="cb67-352"><a href="#cb67-352" aria-hidden="true" tabindex="-1"></a>        <span class="at">idents =</span> <span class="fu">c</span>(<span class="st">"CD14 Mono_CTRL"</span>, <span class="st">"CD14 Mono_STIM"</span>), </span>
<span id="cb67-353"><a href="#cb67-353" aria-hidden="true" tabindex="-1"></a>        <span class="at">group.by =</span> <span class="st">"donor_id.stim"</span>, </span>
<span id="cb67-354"><a href="#cb67-354" aria-hidden="true" tabindex="-1"></a>        <span class="at">ncol =</span> <span class="dv">1</span>) </span>
<span id="cb67-355"><a href="#cb67-355" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb67-356"><a href="#cb67-356" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-357"><a href="#cb67-357" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; Here, SRGN and HLA-DRA both have very small p-values in the single-cell analysis (on the orders of&nbsp;10^−21^&nbsp;and&nbsp;10^−9^), but much larger p-values around 0.18 in the pseudobulk analysis. While there appears to be a difference between control and simulated cells when ignoring sample information, the signal is much weaker on the sample level, and we can see notable variability from sample to sample.</span></span>
<span id="cb67-358"><a href="#cb67-358" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-359"><a href="#cb67-359" aria-hidden="true" tabindex="-1"></a>所以，从这个例子中可以看出**pseudobulk后的差异分析的结果更加准确**。</span>
<span id="cb67-360"><a href="#cb67-360" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-361"><a href="#cb67-361" aria-hidden="true" tabindex="-1"></a>::: {.callout-note collapse="true" icon="false"}</span>
<span id="cb67-362"><a href="#cb67-362" aria-hidden="true" tabindex="-1"></a><span class="fu">## Session Info</span></span>
<span id="cb67-363"><a href="#cb67-363" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-366"><a href="#cb67-366" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb67-367"><a href="#cb67-367" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb67-368"><a href="#cb67-368" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span>
<span id="cb67-369"><a href="#cb67-369" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb67-370"><a href="#cb67-370" aria-hidden="true" tabindex="-1"></a>:::</span>
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">
      <ul class="footer-items list-unstyled">
<li class="nav-item">
 This book was built with Quarto
  </li>  
</ul>
</div>   
    <div class="nav-footer-center">Copyright 2023, Du Junhong</div>
    <div class="nav-footer-right">
      <ul class="footer-items list-unstyled">
<li class="nav-item compact">
    <a class="nav-link" href="https://github.com/djhcod">
      <i class="bi bi-github" role="img">
</i> 
    </a>
  </li>  
</ul>
</div>
  </div>
</footer>


</body></html>