<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>R语言学习笔记 - 12&nbsp; 整合（integration）</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../single_cell/seurat/integrative_analysis_in_seurat_v5.html" rel="next">
<link href="../../single_cell/seurat/cell_cycle_regression.html" rel="prev">
<link href="../../images/logo.jpg" rel="icon" type="image/jpeg">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark"><script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><script type="application/json" class="js-hypothesis-config">
{
  "theme": "classic",
  "openSidebar": false,
  "showHighlights": "always",
  "enableExperimentalNewNoteButton": true
}
</script><script async="" src="https://hypothes.is/embed.js"></script>
<meta name="twitter:title" content="R语言学习笔记 - 12&nbsp; 整合（integration）">
<meta name="twitter:description" content="">
<meta name="twitter:image" content="https://djhcod.github.io/r-notes/single_cell/seurat/images/截屏2023-11-27 12.49.37.png">
<meta name="twitter:image-height" content="330">
<meta name="twitter:image-width" content="1732">
<meta name="twitter:card" content="summary_large_image">
</head>
<body class="nav-sidebar docked nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="navbar navbar-expand-lg navbar-dark "><div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a href="../../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../../images/logo.jpg" alt="" class="navbar-logo"></a>
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">R语言学习笔记</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
<li class="nav-item">
    <a class="nav-link active" href="../../index.html" rel="" target="" aria-current="page">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../intro.html" rel="" target="">
 <span class="menu-text">Introduction</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-parts" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Parts</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-parts">
<li>
    <a class="dropdown-item" href="../../r_basic/r_basics.html" rel="" target="">
 <span class="dropdown-text">R语言基础</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../single_cell/seurat/seurat.html" rel="" target="">
 <span class="dropdown-text">Seurat v5官方文档</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../single_cell/scRNA-seq_online/00_intro.html" rel="" target="">
 <span class="dropdown-text">scRNA-seq_online学习材料</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../single_cell/sc_supplementary/read_sc_data.html" rel="" target="">
 <span class="dropdown-text">单细胞补充内容</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../quarto_foundation/quarto_foundation.html" rel="" target="">
 <span class="dropdown-text">Quarto基础</span></a>
  </li>  
    </ul>
</li>
  <li class="nav-item">
    <a class="nav-link" href="../../references.html" rel="" target="">
 <span class="menu-text">References</span></a>
  </li>  
</ul>
<ul class="navbar-nav navbar-nav-scroll ms-auto">
<li class="nav-item">
    <a class="nav-link" href="https://bookdown.org" rel="" target=""><i class="bi bi-book-fill" role="img">
</i> 
 <span class="menu-text">Bookdown</span></a>
  </li>  
</ul>
<div class="quarto-navbar-tools tools-wide">
    <a href="https://github.com/djhcod/r-notes" rel="" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
    <div class="dropdown">
      <a href="" title="Share" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Share"><i class="bi bi-share"></i></a>
      <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="quarto-navigation-tool-dropdown-0">
<li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://twitter.com/intent/tweet?url=%7Curl%7C">
              <i class="bi bi-bi-twitter pe-1"></i>
            Twitter
            </a>
          </li>
          <li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://www.facebook.com/sharer/sharer.php?u=%7Curl%7C">
              <i class="bi bi-bi-facebook pe-1"></i>
            Facebook
            </a>
          </li>
      </ul>
</div>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../single_cell/seurat/seurat.html">Seurat v5官方文档</a></li><li class="breadcrumb-item"><a href="../../single_cell/seurat/integration.html"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">整合（integration）</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto"><div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">主页</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">前言</span></a>
  </div>
</li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../r_basic/r_basics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">R语言基础</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/environment_configuration.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Rstudio环境配置</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/data_input_output.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">数据的读取与输出</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/basic_data_function.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">数据处理基本函数</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/character.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">字符的处理</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/loop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">循环</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../single_cell/seurat/seurat.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Seurat v5官方文档</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/seurat_command_list.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Seurat常用函数清单</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/seurat_tutorial.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Seurat细胞分群官方教程</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/data_visualization_methods_in_seurat.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Seurat中的数据可视化方法</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/sctransform.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">基于SCTransform的单细胞数据标准化</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/cell_cycle_regression.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">消除细胞周期的影响</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/integration.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">整合（integration）</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/integrative_analysis_in_seurat_v5.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Seurat v5单细胞数据整合分析</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/marker_gene_identification.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">寻找marker gene</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/de_vignette.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">寻找不同样本类型间同一细胞类型内的差异基因</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/mapping_and_annotating_query_datasets.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">基于参考集的细胞注释</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../single_cell/scRNA-seq_online/00_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">scRNA-seq_online学习材料</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/01_intro_to_scRNA-seq.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">单细胞测序技术介绍</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/02_SC_generation_of_count_matrix.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">单细胞RNA测序—从raw data到count matrix</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/03_SC_quality_control-setup.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">数据导入与Seurat对象构建</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/04_SC_quality_control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">质控</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/postQC_workflow.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">21</span>&nbsp; <span class="chapter-title">Single-cell RNA-seq Clustering Workflow</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/05_theory_of_PCA.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">22</span>&nbsp; <span class="chapter-title">PCA原理</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/06_SC_SCT_normalization.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">23</span>&nbsp; <span class="chapter-title">Normalization and regressing out unwanted variation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/06_integration.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">24</span>&nbsp; <span class="chapter-title">单细胞数据整合（Integration）</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/07_SC_clustering_cells_SCT.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">25</span>&nbsp; <span class="chapter-title">细胞聚类（clustering analysis）</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/08_SC_clustering_quality_control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">26</span>&nbsp; <span class="chapter-title">细胞分群质量评估</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/09_merged_SC_marker_identification.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">27</span>&nbsp; <span class="chapter-title">寻找marker基因+细胞注释</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="false">
 <span class="menu-text">单细胞补充内容</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/sc_supplementary/read_sc_data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">28</span>&nbsp; <span class="chapter-title"><strong>读取非标准格式的单细胞数据</strong></span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/sc_supplementary/universal_marker.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">29</span>&nbsp; <span class="chapter-title">细胞分群通用marker</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/sc_supplementary/integrated_analysis_multiple_single_cell_datasets.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">30</span>&nbsp; <span class="chapter-title">多个单细胞数据集整合分析</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../quarto_foundation/quarto_foundation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Quarto基础</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/yaml_settings.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">31</span>&nbsp; <span class="chapter-title">YAML设置</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/images_settings.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">32</span>&nbsp; <span class="chapter-title">图片设置</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/cross_references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">33</span>&nbsp; <span class="chapter-title">交叉引用</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/insert_other_content.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">34</span>&nbsp; <span class="chapter-title">插入其他内容</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/quarto_shiny.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">35</span>&nbsp; <span class="chapter-title">在Quarto中嵌入Shiny应用程序</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/quarto_books.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">36</span>&nbsp; <span class="chapter-title">Quarto Books</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/github_pages.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">37</span>&nbsp; <span class="chapter-title">发布到GitHub Pages</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="false">
 <span class="menu-text">参考文献</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active" data-toc-expanded="1"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#%E6%95%B0%E6%8D%AE%E8%AF%BB%E5%8F%96%E5%92%8C%E5%88%86%E5%B1%82" id="toc-数据读取和分层" class="nav-link active" data-scroll-target="#%E6%95%B0%E6%8D%AE%E8%AF%BB%E5%8F%96%E5%92%8C%E5%88%86%E5%B1%82"><span class="header-section-number">12.1</span> 数据读取和分层</a></li>
  <li><a href="#%E4%B8%8D%E8%BF%9B%E8%A1%8C%E6%95%B4%E5%90%88%E7%9A%84%E6%83%85%E5%86%B5%E4%B8%8B%E7%9A%84%E6%95%B0%E6%8D%AE%E5%A4%84%E7%90%86" id="toc-不进行整合的情况下的数据处理" class="nav-link" data-scroll-target="#%E4%B8%8D%E8%BF%9B%E8%A1%8C%E6%95%B4%E5%90%88%E7%9A%84%E6%83%85%E5%86%B5%E4%B8%8B%E7%9A%84%E6%95%B0%E6%8D%AE%E5%A4%84%E7%90%86"><span class="header-section-number">12.2</span> 不进行整合的情况下的数据处理</a></li>
  <li>
<a href="#%E8%BF%9B%E8%A1%8C%E6%95%B0%E6%8D%AE%E6%95%B4%E5%90%88" id="toc-进行数据整合" class="nav-link" data-scroll-target="#%E8%BF%9B%E8%A1%8C%E6%95%B0%E6%8D%AE%E6%95%B4%E5%90%88"><span class="header-section-number">12.3</span> 进行数据整合</a>
  <ul class="collapse">
<li><a href="#%E6%95%B4%E5%90%88%E5%90%8E%E9%87%8D%E6%96%B0%E8%81%9A%E7%B1%BB%E9%99%8D%E7%BB%B4" id="toc-整合后重新聚类降维" class="nav-link" data-scroll-target="#%E6%95%B4%E5%90%88%E5%90%8E%E9%87%8D%E6%96%B0%E8%81%9A%E7%B1%BB%E9%99%8D%E7%BB%B4">整合后重新聚类、降维</a></li>
  </ul>
</li>
  <li>
<a href="#%E6%89%A7%E8%A1%8Csctransform%E6%A0%87%E5%87%86%E5%8C%96%E6%B5%81%E7%A8%8B%E4%B9%8B%E5%90%8E%E7%9A%84%E6%95%B4%E5%90%88" id="toc-执行sctransform标准化流程之后的整合" class="nav-link" data-scroll-target="#%E6%89%A7%E8%A1%8Csctransform%E6%A0%87%E5%87%86%E5%8C%96%E6%B5%81%E7%A8%8B%E4%B9%8B%E5%90%8E%E7%9A%84%E6%95%B4%E5%90%88"><span class="header-section-number">12.4</span> 执行<code>SCTransform</code>标准化流程之后的整合</a>
  <ul class="collapse">
<li><a href="#%E4%B8%8D%E8%BF%9B%E8%A1%8C%E6%95%B4%E5%90%88%E7%9A%84%E6%83%85%E5%86%B5%E4%B8%8B%E7%9A%84%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90" id="toc-不进行整合的情况下的数据分析" class="nav-link" data-scroll-target="#%E4%B8%8D%E8%BF%9B%E8%A1%8C%E6%95%B4%E5%90%88%E7%9A%84%E6%83%85%E5%86%B5%E4%B8%8B%E7%9A%84%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90">不进行整合的情况下的数据分析</a></li>
  <li><a href="#sec-integration_after_sct" id="toc-sec-integration_after_sct" class="nav-link" data-scroll-target="#sec-integration_after_sct">进行整合</a></li>
  <li><a href="#%E6%95%B4%E5%90%88%E5%90%8E%E8%81%9A%E7%B1%BB" id="toc-整合后聚类" class="nav-link" data-scroll-target="#%E6%95%B4%E5%90%88%E5%90%8E%E8%81%9A%E7%B1%BB">整合后聚类</a></li>
  <li><a href="#%E4%BF%9D%E5%AD%98%E6%95%B4%E5%90%88%E5%90%8E%E7%9A%84seurat%E5%AF%B9%E8%B1%A1" id="toc-保存整合后的seurat对象" class="nav-link" data-scroll-target="#%E4%BF%9D%E5%AD%98%E6%95%B4%E5%90%88%E5%90%8E%E7%9A%84seurat%E5%AF%B9%E8%B1%A1">保存整合后的Seurat对象</a></li>
  </ul>
</li>
  </ul><div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/djhcod/r-notes/blob/main/single_cell/seurat/integration.qmd" class="toc-action">View source</a></p><p><a href="https://github.com/djhcod/r-notes/issues/new" class="toc-action">Report an issue</a></p><p><a href="https://github.com/djhcod/r-notes/edit/main/single_cell/seurat/integration.qmd" class="toc-action">Edit this page</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title"><span id="sec-Single-cell_data_integration" class="quarto-section-identifier"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">整合（integration）</span></span></h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><blockquote class="blockquote">
<p>原文：<a href="https://satijalab.org/seurat/articles/integration_introduction"><em>Introduction to scRNA-seq integration</em></a></p>
<p>原文发布日期：2023年10月31日</p>
</blockquote>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>本篇主要介绍来自不同样本类型的单细胞数据的整合。对于如何整合不同测序技术的单细胞数据集，参考Seurat官方文档：<a href="https://satijalab.org/seurat/articles/seurat5_integration"><em>Integrative analysis in Seurat v5</em></a>。</p>
</div>
</div>
<p>Integration of single-cell sequencing datasets, for example across experimental batches, donors, or conditions, is often an important step in scRNA-seq workflows. Integrative analysis can help to <strong>match shared cell types and states across datasets</strong>, which can boost statistical power, and most importantly, facilitate accurate comparative analysis across datasets. In previous versions of Seurat we introduced methods for integrative analysis, including our ‘anchor-based’ integration workflow. Many labs have also published powerful and pioneering methods, including&nbsp;<a href="https://portals.broadinstitute.org/harmony/">Harmony</a>&nbsp;and&nbsp;<a href="https://docs.scvi-tools.org/en/stable/index.html">scVI</a>, for integrative analysis.</p>
<p>数据整合的目标：</p>
<p>The following tutorial is designed to give you an overview of the kinds of comparative analyses on complex cell types that are possible using the Seurat integration procedure. Here, we address a few key goals:</p>
<ul>
<li><p>Identify <strong>cell subpopulations</strong> that are present in both datasets</p></li>
<li><p>Obtain <strong>cell type markers</strong> that are conserved in both control and stimulated cells</p></li>
<li><p><strong>Compare the datasets</strong> to find cell-type specific responses to stimulation</p></li>
</ul>
<section id="数据读取和分层" class="level2" data-number="12.1"><h2 data-number="12.1" class="anchored" data-anchor-id="数据读取和分层">
<span class="header-section-number">12.1</span> 数据读取和分层</h2>
<div class="callout callout-style-simple callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-2-contents" aria-controls="callout-2" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
在线读取（可能需要全局代理）
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-2" class="callout-2-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-hash="integration_cache/html/unnamed-chunk-1_4259f900996d24c302bead7d4bd5c2b3">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SeuratData)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">InstallData</span>(<span class="st">"ifnb"</span>)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>ifnb <span class="ot">&lt;-</span> <span class="fu">LoadData</span>(<span class="st">"ifnb"</span>)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<p>从本地下载好的数据读取：</p>
<div class="cell" data-hash="integration_cache/html/unnamed-chunk-2_76f2aa1541f13daa997395432581125e">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span></span>
<span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="st">"data/seurat_official/pbmc_ifnb.rds"</span><span class="op">)</span></span>
<span><span class="va">ifnb</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
14053 features across 13999 samples within 1 assay 
Active assay: RNA (14053 features, 0 variable features)
 2 layers present: counts, data</code></pre>
</div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">ifnb</span><span class="op">@</span><span class="va">meta.data</span>, <span class="fl">5</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                  orig.ident nCount_RNA nFeature_RNA stim seurat_annotations
AAACATACATTTCC.1 IMMUNE_CTRL       3017          877 CTRL          CD14 Mono
AAACATACCAGAAA.1 IMMUNE_CTRL       2481          713 CTRL          CD14 Mono
AAACATACCTCGCT.1 IMMUNE_CTRL       3420          850 CTRL          CD14 Mono
AAACATACCTGGTA.1 IMMUNE_CTRL       3156         1109 CTRL                pDC
AAACATACGATGAA.1 IMMUNE_CTRL       1868          634 CTRL       CD4 Memory T</code></pre>
</div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">ifnb</span><span class="op">$</span><span class="va">stim</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
CTRL STIM 
6548 7451 </code></pre>
</div>
</div>
<p><img src="images/截屏2023-11-27 12.49.37.png" class="img-fluid"></p>
<p>The object contains data from human PBMC from two conditions, <strong>interferon-stimulated</strong> and <strong>control</strong> cells (stored in the&nbsp;<code>stim</code> column in the object metadata). We will aim to integrate the two conditions together, so that we can jointly identify cell subpopulations across datasets, and then explore how each group differs across conditions</p>
<p>In previous versions of Seurat, we would require the data to be represented as two different Seurat objects. <strong>In Seurat v5, we keep all the data in one object, but simply split it into multiple ‘layers’</strong>. To learn more about layers, check out our&nbsp;<a href="https://satijalab.org/seurat/articles/interaction_vignette">Seurat object interaction vignette</a>.</p>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Important
</div>
</div>
<div class="callout-body-container callout-body">
<p>Seurat v5 assays store data in <strong>layers</strong>. These layers can store:</p>
<ul>
<li><p>raw, un-normalized counts (<code>layer='counts'</code>)</p></li>
<li><p>normalized data (<code>layer='data'</code>)</p></li>
<li><p>z-scored/variance-stabilized data (<code>layer='scale.data'</code>).</p></li>
</ul>
</div>
</div>
<p><strong>split the RNA measurements into two layers one for control cells, one for stimulated cells:</strong></p>
<div class="cell" data-hash="integration_cache/html/unnamed-chunk-3_cdf052a9c3657bcf3f255f18204908cd">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span></span>
<span><span class="va">ifnb</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html">split</a></span><span class="op">(</span><span class="va">ifnb</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span>, </span>
<span>                       f <span class="op">=</span> <span class="va">ifnb</span><span class="op">$</span><span class="va">stim</span><span class="op">)</span> <span class="co"># 按照meta.data中的“stim”列进行分割</span></span>
<span><span class="va">ifnb</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
14053 features across 13999 samples within 1 assay 
Active assay: RNA (14053 features, 0 variable features)
 4 layers present: counts.CTRL, counts.STIM, data.CTRL, data.STIM</code></pre>
</div>
</div>
<p>现在可以发现<code>ifnb</code>被分为了4个layer，此前是2个layer（<code>counts</code>和<code>data</code>）：</p>
<p><img src="images/截屏2023-11-27 11.55.48.png" class="img-fluid"></p>
</section><section id="不进行整合的情况下的数据处理" class="level2" data-number="12.2"><h2 data-number="12.2" class="anchored" data-anchor-id="不进行整合的情况下的数据处理">
<span class="header-section-number">12.2</span> 不进行整合的情况下的数据处理</h2>
<p>进行标准的数据处理流程：</p>
<div class="cell" data-hash="integration_cache/html/unnamed-chunk-4_217249d68077f9b90661a1346f559aed">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html">NormalizeData</a></span><span class="op">(</span><span class="va">ifnb</span><span class="op">)</span></span>
<span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html">FindVariableFeatures</a></span><span class="op">(</span><span class="va">ifnb</span><span class="op">)</span></span>
<span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html">ScaleData</a></span><span class="op">(</span><span class="va">ifnb</span><span class="op">)</span></span>
<span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html">RunPCA</a></span><span class="op">(</span><span class="va">ifnb</span><span class="op">)</span></span>
<span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html">FindNeighbors</a></span><span class="op">(</span><span class="va">ifnb</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>, reduction <span class="op">=</span> <span class="st">"pca"</span><span class="op">)</span></span>
<span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html">FindClusters</a></span><span class="op">(</span><span class="va">ifnb</span>, </span>
<span>                     resolution <span class="op">=</span> <span class="fl">2</span>, </span>
<span>                     cluster.name <span class="op">=</span> <span class="st">"unintegrated_clusters"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 13999
Number of edges: 555146

Running Louvain algorithm...
Maximum modularity in 10 random starts: 0.8153
Number of communities: 26
Elapsed time: 1 seconds</code></pre>
</div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html">RunUMAP</a></span><span class="op">(</span><span class="va">ifnb</span>, </span>
<span>                dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>, </span>
<span>                reduction <span class="op">=</span> <span class="st">"pca"</span>, </span>
<span>                reduction.name <span class="op">=</span> <span class="st">"umap.unintegrated"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>分别按照<strong>样本分组（“stim”</strong>）<strong>和</strong>细胞聚类情况（“seurat_clusters”<strong>）</strong>着色绘制UMAP图：</p>
<div class="cell" data-hash="integration_cache/html/fig-未整合_01ef1bf20181b2d65d4ac058a26a015b">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">ifnb</span>, </span>
<span>        reduction <span class="op">=</span> <span class="st">"umap.unintegrated"</span>, </span>
<span>        group.by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"stim"</span>, <span class="st">"seurat_clusters"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-未整合" class="quarto-figure quarto-figure-left anchored">
<figure class="figure"><p><img src="integration_files/figure-html/fig-未整合-1.png" class="img-fluid figure-img" width="960"></p>
<figcaption class="figure-caption">Figure&nbsp;12.1: 未整合时的细胞分群情况（左：按照刺激条件着色；右：按照细胞聚类情况着色）</figcaption></figure>
</div>
</div>
</div>
<p>可以发现：The resulting clusters are defined <strong>both by cell type</strong> <strong>and stimulation condition</strong>, which creates challenges for downstream analysis.</p>
</section><section id="进行数据整合" class="level2" data-number="12.3"><h2 data-number="12.3" class="anchored" data-anchor-id="进行数据整合">
<span class="header-section-number">12.3</span> 进行数据整合</h2>
<p>We now aim to integrate data from the two conditions, so that cells from the same cell type/subpopulation will cluster together.</p>
<p>We often refer to this procedure as intergration/alignment. When aligning two genome sequences together, identification of shared/homologous regions can help to interpret differences between the sequences as well. Similarly for scRNA-seq integration, our goal is <strong>not to remove biological differences across conditions</strong>, <strong>but to learn shared cell types/states in an initial step-specifically</strong> because that will enable us to compare control stimulated and control profiles for these individual cell types.</p>
<p>The Seurat v5 integration procedure aims to return a single dimensional reduction that captures the shared sources of variance across multiple layers, so that cells in a similar biological state will cluster. The method returns a dimensional reduction (i.e.&nbsp;<code>integrated.cca</code>) which can be used for visualization and unsupervised clustering analysis. For evaluating performance, we can use cell type labels that are pre-loaded in the&nbsp;<code>seurat_annotations</code>&nbsp;metadata column.</p>
<div class="cell" data-hash="integration_cache/html/unnamed-chunk-6_8034ebad4e50827b35b3521ee2994d38">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 整合，比较耗时间，进度条会一直显示0%直至运算完成</span></span>
<span><span class="va">ifnb_integrated</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/IntegrateLayers.html">IntegrateLayers</a></span><span class="op">(</span><span class="va">ifnb</span>, </span>
<span>                                   method <span class="op">=</span> <span class="va">CCAIntegration</span>, </span>
<span>                                   orig.reduction <span class="op">=</span> <span class="st">"pca"</span>, </span>
<span>                                   new.reduction <span class="op">=</span> <span class="st">"integrated.cca"</span>, <span class="co"># 整合后新的降维数据的名称</span></span>
<span>                                   verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>可以看到经过整合的Seurat对象的降维（“reduction”）中多出了整合后的降维（“integrated.cca”）：</p>
<p><img src="images/截屏2023-11-28 09.57.11.png" class="img-fluid"></p>
<p>目前的RNA assay的counts和data数据仍然按照“CTRL”和“STIM”被分成了4个layer。因此，我们通过JoinLayers进一步合并这些layers：</p>
<div class="cell" data-hash="integration_cache/html/unnamed-chunk-7_2122bb36fe223110327dbd4efb42c8cd">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 目前RNA assay包含的layers</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/Layers.html">Layers</a></span><span class="op">(</span><span class="va">ifnb_integrated</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "counts.CTRL" "counts.STIM" "data.CTRL"   "data.STIM"   "scale.data" </code></pre>
</div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 合并RNA assay的layers</span></span>
<span><span class="va">ifnb_integrated</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/SplitLayers.html">JoinLayers</a></span><span class="op">(</span><span class="va">ifnb_integrated</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co"># 再次检查RNA assay包含的layers</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/Layers.html">Layers</a></span><span class="op">(</span><span class="va">ifnb_integrated</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "data"       "counts"     "scale.data"</code></pre>
</div>
</div>
<p>可以看到，目前的RNA assay只包含三个layers：data”、“counts”、“scale.data”。</p>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Warning
</div>
</div>
<div class="callout-body-container callout-body">
<p>Once integrative analysis is complete, you can rejoin the layers - which collapses the individual datasets together and recreates the original&nbsp;<code>counts</code>&nbsp;and&nbsp;<code>data</code>&nbsp;layers.&nbsp;<strong>You will need to do this before performing any differential expression analysis</strong>. However, you can always resplit the layers in case you would like to reperform integrative analysis.</p>
</div>
</div>
<section id="整合后重新聚类降维" class="level3"><h3 class="anchored" data-anchor-id="整合后重新聚类降维">整合后重新聚类、降维</h3>
<div class="cell" data-hash="integration_cache/html/fig-整合后_c2898705b6f8768d8a83a237e02af729">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 重新聚类</span></span>
<span><span class="va">ifnb_integrated</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html">FindNeighbors</a></span><span class="op">(</span><span class="va">ifnb_integrated</span>, </span>
<span>                                 reduction <span class="op">=</span> <span class="st">"integrated.cca"</span>, <span class="co">#更改降维来源为"integrated.cca"</span></span>
<span>                                 dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span></span>
<span><span class="va">ifnb_integrated</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html">FindClusters</a></span><span class="op">(</span><span class="va">ifnb_integrated</span>, resolution <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 13999
Number of edges: 590406

Running Louvain algorithm...
Maximum modularity in 10 random starts: 0.8448
Number of communities: 18
Elapsed time: 1 seconds</code></pre>
</div>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 重新降维</span></span>
<span><span class="va">ifnb_integrated</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html">RunUMAP</a></span><span class="op">(</span><span class="va">ifnb_integrated</span>, </span>
<span>                           dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>, </span>
<span>                           reduction <span class="op">=</span> <span class="st">"integrated.cca"</span><span class="op">)</span> <span class="co">#更改降维来源为"integrated.cca"</span></span>
<span></span>
<span><span class="co"># Visualization：</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">ifnb_integrated</span>, </span>
<span>        reduction <span class="op">=</span> <span class="st">"umap"</span>, </span>
<span>        group.by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"stim"</span>, <span class="st">"seurat_annotations"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-整合后" class="quarto-figure quarto-figure-left anchored">
<figure class="figure"><p><img src="integration_files/figure-html/fig-整合后-1.png" class="img-fluid figure-img" width="960"></p>
<figcaption class="figure-caption">Figure&nbsp;12.2: 整合后的细胞分群情况（左：按照刺激条件着色；右：按照细胞聚类情况着色）</figcaption></figure>
</div>
</div>
</div>
<p>可以看到和 <a href="#fig-%E6%9C%AA%E6%95%B4%E5%90%88">Figure&nbsp;<span class="quarto-unresolved-ref">fig-未整合</span></a> 相比，在整合后，细胞就只按照细胞类型进行聚类了。</p>
<p>也可以按照刺激条件（“stim”）绘制分面图，分别展示刺激组和对照组的细胞分群情况：</p>
<div class="cell" data-hash="integration_cache/html/unnamed-chunk-9_be0b22749ba8a3a65257e09fcc7d5d41">
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">ifnb_integrated</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, split.by <span class="op">=</span> <span class="st">"stim"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integration_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid" width="960"></p>
</div>
</div>
<p>可以看到，和上面的结论一致，两种条件下的细胞分群基本一致。</p>
</section></section><section id="执行sctransform标准化流程之后的整合" class="level2" data-number="12.4"><h2 data-number="12.4" class="anchored" data-anchor-id="执行sctransform标准化流程之后的整合">
<span class="header-section-number">12.4</span> 执行<code>SCTransform</code>标准化流程之后的整合</h2>
<p>As an alternative to log-normalization, Seurat also includes support for preprocessing of scRNA-seq using the&nbsp;<code><a href="https://satijalab.org/seurat/reference/SCTransform.html">SCTransform()</a></code> workflow（ <a href="#sec-sctransform"><span class="quarto-unresolved-ref">sec-sctransform</span></a> ）. The&nbsp;<code>IntegrateLayers</code>&nbsp;function also supports SCTransform-normalized data, by setting the&nbsp;<code>normalization.method</code> parameter, as shown below.</p>
<section id="不进行整合的情况下的数据分析" class="level3"><h3 class="anchored" data-anchor-id="不进行整合的情况下的数据分析">不进行整合的情况下的数据分析</h3>
<div class="cell" data-hash="integration_cache/html/fig-sctransform_未整合_74f63a8e30dc579164b6ca9aeebc1e97">
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/rm.html">rm</a></span><span class="op">(</span>list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ls.html">ls</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 重新载入原始的Seurat对象ifnb</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span></span>
<span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="st">"data/seurat_official/pbmc_ifnb.rds"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 同样先拆分数据集，然后进行无整合情况下的降维</span></span>
<span><span class="va">ifnb</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html">split</a></span><span class="op">(</span><span class="va">ifnb</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span>, f <span class="op">=</span> <span class="va">ifnb</span><span class="op">$</span><span class="va">stim</span><span class="op">)</span></span>
<span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SCTransform.html">SCTransform</a></span><span class="op">(</span><span class="va">ifnb</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html">RunPCA</a></span><span class="op">(</span><span class="va">ifnb</span><span class="op">)</span></span>
<span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html">RunUMAP</a></span><span class="op">(</span><span class="va">ifnb</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">ifnb</span>, </span>
<span>        reduction <span class="op">=</span> <span class="st">"umap"</span>, </span>
<span>        group.by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"stim"</span>, <span class="st">"seurat_annotations"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-sctransform_未整合" class="quarto-figure quarto-figure-left anchored">
<figure class="figure"><p><img src="integration_files/figure-html/fig-sctransform_未整合-1.png" class="img-fluid figure-img" width="960"></p>
<figcaption class="figure-caption">Figure&nbsp;12.3: 未整合时的细胞分群情况（左：按照刺激条件着色；右：按照细胞聚类情况着色）</figcaption></figure>
</div>
</div>
</div>
<p>可以看到，如果不进行整合，不同样本（STIM vs.&nbsp;STIM）的细胞类型差异很大。</p>
</section><section id="sec-integration_after_sct" class="level3"><h3 class="anchored" data-anchor-id="sec-integration_after_sct">进行整合</h3>
<p>同样通过<code>IntegrateLayers</code>函数进行数据整合，只不过需要将默认的标准化方法由”LogNormalize”指定为”SCT”（<code>normalization.method = "SCT"</code>）：</p>
<div class="cell" data-hash="integration_cache/html/unnamed-chunk-11_a27731d1c41cd8fca82819ab470808d2">
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">ifnb_integrated</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/IntegrateLayers.html">IntegrateLayers</a></span><span class="op">(</span><span class="va">ifnb</span>, </span>
<span>                                   method <span class="op">=</span> <span class="va">CCAIntegration</span>, </span>
<span>                                   normalization.method <span class="op">=</span> <span class="st">"SCT"</span>, </span>
<span>                                   verbose <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span><span class="co"># 整合后重新合并RNA的layers</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/Layers.html">Layers</a></span><span class="op">(</span><span class="va">ifnb_integrated</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "counts.CTRL" "counts.STIM" "data.CTRL"   "data.STIM"  </code></pre>
</div>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">ifnb_integrated</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/SplitLayers.html">JoinLayers</a></span><span class="op">(</span><span class="va">ifnb_integrated</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/Layers.html">Layers</a></span><span class="op">(</span><span class="va">ifnb_integrated</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "data"   "counts"</code></pre>
</div>
</div>
<p><img src="images/截屏2023-11-28 09.16.10.png" class="img-fluid"></p>
<p>可以看到经过整合的Seurat对象的降维（“reduction”）信息中多出了整合后的降维（“integrated.dr”）。</p>
</section><section id="整合后聚类" class="level3"><h3 class="anchored" data-anchor-id="整合后聚类">整合后聚类</h3>
<div class="cell" data-hash="integration_cache/html/unnamed-chunk-12_2f3efe4fc091251dd7dee2a42cebf865">
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">ifnb_integrated</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html">FindNeighbors</a></span><span class="op">(</span><span class="va">ifnb_integrated</span>, </span>
<span>                                 reduction <span class="op">=</span> <span class="st">"integrated.dr"</span>, <span class="co">#更改降维来源为"integrated.dr"</span></span>
<span>                                 dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span></span>
<span><span class="va">ifnb_integrated</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html">FindClusters</a></span><span class="op">(</span><span class="va">ifnb_integrated</span>, </span>
<span>                                resolution <span class="op">=</span> <span class="fl">0.6</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 13999
Number of edges: 527905

Running Louvain algorithm...
Maximum modularity in 10 random starts: 0.9058
Number of communities: 19
Elapsed time: 1 seconds</code></pre>
</div>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">ifnb_integrated</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html">RunUMAP</a></span><span class="op">(</span><span class="va">ifnb_integrated</span>, </span>
<span>                           dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>, </span>
<span>                           reduction <span class="op">=</span> <span class="st">"integrated.dr"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">ifnb_integrated</span>, </span>
<span>        reduction <span class="op">=</span> <span class="st">"umap"</span>, </span>
<span>        group.by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"stim"</span>, <span class="st">"seurat_annotations"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integration_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid" width="960"></p>
</div>
</div>
<p>可以看到和 <a href="#fig-sctransform_%E6%9C%AA%E6%95%B4%E5%90%88">Figure&nbsp;<span class="quarto-unresolved-ref">fig-sctransform_未整合</span></a> 相比，整合后在样本间的细胞类型基本均匀分布。</p>
</section><section id="保存整合后的seurat对象" class="level3"><h3 class="anchored" data-anchor-id="保存整合后的seurat对象">保存整合后的Seurat对象</h3>
<div class="cell" data-hash="integration_cache/html/unnamed-chunk-13_4933610c119e6c9848b45d3e5beb2f07">
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">ifnb_integrated</span>, file <span class="op">=</span> <span class="st">"output/seurat_official/ifnb_integrated.rds"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-5-contents" aria-controls="callout-5" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Session Info
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-5" class="callout-5-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-hash="integration_cache/html/unnamed-chunk-14_9c18d6be3134307a483f97d7659009b4">
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.3.2 (2023-10-31)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Sonoma 14.2.1

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: Asia/Shanghai
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] patchwork_1.1.3    cowplot_1.1.2      ggplot2_3.4.4      Seurat_5.0.1      
[5] SeuratObject_5.0.1 sp_2.1-2          

loaded via a namespace (and not attached):
  [1] RColorBrewer_1.1-3          rstudioapi_0.15.0          
  [3] jsonlite_1.8.8              magrittr_2.0.3             
  [5] spatstat.utils_3.0-4        farver_2.1.1               
  [7] rmarkdown_2.25              zlibbioc_1.48.0            
  [9] vctrs_0.6.5                 ROCR_1.0-11                
 [11] DelayedMatrixStats_1.24.0   spatstat.explore_3.2-5     
 [13] RCurl_1.98-1.13             S4Arrays_1.2.0             
 [15] htmltools_0.5.7             SparseArray_1.2.2          
 [17] sctransform_0.4.1           parallelly_1.36.0          
 [19] KernSmooth_2.23-22          htmlwidgets_1.6.4          
 [21] ica_1.0-3                   plyr_1.8.9                 
 [23] plotly_4.10.3               zoo_1.8-12                 
 [25] igraph_1.6.0                mime_0.12                  
 [27] lifecycle_1.0.4             pkgconfig_2.0.3            
 [29] Matrix_1.6-4                R6_2.5.1                   
 [31] fastmap_1.1.1               GenomeInfoDbData_1.2.11    
 [33] MatrixGenerics_1.14.0       fitdistrplus_1.1-11        
 [35] future_1.33.0               shiny_1.8.0                
 [37] digest_0.6.33               colorspace_2.1-0           
 [39] S4Vectors_0.40.2            tensor_1.5                 
 [41] RSpectra_0.16-1             irlba_2.3.5.1              
 [43] GenomicRanges_1.54.1        labeling_0.4.3             
 [45] progressr_0.14.0            fansi_1.0.6                
 [47] spatstat.sparse_3.0-3       httr_1.4.7                 
 [49] polyclip_1.10-6             abind_1.4-5                
 [51] compiler_4.3.2              withr_2.5.2                
 [53] fastDummies_1.7.3           MASS_7.3-60                
 [55] DelayedArray_0.28.0         tools_4.3.2                
 [57] lmtest_0.9-40               httpuv_1.6.13              
 [59] future.apply_1.11.0         goftest_1.2-3              
 [61] glmGamPoi_1.14.0            glue_1.6.2                 
 [63] nlme_3.1-164                promises_1.2.1             
 [65] grid_4.3.2                  Rtsne_0.17                 
 [67] cluster_2.1.6               reshape2_1.4.4             
 [69] generics_0.1.3              gtable_0.3.4               
 [71] spatstat.data_3.0-3         tidyr_1.3.0                
 [73] data.table_1.14.10          XVector_0.42.0             
 [75] utf8_1.2.4                  BiocGenerics_0.48.1        
 [77] spatstat.geom_3.2-7         RcppAnnoy_0.0.21           
 [79] ggrepel_0.9.4               RANN_2.6.1                 
 [81] pillar_1.9.0                stringr_1.5.1              
 [83] spam_2.10-0                 RcppHNSW_0.5.0             
 [85] later_1.3.2                 splines_4.3.2              
 [87] dplyr_1.1.4                 lattice_0.22-5             
 [89] survival_3.5-7              deldir_2.0-2               
 [91] tidyselect_1.2.0            miniUI_0.1.1.1             
 [93] pbapply_1.7-2               knitr_1.45                 
 [95] gridExtra_2.3               IRanges_2.36.0             
 [97] SummarizedExperiment_1.32.0 scattermore_1.2            
 [99] stats4_4.3.2                xfun_0.41                  
[101] Biobase_2.62.0              matrixStats_1.2.0          
[103] stringi_1.8.3               lazyeval_0.2.2             
[105] yaml_2.3.8                  evaluate_0.23              
[107] codetools_0.2-19            tibble_3.2.1               
[109] cli_3.6.2                   uwot_0.1.16                
[111] xtable_1.8-4                reticulate_1.34.0          
[113] munsell_0.5.0               Rcpp_1.0.11                
[115] GenomeInfoDb_1.38.1         globals_0.16.2             
[117] spatstat.random_3.2-2       png_0.1-8                  
[119] parallel_4.3.2              ellipsis_0.3.2             
[121] dotCall64_1.1-1             sparseMatrixStats_1.14.0   
[123] bitops_1.0-7                listenv_0.9.0              
[125] viridisLite_0.4.2           scales_1.3.0               
[127] ggridges_0.5.5              crayon_1.5.2               
[129] leiden_0.4.3.1              purrr_1.0.2                
[131] rlang_1.1.2                </code></pre>
</div>
</div>
</div>
</div>
</div>


<!-- -->

</section></section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  let localAlternateSentinel = 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var filterRegex = new RegExp("https:\/\/djhcod\.github\.io\/r-notes\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
          // target, if specified
          link.setAttribute("target", "_blank");
      }
    }
});
</script><script src="https://utteranc.es/client.js" repo="https://github.com/djhcod/r-notes" issue-term="pathname" theme="github-light" crossorigin="anonymous" async="">
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="../../single_cell/seurat/cell_cycle_regression.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">消除细胞周期的影响</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../../single_cell/seurat/integrative_analysis_in_seurat_v5.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Seurat v5单细胞数据整合分析</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb34" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu"># 整合（integration） {#sec-Single-cell_data_integration}</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; 原文：</span><span class="co">[</span><span class="ot">*Introduction to scRNA-seq integration*</span><span class="co">](https://satijalab.org/seurat/articles/integration_introduction)</span></span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a><span class="at">&gt;</span></span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; 原文发布日期：2023年10月31日</span></span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a>::: callout-tip</span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a>本篇主要介绍来自不同样本类型的单细胞数据的整合。对于如何整合不同测序技术的单细胞数据集，参考Seurat官方文档：<span class="co">[</span><span class="ot">*Integrative analysis in Seurat v5*</span><span class="co">](https://satijalab.org/seurat/articles/seurat5_integration)</span>。</span>
<span id="cb34-9"><a href="#cb34-9" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb34-10"><a href="#cb34-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-11"><a href="#cb34-11" aria-hidden="true" tabindex="-1"></a>Integration of single-cell sequencing datasets, for example across experimental batches, donors, or conditions, is often an important step in scRNA-seq workflows. Integrative analysis can help to **match shared cell types and states across datasets**, which can boost statistical power, and most importantly, facilitate accurate comparative analysis across datasets. In previous versions of Seurat we introduced methods for integrative analysis, including our 'anchor-based' integration workflow. Many labs have also published powerful and pioneering methods, including&nbsp;<span class="co">[</span><span class="ot">Harmony</span><span class="co">](https://portals.broadinstitute.org/harmony/)</span>&nbsp;and&nbsp;<span class="co">[</span><span class="ot">scVI</span><span class="co">](https://docs.scvi-tools.org/en/stable/index.html)</span>, for integrative analysis.</span>
<span id="cb34-12"><a href="#cb34-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-13"><a href="#cb34-13" aria-hidden="true" tabindex="-1"></a>数据整合的目标：</span>
<span id="cb34-14"><a href="#cb34-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-15"><a href="#cb34-15" aria-hidden="true" tabindex="-1"></a>The following tutorial is designed to give you an overview of the kinds of comparative analyses on complex cell types that are possible using the Seurat integration procedure. Here, we address a few key goals:</span>
<span id="cb34-16"><a href="#cb34-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-17"><a href="#cb34-17" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Identify **cell subpopulations** that are present in both datasets</span>
<span id="cb34-18"><a href="#cb34-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-19"><a href="#cb34-19" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Obtain **cell type markers** that are conserved in both control and stimulated cells</span>
<span id="cb34-20"><a href="#cb34-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-21"><a href="#cb34-21" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>**Compare the datasets** to find cell-type specific responses to stimulation</span>
<span id="cb34-22"><a href="#cb34-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-23"><a href="#cb34-23" aria-hidden="true" tabindex="-1"></a><span class="fu">## 数据读取和分层</span></span>
<span id="cb34-24"><a href="#cb34-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-25"><a href="#cb34-25" aria-hidden="true" tabindex="-1"></a>::: {.callout-note collapse="true" appearance="minimal" icon="false"}</span>
<span id="cb34-26"><a href="#cb34-26" aria-hidden="true" tabindex="-1"></a><span class="fu">### 在线读取（可能需要全局代理）</span></span>
<span id="cb34-27"><a href="#cb34-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-30"><a href="#cb34-30" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb34-31"><a href="#cb34-31" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb34-32"><a href="#cb34-32" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: fenced</span></span>
<span id="cb34-33"><a href="#cb34-33" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb34-34"><a href="#cb34-34" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SeuratData)</span>
<span id="cb34-35"><a href="#cb34-35" aria-hidden="true" tabindex="-1"></a><span class="fu">InstallData</span>(<span class="st">"ifnb"</span>)</span>
<span id="cb34-36"><a href="#cb34-36" aria-hidden="true" tabindex="-1"></a>ifnb <span class="ot">&lt;-</span> <span class="fu">LoadData</span>(<span class="st">"ifnb"</span>)</span>
<span id="cb34-37"><a href="#cb34-37" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb34-38"><a href="#cb34-38" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb34-39"><a href="#cb34-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-40"><a href="#cb34-40" aria-hidden="true" tabindex="-1"></a>从本地下载好的数据读取：</span>
<span id="cb34-41"><a href="#cb34-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-44"><a href="#cb34-44" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb34-45"><a href="#cb34-45" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb34-46"><a href="#cb34-46" aria-hidden="true" tabindex="-1"></a>ifnb <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"data/seurat_official/pbmc_ifnb.rds"</span>)</span>
<span id="cb34-47"><a href="#cb34-47" aria-hidden="true" tabindex="-1"></a>ifnb</span>
<span id="cb34-48"><a href="#cb34-48" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(ifnb<span class="sc">@</span>meta.data, <span class="dv">5</span>)</span>
<span id="cb34-49"><a href="#cb34-49" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(ifnb<span class="sc">$</span>stim)</span>
<span id="cb34-50"><a href="#cb34-50" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb34-51"><a href="#cb34-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-52"><a href="#cb34-52" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/截屏2023-11-27%2012.49.37.png)</span></span>
<span id="cb34-53"><a href="#cb34-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-54"><a href="#cb34-54" aria-hidden="true" tabindex="-1"></a>The object contains data from human PBMC from two conditions, **interferon-stimulated** and **control** cells (stored in the&nbsp;<span class="in">`stim`</span> column in the object metadata). We will aim to integrate the two conditions together, so that we can jointly identify cell subpopulations across datasets, and then explore how each group differs across conditions</span>
<span id="cb34-55"><a href="#cb34-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-56"><a href="#cb34-56" aria-hidden="true" tabindex="-1"></a>In previous versions of Seurat, we would require the data to be represented as two different Seurat objects. **In Seurat v5, we keep all the data in one object, but simply split it into multiple 'layers'**. To learn more about layers, check out our&nbsp;<span class="co">[</span><span class="ot">Seurat object interaction vignette</span><span class="co">](https://satijalab.org/seurat/articles/interaction_vignette)</span>.</span>
<span id="cb34-57"><a href="#cb34-57" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-58"><a href="#cb34-58" aria-hidden="true" tabindex="-1"></a>::: callout-important</span>
<span id="cb34-59"><a href="#cb34-59" aria-hidden="true" tabindex="-1"></a>Seurat v5 assays store data in **layers**. These layers can store:</span>
<span id="cb34-60"><a href="#cb34-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-61"><a href="#cb34-61" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>raw, un-normalized counts (<span class="in">`layer='counts'`</span>)</span>
<span id="cb34-62"><a href="#cb34-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-63"><a href="#cb34-63" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>normalized data (<span class="in">`layer='data'`</span>)</span>
<span id="cb34-64"><a href="#cb34-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-65"><a href="#cb34-65" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>z-scored/variance-stabilized data (<span class="in">`layer='scale.data'`</span>).</span>
<span id="cb34-66"><a href="#cb34-66" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb34-67"><a href="#cb34-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-68"><a href="#cb34-68" aria-hidden="true" tabindex="-1"></a>**split the RNA measurements into two layers one for control cells, one for stimulated cells:**</span>
<span id="cb34-69"><a href="#cb34-69" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-72"><a href="#cb34-72" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb34-73"><a href="#cb34-73" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb34-74"><a href="#cb34-74" aria-hidden="true" tabindex="-1"></a>ifnb[[<span class="st">"RNA"</span>]] <span class="ot">&lt;-</span> <span class="fu">split</span>(ifnb[[<span class="st">"RNA"</span>]], </span>
<span id="cb34-75"><a href="#cb34-75" aria-hidden="true" tabindex="-1"></a>                       <span class="at">f =</span> ifnb<span class="sc">$</span>stim) <span class="co"># 按照meta.data中的“stim”列进行分割</span></span>
<span id="cb34-76"><a href="#cb34-76" aria-hidden="true" tabindex="-1"></a>ifnb</span>
<span id="cb34-77"><a href="#cb34-77" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb34-78"><a href="#cb34-78" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-79"><a href="#cb34-79" aria-hidden="true" tabindex="-1"></a>现在可以发现<span class="in">`ifnb`</span>被分为了4个layer，此前是2个layer（<span class="in">`counts`</span>和<span class="in">`data`</span>）：</span>
<span id="cb34-80"><a href="#cb34-80" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-81"><a href="#cb34-81" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/截屏2023-11-27%2011.55.48.png)</span></span>
<span id="cb34-82"><a href="#cb34-82" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-83"><a href="#cb34-83" aria-hidden="true" tabindex="-1"></a><span class="fu">## 不进行整合的情况下的数据处理</span></span>
<span id="cb34-84"><a href="#cb34-84" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-85"><a href="#cb34-85" aria-hidden="true" tabindex="-1"></a>进行标准的数据处理流程：</span>
<span id="cb34-86"><a href="#cb34-86" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-89"><a href="#cb34-89" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb34-90"><a href="#cb34-90" aria-hidden="true" tabindex="-1"></a>ifnb <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(ifnb)</span>
<span id="cb34-91"><a href="#cb34-91" aria-hidden="true" tabindex="-1"></a>ifnb <span class="ot">&lt;-</span> <span class="fu">FindVariableFeatures</span>(ifnb)</span>
<span id="cb34-92"><a href="#cb34-92" aria-hidden="true" tabindex="-1"></a>ifnb <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(ifnb)</span>
<span id="cb34-93"><a href="#cb34-93" aria-hidden="true" tabindex="-1"></a>ifnb <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(ifnb)</span>
<span id="cb34-94"><a href="#cb34-94" aria-hidden="true" tabindex="-1"></a>ifnb <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(ifnb, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>, <span class="at">reduction =</span> <span class="st">"pca"</span>)</span>
<span id="cb34-95"><a href="#cb34-95" aria-hidden="true" tabindex="-1"></a>ifnb <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(ifnb, </span>
<span id="cb34-96"><a href="#cb34-96" aria-hidden="true" tabindex="-1"></a>                     <span class="at">resolution =</span> <span class="dv">2</span>, </span>
<span id="cb34-97"><a href="#cb34-97" aria-hidden="true" tabindex="-1"></a>                     <span class="at">cluster.name =</span> <span class="st">"unintegrated_clusters"</span>)</span>
<span id="cb34-98"><a href="#cb34-98" aria-hidden="true" tabindex="-1"></a>ifnb <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(ifnb, </span>
<span id="cb34-99"><a href="#cb34-99" aria-hidden="true" tabindex="-1"></a>                <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>, </span>
<span id="cb34-100"><a href="#cb34-100" aria-hidden="true" tabindex="-1"></a>                <span class="at">reduction =</span> <span class="st">"pca"</span>, </span>
<span id="cb34-101"><a href="#cb34-101" aria-hidden="true" tabindex="-1"></a>                <span class="at">reduction.name =</span> <span class="st">"umap.unintegrated"</span>)</span>
<span id="cb34-102"><a href="#cb34-102" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb34-103"><a href="#cb34-103" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-104"><a href="#cb34-104" aria-hidden="true" tabindex="-1"></a>分别按照**样本分组（"stim"**）**和**细胞聚类情况（"seurat_clusters"**）**着色绘制UMAP图：</span>
<span id="cb34-105"><a href="#cb34-105" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-108"><a href="#cb34-108" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb34-109"><a href="#cb34-109" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-未整合</span></span>
<span id="cb34-110"><a href="#cb34-110" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: 未整合时的细胞分群情况（左：按照刺激条件着色；右：按照细胞聚类情况着色）</span></span>
<span id="cb34-111"><a href="#cb34-111" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 10</span></span>
<span id="cb34-112"><a href="#cb34-112" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(ifnb, </span>
<span id="cb34-113"><a href="#cb34-113" aria-hidden="true" tabindex="-1"></a>        <span class="at">reduction =</span> <span class="st">"umap.unintegrated"</span>, </span>
<span id="cb34-114"><a href="#cb34-114" aria-hidden="true" tabindex="-1"></a>        <span class="at">group.by =</span> <span class="fu">c</span>(<span class="st">"stim"</span>, <span class="st">"seurat_clusters"</span>))</span>
<span id="cb34-115"><a href="#cb34-115" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb34-116"><a href="#cb34-116" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-117"><a href="#cb34-117" aria-hidden="true" tabindex="-1"></a>可以发现：The resulting clusters are defined **both by cell type** **and stimulation condition**, which creates challenges for downstream analysis.</span>
<span id="cb34-118"><a href="#cb34-118" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-119"><a href="#cb34-119" aria-hidden="true" tabindex="-1"></a><span class="fu">## 进行数据整合</span></span>
<span id="cb34-120"><a href="#cb34-120" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-121"><a href="#cb34-121" aria-hidden="true" tabindex="-1"></a>We now aim to integrate data from the two conditions, so that cells from the same cell type/subpopulation will cluster together.</span>
<span id="cb34-122"><a href="#cb34-122" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-123"><a href="#cb34-123" aria-hidden="true" tabindex="-1"></a>We often refer to this procedure as intergration/alignment. When aligning two genome sequences together, identification of shared/homologous regions can help to interpret differences between the sequences as well. Similarly for scRNA-seq integration, our goal is **not to remove biological differences across conditions**, **but to learn shared cell types/states in an initial step-specifically** because that will enable us to compare control stimulated and control profiles for these individual cell types.</span>
<span id="cb34-124"><a href="#cb34-124" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-125"><a href="#cb34-125" aria-hidden="true" tabindex="-1"></a>The Seurat v5 integration procedure aims to return a single dimensional reduction that captures the shared sources of variance across multiple layers, so that cells in a similar biological state will cluster. The method returns a dimensional reduction (i.e.&nbsp;<span class="in">`integrated.cca`</span>) which can be used for visualization and unsupervised clustering analysis. For evaluating performance, we can use cell type labels that are pre-loaded in the&nbsp;<span class="in">`seurat_annotations`</span>&nbsp;metadata column.</span>
<span id="cb34-126"><a href="#cb34-126" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-129"><a href="#cb34-129" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb34-130"><a href="#cb34-130" aria-hidden="true" tabindex="-1"></a><span class="co"># 整合，比较耗时间，进度条会一直显示0%直至运算完成</span></span>
<span id="cb34-131"><a href="#cb34-131" aria-hidden="true" tabindex="-1"></a>ifnb_integrated <span class="ot">&lt;-</span> <span class="fu">IntegrateLayers</span>(ifnb, </span>
<span id="cb34-132"><a href="#cb34-132" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">method =</span> CCAIntegration, </span>
<span id="cb34-133"><a href="#cb34-133" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">orig.reduction =</span> <span class="st">"pca"</span>, </span>
<span id="cb34-134"><a href="#cb34-134" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">new.reduction =</span> <span class="st">"integrated.cca"</span>, <span class="co"># 整合后新的降维数据的名称</span></span>
<span id="cb34-135"><a href="#cb34-135" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb34-136"><a href="#cb34-136" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb34-137"><a href="#cb34-137" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-138"><a href="#cb34-138" aria-hidden="true" tabindex="-1"></a>可以看到经过整合的Seurat对象的降维（"reduction"）中多出了整合后的降维（"integrated.cca"）：</span>
<span id="cb34-139"><a href="#cb34-139" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-140"><a href="#cb34-140" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/截屏2023-11-28%2009.57.11.png)</span></span>
<span id="cb34-141"><a href="#cb34-141" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-142"><a href="#cb34-142" aria-hidden="true" tabindex="-1"></a>目前的RNA assay的counts和data数据仍然按照“CTRL”和“STIM”被分成了4个layer。因此，我们通过JoinLayers进一步合并这些layers：</span>
<span id="cb34-143"><a href="#cb34-143" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-146"><a href="#cb34-146" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb34-147"><a href="#cb34-147" aria-hidden="true" tabindex="-1"></a><span class="co"># 目前RNA assay包含的layers</span></span>
<span id="cb34-148"><a href="#cb34-148" aria-hidden="true" tabindex="-1"></a><span class="fu">Layers</span>(ifnb_integrated[[<span class="st">"RNA"</span>]])</span>
<span id="cb34-149"><a href="#cb34-149" aria-hidden="true" tabindex="-1"></a><span class="co"># 合并RNA assay的layers</span></span>
<span id="cb34-150"><a href="#cb34-150" aria-hidden="true" tabindex="-1"></a>ifnb_integrated[[<span class="st">"RNA"</span>]] <span class="ot">&lt;-</span> <span class="fu">JoinLayers</span>(ifnb_integrated[[<span class="st">"RNA"</span>]])</span>
<span id="cb34-151"><a href="#cb34-151" aria-hidden="true" tabindex="-1"></a><span class="co"># 再次检查RNA assay包含的layers</span></span>
<span id="cb34-152"><a href="#cb34-152" aria-hidden="true" tabindex="-1"></a><span class="fu">Layers</span>(ifnb_integrated[[<span class="st">"RNA"</span>]])</span>
<span id="cb34-153"><a href="#cb34-153" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb34-154"><a href="#cb34-154" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-155"><a href="#cb34-155" aria-hidden="true" tabindex="-1"></a>可以看到，目前的RNA assay只包含三个layers：data"、"counts"、"scale.data"。</span>
<span id="cb34-156"><a href="#cb34-156" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-157"><a href="#cb34-157" aria-hidden="true" tabindex="-1"></a>::: callout-warning</span>
<span id="cb34-158"><a href="#cb34-158" aria-hidden="true" tabindex="-1"></a>Once integrative analysis is complete, you can rejoin the layers - which collapses the individual datasets together and recreates the original&nbsp;<span class="in">`counts`</span>&nbsp;and&nbsp;<span class="in">`data`</span>&nbsp;layers.&nbsp;**You will need to do this before performing any differential expression analysis**. However, you can always resplit the layers in case you would like to reperform integrative analysis.</span>
<span id="cb34-159"><a href="#cb34-159" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb34-160"><a href="#cb34-160" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-161"><a href="#cb34-161" aria-hidden="true" tabindex="-1"></a><span class="fu">### 整合后重新聚类、降维</span></span>
<span id="cb34-162"><a href="#cb34-162" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-165"><a href="#cb34-165" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb34-166"><a href="#cb34-166" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-整合后</span></span>
<span id="cb34-167"><a href="#cb34-167" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: 整合后的细胞分群情况（左：按照刺激条件着色；右：按照细胞聚类情况着色）</span></span>
<span id="cb34-168"><a href="#cb34-168" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 10</span></span>
<span id="cb34-169"><a href="#cb34-169" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-170"><a href="#cb34-170" aria-hidden="true" tabindex="-1"></a><span class="co"># 重新聚类</span></span>
<span id="cb34-171"><a href="#cb34-171" aria-hidden="true" tabindex="-1"></a>ifnb_integrated <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(ifnb_integrated, </span>
<span id="cb34-172"><a href="#cb34-172" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">reduction =</span> <span class="st">"integrated.cca"</span>, <span class="co">#更改降维来源为"integrated.cca"</span></span>
<span id="cb34-173"><a href="#cb34-173" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>)</span>
<span id="cb34-174"><a href="#cb34-174" aria-hidden="true" tabindex="-1"></a>ifnb_integrated <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(ifnb_integrated, <span class="at">resolution =</span> <span class="dv">1</span>)</span>
<span id="cb34-175"><a href="#cb34-175" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-176"><a href="#cb34-176" aria-hidden="true" tabindex="-1"></a><span class="co"># 重新降维</span></span>
<span id="cb34-177"><a href="#cb34-177" aria-hidden="true" tabindex="-1"></a>ifnb_integrated <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(ifnb_integrated, </span>
<span id="cb34-178"><a href="#cb34-178" aria-hidden="true" tabindex="-1"></a>                           <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>, </span>
<span id="cb34-179"><a href="#cb34-179" aria-hidden="true" tabindex="-1"></a>                           <span class="at">reduction =</span> <span class="st">"integrated.cca"</span>) <span class="co">#更改降维来源为"integrated.cca"</span></span>
<span id="cb34-180"><a href="#cb34-180" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-181"><a href="#cb34-181" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualization：</span></span>
<span id="cb34-182"><a href="#cb34-182" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(ifnb_integrated, </span>
<span id="cb34-183"><a href="#cb34-183" aria-hidden="true" tabindex="-1"></a>        <span class="at">reduction =</span> <span class="st">"umap"</span>, </span>
<span id="cb34-184"><a href="#cb34-184" aria-hidden="true" tabindex="-1"></a>        <span class="at">group.by =</span> <span class="fu">c</span>(<span class="st">"stim"</span>, <span class="st">"seurat_annotations"</span>))</span>
<span id="cb34-185"><a href="#cb34-185" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb34-186"><a href="#cb34-186" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-187"><a href="#cb34-187" aria-hidden="true" tabindex="-1"></a>可以看到和 @fig-未整合 相比，在整合后，细胞就只按照细胞类型进行聚类了。</span>
<span id="cb34-188"><a href="#cb34-188" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-189"><a href="#cb34-189" aria-hidden="true" tabindex="-1"></a>也可以按照刺激条件（"stim"）绘制分面图，分别展示刺激组和对照组的细胞分群情况：</span>
<span id="cb34-190"><a href="#cb34-190" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-193"><a href="#cb34-193" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb34-194"><a href="#cb34-194" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 10</span></span>
<span id="cb34-195"><a href="#cb34-195" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(ifnb_integrated, <span class="at">reduction =</span> <span class="st">"umap"</span>, <span class="at">split.by =</span> <span class="st">"stim"</span>)</span>
<span id="cb34-196"><a href="#cb34-196" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb34-197"><a href="#cb34-197" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-198"><a href="#cb34-198" aria-hidden="true" tabindex="-1"></a>可以看到，和上面的结论一致，两种条件下的细胞分群基本一致。</span>
<span id="cb34-199"><a href="#cb34-199" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-200"><a href="#cb34-200" aria-hidden="true" tabindex="-1"></a><span class="fu">## 执行`SCTransform`标准化流程之后的整合</span></span>
<span id="cb34-201"><a href="#cb34-201" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-202"><a href="#cb34-202" aria-hidden="true" tabindex="-1"></a>As an alternative to log-normalization, Seurat also includes support for preprocessing of scRNA-seq using the&nbsp;<span class="in">`SCTransform()`</span> workflow（ @sec-sctransform ）. The&nbsp;<span class="in">`IntegrateLayers`</span>&nbsp;function also supports SCTransform-normalized data, by setting the&nbsp;<span class="in">`normalization.method`</span> parameter, as shown below.</span>
<span id="cb34-203"><a href="#cb34-203" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-204"><a href="#cb34-204" aria-hidden="true" tabindex="-1"></a><span class="fu">### 不进行整合的情况下的数据分析</span></span>
<span id="cb34-205"><a href="#cb34-205" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-208"><a href="#cb34-208" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb34-209"><a href="#cb34-209" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-sctransform_未整合</span></span>
<span id="cb34-210"><a href="#cb34-210" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: 未整合时的细胞分群情况（左：按照刺激条件着色；右：按照细胞聚类情况着色）</span></span>
<span id="cb34-211"><a href="#cb34-211" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 10</span></span>
<span id="cb34-212"><a href="#cb34-212" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-213"><a href="#cb34-213" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(<span class="at">list =</span> <span class="fu">ls</span>())</span>
<span id="cb34-214"><a href="#cb34-214" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-215"><a href="#cb34-215" aria-hidden="true" tabindex="-1"></a><span class="co"># 重新载入原始的Seurat对象ifnb</span></span>
<span id="cb34-216"><a href="#cb34-216" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb34-217"><a href="#cb34-217" aria-hidden="true" tabindex="-1"></a>ifnb <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"data/seurat_official/pbmc_ifnb.rds"</span>)</span>
<span id="cb34-218"><a href="#cb34-218" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-219"><a href="#cb34-219" aria-hidden="true" tabindex="-1"></a><span class="co"># 同样先拆分数据集，然后进行无整合情况下的降维</span></span>
<span id="cb34-220"><a href="#cb34-220" aria-hidden="true" tabindex="-1"></a>ifnb[[<span class="st">"RNA"</span>]] <span class="ot">&lt;-</span> <span class="fu">split</span>(ifnb[[<span class="st">"RNA"</span>]], <span class="at">f =</span> ifnb<span class="sc">$</span>stim)</span>
<span id="cb34-221"><a href="#cb34-221" aria-hidden="true" tabindex="-1"></a>ifnb <span class="ot">&lt;-</span> <span class="fu">SCTransform</span>(ifnb, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb34-222"><a href="#cb34-222" aria-hidden="true" tabindex="-1"></a>ifnb <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(ifnb)</span>
<span id="cb34-223"><a href="#cb34-223" aria-hidden="true" tabindex="-1"></a>ifnb <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(ifnb, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>)</span>
<span id="cb34-224"><a href="#cb34-224" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(ifnb, </span>
<span id="cb34-225"><a href="#cb34-225" aria-hidden="true" tabindex="-1"></a>        <span class="at">reduction =</span> <span class="st">"umap"</span>, </span>
<span id="cb34-226"><a href="#cb34-226" aria-hidden="true" tabindex="-1"></a>        <span class="at">group.by =</span> <span class="fu">c</span>(<span class="st">"stim"</span>, <span class="st">"seurat_annotations"</span>))</span>
<span id="cb34-227"><a href="#cb34-227" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb34-228"><a href="#cb34-228" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-229"><a href="#cb34-229" aria-hidden="true" tabindex="-1"></a>可以看到，如果不进行整合，不同样本（STIM vs. STIM）的细胞类型差异很大。</span>
<span id="cb34-230"><a href="#cb34-230" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-231"><a href="#cb34-231" aria-hidden="true" tabindex="-1"></a><span class="fu">### 进行整合 {#sec-integration_after_sct}</span></span>
<span id="cb34-232"><a href="#cb34-232" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-233"><a href="#cb34-233" aria-hidden="true" tabindex="-1"></a>同样通过<span class="in">`IntegrateLayers`</span>函数进行数据整合，只不过需要将默认的标准化方法由"LogNormalize"指定为"SCT"（<span class="in">`normalization.method = "SCT"`</span>）：</span>
<span id="cb34-234"><a href="#cb34-234" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-237"><a href="#cb34-237" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb34-238"><a href="#cb34-238" aria-hidden="true" tabindex="-1"></a>ifnb_integrated <span class="ot">&lt;-</span> <span class="fu">IntegrateLayers</span>(ifnb, </span>
<span id="cb34-239"><a href="#cb34-239" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">method =</span> CCAIntegration, </span>
<span id="cb34-240"><a href="#cb34-240" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">normalization.method =</span> <span class="st">"SCT"</span>, </span>
<span id="cb34-241"><a href="#cb34-241" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">verbose =</span> F)</span>
<span id="cb34-242"><a href="#cb34-242" aria-hidden="true" tabindex="-1"></a><span class="co"># 整合后重新合并RNA的layers</span></span>
<span id="cb34-243"><a href="#cb34-243" aria-hidden="true" tabindex="-1"></a><span class="fu">Layers</span>(ifnb_integrated[[<span class="st">"RNA"</span>]])</span>
<span id="cb34-244"><a href="#cb34-244" aria-hidden="true" tabindex="-1"></a>ifnb_integrated[[<span class="st">"RNA"</span>]] <span class="ot">&lt;-</span> <span class="fu">JoinLayers</span>(ifnb_integrated[[<span class="st">"RNA"</span>]])</span>
<span id="cb34-245"><a href="#cb34-245" aria-hidden="true" tabindex="-1"></a><span class="fu">Layers</span>(ifnb_integrated[[<span class="st">"RNA"</span>]])</span>
<span id="cb34-246"><a href="#cb34-246" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb34-247"><a href="#cb34-247" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-248"><a href="#cb34-248" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/截屏2023-11-28%2009.16.10.png)</span></span>
<span id="cb34-249"><a href="#cb34-249" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-250"><a href="#cb34-250" aria-hidden="true" tabindex="-1"></a>可以看到经过整合的Seurat对象的降维（"reduction"）信息中多出了整合后的降维（"integrated.dr"）。</span>
<span id="cb34-251"><a href="#cb34-251" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-252"><a href="#cb34-252" aria-hidden="true" tabindex="-1"></a><span class="fu">### 整合后聚类</span></span>
<span id="cb34-253"><a href="#cb34-253" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-256"><a href="#cb34-256" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb34-257"><a href="#cb34-257" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 10</span></span>
<span id="cb34-258"><a href="#cb34-258" aria-hidden="true" tabindex="-1"></a>ifnb_integrated <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(ifnb_integrated, </span>
<span id="cb34-259"><a href="#cb34-259" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">reduction =</span> <span class="st">"integrated.dr"</span>, <span class="co">#更改降维来源为"integrated.dr"</span></span>
<span id="cb34-260"><a href="#cb34-260" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>)</span>
<span id="cb34-261"><a href="#cb34-261" aria-hidden="true" tabindex="-1"></a>ifnb_integrated <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(ifnb_integrated, </span>
<span id="cb34-262"><a href="#cb34-262" aria-hidden="true" tabindex="-1"></a>                                <span class="at">resolution =</span> <span class="fl">0.6</span>)</span>
<span id="cb34-263"><a href="#cb34-263" aria-hidden="true" tabindex="-1"></a>ifnb_integrated <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(ifnb_integrated, </span>
<span id="cb34-264"><a href="#cb34-264" aria-hidden="true" tabindex="-1"></a>                           <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>, </span>
<span id="cb34-265"><a href="#cb34-265" aria-hidden="true" tabindex="-1"></a>                           <span class="at">reduction =</span> <span class="st">"integrated.dr"</span>)</span>
<span id="cb34-266"><a href="#cb34-266" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(ifnb_integrated, </span>
<span id="cb34-267"><a href="#cb34-267" aria-hidden="true" tabindex="-1"></a>        <span class="at">reduction =</span> <span class="st">"umap"</span>, </span>
<span id="cb34-268"><a href="#cb34-268" aria-hidden="true" tabindex="-1"></a>        <span class="at">group.by =</span> <span class="fu">c</span>(<span class="st">"stim"</span>, <span class="st">"seurat_annotations"</span>))</span>
<span id="cb34-269"><a href="#cb34-269" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb34-270"><a href="#cb34-270" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-271"><a href="#cb34-271" aria-hidden="true" tabindex="-1"></a>可以看到和 @fig-sctransform_未整合 相比，整合后在样本间的细胞类型基本均匀分布。</span>
<span id="cb34-272"><a href="#cb34-272" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-273"><a href="#cb34-273" aria-hidden="true" tabindex="-1"></a><span class="fu">### 保存整合后的Seurat对象</span></span>
<span id="cb34-274"><a href="#cb34-274" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-277"><a href="#cb34-277" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb34-278"><a href="#cb34-278" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(ifnb_integrated, <span class="at">file =</span> <span class="st">"output/seurat_official/ifnb_integrated.rds"</span>)</span>
<span id="cb34-279"><a href="#cb34-279" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb34-280"><a href="#cb34-280" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-281"><a href="#cb34-281" aria-hidden="true" tabindex="-1"></a>------------------------------------------------------------------------</span>
<span id="cb34-282"><a href="#cb34-282" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-283"><a href="#cb34-283" aria-hidden="true" tabindex="-1"></a>::: {.callout-note collapse="true" icon="false"}</span>
<span id="cb34-284"><a href="#cb34-284" aria-hidden="true" tabindex="-1"></a><span class="fu">## Session Info</span></span>
<span id="cb34-285"><a href="#cb34-285" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-288"><a href="#cb34-288" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb34-289"><a href="#cb34-289" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span>
<span id="cb34-290"><a href="#cb34-290" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb34-291"><a href="#cb34-291" aria-hidden="true" tabindex="-1"></a>:::</span>
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">
      <ul class="footer-items list-unstyled">
<li class="nav-item">
 This book was built with Quarto
  </li>  
</ul>
</div>   
    <div class="nav-footer-center">Copyright 2023, Du Junhong</div>
    <div class="nav-footer-right">
      <ul class="footer-items list-unstyled">
<li class="nav-item compact">
    <a class="nav-link" href="https://github.com/djhcod">
      <i class="bi bi-github" role="img">
</i> 
    </a>
  </li>  
</ul>
</div>
  </div>
</footer>


</body></html>