<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.433">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>R语言学习笔记 - 12&nbsp; 基于参考集的细胞注释</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../single_cell/seurat/cell_cycle_regression.html" rel="next">
<link href="../../single_cell/seurat/integrative_analysis_in_seurat_v5.html" rel="prev">
<link href="../../images/logo.jpg" rel="icon" type="image/jpeg">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark"><script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><script type="application/json" class="js-hypothesis-config">
{
  "theme": "classic",
  "openSidebar": false,
  "showHighlights": "always",
  "enableExperimentalNewNoteButton": true
}
</script><script async="" src="https://hypothes.is/embed.js"></script>
<meta name="twitter:title" content="R语言学习笔记 - 12&nbsp; 基于参考集的细胞注释">
<meta name="twitter:description" content="">
<meta name="twitter:image" content="https://djhcod.github.io/r-notes/single_cell/seurat/images/截屏2023-11-28 11.10.36.png">
<meta name="twitter:image-height" content="426">
<meta name="twitter:image-width" content="1748">
<meta name="twitter:card" content="summary_large_image">
</head>
<body class="nav-sidebar docked nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="navbar navbar-expand-lg navbar-dark "><div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a href="../../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../../images/logo.jpg" alt="" class="navbar-logo"></a>
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">R语言学习笔记</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
<li class="nav-item">
    <a class="nav-link active" href="../../index.html" rel="" target="" aria-current="page">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../intro.html" rel="" target="">
 <span class="menu-text">Introduction</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-parts" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Parts</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-parts">
<li>
    <a class="dropdown-item" href="../../r_basic/r_basics.html" rel="" target="">
 <span class="dropdown-text">R语言基础</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../single_cell/seurat/seurat.html" rel="" target="">
 <span class="dropdown-text">Seurat v5官方文档</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../single_cell/scRNA-seq_online/00_intro.html" rel="" target="">
 <span class="dropdown-text">Introduction</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../quarto_foundation/quarto_foundation.html" rel="" target="">
 <span class="dropdown-text">Quarto基础</span></a>
  </li>  
    </ul>
</li>
  <li class="nav-item">
    <a class="nav-link" href="../../references.html" rel="" target="">
 <span class="menu-text">References</span></a>
  </li>  
</ul>
<ul class="navbar-nav navbar-nav-scroll ms-auto">
<li class="nav-item">
    <a class="nav-link" href="https://bookdown.org" rel="" target=""><i class="bi bi-book-fill" role="img">
</i> 
 <span class="menu-text">Bookdown</span></a>
  </li>  
</ul>
<div class="quarto-navbar-tools tools-wide">
    <a href="https://github.com/djhcod/r-notes" rel="" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
    <div class="dropdown">
      <a href="" title="Share" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Share"><i class="bi bi-share"></i></a>
      <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="quarto-navigation-tool-dropdown-0">
<li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://twitter.com/intent/tweet?url=%7Curl%7C">
              <i class="bi bi-bi-twitter pe-1"></i>
            Twitter
            </a>
          </li>
          <li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://www.facebook.com/sharer/sharer.php?u=%7Curl%7C">
              <i class="bi bi-bi-facebook pe-1"></i>
            Facebook
            </a>
          </li>
      </ul>
</div>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../single_cell/seurat/seurat.html">Seurat v5官方文档</a></li><li class="breadcrumb-item"><a href="../../single_cell/seurat/mapping_and_annotating_query_datasets.html"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">基于参考集的细胞注释</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto"><div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">主页</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">前言</span></a>
  </div>
</li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../r_basic/r_basics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">R语言基础</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/environment_configuration.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Rstudio环境配置</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/data_input_output.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">数据的读取与输出</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/basic_data_function.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">数据处理基本函数</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/character.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">字符的处理</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../single_cell/seurat/seurat.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Seurat v5官方文档</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/seurat_command_list.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Seurat常用函数清单</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/seurat_tutorial.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Seurat细胞分群官方教程</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/data_visualization_methods_in_seurat.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Seurat中的数据可视化方法</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/sctransform.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">基于SCTransform的单细胞数据标准化</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/integration.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">单细胞数据整合（integration）</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/integrative_analysis_in_seurat_v5.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Seurat v5单细胞数据整合分析</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/mapping_and_annotating_query_datasets.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">基于参考集的细胞注释</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/cell_cycle_regression.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">消除细胞周期的影响</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/de_vignette.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">差异表达分析</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../single_cell/scRNA-seq_online/00_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/01_intro_to_scRNA-seq.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">01_intro_to_scRNA-seq.html</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../quarto_foundation/quarto_foundation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Quarto基础</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/yaml_settings.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">YAML设置</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/images_settings.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">图片设置</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/cross_references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">交叉引用</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/insert_other_content.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">插入其他内容</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/quarto_books.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">Quarto Books</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/github_pages.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">21</span>&nbsp; <span class="chapter-title">发布到GitHub Pages</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="false">
 <span class="menu-text">参考文献</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active" data-toc-expanded="1"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li>
<a href="#%E6%9E%84%E5%BB%BA%E5%8F%82%E8%80%83%E6%95%B0%E6%8D%AE%E9%9B%86" id="toc-构建参考数据集" class="nav-link active" data-scroll-target="#%E6%9E%84%E5%BB%BA%E5%8F%82%E8%80%83%E6%95%B0%E6%8D%AE%E9%9B%86"><span class="header-section-number">12.1</span> 构建参考数据集</a>
  <ul class="collapse">
<li><a href="#%E6%95%B0%E6%8D%AE%E8%AF%BB%E5%8F%96" id="toc-数据读取" class="nav-link" data-scroll-target="#%E6%95%B0%E6%8D%AE%E8%AF%BB%E5%8F%96">数据读取</a></li>
  <li><a href="#%E6%95%B0%E6%8D%AE%E9%A2%84%E5%A4%84%E7%90%86" id="toc-数据预处理" class="nav-link" data-scroll-target="#%E6%95%B0%E6%8D%AE%E9%A2%84%E5%A4%84%E7%90%86">数据预处理</a></li>
  </ul>
</li>
  <li>
<a href="#%E5%9F%BA%E4%BA%8E%E5%8F%82%E8%80%83%E9%9B%86%E7%9A%84%E7%BB%86%E8%83%9E%E6%B3%A8%E9%87%8A" id="toc-基于参考集的细胞注释" class="nav-link" data-scroll-target="#%E5%9F%BA%E4%BA%8E%E5%8F%82%E8%80%83%E9%9B%86%E7%9A%84%E7%BB%86%E8%83%9E%E6%B3%A8%E9%87%8A"><span class="header-section-number">12.2</span> 基于参考集的细胞注释</a>
  <ul class="collapse">
<li><a href="#umap%E6%98%A0%E5%B0%84" id="toc-umap映射" class="nav-link" data-scroll-target="#umap%E6%98%A0%E5%B0%84">UMAP映射</a></li>
  </ul>
</li>
  </ul><div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/djhcod/r-notes/blob/main/single_cell/seurat/mapping_and_annotating_query_datasets.qmd" class="toc-action">View source</a></p><p><a href="https://github.com/djhcod/r-notes/issues/new" class="toc-action">Report an issue</a></p><p><a href="https://github.com/djhcod/r-notes/edit/main/single_cell/seurat/mapping_and_annotating_query_datasets.qmd" class="toc-action">Edit this page</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title"><span id="sec-基于参考集的细胞注释" class="quarto-section-identifier"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">基于参考集的细胞注释</span></span></h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><p>原文：<a href="https://satijalab.org/seurat/articles/integration_mapping"><em>Mapping and annotating query datasets</em></a></p>
<p>原文发布日期：2023年10月31日</p>
<p>In this vignette, we first build an <strong>integrated reference</strong> and then demonstrate how to leverage this reference to annotate new <strong>query datasets</strong>. Generating an integrated reference follows the same workflow described in more detail in the integration introduction ( <a href="#sec-%E5%8D%95%E7%BB%86%E8%83%9E%E6%95%B0%E6%8D%AE%E6%95%B4%E5%90%88"><span>Chapter&nbsp;10</span></a> ). Once generated, this reference can be used to analyze additional query datasets through tasks like cell type label transfer and projecting query cells onto reference UMAPs. Notably, this does not require correction of the underlying raw query data and can therefore be an efficient strategy if a high quality reference is available.</p>
<section id="构建参考数据集" class="level2" data-number="12.1"><h2 data-number="12.1" class="anchored" data-anchor-id="构建参考数据集">
<span class="header-section-number">12.1</span> 构建参考数据集</h2>
<p>For the purposes of this example, we’ve chosen human pancreatic islet cell datasets produced across four technologies, <strong>CelSeq (<a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE81076">GSE81076</a>) CelSeq2 (<a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE85241">GSE85241</a>), Fluidigm C1 (<a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE86469">GSE86469</a>), and SMART-Seq2 (E-MTAB-5061)</strong>. For convenience, we distribute this dataset through our&nbsp;<code>SeuratData</code>&nbsp;package. The metadata contains the technology (<code>tech</code>&nbsp;column) and cell type annotations (<code>celltype</code>&nbsp;column) for each cell in the four datasets.</p>
<section id="数据读取" class="level3"><h3 class="anchored" data-anchor-id="数据读取">数据读取</h3>
<div class="callout callout-style-simple callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
在线读取（可能需要全局代理）
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-hash="mapping_and_annotating_query_datasets_cache/html/unnamed-chunk-1_6de20f6101f06fc620404606c1ba0314">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SeuratData)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">InstallData</span>(<span class="st">"panc8"</span>)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>panc8 <span class="ot">&lt;-</span> <span class="fu">LoadData</span>(<span class="st">"panc8"</span>)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<p>从本地下载好的数据读取：</p>
<div class="cell" data-hash="mapping_and_annotating_query_datasets_cache/html/unnamed-chunk-2_c752d7f40dce91c842b60967d4b6a191">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">panc8</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="st">"data/panc8.rds"</span><span class="op">)</span></span>
<span><span class="va">panc8</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
34363 features across 14890 samples within 1 assay 
Active assay: RNA (34363 features, 0 variable features)
 2 layers present: counts, data</code></pre>
</div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">panc8</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "D101_5"  "D101_7"  "D101_10" "D101_13" "D101_14"</code></pre>
</div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">panc8</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "A1BG-AS1" "A1BG"     "A1CF"     "A2M-AS1"  "A2ML1"   </code></pre>
</div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">panc8</span><span class="op">@</span><span class="va">meta.data</span>, <span class="fl">5</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        orig.ident nCount_RNA nFeature_RNA   tech replicate assigned_cluster
D101_5        D101   4615.810         1986 celseq    celseq             &lt;NA&gt;
D101_7        D101  29001.563         4209 celseq    celseq             &lt;NA&gt;
D101_10       D101   6707.857         2408 celseq    celseq             &lt;NA&gt;
D101_13       D101   8797.224         2964 celseq    celseq             &lt;NA&gt;
D101_14       D101   5032.558         2264 celseq    celseq             &lt;NA&gt;
        celltype dataset
D101_5     gamma  celseq
D101_7    acinar  celseq
D101_10    alpha  celseq
D101_13    delta  celseq
D101_14     beta  celseq</code></pre>
</div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">panc8</span><span class="op">$</span><span class="va">tech</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    celseq    celseq2 fluidigmc1     indrop  smartseq2 
      1004       2285        638       8569       2394 </code></pre>
</div>
</div>
<p>可以看到，该数据包含了5种单细胞转录组测序技术获得的单细胞数据。</p>
<p>As a demonstration, we will use a subset of technologies to construct a reference. We will then map the remaining datasets onto this reference. we will use data from <strong>2 technologies</strong> (celseq2和smartseq2) for the reference。</p>
<div class="cell" data-hash="mapping_and_annotating_query_datasets_cache/html/unnamed-chunk-3_5962f7fc928ebb3d42b0f6664b422df2">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span></span>
<span><span class="va">pancreas.ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="va">panc8</span>, <span class="va">tech</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"celseq2"</span>, <span class="st">"smartseq2"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># 按照不同的测序技术将表达矩阵分为不同的layer</span></span>
<span><span class="va">pancreas.ref</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html">split</a></span><span class="op">(</span><span class="va">pancreas.ref</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span>, </span>
<span>                               f <span class="op">=</span> <span class="va">pancreas.ref</span><span class="op">$</span><span class="va">tech</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><img src="images/截屏2023-11-28 11.10.36.png" class="img-fluid"></p>
</section><section id="数据预处理" class="level3"><h3 class="anchored" data-anchor-id="数据预处理">数据预处理</h3>
<p>标准化、找高变基因、归一化、降维、聚类、可视化：</p>
<div class="cell" data-hash="mapping_and_annotating_query_datasets_cache/html/unnamed-chunk-4_a775da7a71812dee8c93a2c375afb818">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pancreas.ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html">NormalizeData</a></span><span class="op">(</span><span class="va">pancreas.ref</span><span class="op">)</span></span>
<span><span class="va">pancreas.ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html">FindVariableFeatures</a></span><span class="op">(</span><span class="va">pancreas.ref</span><span class="op">)</span></span>
<span><span class="va">pancreas.ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html">ScaleData</a></span><span class="op">(</span><span class="va">pancreas.ref</span><span class="op">)</span></span>
<span><span class="va">pancreas.ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html">RunPCA</a></span><span class="op">(</span><span class="va">pancreas.ref</span><span class="op">)</span></span>
<span><span class="va">pancreas.ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html">FindNeighbors</a></span><span class="op">(</span><span class="va">pancreas.ref</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span></span>
<span><span class="va">pancreas.ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html">FindClusters</a></span><span class="op">(</span><span class="va">pancreas.ref</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 4679
Number of edges: 174953

Running Louvain algorithm...
Maximum modularity in 10 random starts: 0.9180
Number of communities: 19
Elapsed time: 0 seconds</code></pre>
</div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pancreas.ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html">RunUMAP</a></span><span class="op">(</span><span class="va">pancreas.ref</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">pancreas.ref</span>, group.by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"celltytpe"</span>, <span class="st">"tech"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="mapping_and_annotating_query_datasets_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell" data-hash="mapping_and_annotating_query_datasets_cache/html/unnamed-chunk-5_35e4d50a2d0b695d4c5f3f2d3a232c6d">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">pancreas.ref</span>, split.by <span class="op">=</span> <span class="st">"tech"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="mapping_and_annotating_query_datasets_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid" width="960"></p>
</div>
</div>
<p>可以看到，不同的测序技术间的细胞类型差异较大。因此需要对数据进行整合，方法同 <a href="#sec-%E5%8D%95%E7%BB%86%E8%83%9E%E6%95%B0%E6%8D%AE%E6%95%B4%E5%90%88"><span>Chapter&nbsp;10</span></a> 一致。</p>
<div class="cell" data-hash="mapping_and_annotating_query_datasets_cache/html/unnamed-chunk-6_e0d4881dd9fbf4245ea54310abd6b465">
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pancreas.ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/IntegrateLayers.html">IntegrateLayers</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas.ref</span>, </span>
<span>                                method <span class="op">=</span> <span class="va">CCAIntegration</span>, </span>
<span>                                orig.reduction <span class="op">=</span> <span class="st">"pca"</span>,</span>
<span>                                new.reduction <span class="op">=</span> <span class="st">"integrated.cca"</span>, </span>
<span>                                verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co"># 重新聚类</span></span>
<span><span class="va">pancreas.ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html">FindNeighbors</a></span><span class="op">(</span><span class="va">pancreas.ref</span>, </span>
<span>                              reduction <span class="op">=</span> <span class="st">"integrated.cca"</span>,<span class="co">#更改降维来源为"integrated.cca"</span></span>
<span>                              dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span></span>
<span><span class="va">pancreas.ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html">FindClusters</a></span><span class="op">(</span><span class="va">pancreas.ref</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 4679
Number of edges: 190152

Running Louvain algorithm...
Maximum modularity in 10 random starts: 0.8680
Number of communities: 15
Elapsed time: 0 seconds</code></pre>
</div>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 重新降维</span></span>
<span><span class="va">pancreas.ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html">RunUMAP</a></span><span class="op">(</span><span class="va">pancreas.ref</span>, </span>
<span>                        reduction <span class="op">=</span> <span class="st">"integrated.cca"</span>, <span class="co">#更改降维来源为"integrated.cca"</span></span>
<span>                        dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">pancreas.ref</span>, group.by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"tech"</span>, <span class="st">"celltype"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="mapping_and_annotating_query_datasets_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid" width="960"></p>
</div>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">pancreas.ref</span>, split.by <span class="op">=</span> <span class="st">"tech"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="mapping_and_annotating_query_datasets_files/figure-html/unnamed-chunk-6-2.png" class="img-fluid" width="960"></p>
</div>
</div>
<p>可以看到，和此前相比，整合后不再有不同测序技术间细胞类型的差异。</p>
</section></section><section id="基于参考集的细胞注释" class="level2" data-number="12.2"><h2 data-number="12.2" class="anchored" data-anchor-id="基于参考集的细胞注释">
<span class="header-section-number">12.2</span> 基于参考集的细胞注释</h2>
<p>Seurat also supports the projection of reference data (or meta data) onto a query object. While many of the methods are conserved (both procedures begin by identifying anchors), there are two important distinctions between data transfer and integration:</p>
<ol type="1">
<li><p>In data transfer, Seurat does not correct or modify the query expression data.</p></li>
<li><p>In data transfer, Seurat has an option (set by default) to project the PCA structure of a reference onto the query, instead of learning a joint structure with CCA. We generally suggest using this option when projecting data between scRNA-seq datasets.</p></li>
</ol>
<blockquote class="blockquote">
<p>Seurat支持将参考数据集的注释信息（meta.data）映射到查询数据集上。基于Seurat的<strong>数据注释映射（data transfer）</strong>和上面的<strong>数据整合（integration）</strong>之间许多步骤都是类似的（例如这两个过程都是从识别锚点开始的），但它们之间有两个重要的区别:</p>
<ol type="1">
<li>
<strong>在data transfer中，Seurat不矫正或修改待查询数据集的表达矩阵</strong>。</li>
<li>在 data transfer中，Seurat有一个选项（默认设置），可以将参考基因集的PCA结构投影到查询对象上，而不是使用CCA学习联合结构。</li>
</ol>
</blockquote>
<p>After finding anchors, we use the&nbsp;<code><a href="https://satijalab.org/seurat/reference/TransferData.html">TransferData()</a></code>&nbsp;function to classify the query cells based on reference data (a vector of reference cell type labels).&nbsp;<code><a href="https://satijalab.org/seurat/reference/TransferData.html">TransferData()</a></code>&nbsp;returns a matrix with <strong>predicted IDs</strong> and <strong>prediction scores</strong>, which we can add to the query metadata.</p>
<p>这里为了演示，还是选取panc8中的两个测序技术（“fluidigmc1”和”celseq”）的数据作为查询数据集。</p>
<div class="cell" data-hash="mapping_and_annotating_query_datasets_cache/html/unnamed-chunk-7_f37063cf4ca68a657a776883b9ed7e06">
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pancreas.query</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="va">panc8</span>, <span class="va">tech</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"fluidigmc1"</span>, <span class="st">"celseq"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># 标准化查询数据集</span></span>
<span><span class="va">pancreas.query</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html">NormalizeData</a></span><span class="op">(</span><span class="va">pancreas.query</span><span class="op">)</span></span>
<span><span class="co"># 寻找transfer锚点</span></span>
<span><span class="va">pancreas.anchors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindTransferAnchors.html">FindTransferAnchors</a></span><span class="op">(</span>reference <span class="op">=</span> <span class="va">pancreas.ref</span>, </span>
<span>                                        query <span class="op">=</span> <span class="va">pancreas.query</span>, </span>
<span>                                        dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>,</span>
<span>                                        reference.reduction <span class="op">=</span> <span class="st">"pca"</span><span class="op">)</span></span>
<span><span class="co"># 映射数据</span></span>
<span><span class="va">predictions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/TransferData.html">TransferData</a></span><span class="op">(</span>anchorset <span class="op">=</span> <span class="va">pancreas.anchors</span>, </span>
<span>                            refdata <span class="op">=</span> <span class="va">pancreas.ref</span><span class="op">$</span><span class="va">celltype</span>, </span>
<span>                            dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">predictions</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1642   15</code></pre>
</div>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">predictions</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        predicted.id prediction.score.alpha prediction.score.endothelial
D101_5         gamma                      0                            0
D101_7        acinar                      0                            0
D101_10        alpha                      1                            0
D101_13        delta                      0                            0
D101_14         beta                      0                            0</code></pre>
</div>
</div>
<blockquote class="blockquote">
<p>可以看到映射之后生成的<code>predictions</code>是一个数据框，将查询数据集中的每一个细胞和预测的细胞类型（predicted.id）一一对应，并给出了这种预测的分数。</p>
</blockquote>
<p>接下来，只需将<code>predictions</code>作为metadata添加到查询数据集：</p>
<div class="cell" data-hash="mapping_and_annotating_query_datasets_cache/html/unnamed-chunk-8_6f56c4ed33546ce3778492ab2dd422e7">
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pancreas.query</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/AddMetaData.html">AddMetaData</a></span><span class="op">(</span><span class="va">pancreas.query</span>, </span>
<span>                              metadata <span class="op">=</span> <span class="va">predictions</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">pancreas.query</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "orig.ident"                          "nCount_RNA"                         
 [3] "nFeature_RNA"                        "tech"                               
 [5] "replicate"                           "assigned_cluster"                   
 [7] "celltype"                            "dataset"                            
 [9] "predicted.id"                        "prediction.score.alpha"             
[11] "prediction.score.endothelial"        "prediction.score.delta"             
[13] "prediction.score.beta"               "prediction.score.ductal"            
[15] "prediction.score.acinar"             "prediction.score.mast"              
[17] "prediction.score.gamma"              "prediction.score.activated_stellate"
[19] "prediction.score.macrophage"         "prediction.score.quiescent_stellate"
[21] "prediction.score.epsilon"            "prediction.score.schwann"           
[23] "prediction.score.max"               </code></pre>
</div>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">pancreas.query</span><span class="op">$</span><span class="va">predicted.id</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
            acinar activated_stellate              alpha               beta 
               262                 39                436                419 
             delta             ductal        endothelial              gamma 
                73                330                 19                 41 
        macrophage               mast            schwann 
                15                  2                  6 </code></pre>
</div>
</div>
<blockquote class="blockquote">
<p>现在的查询数据集中就多出了映射后的细胞注释信息。</p>
</blockquote>
<p>Because we have the original label annotations from our full integrated analysis, we can evaluate how well our predicted cell type annotations match the full reference.</p>
<div class="cell" data-hash="mapping_and_annotating_query_datasets_cache/html/unnamed-chunk-9_59daf3260cd915790cc68d3751cd3649">
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">pancreas.query</span><span class="op">$</span><span class="va">predicted.id</span> <span class="op">==</span> <span class="va">pancreas.query</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
FALSE  TRUE 
   63  1579 </code></pre>
</div>
</div>
<blockquote class="blockquote">
<p>In this example, we find that there is a high agreement in cell type classification, with over 96%（1579/1642） of cells being labeled correctly.</p>
</blockquote>
<p>To verify this further, we can examine some canonical cell type markers for specific pancreatic islet cell populations.</p>
<div class="cell" data-hash="mapping_and_annotating_query_datasets_cache/html/unnamed-chunk-10_41dc34920e0a96b0394a45206f82e5e6">
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html">VlnPlot</a></span><span class="op">(</span><span class="va">pancreas.query</span>, </span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"REG1A"</span>, <span class="st">"PPY"</span>, <span class="st">"SST"</span>, <span class="st">"GHRL"</span>, <span class="st">"VWF"</span>, <span class="st">"SOX10"</span><span class="op">)</span>, </span>
<span>        group.by <span class="op">=</span> <span class="st">"predicted.id"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="mapping_and_annotating_query_datasets_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid" width="960"></p>
</div>
</div>
<blockquote class="blockquote">
<p>Note that even though some of these cell types are only represented by one or two cells (e.g.&nbsp;epsilon cells), we are still able to classify them correctly.</p>
</blockquote>
<section id="umap映射" class="level3"><h3 class="anchored" data-anchor-id="umap映射">UMAP映射</h3>
<p>We also enable projection of a query onto the reference UMAP structure. This can be achieved by computing the reference UMAP model and then calling&nbsp;<code><a href="https://satijalab.org/seurat/reference/MapQuery.html">MapQuery()</a></code>&nbsp;instead of&nbsp;<code><a href="https://satijalab.org/seurat/reference/TransferData.html">TransferData()</a></code>.</p>
<div class="cell" data-hash="mapping_and_annotating_query_datasets_cache/html/unnamed-chunk-11_1a370f75ab6359fb6b57f4f95bb635a9">
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pancreas.ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html">RunUMAP</a></span><span class="op">(</span><span class="va">pancreas.ref</span>, </span>
<span>                        dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>, </span>
<span>                        reduction <span class="op">=</span> <span class="st">"integrated.cca"</span>, </span>
<span>                        return.model <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">pancreas.query</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/MapQuery.html">MapQuery</a></span><span class="op">(</span>anchorset <span class="op">=</span> <span class="va">pancreas.anchors</span>, </span>
<span>                           query <span class="op">=</span> <span class="va">pancreas.query</span>, </span>
<span>                           reference <span class="op">=</span> <span class="va">pancreas.ref</span>, </span>
<span>                           refdata <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>celltype <span class="op">=</span> <span class="st">"celltype"</span><span class="op">)</span>, <span class="co">#需要transfer的参考数据集的列</span></span>
<span>                           reference.reduction <span class="op">=</span> <span class="st">"pca"</span>, </span>
<span>                           reduction.model <span class="op">=</span> <span class="st">"umap"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>可以看到现在的查询数据集pancreas.query中有了降维信息（reduction）。这些信息实际上是映射的参考数据集pancreas.ref的降维信息：</p>
<p><img src="images/截屏2023-11-28 12.25.34.png" class="img-fluid"></p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-2-contents" aria-controls="callout-2" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
What is MapQuery doing?
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-2" class="callout-2-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p><code><a href="https://satijalab.org/seurat/reference/MapQuery.html">MapQuery()</a></code>打包了三个函数的功能:&nbsp;<code><a href="https://satijalab.org/seurat/reference/TransferData.html">TransferData()</a></code>,&nbsp;<code><a href="https://satijalab.org/seurat/reference/IntegrateEmbeddings.html">IntegrateEmbeddings()</a></code>, and&nbsp;<code><a href="https://satijalab.org/seurat/reference/ProjectUMAP.html">ProjectUMAP()</a></code>.&nbsp;<code><a href="https://satijalab.org/seurat/reference/TransferData.html">TransferData()</a></code>&nbsp;is used to transfer cell type labels and impute the ADT values;&nbsp;<code><a href="https://satijalab.org/seurat/reference/IntegrateEmbeddings.html">IntegrateEmbeddings()</a></code>&nbsp;is used to integrate reference with query by correcting the query’s projected low-dimensional embeddings; and finally&nbsp;<code><a href="https://satijalab.org/seurat/reference/ProjectUMAP.html">ProjectUMAP()</a></code>&nbsp;is used to project the query data onto the UMAP structure of the reference. 所以，运行<code><a href="https://satijalab.org/seurat/reference/MapQuery.html">MapQuery()</a></code>的效果和运行下面的脚本一样:</p>
<div class="cell" data-hash="mapping_and_annotating_query_datasets_cache/html/unnamed-chunk-12_2812163ba90a72eacd3c443040b4f9ac">
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pancreas.query</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/TransferData.html">TransferData</a></span><span class="op">(</span>anchorset <span class="op">=</span> <span class="va">pancreas.anchors</span>, </span>
<span>                               reference <span class="op">=</span> <span class="va">pancreas.ref</span>, </span>
<span>                               query <span class="op">=</span> <span class="va">pancreas.query</span>,</span>
<span>                               refdata <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>celltype <span class="op">=</span> <span class="st">"celltype"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">pancreas.query</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/IntegrateEmbeddings.html">IntegrateEmbeddings</a></span><span class="op">(</span>anchorset <span class="op">=</span> <span class="va">pancreas.anchors</span>, </span>
<span>                                      reference <span class="op">=</span> <span class="va">pancreas.ref</span>, </span>
<span>                                      query <span class="op">=</span> <span class="va">pancreas.query</span>,</span>
<span>                                      new.reduction.name <span class="op">=</span> <span class="st">"ref.pca"</span><span class="op">)</span></span>
<span><span class="va">pancreas.query</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ProjectUMAP.html">ProjectUMAP</a></span><span class="op">(</span>query <span class="op">=</span> <span class="va">pancreas.query</span>, </span>
<span>                              query.reduction <span class="op">=</span> <span class="st">"ref.pca"</span>, </span>
<span>                              reference <span class="op">=</span> <span class="va">pancreas.ref</span>,</span>
<span>                              reference.reduction <span class="op">=</span> <span class="st">"pca"</span>, </span>
<span>                              reduction.model <span class="op">=</span> <span class="st">"umap"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<p>We can now visualize the query cells alongside our reference.</p>
<div>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">pancreas.ref</span>, </span>
<span>        reduction <span class="op">=</span> <span class="st">"umap"</span>, </span>
<span>        group.by <span class="op">=</span> <span class="st">"celltype"</span>, </span>
<span>        label <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>        label.size <span class="op">=</span> <span class="fl">6</span>,</span>
<span>        repel <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html">NoLegend</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Reference annotations"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">pancreas.query</span>, </span>
<span>        reduction <span class="op">=</span> <span class="st">"ref.umap"</span>, </span>
<span>        group.by <span class="op">=</span> <span class="st">"predicted.celltype"</span>, </span>
<span>        label <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>        label.size <span class="op">=</span> <span class="fl">6</span>, </span>
<span>        repel <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html">NoLegend</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Query transferred labels"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell quarto-layout-panel" data-hash="mapping_and_annotating_query_datasets_cache/html/unnamed-chunk-13_47bb9ad59c1063fa0f464ea8dcdb6a0f">
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="cell-output-display quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img src="mapping_and_annotating_query_datasets_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid" width="960"></p>
</div>
<div class="cell-output-display quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img src="mapping_and_annotating_query_datasets_files/figure-html/unnamed-chunk-13-2.png" class="img-fluid" width="960"></p>
</div>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>本篇主要介绍了基于单细胞转录组测序（scRNA-seq）数据的参考数据集的制作和映射。Seurat现在还提供了一种’bridge integration’的方法，可以将其他单细胞组学数据（如scATAC-seq、scDNAme、CyTOF）映射到scRNA-seq参考数据集上。详见：<a href="https://satijalab.org/seurat/articles/seurat5_integration_bridge"><em>Dictionary Learning for cross-modality integration</em></a>。</p>
</div>
</div>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-4-contents" aria-controls="callout-4" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Session Info
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-4" class="callout-4-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-hash="mapping_and_annotating_query_datasets_cache/html/unnamed-chunk-14_34ab500b2da253f23d2a455fdfa5ccf3">
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.3.2 (2023-10-31)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Sonoma 14.1.1

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: Asia/Shanghai
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] ggplot2_3.4.4      Seurat_5.0.1       SeuratObject_5.0.1 sp_2.1-1          

loaded via a namespace (and not attached):
  [1] deldir_2.0-2           pbapply_1.7-2          gridExtra_2.3         
  [4] rlang_1.1.2            magrittr_2.0.3         RcppAnnoy_0.0.21      
  [7] spatstat.geom_3.2-7    matrixStats_1.1.0      ggridges_0.5.4        
 [10] compiler_4.3.2         png_0.1-8              vctrs_0.6.4           
 [13] reshape2_1.4.4         stringr_1.5.1          pkgconfig_2.0.3       
 [16] fastmap_1.1.1          ellipsis_0.3.2         labeling_0.4.3        
 [19] utf8_1.2.4             promises_1.2.1         rmarkdown_2.25        
 [22] purrr_1.0.2            xfun_0.41              jsonlite_1.8.7        
 [25] goftest_1.2-3          later_1.3.1            spatstat.utils_3.0-4  
 [28] irlba_2.3.5.1          parallel_4.3.2         cluster_2.1.4         
 [31] R6_2.5.1               ica_1.0-3              stringi_1.8.2         
 [34] RColorBrewer_1.1-3     spatstat.data_3.0-3    reticulate_1.34.0     
 [37] parallelly_1.36.0      lmtest_0.9-40          scattermore_1.2       
 [40] Rcpp_1.0.11            knitr_1.45             tensor_1.5            
 [43] future.apply_1.11.0    zoo_1.8-12             sctransform_0.4.1     
 [46] httpuv_1.6.12          Matrix_1.6-3           splines_4.3.2         
 [49] igraph_1.5.1           tidyselect_1.2.0       abind_1.4-5           
 [52] rstudioapi_0.15.0      yaml_2.3.7             spatstat.random_3.2-1 
 [55] codetools_0.2-19       miniUI_0.1.1.1         spatstat.explore_3.2-5
 [58] listenv_0.9.0          lattice_0.22-5         tibble_3.2.1          
 [61] plyr_1.8.9             withr_2.5.2            shiny_1.8.0           
 [64] ROCR_1.0-11            evaluate_0.23          Rtsne_0.16            
 [67] future_1.33.0          fastDummies_1.7.3      survival_3.5-7        
 [70] polyclip_1.10-6        fitdistrplus_1.1-11    pillar_1.9.0          
 [73] KernSmooth_2.23-22     plotly_4.10.3          generics_0.1.3        
 [76] RcppHNSW_0.5.0         munsell_0.5.0          scales_1.2.1          
 [79] globals_0.16.2         xtable_1.8-4           glue_1.6.2            
 [82] lazyeval_0.2.2         tools_4.3.2            data.table_1.14.8     
 [85] RSpectra_0.16-1        RANN_2.6.1             leiden_0.4.3.1        
 [88] dotCall64_1.1-0        cowplot_1.1.1          grid_4.3.2            
 [91] tidyr_1.3.0            colorspace_2.1-0       nlme_3.1-163          
 [94] patchwork_1.1.3        cli_3.6.1              spatstat.sparse_3.0-3 
 [97] spam_2.10-0            fansi_1.0.5            viridisLite_0.4.2     
[100] dplyr_1.1.4            uwot_0.1.16            gtable_0.3.4          
[103] digest_0.6.33          progressr_0.14.0       ggrepel_0.9.4         
[106] farver_2.1.1           htmlwidgets_1.6.3      htmltools_0.5.7       
[109] lifecycle_1.0.4        httr_1.4.7             mime_0.12             
[112] MASS_7.3-60           </code></pre>
</div>
</div>
</div>
</div>
</div>


<!-- -->

</section></section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  let localAlternateSentinel = 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var filterRegex = new RegExp("https:\/\/djhcod\.github\.io\/r-notes\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
          // target, if specified
          link.setAttribute("target", "_blank");
      }
    }
});
</script><script src="https://utteranc.es/client.js" repo="https://github.com/djhcod/r-notes" issue-term="pathname" theme="github-light" crossorigin="anonymous" async="">
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="../../single_cell/seurat/integrative_analysis_in_seurat_v5.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Seurat v5单细胞数据整合分析</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../../single_cell/seurat/cell_cycle_regression.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">消除细胞周期的影响</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb37" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="fu"># 基于参考集的细胞注释 {#sec-基于参考集的细胞注释}</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>原文：<span class="co">[</span><span class="ot">*Mapping and annotating query datasets*</span><span class="co">](https://satijalab.org/seurat/articles/integration_mapping)</span></span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a>原文发布日期：2023年10月31日</span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a>In this vignette, we first build an **integrated reference** and then demonstrate how to leverage this reference to annotate new **query datasets**. Generating an integrated reference follows the same workflow described in more detail in the integration introduction ( @sec-单细胞数据整合 ). Once generated, this reference can be used to analyze additional query datasets through tasks like cell type label transfer and projecting query cells onto reference UMAPs. Notably, this does not require correction of the underlying raw query data and can therefore be an efficient strategy if a high quality reference is available.</span>
<span id="cb37-8"><a href="#cb37-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-9"><a href="#cb37-9" aria-hidden="true" tabindex="-1"></a><span class="fu">## 构建参考数据集</span></span>
<span id="cb37-10"><a href="#cb37-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-11"><a href="#cb37-11" aria-hidden="true" tabindex="-1"></a>For the purposes of this example, we've chosen human pancreatic islet cell datasets produced across four technologies, **CelSeq ([GSE81076](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE81076)) CelSeq2 ([GSE85241](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE85241)), Fluidigm C1 ([GSE86469](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE86469)), and SMART-Seq2 (E-MTAB-5061)**. For convenience, we distribute this dataset through our&nbsp;<span class="in">`SeuratData`</span>&nbsp;package. The metadata contains the technology (<span class="in">`tech`</span>&nbsp;column) and cell type annotations (<span class="in">`celltype`</span>&nbsp;column) for each cell in the four datasets.</span>
<span id="cb37-12"><a href="#cb37-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-13"><a href="#cb37-13" aria-hidden="true" tabindex="-1"></a><span class="fu">### 数据读取</span></span>
<span id="cb37-14"><a href="#cb37-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-15"><a href="#cb37-15" aria-hidden="true" tabindex="-1"></a>::: {.callout-note collapse="true" appearance="minimal" icon="false"}</span>
<span id="cb37-16"><a href="#cb37-16" aria-hidden="true" tabindex="-1"></a><span class="fu">#### 在线读取（可能需要全局代理）</span></span>
<span id="cb37-17"><a href="#cb37-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-20"><a href="#cb37-20" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb37-21"><a href="#cb37-21" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb37-22"><a href="#cb37-22" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: fenced</span></span>
<span id="cb37-23"><a href="#cb37-23" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SeuratData)</span>
<span id="cb37-24"><a href="#cb37-24" aria-hidden="true" tabindex="-1"></a><span class="fu">InstallData</span>(<span class="st">"panc8"</span>)</span>
<span id="cb37-25"><a href="#cb37-25" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb37-26"><a href="#cb37-26" aria-hidden="true" tabindex="-1"></a>panc8 <span class="ot">&lt;-</span> <span class="fu">LoadData</span>(<span class="st">"panc8"</span>)</span>
<span id="cb37-27"><a href="#cb37-27" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb37-28"><a href="#cb37-28" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb37-29"><a href="#cb37-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-30"><a href="#cb37-30" aria-hidden="true" tabindex="-1"></a>从本地下载好的数据读取：</span>
<span id="cb37-31"><a href="#cb37-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-34"><a href="#cb37-34" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb37-35"><a href="#cb37-35" aria-hidden="true" tabindex="-1"></a>panc8 <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"data/panc8.rds"</span>)</span>
<span id="cb37-36"><a href="#cb37-36" aria-hidden="true" tabindex="-1"></a>panc8</span>
<span id="cb37-37"><a href="#cb37-37" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(panc8)[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span>
<span id="cb37-38"><a href="#cb37-38" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(panc8)[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span>
<span id="cb37-39"><a href="#cb37-39" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(panc8<span class="sc">@</span>meta.data, <span class="dv">5</span>)</span>
<span id="cb37-40"><a href="#cb37-40" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(panc8<span class="sc">$</span>tech)</span>
<span id="cb37-41"><a href="#cb37-41" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb37-42"><a href="#cb37-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-43"><a href="#cb37-43" aria-hidden="true" tabindex="-1"></a>可以看到，该数据包含了5种单细胞转录组测序技术获得的单细胞数据。</span>
<span id="cb37-44"><a href="#cb37-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-45"><a href="#cb37-45" aria-hidden="true" tabindex="-1"></a>As a demonstration, we will use a subset of technologies to construct a reference. We will then map the remaining datasets onto this reference. we will use data from **2 technologies** (celseq2和smartseq2) for the reference。</span>
<span id="cb37-46"><a href="#cb37-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-49"><a href="#cb37-49" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb37-50"><a href="#cb37-50" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb37-51"><a href="#cb37-51" aria-hidden="true" tabindex="-1"></a>pancreas.ref <span class="ot">&lt;-</span> <span class="fu">subset</span>(panc8, tech <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"celseq2"</span>, <span class="st">"smartseq2"</span>))</span>
<span id="cb37-52"><a href="#cb37-52" aria-hidden="true" tabindex="-1"></a><span class="co"># 按照不同的测序技术将表达矩阵分为不同的layer</span></span>
<span id="cb37-53"><a href="#cb37-53" aria-hidden="true" tabindex="-1"></a>pancreas.ref[[<span class="st">"RNA"</span>]] <span class="ot">&lt;-</span> <span class="fu">split</span>(pancreas.ref[[<span class="st">"RNA"</span>]], </span>
<span id="cb37-54"><a href="#cb37-54" aria-hidden="true" tabindex="-1"></a>                               <span class="at">f =</span> pancreas.ref<span class="sc">$</span>tech)</span>
<span id="cb37-55"><a href="#cb37-55" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb37-56"><a href="#cb37-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-57"><a href="#cb37-57" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/截屏2023-11-28%2011.10.36.png)</span></span>
<span id="cb37-58"><a href="#cb37-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-59"><a href="#cb37-59" aria-hidden="true" tabindex="-1"></a><span class="fu">### 数据预处理</span></span>
<span id="cb37-60"><a href="#cb37-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-61"><a href="#cb37-61" aria-hidden="true" tabindex="-1"></a>标准化、找高变基因、归一化、降维、聚类、可视化：</span>
<span id="cb37-62"><a href="#cb37-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-65"><a href="#cb37-65" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb37-66"><a href="#cb37-66" aria-hidden="true" tabindex="-1"></a>pancreas.ref <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(pancreas.ref)</span>
<span id="cb37-67"><a href="#cb37-67" aria-hidden="true" tabindex="-1"></a>pancreas.ref <span class="ot">&lt;-</span> <span class="fu">FindVariableFeatures</span>(pancreas.ref)</span>
<span id="cb37-68"><a href="#cb37-68" aria-hidden="true" tabindex="-1"></a>pancreas.ref <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(pancreas.ref)</span>
<span id="cb37-69"><a href="#cb37-69" aria-hidden="true" tabindex="-1"></a>pancreas.ref <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(pancreas.ref)</span>
<span id="cb37-70"><a href="#cb37-70" aria-hidden="true" tabindex="-1"></a>pancreas.ref <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(pancreas.ref, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>)</span>
<span id="cb37-71"><a href="#cb37-71" aria-hidden="true" tabindex="-1"></a>pancreas.ref <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(pancreas.ref)</span>
<span id="cb37-72"><a href="#cb37-72" aria-hidden="true" tabindex="-1"></a>pancreas.ref <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(pancreas.ref, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>)</span>
<span id="cb37-73"><a href="#cb37-73" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(pancreas.ref, <span class="at">group.by =</span> <span class="fu">c</span>(<span class="st">"celltytpe"</span>, <span class="st">"tech"</span>))</span>
<span id="cb37-74"><a href="#cb37-74" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb37-75"><a href="#cb37-75" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-78"><a href="#cb37-78" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb37-79"><a href="#cb37-79" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 10</span></span>
<span id="cb37-80"><a href="#cb37-80" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(pancreas.ref, <span class="at">split.by =</span> <span class="st">"tech"</span>)</span>
<span id="cb37-81"><a href="#cb37-81" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb37-82"><a href="#cb37-82" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-83"><a href="#cb37-83" aria-hidden="true" tabindex="-1"></a>可以看到，不同的测序技术间的细胞类型差异较大。因此需要对数据进行整合，方法同 @sec-单细胞数据整合 一致。</span>
<span id="cb37-84"><a href="#cb37-84" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-87"><a href="#cb37-87" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb37-88"><a href="#cb37-88" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 10</span></span>
<span id="cb37-89"><a href="#cb37-89" aria-hidden="true" tabindex="-1"></a>pancreas.ref <span class="ot">&lt;-</span> <span class="fu">IntegrateLayers</span>(<span class="at">object =</span> pancreas.ref, </span>
<span id="cb37-90"><a href="#cb37-90" aria-hidden="true" tabindex="-1"></a>                                <span class="at">method =</span> CCAIntegration, </span>
<span id="cb37-91"><a href="#cb37-91" aria-hidden="true" tabindex="-1"></a>                                <span class="at">orig.reduction =</span> <span class="st">"pca"</span>,</span>
<span id="cb37-92"><a href="#cb37-92" aria-hidden="true" tabindex="-1"></a>                                <span class="at">new.reduction =</span> <span class="st">"integrated.cca"</span>, </span>
<span id="cb37-93"><a href="#cb37-93" aria-hidden="true" tabindex="-1"></a>                                <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb37-94"><a href="#cb37-94" aria-hidden="true" tabindex="-1"></a><span class="co"># 重新聚类</span></span>
<span id="cb37-95"><a href="#cb37-95" aria-hidden="true" tabindex="-1"></a>pancreas.ref <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(pancreas.ref, </span>
<span id="cb37-96"><a href="#cb37-96" aria-hidden="true" tabindex="-1"></a>                              <span class="at">reduction =</span> <span class="st">"integrated.cca"</span>,<span class="co">#更改降维来源为"integrated.cca"</span></span>
<span id="cb37-97"><a href="#cb37-97" aria-hidden="true" tabindex="-1"></a>                              <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>)</span>
<span id="cb37-98"><a href="#cb37-98" aria-hidden="true" tabindex="-1"></a>pancreas.ref <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(pancreas.ref)</span>
<span id="cb37-99"><a href="#cb37-99" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-100"><a href="#cb37-100" aria-hidden="true" tabindex="-1"></a><span class="co"># 重新降维</span></span>
<span id="cb37-101"><a href="#cb37-101" aria-hidden="true" tabindex="-1"></a>pancreas.ref <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(pancreas.ref, </span>
<span id="cb37-102"><a href="#cb37-102" aria-hidden="true" tabindex="-1"></a>                        <span class="at">reduction =</span> <span class="st">"integrated.cca"</span>, <span class="co">#更改降维来源为"integrated.cca"</span></span>
<span id="cb37-103"><a href="#cb37-103" aria-hidden="true" tabindex="-1"></a>                        <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>)</span>
<span id="cb37-104"><a href="#cb37-104" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(pancreas.ref, <span class="at">group.by =</span> <span class="fu">c</span>(<span class="st">"tech"</span>, <span class="st">"celltype"</span>))</span>
<span id="cb37-105"><a href="#cb37-105" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(pancreas.ref, <span class="at">split.by =</span> <span class="st">"tech"</span>)</span>
<span id="cb37-106"><a href="#cb37-106" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb37-107"><a href="#cb37-107" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-108"><a href="#cb37-108" aria-hidden="true" tabindex="-1"></a>可以看到，和此前相比，整合后不再有不同测序技术间细胞类型的差异。</span>
<span id="cb37-109"><a href="#cb37-109" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-110"><a href="#cb37-110" aria-hidden="true" tabindex="-1"></a><span class="fu">## 基于参考集的细胞注释</span></span>
<span id="cb37-111"><a href="#cb37-111" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-112"><a href="#cb37-112" aria-hidden="true" tabindex="-1"></a>Seurat also supports the projection of reference data (or meta data) onto a query object. While many of the methods are conserved (both procedures begin by identifying anchors), there are two important distinctions between data transfer and integration:</span>
<span id="cb37-113"><a href="#cb37-113" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-114"><a href="#cb37-114" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>In data transfer, Seurat does not correct or modify the query expression data.</span>
<span id="cb37-115"><a href="#cb37-115" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-116"><a href="#cb37-116" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>In data transfer, Seurat has an option (set by default) to project the PCA structure of a reference onto the query, instead of learning a joint structure with CCA. We generally suggest using this option when projecting data between scRNA-seq datasets.</span>
<span id="cb37-117"><a href="#cb37-117" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-118"><a href="#cb37-118" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; Seurat支持将参考数据集的注释信息（meta.data）映射到查询数据集上。基于Seurat的**数据注释映射（data transfer）**和上面的**数据整合（integration）**之间许多步骤都是类似的（例如这两个过程都是从识别锚点开始的），但它们之间有两个重要的区别:</span></span>
<span id="cb37-119"><a href="#cb37-119" aria-hidden="true" tabindex="-1"></a><span class="at">&gt;</span></span>
<span id="cb37-120"><a href="#cb37-120" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; 1.  **在data transfer中，Seurat不矫正或修改待查询数据集的表达矩阵**。</span></span>
<span id="cb37-121"><a href="#cb37-121" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; 2.  在 data transfer中，Seurat有一个选项（默认设置），可以将参考基因集的PCA结构投影到查询对象上，而不是使用CCA学习联合结构。</span></span>
<span id="cb37-122"><a href="#cb37-122" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-123"><a href="#cb37-123" aria-hidden="true" tabindex="-1"></a>After finding anchors, we use the&nbsp;<span class="in">`TransferData()`</span>&nbsp;function to classify the query cells based on reference data (a vector of reference cell type labels).&nbsp;<span class="in">`TransferData()`</span>&nbsp;returns a matrix with **predicted IDs** and **prediction scores**, which we can add to the query metadata.</span>
<span id="cb37-124"><a href="#cb37-124" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-125"><a href="#cb37-125" aria-hidden="true" tabindex="-1"></a>这里为了演示，还是选取panc8中的两个测序技术（"fluidigmc1"和"celseq"）的数据作为查询数据集。</span>
<span id="cb37-126"><a href="#cb37-126" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-129"><a href="#cb37-129" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb37-130"><a href="#cb37-130" aria-hidden="true" tabindex="-1"></a>pancreas.query <span class="ot">&lt;-</span> <span class="fu">subset</span>(panc8, tech <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"fluidigmc1"</span>, <span class="st">"celseq"</span>))</span>
<span id="cb37-131"><a href="#cb37-131" aria-hidden="true" tabindex="-1"></a><span class="co"># 标准化查询数据集</span></span>
<span id="cb37-132"><a href="#cb37-132" aria-hidden="true" tabindex="-1"></a>pancreas.query <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(pancreas.query)</span>
<span id="cb37-133"><a href="#cb37-133" aria-hidden="true" tabindex="-1"></a><span class="co"># 寻找transfer锚点</span></span>
<span id="cb37-134"><a href="#cb37-134" aria-hidden="true" tabindex="-1"></a>pancreas.anchors <span class="ot">&lt;-</span> <span class="fu">FindTransferAnchors</span>(<span class="at">reference =</span> pancreas.ref, </span>
<span id="cb37-135"><a href="#cb37-135" aria-hidden="true" tabindex="-1"></a>                                        <span class="at">query =</span> pancreas.query, </span>
<span id="cb37-136"><a href="#cb37-136" aria-hidden="true" tabindex="-1"></a>                                        <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>,</span>
<span id="cb37-137"><a href="#cb37-137" aria-hidden="true" tabindex="-1"></a>                                        <span class="at">reference.reduction =</span> <span class="st">"pca"</span>)</span>
<span id="cb37-138"><a href="#cb37-138" aria-hidden="true" tabindex="-1"></a><span class="co"># 映射数据</span></span>
<span id="cb37-139"><a href="#cb37-139" aria-hidden="true" tabindex="-1"></a>predictions <span class="ot">&lt;-</span> <span class="fu">TransferData</span>(<span class="at">anchorset =</span> pancreas.anchors, </span>
<span id="cb37-140"><a href="#cb37-140" aria-hidden="true" tabindex="-1"></a>                            <span class="at">refdata =</span> pancreas.ref<span class="sc">$</span>celltype, </span>
<span id="cb37-141"><a href="#cb37-141" aria-hidden="true" tabindex="-1"></a>                            <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>)</span>
<span id="cb37-142"><a href="#cb37-142" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(predictions)</span>
<span id="cb37-143"><a href="#cb37-143" aria-hidden="true" tabindex="-1"></a>predictions[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]</span>
<span id="cb37-144"><a href="#cb37-144" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb37-145"><a href="#cb37-145" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-146"><a href="#cb37-146" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; 可以看到映射之后生成的</span><span class="in">`predictions`</span><span class="at">是一个数据框，将查询数据集中的每一个细胞和预测的细胞类型（predicted.id）一一对应，并给出了这种预测的分数。</span></span>
<span id="cb37-147"><a href="#cb37-147" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-148"><a href="#cb37-148" aria-hidden="true" tabindex="-1"></a>接下来，只需将<span class="in">`predictions`</span>作为metadata添加到查询数据集：</span>
<span id="cb37-149"><a href="#cb37-149" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-152"><a href="#cb37-152" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb37-153"><a href="#cb37-153" aria-hidden="true" tabindex="-1"></a>pancreas.query <span class="ot">&lt;-</span> <span class="fu">AddMetaData</span>(pancreas.query, </span>
<span id="cb37-154"><a href="#cb37-154" aria-hidden="true" tabindex="-1"></a>                              <span class="at">metadata =</span> predictions)</span>
<span id="cb37-155"><a href="#cb37-155" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(pancreas.query<span class="sc">@</span>meta.data)</span>
<span id="cb37-156"><a href="#cb37-156" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(pancreas.query<span class="sc">$</span>predicted.id)</span>
<span id="cb37-157"><a href="#cb37-157" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb37-158"><a href="#cb37-158" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-159"><a href="#cb37-159" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; 现在的查询数据集中就多出了映射后的细胞注释信息。</span></span>
<span id="cb37-160"><a href="#cb37-160" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-161"><a href="#cb37-161" aria-hidden="true" tabindex="-1"></a>Because we have the original label annotations from our full integrated analysis, we can evaluate how well our predicted cell type annotations match the full reference.</span>
<span id="cb37-162"><a href="#cb37-162" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-165"><a href="#cb37-165" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb37-166"><a href="#cb37-166" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(pancreas.query<span class="sc">$</span>predicted.id <span class="sc">==</span> pancreas.query<span class="sc">$</span>celltype)</span>
<span id="cb37-167"><a href="#cb37-167" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb37-168"><a href="#cb37-168" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-169"><a href="#cb37-169" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; In this example, we find that there is a high agreement in cell type classification, with over 96%（1579/1642） of cells being labeled correctly.</span></span>
<span id="cb37-170"><a href="#cb37-170" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-171"><a href="#cb37-171" aria-hidden="true" tabindex="-1"></a>To verify this further, we can examine some canonical cell type markers for specific pancreatic islet cell populations.</span>
<span id="cb37-172"><a href="#cb37-172" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-175"><a href="#cb37-175" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb37-176"><a href="#cb37-176" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 10</span></span>
<span id="cb37-177"><a href="#cb37-177" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 8</span></span>
<span id="cb37-178"><a href="#cb37-178" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(pancreas.query, </span>
<span id="cb37-179"><a href="#cb37-179" aria-hidden="true" tabindex="-1"></a>        <span class="fu">c</span>(<span class="st">"REG1A"</span>, <span class="st">"PPY"</span>, <span class="st">"SST"</span>, <span class="st">"GHRL"</span>, <span class="st">"VWF"</span>, <span class="st">"SOX10"</span>), </span>
<span id="cb37-180"><a href="#cb37-180" aria-hidden="true" tabindex="-1"></a>        <span class="at">group.by =</span> <span class="st">"predicted.id"</span>)</span>
<span id="cb37-181"><a href="#cb37-181" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb37-182"><a href="#cb37-182" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-183"><a href="#cb37-183" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; Note that even though some of these cell types are only represented by one or two cells (e.g.&nbsp;epsilon cells), we are still able to classify them correctly.</span></span>
<span id="cb37-184"><a href="#cb37-184" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-185"><a href="#cb37-185" aria-hidden="true" tabindex="-1"></a><span class="fu">### UMAP映射</span></span>
<span id="cb37-186"><a href="#cb37-186" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-187"><a href="#cb37-187" aria-hidden="true" tabindex="-1"></a>We also enable projection of a query onto the reference UMAP structure. This can be achieved by computing the reference UMAP model and then calling&nbsp;<span class="in">`MapQuery()`</span>&nbsp;instead of&nbsp;<span class="in">`TransferData()`</span>.</span>
<span id="cb37-188"><a href="#cb37-188" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-191"><a href="#cb37-191" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb37-192"><a href="#cb37-192" aria-hidden="true" tabindex="-1"></a>pancreas.ref <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(pancreas.ref, </span>
<span id="cb37-193"><a href="#cb37-193" aria-hidden="true" tabindex="-1"></a>                        <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>, </span>
<span id="cb37-194"><a href="#cb37-194" aria-hidden="true" tabindex="-1"></a>                        <span class="at">reduction =</span> <span class="st">"integrated.cca"</span>, </span>
<span id="cb37-195"><a href="#cb37-195" aria-hidden="true" tabindex="-1"></a>                        <span class="at">return.model =</span> <span class="cn">TRUE</span>)</span>
<span id="cb37-196"><a href="#cb37-196" aria-hidden="true" tabindex="-1"></a>pancreas.query <span class="ot">&lt;-</span> <span class="fu">MapQuery</span>(<span class="at">anchorset =</span> pancreas.anchors, </span>
<span id="cb37-197"><a href="#cb37-197" aria-hidden="true" tabindex="-1"></a>                           <span class="at">query =</span> pancreas.query, </span>
<span id="cb37-198"><a href="#cb37-198" aria-hidden="true" tabindex="-1"></a>                           <span class="at">reference =</span> pancreas.ref, </span>
<span id="cb37-199"><a href="#cb37-199" aria-hidden="true" tabindex="-1"></a>                           <span class="at">refdata =</span> <span class="fu">list</span>(<span class="at">celltype =</span> <span class="st">"celltype"</span>), <span class="co">#需要transfer的参考数据集的列</span></span>
<span id="cb37-200"><a href="#cb37-200" aria-hidden="true" tabindex="-1"></a>                           <span class="at">reference.reduction =</span> <span class="st">"pca"</span>, </span>
<span id="cb37-201"><a href="#cb37-201" aria-hidden="true" tabindex="-1"></a>                           <span class="at">reduction.model =</span> <span class="st">"umap"</span>)</span>
<span id="cb37-202"><a href="#cb37-202" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb37-203"><a href="#cb37-203" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-204"><a href="#cb37-204" aria-hidden="true" tabindex="-1"></a>可以看到现在的查询数据集pancreas.query中有了降维信息（reduction）。这些信息实际上是映射的参考数据集pancreas.ref的降维信息：</span>
<span id="cb37-205"><a href="#cb37-205" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-206"><a href="#cb37-206" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/截屏2023-11-28%2012.25.34.png)</span></span>
<span id="cb37-207"><a href="#cb37-207" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-208"><a href="#cb37-208" aria-hidden="true" tabindex="-1"></a>::: {.callout-tip collapse="true"}</span>
<span id="cb37-209"><a href="#cb37-209" aria-hidden="true" tabindex="-1"></a><span class="fu">###### What is MapQuery doing?</span></span>
<span id="cb37-210"><a href="#cb37-210" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-211"><a href="#cb37-211" aria-hidden="true" tabindex="-1"></a><span class="in">`MapQuery()`</span>打包了三个函数的功能:&nbsp;<span class="in">`TransferData()`</span>,&nbsp;<span class="in">`IntegrateEmbeddings()`</span>, and&nbsp;<span class="in">`ProjectUMAP()`</span>.&nbsp;<span class="in">`TransferData()`</span>&nbsp;is used to transfer cell type labels and impute the ADT values;&nbsp;<span class="in">`IntegrateEmbeddings()`</span>&nbsp;is used to integrate reference with query by correcting the query's projected low-dimensional embeddings; and finally&nbsp;<span class="in">`ProjectUMAP()`</span>&nbsp;is used to project the query data onto the UMAP structure of the reference. 所以，运行<span class="in">`MapQuery()`</span>的效果和运行下面的脚本一样:</span>
<span id="cb37-212"><a href="#cb37-212" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-215"><a href="#cb37-215" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb37-216"><a href="#cb37-216" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb37-217"><a href="#cb37-217" aria-hidden="true" tabindex="-1"></a>pancreas.query <span class="ot">&lt;-</span> <span class="fu">TransferData</span>(<span class="at">anchorset =</span> pancreas.anchors, </span>
<span id="cb37-218"><a href="#cb37-218" aria-hidden="true" tabindex="-1"></a>                               <span class="at">reference =</span> pancreas.ref, </span>
<span id="cb37-219"><a href="#cb37-219" aria-hidden="true" tabindex="-1"></a>                               <span class="at">query =</span> pancreas.query,</span>
<span id="cb37-220"><a href="#cb37-220" aria-hidden="true" tabindex="-1"></a>                               <span class="at">refdata =</span> <span class="fu">list</span>(<span class="at">celltype =</span> <span class="st">"celltype"</span>))</span>
<span id="cb37-221"><a href="#cb37-221" aria-hidden="true" tabindex="-1"></a>pancreas.query <span class="ot">&lt;-</span> <span class="fu">IntegrateEmbeddings</span>(<span class="at">anchorset =</span> pancreas.anchors, </span>
<span id="cb37-222"><a href="#cb37-222" aria-hidden="true" tabindex="-1"></a>                                      <span class="at">reference =</span> pancreas.ref, </span>
<span id="cb37-223"><a href="#cb37-223" aria-hidden="true" tabindex="-1"></a>                                      <span class="at">query =</span> pancreas.query,</span>
<span id="cb37-224"><a href="#cb37-224" aria-hidden="true" tabindex="-1"></a>                                      <span class="at">new.reduction.name =</span> <span class="st">"ref.pca"</span>)</span>
<span id="cb37-225"><a href="#cb37-225" aria-hidden="true" tabindex="-1"></a>pancreas.query <span class="ot">&lt;-</span> <span class="fu">ProjectUMAP</span>(<span class="at">query =</span> pancreas.query, </span>
<span id="cb37-226"><a href="#cb37-226" aria-hidden="true" tabindex="-1"></a>                              <span class="at">query.reduction =</span> <span class="st">"ref.pca"</span>, </span>
<span id="cb37-227"><a href="#cb37-227" aria-hidden="true" tabindex="-1"></a>                              <span class="at">reference =</span> pancreas.ref,</span>
<span id="cb37-228"><a href="#cb37-228" aria-hidden="true" tabindex="-1"></a>                              <span class="at">reference.reduction =</span> <span class="st">"pca"</span>, </span>
<span id="cb37-229"><a href="#cb37-229" aria-hidden="true" tabindex="-1"></a>                              <span class="at">reduction.model =</span> <span class="st">"umap"</span>)</span>
<span id="cb37-230"><a href="#cb37-230" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb37-231"><a href="#cb37-231" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb37-232"><a href="#cb37-232" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-233"><a href="#cb37-233" aria-hidden="true" tabindex="-1"></a>We can now visualize the query cells alongside our reference.</span>
<span id="cb37-234"><a href="#cb37-234" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-237"><a href="#cb37-237" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb37-238"><a href="#cb37-238" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 10</span></span>
<span id="cb37-239"><a href="#cb37-239" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 8</span></span>
<span id="cb37-240"><a href="#cb37-240" aria-hidden="true" tabindex="-1"></a><span class="co">#| layout-ncol: 2</span></span>
<span id="cb37-241"><a href="#cb37-241" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb37-242"><a href="#cb37-242" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(pancreas.ref, </span>
<span id="cb37-243"><a href="#cb37-243" aria-hidden="true" tabindex="-1"></a>        <span class="at">reduction =</span> <span class="st">"umap"</span>, </span>
<span id="cb37-244"><a href="#cb37-244" aria-hidden="true" tabindex="-1"></a>        <span class="at">group.by =</span> <span class="st">"celltype"</span>, </span>
<span id="cb37-245"><a href="#cb37-245" aria-hidden="true" tabindex="-1"></a>        <span class="at">label =</span> <span class="cn">TRUE</span>, </span>
<span id="cb37-246"><a href="#cb37-246" aria-hidden="true" tabindex="-1"></a>        <span class="at">label.size =</span> <span class="dv">6</span>,</span>
<span id="cb37-247"><a href="#cb37-247" aria-hidden="true" tabindex="-1"></a>        <span class="at">repel =</span> <span class="cn">TRUE</span>) <span class="sc">+</span> </span>
<span id="cb37-248"><a href="#cb37-248" aria-hidden="true" tabindex="-1"></a>  <span class="fu">NoLegend</span>() <span class="sc">+</span> </span>
<span id="cb37-249"><a href="#cb37-249" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Reference annotations"</span>)</span>
<span id="cb37-250"><a href="#cb37-250" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(pancreas.query, </span>
<span id="cb37-251"><a href="#cb37-251" aria-hidden="true" tabindex="-1"></a>        <span class="at">reduction =</span> <span class="st">"ref.umap"</span>, </span>
<span id="cb37-252"><a href="#cb37-252" aria-hidden="true" tabindex="-1"></a>        <span class="at">group.by =</span> <span class="st">"predicted.celltype"</span>, </span>
<span id="cb37-253"><a href="#cb37-253" aria-hidden="true" tabindex="-1"></a>        <span class="at">label =</span> <span class="cn">TRUE</span>,</span>
<span id="cb37-254"><a href="#cb37-254" aria-hidden="true" tabindex="-1"></a>        <span class="at">label.size =</span> <span class="dv">6</span>, </span>
<span id="cb37-255"><a href="#cb37-255" aria-hidden="true" tabindex="-1"></a>        <span class="at">repel =</span> <span class="cn">TRUE</span>) <span class="sc">+</span> </span>
<span id="cb37-256"><a href="#cb37-256" aria-hidden="true" tabindex="-1"></a>  <span class="fu">NoLegend</span>() <span class="sc">+</span> </span>
<span id="cb37-257"><a href="#cb37-257" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Query transferred labels"</span>)</span>
<span id="cb37-258"><a href="#cb37-258" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb37-259"><a href="#cb37-259" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-260"><a href="#cb37-260" aria-hidden="true" tabindex="-1"></a>::: callout-tip</span>
<span id="cb37-261"><a href="#cb37-261" aria-hidden="true" tabindex="-1"></a>本篇主要介绍了基于单细胞转录组测序（scRNA-seq）数据的参考数据集的制作和映射。Seurat现在还提供了一种'bridge integration'的方法，可以将其他单细胞组学数据（如scATAC-seq、scDNAme、CyTOF）映射到scRNA-seq参考数据集上。详见：<span class="co">[</span><span class="ot">*Dictionary Learning for cross-modality integration*</span><span class="co">](https://satijalab.org/seurat/articles/seurat5_integration_bridge)</span>。</span>
<span id="cb37-262"><a href="#cb37-262" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb37-263"><a href="#cb37-263" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-264"><a href="#cb37-264" aria-hidden="true" tabindex="-1"></a>::: {.callout-note collapse="true" icon="false"}</span>
<span id="cb37-265"><a href="#cb37-265" aria-hidden="true" tabindex="-1"></a><span class="fu">## Session Info</span></span>
<span id="cb37-266"><a href="#cb37-266" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-269"><a href="#cb37-269" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb37-270"><a href="#cb37-270" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span>
<span id="cb37-271"><a href="#cb37-271" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb37-272"><a href="#cb37-272" aria-hidden="true" tabindex="-1"></a>:::</span>
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">
      <ul class="footer-items list-unstyled">
<li class="nav-item">
 This book was built with Quarto
  </li>  
</ul>
</div>   
    <div class="nav-footer-center">Copyright 2023, Du Junhong</div>
    <div class="nav-footer-right">
      <ul class="footer-items list-unstyled">
<li class="nav-item compact">
    <a class="nav-link" href="https://github.com/djhcod">
      <i class="bi bi-github" role="img">
</i> 
    </a>
  </li>  
</ul>
</div>
  </div>
</footer>


</body></html>