<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.549">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>R语言学习笔记：从数据清洗到高级统计学和生物信息学 - Seurat细胞分群官方教程</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../single_cell/seurat/data_visualization_methods_in_seurat.html" rel="next">
<link href="../../single_cell/seurat/seurat_command_list.html" rel="prev">
<link href="../../images/logo.jpg" rel="icon" type="image/jpeg">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark"><script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><script async="" src="https://www.googletagmanager.com/gtag/js?id=G-G2HZEXHNCE"></script><script type="text/javascript">

window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', 'G-G2HZEXHNCE', { 'anonymize_ip': true});
</script><script type="application/json" class="js-hypothesis-config">
{
  "theme": "classic",
  "openSidebar": false,
  "showHighlights": "always",
  "enableExperimentalNewNoteButton": true
}
</script><script async="" src="https://hypothes.is/embed.js"></script><script>
  window.document.addEventListener("DOMContentLoaded", function (_event) {
    document.body.classList.add('hypothesis-enabled');
  });
</script>
<meta property="og:title" content="R语言学习笔记：从数据清洗到高级统计学和生物信息学 - Seurat细胞分群官方教程">
<meta property="og:description" content="">
<meta property="og:site_name" content="R语言学习笔记：从数据清洗到高级统计学和生物信息学">
<meta name="twitter:title" content="R语言学习笔记：从数据清洗到高级统计学和生物信息学 - Seurat细胞分群官方教程">
<meta name="twitter:description" content="">
<meta name="twitter:card" content="summary">
</head>
<body class="nav-sidebar docked nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="navbar navbar-expand-lg " data-bs-theme="dark"><div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="../../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../../images/logo.jpg" alt="" class="navbar-logo"></a>
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">R语言学习笔记</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
<li class="nav-item">
    <a class="nav-link active" href="../../index.html" aria-current="page"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../r_basic/r_basics.html"> 
<span class="menu-text">R语言基础</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">单细胞数据分析</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-">
<li>
    <a class="dropdown-item" href="../../single_cell/single_cell_intro.html">
 <span class="dropdown-text">单细胞数据分析介绍</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../single_cell/seurat/seurat_intro.html">
 <span class="dropdown-text">Seurat v5官方文档</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../single_cell/scRNA-seq_online/00_intro.html">
 <span class="dropdown-text">scRNA-seq_online学习材料</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../single_cell/sc_supplementary/sc_supplementary_intro.html">
 <span class="dropdown-text">单细胞补充内容</span></a>
  </li>  
    </ul>
</li>
  <li class="nav-item">
    <a class="nav-link" href="../../quarto_foundation/quarto_foundation.html"> 
<span class="menu-text">Quarto基础</span></a>
  </li>  
</ul>
<ul class="navbar-nav navbar-nav-scroll ms-auto">
<li class="nav-item">
    <a class="nav-link" href="https://bookdown.org"> <i class="bi bi-book-fill" role="img">
</i> 
<span class="menu-text">Bookdown</span></a>
  </li>  
</ul>
</div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools tools-wide">
    <a href="https://github.com/djhcod/r-notes" title="" class="quarto-navigation-tool px-1" aria-label=""><i class="bi bi-github"></i></a>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
  <a href="" class="quarto-reader-toggle quarto-navigation-tool px-1" onclick="window.quartoToggleReader(); return false;" title="Toggle reader mode">
  <div class="quarto-reader-toggle-btn">
  <i class="bi"></i>
  </div>
</a>
</div>
      </div> <!-- /container-fluid -->
    </nav><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../single_cell/single_cell_intro.html">单细胞数据分析</a></li><li class="breadcrumb-item"><a href="../../single_cell/seurat/seurat_intro.html">Seurat v5官方文档</a></li><li class="breadcrumb-item"><a href="../../single_cell/seurat/seurat_tutorial.html">Seurat细胞分群官方教程</a></li></ol></nav>
        <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto"><div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">主页</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../r_basic/r_basics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">R语言基础</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/environment_configuration.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Rstudio环境配置</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/data_input_output.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">数据的读取与输出</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/basic_data_function.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">数据处理基本函数</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/dplyr.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">基于dplyr包的数据处理</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/character.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">字符的处理</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/loop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">循环</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../single_cell/single_cell_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">单细胞数据分析</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">单细胞测序技术介绍</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/generation_of_count_matrix.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">单细胞RNA测序—从raw data到count matrix</span></a>
  </div>
</li>
          <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../single_cell/seurat/seurat_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Seurat v5官方文档</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth2 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/seurat_command_list.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Seurat常用函数清单</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/seurat_tutorial.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Seurat细胞分群官方教程</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/data_visualization_methods_in_seurat.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Seurat中的数据可视化方法</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/sctransform.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">基于SCTransform的单细胞数据标准化</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/cell_cycle_regression.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">消除细胞周期的影响</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/integration.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">整合（integration）</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/integrative_analysis_in_seurat_v5.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Seurat v5单细胞数据整合分析</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/marker_gene_identification.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">寻找marker gene</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/de_vignette.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">寻找不同样本类型间同一细胞类型内的差异基因</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/mapping_and_annotating_query_datasets.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">基于参考集的细胞注释</span></a>
  </div>
</li>
      </ul>
</li>
          <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../single_cell/scRNA-seq_online/00_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">scRNA-seq_online学习材料</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth2 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/03_SC_quality_control-setup.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">数据导入与Seurat对象构建</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/04_SC_quality_control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">质控</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/postQC_workflow.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Single-cell RNA-seq Clustering Workflow</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/05_theory_of_PCA.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">PCA原理</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/06_SC_SCT_normalization.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Normalization and regressing out unwanted variation</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/06_integration.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">单细胞数据整合（Integration）</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/07_SC_clustering_cells_SCT.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">细胞聚类（clustering analysis）</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/08_SC_clustering_quality_control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">细胞分群质量评估</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/09_merged_SC_marker_identification.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">寻找marker基因+细胞注释</span></a>
  </div>
</li>
      </ul>
</li>
          <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../single_cell/sc_supplementary/sc_supplementary_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">单细胞补充内容</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth2 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/sc_supplementary/read_sc_data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">读取非标准格式的单细胞数据</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/sc_supplementary/universal_marker.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">细胞分群通用marker</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/sc_supplementary/integrated_analysis_multiple_single_cell_datasets_1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">多个单细胞数据集整合分析（上）</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/sc_supplementary/integrated_analysis_multiple_single_cell_datasets_2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">多个单细胞数据集整合分析（下）</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/sc_supplementary/DecontX.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">预测和去除单细胞转录组的环境游离RNA污染</span></a>
  </div>
</li>
      </ul>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../quarto_foundation/quarto_foundation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Quarto基础</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/yaml_settings.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">YAML设置</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/images_settings.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">图片设置</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/cross_references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">交叉引用</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/insert_other_content.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">插入其他内容</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/quarto_shiny.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">在Quarto中嵌入Shiny应用程序</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/quarto_books.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Quarto Books</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/github_pages.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">发布到GitHub Pages</span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active" data-toc-expanded="1"><h2 id="toc-title">On this page</h2>
   
  <ul>
<li><a href="#seurat%E5%AF%B9%E8%B1%A1%E6%9E%84%E5%BB%BA" id="toc-seurat对象构建" class="nav-link active" data-scroll-target="#seurat%E5%AF%B9%E8%B1%A1%E6%9E%84%E5%BB%BA"><span class="header-section-number">1</span> Seurat对象构建</a></li>
  <li>
<a href="#%E8%B4%A8%E6%8E%A7" id="toc-质控" class="nav-link" data-scroll-target="#%E8%B4%A8%E6%8E%A7"><span class="header-section-number">2</span> 质控</a>
  <ul class="collapse">
<li><a href="#%E8%AE%A1%E7%AE%97%E7%BA%BF%E7%B2%92%E4%BD%93%E5%9F%BA%E5%9B%A0%E6%AF%94%E4%BE%8B" id="toc-计算线粒体基因比例" class="nav-link" data-scroll-target="#%E8%AE%A1%E7%AE%97%E7%BA%BF%E7%B2%92%E4%BD%93%E5%9F%BA%E5%9B%A0%E6%AF%94%E4%BE%8B"><span class="header-section-number">2.1</span> 计算线粒体基因比例</a></li>
  <li><a href="#%E5%8F%AF%E8%A7%86%E5%8C%96%E8%B4%A8%E6%8E%A7%E6%8C%87%E6%A0%87" id="toc-可视化质控指标" class="nav-link" data-scroll-target="#%E5%8F%AF%E8%A7%86%E5%8C%96%E8%B4%A8%E6%8E%A7%E6%8C%87%E6%A0%87"><span class="header-section-number">2.2</span> 可视化质控指标</a></li>
  <li><a href="#%E8%BF%87%E6%BB%A4%E7%BB%86%E8%83%9E" id="toc-过滤细胞" class="nav-link" data-scroll-target="#%E8%BF%87%E6%BB%A4%E7%BB%86%E8%83%9E"><span class="header-section-number">2.3</span> 过滤细胞</a></li>
  </ul>
</li>
  <li><a href="#%E6%A0%87%E5%87%86%E5%8C%96normalizing" id="toc-标准化normalizing" class="nav-link" data-scroll-target="#%E6%A0%87%E5%87%86%E5%8C%96normalizing"><span class="header-section-number">3</span> 标准化（Normalizing）</a></li>
  <li><a href="#sec-highly_variable_features" id="toc-sec-highly_variable_features" class="nav-link" data-scroll-target="#sec-highly_variable_features"><span class="header-section-number">4</span> 识别高变基因（highly variable features）</a></li>
  <li><a href="#sec-scaledata" id="toc-sec-scaledata" class="nav-link" data-scroll-target="#sec-scaledata"><span class="header-section-number">5</span> 归一化（Scaling）</a></li>
  <li>
<a href="#%E7%BA%BF%E6%80%A7%E9%99%8D%E7%BB%B4" id="toc-线性降维" class="nav-link" data-scroll-target="#%E7%BA%BF%E6%80%A7%E9%99%8D%E7%BB%B4"><span class="header-section-number">6</span> 线性降维</a>
  <ul class="collapse">
<li><a href="#%E9%99%8D%E7%BB%B4%E5%8F%AF%E8%A7%86%E5%8C%96" id="toc-降维可视化" class="nav-link" data-scroll-target="#%E9%99%8D%E7%BB%B4%E5%8F%AF%E8%A7%86%E5%8C%96"><span class="header-section-number">6.1</span> 降维可视化</a></li>
  </ul>
</li>
  <li><a href="#sec-Determine_pcs_for_subsequent_analyses" id="toc-sec-Determine_pcs_for_subsequent_analyses" class="nav-link" data-scroll-target="#sec-Determine_pcs_for_subsequent_analyses"><span class="header-section-number">7</span> 决定后续分析的主成分</a></li>
  <li><a href="#sec-clustering_seurat" id="toc-sec-clustering_seurat" class="nav-link" data-scroll-target="#sec-clustering_seurat"><span class="header-section-number">8</span> 细胞聚类（cluster）</a></li>
  <li>
<a href="#%E9%9D%9E%E7%BA%BF%E6%80%A7%E9%99%8D%E7%BB%B4" id="toc-非线性降维" class="nav-link" data-scroll-target="#%E9%9D%9E%E7%BA%BF%E6%80%A7%E9%99%8D%E7%BB%B4"><span class="header-section-number">9</span> 非线性降维</a>
  <ul class="collapse">
<li><a href="#%E9%9D%9E%E7%BA%BF%E6%80%A7%E9%99%8D%E7%BB%B4%E5%8F%AF%E8%A7%86%E5%8C%96" id="toc-非线性降维可视化" class="nav-link" data-scroll-target="#%E9%9D%9E%E7%BA%BF%E6%80%A7%E9%99%8D%E7%BB%B4%E5%8F%AF%E8%A7%86%E5%8C%96"><span class="header-section-number">9.1</span> 非线性降维可视化</a></li>
  </ul>
</li>
  <li><a href="#sec-seurat_cluster_biomarkers" id="toc-sec-seurat_cluster_biomarkers" class="nav-link" data-scroll-target="#sec-seurat_cluster_biomarkers"><span class="header-section-number">10</span> 寻找细胞群间的差异表达基因（cluster biomarkers）</a></li>
  <li>
<a href="#%E5%8F%AF%E8%A7%86%E5%8C%96marker%E5%9F%BA%E5%9B%A0%E7%9A%84%E8%A1%A8%E8%BE%BE" id="toc-可视化marker基因的表达" class="nav-link" data-scroll-target="#%E5%8F%AF%E8%A7%86%E5%8C%96marker%E5%9F%BA%E5%9B%A0%E7%9A%84%E8%A1%A8%E8%BE%BE"><span class="header-section-number">11</span> 可视化marker基因的表达</a>
  <ul class="collapse">
<li><a href="#%E5%B0%8F%E6%8F%90%E7%90%B4%E5%9B%BE" id="toc-小提琴图" class="nav-link" data-scroll-target="#%E5%B0%8F%E6%8F%90%E7%90%B4%E5%9B%BE"><span class="header-section-number">11.1</span> 小提琴图：</a></li>
  <li><a href="#umap%E5%9B%BE" id="toc-umap图" class="nav-link" data-scroll-target="#umap%E5%9B%BE"><span class="header-section-number">11.2</span> UMAP图：</a></li>
  <li><a href="#%E7%83%AD%E5%9B%BE" id="toc-热图" class="nav-link" data-scroll-target="#%E7%83%AD%E5%9B%BE"><span class="header-section-number">11.3</span> 热图：</a></li>
  </ul>
</li>
  <li><a href="#%E7%BB%86%E8%83%9E%E6%B3%A8%E9%87%8A" id="toc-细胞注释" class="nav-link" data-scroll-target="#%E7%BB%86%E8%83%9E%E6%B3%A8%E9%87%8A"><span class="header-section-number">12</span> 细胞注释</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul><div class="toc-actions"><ul><li><a href="https://github.com/djhcod/r-notes/blob/main/single_cell/seurat/seurat_tutorial.qmd" class="toc-action"><i class="bi bi-github"></i>View source</a></li><li><a href="https://github.com/djhcod/r-notes/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../single_cell/single_cell_intro.html">单细胞数据分析</a></li><li class="breadcrumb-item"><a href="../../single_cell/seurat/seurat_intro.html">Seurat v5官方文档</a></li><li class="breadcrumb-item"><a href="../../single_cell/seurat/seurat_tutorial.html">Seurat细胞分群官方教程</a></li></ol></nav><div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title">Seurat细胞分群官方教程</h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header><blockquote class="blockquote">
<p>原文：<a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial"><em>Seurat - Guided Clustering Tutorial</em></a></p>
<p>原文发布日期：2023年10月31日</p>
</blockquote>
<section id="seurat对象构建" class="level1" data-number="1"><h1 data-number="1">
<span class="header-section-number">1</span> Seurat对象构建</h1>
<p>数据源是来自10X Genomics的<strong>外周血单核细胞（peripheral blood mononuclear cells，PBMC）数据集</strong>。该数据集基于Illumina NextSeq 500平台对2700个单细胞进行了测序。数据可在<a href="https://cf.10xgenomics.com/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz">此链接</a>下载。</p>
<div class="quarto-figure quarto-figure-left">
<figure class="figure"><p><img src="images/2023-11-20 13.40.27.png" class="img-fluid figure-img" width="449"></p>
<figcaption>10X数据的标准文件</figcaption></figure>
</div>
<p>该数据已经通过<a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger">cellranger</a>上游数据处理流程的处理，返回的数据是一个由唯一分子识别（unique molecular identified，UMI）构成的count矩阵。该矩阵中的值表示在每个细胞（列）中检测到的每个特征（即基因；行）的分子数量。数据以10X的标准形式储存，包括：</p>
<ul>
<li>
<p>barcode文件：细胞条码。</p>
<p><img src="images/2023-11-20 14.04.06.png" class="img-fluid" width="161"></p>
</li>
<li>
<p>genes文件：基因名文件。</p>
<p><img src="images/2023-11-20 14.04.15.png" class="img-fluid" width="229"></p>
</li>
<li><p>matrix文件：表达矩阵（稀疏矩阵）。</p></li>
</ul>
<div class="cell" data-hash="seurat_tutorial_cache/html/unnamed-chunk-1_3b646a7e4b1c2ed5ab007863cec18426">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 读取PBMC数据集</span></span>
<span><span class="va">pbmc.data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/Read10X.html">Read10X</a></span><span class="op">(</span>data.dir <span class="op">=</span> <span class="st">"data/seurat_official/filtered_gene_bc_matrices/hg19"</span><span class="op">)</span></span>
<span><span class="co"># Initialize the Seurat object with the raw (non-normalized data).</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/CreateSeuratObject.html">CreateSeuratObject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">pbmc.data</span>, </span>
<span>                           project <span class="op">=</span> <span class="st">"pbmc3k"</span>, </span>
<span>                           min.cells <span class="op">=</span> <span class="fl">3</span>, </span>
<span>                           min.features <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></span>
<span><span class="va">pbmc</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
13714 features across 2700 samples within 1 assay 
Active assay: RNA (13714 features, 0 variable features)
 1 layer present: counts</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-caution callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Caution
</div>
</div>
<div class="callout-body-container callout-body">
<p>有的cellranger处理数据以<a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/advanced/h5_matrices">h5文件格式</a>输出，需要使用seurat中的<a href="https://satijalab.org/seurat/reference/read10x_h5"><code>Read10X_h5()</code></a>函数读取该格式。</p>
</div>
</div>
</section><section id="质控" class="level1" data-number="2"><h1 data-number="2">
<span class="header-section-number">2</span> 质控</h1>
<p>在 <span class="citation" data-cites="ilicic2016">(<a href="#ref-ilicic2016" role="doc-biblioref">Ilicic et al. 2016</a>)</span> 中概括了目前常用的单细胞数据分析中识别低质量细胞的标准，包括：</p>
<ul>
<li>
<p>The <strong>number of unique genes</strong> detected in each cell.</p>
<ul>
<li><p><strong>Low-quality cells</strong> or <strong>empty droplets</strong> will often have very few genes</p></li>
<li><p><strong>Cell doublets</strong> or multiplets may exhibit an aberrantly high gene count</p></li>
</ul>
</li>
<li><p>Similarly, the total number of molecules detected within a cell (correlates strongly with unique genes)</p></li>
<li>
<p>The percentage of reads that map to the mitochondrial genome</p>
<ul>
<li><p>Low-quality / dying cells often exhibit extensive mitochondrial contamination</p></li>
<li><p>We calculate mitochondrial QC metrics with the&nbsp;<code><a href="https://satijalab.org/seurat/reference/PercentageFeatureSet.html">PercentageFeatureSet()</a></code>&nbsp;function, which calculates the percentage of counts originating from a set of features</p></li>
<li><p>We use the set of all genes starting with&nbsp;<code>MT-</code>&nbsp;as a set of mitochondrial genes</p></li>
</ul>
</li>
</ul>
<section id="计算线粒体基因比例" class="level2" data-number="2.1"><h2 data-number="2.1" class="anchored" data-anchor-id="计算线粒体基因比例">
<span class="header-section-number">2.1</span> 计算线粒体基因比例</h2>
<p>通过<code><a href="https://satijalab.org/seurat/reference/PercentageFeatureSet.html">PercentageFeatureSet()</a></code>函数计算每个细胞中线粒体基因的比例，并将其返回到Seurat对象的<code>meta.data</code>中，形成一个新列”percent.mt”。</p>
<div class="cell" data-hash="seurat_tutorial_cache/html/unnamed-chunk-2_07e0e563e9f40722f23de26b873f4a9e">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pbmc</span><span class="op">[[</span><span class="st">"percent.mt"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/PercentageFeatureSet.html">PercentageFeatureSet</a></span><span class="op">(</span><span class="va">pbmc</span>, pattern <span class="op">=</span> <span class="st">"^MT-"</span><span class="op">)</span></span>
<span><span class="co"># Show QC metrics for the first 5 cells</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">@</span><span class="va">meta.data</span>, <span class="fl">5</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                 orig.ident nCount_RNA nFeature_RNA percent.mt
AAACATACAACCAC-1     pbmc3k       2419          779  3.0177759
AAACATTGAGCTAC-1     pbmc3k       4903         1352  3.7935958
AAACATTGATCAGC-1     pbmc3k       3147         1129  0.8897363
AAACCGTGCTTCCG-1     pbmc3k       2639          960  1.7430845
AAACCGTGTATGCG-1     pbmc3k        980          521  1.2244898</code></pre>
</div>
</div>
</section><section id="可视化质控指标" class="level2" data-number="2.2"><h2 data-number="2.2" class="anchored" data-anchor-id="可视化质控指标">
<span class="header-section-number">2.2</span> 可视化质控指标</h2>
<p>通过<code>VlnPlot</code>函数绘制小提琴图展示每个细胞中UMI（nCount_RNA）、基因（percent.mt）和线粒体基因（percent.mt）的数量。</p>
<div class="cell" data-hash="seurat_tutorial_cache/html/unnamed-chunk-3_d03b7ba047e36ea099482954ebca6370">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Visualize QC metrics as a violin plot</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html">VlnPlot</a></span><span class="op">(</span><span class="va">pbmc</span>, </span>
<span>        features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"nFeature_RNA"</span>, <span class="st">"nCount_RNA"</span>, <span class="st">"percent.mt"</span><span class="op">)</span>, </span>
<span>        ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="seurat_tutorial_files/figure-html/unnamed-chunk-3-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>通过<code>FeatureScatter</code>函数展示UMI和线粒体基因数量多关系、UMI和总基因数量多关系。</p>
<div class="cell" data-hash="seurat_tutorial_cache/html/unnamed-chunk-4_1ad2434798ee831bce64ae2574e1aaba">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">plot1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FeatureScatter.html">FeatureScatter</a></span><span class="op">(</span><span class="va">pbmc</span>, </span>
<span>                        feature1 <span class="op">=</span> <span class="st">"nCount_RNA"</span>, </span>
<span>                        feature2 <span class="op">=</span> <span class="st">"percent.mt"</span><span class="op">)</span></span>
<span><span class="va">plot2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FeatureScatter.html">FeatureScatter</a></span><span class="op">(</span><span class="va">pbmc</span>, </span>
<span>                        feature1 <span class="op">=</span> <span class="st">"nCount_RNA"</span>, </span>
<span>                        feature2 <span class="op">=</span> <span class="st">"nFeature_RNA"</span><span class="op">)</span></span>
<span><span class="va">plot1</span> <span class="op">+</span> <span class="va">plot2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="seurat_tutorial_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="过滤细胞" class="level2" data-number="2.3"><h2 data-number="2.3" class="anchored" data-anchor-id="过滤细胞">
<span class="header-section-number">2.3</span> 过滤细胞</h2>
<p>在本案例中使用的质控标准：</p>
<ul>
<li><p>filter cells that have <strong>unique feature counts over 2,500 or less than 200</strong></p></li>
<li><p>filter cells that have <strong>&gt;5% mitochondrial counts</strong></p></li>
</ul>
<div class="cell" data-hash="seurat_tutorial_cache/html/unnamed-chunk-5_dfcbf4d8d9933fdd0cffca2124c419ed">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="va">pbmc</span>, </span>
<span>               subset <span class="op">=</span> <span class="va">nFeature_RNA</span> <span class="op">&gt;</span> <span class="fl">200</span> <span class="op">&amp;</span> <span class="va">nFeature_RNA</span> <span class="op">&lt;</span> <span class="fl">2500</span> <span class="op">&amp;</span> <span class="va">percent.mt</span> <span class="op">&lt;</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">pbmc</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
13714 features across 2638 samples within 1 assay 
Active assay: RNA (13714 features, 0 variable features)
 1 layer present: counts</code></pre>
</div>
</div>
</section></section><section id="标准化normalizing" class="level1" data-number="3"><h1 data-number="3">
<span class="header-section-number">3</span> 标准化（Normalizing）</h1>
<p>After removing unwanted cells from the dataset, the next step is to normalize the data. By default, we employ a <strong>global-scaling normalization method “LogNormalize”</strong> that normalizes the feature expression measurements for each cell by the total expression, multiplies this by a <strong>scale factor</strong> (10,000 by default), and log-transforms the result. 标准化的方法是每个细胞中的基因表达量除以该细胞的基因总表达量，然后乘以比例因子（通常是1000），最后将这个结果取对数。</p>
<div class="cell" data-hash="seurat_tutorial_cache/html/unnamed-chunk-6_7a0340d5c3717543be1a13f1700328ba">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html">NormalizeData</a></span><span class="op">(</span><span class="va">pbmc</span>, </span>
<span>                      normalization.method <span class="op">=</span> <span class="st">"LogNormalize"</span>, </span>
<span>                      scale.factor <span class="op">=</span> <span class="fl">10000</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In Seurat v5, Normalized values are stored in&nbsp;<code>pbmc[["RNA"]]$data</code>：</p>
<p><img src="images/截屏2023-11-21 11.17.58.png" class="img-fluid" width="599"></p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>While this method of normalization is standard and widely used in scRNA-seq analysis, global-scaling relies on an <strong>assumption that each cell originally contains the same number of RNA molecules</strong>. We and others have developed alternative workflows for the single cell preprocessing that do not make these assumptions.</p>
<p>SCTransform() normalization workflow 就是这样的一种工作流. 针对这一方法的原始文献：<span class="citation" data-cites="choudhary2022">(<a href="#ref-choudhary2022" role="doc-biblioref">Choudhary and Satija 2022</a>)</span> 。基于<code>SCTransform</code>的数据标准化流程中，无需使用上面的<code>NormalizeData</code>,&nbsp;和下面的<code>FindVariableFeatures</code>和<code>ScaleData</code>函数。详见<a href="../../single_cell/seurat/sctransform.html">后面的章节</a>。</p>
</div>
</div>
</section><section id="sec-highly_variable_features" class="level1" data-number="4"><h1 data-number="4">
<span class="header-section-number">4</span> 识别高变基因（highly variable features）</h1>
<p>We next calculate a subset of features that exhibit <strong>high cell-to-cell variation</strong> in the dataset (i.e, they are highly expressed in some cells, and lowly expressed in others). We and others <span class="citation" data-cites="brennecke2013">(<a href="#ref-brennecke2013" role="doc-biblioref">Brennecke et al. 2013</a>)</span> have found that focusing on these genes in downstream analysis helps to <strong>highlight biological signal</strong> in single-cell datasets.</p>
<p>Our procedure in Seurat is described in detail here <span class="citation" data-cites="stuart2019">(<a href="#ref-stuart2019" role="doc-biblioref">Stuart et al. 2019</a>)</span> , and improves on previous versions by directly modeling the mean-variance relationship inherent in single-cell data, and is implemented in the&nbsp;<code><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html">FindVariableFeatures()</a></code>&nbsp;function. <strong>By default, we return 2,000 features per dataset</strong>. These will be used in downstream analysis, like PCA.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html">FindVariableFeatures</a></span><span class="op">(</span><span class="va">pbmc</span>, </span>
<span>                             selection.method <span class="op">=</span> <span class="st">"vst"</span>, </span>
<span>                             nfeatures <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Identify the 10 most highly variable genes</span></span>
<span><span class="va">top10</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/VariableFeatures.html">VariableFeatures</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span>, <span class="fl">10</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot variable features with and without labels</span></span>
<span><span class="va">plot1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/VariableFeaturePlot.html">VariableFeaturePlot</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span>
<span><span class="va">plot1</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/LabelPoints.html">LabelPoints</a></span><span class="op">(</span>plot <span class="op">=</span> <span class="va">plot1</span>, points <span class="op">=</span> <span class="va">top10</span>, repel <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="fig-识别高变基因" class="quarto-layout-panel">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-识别高变基因-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="quarto-layout-row">
<div class="cell-output-display quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-识别高变基因" style="flex-basis: 50.0%;justify-content: flex-start;">
<div id="fig-识别高变基因-1" class="quarto-figure quarto-figure-left quarto-float anchored">
<figure class="quarto-float quarto-subfloat-fig figure"><div aria-describedby="fig-识别高变基因-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="seurat_tutorial_files/figure-html/fig-识别高变基因-1.png" class="img-fluid figure-img" data-ref-parent="fig-识别高变基因" width="672">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-识别高变基因-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(A) 前2000个高变基因
</figcaption></figure>
</div>
</div>
<div class="cell-output-display quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-识别高变基因" style="flex-basis: 50.0%;justify-content: flex-start;">
<div id="fig-识别高变基因-2" class="quarto-figure quarto-figure-left quarto-float anchored">
<figure class="quarto-float quarto-subfloat-fig figure"><div aria-describedby="fig-识别高变基因-2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="seurat_tutorial_files/figure-html/fig-识别高变基因-2.png" class="img-fluid figure-img" data-ref-parent="fig-识别高变基因" width="672">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-识别高变基因-2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(B) 前2000个高变基因（标注了前10个高变基因）
</figcaption></figure>
</div>
</div>
</div>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-识别高变基因-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;1: 识别高变基因
</figcaption></figure>
</div>
</section><section id="sec-scaledata" class="level1" data-number="5"><h1 data-number="5">
<span class="header-section-number">5</span> 归一化（Scaling）</h1>
<p>Next, we apply a <strong>linear transformation</strong> (‘scaling’) that is a standard pre-processing step <strong>prior to dimensional reduction</strong> techniques like PCA. The&nbsp;<code><a href="https://satijalab.org/seurat/reference/ScaleData.html">ScaleData()</a></code>&nbsp;function:</p>
<ul>
<li><p>Shifts the expression of each gene, so that <strong>the mean expression across cells is 0（每个基因在所有细胞的平均表达量为0）</strong></p></li>
<li><p>Scales the expression of each gene, so that <strong>the variance across cells is 1.</strong> This step gives equal weight in downstream analyses, so that <strong>highly-expressed genes do not dominate</strong></p></li>
</ul>
<div class="cell" data-hash="seurat_tutorial_cache/html/unnamed-chunk-8_a2d38a6579018310d312e22ba2a0dace">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html">ScaleData</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The results of this are stored in&nbsp;<code>pbmc[["RNA"]]$scale.data</code>:</p>
<p><img src="images/2023-11-21 12.04.27.png" class="img-fluid"></p>
<p><code>ScaleData</code>函数默认情况下使用<code>FindVariableFeatures</code>函数确定的高变基因作为scale依据。可以通过<code>features</code>参数来自定义scale依据，比如这里我们可以让所有基因都参与scale：</p>
<div class="cell" data-hash="seurat_tutorial_cache/html/unnamed-chunk-9_9f45b8f1de0441d6a244fe4c48e23af5">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 不运行</span></span>
<span><span class="va">all.genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html">ScaleData</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="va">all.genes</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
去除单细胞数据中的非期望变异来源
</div>
</div>
<div class="callout-body-container callout-body">
<p>In Seurat, we also use the&nbsp;<code><a href="https://satijalab.org/seurat/reference/ScaleData.html">ScaleData()</a></code>&nbsp;function to remove unwanted sources of variation from a single-cell dataset. For example, we could ‘regress out’ heterogeneity associated with (for example)&nbsp;cell cycle stage (见<a href="../../single_cell/seurat/cell_cycle_regression.html">消除细胞周期的影响</a>), or mitochondrial contamination i.e.:</p>
<div class="cell" data-hash="seurat_tutorial_cache/html/unnamed-chunk-10_2ff592264acb0c2d670aef7190c6f22f">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html">ScaleData</a></span><span class="op">(</span><span class="va">pbmc</span>, vars.to.regress <span class="op">=</span> <span class="st">"percent.mt"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>However, particularly for advanced users who would like to use this functionality, <strong>we strongly recommend the use of our new normalization workflow,&nbsp;<code><a href="https://satijalab.org/seurat/reference/SCTransform.html">SCTransform()</a></code></strong>. The method is described in our&nbsp;paper <span class="citation" data-cites="choudhary2022">(<a href="#ref-choudhary2022" role="doc-biblioref">Choudhary and Satija 2022</a>)</span>. As with&nbsp;<code><a href="https://satijalab.org/seurat/reference/ScaleData.html">ScaleData()</a></code>, the function&nbsp;<code><a href="https://satijalab.org/seurat/reference/SCTransform.html">SCTransform()</a></code>&nbsp;also includes a&nbsp;<code>vars.to.regress</code>&nbsp;parameter (见<a href="../../single_cell/seurat/sctransform.html#sec-perform_sctransform">运行<code>SCTransform</code></a>).&nbsp;</p>
</div>
</div>
</section><section id="线性降维" class="level1" data-number="6"><h1 data-number="6">
<span class="header-section-number">6</span> 线性降维</h1>
<p>Next we perform PCA on the scaled data. By default, <strong>only the previously determined variable features are used as input,</strong> but can be defined using&nbsp;<code>features</code>&nbsp;argument if you wish to choose a different subset (if you do want to use a custom subset of features, make sure you pass these to&nbsp;<code>ScaleData</code>&nbsp;first).</p>
<div class="cell" data-hash="seurat_tutorial_cache/html/unnamed-chunk-11_8ac4bd088186ef8b33b82ba291959b33">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html">RunPCA</a></span><span class="op">(</span><span class="va">pbmc</span>, </span>
<span>               features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/VariableFeatures.html">VariableFeatures</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>For the first principal components, Seurat outputs a list of genes with the most positive and negative loadings, representing modules of genes that exhibit either correlation (or anti-correlation) across single-cells in the dataset.</p>
<div class="cell" data-hash="seurat_tutorial_cache/html/unnamed-chunk-12_43b495d85ecaf24d4d3ed11f0708baf6">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># PC_ 1 </span></span>
<span><span class="co"># Positive:  CST3, TYROBP, LST1, AIF1, FTL, FTH1, LYZ, FCN1, S100A9, TYMP </span></span>
<span><span class="co">#      FCER1G, CFD, LGALS1, LGALS2, SERPINA1, S100A8, CTSS, IFITM3, SPI1, CFP </span></span>
<span><span class="co">#      PSAP, IFI30, COTL1, SAT1, S100A11, NPC2, GRN, LGALS3, GSTP1, PYCARD </span></span>
<span><span class="co"># Negative:  MALAT1, LTB, IL32, IL7R, CD2, B2M, ACAP1, CTSW, STK17A, CD27 </span></span>
<span><span class="co">#      CD247, CCL5, GIMAP5, GZMA, AQP3, CST7, TRAF3IP3, SELL, GZMK, HOPX </span></span>
<span><span class="co">#      MAL, MYC, ITM2A, ETS1, LYAR, GIMAP7, KLRG1, NKG7, ZAP70, BEX2 </span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>完成PCA分析的Seurat对象：</p>
<p><img src="images/截屏2023-11-21 15.04.58.png" class="img-fluid"></p>
<section id="降维可视化" class="level2" data-number="6.1"><h2 data-number="6.1" class="anchored" data-anchor-id="降维可视化">
<span class="header-section-number">6.1</span> 降维可视化</h2>
<p>Seurat provides several useful ways of visualizing both cells and features that define the PCA, including&nbsp;<code>VizDimReduction()</code>,&nbsp;<code><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot()</a></code>, and&nbsp;<code><a href="https://satijalab.org/seurat/reference/DimHeatmap.html">DimHeatmap()</a></code> .</p>
<div class="cell" data-hash="seurat_tutorial_cache/html/unnamed-chunk-13_2616cc484150c482ad5e4a2895baaf39">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/VizDimLoadings.html">VizDimLoadings</a></span><span class="op">(</span><span class="va">pbmc</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">2</span>, reduction <span class="op">=</span> <span class="st">"pca"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="seurat_tutorial_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell" data-hash="seurat_tutorial_cache/html/unnamed-chunk-14_81db6ff90e35f80e6120575d331b06ba">
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">pbmc</span>, reduction <span class="op">=</span> <span class="st">"pca"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html">NoLegend</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="seurat_tutorial_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>In particular <code><a href="https://satijalab.org/seurat/reference/DimHeatmap.html">DimHeatmap()</a></code> allows for easy exploration of the primary sources of heterogeneity in a dataset, and can be useful when trying to <strong>decide which PCs to include for further downstream analyses</strong>. <strong>Both cells and features are ordered according to their PCA scores</strong>.</p>
<p>Setting&nbsp;<code>cells</code>&nbsp;to a number, will plot the ‘extreme’ cells on both ends of the spectrum, which dramatically speeds plotting for large datasets. Though clearly a supervised analysis, we find this to be a valuable tool for exploring correlated feature sets.</p>
<div class="cell" data-hash="seurat_tutorial_cache/html/unnamed-chunk-15_0ca22b5fce3afe2ef9f60c94ae681540">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimHeatmap.html">DimHeatmap</a></span><span class="op">(</span><span class="va">pbmc</span>, dims <span class="op">=</span> <span class="fl">1</span>, cells <span class="op">=</span> <span class="fl">1000</span>, balanced <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="seurat_tutorial_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>通过<code>dims</code>参数指定一个范围内的主成分，可以用来决定在后续的分析中应该包括哪些主成分。</p>
<div class="cell" data-hash="seurat_tutorial_cache/html/unnamed-chunk-16_528759ff6a19d1021dfea40c4476e76b">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimHeatmap.html">DimHeatmap</a></span><span class="op">(</span><span class="va">pbmc</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">15</span>, cells <span class="op">=</span> <span class="fl">1000</span>, balanced <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="seurat_tutorial_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section></section><section id="sec-Determine_pcs_for_subsequent_analyses" class="level1" data-number="7"><h1 data-number="7">
<span class="header-section-number">7</span> 决定后续分析的主成分</h1>
<p>To overcome the extensive technical noise in any single feature for scRNA-seq data, <strong>Seurat clusters cells based on their PCA scores</strong>, with each PC essentially representing a ‘metafeature’ that combines information across a correlated feature set. The top principal components therefore represent a robust compression of the dataset. However, how many components should we choose to include? 10? 20? 100?</p>
<p>In Macosko&nbsp;<em>et al</em> <span class="citation" data-cites="macosko2015">(<a href="#ref-macosko2015" role="doc-biblioref">Macosko et al. 2015</a>)</span>, we implemented a resampling test inspired by the JackStraw procedure. While still available in Seurat, this is a slow and computationally expensive procedure, and we is no longer routinely used in single cell analysis.</p>
<p>An alternative heuristic method generates an ‘<strong>Elbow plot</strong>’: a ranking of principle components based on the percentage of variance explained by each one. In this example, w<strong>e can observe an ‘elbow’ around PC9-10, suggesting that the majority of true signal is captured in the first 10 PCs</strong>.</p>
<div class="cell" data-hash="seurat_tutorial_cache/html/unnamed-chunk-17_c4f190b4a1a6a9e4bb222b03d5ea2452">
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/ElbowPlot.html">ElbowPlot</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="seurat_tutorial_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-4-contents" aria-controls="callout-4" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Identifying the true dimensionality of a dataset
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-4" class="callout-4-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p><strong>Identifying the true dimensionality of a dataset can be challenging/uncertain for the user.</strong> We therefore suggest these multiple approaches for users. The first is more supervised, exploring PCs to determine relevant sources of heterogeneity, and could be used in conjunction with GSEA for example. The second (<code>ElbowPlot</code>). The third is a heuristic that is commonly used, and can be calculated instantly. In this example, we might have been justified in choosing anything between PC 7-12 as a cutoff.</p>
<p>We chose 10 here, but encourage users to consider the following:</p>
<ul>
<li><p>Dendritic cell and NK aficionados may recognize that genes strongly associated with PCs 12 and 13 define rare immune subsets (i.e.&nbsp;MZB1 is a marker for plasmacytoid DCs). However, these groups are so rare, they are difficult to distinguish from background noise for a dataset of this size without prior knowledge.</p></li>
<li><p>We encourage users to repeat downstream analyses with a different number of PCs (10, 15, or even 50!). As you will observe, <strong>the results often do not differ dramatically.</strong></p></li>
<li><p>We advise users to <strong>err on the higher side when choosing this parameter（建议选择偏高的主成分数量）</strong>. For example, performing downstream analyses with only 5 PCs does significantly and adversely affect results.</p></li>
</ul>
</div>
</div>
</div>
</section><section id="sec-clustering_seurat" class="level1" data-number="8"><h1 data-number="8">
<span class="header-section-number">8</span> 细胞聚类（cluster）</h1>
<p>Seurat applies a <strong>graph-based clustering approach</strong>, building upon initial strategies in Macosko&nbsp;<em>et al</em> <span class="citation" data-cites="macosko2015">(<a href="#ref-macosko2015" role="doc-biblioref">Macosko et al. 2015</a>)</span>. Importantly, the <em>distance metric</em> which drives the clustering analysis (based on previously identified PCs) remains the same. However, our approach to partitioning the cellular distance matrix into clusters has dramatically improved. Our approach was heavily inspired by recent manuscripts which applied graph-based clustering approaches to scRNA-seq data and CyTOF data <span class="citation" data-cites="levine2015">(<a href="#ref-levine2015" role="doc-biblioref">Levine et al. 2015</a>)</span>. Briefly, these methods embed cells in a graph structure - for example a <strong>K-nearest neighbor (KNN) graph</strong>, with edges drawn between cells with similar feature expression patterns, and then attempt to partition this graph into highly interconnected ‘quasi-cliques’ or ‘communities’.</p>
<p>As in PhenoGraph, we first construct a KNN graph based on the euclidean distance in PCA space, and refine the edge weights between any two cells based on the shared overlap in their local neighborhoods (Jaccard similarity). This step is performed using the&nbsp;<strong><code><a href="https://satijalab.org/seurat/reference/FindNeighbors.html">FindNeighbors()</a></code></strong>&nbsp;function, and takes as input the <strong>previously defined dimensionality of the dataset</strong> (first 10 PCs).</p>
<p>To cluster the cells, we next apply modularity optimization techniques such as the Louvain algorithm (default) or SLM <span class="citation" data-cites="blondel2008">(<a href="#ref-blondel2008" role="doc-biblioref">Blondel et al. 2008</a>)</span>, to iteratively group cells together, with the goal of optimizing the standard modularity function. The&nbsp;<strong><code><a href="https://satijalab.org/seurat/reference/FindClusters.html">FindClusters()</a></code></strong>&nbsp;function implements this procedure, and contains a <strong>resolution parameter</strong> that sets the ‘granularity’ of the downstream clustering, with <strong>increased values leading to a greater number of clusters</strong>. We find that setting this parameter <strong>between 0.4-1.2</strong> typically returns good results for single-cell datasets of <strong>around 3K cells</strong>. Optimal resolution often increases for larger datasets.</p>
<div class="cell" data-hash="seurat_tutorial_cache/html/unnamed-chunk-18_a9d668da31c33e5559d243b8ddd29a27">
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html">FindNeighbors</a></span><span class="op">(</span><span class="va">pbmc</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html">FindClusters</a></span><span class="op">(</span><span class="va">pbmc</span>, resolution <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 2638
Number of edges: 95905

Running Louvain algorithm...
Maximum modularity in 10 random starts: 0.8735
Number of communities: 9
Elapsed time: 0 seconds</code></pre>
</div>
</div>
<p>The clusters can be found using the&nbsp;<code><a href="https://rdrr.io/pkg/SeuratObject/man/Idents.html">Idents()</a></code>&nbsp;function. 或者通过<code>seurat@active.ident</code>获取。</p>
<div class="cell" data-hash="seurat_tutorial_cache/html/unnamed-chunk-19_f4735efd5619500afdb4770a894dba58">
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Look at cluster IDs of the first 5 cells</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/Idents.html">Idents</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span>, <span class="fl">5</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>AAACATACAACCAC-1 AAACATTGAGCTAC-1 AAACATTGATCAGC-1 AAACCGTGCTTCCG-1 
               0                3                2                1 
AAACCGTGTATGCG-1 
               6 
Levels: 0 1 2 3 4 5 6 7 8</code></pre>
</div>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 等价于：head(pbmc@active.ident, 5)</span></span>
<span><span class="co"># 类似于：pbmc@meta.data[["seurat_clusters"]][1:5]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>现在的meta.data中出现了RNA_snn_res.0.5列和seurat_clusters列，分别代表了在分辨率为0.5的情况下的细胞cluster id以及目前激活的cluster id。由于目前只有一个0.5的分辨率，所以这两列是一样的。</p>
<div class="cell" data-hash="seurat_tutorial_cache/html/unnamed-chunk-20_629124c70c4a5355b739ab3821fa0624">
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                 orig.ident nCount_RNA nFeature_RNA percent.mt RNA_snn_res.0.5
AAACATACAACCAC-1     pbmc3k       2419          779  3.0177759               0
AAACATTGAGCTAC-1     pbmc3k       4903         1352  3.7935958               3
AAACATTGATCAGC-1     pbmc3k       3147         1129  0.8897363               2
AAACCGTGCTTCCG-1     pbmc3k       2639          960  1.7430845               1
AAACCGTGTATGCG-1     pbmc3k        980          521  1.2244898               6
AAACGCACTGGTAC-1     pbmc3k       2163          781  1.6643551               2
                 seurat_clusters
AAACATACAACCAC-1               0
AAACATTGAGCTAC-1               3
AAACATTGATCAGC-1               2
AAACCGTGCTTCCG-1               1
AAACCGTGTATGCG-1               6
AAACGCACTGGTAC-1               2</code></pre>
</div>
</div>
</section><section id="非线性降维" class="level1" data-number="9"><h1 data-number="9">
<span class="header-section-number">9</span> 非线性降维</h1>
<p>Seurat offers several non-linear dimensional reduction techniques, such as <strong>tSNE</strong> and <strong>UMAP</strong>, to visualize and explore these datasets. The goal of these algorithms is to learn underlying structure in the dataset, in order to <strong>place similar cells together</strong> in low-dimensional space. Therefore, cells that are grouped together within graph-based clusters determined above should co-localize on these dimension reduction plots.</p>
<p>While we and others have routinely found 2D visualization techniques like tSNE and UMAP to be valuable tools for exploring datasets, all visualization techniques have limitations, and cannot fully represent the complexity of the underlying data. In particular, these methods aim to preserve local distances in the dataset (i.e.&nbsp;ensuring that cells with very similar gene expression profiles co-localize), but often do not preserve more global relationships. <strong>We encourage users to leverage techniques like UMAP for visualization</strong>, but to avoid drawing biological conclusions solely on the basis of visualization techniques.</p>
<div class="callout callout-style-default callout-caution callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Caution
</div>
</div>
<div class="callout-body-container callout-body">
<p>需要先进行线性降维<code>RunPCA</code>，再进行非线性降维<code>RunUMAP</code>。因为<code>RunUMAP</code>默认将PCA的结果作为输入（<code>reduction = "pca"</code>）。</p>
</div>
</div>
<div class="cell" data-hash="seurat_tutorial_cache/html/unnamed-chunk-21_65da30ecac7a217cf74f4c03ba2dcbc6">
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html">RunUMAP</a></span><span class="op">(</span><span class="va">pbmc</span>, reduction <span class="op">=</span> <span class="st">"pca"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>可以看到在Seurat对象的reductions中多了umap项：</p>
<p><img src="images/截屏2023-11-21 15.14.18.png" class="img-fluid"></p>
<section id="非线性降维可视化" class="level2" data-number="9.1"><h2 data-number="9.1" class="anchored" data-anchor-id="非线性降维可视化">
<span class="header-section-number">9.1</span> 非线性降维可视化</h2>
<div class="cell" data-hash="seurat_tutorial_cache/html/unnamed-chunk-22_73b1f061bc72bfca59342a67a0d6efd0">
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">pbmc</span>, reduction <span class="op">=</span> <span class="st">"umap"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="seurat_tutorial_files/figure-html/unnamed-chunk-22-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>You can save the object at this point so that it can easily be loaded back in without having to rerun the computationally intensive steps performed above, or easily shared with collaborators.</p>
<div class="cell" data-hash="seurat_tutorial_cache/html/unnamed-chunk-23_ddc2c590f4d5c2ae5e3fb6d808f6b5bc">
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">pbmc</span>, file <span class="op">=</span> <span class="st">"output/seurat_official/pbmc_tutorial.rds"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section></section><section id="sec-seurat_cluster_biomarkers" class="level1" data-number="10"><h1 data-number="10">
<span class="header-section-number">10</span> 寻找细胞群间的差异表达基因（cluster biomarkers）</h1>
<p>Seurat can help you find markers that define clusters via <strong>differential expression (DE)</strong>. By default, it identifies positive and negative markers of a single cluster (<strong>specified in&nbsp;<code>ident.1</code></strong>), <strong>compared to all other cells</strong>.&nbsp;<code><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html">FindAllMarkers()</a></code>&nbsp;automates this process for all clusters, but you can also test groups of clusters vs.&nbsp;each other, or against all cells.</p>
<div class="callout callout-style-default callout-caution callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Caution
</div>
</div>
<div class="callout-body-container callout-body">
<p>In Seurat v5, we use the <strong>presto package</strong> (as described&nbsp;<a href="https://www.biorxiv.org/content/10.1101/653253v1">here</a>&nbsp;and available for installation&nbsp;<a href="https://github.com/immunogenomics/presto">here</a>), to dramatically improve the speed of DE analysis, particularly for large datasets.</p>
<div class="cell" data-hash="seurat_tutorial_cache/html/unnamed-chunk-24_f5df49f361f1cc176eb36965ec9e495d">
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html">install_github</a></span><span class="op">(</span><span class="st">'immunogenomics/presto'</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>For users who are not using presto, you can examine the documentation for this function (<a href="https://satijalab.org/seurat/reference/findmarkers"><code>?FindMarkers</code></a>) to explore the&nbsp;<code>min.pct</code>&nbsp;and&nbsp;<code>logfc.threshold</code>&nbsp;parameters, which can be increased in order to increase the speed of DE testing.</p>
</div>
</div>
<p>Find all markers of cluster 2：</p>
<div class="cell" data-hash="seurat_tutorial_cache/html/unnamed-chunk-25_66bf28d363a1dec24425a9c9a1982308">
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">cluster2.markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindMarkers.html">FindMarkers</a></span><span class="op">(</span><span class="va">pbmc</span>, ident.1 <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">cluster2.markers</span>, n <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>            p_val avg_log2FC pct.1 pct.2    p_val_adj
LTB  1.709675e-83   1.330256 0.982 0.647 2.344648e-79
IL32 5.076510e-83   1.242930 0.947 0.471 6.961926e-79
LDHB 2.467055e-68   1.044820 0.967 0.615 3.383320e-64
CD3D 1.817480e-66   1.058609 0.920 0.438 2.492492e-62
IL7R 8.698894e-61   1.389909 0.744 0.333 1.192966e-56</code></pre>
</div>
</div>
<p>Find all markers distinguishing cluster 5 from clusters 0 and 3：</p>
<div class="cell" data-hash="seurat_tutorial_cache/html/unnamed-chunk-26_656efe765281281756394dad6e83d1a6">
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">cluster5.markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindMarkers.html">FindMarkers</a></span><span class="op">(</span><span class="va">pbmc</span>, ident.1 <span class="op">=</span> <span class="fl">5</span>, ident.2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">cluster5.markers</span>, n <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                      p_val avg_log2FC pct.1 pct.2     p_val_adj
FCGR3A        5.972471e-204   6.795991 0.975 0.041 8.190647e-200
IFITM3        5.671364e-195   6.201036 0.975 0.048 7.777708e-191
CFD           2.389538e-193   6.081028 0.937 0.038 3.277012e-189
CD68          1.800066e-189   5.472200 0.925 0.036 2.468611e-185
RP11-290F20.3 6.852416e-189   6.390800 0.843 0.015 9.397404e-185</code></pre>
</div>
</div>
<p>Find markers for every cluster compared to all remaining cells, report only the positive ones：</p>
<div class="cell" data-hash="seurat_tutorial_cache/html/unnamed-chunk-27_5853f4d447b5e2dd69ddb952066cf840">
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pbmc.markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html">FindAllMarkers</a></span><span class="op">(</span><span class="va">pbmc</span>, only.pos <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></span>
<span><span class="va">pbmc.markers</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">avg_log2FC</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 7,024 × 7
# Groups:   cluster [9]
       p_val avg_log2FC pct.1 pct.2 p_val_adj cluster gene     
       &lt;dbl&gt;      &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt; &lt;fct&gt;   &lt;chr&gt;    
 1 5.32e-114       1.19 0.912 0.591 7.29e-110 0       LDHB     
 2 1.31e- 83       2.35 0.439 0.11  1.79e- 79 0       CCR7     
 3 2.61e- 78       1.06 0.85  0.403 3.58e- 74 0       CD3D     
 4 5.89e- 55       1.03 0.731 0.398 8.07e- 51 0       CD3E     
 5 3.91e- 50       2.11 0.338 0.104 5.36e- 46 0       LEF1     
 6 2.53e- 47       1.23 0.624 0.36  3.47e- 43 0       NOSIP    
 7 5.11e- 46       2.04 0.335 0.109 7.01e- 42 0       PRKCQ-AS1
 8 5.49e- 43       1.51 0.438 0.186 7.52e- 39 0       PIK3IP1  
 9 9.17e- 41       2.73 0.199 0.04  1.26e- 36 0       FHIT     
10 1.26e- 33       1.32 0.39  0.177 1.72e- 29 0       TCF7     
# ℹ 7,014 more rows</code></pre>
</div>
</div>
<p>Seurat has several tests for differential expression which can be set with the <code>test.use</code> parameter (see <a href="../../single_cell/seurat/de_vignette.html#sec-perform_de_analysis">执行差异分析</a> for details). 默认是用的”wilcox”方法。For example, the ROC test（<code>test.use = "roc"</code>） returns the ‘classification power’ for any individual marker (ranging from 0 - random, to 1 - perfect)：</p>
<div class="cell" data-hash="seurat_tutorial_cache/html/unnamed-chunk-28_5644535bdfbf12ebc91b387175bdda87">
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">cluster0.markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindMarkers.html">FindMarkers</a></span><span class="op">(</span><span class="va">pbmc</span>, </span>
<span>                                ident.1 <span class="op">=</span> <span class="fl">0</span>, </span>
<span>                                logfc.threshold <span class="op">=</span> <span class="fl">0.25</span>, </span>
<span>                                test.use <span class="op">=</span> <span class="st">"roc"</span>, </span>
<span>                                only.pos <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">cluster0.markers</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      myAUC  avg_diff power avg_log2FC pct.1 pct.2
RPS6  0.828 0.4737648 0.656  0.6898595 1.000 0.995
RPS12 0.827 0.5065637 0.654  0.7396535 1.000 0.991
RPL32 0.822 0.4338741 0.644  0.6330085 0.999 0.995
RPS27 0.819 0.4971128 0.638  0.7261693 0.999 0.992
RPS14 0.814 0.4354967 0.628  0.6367519 1.000 0.994
RPS25 0.809 0.5345947 0.618  0.7911714 0.997 0.975</code></pre>
</div>
</div>
</section><section id="可视化marker基因的表达" class="level1" data-number="11"><h1 data-number="11">
<span class="header-section-number">11</span> 可视化marker基因的表达</h1>
<p>We include several tools for visualizing marker expression.&nbsp;<code><a href="https://satijalab.org/seurat/reference/VlnPlot.html">VlnPlot()</a></code>&nbsp;(shows expression probability distributions across clusters), and&nbsp;<code><a href="https://satijalab.org/seurat/reference/FeaturePlot.html">FeaturePlot()</a></code>&nbsp;(visualizes feature expression on a tSNE or PCA plot) are our most commonly used visualizations. We also suggest exploring&nbsp;<code><a href="https://satijalab.org/seurat/reference/RidgePlot.html">RidgePlot()</a></code>,&nbsp;<code><a href="https://satijalab.org/seurat/reference/CellScatter.html">CellScatter()</a></code>, and&nbsp;<code><a href="https://satijalab.org/seurat/reference/DotPlot.html">DotPlot()</a></code>&nbsp;as additional methods to view your dataset.</p>
<section id="小提琴图" class="level2" data-number="11.1"><h2 data-number="11.1" class="anchored" data-anchor-id="小提琴图">
<span class="header-section-number">11.1</span> 小提琴图：</h2>
<div class="cell" data-hash="seurat_tutorial_cache/html/unnamed-chunk-29_271b7871dd18c41d80ecf1834a36b22f">
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html">VlnPlot</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"MS4A1"</span>, <span class="st">"CD79A"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="seurat_tutorial_files/figure-html/unnamed-chunk-29-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>You can plot raw counts as well（<code>layer = "counts"</code>）：</p>
<div class="cell" data-hash="seurat_tutorial_cache/html/unnamed-chunk-30_1b344cd815ec5bd137a87c629fdae57b">
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html">VlnPlot</a></span><span class="op">(</span><span class="va">pbmc</span>, </span>
<span>        features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"NKG7"</span>, <span class="st">"PF4"</span><span class="op">)</span>, </span>
<span>        layer <span class="op">=</span> <span class="st">"counts"</span>, <span class="co"># Layer to pull expression data from (e.g. "counts" or "data")</span></span>
<span>        log <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="seurat_tutorial_files/figure-html/unnamed-chunk-30-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="umap图" class="level2" data-number="11.2"><h2 data-number="11.2" class="anchored" data-anchor-id="umap图">
<span class="header-section-number">11.2</span> UMAP图：</h2>
<div class="cell" data-hash="seurat_tutorial_cache/html/unnamed-chunk-31_b568358fb54d8974bb24e1d327c5e483">
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html">FeaturePlot</a></span><span class="op">(</span><span class="va">pbmc</span>,</span>
<span>            features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"MS4A1"</span>, <span class="st">"GNLY"</span>, <span class="st">"CD3E"</span>, <span class="st">"CD14"</span>, <span class="st">"FCER1A"</span>, <span class="st">"FCGR3A"</span>, <span class="st">"LYZ"</span>,</span>
<span>                         <span class="st">"PPBP"</span>, <span class="st">"CD8A"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="seurat_tutorial_files/figure-html/unnamed-chunk-31-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="热图" class="level2" data-number="11.3"><h2 data-number="11.3" class="anchored" data-anchor-id="热图">
<span class="header-section-number">11.3</span> 热图：</h2>
<p><code><a href="https://satijalab.org/seurat/reference/DoHeatmap.html">DoHeatmap()</a></code>&nbsp;generates an expression heatmap for given cells and features. In this case, we are plotting the top 20 markers (or all markers if less than 20) for each cluster.</p>
<div class="cell" data-hash="seurat_tutorial_cache/html/unnamed-chunk-32_7a2189a3ec821fdedc72854657cce50c">
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pbmc.markers</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">avg_log2FC</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html">slice_head</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="co"># 选取开头的10行</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">ungroup</a></span><span class="op">(</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">top10</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DoHeatmap.html">DoHeatmap</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="va">top10</span><span class="op">$</span><span class="va">gene</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html">NoLegend</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="seurat_tutorial_files/figure-html/unnamed-chunk-32-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section></section><section id="细胞注释" class="level1" data-number="12"><h1 data-number="12">
<span class="header-section-number">12</span> 细胞注释</h1>
<p>本案例使用的细胞注释：</p>
<table class="table">
<thead><tr class="header">
<th style="text-align: left;">Cluster ID</th>
<th style="text-align: left;">Markers</th>
<th style="text-align: left;">Cell Type</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">0</td>
<td style="text-align: left;">IL7R, CCR7</td>
<td style="text-align: left;">Naive CD4+ T</td>
</tr>
<tr class="even">
<td style="text-align: left;">1</td>
<td style="text-align: left;">CD14, LYZ</td>
<td style="text-align: left;">CD14+ Mono</td>
</tr>
<tr class="odd">
<td style="text-align: left;">2</td>
<td style="text-align: left;">IL7R, S100A4</td>
<td style="text-align: left;">Memory CD4+</td>
</tr>
<tr class="even">
<td style="text-align: left;">3</td>
<td style="text-align: left;">MS4A1</td>
<td style="text-align: left;">B</td>
</tr>
<tr class="odd">
<td style="text-align: left;">4</td>
<td style="text-align: left;">CD8A</td>
<td style="text-align: left;">CD8+ T</td>
</tr>
<tr class="even">
<td style="text-align: left;">5</td>
<td style="text-align: left;">FCGR3A, MS4A7</td>
<td style="text-align: left;">FCGR3A+ Mono</td>
</tr>
<tr class="odd">
<td style="text-align: left;">6</td>
<td style="text-align: left;">GNLY, NKG7</td>
<td style="text-align: left;">NK</td>
</tr>
<tr class="even">
<td style="text-align: left;">7</td>
<td style="text-align: left;">FCER1A, CST3</td>
<td style="text-align: left;">DC</td>
</tr>
<tr class="odd">
<td style="text-align: left;">8</td>
<td style="text-align: left;">PPBP</td>
<td style="text-align: left;">Platelet</td>
</tr>
</tbody>
</table>
<p>目前的<code>pbmc@active.ident</code>（或<code>Idents(pbmc)</code>）和<code>meta.data</code>中的”seurat_clusters”列储存了激活的分群信息：</p>
<div class="cell" data-hash="seurat_tutorial_cache/html/unnamed-chunk-33_538ec9682749e686f13a2fd6ddff4ec7">
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/Idents.html">Idents</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>AAACATACAACCAC-1 AAACATTGAGCTAC-1 AAACATTGATCAGC-1 AAACCGTGCTTCCG-1 
               0                3                2                1 
AAACCGTGTATGCG-1 AAACGCACTGGTAC-1 
               6                2 
Levels: 0 1 2 3 4 5 6 7 8</code></pre>
</div>
</div>
<p>通过上面表格的注释依据对细胞群重命名，将命名信息储存在”new.cluster.ids”中，然后返回到meta.data中，形成新的”new.cluster.ids”列：</p>
<div class="cell" data-hash="seurat_tutorial_cache/html/unnamed-chunk-34_ca9f996e8668e83bc5b40443fbfe4199">
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">new.cluster.ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Naive CD4 T"</span>, <span class="st">"CD14+ Mono"</span>, <span class="st">"Memory CD4 T"</span>, <span class="st">"B"</span>, <span class="st">"CD8 T"</span>, </span>
<span>                     <span class="st">"FCGR3A+ Mono"</span>, <span class="st">"NK"</span>, <span class="st">"DC"</span>, <span class="st">"Platelet"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">new.cluster.ids</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/Idents.html">RenameIdents</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="va">new.cluster.ids</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">pbmc</span>, </span>
<span>        reduction <span class="op">=</span> <span class="st">"umap"</span>, </span>
<span>        label <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>        pt.size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html">NoLegend</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="seurat_tutorial_files/figure-html/unnamed-chunk-34-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>目前的分群信息：</p>
<div class="cell" data-hash="seurat_tutorial_cache/html/unnamed-chunk-35_c506b96aaa329f78dd9b5f34ea8f548f">
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/Idents.html">Idents</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>AAACATACAACCAC-1 AAACATTGAGCTAC-1 AAACATTGATCAGC-1 AAACCGTGCTTCCG-1 
     Naive CD4 T                B     Memory CD4 T       CD14+ Mono 
AAACCGTGTATGCG-1 AAACGCACTGGTAC-1 
              NK     Memory CD4 T 
9 Levels: Naive CD4 T CD14+ Mono Memory CD4 T B CD8 T FCGR3A+ Mono NK ... Platelet</code></pre>
</div>
</div>
<p><code>DimPlot</code>基于<code>ggplot2</code>绘图框架，所以可以用<code>ggplot2</code>语法对图像进行自定义调整：</p>
<div class="cell" data-hash="seurat_tutorial_cache/html/unnamed-chunk-36_14fcc68bc951760e932498084180de71">
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">pbmc</span>, </span>
<span>        reduction <span class="op">=</span> <span class="st">"umap"</span>, </span>
<span>        label <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>        label.size <span class="op">=</span> <span class="fl">4.5</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span><span class="op">(</span><span class="st">"UMAP 1"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span><span class="op">(</span><span class="st">"UMAP 2"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>axis.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">18</span><span class="op">)</span>, </span>
<span>        legend.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">18</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html">guides</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html">guide_legend</a></span><span class="op">(</span>override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="seurat_tutorial_files/figure-html/unnamed-chunk-36-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>保存Seurat对象：</p>
<div class="cell" data-hash="seurat_tutorial_cache/html/unnamed-chunk-37_10fe722fdd17fce3ec0e874ef7e19fcf">
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">pbmc</span>, file <span class="op">=</span> <span class="st">"output/seurat_official/pbmc3k_final.rds"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-7-contents" aria-controls="callout-7" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Session Info
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-7" class="callout-7-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-hash="seurat_tutorial_cache/html/unnamed-chunk-38_3c48381dd1e75aa5679a59640683acfc">
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.3.2 (2023-10-31)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Sonoma 14.3

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: Asia/Shanghai
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] ggplot2_3.4.4      dplyr_1.1.4        Seurat_5.0.1       SeuratObject_5.0.1
[5] sp_2.1-2          

loaded via a namespace (and not attached):
  [1] RColorBrewer_1.1-3     rstudioapi_0.15.0      jsonlite_1.8.8        
  [4] magrittr_2.0.3         spatstat.utils_3.0-4   ggbeeswarm_0.7.2      
  [7] farver_2.1.1           rmarkdown_2.25         vctrs_0.6.5           
 [10] ROCR_1.0-11            spatstat.explore_3.2-5 htmltools_0.5.7       
 [13] sctransform_0.4.1      parallelly_1.36.0      KernSmooth_2.23-22    
 [16] htmlwidgets_1.6.4      ica_1.0-3              plyr_1.8.9            
 [19] plotly_4.10.4          zoo_1.8-12             igraph_1.6.0          
 [22] mime_0.12              lifecycle_1.0.4        pkgconfig_2.0.3       
 [25] Matrix_1.6-5           R6_2.5.1               fastmap_1.1.1         
 [28] fitdistrplus_1.1-11    future_1.33.1          shiny_1.8.0           
 [31] digest_0.6.34          colorspace_2.1-0       patchwork_1.2.0       
 [34] tensor_1.5             RSpectra_0.16-1        irlba_2.3.5.1         
 [37] labeling_0.4.3         progressr_0.14.0       fansi_1.0.6           
 [40] spatstat.sparse_3.0-3  httr_1.4.7             polyclip_1.10-6       
 [43] abind_1.4-5            compiler_4.3.2         withr_3.0.0           
 [46] fastDummies_1.7.3      R.utils_2.12.3         MASS_7.3-60.0.1       
 [49] tools_4.3.2            vipor_0.4.7            lmtest_0.9-40         
 [52] beeswarm_0.4.0         httpuv_1.6.13          future.apply_1.11.1   
 [55] goftest_1.2-3          R.oo_1.25.0            glue_1.7.0            
 [58] nlme_3.1-164           promises_1.2.1         grid_4.3.2            
 [61] Rtsne_0.17             cluster_2.1.6          reshape2_1.4.4        
 [64] generics_0.1.3         gtable_0.3.4           spatstat.data_3.0-4   
 [67] R.methodsS3_1.8.2      tidyr_1.3.0            data.table_1.14.10    
 [70] utf8_1.2.4             spatstat.geom_3.2-7    RcppAnnoy_0.0.21      
 [73] ggrepel_0.9.5          RANN_2.6.1             pillar_1.9.0          
 [76] stringr_1.5.1          spam_2.10-0            RcppHNSW_0.5.0        
 [79] limma_3.58.1           later_1.3.2            splines_4.3.2         
 [82] lattice_0.22-5         survival_3.5-7         deldir_2.0-2          
 [85] tidyselect_1.2.0       miniUI_0.1.1.1         pbapply_1.7-2         
 [88] knitr_1.45             gridExtra_2.3          scattermore_1.2       
 [91] xfun_0.41              statmod_1.5.0          matrixStats_1.2.0     
 [94] stringi_1.8.3          lazyeval_0.2.2         yaml_2.3.8            
 [97] evaluate_0.23          codetools_0.2-19       tibble_3.2.1          
[100] cli_3.6.2              uwot_0.1.16            xtable_1.8-4          
[103] reticulate_1.34.0      munsell_0.5.0          Rcpp_1.0.12           
[106] globals_0.16.2         spatstat.random_3.2-2  png_0.1-8             
[109] ggrastr_1.0.2          parallel_4.3.2         ellipsis_0.3.2        
[112] presto_1.0.0           dotCall64_1.1-1        listenv_0.9.0         
[115] viridisLite_0.4.2      scales_1.3.0           ggridges_0.5.5        
[118] leiden_0.4.3.1         purrr_1.0.2            rlang_1.1.3           
[121] cowplot_1.1.2         </code></pre>
</div>
</div>
</div>
</div>
</div>
</section><section id="references" class="level1 unnumbered"><h1 class="unnumbered">References</h1>
<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-blondel2008" class="csl-entry" role="listitem">
Blondel, Vincent D, Jean-Loup Guillaume, Renaud Lambiotte, and Etienne Lefebvre. 2008. <span>“Fast Unfolding of Communities in Large Networks.”</span> <em>Journal of Statistical Mechanics: Theory and Experiment</em> 2008 (10): P10008. <a href="https://doi.org/10.1088/1742-5468/2008/10/p10008">https://doi.org/10.1088/1742-5468/2008/10/p10008</a>.
</div>
<div id="ref-brennecke2013" class="csl-entry" role="listitem">
Brennecke, Philip, Simon Anders, Jong Kyoung Kim, Aleksandra A Kołodziejczyk, Xiuwei Zhang, Valentina Proserpio, Bianka Baying, et al. 2013. <span>“Accounting for Technical Noise in Single-Cell RNA-Seq Experiments.”</span> <em>Nature Methods</em> 10 (11): 1093–95. <a href="https://doi.org/10.1038/nmeth.2645">https://doi.org/10.1038/nmeth.2645</a>.
</div>
<div id="ref-choudhary2022" class="csl-entry" role="listitem">
Choudhary, Saket, and Rahul Satija. 2022. <span>“Comparison and Evaluation of Statistical Error Models for scRNA-Seq.”</span> <em>Genome Biology</em> 23 (1). <a href="https://doi.org/10.1186/s13059-021-02584-9">https://doi.org/10.1186/s13059-021-02584-9</a>.
</div>
<div id="ref-ilicic2016" class="csl-entry" role="listitem">
Ilicic, Tomislav, Jong Kyoung Kim, Aleksandra A. Kolodziejczyk, Frederik Otzen Bagger, Davis James McCarthy, John C. Marioni, and Sarah A. Teichmann. 2016. <span>“Classification of Low Quality Cells from Single-Cell RNA-Seq Data.”</span> <em>Genome Biology</em> 17 (1). <a href="https://doi.org/10.1186/s13059-016-0888-1">https://doi.org/10.1186/s13059-016-0888-1</a>.
</div>
<div id="ref-levine2015" class="csl-entry" role="listitem">
Levine, Jacob&nbsp;H., Erin&nbsp;F. Simonds, Sean&nbsp;C. Bendall, Kara&nbsp;L. Davis, El-ad&nbsp;D. Amir, Michelle&nbsp;D. Tadmor, Oren Litvin, et al. 2015. <span>“Data-Driven Phenotypic Dissection of AML Reveals Progenitor-Like Cells That Correlate with Prognosis.”</span> <em>Cell</em> 162 (1): 184–97. <a href="https://doi.org/10.1016/j.cell.2015.05.047">https://doi.org/10.1016/j.cell.2015.05.047</a>.
</div>
<div id="ref-macosko2015" class="csl-entry" role="listitem">
Macosko, Evan&nbsp;Z., Anindita Basu, Rahul Satija, James Nemesh, Karthik Shekhar, Melissa Goldman, Itay Tirosh, et al. 2015. <span>“Highly Parallel Genome-Wide Expression Profiling of Individual Cells Using Nanoliter Droplets.”</span> <em>Cell</em> 161 (5): 1202–14. <a href="https://doi.org/10.1016/j.cell.2015.05.002">https://doi.org/10.1016/j.cell.2015.05.002</a>.
</div>
<div id="ref-stuart2019" class="csl-entry" role="listitem">
Stuart, Tim, Andrew Butler, Paul Hoffman, Christoph Hafemeister, Efthymia Papalexi, William M. Mauck, Yuhan Hao, Marlon Stoeckius, Peter Smibert, and Rahul Satija. 2019. <span>“Comprehensive Integration of Single-Cell Data.”</span> <em>Cell</em> 177 (7): 1888–1902.e21. <a href="https://doi.org/10.1016/j.cell.2019.05.031">https://doi.org/10.1016/j.cell.2019.05.031</a>.
</div>
</div>


<!-- -->

</section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/djhcod\.github\.io\/r-notes\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
          // target, if specified
          link.setAttribute("target", "_blank");
          if (link.getAttribute("rel") === null) {
            link.setAttribute("rel", "noopener");
          }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><script src="https://giscus.app/client.js" data-repo="djhcod/r-notes" data-repo-id="R_kgDOKw7GMg" data-category="General" data-category-id="DIC_kwDOKw7GMs4Cbz4B" data-mapping="title" data-reactions-enabled="1" data-emit-metadata="0" data-input-position="top" data-theme="light" data-lang="en" crossorigin="anonymous" async="">
</script><input type="hidden" id="giscus-base-theme" value="light"><input type="hidden" id="giscus-alt-theme" value="dark"><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="../../single_cell/seurat/seurat_command_list.html" class="pagination-link  aria-label=" seurat="">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Seurat常用函数清单</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../../single_cell/seurat/data_visualization_methods_in_seurat.html" class="pagination-link" aria-label="Seurat中的数据可视化方法">
        <span class="nav-page-text">Seurat中的数据可视化方法</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb53" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> "Seurat细胞分群官方教程"</span></span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a><span class="an">editor_options:</span><span class="co"> </span></span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a><span class="co">  chunk_output_type: console</span></span>
<span id="cb53-5"><a href="#cb53-5" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb53-6"><a href="#cb53-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-7"><a href="#cb53-7" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; 原文：</span><span class="co">[</span><span class="ot">*Seurat - Guided Clustering Tutorial*</span><span class="co">](https://satijalab.org/seurat/articles/pbmc3k_tutorial)</span></span>
<span id="cb53-8"><a href="#cb53-8" aria-hidden="true" tabindex="-1"></a><span class="at">&gt;</span></span>
<span id="cb53-9"><a href="#cb53-9" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; 原文发布日期：2023年10月31日</span></span>
<span id="cb53-10"><a href="#cb53-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-11"><a href="#cb53-11" aria-hidden="true" tabindex="-1"></a><span class="fu"># Seurat对象构建</span></span>
<span id="cb53-12"><a href="#cb53-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-13"><a href="#cb53-13" aria-hidden="true" tabindex="-1"></a>数据源是来自10X Genomics的**外周血单核细胞（peripheral blood mononuclear cells，PBMC）数据集**。该数据集基于Illumina NextSeq 500平台对2700个单细胞进行了测序。数据可在<span class="co">[</span><span class="ot">此链接</span><span class="co">](https://cf.10xgenomics.com/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz)</span>下载。</span>
<span id="cb53-14"><a href="#cb53-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-15"><a href="#cb53-15" aria-hidden="true" tabindex="-1"></a><span class="al">![10X数据的标准文件](images/2023-11-20%2013.40.27.png)</span>{width="449"}</span>
<span id="cb53-16"><a href="#cb53-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-17"><a href="#cb53-17" aria-hidden="true" tabindex="-1"></a>该数据已经通过<span class="co">[</span><span class="ot">cellranger</span><span class="co">](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger)</span>上游数据处理流程的处理，返回的数据是一个由唯一分子识别（unique molecular identified，UMI）构成的count矩阵。该矩阵中的值表示在每个细胞（列）中检测到的每个特征（即基因；行）的分子数量。数据以10X的标准形式储存，包括：</span>
<span id="cb53-18"><a href="#cb53-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-19"><a href="#cb53-19" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>barcode文件：细胞条码。</span>
<span id="cb53-20"><a href="#cb53-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-21"><a href="#cb53-21" aria-hidden="true" tabindex="-1"></a>    <span class="al">![](images/2023-11-20%2014.04.06.png)</span>{width="161"}</span>
<span id="cb53-22"><a href="#cb53-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-23"><a href="#cb53-23" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>genes文件：基因名文件。</span>
<span id="cb53-24"><a href="#cb53-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-25"><a href="#cb53-25" aria-hidden="true" tabindex="-1"></a>    <span class="al">![](images/2023-11-20%2014.04.15.png)</span>{width="229"}</span>
<span id="cb53-26"><a href="#cb53-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-27"><a href="#cb53-27" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>matrix文件：表达矩阵（稀疏矩阵）。</span>
<span id="cb53-28"><a href="#cb53-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-31"><a href="#cb53-31" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb53-32"><a href="#cb53-32" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb53-33"><a href="#cb53-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-34"><a href="#cb53-34" aria-hidden="true" tabindex="-1"></a><span class="co"># 读取PBMC数据集</span></span>
<span id="cb53-35"><a href="#cb53-35" aria-hidden="true" tabindex="-1"></a>pbmc.data <span class="ot">&lt;-</span> <span class="fu">Read10X</span>(<span class="at">data.dir =</span> <span class="st">"data/seurat_official/filtered_gene_bc_matrices/hg19"</span>)</span>
<span id="cb53-36"><a href="#cb53-36" aria-hidden="true" tabindex="-1"></a><span class="co"># Initialize the Seurat object with the raw (non-normalized data).</span></span>
<span id="cb53-37"><a href="#cb53-37" aria-hidden="true" tabindex="-1"></a>pbmc <span class="ot">&lt;-</span> <span class="fu">CreateSeuratObject</span>(<span class="at">counts =</span> pbmc.data, </span>
<span id="cb53-38"><a href="#cb53-38" aria-hidden="true" tabindex="-1"></a>                           <span class="at">project =</span> <span class="st">"pbmc3k"</span>, </span>
<span id="cb53-39"><a href="#cb53-39" aria-hidden="true" tabindex="-1"></a>                           <span class="at">min.cells =</span> <span class="dv">3</span>, </span>
<span id="cb53-40"><a href="#cb53-40" aria-hidden="true" tabindex="-1"></a>                           <span class="at">min.features =</span> <span class="dv">200</span>)</span>
<span id="cb53-41"><a href="#cb53-41" aria-hidden="true" tabindex="-1"></a>pbmc</span>
<span id="cb53-42"><a href="#cb53-42" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb53-43"><a href="#cb53-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-44"><a href="#cb53-44" aria-hidden="true" tabindex="-1"></a>::: callout-caution</span>
<span id="cb53-45"><a href="#cb53-45" aria-hidden="true" tabindex="-1"></a>有的cellranger处理数据以<span class="co">[</span><span class="ot">h5文件格式</span><span class="co">](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/advanced/h5_matrices)</span>输出，需要使用seurat中的<span class="co">[</span><span class="ot">`Read10X_h5()`</span><span class="co">](https://satijalab.org/seurat/reference/read10x_h5)</span>函数读取该格式。</span>
<span id="cb53-46"><a href="#cb53-46" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb53-47"><a href="#cb53-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-48"><a href="#cb53-48" aria-hidden="true" tabindex="-1"></a><span class="fu"># 质控</span></span>
<span id="cb53-49"><a href="#cb53-49" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-50"><a href="#cb53-50" aria-hidden="true" tabindex="-1"></a>在 <span class="co">[</span><span class="ot">@ilicic2016</span><span class="co">]</span> 中概括了目前常用的单细胞数据分析中识别低质量细胞的标准，包括：</span>
<span id="cb53-51"><a href="#cb53-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-52"><a href="#cb53-52" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>The **number of unique genes** detected in each cell.</span>
<span id="cb53-53"><a href="#cb53-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-54"><a href="#cb53-54" aria-hidden="true" tabindex="-1"></a><span class="ss">    -   </span>**Low-quality cells** or **empty droplets** will often have very few genes</span>
<span id="cb53-55"><a href="#cb53-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-56"><a href="#cb53-56" aria-hidden="true" tabindex="-1"></a><span class="ss">    -   </span>**Cell doublets** or multiplets may exhibit an aberrantly high gene count</span>
<span id="cb53-57"><a href="#cb53-57" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-58"><a href="#cb53-58" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Similarly, the total number of molecules detected within a cell (correlates strongly with unique genes)</span>
<span id="cb53-59"><a href="#cb53-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-60"><a href="#cb53-60" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>The percentage of reads that map to the mitochondrial genome</span>
<span id="cb53-61"><a href="#cb53-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-62"><a href="#cb53-62" aria-hidden="true" tabindex="-1"></a><span class="ss">    -   </span>Low-quality / dying cells often exhibit extensive mitochondrial contamination</span>
<span id="cb53-63"><a href="#cb53-63" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-64"><a href="#cb53-64" aria-hidden="true" tabindex="-1"></a><span class="ss">    -   </span>We calculate mitochondrial QC metrics with the&nbsp;<span class="in">`PercentageFeatureSet()`</span>&nbsp;function, which calculates the percentage of counts originating from a set of features</span>
<span id="cb53-65"><a href="#cb53-65" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-66"><a href="#cb53-66" aria-hidden="true" tabindex="-1"></a><span class="ss">    -   </span>We use the set of all genes starting with&nbsp;<span class="in">`MT-`</span>&nbsp;as a set of mitochondrial genes</span>
<span id="cb53-67"><a href="#cb53-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-68"><a href="#cb53-68" aria-hidden="true" tabindex="-1"></a><span class="fu">## 计算线粒体基因比例</span></span>
<span id="cb53-69"><a href="#cb53-69" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-70"><a href="#cb53-70" aria-hidden="true" tabindex="-1"></a>通过<span class="in">`PercentageFeatureSet()`</span>函数计算每个细胞中线粒体基因的比例，并将其返回到Seurat对象的<span class="in">`meta.data`</span>中，形成一个新列"percent.mt"。</span>
<span id="cb53-71"><a href="#cb53-71" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-74"><a href="#cb53-74" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb53-75"><a href="#cb53-75" aria-hidden="true" tabindex="-1"></a>pbmc[[<span class="st">"percent.mt"</span>]] <span class="ot">&lt;-</span> <span class="fu">PercentageFeatureSet</span>(pbmc, <span class="at">pattern =</span> <span class="st">"^MT-"</span>)</span>
<span id="cb53-76"><a href="#cb53-76" aria-hidden="true" tabindex="-1"></a><span class="co"># Show QC metrics for the first 5 cells</span></span>
<span id="cb53-77"><a href="#cb53-77" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(pbmc<span class="sc">@</span>meta.data, <span class="dv">5</span>)</span>
<span id="cb53-78"><a href="#cb53-78" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb53-79"><a href="#cb53-79" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-80"><a href="#cb53-80" aria-hidden="true" tabindex="-1"></a><span class="fu">## 可视化质控指标</span></span>
<span id="cb53-81"><a href="#cb53-81" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-82"><a href="#cb53-82" aria-hidden="true" tabindex="-1"></a>通过<span class="in">`VlnPlot`</span>函数绘制小提琴图展示每个细胞中UMI（nCount_RNA）、基因（percent.mt）和线粒体基因（percent.mt）的数量。</span>
<span id="cb53-83"><a href="#cb53-83" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-86"><a href="#cb53-86" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb53-87"><a href="#cb53-87" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualize QC metrics as a violin plot</span></span>
<span id="cb53-88"><a href="#cb53-88" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(pbmc, </span>
<span id="cb53-89"><a href="#cb53-89" aria-hidden="true" tabindex="-1"></a>        <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"nFeature_RNA"</span>, <span class="st">"nCount_RNA"</span>, <span class="st">"percent.mt"</span>), </span>
<span id="cb53-90"><a href="#cb53-90" aria-hidden="true" tabindex="-1"></a>        <span class="at">ncol =</span> <span class="dv">3</span>)</span>
<span id="cb53-91"><a href="#cb53-91" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb53-92"><a href="#cb53-92" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-93"><a href="#cb53-93" aria-hidden="true" tabindex="-1"></a>通过<span class="in">`FeatureScatter`</span>函数展示UMI和线粒体基因数量多关系、UMI和总基因数量多关系。</span>
<span id="cb53-94"><a href="#cb53-94" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-97"><a href="#cb53-97" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb53-98"><a href="#cb53-98" aria-hidden="true" tabindex="-1"></a>plot1 <span class="ot">&lt;-</span> <span class="fu">FeatureScatter</span>(pbmc, </span>
<span id="cb53-99"><a href="#cb53-99" aria-hidden="true" tabindex="-1"></a>                        <span class="at">feature1 =</span> <span class="st">"nCount_RNA"</span>, </span>
<span id="cb53-100"><a href="#cb53-100" aria-hidden="true" tabindex="-1"></a>                        <span class="at">feature2 =</span> <span class="st">"percent.mt"</span>)</span>
<span id="cb53-101"><a href="#cb53-101" aria-hidden="true" tabindex="-1"></a>plot2 <span class="ot">&lt;-</span> <span class="fu">FeatureScatter</span>(pbmc, </span>
<span id="cb53-102"><a href="#cb53-102" aria-hidden="true" tabindex="-1"></a>                        <span class="at">feature1 =</span> <span class="st">"nCount_RNA"</span>, </span>
<span id="cb53-103"><a href="#cb53-103" aria-hidden="true" tabindex="-1"></a>                        <span class="at">feature2 =</span> <span class="st">"nFeature_RNA"</span>)</span>
<span id="cb53-104"><a href="#cb53-104" aria-hidden="true" tabindex="-1"></a>plot1 <span class="sc">+</span> plot2</span>
<span id="cb53-105"><a href="#cb53-105" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb53-106"><a href="#cb53-106" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-107"><a href="#cb53-107" aria-hidden="true" tabindex="-1"></a><span class="fu">## 过滤细胞</span></span>
<span id="cb53-108"><a href="#cb53-108" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-109"><a href="#cb53-109" aria-hidden="true" tabindex="-1"></a>在本案例中使用的质控标准：</span>
<span id="cb53-110"><a href="#cb53-110" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-111"><a href="#cb53-111" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>filter cells that have **unique feature counts over 2,500 or less than 200**</span>
<span id="cb53-112"><a href="#cb53-112" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-113"><a href="#cb53-113" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>filter cells that have **\&gt;5% mitochondrial counts**</span>
<span id="cb53-114"><a href="#cb53-114" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-117"><a href="#cb53-117" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb53-118"><a href="#cb53-118" aria-hidden="true" tabindex="-1"></a>pbmc <span class="ot">&lt;-</span> <span class="fu">subset</span>(pbmc, </span>
<span id="cb53-119"><a href="#cb53-119" aria-hidden="true" tabindex="-1"></a>               <span class="at">subset =</span> nFeature_RNA <span class="sc">&gt;</span> <span class="dv">200</span> <span class="sc">&amp;</span> nFeature_RNA <span class="sc">&lt;</span> <span class="dv">2500</span> <span class="sc">&amp;</span> percent.mt <span class="sc">&lt;</span> <span class="dv">5</span>)</span>
<span id="cb53-120"><a href="#cb53-120" aria-hidden="true" tabindex="-1"></a>pbmc</span>
<span id="cb53-121"><a href="#cb53-121" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb53-122"><a href="#cb53-122" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-123"><a href="#cb53-123" aria-hidden="true" tabindex="-1"></a><span class="fu"># 标准化（Normalizing）</span></span>
<span id="cb53-124"><a href="#cb53-124" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-125"><a href="#cb53-125" aria-hidden="true" tabindex="-1"></a>After removing unwanted cells from the dataset, the next step is to normalize the data. By default, we employ a **global-scaling normalization method "LogNormalize"** that normalizes the feature expression measurements for each cell by the total expression, multiplies this by a **scale factor** (10,000 by default), and log-transforms the result. 标准化的方法是每个细胞中的基因表达量除以该细胞的基因总表达量，然后乘以比例因子（通常是1000），最后将这个结果取对数。</span>
<span id="cb53-126"><a href="#cb53-126" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-129"><a href="#cb53-129" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb53-130"><a href="#cb53-130" aria-hidden="true" tabindex="-1"></a>pbmc <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(pbmc, </span>
<span id="cb53-131"><a href="#cb53-131" aria-hidden="true" tabindex="-1"></a>                      <span class="at">normalization.method =</span> <span class="st">"LogNormalize"</span>, </span>
<span id="cb53-132"><a href="#cb53-132" aria-hidden="true" tabindex="-1"></a>                      <span class="at">scale.factor =</span> <span class="dv">10000</span>)</span>
<span id="cb53-133"><a href="#cb53-133" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb53-134"><a href="#cb53-134" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-135"><a href="#cb53-135" aria-hidden="true" tabindex="-1"></a>In Seurat v5, Normalized values are stored in&nbsp;<span class="in">`pbmc[["RNA"]]$data`</span>：</span>
<span id="cb53-136"><a href="#cb53-136" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-137"><a href="#cb53-137" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/截屏2023-11-21%2011.17.58.png)</span>{width="599"}</span>
<span id="cb53-138"><a href="#cb53-138" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-139"><a href="#cb53-139" aria-hidden="true" tabindex="-1"></a>::: callout-tip</span>
<span id="cb53-140"><a href="#cb53-140" aria-hidden="true" tabindex="-1"></a>While this method of normalization is standard and widely used in scRNA-seq analysis, global-scaling relies on an **assumption that each cell originally contains the same number of RNA molecules**. We and others have developed alternative workflows for the single cell preprocessing that do not make these assumptions.</span>
<span id="cb53-141"><a href="#cb53-141" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-142"><a href="#cb53-142" aria-hidden="true" tabindex="-1"></a>SCTransform() normalization workflow 就是这样的一种工作流. 针对这一方法的原始文献：<span class="co">[</span><span class="ot">@choudhary2022</span><span class="co">]</span> 。基于<span class="in">`SCTransform`</span>的数据标准化流程中，无需使用上面的<span class="in">`NormalizeData`</span>,&nbsp;和下面的<span class="in">`FindVariableFeatures`</span>和<span class="in">`ScaleData`</span>函数。详见<span class="co">[</span><span class="ot">后面的章节</span><span class="co">](/single_cell/seurat/sctransform.qmd)</span>。</span>
<span id="cb53-143"><a href="#cb53-143" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb53-144"><a href="#cb53-144" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-145"><a href="#cb53-145" aria-hidden="true" tabindex="-1"></a><span class="fu"># 识别高变基因（highly variable features） {#sec-highly_variable_features}</span></span>
<span id="cb53-146"><a href="#cb53-146" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-147"><a href="#cb53-147" aria-hidden="true" tabindex="-1"></a>We next calculate a subset of features that exhibit **high cell-to-cell variation** in the dataset (i.e, they are highly expressed in some cells, and lowly expressed in others). We and others [@brennecke2013] have found that focusing on these genes in downstream analysis helps to **highlight biological signal** in single-cell datasets.</span>
<span id="cb53-148"><a href="#cb53-148" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-149"><a href="#cb53-149" aria-hidden="true" tabindex="-1"></a>Our procedure in Seurat is described in detail here <span class="co">[</span><span class="ot">@stuart2019</span><span class="co">]</span> , and improves on previous versions by directly modeling the mean-variance relationship inherent in single-cell data, and is implemented in the&nbsp;<span class="in">`FindVariableFeatures()`</span>&nbsp;function. **By default, we return 2,000 features per dataset**. These will be used in downstream analysis, like PCA.</span>
<span id="cb53-150"><a href="#cb53-150" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-153"><a href="#cb53-153" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb53-154"><a href="#cb53-154" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-识别高变基因</span></span>
<span id="cb53-155"><a href="#cb53-155" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "识别高变基因"</span></span>
<span id="cb53-156"><a href="#cb53-156" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-subcap: </span></span>
<span id="cb53-157"><a href="#cb53-157" aria-hidden="true" tabindex="-1"></a><span class="co">#|   - "前2000个高变基因"</span></span>
<span id="cb53-158"><a href="#cb53-158" aria-hidden="true" tabindex="-1"></a><span class="co">#|   - "前2000个高变基因（标注了前10个高变基因）"</span></span>
<span id="cb53-159"><a href="#cb53-159" aria-hidden="true" tabindex="-1"></a><span class="co">#| layout-ncol: 2</span></span>
<span id="cb53-160"><a href="#cb53-160" aria-hidden="true" tabindex="-1"></a>pbmc <span class="ot">&lt;-</span> <span class="fu">FindVariableFeatures</span>(pbmc, </span>
<span id="cb53-161"><a href="#cb53-161" aria-hidden="true" tabindex="-1"></a>                             <span class="at">selection.method =</span> <span class="st">"vst"</span>, </span>
<span id="cb53-162"><a href="#cb53-162" aria-hidden="true" tabindex="-1"></a>                             <span class="at">nfeatures =</span> <span class="dv">2000</span>)</span>
<span id="cb53-163"><a href="#cb53-163" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-164"><a href="#cb53-164" aria-hidden="true" tabindex="-1"></a><span class="co"># Identify the 10 most highly variable genes</span></span>
<span id="cb53-165"><a href="#cb53-165" aria-hidden="true" tabindex="-1"></a>top10 <span class="ot">&lt;-</span> <span class="fu">head</span>(<span class="fu">VariableFeatures</span>(pbmc), <span class="dv">10</span>)</span>
<span id="cb53-166"><a href="#cb53-166" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-167"><a href="#cb53-167" aria-hidden="true" tabindex="-1"></a><span class="co"># plot variable features with and without labels</span></span>
<span id="cb53-168"><a href="#cb53-168" aria-hidden="true" tabindex="-1"></a>plot1 <span class="ot">&lt;-</span> <span class="fu">VariableFeaturePlot</span>(pbmc)</span>
<span id="cb53-169"><a href="#cb53-169" aria-hidden="true" tabindex="-1"></a>plot1</span>
<span id="cb53-170"><a href="#cb53-170" aria-hidden="true" tabindex="-1"></a><span class="fu">LabelPoints</span>(<span class="at">plot =</span> plot1, <span class="at">points =</span> top10, <span class="at">repel =</span> <span class="cn">TRUE</span>)</span>
<span id="cb53-171"><a href="#cb53-171" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb53-172"><a href="#cb53-172" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-173"><a href="#cb53-173" aria-hidden="true" tabindex="-1"></a><span class="fu"># 归一化（Scaling） {#sec-scaledata}</span></span>
<span id="cb53-174"><a href="#cb53-174" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-175"><a href="#cb53-175" aria-hidden="true" tabindex="-1"></a>Next, we apply a **linear transformation** ('scaling') that is a standard pre-processing step **prior to dimensional reduction** techniques like PCA. The&nbsp;<span class="in">`ScaleData()`</span>&nbsp;function:</span>
<span id="cb53-176"><a href="#cb53-176" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-177"><a href="#cb53-177" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Shifts the expression of each gene, so that **the mean expression across cells is 0（每个基因在所有细胞的平均表达量为0）**</span>
<span id="cb53-178"><a href="#cb53-178" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-179"><a href="#cb53-179" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Scales the expression of each gene, so that **the variance across cells is 1.** This step gives equal weight in downstream analyses, so that **highly-expressed genes do not dominate**</span>
<span id="cb53-180"><a href="#cb53-180" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-183"><a href="#cb53-183" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb53-184"><a href="#cb53-184" aria-hidden="true" tabindex="-1"></a>pbmc <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(pbmc)</span>
<span id="cb53-185"><a href="#cb53-185" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb53-186"><a href="#cb53-186" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-187"><a href="#cb53-187" aria-hidden="true" tabindex="-1"></a>The results of this are stored in&nbsp;<span class="in">`pbmc[["RNA"]]$scale.data`</span>:</span>
<span id="cb53-188"><a href="#cb53-188" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-189"><a href="#cb53-189" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/2023-11-21%2012.04.27.png)</span></span>
<span id="cb53-190"><a href="#cb53-190" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-191"><a href="#cb53-191" aria-hidden="true" tabindex="-1"></a><span class="in">`ScaleData`</span>函数默认情况下使用<span class="in">`FindVariableFeatures`</span>函数确定的高变基因作为scale依据。可以通过<span class="in">`features`</span>参数来自定义scale依据，比如这里我们可以让所有基因都参与scale：</span>
<span id="cb53-192"><a href="#cb53-192" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-195"><a href="#cb53-195" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb53-196"><a href="#cb53-196" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb53-197"><a href="#cb53-197" aria-hidden="true" tabindex="-1"></a><span class="co"># 不运行</span></span>
<span id="cb53-198"><a href="#cb53-198" aria-hidden="true" tabindex="-1"></a>all.genes <span class="ot">&lt;-</span> <span class="fu">rownames</span>(pbmc)</span>
<span id="cb53-199"><a href="#cb53-199" aria-hidden="true" tabindex="-1"></a>pbmc <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(pbmc, <span class="at">features =</span> all.genes)</span>
<span id="cb53-200"><a href="#cb53-200" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb53-201"><a href="#cb53-201" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-202"><a href="#cb53-202" aria-hidden="true" tabindex="-1"></a>::: callout-tip</span>
<span id="cb53-203"><a href="#cb53-203" aria-hidden="true" tabindex="-1"></a><span class="fu">###### 去除单细胞数据中的非期望变异来源</span></span>
<span id="cb53-204"><a href="#cb53-204" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-205"><a href="#cb53-205" aria-hidden="true" tabindex="-1"></a>In Seurat, we also use the&nbsp;<span class="in">`ScaleData()`</span>&nbsp;function to remove unwanted sources of variation from a single-cell dataset. For example, we could 'regress out' heterogeneity associated with (for example)&nbsp;cell cycle stage (见<span class="co">[</span><span class="ot">消除细胞周期的影响</span><span class="co">](/single_cell/seurat/cell_cycle_regression.qmd)</span>), or mitochondrial contamination i.e.:</span>
<span id="cb53-206"><a href="#cb53-206" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-209"><a href="#cb53-209" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb53-210"><a href="#cb53-210" aria-hidden="true" tabindex="-1"></a>pbmc <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(pbmc, <span class="at">vars.to.regress =</span> <span class="st">"percent.mt"</span>)</span>
<span id="cb53-211"><a href="#cb53-211" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb53-212"><a href="#cb53-212" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-213"><a href="#cb53-213" aria-hidden="true" tabindex="-1"></a>However, particularly for advanced users who would like to use this functionality, **we strongly recommend the use of our new normalization workflow,&nbsp;`SCTransform()`**. The method is described in our&nbsp;paper <span class="co">[</span><span class="ot">@choudhary2022</span><span class="co">]</span>. As with&nbsp;<span class="in">`ScaleData()`</span>, the function&nbsp;<span class="in">`SCTransform()`</span>&nbsp;also includes a&nbsp;<span class="in">`vars.to.regress`</span>&nbsp;parameter (见<span class="co">[</span><span class="ot">运行`SCTransform`</span><span class="co">](/single_cell/seurat/sctransform.qmd#sec-perform_sctransform)</span>).&nbsp;</span>
<span id="cb53-214"><a href="#cb53-214" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb53-215"><a href="#cb53-215" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-216"><a href="#cb53-216" aria-hidden="true" tabindex="-1"></a><span class="fu"># 线性降维</span></span>
<span id="cb53-217"><a href="#cb53-217" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-218"><a href="#cb53-218" aria-hidden="true" tabindex="-1"></a>Next we perform PCA on the scaled data. By default, **only the previously determined variable features are used as input,** but can be defined using&nbsp;<span class="in">`features`</span>&nbsp;argument if you wish to choose a different subset (if you do want to use a custom subset of features, make sure you pass these to&nbsp;<span class="in">`ScaleData`</span>&nbsp;first).</span>
<span id="cb53-219"><a href="#cb53-219" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-222"><a href="#cb53-222" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb53-223"><a href="#cb53-223" aria-hidden="true" tabindex="-1"></a>pbmc <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(pbmc, </span>
<span id="cb53-224"><a href="#cb53-224" aria-hidden="true" tabindex="-1"></a>               <span class="at">features =</span> <span class="fu">VariableFeatures</span>(<span class="at">object =</span> pbmc))</span>
<span id="cb53-225"><a href="#cb53-225" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb53-226"><a href="#cb53-226" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-227"><a href="#cb53-227" aria-hidden="true" tabindex="-1"></a>For the first principal components, Seurat outputs a list of genes with the most positive and negative loadings, representing modules of genes that exhibit either correlation (or anti-correlation) across single-cells in the dataset.</span>
<span id="cb53-228"><a href="#cb53-228" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-231"><a href="#cb53-231" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb53-232"><a href="#cb53-232" aria-hidden="true" tabindex="-1"></a><span class="co"># PC_ 1 </span></span>
<span id="cb53-233"><a href="#cb53-233" aria-hidden="true" tabindex="-1"></a><span class="co"># Positive:  CST3, TYROBP, LST1, AIF1, FTL, FTH1, LYZ, FCN1, S100A9, TYMP </span></span>
<span id="cb53-234"><a href="#cb53-234" aria-hidden="true" tabindex="-1"></a><span class="co">#      FCER1G, CFD, LGALS1, LGALS2, SERPINA1, S100A8, CTSS, IFITM3, SPI1, CFP </span></span>
<span id="cb53-235"><a href="#cb53-235" aria-hidden="true" tabindex="-1"></a><span class="co">#      PSAP, IFI30, COTL1, SAT1, S100A11, NPC2, GRN, LGALS3, GSTP1, PYCARD </span></span>
<span id="cb53-236"><a href="#cb53-236" aria-hidden="true" tabindex="-1"></a><span class="co"># Negative:  MALAT1, LTB, IL32, IL7R, CD2, B2M, ACAP1, CTSW, STK17A, CD27 </span></span>
<span id="cb53-237"><a href="#cb53-237" aria-hidden="true" tabindex="-1"></a><span class="co">#      CD247, CCL5, GIMAP5, GZMA, AQP3, CST7, TRAF3IP3, SELL, GZMK, HOPX </span></span>
<span id="cb53-238"><a href="#cb53-238" aria-hidden="true" tabindex="-1"></a><span class="co">#      MAL, MYC, ITM2A, ETS1, LYAR, GIMAP7, KLRG1, NKG7, ZAP70, BEX2 </span></span>
<span id="cb53-239"><a href="#cb53-239" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb53-240"><a href="#cb53-240" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-241"><a href="#cb53-241" aria-hidden="true" tabindex="-1"></a>完成PCA分析的Seurat对象：</span>
<span id="cb53-242"><a href="#cb53-242" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-243"><a href="#cb53-243" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/截屏2023-11-21%2015.04.58.png)</span></span>
<span id="cb53-244"><a href="#cb53-244" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-245"><a href="#cb53-245" aria-hidden="true" tabindex="-1"></a><span class="fu">## 降维可视化</span></span>
<span id="cb53-246"><a href="#cb53-246" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-247"><a href="#cb53-247" aria-hidden="true" tabindex="-1"></a>Seurat provides several useful ways of visualizing both cells and features that define the PCA, including&nbsp;<span class="in">`VizDimReduction()`</span>,&nbsp;<span class="in">`DimPlot()`</span>, and&nbsp;<span class="in">`DimHeatmap()`</span> .</span>
<span id="cb53-248"><a href="#cb53-248" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-251"><a href="#cb53-251" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb53-252"><a href="#cb53-252" aria-hidden="true" tabindex="-1"></a><span class="fu">VizDimLoadings</span>(pbmc, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>, <span class="at">reduction =</span> <span class="st">"pca"</span>)</span>
<span id="cb53-253"><a href="#cb53-253" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb53-254"><a href="#cb53-254" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-257"><a href="#cb53-257" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb53-258"><a href="#cb53-258" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(pbmc, <span class="at">reduction =</span> <span class="st">"pca"</span>) <span class="sc">+</span> <span class="fu">NoLegend</span>()</span>
<span id="cb53-259"><a href="#cb53-259" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb53-260"><a href="#cb53-260" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-261"><a href="#cb53-261" aria-hidden="true" tabindex="-1"></a>In particular <span class="in">`DimHeatmap()`</span> allows for easy exploration of the primary sources of heterogeneity in a dataset, and can be useful when trying to **decide which PCs to include for further downstream analyses**. **Both cells and features are ordered according to their PCA scores**.</span>
<span id="cb53-262"><a href="#cb53-262" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-263"><a href="#cb53-263" aria-hidden="true" tabindex="-1"></a>Setting&nbsp;<span class="in">`cells`</span>&nbsp;to a number, will plot the 'extreme' cells on both ends of the spectrum, which dramatically speeds plotting for large datasets. Though clearly a supervised analysis, we find this to be a valuable tool for exploring correlated feature sets.</span>
<span id="cb53-264"><a href="#cb53-264" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-267"><a href="#cb53-267" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb53-268"><a href="#cb53-268" aria-hidden="true" tabindex="-1"></a><span class="fu">DimHeatmap</span>(pbmc, <span class="at">dims =</span> <span class="dv">1</span>, <span class="at">cells =</span> <span class="dv">1000</span>, <span class="at">balanced =</span> <span class="cn">TRUE</span>)</span>
<span id="cb53-269"><a href="#cb53-269" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb53-270"><a href="#cb53-270" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-271"><a href="#cb53-271" aria-hidden="true" tabindex="-1"></a>通过<span class="in">`dims`</span>参数指定一个范围内的主成分，可以用来决定在后续的分析中应该包括哪些主成分。</span>
<span id="cb53-272"><a href="#cb53-272" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-275"><a href="#cb53-275" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb53-276"><a href="#cb53-276" aria-hidden="true" tabindex="-1"></a><span class="fu">DimHeatmap</span>(pbmc, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">15</span>, <span class="at">cells =</span> <span class="dv">1000</span>, <span class="at">balanced =</span> <span class="cn">TRUE</span>)</span>
<span id="cb53-277"><a href="#cb53-277" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb53-278"><a href="#cb53-278" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-279"><a href="#cb53-279" aria-hidden="true" tabindex="-1"></a><span class="fu"># 决定后续分析的主成分 {#sec-Determine_pcs_for_subsequent_analyses}</span></span>
<span id="cb53-280"><a href="#cb53-280" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-281"><a href="#cb53-281" aria-hidden="true" tabindex="-1"></a>To overcome the extensive technical noise in any single feature for scRNA-seq data, **Seurat clusters cells based on their PCA scores**, with each PC essentially representing a 'metafeature' that combines information across a correlated feature set. The top principal components therefore represent a robust compression of the dataset. However, how many components should we choose to include? 10? 20? 100?</span>
<span id="cb53-282"><a href="#cb53-282" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-283"><a href="#cb53-283" aria-hidden="true" tabindex="-1"></a>In Macosko&nbsp;*et al* <span class="co">[</span><span class="ot">@macosko2015</span><span class="co">]</span>, we implemented a resampling test inspired by the JackStraw procedure. While still available in Seurat, this is a slow and computationally expensive procedure, and we is no longer routinely used in single cell analysis.</span>
<span id="cb53-284"><a href="#cb53-284" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-285"><a href="#cb53-285" aria-hidden="true" tabindex="-1"></a>An alternative heuristic method generates an '**Elbow plot**': a ranking of principle components based on the percentage of variance explained by each one. In this example, w**e can observe an 'elbow' around PC9-10, suggesting that the majority of true signal is captured in the first 10 PCs**.</span>
<span id="cb53-286"><a href="#cb53-286" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-289"><a href="#cb53-289" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb53-290"><a href="#cb53-290" aria-hidden="true" tabindex="-1"></a><span class="fu">ElbowPlot</span>(pbmc)</span>
<span id="cb53-291"><a href="#cb53-291" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb53-292"><a href="#cb53-292" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-293"><a href="#cb53-293" aria-hidden="true" tabindex="-1"></a>::: {.callout-tip collapse="true"}</span>
<span id="cb53-294"><a href="#cb53-294" aria-hidden="true" tabindex="-1"></a><span class="fu">###### Identifying the true dimensionality of a dataset</span></span>
<span id="cb53-295"><a href="#cb53-295" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-296"><a href="#cb53-296" aria-hidden="true" tabindex="-1"></a>**Identifying the true dimensionality of a dataset can be challenging/uncertain for the user.** We therefore suggest these multiple approaches for users. The first is more supervised, exploring PCs to determine relevant sources of heterogeneity, and could be used in conjunction with GSEA for example. The second (<span class="in">`ElbowPlot`</span>). The third is a heuristic that is commonly used, and can be calculated instantly. In this example, we might have been justified in choosing anything between PC 7-12 as a cutoff.</span>
<span id="cb53-297"><a href="#cb53-297" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-298"><a href="#cb53-298" aria-hidden="true" tabindex="-1"></a>We chose 10 here, but encourage users to consider the following:</span>
<span id="cb53-299"><a href="#cb53-299" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-300"><a href="#cb53-300" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Dendritic cell and NK aficionados may recognize that genes strongly associated with PCs 12 and 13 define rare immune subsets (i.e.&nbsp;MZB1 is a marker for plasmacytoid DCs). However, these groups are so rare, they are difficult to distinguish from background noise for a dataset of this size without prior knowledge.</span>
<span id="cb53-301"><a href="#cb53-301" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-302"><a href="#cb53-302" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>We encourage users to repeat downstream analyses with a different number of PCs (10, 15, or even 50!). As you will observe, **the results often do not differ dramatically.**</span>
<span id="cb53-303"><a href="#cb53-303" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-304"><a href="#cb53-304" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>We advise users to **err on the higher side when choosing this parameter（建议选择偏高的主成分数量）**. For example, performing downstream analyses with only 5 PCs does significantly and adversely affect results.</span>
<span id="cb53-305"><a href="#cb53-305" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb53-306"><a href="#cb53-306" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-307"><a href="#cb53-307" aria-hidden="true" tabindex="-1"></a><span class="fu"># 细胞聚类（cluster） {#sec-clustering_seurat}</span></span>
<span id="cb53-308"><a href="#cb53-308" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-309"><a href="#cb53-309" aria-hidden="true" tabindex="-1"></a>Seurat applies a **graph-based clustering approach**, building upon initial strategies in Macosko&nbsp;*et al* [@macosko2015]. Importantly, the *distance metric* which drives the clustering analysis (based on previously identified PCs) remains the same. However, our approach to partitioning the cellular distance matrix into clusters has dramatically improved. Our approach was heavily inspired by recent manuscripts which applied graph-based clustering approaches to scRNA-seq data and CyTOF data [@levine2015]. Briefly, these methods embed cells in a graph structure - for example a **K-nearest neighbor (KNN) graph**, with edges drawn between cells with similar feature expression patterns, and then attempt to partition this graph into highly interconnected 'quasi-cliques' or 'communities'.</span>
<span id="cb53-310"><a href="#cb53-310" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-311"><a href="#cb53-311" aria-hidden="true" tabindex="-1"></a>As in PhenoGraph, we first construct a KNN graph based on the euclidean distance in PCA space, and refine the edge weights between any two cells based on the shared overlap in their local neighborhoods (Jaccard similarity). This step is performed using the&nbsp;**`FindNeighbors()`**&nbsp;function, and takes as input the **previously defined dimensionality of the dataset** (first 10 PCs).</span>
<span id="cb53-312"><a href="#cb53-312" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-313"><a href="#cb53-313" aria-hidden="true" tabindex="-1"></a>To cluster the cells, we next apply modularity optimization techniques such as the Louvain algorithm (default) or SLM <span class="co">[</span><span class="ot">@blondel2008</span><span class="co">]</span>, to iteratively group cells together, with the goal of optimizing the standard modularity function. The&nbsp;**`FindClusters()`**&nbsp;function implements this procedure, and contains a **resolution parameter** that sets the 'granularity' of the downstream clustering, with **increased values leading to a greater number of clusters**. We find that setting this parameter **between 0.4-1.2** typically returns good results for single-cell datasets of **around 3K cells**. Optimal resolution often increases for larger datasets.</span>
<span id="cb53-314"><a href="#cb53-314" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-317"><a href="#cb53-317" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb53-318"><a href="#cb53-318" aria-hidden="true" tabindex="-1"></a>pbmc <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(pbmc, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>)</span>
<span id="cb53-319"><a href="#cb53-319" aria-hidden="true" tabindex="-1"></a>pbmc <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(pbmc, <span class="at">resolution =</span> <span class="fl">0.5</span>)</span>
<span id="cb53-320"><a href="#cb53-320" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb53-321"><a href="#cb53-321" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-322"><a href="#cb53-322" aria-hidden="true" tabindex="-1"></a>The clusters can be found using the&nbsp;<span class="in">`Idents()`</span>&nbsp;function. 或者通过<span class="in">`seurat@active.ident`</span>获取。</span>
<span id="cb53-323"><a href="#cb53-323" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-326"><a href="#cb53-326" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb53-327"><a href="#cb53-327" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at cluster IDs of the first 5 cells</span></span>
<span id="cb53-328"><a href="#cb53-328" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">Idents</span>(pbmc), <span class="dv">5</span>)</span>
<span id="cb53-329"><a href="#cb53-329" aria-hidden="true" tabindex="-1"></a><span class="co"># 等价于：head(pbmc@active.ident, 5)</span></span>
<span id="cb53-330"><a href="#cb53-330" aria-hidden="true" tabindex="-1"></a><span class="co"># 类似于：pbmc@meta.data[["seurat_clusters"]][1:5]</span></span>
<span id="cb53-331"><a href="#cb53-331" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb53-332"><a href="#cb53-332" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-333"><a href="#cb53-333" aria-hidden="true" tabindex="-1"></a>现在的meta.data中出现了RNA_snn_res.0.5列和seurat_clusters列，分别代表了在分辨率为0.5的情况下的细胞cluster id以及目前激活的cluster id。由于目前只有一个0.5的分辨率，所以这两列是一样的。</span>
<span id="cb53-334"><a href="#cb53-334" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-337"><a href="#cb53-337" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb53-338"><a href="#cb53-338" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(pbmc<span class="sc">@</span>meta.data)</span>
<span id="cb53-339"><a href="#cb53-339" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb53-340"><a href="#cb53-340" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-341"><a href="#cb53-341" aria-hidden="true" tabindex="-1"></a><span class="fu"># 非线性降维</span></span>
<span id="cb53-342"><a href="#cb53-342" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-343"><a href="#cb53-343" aria-hidden="true" tabindex="-1"></a>Seurat offers several non-linear dimensional reduction techniques, such as **tSNE** and **UMAP**, to visualize and explore these datasets. The goal of these algorithms is to learn underlying structure in the dataset, in order to **place similar cells together** in low-dimensional space. Therefore, cells that are grouped together within graph-based clusters determined above should co-localize on these dimension reduction plots.</span>
<span id="cb53-344"><a href="#cb53-344" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-345"><a href="#cb53-345" aria-hidden="true" tabindex="-1"></a>While we and others have routinely found 2D visualization techniques like tSNE and UMAP to be valuable tools for exploring datasets, all visualization techniques have limitations, and cannot fully represent the complexity of the underlying data. In particular, these methods aim to preserve local distances in the dataset (i.e.&nbsp;ensuring that cells with very similar gene expression profiles co-localize), but often do not preserve more global relationships. **We encourage users to leverage techniques like UMAP for visualization**, but to avoid drawing biological conclusions solely on the basis of visualization techniques.</span>
<span id="cb53-346"><a href="#cb53-346" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-347"><a href="#cb53-347" aria-hidden="true" tabindex="-1"></a>::: callout-caution</span>
<span id="cb53-348"><a href="#cb53-348" aria-hidden="true" tabindex="-1"></a>需要先进行线性降维<span class="in">`RunPCA`</span>，再进行非线性降维<span class="in">`RunUMAP`</span>。因为<span class="in">`RunUMAP`</span>默认将PCA的结果作为输入（<span class="in">`reduction = "pca"`</span>）。</span>
<span id="cb53-349"><a href="#cb53-349" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb53-350"><a href="#cb53-350" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-353"><a href="#cb53-353" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb53-354"><a href="#cb53-354" aria-hidden="true" tabindex="-1"></a>pbmc <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(pbmc, <span class="at">reduction =</span> <span class="st">"pca"</span>, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>)</span>
<span id="cb53-355"><a href="#cb53-355" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb53-356"><a href="#cb53-356" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-357"><a href="#cb53-357" aria-hidden="true" tabindex="-1"></a>可以看到在Seurat对象的reductions中多了umap项：</span>
<span id="cb53-358"><a href="#cb53-358" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-359"><a href="#cb53-359" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/截屏2023-11-21%2015.14.18.png)</span></span>
<span id="cb53-360"><a href="#cb53-360" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-361"><a href="#cb53-361" aria-hidden="true" tabindex="-1"></a><span class="fu">## 非线性降维可视化</span></span>
<span id="cb53-362"><a href="#cb53-362" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-365"><a href="#cb53-365" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb53-366"><a href="#cb53-366" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(pbmc, <span class="at">reduction =</span> <span class="st">"umap"</span>)</span>
<span id="cb53-367"><a href="#cb53-367" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb53-368"><a href="#cb53-368" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-369"><a href="#cb53-369" aria-hidden="true" tabindex="-1"></a>You can save the object at this point so that it can easily be loaded back in without having to rerun the computationally intensive steps performed above, or easily shared with collaborators.</span>
<span id="cb53-370"><a href="#cb53-370" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-373"><a href="#cb53-373" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb53-374"><a href="#cb53-374" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb53-375"><a href="#cb53-375" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(pbmc, <span class="at">file =</span> <span class="st">"output/seurat_official/pbmc_tutorial.rds"</span>)</span>
<span id="cb53-376"><a href="#cb53-376" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb53-377"><a href="#cb53-377" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-378"><a href="#cb53-378" aria-hidden="true" tabindex="-1"></a><span class="fu"># 寻找细胞群间的差异表达基因（cluster biomarkers） {#sec-seurat_cluster_biomarkers}</span></span>
<span id="cb53-379"><a href="#cb53-379" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-380"><a href="#cb53-380" aria-hidden="true" tabindex="-1"></a>Seurat can help you find markers that define clusters via **differential expression (DE)**. By default, it identifies positive and negative markers of a single cluster (**specified in&nbsp;`ident.1`**), **compared to all other cells**.&nbsp;<span class="in">`FindAllMarkers()`</span>&nbsp;automates this process for all clusters, but you can also test groups of clusters vs.&nbsp;each other, or against all cells.</span>
<span id="cb53-381"><a href="#cb53-381" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-382"><a href="#cb53-382" aria-hidden="true" tabindex="-1"></a>::: callout-caution</span>
<span id="cb53-383"><a href="#cb53-383" aria-hidden="true" tabindex="-1"></a>In Seurat v5, we use the **presto package** (as described&nbsp;<span class="co">[</span><span class="ot">here</span><span class="co">](https://www.biorxiv.org/content/10.1101/653253v1)</span>&nbsp;and available for installation&nbsp;<span class="co">[</span><span class="ot">here</span><span class="co">](https://github.com/immunogenomics/presto)</span>), to dramatically improve the speed of DE analysis, particularly for large datasets.</span>
<span id="cb53-384"><a href="#cb53-384" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-387"><a href="#cb53-387" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb53-388"><a href="#cb53-388" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb53-389"><a href="#cb53-389" aria-hidden="true" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">'immunogenomics/presto'</span>)</span>
<span id="cb53-390"><a href="#cb53-390" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb53-391"><a href="#cb53-391" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-392"><a href="#cb53-392" aria-hidden="true" tabindex="-1"></a>For users who are not using presto, you can examine the documentation for this function (<span class="co">[</span><span class="ot">`?FindMarkers`</span><span class="co">](https://satijalab.org/seurat/reference/findmarkers)</span>) to explore the&nbsp;<span class="in">`min.pct`</span>&nbsp;and&nbsp;<span class="in">`logfc.threshold`</span>&nbsp;parameters, which can be increased in order to increase the speed of DE testing.</span>
<span id="cb53-393"><a href="#cb53-393" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb53-394"><a href="#cb53-394" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-395"><a href="#cb53-395" aria-hidden="true" tabindex="-1"></a>Find all markers of cluster 2：</span>
<span id="cb53-396"><a href="#cb53-396" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-399"><a href="#cb53-399" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb53-400"><a href="#cb53-400" aria-hidden="true" tabindex="-1"></a>cluster2.markers <span class="ot">&lt;-</span> <span class="fu">FindMarkers</span>(pbmc, <span class="at">ident.1 =</span> <span class="dv">2</span>)</span>
<span id="cb53-401"><a href="#cb53-401" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(cluster2.markers, <span class="at">n =</span> <span class="dv">5</span>)</span>
<span id="cb53-402"><a href="#cb53-402" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb53-403"><a href="#cb53-403" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-404"><a href="#cb53-404" aria-hidden="true" tabindex="-1"></a>Find all markers distinguishing cluster 5 from clusters 0 and 3：</span>
<span id="cb53-405"><a href="#cb53-405" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-408"><a href="#cb53-408" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb53-409"><a href="#cb53-409" aria-hidden="true" tabindex="-1"></a>cluster5.markers <span class="ot">&lt;-</span> <span class="fu">FindMarkers</span>(pbmc, <span class="at">ident.1 =</span> <span class="dv">5</span>, <span class="at">ident.2 =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">3</span>))</span>
<span id="cb53-410"><a href="#cb53-410" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(cluster5.markers, <span class="at">n =</span> <span class="dv">5</span>)</span>
<span id="cb53-411"><a href="#cb53-411" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb53-412"><a href="#cb53-412" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-413"><a href="#cb53-413" aria-hidden="true" tabindex="-1"></a>Find markers for every cluster compared to all remaining cells, report only the positive ones：</span>
<span id="cb53-414"><a href="#cb53-414" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-417"><a href="#cb53-417" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb53-418"><a href="#cb53-418" aria-hidden="true" tabindex="-1"></a>pbmc.markers <span class="ot">&lt;-</span> <span class="fu">FindAllMarkers</span>(pbmc, <span class="at">only.pos =</span> <span class="cn">TRUE</span>)</span>
<span id="cb53-419"><a href="#cb53-419" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb53-420"><a href="#cb53-420" aria-hidden="true" tabindex="-1"></a>pbmc.markers <span class="sc">%&gt;%</span></span>
<span id="cb53-421"><a href="#cb53-421" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(cluster) <span class="sc">%&gt;%</span></span>
<span id="cb53-422"><a href="#cb53-422" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(avg_log2FC <span class="sc">&gt;</span> <span class="dv">1</span>)</span>
<span id="cb53-423"><a href="#cb53-423" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb53-424"><a href="#cb53-424" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-425"><a href="#cb53-425" aria-hidden="true" tabindex="-1"></a>Seurat has several tests for differential expression which can be set with the <span class="in">`test.use`</span> parameter (see <span class="co">[</span><span class="ot">执行差异分析</span><span class="co">](/single_cell/seurat/de_vignette.qmd#sec-perform_de_analysis)</span> for details). 默认是用的"wilcox"方法。For example, the ROC test（<span class="in">`test.use = "roc"`</span>） returns the 'classification power' for any individual marker (ranging from 0 - random, to 1 - perfect)：</span>
<span id="cb53-426"><a href="#cb53-426" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-429"><a href="#cb53-429" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb53-430"><a href="#cb53-430" aria-hidden="true" tabindex="-1"></a>cluster0.markers <span class="ot">&lt;-</span> <span class="fu">FindMarkers</span>(pbmc, </span>
<span id="cb53-431"><a href="#cb53-431" aria-hidden="true" tabindex="-1"></a>                                <span class="at">ident.1 =</span> <span class="dv">0</span>, </span>
<span id="cb53-432"><a href="#cb53-432" aria-hidden="true" tabindex="-1"></a>                                <span class="at">logfc.threshold =</span> <span class="fl">0.25</span>, </span>
<span id="cb53-433"><a href="#cb53-433" aria-hidden="true" tabindex="-1"></a>                                <span class="at">test.use =</span> <span class="st">"roc"</span>, </span>
<span id="cb53-434"><a href="#cb53-434" aria-hidden="true" tabindex="-1"></a>                                <span class="at">only.pos =</span> <span class="cn">TRUE</span>)</span>
<span id="cb53-435"><a href="#cb53-435" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(cluster0.markers)</span>
<span id="cb53-436"><a href="#cb53-436" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb53-437"><a href="#cb53-437" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-438"><a href="#cb53-438" aria-hidden="true" tabindex="-1"></a><span class="fu"># 可视化marker基因的表达</span></span>
<span id="cb53-439"><a href="#cb53-439" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-440"><a href="#cb53-440" aria-hidden="true" tabindex="-1"></a>We include several tools for visualizing marker expression.&nbsp;<span class="in">`VlnPlot()`</span>&nbsp;(shows expression probability distributions across clusters), and&nbsp;<span class="in">`FeaturePlot()`</span>&nbsp;(visualizes feature expression on a tSNE or PCA plot) are our most commonly used visualizations. We also suggest exploring&nbsp;<span class="in">`RidgePlot()`</span>,&nbsp;<span class="in">`CellScatter()`</span>, and&nbsp;<span class="in">`DotPlot()`</span>&nbsp;as additional methods to view your dataset.</span>
<span id="cb53-441"><a href="#cb53-441" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-442"><a href="#cb53-442" aria-hidden="true" tabindex="-1"></a><span class="fu">## 小提琴图：</span></span>
<span id="cb53-443"><a href="#cb53-443" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-446"><a href="#cb53-446" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb53-447"><a href="#cb53-447" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(pbmc, <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"MS4A1"</span>, <span class="st">"CD79A"</span>))</span>
<span id="cb53-448"><a href="#cb53-448" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb53-449"><a href="#cb53-449" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-450"><a href="#cb53-450" aria-hidden="true" tabindex="-1"></a>You can plot raw counts as well（<span class="in">`layer = "counts"`</span>）：</span>
<span id="cb53-451"><a href="#cb53-451" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-454"><a href="#cb53-454" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb53-455"><a href="#cb53-455" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(pbmc, </span>
<span id="cb53-456"><a href="#cb53-456" aria-hidden="true" tabindex="-1"></a>        <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"NKG7"</span>, <span class="st">"PF4"</span>), </span>
<span id="cb53-457"><a href="#cb53-457" aria-hidden="true" tabindex="-1"></a>        <span class="at">layer =</span> <span class="st">"counts"</span>, <span class="co"># Layer to pull expression data from (e.g. "counts" or "data")</span></span>
<span id="cb53-458"><a href="#cb53-458" aria-hidden="true" tabindex="-1"></a>        <span class="at">log =</span> <span class="cn">TRUE</span>)</span>
<span id="cb53-459"><a href="#cb53-459" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb53-460"><a href="#cb53-460" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-461"><a href="#cb53-461" aria-hidden="true" tabindex="-1"></a><span class="fu">## UMAP图：</span></span>
<span id="cb53-462"><a href="#cb53-462" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-465"><a href="#cb53-465" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb53-466"><a href="#cb53-466" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>(pbmc,</span>
<span id="cb53-467"><a href="#cb53-467" aria-hidden="true" tabindex="-1"></a>            <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"MS4A1"</span>, <span class="st">"GNLY"</span>, <span class="st">"CD3E"</span>, <span class="st">"CD14"</span>, <span class="st">"FCER1A"</span>, <span class="st">"FCGR3A"</span>, <span class="st">"LYZ"</span>,</span>
<span id="cb53-468"><a href="#cb53-468" aria-hidden="true" tabindex="-1"></a>                         <span class="st">"PPBP"</span>, <span class="st">"CD8A"</span>))</span>
<span id="cb53-469"><a href="#cb53-469" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb53-470"><a href="#cb53-470" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-471"><a href="#cb53-471" aria-hidden="true" tabindex="-1"></a><span class="fu">## 热图：</span></span>
<span id="cb53-472"><a href="#cb53-472" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-473"><a href="#cb53-473" aria-hidden="true" tabindex="-1"></a><span class="in">`DoHeatmap()`</span>&nbsp;generates an expression heatmap for given cells and features. In this case, we are plotting the top 20 markers (or all markers if less than 20) for each cluster.</span>
<span id="cb53-474"><a href="#cb53-474" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-477"><a href="#cb53-477" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb53-478"><a href="#cb53-478" aria-hidden="true" tabindex="-1"></a>pbmc.markers <span class="sc">%&gt;%</span></span>
<span id="cb53-479"><a href="#cb53-479" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(cluster) <span class="sc">%&gt;%</span></span>
<span id="cb53-480"><a href="#cb53-480" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(avg_log2FC <span class="sc">&gt;</span> <span class="dv">1</span>) <span class="sc">%&gt;%</span></span>
<span id="cb53-481"><a href="#cb53-481" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice_head</span>(<span class="at">n =</span> <span class="dv">10</span>) <span class="sc">%&gt;%</span> <span class="co"># 选取开头的10行</span></span>
<span id="cb53-482"><a href="#cb53-482" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ungroup</span>() <span class="ot">-&gt;</span> top10</span>
<span id="cb53-483"><a href="#cb53-483" aria-hidden="true" tabindex="-1"></a><span class="fu">DoHeatmap</span>(pbmc, <span class="at">features =</span> top10<span class="sc">$</span>gene) <span class="sc">+</span> <span class="fu">NoLegend</span>()</span>
<span id="cb53-484"><a href="#cb53-484" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb53-485"><a href="#cb53-485" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-486"><a href="#cb53-486" aria-hidden="true" tabindex="-1"></a><span class="fu"># 细胞注释</span></span>
<span id="cb53-487"><a href="#cb53-487" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-488"><a href="#cb53-488" aria-hidden="true" tabindex="-1"></a>本案例使用的细胞注释：</span>
<span id="cb53-489"><a href="#cb53-489" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-490"><a href="#cb53-490" aria-hidden="true" tabindex="-1"></a>| Cluster ID | Markers       | Cell Type    |</span>
<span id="cb53-491"><a href="#cb53-491" aria-hidden="true" tabindex="-1"></a>|:-----------|:--------------|:-------------|</span>
<span id="cb53-492"><a href="#cb53-492" aria-hidden="true" tabindex="-1"></a>| 0          | IL7R, CCR7    | Naive CD4+ T |</span>
<span id="cb53-493"><a href="#cb53-493" aria-hidden="true" tabindex="-1"></a>| 1          | CD14, LYZ     | CD14+ Mono   |</span>
<span id="cb53-494"><a href="#cb53-494" aria-hidden="true" tabindex="-1"></a>| 2          | IL7R, S100A4  | Memory CD4+  |</span>
<span id="cb53-495"><a href="#cb53-495" aria-hidden="true" tabindex="-1"></a>| 3          | MS4A1         | B            |</span>
<span id="cb53-496"><a href="#cb53-496" aria-hidden="true" tabindex="-1"></a>| 4          | CD8A          | CD8+ T       |</span>
<span id="cb53-497"><a href="#cb53-497" aria-hidden="true" tabindex="-1"></a>| 5          | FCGR3A, MS4A7 | FCGR3A+ Mono |</span>
<span id="cb53-498"><a href="#cb53-498" aria-hidden="true" tabindex="-1"></a>| 6          | GNLY, NKG7    | NK           |</span>
<span id="cb53-499"><a href="#cb53-499" aria-hidden="true" tabindex="-1"></a>| 7          | FCER1A, CST3  | DC           |</span>
<span id="cb53-500"><a href="#cb53-500" aria-hidden="true" tabindex="-1"></a>| 8          | PPBP          | Platelet     |</span>
<span id="cb53-501"><a href="#cb53-501" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-502"><a href="#cb53-502" aria-hidden="true" tabindex="-1"></a>目前的<span class="in">`pbmc@active.ident`</span>（或<span class="in">`Idents(pbmc)`</span>）和<span class="in">`meta.data`</span>中的"seurat_clusters"列储存了激活的分群信息：</span>
<span id="cb53-503"><a href="#cb53-503" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-506"><a href="#cb53-506" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb53-507"><a href="#cb53-507" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(pbmc) <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span>
<span id="cb53-508"><a href="#cb53-508" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb53-509"><a href="#cb53-509" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-510"><a href="#cb53-510" aria-hidden="true" tabindex="-1"></a>通过上面表格的注释依据对细胞群重命名，将命名信息储存在"new.cluster.ids"中，然后返回到meta.data中，形成新的"new.cluster.ids"列：</span>
<span id="cb53-511"><a href="#cb53-511" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-514"><a href="#cb53-514" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb53-515"><a href="#cb53-515" aria-hidden="true" tabindex="-1"></a>new.cluster.ids <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Naive CD4 T"</span>, <span class="st">"CD14+ Mono"</span>, <span class="st">"Memory CD4 T"</span>, <span class="st">"B"</span>, <span class="st">"CD8 T"</span>, </span>
<span id="cb53-516"><a href="#cb53-516" aria-hidden="true" tabindex="-1"></a>                     <span class="st">"FCGR3A+ Mono"</span>, <span class="st">"NK"</span>, <span class="st">"DC"</span>, <span class="st">"Platelet"</span>)</span>
<span id="cb53-517"><a href="#cb53-517" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(new.cluster.ids) <span class="ot">&lt;-</span> <span class="fu">levels</span>(pbmc)</span>
<span id="cb53-518"><a href="#cb53-518" aria-hidden="true" tabindex="-1"></a>pbmc <span class="ot">&lt;-</span> <span class="fu">RenameIdents</span>(pbmc, new.cluster.ids)</span>
<span id="cb53-519"><a href="#cb53-519" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(pbmc, </span>
<span id="cb53-520"><a href="#cb53-520" aria-hidden="true" tabindex="-1"></a>        <span class="at">reduction =</span> <span class="st">"umap"</span>, </span>
<span id="cb53-521"><a href="#cb53-521" aria-hidden="true" tabindex="-1"></a>        <span class="at">label =</span> <span class="cn">TRUE</span>, </span>
<span id="cb53-522"><a href="#cb53-522" aria-hidden="true" tabindex="-1"></a>        <span class="at">pt.size =</span> <span class="fl">0.5</span>) <span class="sc">+</span> </span>
<span id="cb53-523"><a href="#cb53-523" aria-hidden="true" tabindex="-1"></a>  <span class="fu">NoLegend</span>()</span>
<span id="cb53-524"><a href="#cb53-524" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb53-525"><a href="#cb53-525" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-526"><a href="#cb53-526" aria-hidden="true" tabindex="-1"></a>目前的分群信息：</span>
<span id="cb53-527"><a href="#cb53-527" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-530"><a href="#cb53-530" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb53-531"><a href="#cb53-531" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(pbmc) <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span>
<span id="cb53-532"><a href="#cb53-532" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb53-533"><a href="#cb53-533" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-534"><a href="#cb53-534" aria-hidden="true" tabindex="-1"></a><span class="in">`DimPlot`</span>基于<span class="in">`ggplot2`</span>绘图框架，所以可以用<span class="in">`ggplot2`</span>语法对图像进行自定义调整：</span>
<span id="cb53-535"><a href="#cb53-535" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-538"><a href="#cb53-538" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb53-539"><a href="#cb53-539" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb53-540"><a href="#cb53-540" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(pbmc, </span>
<span id="cb53-541"><a href="#cb53-541" aria-hidden="true" tabindex="-1"></a>        <span class="at">reduction =</span> <span class="st">"umap"</span>, </span>
<span id="cb53-542"><a href="#cb53-542" aria-hidden="true" tabindex="-1"></a>        <span class="at">label =</span> <span class="cn">TRUE</span>, </span>
<span id="cb53-543"><a href="#cb53-543" aria-hidden="true" tabindex="-1"></a>        <span class="at">label.size =</span> <span class="fl">4.5</span>) <span class="sc">+</span> </span>
<span id="cb53-544"><a href="#cb53-544" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"UMAP 1"</span>) <span class="sc">+</span> </span>
<span id="cb53-545"><a href="#cb53-545" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"UMAP 2"</span>) <span class="sc">+</span></span>
<span id="cb53-546"><a href="#cb53-546" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.title =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">18</span>), </span>
<span id="cb53-547"><a href="#cb53-547" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">18</span>)) <span class="sc">+</span> </span>
<span id="cb53-548"><a href="#cb53-548" aria-hidden="true" tabindex="-1"></a>  <span class="fu">guides</span>(<span class="at">colour =</span> <span class="fu">guide_legend</span>(<span class="at">override.aes =</span> <span class="fu">list</span>(<span class="at">size =</span> <span class="dv">10</span>)))</span>
<span id="cb53-549"><a href="#cb53-549" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb53-550"><a href="#cb53-550" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-551"><a href="#cb53-551" aria-hidden="true" tabindex="-1"></a>保存Seurat对象：</span>
<span id="cb53-552"><a href="#cb53-552" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-555"><a href="#cb53-555" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb53-556"><a href="#cb53-556" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(pbmc, <span class="at">file =</span> <span class="st">"output/seurat_official/pbmc3k_final.rds"</span>)</span>
<span id="cb53-557"><a href="#cb53-557" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb53-558"><a href="#cb53-558" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-559"><a href="#cb53-559" aria-hidden="true" tabindex="-1"></a>------------------------------------------------------------------------</span>
<span id="cb53-560"><a href="#cb53-560" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-561"><a href="#cb53-561" aria-hidden="true" tabindex="-1"></a>::: {.callout-note collapse="true" icon="false"}</span>
<span id="cb53-562"><a href="#cb53-562" aria-hidden="true" tabindex="-1"></a><span class="fu">## Session Info</span></span>
<span id="cb53-563"><a href="#cb53-563" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-566"><a href="#cb53-566" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb53-567"><a href="#cb53-567" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span>
<span id="cb53-568"><a href="#cb53-568" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb53-569"><a href="#cb53-569" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb53-570"><a href="#cb53-570" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-571"><a href="#cb53-571" aria-hidden="true" tabindex="-1"></a><span class="fu"># References {.unnumbered}</span></span>
<span id="cb53-572"><a href="#cb53-572" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-573"><a href="#cb53-573" aria-hidden="true" tabindex="-1"></a>::: {#refs}</span>
<span id="cb53-574"><a href="#cb53-574" aria-hidden="true" tabindex="-1"></a>:::</span>
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">
      <ul class="footer-items list-unstyled">
<li class="nav-item">
 This website was built with Quarto
  </li>  
</ul>
</div>   
    <div class="nav-footer-center">
<p>Copyright 2024, Du Junhong</p>
<div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/djhcod/r-notes/blob/main/single_cell/seurat/seurat_tutorial.qmd" class="toc-action"><i class="bi bi-github"></i>View source</a></li><li><a href="https://github.com/djhcod/r-notes/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></div>
    <div class="nav-footer-right">
      <ul class="footer-items list-unstyled">
<li class="nav-item compact">
    <a class="nav-link" href="https://github.com/djhcod">
      <i class="bi bi-github" role="img">
</i> 
    </a>
  </li>  
</ul>
</div>
  </div>
</footer>


</body></html>