---
title: "scRNA-seq_online学习材料"
toc: false
number-sections: false
---

> 本节内容来自Mary Piper等编写的[*scRNA-seq_online: scRNA-seq Lessons from HCBC (first release)*](https://doi.org/10.5281/zenodo.5826256)
>
> 代码文件下载自GitHub仓库：[scRNA-seq_online](https://github.com/hbctraining/scRNA-seq_online)
>
> 在线版本：<https://hbctraining.github.io/scRNA-seq_online/>
>
> GitHub仓库更新日期：2023年12月13日

This repository has teaching materials for a hands-on *Introduction to single-cell RNA-seq analysis workshop*. This workshop will instruct participants on how to design a single-cell RNA-seq experiment, and how to efficiently manage and analyze the data starting from count matrices. This will be a hands-on workshop in which we will focus on using the Seurat package using R/RStudio. Working knowledge of R is required or completion of the [Introduction to R workshop](https://hbctraining.github.io/Intro-to-R/).

# Learning Objectives

-   Explain common considerations when designing a single-cell RNA-seq experiment
-   Discuss the steps involved in taking raw single-cell RNA-sequencing data and generating a count (gene expression) matrix
-   Compute and assess QC metrics at every step in the workflow
-   Cluster cells based on expression data and derive the identity of the different cell types present
-   Perform integration of different sample conditions

------------------------------------------------------------------------

# Resources

## 其他scRNA-seq数据分析课程:

-   [Seurat vignettes](https://satijalab.org/seurat/vignettes.html)

-   [Seurat cheatsheet](https://satijalab.org/seurat/essential_commands.html)

-   《[Analysis of single cell RNA-seq data](https://www.singlecellcourse.org/index.html)》

-   《[ANALYSIS OF SINGLE CELL RNA-SEQ DATA](https://broadinstitute.github.io/2020_scWorkshop/)》

-   [NBIS Workshop: Single Cell RNA-seq Analysis](https://nbisweden.github.io/workshop-scRNAseq/)

-   《[Single-cell RNA-seq YouTube视频合集](https://liulab-dfci.github.io/bioinfo-combio/scrna1.html)》

-   [Ligand-receptor analysis with CellphoneDB](https://www.nature.com/articles/s41576-020-00292-x)

-   [Best practices for single-cell analysis across modalities](https://www.nature.com/articles/s41576-023-00586-w)

## Resources for scRNA-seq Sample Prep:

-   [Research protocol分享平台](https://www.protocols.io)

-   [“Sampling time-dependent artifacts in single-cell genomics studies.”](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-02032-0) *Massoni-Badosa et al.*2019

-   [“Dissociation of solid tumor tissues with cold active protease for single-cell RNA-seq minimizes conserved collagenase-associated stress responses.”](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1830-0) *O’Flanagan et al.* 2020

-   [“Systematic assessment of tissue dissociation and storage biases in single-cell and single-nucleus RNA-seq workflows.”](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-02048-6) *Denisenko et al.* 2020

## Highlighted papers for single-nuclei RNA-seq:

-   [Single-nucleus and single-cell transcriptomes compared in matched cortical cell types](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6306246/)

-   [A single-cell and single-nucleus RNA-Seq toolbox for fresh and frozen human tumors](https://www.nature.com/articles/s41591-020-0844-1)
