# scRNA-seq_online学习材料

> 本节内容来自Mary Piper等编写的[*scRNA-seq_online: scRNA-seq Lessons from HCBC (first release)*](https://doi.org/10.5281/zenodo.5826256)
>
> 代码文件下载自GitHub仓库：[scRNA-seq_online](https://github.com/hbctraining/scRNA-seq_online)
>
> 在线版本：<https://hbctraining.github.io/scRNA-seq_online/>
>
> GitHub仓库更新日期：2023年12月13日

This repository has teaching materials for a hands-on *Introduction to single-cell RNA-seq analysis workshop*. This workshop will instruct participants on how to design a single-cell RNA-seq experiment, and how to efficiently manage and analyze the data starting from count matrices. This will be a hands-on workshop in which we will focus on using the Seurat package using R/RStudio. Working knowledge of R is required or completion of the [Introduction to R workshop](https://hbctraining.github.io/Intro-to-R/).

### Learning Objectives

-   Explain common considerations when designing a single-cell RNA-seq experiment
-   Discuss the steps involved in taking raw single-cell RNA-sequencing data and generating a count (gene expression) matrix
-   Compute and assess QC metrics at every step in the workflow
-   Cluster cells based on expression data and derive the identity of the different cell types present
-   Perform integration of different sample conditions
