<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>R语言学习笔记 - 30&nbsp; 多个单细胞数据集整合分析</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../quarto_foundation/quarto_foundation.html" rel="next">
<link href="../../single_cell/sc_supplementary/universal_marker.html" rel="prev">
<link href="../../images/logo.jpg" rel="icon" type="image/jpeg">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark"><script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><script type="application/json" class="js-hypothesis-config">
{
  "theme": "classic",
  "openSidebar": false,
  "showHighlights": "always",
  "enableExperimentalNewNoteButton": true
}
</script><script async="" src="https://hypothes.is/embed.js"></script>
<meta name="twitter:title" content="R语言学习笔记 - 30&nbsp; 多个单细胞数据集整合分析">
<meta name="twitter:description" content="">
<meta name="twitter:image" content="https://djhcod.github.io/r-notes/single_cell/sc_supplementary/images/截屏2024-01-10 12.00.48.png">
<meta name="twitter:image-height" content="886">
<meta name="twitter:image-width" content="1446">
<meta name="twitter:card" content="summary_large_image">
</head>
<body class="nav-sidebar docked nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="navbar navbar-expand-lg navbar-dark "><div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a href="../../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../../images/logo.jpg" alt="" class="navbar-logo"></a>
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">R语言学习笔记</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
<li class="nav-item">
    <a class="nav-link active" href="../../index.html" rel="" target="" aria-current="page">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../intro.html" rel="" target="">
 <span class="menu-text">Introduction</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-parts" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Parts</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-parts">
<li>
    <a class="dropdown-item" href="../../r_basic/r_basics.html" rel="" target="">
 <span class="dropdown-text">R语言基础</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../single_cell/seurat/seurat.html" rel="" target="">
 <span class="dropdown-text">Seurat v5官方文档</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../single_cell/scRNA-seq_online/00_intro.html" rel="" target="">
 <span class="dropdown-text">scRNA-seq_online学习材料</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../single_cell/sc_supplementary/read_sc_data.html" rel="" target="">
 <span class="dropdown-text">单细胞补充内容</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../quarto_foundation/quarto_foundation.html" rel="" target="">
 <span class="dropdown-text">Quarto基础</span></a>
  </li>  
    </ul>
</li>
  <li class="nav-item">
    <a class="nav-link" href="../../references.html" rel="" target="">
 <span class="menu-text">References</span></a>
  </li>  
</ul>
<ul class="navbar-nav navbar-nav-scroll ms-auto">
<li class="nav-item">
    <a class="nav-link" href="https://bookdown.org" rel="" target=""><i class="bi bi-book-fill" role="img">
</i> 
 <span class="menu-text">Bookdown</span></a>
  </li>  
</ul>
<div class="quarto-navbar-tools tools-wide">
    <a href="https://github.com/djhcod/r-notes" rel="" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
    <div class="dropdown">
      <a href="" title="Share" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Share"><i class="bi bi-share"></i></a>
      <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="quarto-navigation-tool-dropdown-0">
<li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://twitter.com/intent/tweet?url=%7Curl%7C">
              <i class="bi bi-bi-twitter pe-1"></i>
            Twitter
            </a>
          </li>
          <li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://www.facebook.com/sharer/sharer.php?u=%7Curl%7C">
              <i class="bi bi-bi-facebook pe-1"></i>
            Facebook
            </a>
          </li>
      </ul>
</div>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../single_cell/sc_supplementary/read_sc_data.html">单细胞补充内容</a></li><li class="breadcrumb-item"><a href="../../single_cell/sc_supplementary/integrated_analysis_multiple_single_cell_datasets.html"><span class="chapter-number">30</span>&nbsp; <span class="chapter-title">多个单细胞数据集整合分析</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto"><div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">主页</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">前言</span></a>
  </div>
</li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../r_basic/r_basics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">R语言基础</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/environment_configuration.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Rstudio环境配置</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/data_input_output.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">数据的读取与输出</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/basic_data_function.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">数据处理基本函数</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/character.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">字符的处理</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/loop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">循环</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../single_cell/seurat/seurat.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Seurat v5官方文档</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/seurat_command_list.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Seurat常用函数清单</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/seurat_tutorial.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Seurat细胞分群官方教程</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/data_visualization_methods_in_seurat.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Seurat中的数据可视化方法</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/sctransform.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">基于SCTransform的单细胞数据标准化</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/cell_cycle_regression.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">消除细胞周期的影响</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/integration.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">整合（integration）</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/integrative_analysis_in_seurat_v5.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Seurat v5单细胞数据整合分析</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/marker_gene_identification.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">寻找marker gene</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/de_vignette.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">寻找不同样本类型间同一细胞类型内的差异基因</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/mapping_and_annotating_query_datasets.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">基于参考集的细胞注释</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../single_cell/scRNA-seq_online/00_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">scRNA-seq_online学习材料</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/01_intro_to_scRNA-seq.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">单细胞测序技术介绍</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/02_SC_generation_of_count_matrix.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">单细胞RNA测序—从raw data到count matrix</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/03_SC_quality_control-setup.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">数据导入与Seurat对象构建</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/04_SC_quality_control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">质控</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/postQC_workflow.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">21</span>&nbsp; <span class="chapter-title">Single-cell RNA-seq Clustering Workflow</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/05_theory_of_PCA.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">22</span>&nbsp; <span class="chapter-title">PCA原理</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/06_SC_SCT_normalization.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">23</span>&nbsp; <span class="chapter-title">Normalization and regressing out unwanted variation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/06_integration.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">24</span>&nbsp; <span class="chapter-title">单细胞数据整合（Integration）</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/07_SC_clustering_cells_SCT.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">25</span>&nbsp; <span class="chapter-title">细胞聚类（clustering analysis）</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/08_SC_clustering_quality_control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">26</span>&nbsp; <span class="chapter-title">细胞分群质量评估</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/09_merged_SC_marker_identification.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">27</span>&nbsp; <span class="chapter-title">寻找marker基因+细胞注释</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
 <span class="menu-text">单细胞补充内容</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/sc_supplementary/read_sc_data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">28</span>&nbsp; <span class="chapter-title"><strong>读取非标准格式的单细胞数据</strong></span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/sc_supplementary/universal_marker.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">29</span>&nbsp; <span class="chapter-title">细胞分群通用marker</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/sc_supplementary/integrated_analysis_multiple_single_cell_datasets.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">30</span>&nbsp; <span class="chapter-title">多个单细胞数据集整合分析</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../quarto_foundation/quarto_foundation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Quarto基础</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/yaml_settings.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">31</span>&nbsp; <span class="chapter-title">YAML设置</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/images_settings.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">32</span>&nbsp; <span class="chapter-title">图片设置</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/cross_references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">33</span>&nbsp; <span class="chapter-title">交叉引用</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/insert_other_content.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">34</span>&nbsp; <span class="chapter-title">插入其他内容</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/quarto_shiny.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">35</span>&nbsp; <span class="chapter-title">在Quarto中嵌入Shiny应用程序</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/quarto_books.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">36</span>&nbsp; <span class="chapter-title">Quarto Books</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/github_pages.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">37</span>&nbsp; <span class="chapter-title">发布到GitHub Pages</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="false">
 <span class="menu-text">参考文献</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active" data-toc-expanded="1"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#%E5%8A%A0%E8%BD%BD%E5%8C%85" id="toc-加载包" class="nav-link active" data-scroll-target="#%E5%8A%A0%E8%BD%BD%E5%8C%85"><span class="header-section-number">30.1</span> 加载包</a></li>
  <li>
<a href="#gse150430" id="toc-gse150430" class="nav-link" data-scroll-target="#gse150430"><span class="header-section-number">30.2</span> GSE150430</a>
  <ul class="collapse">
<li><a href="#%E6%95%B0%E6%8D%AE%E5%AF%BC%E5%85%A5" id="toc-数据导入" class="nav-link" data-scroll-target="#%E6%95%B0%E6%8D%AE%E5%AF%BC%E5%85%A5">数据导入</a></li>
  <li><a href="#%E8%B4%A8%E6%8E%A7" id="toc-质控" class="nav-link" data-scroll-target="#%E8%B4%A8%E6%8E%A7">质控</a></li>
  <li><a href="#%E5%BD%92%E4%B8%80%E5%8C%96%E5%8F%8A%E6%B6%88%E9%99%A4%E9%9D%9E%E6%9C%9F%E6%9C%9B%E5%8F%98%E5%BC%82%E6%9D%A5%E6%BA%90" id="toc-归一化及消除非期望变异来源" class="nav-link" data-scroll-target="#%E5%BD%92%E4%B8%80%E5%8C%96%E5%8F%8A%E6%B6%88%E9%99%A4%E9%9D%9E%E6%9C%9F%E6%9C%9B%E5%8F%98%E5%BC%82%E6%9D%A5%E6%BA%90">归一化及消除非期望变异来源</a></li>
  <li><a href="#%E6%95%B0%E6%8D%AE%E6%95%B4%E5%90%88" id="toc-数据整合" class="nav-link" data-scroll-target="#%E6%95%B0%E6%8D%AE%E6%95%B4%E5%90%88">数据整合</a></li>
  <li><a href="#%E8%81%9A%E7%B1%BB" id="toc-聚类" class="nav-link" data-scroll-target="#%E8%81%9A%E7%B1%BB">聚类</a></li>
  <li><a href="#%E7%BB%86%E8%83%9E%E5%88%86%E7%BE%A4%E8%B4%A8%E9%87%8F%E8%AF%84%E4%BC%B0" id="toc-细胞分群质量评估" class="nav-link" data-scroll-target="#%E7%BB%86%E8%83%9E%E5%88%86%E7%BE%A4%E8%B4%A8%E9%87%8F%E8%AF%84%E4%BC%B0"><strong>细胞分群质量评估</strong></a></li>
  <li><a href="#%E7%BB%86%E8%83%9E%E6%B3%A8%E9%87%8A" id="toc-细胞注释" class="nav-link" data-scroll-target="#%E7%BB%86%E8%83%9E%E6%B3%A8%E9%87%8A">细胞注释</a></li>
  <li><a href="#%E6%8F%90%E5%8F%96%E9%AB%93%E7%B3%BB%E7%BB%86%E8%83%9E" id="toc-提取髓系细胞" class="nav-link" data-scroll-target="#%E6%8F%90%E5%8F%96%E9%AB%93%E7%B3%BB%E7%BB%86%E8%83%9E">提取髓系细胞</a></li>
  <li><a href="#%E9%AB%93%E7%B3%BB%E7%BB%86%E8%83%9E%E8%BF%9B%E4%B8%80%E6%AD%A5%E5%BD%92%E4%B8%80%E5%8C%96%E6%95%B4%E5%90%88%E9%99%8D%E7%BB%B4%E5%88%86%E7%BE%A4" id="toc-髓系细胞进一步归一化整合降维分群" class="nav-link" data-scroll-target="#%E9%AB%93%E7%B3%BB%E7%BB%86%E8%83%9E%E8%BF%9B%E4%B8%80%E6%AD%A5%E5%BD%92%E4%B8%80%E5%8C%96%E6%95%B4%E5%90%88%E9%99%8D%E7%BB%B4%E5%88%86%E7%BE%A4">髓系细胞进一步归一化、整合、降维、分群</a></li>
  </ul>
</li>
  </ul><div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/djhcod/r-notes/blob/main/single_cell/sc_supplementary/integrated_analysis_multiple_single_cell_datasets.qmd" class="toc-action">View source</a></p><p><a href="https://github.com/djhcod/r-notes/issues/new" class="toc-action">Report an issue</a></p><p><a href="https://github.com/djhcod/r-notes/edit/main/single_cell/sc_supplementary/integrated_analysis_multiple_single_cell_datasets.qmd" class="toc-action">Edit this page</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title">
<span class="chapter-number">30</span>&nbsp; <span class="chapter-title">多个单细胞数据集整合分析</span>
</h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><blockquote class="blockquote">
<p>参考：<a href="https://mp.weixin.qq.com/s/51PtUi7ZKp1nbrFGElU74A">单细胞多数据集整合示例</a></p>
</blockquote>
<p>有时候，为了扩大数据量和得到更加可靠的结论，我们可能获取和下载了多个单细胞数据集。如果我们只关注某一细胞类型，如T cells或B cells或髓系细胞，那么就需要合并分析这些数据集，提取感兴趣的细胞亚群。这时候有两种方法可以选择：</p>
<ul>
<li><p><strong>对各个数据集分别降维分群，然后提取各个数据集的感兴趣细胞亚群并整合到一起</strong></p></li>
<li><p><strong>在一开始分析的时候就对这些数据集进行整合，然后分群，提起感兴趣细胞群</strong></p></li>
</ul>
<p><strong>这个部分怎么分析并没有一个特定的答案或者规则，具体还是得看数据集适合以上提出的哪种分析顺序。</strong></p>
<p>例如<span class="citation" data-cites="gong2023">(<a href="../../references.html#ref-gong2023" role="doc-biblioref">Gong et al. 2023</a>)</span>就用了第二种方法。选取了三个鼻咽癌单细胞转录组数据： GSE150825, GSE150430, GSE162025。首先对三组数据进行整合，然后选取其中的一部分T细胞进行进一步分析。</p>
<p><img src="images/截屏2024-01-10 12.00.48.png" class="img-fluid" width="600"></p>
<p>这里我们介绍第一种方式，即<strong>对各数据集分别降维分群，然后提取并整合各个数据集的感兴趣细胞亚群。</strong></p>
<section id="加载包" class="level2" data-number="30.1"><h2 data-number="30.1" class="anchored" data-anchor-id="加载包">
<span class="header-section-number">30.1</span> 加载包</h2>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_cache/html/unnamed-chunk-1_20288cda55e52aecc5327d3b6ed297cd">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">beepr</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/">cowplot</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/lazappi/clustree">clustree</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="gse150430" class="level2" data-number="30.2"><h2 data-number="30.2" class="anchored" data-anchor-id="gse150430">
<span class="header-section-number">30.2</span> GSE150430</h2>
<p>这里我们直接载入 <a href="read_sc_data.html#sec-reading_non_standard_large_volume_text"><span>Section&nbsp;28.4</span></a> 中构建好的Seurat对象，该案例数据的介绍及读取过程参见该章节。</p>
<section id="数据导入" class="level3"><h3 class="anchored" data-anchor-id="数据导入">数据导入</h3>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_cache/html/unnamed-chunk-2_577cf27b7aa4c1dfff151fdcbc8276d7">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">merged_seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="st">"output/sc_supplementary/GSE150430_merged_seurat.rds"</span><span class="op">)</span></span>
<span><span class="va">merged_seurat</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
24720 features across 48584 samples within 1 assay 
Active assay: RNA (24720 features, 0 variable features)
 16 layers present: counts.1, counts.2, counts.3, counts.4, counts.5, counts.6, counts.7, counts.8, counts.9, counts.10, counts.11, counts.12, counts.13, counts.14, counts.15, counts.16</code></pre>
</div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">merged_seurat</span>, <span class="fl">4</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                       orig.ident nCount_RNA nFeature_RNA samples
N01_CAGAATCAGTAATCCC.1        N01   3470.970         4173     N01
N01_TGCGGGTAGTACGTAA.1        N01   3155.725         2994     N01
N01_AAACCTGAGGGTTTCT.1        N01   2347.034         1946     N01
N01_AAACCTGAGTTCCACA.1        N01   2362.495         1993     N01</code></pre>
</div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">$</span><span class="va">samples</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "N01" "P01" "P02" "P03" "P04" "P05" "P06" "P07" "P08" "P09" "P10" "P11"
[13] "P12" "P13" "P14" "P15"</code></pre>
</div>
</div>
</section><section id="质控" class="level3"><h3 class="anchored" data-anchor-id="质控">质控</h3>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_cache/html/unnamed-chunk-3_582d0968d080ffda66e06763b04af6f3">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Add number of genes per UMI for each cell to metadata</span></span>
<span><span class="va">merged_seurat</span><span class="op">$</span><span class="va">log10GenesPerUMI</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">$</span><span class="va">nFeature_RNA</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">$</span><span class="va">nCount_RNA</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">$</span><span class="va">log10GenesPerUMI</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
 0.8897  0.9488  0.9670  0.9773  0.9971  1.1429 </code></pre>
</div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Compute percent mito ratio</span></span>
<span><span class="va">merged_seurat</span><span class="op">$</span><span class="va">mitoRatio</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/PercentageFeatureSet.html">PercentageFeatureSet</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">merged_seurat</span>, </span>
<span>                                                pattern <span class="op">=</span> <span class="st">"^MT-"</span><span class="op">)</span></span>
<span><span class="va">merged_seurat</span><span class="op">$</span><span class="va">mitoRatio</span> <span class="op">&lt;-</span> <span class="va">merged_seurat</span><span class="op">$</span><span class="va">mitoRatio</span> <span class="op">/</span> <span class="fl">100</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">$</span><span class="va">mitoRatio</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     Min.   1st Qu.    Median      Mean   3rd Qu.      Max. 
0.0006859 0.0106660 0.0127552 0.0129526 0.0150420 0.0407058 </code></pre>
</div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/boxplot.html">boxplot</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">$</span><span class="va">mitoRatio</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integrated_analysis_multiple_single_cell_datasets_files/figure-html/unnamed-chunk-3-1.png" class="img-fluid" width="288"></p>
</div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Add cell IDs to metadata</span></span>
<span><span class="va">merged_seurat</span><span class="op">$</span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Visualize the number of cell counts per sample</span></span>
<span><span class="va">merged_seurat</span><span class="op">@</span><span class="va">meta.data</span>  <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">samples</span>, fill <span class="op">=</span> <span class="va">samples</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, vjust <span class="op">=</span> <span class="fl">1</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span>, face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"NCells"</span><span class="op">)</span></span>
<span><span class="co"># Visualize the number UMIs/transcripts per cell</span></span>
<span><span class="va">merged_seurat</span><span class="op">@</span><span class="va">meta.data</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">samples</span>, x <span class="op">=</span> <span class="va">nCount_RNA</span>, fill <span class="op">=</span> <span class="va">samples</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span><span class="op">(</span><span class="st">"Cell density"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span>
<span><span class="co"># Visualize the distribution of genes detected per cell via histogram</span></span>
<span><span class="va">merged_seurat</span><span class="op">@</span><span class="va">meta.data</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">samples</span>, x <span class="op">=</span> <span class="va">nFeature_RNA</span>, fill<span class="op">=</span> <span class="va">samples</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></span>
<span><span class="co"># Visualize the overall complexity of the gene expression by visualizing the genes detected per UMI (novelty score)</span></span>
<span><span class="va">merged_seurat</span><span class="op">@</span><span class="va">meta.data</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">log10GenesPerUMI</span>, color <span class="op">=</span> <span class="va">samples</span>, fill<span class="op">=</span><span class="va">samples</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span></span>
<span><span class="co"># Visualize the distribution of mitochondrial gene expression detected per cell</span></span>
<span><span class="va">merged_seurat</span><span class="op">@</span><span class="va">meta.data</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">samples</span>, x <span class="op">=</span> <span class="va">mitoRatio</span>, fill <span class="op">=</span> <span class="va">samples</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell quarto-layout-panel" data-hash="integrated_analysis_multiple_single_cell_datasets_cache/html/unnamed-chunk-4_4b3691f7f80d256dbb59662a6d2de2f8">
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="cell-output-display quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img src="integrated_analysis_multiple_single_cell_datasets_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img src="integrated_analysis_multiple_single_cell_datasets_files/figure-html/unnamed-chunk-4-2.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="cell-output-display quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img src="integrated_analysis_multiple_single_cell_datasets_files/figure-html/unnamed-chunk-4-3.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img src="integrated_analysis_multiple_single_cell_datasets_files/figure-html/unnamed-chunk-4-4.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="cell-output-display quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img src="integrated_analysis_multiple_single_cell_datasets_files/figure-html/unnamed-chunk-4-5.png" class="img-fluid" width="672"></p>
</div>
</div>
</div>
</div>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_cache/html/unnamed-chunk-5_9ae19beac1e63c30ad5a852bb811364d">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Visualize the correlation between genes detected and number of UMIs and determine whether strong presence of cells with low numbers of genes/UMIs</span></span>
<span><span class="va">merged_seurat</span><span class="op">@</span><span class="va">meta.data</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">nCount_RNA</span>, y <span class="op">=</span> <span class="va">nFeature_RNA</span>, color <span class="op">=</span> <span class="va">mitoRatio</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html">scale_colour_gradient</a></span><span class="op">(</span>low <span class="op">=</span> <span class="st">"gray90"</span>, high <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html">stat_smooth</a></span><span class="op">(</span>method <span class="op">=</span> <span class="va">lm</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">500</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">250</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">samples</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integrated_analysis_multiple_single_cell_datasets_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid" width="960"></p>
</div>
</div>
<p>可以看到该数据集已经进行了质控，各项指标均在正常范围内，因此，我们可以跳过细胞/基因过滤，直接进入下面的环节。</p>
</section><section id="归一化及消除非期望变异来源" class="level3"><h3 class="anchored" data-anchor-id="归一化及消除非期望变异来源">归一化及消除非期望变异来源</h3>
<div class="callout callout-style-default callout-caution callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Caution
</div>
</div>
<div class="callout-body-container callout-body">
<p>由于这里<code>SCTransform</code>的数据量太大，在我的16GB MacBook Pro上超出了R内存分配上限，出现了“Error: vector memory exhausted (limit reached?)”报错。因此，这里在运行下面的脚本之前通过 <a href="../../r_basic/environment_configuration.html#sec-Raising_memory_limit"><span>Section&nbsp;2.8.2</span></a> 的方法进行了处理。</p>
</div>
</div>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_cache/html/unnamed-chunk-6_c5c0b948a9b31125fd1168ef63534435">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># SCTranform</span></span>
<span><span class="va">merged_seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SCTransform.html">SCTransform</a></span><span class="op">(</span><span class="va">merged_seurat</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>; <span class="fu"><a href="https://rdrr.io/pkg/beepr/man/beep.html">beep</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/gc.html">gc</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># 释放未使用内存</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>            used   (Mb) gc trigger    (Mb) limit (Mb)   max used    (Mb)
Ncells   8269831  441.7   12102744   646.4         NA   12102744   646.4
Vcells 988419958 7541.1 3232715316 24663.7     102400 3232715179 24663.7</code></pre>
</div>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">merged_seurat</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
46357 features across 48584 samples within 2 assays 
Active assay: SCT (21637 features, 3000 variable features)
 3 layers present: counts, data, scale.data
 1 other assay present: RNA</code></pre>
</div>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Check which assays are stored in objects</span></span>
<span><span class="va">merged_seurat</span><span class="op">@</span><span class="va">assays</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$RNA
Assay (v5) data with 24720 features for 48584 cells
First 10 features:
 RP11-34P13.7, RP11-34P13.8, AL627309.1, AP006222.2, RP4-669L17.10,
RP4-669L17.2, RP5-857K21.4, RP11-206L10.3, RP11-206L10.5, RP11-206L10.2 
Layers:
 counts.1, counts.2, counts.3, counts.4, counts.5, counts.6, counts.7,
counts.8, counts.9, counts.10, counts.11, counts.12, counts.13,
counts.14, counts.15, counts.16 

$SCT
SCTAssay data with 21637 features for 48584 cells, and 16 SCTModel(s) 
Top 10 variable features:
 RNASE6, PHACTR1, NR4A3, C1orf162, HMOX1, IFIT2, CD7, CD86, CLNK, CRIP1 </code></pre>
</div>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 查看目前默认的assay</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/DefaultAssay.html">DefaultAssay</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "SCT"</code></pre>
</div>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 查看默认assay的layers</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/Layers.html">Layers</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "counts"     "data"       "scale.data"</code></pre>
</div>
</div>
<section id="评估细胞周期的影响" class="level4"><h4 class="anchored" data-anchor-id="评估细胞周期的影响">评估细胞周期的影响</h4>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_cache/html/unnamed-chunk-7_3f5c218f807e2d572d91db2d92b80bbd">
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Load cell cycle markers</span></span>
<span><span class="va">s.genes</span> <span class="op">&lt;-</span> <span class="va">cc.genes</span><span class="op">$</span><span class="va">s.genes</span></span>
<span><span class="va">g2m.genes</span> <span class="op">&lt;-</span> <span class="va">cc.genes</span><span class="op">$</span><span class="va">g2m.genes</span></span>
<span></span>
<span><span class="co"># Score cells for cell cycle</span></span>
<span><span class="va">merged_seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/CellCycleScoring.html">CellCycleScoring</a></span><span class="op">(</span><span class="va">merged_seurat</span>, </span>
<span>                                  g2m.features <span class="op">=</span> <span class="va">g2m.genes</span>, </span>
<span>                                  s.features <span class="op">=</span> <span class="va">s.genes</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 现在的meta.data中多出了细胞周期评分“S.Score”和“G2M.Score”，以及推断的细胞所处的周期“Phase”</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                       orig.ident nCount_RNA nFeature_RNA samples
N01_CAGAATCAGTAATCCC.1        N01   3470.970         4173     N01
N01_TGCGGGTAGTACGTAA.1        N01   3155.725         2994     N01
N01_AAACCTGAGGGTTTCT.1        N01   2347.034         1946     N01
N01_AAACCTGAGTTCCACA.1        N01   2362.495         1993     N01
N01_AAACCTGGTTCTCATT.1        N01   2130.465         1459     N01
N01_AAACGGGGTCCTCCAT.1        N01   2772.010         2573     N01
                       log10GenesPerUMI   mitoRatio                  cells
N01_CAGAATCAGTAATCCC.1        1.0225953 0.008644269 N01_CAGAATCAGTAATCCC.1
N01_TGCGGGTAGTACGTAA.1        0.9934705 0.009895666 N01_TGCGGGTAGTACGTAA.1
N01_AAACCTGAGGGTTTCT.1        0.9758564 0.015183419 N01_AAACCTGAGGGTTTCT.1
N01_AAACCTGAGTTCCACA.1        0.9781039 0.014965958 N01_AAACCTGAGTTCCACA.1
N01_AAACCTGGTTCTCATT.1        0.9506023 0.017437508 N01_AAACCTGGTTCTCATT.1
N01_AAACGGGGTCCTCCAT.1        0.9906021 0.011233365 N01_AAACGGGGTCCTCCAT.1
                       nCount_SCT nFeature_SCT     S.Score    G2M.Score Phase
N01_CAGAATCAGTAATCCC.1       1840         1674 -0.05179250 -0.070444626    G1
N01_TGCGGGTAGTACGTAA.1       2840         2604 -0.04341940 -0.074302336    G1
N01_AAACCTGAGGGTTTCT.1       2443         1931 -0.02396942 -0.007239586    G1
N01_AAACCTGAGTTCCACA.1       2488         1972 -0.01294260 -0.021167805    G1
N01_AAACCTGGTTCTCATT.1       2104         1451 -0.05440456 -0.049531592    G1
N01_AAACGGGGTCCTCCAT.1       2892         2520 -0.10189203 -0.087537749    G1</code></pre>
</div>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 查看一下细胞周期的分布情况</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">$</span><span class="va">Phase</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
   G1   G2M     S 
33941  5147  9496 </code></pre>
</div>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 执行PCA</span></span>
<span><span class="va">merged_seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html">RunPCA</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot the PCA colored by cell cycle phase</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">merged_seurat</span>,</span>
<span>              reduction <span class="op">=</span> <span class="st">"pca"</span>,</span>
<span>              group.by<span class="op">=</span> <span class="st">"Phase"</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">merged_seurat</span>,</span>
<span>              reduction <span class="op">=</span> <span class="st">"pca"</span>,</span>
<span>              group.by<span class="op">=</span> <span class="st">"Phase"</span>,</span>
<span>              split.by <span class="op">=</span> <span class="st">"Phase"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">plot_grid</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span>, ncol <span class="op">=</span> <span class="fl">2</span>, labels <span class="op">=</span> <span class="st">"AUTO"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integrated_analysis_multiple_single_cell_datasets_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid" width="960"></p>
</div>
</div>
<p>可以看到细胞周期不是变异来源。</p>
</section><section id="评估线粒体基因的影响" class="level4"><h4 class="anchored" data-anchor-id="评估线粒体基因的影响"><strong>评估线粒体基因的影响</strong></h4>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_cache/html/unnamed-chunk-8_456c177dad0c7674e56e1135d8ec7612">
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Check quartile values</span></span>
<span><span class="va">mito_sum</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">$</span><span class="va">mitoRatio</span><span class="op">)</span></span>
<span><span class="va">mito_sum</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     Min.   1st Qu.    Median      Mean   3rd Qu.      Max. 
0.0006859 0.0106660 0.0127552 0.0129526 0.0150420 0.0407058 </code></pre>
</div>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Turn mitoRatio into categorical factor vector based on quartile values</span></span>
<span><span class="va">merged_seurat</span><span class="op">$</span><span class="va">mitoFr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cut.html">cut</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">$</span><span class="va">mitoRatio</span>, </span>
<span>                            breaks<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="cn">Inf</span>, <span class="va">mito_sum</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, <span class="va">mito_sum</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span>, <span class="va">mito_sum</span><span class="op">[</span><span class="fl">5</span><span class="op">]</span>, <span class="cn">Inf</span><span class="op">)</span>,</span>
<span>                            labels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Low"</span>, <span class="st">"Medium"</span>, <span class="st">"Medium high"</span>, <span class="st">"High"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">$</span><span class="va">mitoFr</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integrated_analysis_multiple_single_cell_datasets_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid" width="288"></p>
</div>
</div>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_cache/html/unnamed-chunk-9_4d84e123eb5af04a80d070d8959071cc">
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Plot the PCA colored by cell cycle phase</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">merged_seurat</span>,</span>
<span>              reduction <span class="op">=</span> <span class="st">"pca"</span>,</span>
<span>              group.by<span class="op">=</span> <span class="st">"mitoFr"</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">merged_seurat</span>,</span>
<span>              reduction <span class="op">=</span> <span class="st">"pca"</span>,</span>
<span>              group.by<span class="op">=</span> <span class="st">"mitoFr"</span>,</span>
<span>              split.by <span class="op">=</span> <span class="st">"mitoFr"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">plot_grid</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span>, ncol <span class="op">=</span> <span class="fl">2</span>, labels <span class="op">=</span> <span class="st">"AUTO"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integrated_analysis_multiple_single_cell_datasets_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid" width="1440"></p>
</div>
</div>
<p>可以看到线粒体基因比例不是变异来源。</p>
<p>由于细胞周期和线粒体基因比例都不是非期望变异来源，所以我们这里不需要再次运行SCTransform来回归这些变量。接下来，直接进入数据整合环节。</p>
</section></section><section id="数据整合" class="level3"><h3 class="anchored" data-anchor-id="数据整合">数据整合</h3>
<section id="不进行整合时检验细胞分群情况" class="level4"><h4 class="anchored" data-anchor-id="不进行整合时检验细胞分群情况"><strong>不进行整合时检验细胞分群情况</strong></h4>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_cache/html/unnamed-chunk-10_e9de6ba7eb1e7293f0d3f51a6ac21307">
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 查看降维信息</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">@</span><span class="va">reductions</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "pca"</code></pre>
</div>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Run UMAP</span></span>
<span><span class="va">merged_seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html">RunUMAP</a></span><span class="op">(</span><span class="va">merged_seurat</span>, </span>
<span>                         dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">40</span>, </span>
<span>                         reduction <span class="op">=</span> <span class="st">"pca"</span>,</span>
<span>                         reduction.name <span class="op">=</span> <span class="st">"umap.unintegrated"</span><span class="op">)</span>; <span class="fu"><a href="https://rdrr.io/pkg/beepr/man/beep.html">beep</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 分群</span></span>
<span><span class="co"># Determine the K-nearest neighbor graph</span></span>
<span><span class="va">merged_seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html">FindNeighbors</a></span><span class="op">(</span><span class="va">merged_seurat</span>, </span>
<span>                               dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">40</span>, </span>
<span>                               reduction <span class="op">=</span> <span class="st">"pca"</span><span class="op">)</span></span>
<span><span class="va">merged_seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html">FindClusters</a></span><span class="op">(</span><span class="va">merged_seurat</span>, </span>
<span>                              cluster.name <span class="op">=</span> <span class="st">"unintegrated_clusters"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 48584
Number of edges: 1937300

Running Louvain algorithm...
Maximum modularity in 10 random starts: 0.9298
Number of communities: 38
Elapsed time: 6 seconds</code></pre>
</div>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Plot UMAP</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">merged_seurat</span>, </span>
<span>              reduction <span class="op">=</span> <span class="st">"umap.unintegrated"</span>,</span>
<span>              group.by <span class="op">=</span> <span class="st">"samples"</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">merged_seurat</span>, </span>
<span>              reduction <span class="op">=</span> <span class="st">"umap.unintegrated"</span>,</span>
<span>              split.by <span class="op">=</span> <span class="st">"samples"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">plot_grid</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span>, </span>
<span>          ncol <span class="op">=</span> <span class="fl">1</span>, labels <span class="op">=</span> <span class="st">"AUTO"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integrated_analysis_multiple_single_cell_datasets_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid" width="1440"></p>
</div>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 由于样本数较多，我们再按照样本类型“Normal” vs. “Cancer”画一下UMAP图</span></span>
<span><span class="va">merged_seurat</span><span class="op">$</span><span class="va">groups</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">$</span><span class="va">samples</span> <span class="op">==</span> <span class="st">"N01"</span>, <span class="st">"Normal"</span>, <span class="st">"Cancer"</span><span class="op">)</span></span>
<span><span class="va">p1a</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">merged_seurat</span>, </span>
<span>               reduction <span class="op">=</span> <span class="st">"umap.unintegrated"</span>,</span>
<span>               group.by <span class="op">=</span> <span class="st">"groups"</span><span class="op">)</span></span>
<span><span class="va">p2a</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">merged_seurat</span>, </span>
<span>               reduction <span class="op">=</span> <span class="st">"umap.unintegrated"</span>,</span>
<span>               split.by <span class="op">=</span> <span class="st">"groups"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">plot_grid</a></span><span class="op">(</span><span class="va">p1a</span>, <span class="va">p2a</span>, </span>
<span>          ncol <span class="op">=</span> <span class="fl">1</span>, labels <span class="op">=</span> <span class="st">"AUTO"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integrated_analysis_multiple_single_cell_datasets_files/figure-html/unnamed-chunk-10-2.png" class="img-fluid" width="1440"></p>
</div>
</div>
</section><section id="整合" class="level4"><h4 class="anchored" data-anchor-id="整合">整合</h4>
<p>这里我们用Harmony整合算法。</p>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_cache/html/unnamed-chunk-11_373e45dbd3b62419634df8f3a76ded11">
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 整合，比较耗时间，进度条会一直显示0%直至运算完成</span></span>
<span><span class="va">seurat_integrated</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/IntegrateLayers.html">IntegrateLayers</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">merged_seurat</span>,</span>
<span>                                     method <span class="op">=</span> <span class="va">HarmonyIntegration</span>,</span>
<span>                                     assay <span class="op">=</span> <span class="st">"SCT"</span>, <span class="co"># Integrating SCTransformed data</span></span>
<span>                                     orig.reduction <span class="op">=</span> <span class="st">"pca"</span>,</span>
<span>                                     verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>; <span class="fu"><a href="https://rdrr.io/pkg/beepr/man/beep.html">beep</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># 整合后合并RNA layer</span></span>
<span><span class="va">seurat_integrated</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/SplitLayers.html">JoinLayers</a></span><span class="op">(</span><span class="va">seurat_integrated</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 查看整合后的降维信息</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">seurat_integrated</span><span class="op">@</span><span class="va">reductions</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "pca"               "umap.unintegrated" "harmony"          </code></pre>
</div>
</div>
</section><section id="整合后检验细胞分群情况" class="level4"><h4 class="anchored" data-anchor-id="整合后检验细胞分群情况"><strong>整合后检验细胞分群情况</strong></h4>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_cache/html/unnamed-chunk-12_92af96521e881042344f53b00ed1723c">
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">123456</span><span class="op">)</span></span>
<span><span class="co"># Run UMAP</span></span>
<span><span class="va">seurat_integrated</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html">RunUMAP</a></span><span class="op">(</span><span class="va">seurat_integrated</span>, </span>
<span>                             dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">40</span>,</span>
<span>                             reduction <span class="op">=</span> <span class="st">"harmony"</span>, <span class="co"># 更改降维来源为整合后的"harmony"</span></span>
<span>                             reduction.name <span class="op">=</span> <span class="st">"umap.integrated"</span><span class="op">)</span>; <span class="fu"><a href="https://rdrr.io/pkg/beepr/man/beep.html">beep</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">seurat_integrated</span><span class="op">@</span><span class="va">reductions</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "pca"               "umap.unintegrated" "harmony"          
[4] "umap.integrated"  </code></pre>
</div>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 分群</span></span>
<span><span class="va">seurat_integrated</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html">FindNeighbors</a></span><span class="op">(</span><span class="va">seurat_integrated</span>, </span>
<span>                                   dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">40</span>, </span>
<span>                                   reduction <span class="op">=</span> <span class="st">"harmony"</span><span class="op">)</span> <span class="co">#更改降维来源为"harmony"</span></span>
<span><span class="va">seurat_integrated</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html">FindClusters</a></span><span class="op">(</span><span class="va">seurat_integrated</span>, </span>
<span>                                  cluster.name <span class="op">=</span> <span class="st">"integrated_clusters"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 48584
Number of edges: 2161498

Running Louvain algorithm...
Maximum modularity in 10 random starts: 0.9121
Number of communities: 24
Elapsed time: 8 seconds</code></pre>
</div>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">seurat_integrated</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "orig.ident"            "nCount_RNA"            "nFeature_RNA"         
 [4] "samples"               "log10GenesPerUMI"      "mitoRatio"            
 [7] "cells"                 "nCount_SCT"            "nFeature_SCT"         
[10] "S.Score"               "G2M.Score"             "Phase"                
[13] "mitoFr"                "unintegrated_clusters" "seurat_clusters"      
[16] "groups"                "integrated_clusters"  </code></pre>
</div>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Plot UMAP                             </span></span>
<span><span class="va">p3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">seurat_integrated</span>, </span>
<span>              reduction <span class="op">=</span> <span class="st">"umap.integrated"</span>, </span>
<span>              group.by <span class="op">=</span> <span class="st">"samples"</span><span class="op">)</span></span>
<span><span class="va">p4</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">seurat_integrated</span>, </span>
<span>              reduction <span class="op">=</span> <span class="st">"umap.integrated"</span>, </span>
<span>              split.by <span class="op">=</span> <span class="st">"samples"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">plot_grid</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p3</span>, <span class="va">p2</span>, <span class="va">p4</span>, </span>
<span>          ncol <span class="op">=</span> <span class="fl">1</span>, </span>
<span>          labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Before Harmony"</span>, <span class="st">"After Harmony"</span>, </span>
<span>                     <span class="st">"Before Harmony"</span>, <span class="st">"After Harmony"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integrated_analysis_multiple_single_cell_datasets_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid" width="960"></p>
</div>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 保存</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">seurat_integrated</span>, </span>
<span>        file <span class="op">=</span> <span class="st">"output/sc_supplementary/GSE150430_seurat_integrated.RDS"</span><span class="op">)</span>; <span class="fu"><a href="https://rdrr.io/pkg/beepr/man/beep.html">beep</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr></section></section><section id="聚类" class="level3"><h3 class="anchored" data-anchor-id="聚类">聚类</h3>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_cache/html/unnamed-chunk-13_d48844b64a2118231b98164a96f5eae5">
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/rm.html">rm</a></span><span class="op">(</span>list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ls.html">ls</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/gc.html">gc</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>             used   (Mb) gc trigger    (Mb) limit (Mb)   max used    (Mb)
Ncells    8423688  449.9   15406287   822.8         NA   12102744   646.4
Vcells 1189109624 9072.2 3232715316 24663.7     102400 3232715179 24663.7</code></pre>
</div>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">seurat_integrated</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="st">"output/sc_supplementary/GSE150430_seurat_integrated.RDS"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_cache/html/unnamed-chunk-14_e79379c1f5294380b04e40f21f239b90">
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Determine the clusters for various resolutions                                </span></span>
<span><span class="va">seurat_integrated</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html">FindClusters</a></span><span class="op">(</span><span class="va">seurat_integrated</span>,</span>
<span>                                  resolution <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.01</span>, <span class="fl">0.05</span>, <span class="fl">0.1</span>, <span class="fl">0.2</span>, <span class="fl">0.3</span>, <span class="fl">0.4</span>, <span class="fl">0.5</span>, <span class="fl">0.8</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>                                  verbose <span class="op">=</span> <span class="cn">F</span><span class="op">)</span>; <span class="fu"><a href="https://rdrr.io/pkg/beepr/man/beep.html">beep</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co"># Explore resolutions</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">seurat_integrated</span><span class="op">@</span><span class="va">meta.data</span>, <span class="fl">5</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                       orig.ident nCount_RNA nFeature_RNA samples
N01_CAGAATCAGTAATCCC.1        N01   3470.970         4173     N01
N01_TGCGGGTAGTACGTAA.1        N01   3155.725         2994     N01
N01_AAACCTGAGGGTTTCT.1        N01   2347.034         1946     N01
N01_AAACCTGAGTTCCACA.1        N01   2362.495         1993     N01
N01_AAACCTGGTTCTCATT.1        N01   2130.465         1459     N01
                       log10GenesPerUMI   mitoRatio                  cells
N01_CAGAATCAGTAATCCC.1        1.0225953 0.008644269 N01_CAGAATCAGTAATCCC.1
N01_TGCGGGTAGTACGTAA.1        0.9934705 0.009895666 N01_TGCGGGTAGTACGTAA.1
N01_AAACCTGAGGGTTTCT.1        0.9758564 0.015183419 N01_AAACCTGAGGGTTTCT.1
N01_AAACCTGAGTTCCACA.1        0.9781039 0.014965958 N01_AAACCTGAGTTCCACA.1
N01_AAACCTGGTTCTCATT.1        0.9506023 0.017437508 N01_AAACCTGGTTCTCATT.1
                       nCount_SCT nFeature_SCT     S.Score    G2M.Score Phase
N01_CAGAATCAGTAATCCC.1       1840         1674 -0.05179250 -0.070444626    G1
N01_TGCGGGTAGTACGTAA.1       2840         2604 -0.04341940 -0.074302336    G1
N01_AAACCTGAGGGTTTCT.1       2443         1931 -0.02396942 -0.007239586    G1
N01_AAACCTGAGTTCCACA.1       2488         1972 -0.01294260 -0.021167805    G1
N01_AAACCTGGTTCTCATT.1       2104         1451 -0.05440456 -0.049531592    G1
                            mitoFr unintegrated_clusters seurat_clusters groups
N01_CAGAATCAGTAATCCC.1         Low                    20              20 Normal
N01_TGCGGGTAGTACGTAA.1         Low                    32              20 Normal
N01_AAACCTGAGGGTTTCT.1        High                     0               0 Normal
N01_AAACCTGAGTTCCACA.1 Medium high                     1               3 Normal
N01_AAACCTGGTTCTCATT.1        High                     1               8 Normal
                       integrated_clusters SCT_snn_res.0.01 SCT_snn_res.0.05
N01_CAGAATCAGTAATCCC.1                  18                2                2
N01_TGCGGGTAGTACGTAA.1                  18                2                2
N01_AAACCTGAGGGTTTCT.1                   0                1                1
N01_AAACCTGAGTTCCACA.1                   0                1                1
N01_AAACCTGGTTCTCATT.1                   7                1                1
                       SCT_snn_res.0.1 SCT_snn_res.0.2 SCT_snn_res.0.3
N01_CAGAATCAGTAATCCC.1               2               2               1
N01_TGCGGGTAGTACGTAA.1               2               2               1
N01_AAACCTGAGGGTTTCT.1               0               0               0
N01_AAACCTGAGTTCCACA.1               0               0               0
N01_AAACCTGGTTCTCATT.1               0               0               0
                       SCT_snn_res.0.4 SCT_snn_res.0.5 SCT_snn_res.0.8
N01_CAGAATCAGTAATCCC.1               1              15              18
N01_TGCGGGTAGTACGTAA.1               1              15              18
N01_AAACCTGAGGGTTTCT.1               0               0               0
N01_AAACCTGAGTTCCACA.1               0               0               0
N01_AAACCTGGTTCTCATT.1               0               8               7
                       SCT_snn_res.1
N01_CAGAATCAGTAATCCC.1            20
N01_TGCGGGTAGTACGTAA.1            20
N01_AAACCTGAGGGTTTCT.1             0
N01_AAACCTGAGTTCCACA.1             3
N01_AAACCTGGTTCTCATT.1             8</code></pre>
</div>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 查看各个分辨率下的细胞分群情况</span></span>
<span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">seurat_integrated</span><span class="op">@</span><span class="va">meta.data</span>, <span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html">starts_with</a></span><span class="op">(</span>match <span class="op">=</span> <span class="st">"SCT_snn_res."</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>  </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">levels</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$SCT_snn_res.0.01
[1] "0" "1" "2" "3" "4" "5"

$SCT_snn_res.0.05
 [1] "0"  "1"  "2"  "3"  "4"  "5"  "6"  "7"  "8"  "9"  "10"

$SCT_snn_res.0.1
 [1] "0"  "1"  "2"  "3"  "4"  "5"  "6"  "7"  "8"  "9"  "10" "11" "12"

$SCT_snn_res.0.2
 [1] "0"  "1"  "2"  "3"  "4"  "5"  "6"  "7"  "8"  "9"  "10" "11" "12"

$SCT_snn_res.0.3
 [1] "0"  "1"  "2"  "3"  "4"  "5"  "6"  "7"  "8"  "9"  "10" "11" "12" "13" "14"
[16] "15" "16"

$SCT_snn_res.0.4
 [1] "0"  "1"  "2"  "3"  "4"  "5"  "6"  "7"  "8"  "9"  "10" "11" "12" "13" "14"
[16] "15" "16" "17"

$SCT_snn_res.0.5
 [1] "0"  "1"  "2"  "3"  "4"  "5"  "6"  "7"  "8"  "9"  "10" "11" "12" "13" "14"
[16] "15" "16" "17" "18" "19" "20"

$SCT_snn_res.0.8
 [1] "0"  "1"  "2"  "3"  "4"  "5"  "6"  "7"  "8"  "9"  "10" "11" "12" "13" "14"
[16] "15" "16" "17" "18" "19" "20" "21" "22" "23"

$SCT_snn_res.1
 [1] "0"  "1"  "2"  "3"  "4"  "5"  "6"  "7"  "8"  "9"  "10" "11" "12" "13" "14"
[16] "15" "16" "17" "18" "19" "20" "21" "22" "23" "24" "25"</code></pre>
</div>
</div>
<p>绘制聚类树展示不同分辨率下的细胞分群情况及相互关系</p>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_cache/html/unnamed-chunk-15_d54bdde33a5360e80e5fed78c9bad763">
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://lazappi.github.io/clustree/reference/clustree.html">clustree</a></span><span class="op">(</span><span class="va">seurat_integrated</span><span class="op">@</span><span class="va">meta.data</span>, </span>
<span>                 prefix <span class="op">=</span> <span class="st">"SCT_snn_res."</span><span class="op">)</span><span class="co">#指定包含聚类信息的列</span></span>
<span><span class="va">tree</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integrated_analysis_multiple_single_cell_datasets_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid" width="960"></p>
</div>
</div>
<p>通常情况下，为了决定合适的聚类分辨率，可使用以下两种策略：</p>
<ol type="1">
<li><p>选择透明箭头出现较少的分辨率聚类结果</p></li>
<li><p>基于marker基因表达选择有生物学意义的分辨率聚类结果</p></li>
</ol>
<p>接下来分析，按照分辨率为0.5进行</p>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_cache/html/unnamed-chunk-16_1d81d6287fe004fb97ad187577cd6274">
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/Idents.html">Idents</a></span><span class="op">(</span><span class="va">seurat_integrated</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"SCT_snn_res.0.5"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>聚类可视化</p>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_cache/html/unnamed-chunk-17_b522c09b9e751372f34629db9e9900c2">
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Plot the UMAP</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">seurat_integrated</span>,</span>
<span>        reduction <span class="op">=</span> <span class="st">"umap.integrated"</span>,</span>
<span>        label <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integrated_analysis_multiple_single_cell_datasets_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section><section id="细胞分群质量评估" class="level3"><h3 class="anchored" data-anchor-id="细胞分群质量评估"><strong>细胞分群质量评估</strong></h3>
<section id="分析样本类型是否影响细胞分群" class="level4"><h4 class="anchored" data-anchor-id="分析样本类型是否影响细胞分群"><strong>分析样本类型是否影响细胞分群</strong></h4>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_cache/html/unnamed-chunk-18_18a174e3f9fd108c4ca0747d58c7254c">
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 先简单查看不同cluster的细胞数</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">seurat_integrated</span><span class="op">@</span><span class="va">active.ident</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    0     1     2     3     4     5     6     7     8     9    10    11    12 
10830  5756  5526  5321  4161  3765  2388  2365  1952  1478   986   792   759 
   13    14    15    16    17    18    19    20 
  707   398   388   232   214   198   189   179 </code></pre>
</div>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 查看不同样本类型中的细胞分群情况</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">seurat_integrated</span>, </span>
<span>        reduction <span class="op">=</span> <span class="st">"umap.integrated"</span>,</span>
<span>        label <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>        split.by <span class="op">=</span> <span class="st">"groups"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html">NoLegend</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integrated_analysis_multiple_single_cell_datasets_files/figure-html/unnamed-chunk-18-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section><section id="分析细胞周期是否影响细胞分群" class="level4"><h4 class="anchored" data-anchor-id="分析细胞周期是否影响细胞分群"><strong>分析细胞周期是否影响细胞分群</strong></h4>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_cache/html/unnamed-chunk-19_724472838921f7f6c7ac7b6cafc3f5a3">
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Explore whether clusters segregate by cell cycle phase</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">seurat_integrated</span>,</span>
<span>        reduction <span class="op">=</span> <span class="st">"umap.integrated"</span>,</span>
<span>        label <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>        split.by <span class="op">=</span> <span class="st">"Phase"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html">NoLegend</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integrated_analysis_multiple_single_cell_datasets_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section><section id="分析其他非期望变异来源是否会影响细胞分群" class="level4"><h4 class="anchored" data-anchor-id="分析其他非期望变异来源是否会影响细胞分群"><strong>分析其他非期望变异来源是否会影响细胞分群</strong></h4>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_cache/html/unnamed-chunk-20_7e6853523adcbb1e3fc71acfcd7ba834">
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Determine metrics to plot present in seurat_clustered@meta.data</span></span>
<span><span class="va">metrics</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"nCount_RNA"</span>, <span class="st">"nFeature_RNA"</span>, <span class="st">"S.Score"</span>, <span class="st">"G2M.Score"</span>, <span class="st">"mitoRatio"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html">FeaturePlot</a></span><span class="op">(</span><span class="va">seurat_integrated</span>, </span>
<span>            reduction <span class="op">=</span> <span class="st">"umap.integrated"</span>, </span>
<span>            features <span class="op">=</span> <span class="va">metrics</span>,</span>
<span>            pt.size <span class="op">=</span> <span class="fl">0.4</span>, </span>
<span>            order <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            min.cutoff <span class="op">=</span> <span class="st">'q10'</span>, </span>
<span>            label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integrated_analysis_multiple_single_cell_datasets_files/figure-html/unnamed-chunk-20-1.png" class="img-fluid" width="960"></p>
</div>
</div>
<p>保存</p>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_cache/html/unnamed-chunk-21_ca553d498cc46e450933d59d127c9b3c">
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">seurat_integrated</span>, </span>
<span>        file <span class="op">=</span> <span class="st">"output/sc_supplementary/GSE150430_seurat_clustered.RDS"</span><span class="op">)</span>; <span class="fu"><a href="https://rdrr.io/pkg/beepr/man/beep.html">beep</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr></section></section><section id="细胞注释" class="level3"><h3 class="anchored" data-anchor-id="细胞注释">细胞注释</h3>
<section id="探索cell-type-markers的表达" class="level4"><h4 class="anchored" data-anchor-id="探索cell-type-markers的表达"><strong>探索cell type markers的表达</strong></h4>
<p>这里的marker基因选用 <a href="universal_marker.html#sec-Universal_Marker_for_initial_Cell_clustering"><span>Section&nbsp;29.3</span></a> 中的细胞初步分群通用marker。</p>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_cache/html/unnamed-chunk-22_f4bed17b3027c83fd7b8510116c2b126">
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/rm.html">rm</a></span><span class="op">(</span>list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ls.html">ls</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/gc.html">gc</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>             used   (Mb) gc trigger    (Mb) limit (Mb)   max used    (Mb)
Ncells    8509717  454.5   15406287   822.8         NA   15406287   822.8
Vcells 1006106395 7676.0 3232715316 24663.7     102400 3232715179 24663.7</code></pre>
</div>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">seurat_clustered</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="st">"output/sc_supplementary/GSE150430_seurat_clustered.RDS"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">genes_to_check</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'PTPRC'</span>, </span>
<span>                   <span class="st">"CD163"</span>,<span class="st">"AIF1"</span>, </span>
<span>                   <span class="st">'CD3D'</span>, <span class="st">'CD3E'</span>, <span class="st">'CD4'</span>, <span class="st">'CD8A'</span>, </span>
<span>                   <span class="st">'CD19'</span>, <span class="st">'CD79A'</span>, <span class="st">'MS4A1'</span>, <span class="st">"SDC1"</span>, <span class="st">"CD27"</span>, <span class="st">"CD38"</span>,</span>
<span>                   <span class="st">'IGHG1'</span>, <span class="st">'MZB1'</span>, <span class="st">'SDC1'</span>, <span class="st">"JCHAIN"</span>, </span>
<span>                   <span class="st">'CD68'</span>, <span class="st">'CD163'</span>, <span class="st">'CD14'</span>, </span>
<span>                   <span class="st">'S100A9'</span>, <span class="st">'S100A8'</span>, <span class="st">'MMP19'</span>,</span>
<span>                   <span class="st">'C1QA'</span>,  <span class="st">'C1QB'</span>, </span>
<span>                   <span class="st">'TPSAB1'</span>, <span class="st">'TPSB2'</span>, </span>
<span>                   <span class="st">'KLRB1'</span>, <span class="st">"KLRD1"</span>, <span class="st">'NCR1'</span>, <span class="st">"GNLY"</span>, <span class="st">"NKG7"</span>, </span>
<span>                   <span class="st">'FGF7'</span>, <span class="st">'MME'</span>, <span class="st">'ACTA2'</span>, <span class="st">"COL3A1"</span>, </span>
<span>                   <span class="st">'PECAM1'</span>, <span class="st">'VWF'</span>, </span>
<span>                   <span class="st">'EPCAM'</span>, <span class="st">'KRT19'</span>, <span class="st">'PROM1'</span>, <span class="st">'ALDH1A1'</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DotPlot.html">DotPlot</a></span><span class="op">(</span><span class="va">seurat_clustered</span>, </span>
<span>        features <span class="op">=</span> <span class="va">genes_to_check</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html">coord_flip</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integrated_analysis_multiple_single_cell_datasets_files/figure-html/unnamed-chunk-22-1.png" class="img-fluid" width="768"></p>
</div>
</div>
</section><section id="手动注释" class="level4"><h4 class="anchored" data-anchor-id="手动注释">手动注释</h4>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_cache/html/unnamed-chunk-23_4f847f0f6aac71fe74fe2044add4b0d5">
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">seurat_clustered</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/Idents.html">RenameIdents</a></span><span class="op">(</span><span class="va">seurat_clustered</span>, </span>
<span>                                 <span class="st">"0"</span> <span class="op">=</span> <span class="st">"B cells"</span>,</span>
<span>                                 <span class="st">"1"</span> <span class="op">=</span> <span class="st">"T cells"</span>,</span>
<span>                                 <span class="st">"2"</span> <span class="op">=</span> <span class="st">"Epithelial/cancer"</span>,</span>
<span>                                 <span class="st">"3"</span> <span class="op">=</span> <span class="st">"T cells"</span>,</span>
<span>                                 <span class="st">"4"</span> <span class="op">=</span> <span class="st">"T cells"</span>,</span>
<span>                                 <span class="st">"5"</span> <span class="op">=</span> <span class="st">"Macrophages"</span>,</span>
<span>                                 <span class="st">"6"</span> <span class="op">=</span> <span class="st">"T cells"</span>,</span>
<span>                                 <span class="st">"7"</span> <span class="op">=</span> <span class="st">"Plasma cells"</span>,</span>
<span>                                 <span class="st">"8"</span> <span class="op">=</span> <span class="st">"B cells"</span>,</span>
<span>                                 <span class="st">"9"</span> <span class="op">=</span> <span class="st">"Epithelial/cancer"</span>,</span>
<span>                                 <span class="st">"10"</span> <span class="op">=</span> <span class="st">"B cells"</span>,</span>
<span>                                 <span class="st">"11"</span> <span class="op">=</span> <span class="st">"B cells"</span>,</span>
<span>                                 <span class="st">"12"</span> <span class="op">=</span> <span class="st">"Unknown"</span>,</span>
<span>                                 <span class="st">"13"</span> <span class="op">=</span> <span class="st">"NK cells"</span>,</span>
<span>                                 <span class="st">"14"</span> <span class="op">=</span> <span class="st">"Unknown"</span>,</span>
<span>                                 <span class="st">"15"</span> <span class="op">=</span> <span class="st">"Epithelial/cancer"</span>,</span>
<span>                                 <span class="st">"16"</span> <span class="op">=</span> <span class="st">"Myeloid (unspecific)"</span>,</span>
<span>                                 <span class="st">"17"</span> <span class="op">=</span> <span class="st">"Epithelial/cancer"</span>,</span>
<span>                                 <span class="st">"18"</span> <span class="op">=</span> <span class="st">"Epithelial/cancer"</span>,</span>
<span>                                 <span class="st">"19"</span> <span class="op">=</span> <span class="st">"Epithelial/cancer"</span>,</span>
<span>                                 <span class="st">"20"</span> <span class="op">=</span> <span class="st">"B cells"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/Idents.html">Idents</a></span><span class="op">(</span><span class="va">seurat_clustered</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
             B cells              T cells    Epithelial/cancer 
               14739                17626                 7993 
         Macrophages         Plasma cells              Unknown 
                3765                 2365                 1157 
            NK cells Myeloid (unspecific) 
                 707                  232 </code></pre>
</div>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Plot the UMAP</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">seurat_clustered</span>, </span>
<span>        reduction <span class="op">=</span> <span class="st">"umap.integrated"</span>, </span>
<span>        label <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integrated_analysis_multiple_single_cell_datasets_files/figure-html/unnamed-chunk-23-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>在注释好的数据中再次检查marker基因的表达情况：</p>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_cache/html/unnamed-chunk-24_a9945dd6a5fbf96d0b24dfe4b1df786b">
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DotPlot.html">DotPlot</a></span><span class="op">(</span><span class="va">seurat_clustered</span>, </span>
<span>        features <span class="op">=</span> <span class="va">genes_to_check</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">30</span>,vjust <span class="op">=</span> <span class="fl">0.85</span>,hjust <span class="op">=</span> <span class="fl">0.75</span><span class="op">)</span></span>
<span><span class="op">)</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integrated_analysis_multiple_single_cell_datasets_files/figure-html/unnamed-chunk-24-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>因为接下来我们要提取髓系细胞进一步分析，所以通过几种可视化方法检查一下髓系细胞通用marker（CD163和AIF1）的表达情况：</p>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_cache/html/unnamed-chunk-25_aa4049e0ef188f7a32b266d7f18fe2a0">
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/RidgePlot.html">RidgePlot</a></span><span class="op">(</span><span class="va">seurat_clustered</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CD163"</span>, <span class="st">"AIF1"</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integrated_analysis_multiple_single_cell_datasets_files/figure-html/unnamed-chunk-25-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_cache/html/unnamed-chunk-26_587616531705371641912ceb7dda09e7">
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html">VlnPlot</a></span><span class="op">(</span><span class="va">seurat_clustered</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CD163"</span>, <span class="st">"AIF1"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integrated_analysis_multiple_single_cell_datasets_files/figure-html/unnamed-chunk-26-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_cache/html/unnamed-chunk-27_7546b8544540e92ad1843e2d66d04181">
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html">FeaturePlot</a></span><span class="op">(</span><span class="va">seurat_clustered</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CD163"</span>, <span class="st">"AIF1"</span><span class="op">)</span>, label <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integrated_analysis_multiple_single_cell_datasets_files/figure-html/unnamed-chunk-27-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>保存</p>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_cache/html/unnamed-chunk-28_3a885c4b4ce78be240f393bf0a298618">
<div class="sourceCode" id="cb77"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">seurat_clustered</span>, </span>
<span>        file <span class="op">=</span> <span class="st">"output/sc_supplementary/GSE150430_seurat_annotated.RDS"</span><span class="op">)</span>; <span class="fu"><a href="https://rdrr.io/pkg/beepr/man/beep.html">beep</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section></section><section id="提取髓系细胞" class="level3"><h3 class="anchored" data-anchor-id="提取髓系细胞">提取髓系细胞</h3>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_cache/html/unnamed-chunk-29_d7a7b807e7991dd15ac04027ccbd48a9">
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">sub_seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="va">seurat_clustered</span>, </span>
<span>                     idents <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Macrophages"</span>, </span>
<span>                                <span class="st">"Myeloid (unspecific)"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sub_seurat</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
46357 features across 3997 samples within 2 assays 
Active assay: SCT (21637 features, 3000 variable features)
 3 layers present: counts, data, scale.data
 1 other assay present: RNA
 4 dimensional reductions calculated: pca, umap.unintegrated, harmony, umap.integrated</code></pre>
</div>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">sub_seurat</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "orig.ident"            "nCount_RNA"            "nFeature_RNA"         
 [4] "samples"               "log10GenesPerUMI"      "mitoRatio"            
 [7] "cells"                 "nCount_SCT"            "nFeature_SCT"         
[10] "S.Score"               "G2M.Score"             "Phase"                
[13] "mitoFr"                "unintegrated_clusters" "seurat_clusters"      
[16] "groups"                "integrated_clusters"   "SCT_snn_res.0.01"     
[19] "SCT_snn_res.0.05"      "SCT_snn_res.0.1"       "SCT_snn_res.0.2"      
[22] "SCT_snn_res.0.3"       "SCT_snn_res.0.4"       "SCT_snn_res.0.5"      
[25] "SCT_snn_res.0.8"       "SCT_snn_res.1"        </code></pre>
</div>
<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">myeloid_seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/CreateSeuratObject.html">CreateSeuratObject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">sub_seurat</span><span class="op">@</span><span class="va">assays</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>                                     meta.data <span class="op">=</span> <span class="va">sub_seurat</span><span class="op">@</span><span class="va">meta.data</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">7</span>, <span class="fl">10</span><span class="op">:</span><span class="fl">13</span>, <span class="fl">16</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">myeloid_seurat</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
24720 features across 3997 samples within 1 assay 
Active assay: RNA (24720 features, 0 variable features)
 1 layer present: counts</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Warning
</div>
</div>
<div class="callout-body-container callout-body">
<p>可以看到在重新创建Seurat之后，默认的“RNA” assay之中只有一个layer。为了后面执行<code>SCTransform</code>及整合，<strong>这里需要将其按照样本拆分成不同的layers</strong>。</p>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_cache/html/unnamed-chunk-30_67e9b21961e07d06a9fee5c3d0ba5804">
<div class="sourceCode" id="cb84"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">myeloid_seurat</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html">split</a></span><span class="op">(</span><span class="va">myeloid_seurat</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span>, </span>
<span>                                 f <span class="op">=</span> <span class="va">myeloid_seurat</span><span class="op">$</span><span class="va">samples</span><span class="op">)</span></span>
<span><span class="va">myeloid_seurat</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
24720 features across 3997 samples within 1 assay 
Active assay: RNA (24720 features, 0 variable features)
 16 layers present: counts.N01, counts.P01, counts.P02, counts.P03, counts.P04, counts.P05, counts.P06, counts.P07, counts.P08, counts.P09, counts.P10, counts.P11, counts.P12, counts.P13, counts.P14, counts.P15</code></pre>
</div>
</div>
</div>
</div>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_cache/html/unnamed-chunk-31_f8abf3533e4750c730a5cf6ba0aeef4c">
<div class="sourceCode" id="cb86"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 补充原细胞注释信息</span></span>
<span><span class="va">myeloid_seurat</span><span class="op">$</span><span class="va">old_clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/Idents.html">Idents</a></span><span class="op">(</span><span class="va">sub_seurat</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">myeloid_seurat</span><span class="op">$</span><span class="va">old_clusters</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
         Macrophages Myeloid (unspecific) 
                3765                  232 </code></pre>
</div>
<div class="sourceCode" id="cb88"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 保存</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">myeloid_seurat</span>, file <span class="op">=</span> <span class="st">"output/sc_supplementary/GSE150430_myeloid_seurat.RDS"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr></section><section id="髓系细胞进一步归一化整合降维分群" class="level3"><h3 class="anchored" data-anchor-id="髓系细胞进一步归一化整合降维分群">髓系细胞进一步归一化、整合、降维、分群</h3>
<section id="sctransformpca" class="level4"><h4 class="anchored" data-anchor-id="sctransformpca">SCTransform、PCA</h4>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_cache/html/unnamed-chunk-32_861bbfd09031eb2817d2e2e60edc9039">
<div class="sourceCode" id="cb89"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/rm.html">rm</a></span><span class="op">(</span>list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ls.html">ls</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/gc.html">gc</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>             used   (Mb) gc trigger    (Mb) limit (Mb)   max used    (Mb)
Ncells    8542995  456.3   15406287   822.8         NA   15406287   822.8
Vcells 1003510688 7656.2 3232715316 24663.7     102400 3232715179 24663.7</code></pre>
</div>
<div class="sourceCode" id="cb91"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">myeloid_seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="st">"output/sc_supplementary/GSE150430_myeloid_seurat.RDS"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># SCTranform</span></span>
<span><span class="va">myeloid_seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SCTransform.html">SCTransform</a></span><span class="op">(</span><span class="va">myeloid_seurat</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">myeloid_seurat</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
39690 features across 3997 samples within 2 assays 
Active assay: SCT (14970 features, 3000 variable features)
 3 layers present: counts, data, scale.data
 1 other assay present: RNA</code></pre>
</div>
<div class="sourceCode" id="cb93"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Check which assays are stored in objects</span></span>
<span><span class="va">myeloid_seurat</span><span class="op">@</span><span class="va">assays</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$RNA
Assay (v5) data with 24720 features for 3997 cells
First 10 features:
 RP11-34P13.7, RP11-34P13.8, AL627309.1, AP006222.2, RP4-669L17.10,
RP4-669L17.2, RP5-857K21.4, RP11-206L10.3, RP11-206L10.5, RP11-206L10.2 
Layers:
 counts.N01, counts.P01, counts.P02, counts.P03, counts.P04, counts.P05,
counts.P06, counts.P07, counts.P08, counts.P09, counts.P10, counts.P11,
counts.P12, counts.P13, counts.P14, counts.P15 

$SCT
SCTAssay data with 14970 features for 3997 cells, and 16 SCTModel(s) 
Top 10 variable features:
 IL1RN, CLEC10A, PKIB, HBEGF, MSR1, IER3, LGMN, CA2, CFD, GLUL </code></pre>
</div>
<div class="sourceCode" id="cb95"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 查看目前默认的assay</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/DefaultAssay.html">DefaultAssay</a></span><span class="op">(</span><span class="va">myeloid_seurat</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "SCT"</code></pre>
</div>
<div class="sourceCode" id="cb97"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 查看默认assay的layers</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/Layers.html">Layers</a></span><span class="op">(</span><span class="va">myeloid_seurat</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "counts"     "data"       "scale.data"</code></pre>
</div>
<div class="sourceCode" id="cb99"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 执行PCA</span></span>
<span><span class="va">myeloid_seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html">RunPCA</a></span><span class="op">(</span><span class="va">myeloid_seurat</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="不进行整合时检查细胞分群情况" class="level4"><h4 class="anchored" data-anchor-id="不进行整合时检查细胞分群情况">不进行整合时检查细胞分群情况：</h4>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_cache/html/unnamed-chunk-33_b8da8c25a4cb8cd8ba811a12f315a074">
<div class="sourceCode" id="cb100"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 查看降维信息</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">myeloid_seurat</span><span class="op">@</span><span class="va">reductions</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "pca"</code></pre>
</div>
<div class="sourceCode" id="cb102"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Run UMAP</span></span>
<span><span class="va">myeloid_seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html">RunUMAP</a></span><span class="op">(</span><span class="va">myeloid_seurat</span>, </span>
<span>                          dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">40</span>, </span>
<span>                          reduction <span class="op">=</span> <span class="st">"pca"</span>,</span>
<span>                          reduction.name <span class="op">=</span> <span class="st">"umap.unintegrated"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 分群</span></span>
<span><span class="co"># Determine the K-nearest neighbor graph</span></span>
<span><span class="va">myeloid_seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html">FindNeighbors</a></span><span class="op">(</span><span class="va">myeloid_seurat</span>, </span>
<span>                                dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">40</span>, </span>
<span>                                reduction <span class="op">=</span> <span class="st">"pca"</span><span class="op">)</span></span>
<span><span class="va">myeloid_seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html">FindClusters</a></span><span class="op">(</span><span class="va">myeloid_seurat</span>, </span>
<span>                               cluster.name <span class="op">=</span> <span class="st">"unintegrated_clusters"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 3997
Number of edges: 211202

Running Louvain algorithm...
Maximum modularity in 10 random starts: 0.7202
Number of communities: 8
Elapsed time: 0 seconds</code></pre>
</div>
<div class="sourceCode" id="cb104"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Plot UMAP</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">myeloid_seurat</span>, </span>
<span>              reduction <span class="op">=</span> <span class="st">"umap.unintegrated"</span>,</span>
<span>              group.by <span class="op">=</span> <span class="st">"samples"</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">myeloid_seurat</span>, </span>
<span>              reduction <span class="op">=</span> <span class="st">"umap.unintegrated"</span>,</span>
<span>              split.by <span class="op">=</span> <span class="st">"samples"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">plot_grid</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span>, </span>
<span>          ncol <span class="op">=</span> <span class="fl">1</span>, labels <span class="op">=</span> <span class="st">"AUTO"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integrated_analysis_multiple_single_cell_datasets_files/figure-html/unnamed-chunk-33-1.png" class="img-fluid" width="960"></p>
</div>
</div>
</section><section id="整合-1" class="level4"><h4 class="anchored" data-anchor-id="整合-1">整合</h4>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_cache/html/unnamed-chunk-34_4e7aae10f1621ff25e3e5e7c4cea9f78">
<div class="sourceCode" id="cb105"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 整合</span></span>
<span><span class="va">myeloid_integrated</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/IntegrateLayers.html">IntegrateLayers</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">myeloid_seurat</span>,</span>
<span>                                      method <span class="op">=</span> <span class="va">HarmonyIntegration</span>,</span>
<span>                                      verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># 整合后合并RNA layer</span></span>
<span><span class="va">myeloid_integrated</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/SplitLayers.html">JoinLayers</a></span><span class="op">(</span><span class="va">myeloid_integrated</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 查看整合后的降维信息</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">myeloid_integrated</span><span class="op">@</span><span class="va">reductions</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "pca"               "umap.unintegrated" "harmony"          </code></pre>
</div>
</div>
</section><section id="整合后检验细胞分群情况-1" class="level4"><h4 class="anchored" data-anchor-id="整合后检验细胞分群情况-1"><strong>整合后检验细胞分群情况</strong></h4>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_cache/html/unnamed-chunk-35_c1e130f9efb75a06bb035da9f1a6587e">
<div class="sourceCode" id="cb107"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">123456</span><span class="op">)</span></span>
<span><span class="co"># Run UMAP</span></span>
<span><span class="va">myeloid_integrated</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html">RunUMAP</a></span><span class="op">(</span><span class="va">myeloid_integrated</span>, </span>
<span>                              dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">40</span>,</span>
<span>                              reduction <span class="op">=</span> <span class="st">"harmony"</span>, <span class="co"># 更改降维来源为整合后的"harmony"</span></span>
<span>                              reduction.name <span class="op">=</span> <span class="st">"umap.integrated"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">myeloid_integrated</span><span class="op">@</span><span class="va">reductions</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "pca"               "umap.unintegrated" "harmony"          
[4] "umap.integrated"  </code></pre>
</div>
<div class="sourceCode" id="cb109"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 分群</span></span>
<span><span class="va">myeloid_integrated</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html">FindNeighbors</a></span><span class="op">(</span><span class="va">myeloid_integrated</span>, </span>
<span>                                    dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">40</span>, </span>
<span>                                    reduction <span class="op">=</span> <span class="st">"harmony"</span><span class="op">)</span> <span class="co">#更改降维来源为"harmony"</span></span>
<span><span class="va">myeloid_integrated</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html">FindClusters</a></span><span class="op">(</span><span class="va">myeloid_integrated</span>, </span>
<span>                                   cluster.name <span class="op">=</span> <span class="st">"integrated_clusters"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 3997
Number of edges: 290917

Running Louvain algorithm...
Maximum modularity in 10 random starts: 0.7795
Number of communities: 10
Elapsed time: 0 seconds</code></pre>
</div>
<div class="sourceCode" id="cb111"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">myeloid_integrated</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "orig.ident"            "nCount_RNA"            "nFeature_RNA"         
 [4] "samples"               "log10GenesPerUMI"      "mitoRatio"            
 [7] "cells"                 "S.Score"               "G2M.Score"            
[10] "Phase"                 "mitoFr"                "groups"               
[13] "old_clusters"          "nCount_SCT"            "nFeature_SCT"         
[16] "unintegrated_clusters" "seurat_clusters"       "integrated_clusters"  </code></pre>
</div>
<div class="sourceCode" id="cb113"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Plot UMAP                             </span></span>
<span><span class="va">p3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">myeloid_integrated</span>, </span>
<span>              reduction <span class="op">=</span> <span class="st">"umap.integrated"</span>, </span>
<span>              group.by <span class="op">=</span> <span class="st">"samples"</span><span class="op">)</span></span>
<span><span class="va">p4</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">myeloid_integrated</span>, </span>
<span>              reduction <span class="op">=</span> <span class="st">"umap.integrated"</span>, </span>
<span>              split.by <span class="op">=</span> <span class="st">"samples"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">plot_grid</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p3</span>, <span class="va">p2</span>, <span class="va">p4</span>, </span>
<span>          ncol <span class="op">=</span> <span class="fl">1</span>, </span>
<span>          labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Before Harmony"</span>, <span class="st">"After Harmony"</span>, </span>
<span>                     <span class="st">"Before Harmony"</span>, <span class="st">"After Harmony"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integrated_analysis_multiple_single_cell_datasets_files/figure-html/unnamed-chunk-35-1.png" class="img-fluid" width="960"></p>
</div>
</div>
</section><section id="聚类-1" class="level4"><h4 class="anchored" data-anchor-id="聚类-1">聚类</h4>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_cache/html/unnamed-chunk-36_0f95e3f2313fad4c6ea4f9c9bdfa73b3">
<div class="sourceCode" id="cb114"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Determine the clusters for various resolutions                                </span></span>
<span><span class="va">myeloid_integrated</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html">FindClusters</a></span><span class="op">(</span><span class="va">myeloid_integrated</span>,</span>
<span>                                   resolution <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.01</span>, <span class="fl">0.05</span>, <span class="fl">0.1</span>, <span class="fl">0.2</span>, <span class="fl">0.3</span>, </span>
<span>                                                  <span class="fl">0.4</span>, <span class="fl">0.5</span>, <span class="fl">0.8</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>                                   verbose <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span><span class="co"># Explore resolutions</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">myeloid_integrated</span><span class="op">@</span><span class="va">meta.data</span>, <span class="fl">5</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                       orig.ident nCount_RNA nFeature_RNA samples
N01_AGCGTCGAGTGAAGAG.1        N01   2691.228         2086     N01
N01_GCAATCATCAGCCTAA.1        N01   2139.759         1566     N01
N01_CGTGTAATCCCACTTG.1        N01   2251.166         1803     N01
N01_AAACGGGCATTTCAGG.1        N01   1909.545         1226     N01
N01_AAAGATGCAATGTAAG.1        N01   1796.109         1360     N01
                       log10GenesPerUMI   mitoRatio                  cells
N01_AGCGTCGAGTGAAGAG.1        0.9677441 0.010726702 N01_AGCGTCGAGTGAAGAG.1
N01_GCAATCATCAGCCTAA.1        0.9592918 0.018235699 N01_GCAATCATCAGCCTAA.1
N01_CGTGTAATCCCACTTG.1        0.9712410 0.017482940 N01_CGTGTAATCCCACTTG.1
N01_AAACGGGCATTTCAGG.1        0.9413461 0.011039803 N01_AAACGGGCATTTCAGG.1
N01_AAAGATGCAATGTAAG.1        0.9628822 0.006095955 N01_AAAGATGCAATGTAAG.1
                            S.Score   G2M.Score Phase mitoFr groups
N01_AGCGTCGAGTGAAGAG.1 -0.093324147 -0.04497299    G1 Medium Normal
N01_GCAATCATCAGCCTAA.1  0.008569604 -0.03484456     S   High Normal
N01_CGTGTAATCCCACTTG.1 -0.013729028 -0.05064395    G1   High Normal
N01_AAACGGGCATTTCAGG.1 -0.005136538 -0.03723796    G1 Medium Normal
N01_AAAGATGCAATGTAAG.1 -0.039476505 -0.04625311    G1    Low Normal
                               old_clusters nCount_SCT nFeature_SCT
N01_AGCGTCGAGTGAAGAG.1          Macrophages       2455         1929
N01_GCAATCATCAGCCTAA.1          Macrophages       2252         1505
N01_CGTGTAATCCCACTTG.1          Macrophages       2359         1712
N01_AAACGGGCATTTCAGG.1          Macrophages       2044         1197
N01_AAAGATGCAATGTAAG.1 Myeloid (unspecific)       2271         1322
                       unintegrated_clusters seurat_clusters
N01_AGCGTCGAGTGAAGAG.1                     0               5
N01_GCAATCATCAGCCTAA.1                     0               6
N01_CGTGTAATCCCACTTG.1                     0               5
N01_AAACGGGCATTTCAGG.1                     6               7
N01_AAAGATGCAATGTAAG.1                     5              10
                       integrated_clusters SCT_snn_res.0.01 SCT_snn_res.0.05
N01_AGCGTCGAGTGAAGAG.1                   3                0                0
N01_GCAATCATCAGCCTAA.1                   3                0                0
N01_CGTGTAATCCCACTTG.1                   3                0                0
N01_AAACGGGCATTTCAGG.1                   1                0                0
N01_AAAGATGCAATGTAAG.1                   7                0                1
                       SCT_snn_res.0.1 SCT_snn_res.0.2 SCT_snn_res.0.3
N01_AGCGTCGAGTGAAGAG.1               0               0               3
N01_GCAATCATCAGCCTAA.1               0               0               3
N01_CGTGTAATCCCACTTG.1               0               0               3
N01_AAACGGGCATTTCAGG.1               1               1               1
N01_AAAGATGCAATGTAAG.1               3               4               5
                       SCT_snn_res.0.4 SCT_snn_res.0.5 SCT_snn_res.0.8
N01_AGCGTCGAGTGAAGAG.1               3               3               3
N01_GCAATCATCAGCCTAA.1               3               3               3
N01_CGTGTAATCCCACTTG.1               3               3               3
N01_AAACGGGCATTTCAGG.1               1               1               1
N01_AAAGATGCAATGTAAG.1               6               7               7
                       SCT_snn_res.1
N01_AGCGTCGAGTGAAGAG.1             5
N01_GCAATCATCAGCCTAA.1             6
N01_CGTGTAATCCCACTTG.1             5
N01_AAACGGGCATTTCAGG.1             7
N01_AAAGATGCAATGTAAG.1            10</code></pre>
</div>
<div class="sourceCode" id="cb116"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 查看各个分辨率下的细胞分群情况</span></span>
<span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">myeloid_integrated</span><span class="op">@</span><span class="va">meta.data</span>, </span>
<span>       <span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html">starts_with</a></span><span class="op">(</span>match <span class="op">=</span> <span class="st">"SCT_snn_res."</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>  </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">levels</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$SCT_snn_res.0.01
[1] "0"

$SCT_snn_res.0.05
[1] "0" "1"

$SCT_snn_res.0.1
[1] "0" "1" "2" "3"

$SCT_snn_res.0.2
[1] "0" "1" "2" "3" "4"

$SCT_snn_res.0.3
[1] "0" "1" "2" "3" "4" "5"

$SCT_snn_res.0.4
[1] "0" "1" "2" "3" "4" "5" "6" "7"

$SCT_snn_res.0.5
[1] "0" "1" "2" "3" "4" "5" "6" "7" "8"

$SCT_snn_res.0.8
 [1] "0" "1" "2" "3" "4" "5" "6" "7" "8" "9"

$SCT_snn_res.1
 [1] "0"  "1"  "2"  "3"  "4"  "5"  "6"  "7"  "8"  "9"  "10" "11" "12"</code></pre>
</div>
</div>
<p>绘制聚类树展示不同分辨率下的细胞分群情况及相互关系</p>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_cache/html/unnamed-chunk-37_71331c3da68ac4b1ab6d6b4f3db28c97">
<div class="sourceCode" id="cb118"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://lazappi.github.io/clustree/reference/clustree.html">clustree</a></span><span class="op">(</span><span class="va">myeloid_integrated</span><span class="op">@</span><span class="va">meta.data</span>, </span>
<span>                 prefix <span class="op">=</span> <span class="st">"SCT_snn_res."</span><span class="op">)</span> <span class="co"># 指定包含聚类信息的列</span></span>
<span><span class="va">tree</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integrated_analysis_multiple_single_cell_datasets_files/figure-html/unnamed-chunk-37-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>这里，选取分辨为0.2</p>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_cache/html/unnamed-chunk-38_230c5b3c7c965ec36b76ac6d13a4c020">
<div class="sourceCode" id="cb119"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/Idents.html">Idents</a></span><span class="op">(</span><span class="va">myeloid_integrated</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"SCT_snn_res.0.2"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>聚类可视化</p>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_cache/html/unnamed-chunk-39_bd03bce8bbc6d04e8965505bb67f11ec">
<div class="sourceCode" id="cb120"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Plot the UMAP</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">myeloid_integrated</span>,</span>
<span>        reduction <span class="op">=</span> <span class="st">"umap.integrated"</span>,</span>
<span>        label <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integrated_analysis_multiple_single_cell_datasets_files/figure-html/unnamed-chunk-39-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section><section id="细胞分群质量评估-1" class="level4"><h4 class="anchored" data-anchor-id="细胞分群质量评估-1"><strong>细胞分群质量评估</strong></h4>
<p>分析样本类型是否影响细胞分群</p>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_cache/html/unnamed-chunk-40_00aa48023aa5e37bbcbe741069020617">
<div class="sourceCode" id="cb121"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 先简单查看不同cluster的细胞数</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">myeloid_integrated</span><span class="op">@</span><span class="va">active.ident</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
   0    1    2    3    4 
1708 1133  613  352  191 </code></pre>
</div>
<div class="sourceCode" id="cb123"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 查看不同样本类型中的细胞分群情况</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">myeloid_integrated</span>, </span>
<span>        reduction <span class="op">=</span> <span class="st">"umap.integrated"</span>,</span>
<span>        label <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>        split.by <span class="op">=</span> <span class="st">"groups"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html">NoLegend</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integrated_analysis_multiple_single_cell_datasets_files/figure-html/unnamed-chunk-40-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>分析细胞周期是否影响细胞分群</p>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_cache/html/unnamed-chunk-41_a9db9be204247f0c8afbef85ea5253b3">
<div class="sourceCode" id="cb124"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Explore whether clusters segregate by cell cycle phase</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">myeloid_integrated</span>,</span>
<span>        reduction <span class="op">=</span> <span class="st">"umap.integrated"</span>,</span>
<span>        label <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>        split.by <span class="op">=</span> <span class="st">"Phase"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html">NoLegend</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integrated_analysis_multiple_single_cell_datasets_files/figure-html/unnamed-chunk-41-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>分析其他非期望变异来源是否会影响细胞分群</p>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_cache/html/unnamed-chunk-42_0372577ada928073302956641ba181a3">
<div class="sourceCode" id="cb125"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Determine metrics to plot present in seurat_clustered@meta.data</span></span>
<span><span class="va">metrics</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"nCount_RNA"</span>, <span class="st">"nFeature_RNA"</span>, <span class="st">"S.Score"</span>, <span class="st">"G2M.Score"</span>, <span class="st">"mitoRatio"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html">FeaturePlot</a></span><span class="op">(</span><span class="va">myeloid_integrated</span>, </span>
<span>            reduction <span class="op">=</span> <span class="st">"umap.integrated"</span>, </span>
<span>            features <span class="op">=</span> <span class="va">metrics</span>,</span>
<span>            pt.size <span class="op">=</span> <span class="fl">0.4</span>, </span>
<span>            order <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            min.cutoff <span class="op">=</span> <span class="st">'q10'</span>, </span>
<span>            label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integrated_analysis_multiple_single_cell_datasets_files/figure-html/unnamed-chunk-42-1.png" class="img-fluid" width="960"></p>
</div>
</div>
<p>保存</p>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_cache/html/unnamed-chunk-43_54a349c636840c46c660acc19f5dbc94">
<div class="sourceCode" id="cb126"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">myeloid_integrated</span>, </span>
<span>        file <span class="op">=</span> <span class="st">"output/sc_supplementary/GSE150430_myeloid_clustered.RDS"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="亚群细胞注释" class="level4"><h4 class="anchored" data-anchor-id="亚群细胞注释">亚群细胞注释</h4>


<!-- -->

<div id="refs" class="references csl-bib-body hanging-indent" role="list" style="display: none">
<div id="ref-gong2023" class="csl-entry" role="listitem">
Gong, Lanqi, Jie Luo, Yu Zhang, Yuma Yang, Shanshan Li, Xiaona Fang, Baifeng Zhang, et al. 2023. <span>“Nasopharyngeal Carcinoma Cells Promote Regulatory T Cell Development and Suppressive Activity via CD70-CD27 Interaction.”</span> <em>Nature Communications</em> 14 (1). <a href="https://doi.org/10.1038/s41467-023-37614-6">https://doi.org/10.1038/s41467-023-37614-6</a>.
</div>
</div>
</section></section></section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  let localAlternateSentinel = 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var filterRegex = new RegExp("https:\/\/djhcod\.github\.io\/r-notes\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
          // target, if specified
          link.setAttribute("target", "_blank");
      }
    }
});
</script><script src="https://utteranc.es/client.js" repo="https://github.com/djhcod/r-notes" issue-term="pathname" theme="github-light" crossorigin="anonymous" async="">
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="../../single_cell/sc_supplementary/universal_marker.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">29</span>&nbsp; <span class="chapter-title">细胞分群通用marker</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../../quarto_foundation/quarto_foundation.html" class="pagination-link">
        <span class="nav-page-text">Quarto基础</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb127" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb127-1"><a href="#cb127-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb127-2"><a href="#cb127-2" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> "多个单细胞数据集整合分析"</span></span>
<span id="cb127-3"><a href="#cb127-3" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb127-4"><a href="#cb127-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-5"><a href="#cb127-5" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; 参考：</span><span class="co">[</span><span class="ot">单细胞多数据集整合示例</span><span class="co">](https://mp.weixin.qq.com/s/51PtUi7ZKp1nbrFGElU74A)</span></span>
<span id="cb127-6"><a href="#cb127-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-7"><a href="#cb127-7" aria-hidden="true" tabindex="-1"></a>有时候，为了扩大数据量和得到更加可靠的结论，我们可能获取和下载了多个单细胞数据集。如果我们只关注某一细胞类型，如T cells或B cells或髓系细胞，那么就需要合并分析这些数据集，提取感兴趣的细胞亚群。这时候有两种方法可以选择：</span>
<span id="cb127-8"><a href="#cb127-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-9"><a href="#cb127-9" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>**对各个数据集分别降维分群，然后提取各个数据集的感兴趣细胞亚群并整合到一起**</span>
<span id="cb127-10"><a href="#cb127-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-11"><a href="#cb127-11" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>**在一开始分析的时候就对这些数据集进行整合，然后分群，提起感兴趣细胞群**</span>
<span id="cb127-12"><a href="#cb127-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-13"><a href="#cb127-13" aria-hidden="true" tabindex="-1"></a>**这个部分怎么分析并没有一个特定的答案或者规则，具体还是得看数据集适合以上提出的哪种分析顺序。**</span>
<span id="cb127-14"><a href="#cb127-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-15"><a href="#cb127-15" aria-hidden="true" tabindex="-1"></a>例如<span class="co">[</span><span class="ot">@gong2023</span><span class="co">]</span>就用了第二种方法。选取了三个鼻咽癌单细胞转录组数据： GSE150825, GSE150430, GSE162025。首先对三组数据进行整合，然后选取其中的一部分T细胞进行进一步分析。</span>
<span id="cb127-16"><a href="#cb127-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-17"><a href="#cb127-17" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/截屏2024-01-10%2012.00.48.png)</span>{width="600"}</span>
<span id="cb127-18"><a href="#cb127-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-19"><a href="#cb127-19" aria-hidden="true" tabindex="-1"></a>这里我们介绍第一种方式，即**对各数据集分别降维分群，然后提取并整合各个数据集的感兴趣细胞亚群。**</span>
<span id="cb127-20"><a href="#cb127-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-21"><a href="#cb127-21" aria-hidden="true" tabindex="-1"></a><span class="fu">## 加载包</span></span>
<span id="cb127-22"><a href="#cb127-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-25"><a href="#cb127-25" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb127-26"><a href="#cb127-26" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb127-27"><a href="#cb127-27" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb127-28"><a href="#cb127-28" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(beepr)</span>
<span id="cb127-29"><a href="#cb127-29" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(cowplot)</span>
<span id="cb127-30"><a href="#cb127-30" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb127-31"><a href="#cb127-31" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(clustree)</span>
<span id="cb127-32"><a href="#cb127-32" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb127-33"><a href="#cb127-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-34"><a href="#cb127-34" aria-hidden="true" tabindex="-1"></a><span class="fu">## GSE150430</span></span>
<span id="cb127-35"><a href="#cb127-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-36"><a href="#cb127-36" aria-hidden="true" tabindex="-1"></a>这里我们直接载入 @sec-reading_non_standard_large_volume_text 中构建好的Seurat对象，该案例数据的介绍及读取过程参见该章节。</span>
<span id="cb127-37"><a href="#cb127-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-38"><a href="#cb127-38" aria-hidden="true" tabindex="-1"></a><span class="fu">### 数据导入</span></span>
<span id="cb127-39"><a href="#cb127-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-42"><a href="#cb127-42" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb127-43"><a href="#cb127-43" aria-hidden="true" tabindex="-1"></a>merged_seurat <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"output/sc_supplementary/GSE150430_merged_seurat.rds"</span>)</span>
<span id="cb127-44"><a href="#cb127-44" aria-hidden="true" tabindex="-1"></a>merged_seurat</span>
<span id="cb127-45"><a href="#cb127-45" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(merged_seurat, <span class="dv">4</span>)</span>
<span id="cb127-46"><a href="#cb127-46" aria-hidden="true" tabindex="-1"></a><span class="fu">unique</span>(merged_seurat<span class="sc">$</span>samples)</span>
<span id="cb127-47"><a href="#cb127-47" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb127-48"><a href="#cb127-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-49"><a href="#cb127-49" aria-hidden="true" tabindex="-1"></a><span class="fu">### 质控</span></span>
<span id="cb127-50"><a href="#cb127-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-53"><a href="#cb127-53" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb127-54"><a href="#cb127-54" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 3</span></span>
<span id="cb127-55"><a href="#cb127-55" aria-hidden="true" tabindex="-1"></a><span class="co"># Add number of genes per UMI for each cell to metadata</span></span>
<span id="cb127-56"><a href="#cb127-56" aria-hidden="true" tabindex="-1"></a>merged_seurat<span class="sc">$</span>log10GenesPerUMI <span class="ot">&lt;-</span> <span class="fu">log10</span>(merged_seurat<span class="sc">$</span>nFeature_RNA) <span class="sc">/</span> <span class="fu">log10</span>(merged_seurat<span class="sc">$</span>nCount_RNA)</span>
<span id="cb127-57"><a href="#cb127-57" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(merged_seurat<span class="sc">$</span>log10GenesPerUMI)</span>
<span id="cb127-58"><a href="#cb127-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-59"><a href="#cb127-59" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute percent mito ratio</span></span>
<span id="cb127-60"><a href="#cb127-60" aria-hidden="true" tabindex="-1"></a>merged_seurat<span class="sc">$</span>mitoRatio <span class="ot">&lt;-</span> <span class="fu">PercentageFeatureSet</span>(<span class="at">object =</span> merged_seurat, </span>
<span id="cb127-61"><a href="#cb127-61" aria-hidden="true" tabindex="-1"></a>                                                <span class="at">pattern =</span> <span class="st">"^MT-"</span>)</span>
<span id="cb127-62"><a href="#cb127-62" aria-hidden="true" tabindex="-1"></a>merged_seurat<span class="sc">$</span>mitoRatio <span class="ot">&lt;-</span> merged_seurat<span class="sc">$</span>mitoRatio <span class="sc">/</span> <span class="dv">100</span></span>
<span id="cb127-63"><a href="#cb127-63" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(merged_seurat<span class="sc">$</span>mitoRatio)</span>
<span id="cb127-64"><a href="#cb127-64" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(merged_seurat<span class="sc">$</span>mitoRatio)</span>
<span id="cb127-65"><a href="#cb127-65" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-66"><a href="#cb127-66" aria-hidden="true" tabindex="-1"></a><span class="co"># Add cell IDs to metadata</span></span>
<span id="cb127-67"><a href="#cb127-67" aria-hidden="true" tabindex="-1"></a>merged_seurat<span class="sc">$</span>cells <span class="ot">&lt;-</span> <span class="fu">rownames</span>(merged_seurat<span class="sc">@</span>meta.data)</span>
<span id="cb127-68"><a href="#cb127-68" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb127-69"><a href="#cb127-69" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-72"><a href="#cb127-72" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb127-73"><a href="#cb127-73" aria-hidden="true" tabindex="-1"></a><span class="co">#| layout-ncol: 2</span></span>
<span id="cb127-74"><a href="#cb127-74" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualize the number of cell counts per sample</span></span>
<span id="cb127-75"><a href="#cb127-75" aria-hidden="true" tabindex="-1"></a>merged_seurat<span class="sc">@</span>meta.data  <span class="sc">|&gt;</span> </span>
<span id="cb127-76"><a href="#cb127-76" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> samples, <span class="at">fill =</span> samples)) <span class="sc">+</span> </span>
<span id="cb127-77"><a href="#cb127-77" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_bar</span>() <span class="sc">+</span></span>
<span id="cb127-78"><a href="#cb127-78" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb127-79"><a href="#cb127-79" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">45</span>, <span class="at">vjust =</span> <span class="dv">1</span>, <span class="at">hjust =</span> <span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb127-80"><a href="#cb127-80" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>, <span class="at">face =</span> <span class="st">"bold"</span>)) <span class="sc">+</span></span>
<span id="cb127-81"><a href="#cb127-81" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"NCells"</span>)</span>
<span id="cb127-82"><a href="#cb127-82" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-83"><a href="#cb127-83" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualize the number UMIs/transcripts per cell</span></span>
<span id="cb127-84"><a href="#cb127-84" aria-hidden="true" tabindex="-1"></a>merged_seurat<span class="sc">@</span>meta.data <span class="sc">|&gt;</span> </span>
<span id="cb127-85"><a href="#cb127-85" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">color =</span> samples, <span class="at">x =</span> nCount_RNA, <span class="at">fill =</span> samples)) <span class="sc">+</span> </span>
<span id="cb127-86"><a href="#cb127-86" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_density</span>(<span class="at">alpha =</span> <span class="fl">0.2</span>) <span class="sc">+</span> </span>
<span id="cb127-87"><a href="#cb127-87" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_x_log10</span>() <span class="sc">+</span> </span>
<span id="cb127-88"><a href="#cb127-88" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb127-89"><a href="#cb127-89" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ylab</span>(<span class="st">"Cell density"</span>) <span class="sc">+</span></span>
<span id="cb127-90"><a href="#cb127-90" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">500</span>)</span>
<span id="cb127-91"><a href="#cb127-91" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-92"><a href="#cb127-92" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualize the distribution of genes detected per cell via histogram</span></span>
<span id="cb127-93"><a href="#cb127-93" aria-hidden="true" tabindex="-1"></a>merged_seurat<span class="sc">@</span>meta.data <span class="sc">|&gt;</span></span>
<span id="cb127-94"><a href="#cb127-94" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">color =</span> samples, <span class="at">x =</span> nFeature_RNA, <span class="at">fill=</span> samples)) <span class="sc">+</span> </span>
<span id="cb127-95"><a href="#cb127-95" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_density</span>(<span class="at">alpha =</span> <span class="fl">0.2</span>) <span class="sc">+</span> </span>
<span id="cb127-96"><a href="#cb127-96" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb127-97"><a href="#cb127-97" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_x_log10</span>() <span class="sc">+</span> </span>
<span id="cb127-98"><a href="#cb127-98" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">200</span>)</span>
<span id="cb127-99"><a href="#cb127-99" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-100"><a href="#cb127-100" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualize the overall complexity of the gene expression by visualizing the genes detected per UMI (novelty score)</span></span>
<span id="cb127-101"><a href="#cb127-101" aria-hidden="true" tabindex="-1"></a>merged_seurat<span class="sc">@</span>meta.data <span class="sc">|&gt;</span></span>
<span id="cb127-102"><a href="#cb127-102" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> log10GenesPerUMI, <span class="at">color =</span> samples, <span class="at">fill=</span>samples)) <span class="sc">+</span></span>
<span id="cb127-103"><a href="#cb127-103" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_density</span>(<span class="at">alpha =</span> <span class="fl">0.2</span>) <span class="sc">+</span></span>
<span id="cb127-104"><a href="#cb127-104" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb127-105"><a href="#cb127-105" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fl">0.8</span>)</span>
<span id="cb127-106"><a href="#cb127-106" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-107"><a href="#cb127-107" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualize the distribution of mitochondrial gene expression detected per cell</span></span>
<span id="cb127-108"><a href="#cb127-108" aria-hidden="true" tabindex="-1"></a>merged_seurat<span class="sc">@</span>meta.data <span class="sc">|&gt;</span></span>
<span id="cb127-109"><a href="#cb127-109" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">color =</span> samples, <span class="at">x =</span> mitoRatio, <span class="at">fill =</span> samples)) <span class="sc">+</span> </span>
<span id="cb127-110"><a href="#cb127-110" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_density</span>(<span class="at">alpha =</span> <span class="fl">0.2</span>) <span class="sc">+</span> </span>
<span id="cb127-111"><a href="#cb127-111" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_x_log10</span>() <span class="sc">+</span> </span>
<span id="cb127-112"><a href="#cb127-112" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb127-113"><a href="#cb127-113" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fl">0.2</span>)</span>
<span id="cb127-114"><a href="#cb127-114" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb127-115"><a href="#cb127-115" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-118"><a href="#cb127-118" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb127-119"><a href="#cb127-119" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 10</span></span>
<span id="cb127-120"><a href="#cb127-120" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 10</span></span>
<span id="cb127-121"><a href="#cb127-121" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualize the correlation between genes detected and number of UMIs and determine whether strong presence of cells with low numbers of genes/UMIs</span></span>
<span id="cb127-122"><a href="#cb127-122" aria-hidden="true" tabindex="-1"></a>merged_seurat<span class="sc">@</span>meta.data <span class="sc">|&gt;</span> </span>
<span id="cb127-123"><a href="#cb127-123" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> nCount_RNA, <span class="at">y =</span> nFeature_RNA, <span class="at">color =</span> mitoRatio)) <span class="sc">+</span> </span>
<span id="cb127-124"><a href="#cb127-124" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>() <span class="sc">+</span> </span>
<span id="cb127-125"><a href="#cb127-125" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_colour_gradient</span>(<span class="at">low =</span> <span class="st">"gray90"</span>, <span class="at">high =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb127-126"><a href="#cb127-126" aria-hidden="true" tabindex="-1"></a>    <span class="fu">stat_smooth</span>(<span class="at">method =</span> lm) <span class="sc">+</span></span>
<span id="cb127-127"><a href="#cb127-127" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_x_log10</span>() <span class="sc">+</span> </span>
<span id="cb127-128"><a href="#cb127-128" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_y_log10</span>() <span class="sc">+</span> </span>
<span id="cb127-129"><a href="#cb127-129" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb127-130"><a href="#cb127-130" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">500</span>) <span class="sc">+</span></span>
<span id="cb127-131"><a href="#cb127-131" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">250</span>) <span class="sc">+</span></span>
<span id="cb127-132"><a href="#cb127-132" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_wrap</span>(<span class="sc">~</span>samples)</span>
<span id="cb127-133"><a href="#cb127-133" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb127-134"><a href="#cb127-134" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-135"><a href="#cb127-135" aria-hidden="true" tabindex="-1"></a>可以看到该数据集已经进行了质控，各项指标均在正常范围内，因此，我们可以跳过细胞/基因过滤，直接进入下面的环节。</span>
<span id="cb127-136"><a href="#cb127-136" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-137"><a href="#cb127-137" aria-hidden="true" tabindex="-1"></a><span class="fu">### 归一化及消除非期望变异来源</span></span>
<span id="cb127-138"><a href="#cb127-138" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-139"><a href="#cb127-139" aria-hidden="true" tabindex="-1"></a>::: callout-caution</span>
<span id="cb127-140"><a href="#cb127-140" aria-hidden="true" tabindex="-1"></a>由于这里<span class="in">`SCTransform`</span>的数据量太大，在我的16GB MacBook Pro上超出了R内存分配上限，出现了“Error: vector memory exhausted (limit reached?)”报错。因此，这里在运行下面的脚本之前通过 @sec-Raising_memory_limit 的方法进行了处理。</span>
<span id="cb127-141"><a href="#cb127-141" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb127-142"><a href="#cb127-142" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-145"><a href="#cb127-145" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb127-146"><a href="#cb127-146" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb127-147"><a href="#cb127-147" aria-hidden="true" tabindex="-1"></a><span class="co"># SCTranform</span></span>
<span id="cb127-148"><a href="#cb127-148" aria-hidden="true" tabindex="-1"></a>merged_seurat <span class="ot">&lt;-</span> <span class="fu">SCTransform</span>(merged_seurat, <span class="at">verbose =</span> <span class="cn">FALSE</span>); <span class="fu">beep</span>()</span>
<span id="cb127-149"><a href="#cb127-149" aria-hidden="true" tabindex="-1"></a><span class="fu">gc</span>() <span class="co"># 释放未使用内存</span></span>
<span id="cb127-150"><a href="#cb127-150" aria-hidden="true" tabindex="-1"></a>merged_seurat</span>
<span id="cb127-151"><a href="#cb127-151" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-152"><a href="#cb127-152" aria-hidden="true" tabindex="-1"></a><span class="co"># Check which assays are stored in objects</span></span>
<span id="cb127-153"><a href="#cb127-153" aria-hidden="true" tabindex="-1"></a>merged_seurat<span class="sc">@</span>assays</span>
<span id="cb127-154"><a href="#cb127-154" aria-hidden="true" tabindex="-1"></a><span class="co"># 查看目前默认的assay</span></span>
<span id="cb127-155"><a href="#cb127-155" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(merged_seurat)</span>
<span id="cb127-156"><a href="#cb127-156" aria-hidden="true" tabindex="-1"></a><span class="co"># 查看默认assay的layers</span></span>
<span id="cb127-157"><a href="#cb127-157" aria-hidden="true" tabindex="-1"></a><span class="fu">Layers</span>(merged_seurat)</span>
<span id="cb127-158"><a href="#cb127-158" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb127-159"><a href="#cb127-159" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-160"><a href="#cb127-160" aria-hidden="true" tabindex="-1"></a><span class="fu">#### 评估细胞周期的影响</span></span>
<span id="cb127-161"><a href="#cb127-161" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-164"><a href="#cb127-164" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb127-165"><a href="#cb127-165" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb127-166"><a href="#cb127-166" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 10</span></span>
<span id="cb127-167"><a href="#cb127-167" aria-hidden="true" tabindex="-1"></a><span class="co"># Load cell cycle markers</span></span>
<span id="cb127-168"><a href="#cb127-168" aria-hidden="true" tabindex="-1"></a>s.genes <span class="ot">&lt;-</span> cc.genes<span class="sc">$</span>s.genes</span>
<span id="cb127-169"><a href="#cb127-169" aria-hidden="true" tabindex="-1"></a>g2m.genes <span class="ot">&lt;-</span> cc.genes<span class="sc">$</span>g2m.genes</span>
<span id="cb127-170"><a href="#cb127-170" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-171"><a href="#cb127-171" aria-hidden="true" tabindex="-1"></a><span class="co"># Score cells for cell cycle</span></span>
<span id="cb127-172"><a href="#cb127-172" aria-hidden="true" tabindex="-1"></a>merged_seurat <span class="ot">&lt;-</span> <span class="fu">CellCycleScoring</span>(merged_seurat, </span>
<span id="cb127-173"><a href="#cb127-173" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">g2m.features =</span> g2m.genes, </span>
<span id="cb127-174"><a href="#cb127-174" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">s.features =</span> s.genes)</span>
<span id="cb127-175"><a href="#cb127-175" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-176"><a href="#cb127-176" aria-hidden="true" tabindex="-1"></a><span class="co"># 现在的meta.data中多出了细胞周期评分“S.Score”和“G2M.Score”，以及推断的细胞所处的周期“Phase”</span></span>
<span id="cb127-177"><a href="#cb127-177" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(merged_seurat<span class="sc">@</span>meta.data)</span>
<span id="cb127-178"><a href="#cb127-178" aria-hidden="true" tabindex="-1"></a><span class="co"># 查看一下细胞周期的分布情况</span></span>
<span id="cb127-179"><a href="#cb127-179" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(merged_seurat<span class="sc">$</span>Phase)</span>
<span id="cb127-180"><a href="#cb127-180" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-181"><a href="#cb127-181" aria-hidden="true" tabindex="-1"></a><span class="co"># 执行PCA</span></span>
<span id="cb127-182"><a href="#cb127-182" aria-hidden="true" tabindex="-1"></a>merged_seurat <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(merged_seurat)</span>
<span id="cb127-183"><a href="#cb127-183" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-184"><a href="#cb127-184" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the PCA colored by cell cycle phase</span></span>
<span id="cb127-185"><a href="#cb127-185" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(merged_seurat,</span>
<span id="cb127-186"><a href="#cb127-186" aria-hidden="true" tabindex="-1"></a>              <span class="at">reduction =</span> <span class="st">"pca"</span>,</span>
<span id="cb127-187"><a href="#cb127-187" aria-hidden="true" tabindex="-1"></a>              <span class="at">group.by=</span> <span class="st">"Phase"</span>)</span>
<span id="cb127-188"><a href="#cb127-188" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(merged_seurat,</span>
<span id="cb127-189"><a href="#cb127-189" aria-hidden="true" tabindex="-1"></a>              <span class="at">reduction =</span> <span class="st">"pca"</span>,</span>
<span id="cb127-190"><a href="#cb127-190" aria-hidden="true" tabindex="-1"></a>              <span class="at">group.by=</span> <span class="st">"Phase"</span>,</span>
<span id="cb127-191"><a href="#cb127-191" aria-hidden="true" tabindex="-1"></a>              <span class="at">split.by =</span> <span class="st">"Phase"</span>)</span>
<span id="cb127-192"><a href="#cb127-192" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_grid</span>(p1, p2, <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">labels =</span> <span class="st">"AUTO"</span>)</span>
<span id="cb127-193"><a href="#cb127-193" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb127-194"><a href="#cb127-194" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-195"><a href="#cb127-195" aria-hidden="true" tabindex="-1"></a>可以看到细胞周期不是变异来源。</span>
<span id="cb127-196"><a href="#cb127-196" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-197"><a href="#cb127-197" aria-hidden="true" tabindex="-1"></a><span class="fu">#### **评估线粒体基因的影响**</span></span>
<span id="cb127-198"><a href="#cb127-198" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-201"><a href="#cb127-201" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb127-202"><a href="#cb127-202" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb127-203"><a href="#cb127-203" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 3</span></span>
<span id="cb127-204"><a href="#cb127-204" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 3</span></span>
<span id="cb127-205"><a href="#cb127-205" aria-hidden="true" tabindex="-1"></a><span class="co"># Check quartile values</span></span>
<span id="cb127-206"><a href="#cb127-206" aria-hidden="true" tabindex="-1"></a>mito_sum <span class="ot">&lt;-</span> <span class="fu">summary</span>(merged_seurat<span class="sc">$</span>mitoRatio)</span>
<span id="cb127-207"><a href="#cb127-207" aria-hidden="true" tabindex="-1"></a>mito_sum</span>
<span id="cb127-208"><a href="#cb127-208" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-209"><a href="#cb127-209" aria-hidden="true" tabindex="-1"></a><span class="co"># Turn mitoRatio into categorical factor vector based on quartile values</span></span>
<span id="cb127-210"><a href="#cb127-210" aria-hidden="true" tabindex="-1"></a>merged_seurat<span class="sc">$</span>mitoFr <span class="ot">&lt;-</span> <span class="fu">cut</span>(merged_seurat<span class="sc">$</span>mitoRatio, </span>
<span id="cb127-211"><a href="#cb127-211" aria-hidden="true" tabindex="-1"></a>                            <span class="at">breaks=</span><span class="fu">c</span>(<span class="sc">-</span><span class="cn">Inf</span>, mito_sum[<span class="dv">2</span>], mito_sum[<span class="dv">3</span>], mito_sum[<span class="dv">5</span>], <span class="cn">Inf</span>),</span>
<span id="cb127-212"><a href="#cb127-212" aria-hidden="true" tabindex="-1"></a>                            <span class="at">labels=</span><span class="fu">c</span>(<span class="st">"Low"</span>, <span class="st">"Medium"</span>, <span class="st">"Medium high"</span>, <span class="st">"High"</span>))</span>
<span id="cb127-213"><a href="#cb127-213" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(merged_seurat<span class="sc">$</span>mitoFr)</span>
<span id="cb127-214"><a href="#cb127-214" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb127-215"><a href="#cb127-215" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-218"><a href="#cb127-218" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb127-219"><a href="#cb127-219" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb127-220"><a href="#cb127-220" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 15</span></span>
<span id="cb127-221"><a href="#cb127-221" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the PCA colored by cell cycle phase</span></span>
<span id="cb127-222"><a href="#cb127-222" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(merged_seurat,</span>
<span id="cb127-223"><a href="#cb127-223" aria-hidden="true" tabindex="-1"></a>              <span class="at">reduction =</span> <span class="st">"pca"</span>,</span>
<span id="cb127-224"><a href="#cb127-224" aria-hidden="true" tabindex="-1"></a>              <span class="at">group.by=</span> <span class="st">"mitoFr"</span>)</span>
<span id="cb127-225"><a href="#cb127-225" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(merged_seurat,</span>
<span id="cb127-226"><a href="#cb127-226" aria-hidden="true" tabindex="-1"></a>              <span class="at">reduction =</span> <span class="st">"pca"</span>,</span>
<span id="cb127-227"><a href="#cb127-227" aria-hidden="true" tabindex="-1"></a>              <span class="at">group.by=</span> <span class="st">"mitoFr"</span>,</span>
<span id="cb127-228"><a href="#cb127-228" aria-hidden="true" tabindex="-1"></a>              <span class="at">split.by =</span> <span class="st">"mitoFr"</span>)</span>
<span id="cb127-229"><a href="#cb127-229" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_grid</span>(p1, p2, <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">labels =</span> <span class="st">"AUTO"</span>)</span>
<span id="cb127-230"><a href="#cb127-230" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb127-231"><a href="#cb127-231" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-232"><a href="#cb127-232" aria-hidden="true" tabindex="-1"></a>可以看到线粒体基因比例不是变异来源。</span>
<span id="cb127-233"><a href="#cb127-233" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-234"><a href="#cb127-234" aria-hidden="true" tabindex="-1"></a>由于细胞周期和线粒体基因比例都不是非期望变异来源，所以我们这里不需要再次运行SCTransform来回归这些变量。接下来，直接进入数据整合环节。</span>
<span id="cb127-235"><a href="#cb127-235" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-236"><a href="#cb127-236" aria-hidden="true" tabindex="-1"></a><span class="fu">### 数据整合</span></span>
<span id="cb127-237"><a href="#cb127-237" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-238"><a href="#cb127-238" aria-hidden="true" tabindex="-1"></a><span class="fu">#### **不进行整合时检验细胞分群情况**</span></span>
<span id="cb127-239"><a href="#cb127-239" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-242"><a href="#cb127-242" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb127-243"><a href="#cb127-243" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb127-244"><a href="#cb127-244" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 15</span></span>
<span id="cb127-245"><a href="#cb127-245" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 15</span></span>
<span id="cb127-246"><a href="#cb127-246" aria-hidden="true" tabindex="-1"></a><span class="co"># 查看降维信息</span></span>
<span id="cb127-247"><a href="#cb127-247" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(merged_seurat<span class="sc">@</span>reductions)</span>
<span id="cb127-248"><a href="#cb127-248" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-249"><a href="#cb127-249" aria-hidden="true" tabindex="-1"></a><span class="co"># Run UMAP</span></span>
<span id="cb127-250"><a href="#cb127-250" aria-hidden="true" tabindex="-1"></a>merged_seurat <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(merged_seurat, </span>
<span id="cb127-251"><a href="#cb127-251" aria-hidden="true" tabindex="-1"></a>                         <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">40</span>, </span>
<span id="cb127-252"><a href="#cb127-252" aria-hidden="true" tabindex="-1"></a>                         <span class="at">reduction =</span> <span class="st">"pca"</span>,</span>
<span id="cb127-253"><a href="#cb127-253" aria-hidden="true" tabindex="-1"></a>                         <span class="at">reduction.name =</span> <span class="st">"umap.unintegrated"</span>); <span class="fu">beep</span>()</span>
<span id="cb127-254"><a href="#cb127-254" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-255"><a href="#cb127-255" aria-hidden="true" tabindex="-1"></a><span class="co"># 分群</span></span>
<span id="cb127-256"><a href="#cb127-256" aria-hidden="true" tabindex="-1"></a><span class="co"># Determine the K-nearest neighbor graph</span></span>
<span id="cb127-257"><a href="#cb127-257" aria-hidden="true" tabindex="-1"></a>merged_seurat <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(merged_seurat, </span>
<span id="cb127-258"><a href="#cb127-258" aria-hidden="true" tabindex="-1"></a>                               <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">40</span>, </span>
<span id="cb127-259"><a href="#cb127-259" aria-hidden="true" tabindex="-1"></a>                               <span class="at">reduction =</span> <span class="st">"pca"</span>)</span>
<span id="cb127-260"><a href="#cb127-260" aria-hidden="true" tabindex="-1"></a>merged_seurat <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(merged_seurat, </span>
<span id="cb127-261"><a href="#cb127-261" aria-hidden="true" tabindex="-1"></a>                              <span class="at">cluster.name =</span> <span class="st">"unintegrated_clusters"</span>)</span>
<span id="cb127-262"><a href="#cb127-262" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-263"><a href="#cb127-263" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot UMAP</span></span>
<span id="cb127-264"><a href="#cb127-264" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(merged_seurat, </span>
<span id="cb127-265"><a href="#cb127-265" aria-hidden="true" tabindex="-1"></a>              <span class="at">reduction =</span> <span class="st">"umap.unintegrated"</span>,</span>
<span id="cb127-266"><a href="#cb127-266" aria-hidden="true" tabindex="-1"></a>              <span class="at">group.by =</span> <span class="st">"samples"</span>)</span>
<span id="cb127-267"><a href="#cb127-267" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(merged_seurat, </span>
<span id="cb127-268"><a href="#cb127-268" aria-hidden="true" tabindex="-1"></a>              <span class="at">reduction =</span> <span class="st">"umap.unintegrated"</span>,</span>
<span id="cb127-269"><a href="#cb127-269" aria-hidden="true" tabindex="-1"></a>              <span class="at">split.by =</span> <span class="st">"samples"</span>)</span>
<span id="cb127-270"><a href="#cb127-270" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_grid</span>(p1, p2, </span>
<span id="cb127-271"><a href="#cb127-271" aria-hidden="true" tabindex="-1"></a>          <span class="at">ncol =</span> <span class="dv">1</span>, <span class="at">labels =</span> <span class="st">"AUTO"</span>)</span>
<span id="cb127-272"><a href="#cb127-272" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-273"><a href="#cb127-273" aria-hidden="true" tabindex="-1"></a><span class="co"># 由于样本数较多，我们再按照样本类型“Normal” vs. “Cancer”画一下UMAP图</span></span>
<span id="cb127-274"><a href="#cb127-274" aria-hidden="true" tabindex="-1"></a>merged_seurat<span class="sc">$</span>groups <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(merged_seurat<span class="sc">$</span>samples <span class="sc">==</span> <span class="st">"N01"</span>, <span class="st">"Normal"</span>, <span class="st">"Cancer"</span>)</span>
<span id="cb127-275"><a href="#cb127-275" aria-hidden="true" tabindex="-1"></a>p1a <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(merged_seurat, </span>
<span id="cb127-276"><a href="#cb127-276" aria-hidden="true" tabindex="-1"></a>               <span class="at">reduction =</span> <span class="st">"umap.unintegrated"</span>,</span>
<span id="cb127-277"><a href="#cb127-277" aria-hidden="true" tabindex="-1"></a>               <span class="at">group.by =</span> <span class="st">"groups"</span>)</span>
<span id="cb127-278"><a href="#cb127-278" aria-hidden="true" tabindex="-1"></a>p2a <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(merged_seurat, </span>
<span id="cb127-279"><a href="#cb127-279" aria-hidden="true" tabindex="-1"></a>               <span class="at">reduction =</span> <span class="st">"umap.unintegrated"</span>,</span>
<span id="cb127-280"><a href="#cb127-280" aria-hidden="true" tabindex="-1"></a>               <span class="at">split.by =</span> <span class="st">"groups"</span>)</span>
<span id="cb127-281"><a href="#cb127-281" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_grid</span>(p1a, p2a, </span>
<span id="cb127-282"><a href="#cb127-282" aria-hidden="true" tabindex="-1"></a>          <span class="at">ncol =</span> <span class="dv">1</span>, <span class="at">labels =</span> <span class="st">"AUTO"</span>)</span>
<span id="cb127-283"><a href="#cb127-283" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb127-284"><a href="#cb127-284" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-285"><a href="#cb127-285" aria-hidden="true" tabindex="-1"></a><span class="fu">#### 整合</span></span>
<span id="cb127-286"><a href="#cb127-286" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-287"><a href="#cb127-287" aria-hidden="true" tabindex="-1"></a>这里我们用Harmony整合算法。</span>
<span id="cb127-288"><a href="#cb127-288" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-291"><a href="#cb127-291" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb127-292"><a href="#cb127-292" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb127-293"><a href="#cb127-293" aria-hidden="true" tabindex="-1"></a><span class="co"># 整合，比较耗时间，进度条会一直显示0%直至运算完成</span></span>
<span id="cb127-294"><a href="#cb127-294" aria-hidden="true" tabindex="-1"></a>seurat_integrated <span class="ot">&lt;-</span> <span class="fu">IntegrateLayers</span>(<span class="at">object =</span> merged_seurat,</span>
<span id="cb127-295"><a href="#cb127-295" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">method =</span> HarmonyIntegration,</span>
<span id="cb127-296"><a href="#cb127-296" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">assay =</span> <span class="st">"SCT"</span>, <span class="co"># Integrating SCTransformed data</span></span>
<span id="cb127-297"><a href="#cb127-297" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">orig.reduction =</span> <span class="st">"pca"</span>,</span>
<span id="cb127-298"><a href="#cb127-298" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">verbose =</span> <span class="cn">FALSE</span>); <span class="fu">beep</span>()</span>
<span id="cb127-299"><a href="#cb127-299" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-300"><a href="#cb127-300" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-301"><a href="#cb127-301" aria-hidden="true" tabindex="-1"></a><span class="co"># 整合后合并RNA layer</span></span>
<span id="cb127-302"><a href="#cb127-302" aria-hidden="true" tabindex="-1"></a>seurat_integrated[[<span class="st">"RNA"</span>]] <span class="ot">&lt;-</span> <span class="fu">JoinLayers</span>(seurat_integrated[[<span class="st">"RNA"</span>]])</span>
<span id="cb127-303"><a href="#cb127-303" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-304"><a href="#cb127-304" aria-hidden="true" tabindex="-1"></a><span class="co"># 查看整合后的降维信息</span></span>
<span id="cb127-305"><a href="#cb127-305" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(seurat_integrated<span class="sc">@</span>reductions)</span>
<span id="cb127-306"><a href="#cb127-306" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb127-307"><a href="#cb127-307" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-308"><a href="#cb127-308" aria-hidden="true" tabindex="-1"></a><span class="fu">#### **整合后检验细胞分群情况**</span></span>
<span id="cb127-309"><a href="#cb127-309" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-312"><a href="#cb127-312" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb127-313"><a href="#cb127-313" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb127-314"><a href="#cb127-314" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 10</span></span>
<span id="cb127-315"><a href="#cb127-315" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 20</span></span>
<span id="cb127-316"><a href="#cb127-316" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123456</span>)</span>
<span id="cb127-317"><a href="#cb127-317" aria-hidden="true" tabindex="-1"></a><span class="co"># Run UMAP</span></span>
<span id="cb127-318"><a href="#cb127-318" aria-hidden="true" tabindex="-1"></a>seurat_integrated <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(seurat_integrated, </span>
<span id="cb127-319"><a href="#cb127-319" aria-hidden="true" tabindex="-1"></a>                             <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">40</span>,</span>
<span id="cb127-320"><a href="#cb127-320" aria-hidden="true" tabindex="-1"></a>                             <span class="at">reduction =</span> <span class="st">"harmony"</span>, <span class="co"># 更改降维来源为整合后的"harmony"</span></span>
<span id="cb127-321"><a href="#cb127-321" aria-hidden="true" tabindex="-1"></a>                             <span class="at">reduction.name =</span> <span class="st">"umap.integrated"</span>); <span class="fu">beep</span>()</span>
<span id="cb127-322"><a href="#cb127-322" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(seurat_integrated<span class="sc">@</span>reductions)</span>
<span id="cb127-323"><a href="#cb127-323" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-324"><a href="#cb127-324" aria-hidden="true" tabindex="-1"></a><span class="co"># 分群</span></span>
<span id="cb127-325"><a href="#cb127-325" aria-hidden="true" tabindex="-1"></a>seurat_integrated <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(seurat_integrated, </span>
<span id="cb127-326"><a href="#cb127-326" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">40</span>, </span>
<span id="cb127-327"><a href="#cb127-327" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">reduction =</span> <span class="st">"harmony"</span>) <span class="co">#更改降维来源为"harmony"</span></span>
<span id="cb127-328"><a href="#cb127-328" aria-hidden="true" tabindex="-1"></a>seurat_integrated <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(seurat_integrated, </span>
<span id="cb127-329"><a href="#cb127-329" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">cluster.name =</span> <span class="st">"integrated_clusters"</span>)</span>
<span id="cb127-330"><a href="#cb127-330" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(seurat_integrated<span class="sc">@</span>meta.data)</span>
<span id="cb127-331"><a href="#cb127-331" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-332"><a href="#cb127-332" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot UMAP                             </span></span>
<span id="cb127-333"><a href="#cb127-333" aria-hidden="true" tabindex="-1"></a>p3 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(seurat_integrated, </span>
<span id="cb127-334"><a href="#cb127-334" aria-hidden="true" tabindex="-1"></a>              <span class="at">reduction =</span> <span class="st">"umap.integrated"</span>, </span>
<span id="cb127-335"><a href="#cb127-335" aria-hidden="true" tabindex="-1"></a>              <span class="at">group.by =</span> <span class="st">"samples"</span>)</span>
<span id="cb127-336"><a href="#cb127-336" aria-hidden="true" tabindex="-1"></a>p4 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(seurat_integrated, </span>
<span id="cb127-337"><a href="#cb127-337" aria-hidden="true" tabindex="-1"></a>              <span class="at">reduction =</span> <span class="st">"umap.integrated"</span>, </span>
<span id="cb127-338"><a href="#cb127-338" aria-hidden="true" tabindex="-1"></a>              <span class="at">split.by =</span> <span class="st">"samples"</span>)</span>
<span id="cb127-339"><a href="#cb127-339" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_grid</span>(p1, p3, p2, p4, </span>
<span id="cb127-340"><a href="#cb127-340" aria-hidden="true" tabindex="-1"></a>          <span class="at">ncol =</span> <span class="dv">1</span>, </span>
<span id="cb127-341"><a href="#cb127-341" aria-hidden="true" tabindex="-1"></a>          <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">"Before Harmony"</span>, <span class="st">"After Harmony"</span>, </span>
<span id="cb127-342"><a href="#cb127-342" aria-hidden="true" tabindex="-1"></a>                     <span class="st">"Before Harmony"</span>, <span class="st">"After Harmony"</span>))</span>
<span id="cb127-343"><a href="#cb127-343" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-344"><a href="#cb127-344" aria-hidden="true" tabindex="-1"></a><span class="co"># 保存</span></span>
<span id="cb127-345"><a href="#cb127-345" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(seurat_integrated, </span>
<span id="cb127-346"><a href="#cb127-346" aria-hidden="true" tabindex="-1"></a>        <span class="at">file =</span> <span class="st">"output/sc_supplementary/GSE150430_seurat_integrated.RDS"</span>); <span class="fu">beep</span>()</span>
<span id="cb127-347"><a href="#cb127-347" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb127-348"><a href="#cb127-348" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-349"><a href="#cb127-349" aria-hidden="true" tabindex="-1"></a>------------------------------------------------------------------------</span>
<span id="cb127-350"><a href="#cb127-350" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-351"><a href="#cb127-351" aria-hidden="true" tabindex="-1"></a><span class="fu">### 聚类</span></span>
<span id="cb127-352"><a href="#cb127-352" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-355"><a href="#cb127-355" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb127-356"><a href="#cb127-356" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb127-357"><a href="#cb127-357" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(<span class="at">list =</span> <span class="fu">ls</span>())</span>
<span id="cb127-358"><a href="#cb127-358" aria-hidden="true" tabindex="-1"></a><span class="fu">gc</span>()</span>
<span id="cb127-359"><a href="#cb127-359" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-360"><a href="#cb127-360" aria-hidden="true" tabindex="-1"></a>seurat_integrated <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"output/sc_supplementary/GSE150430_seurat_integrated.RDS"</span>)</span>
<span id="cb127-361"><a href="#cb127-361" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb127-362"><a href="#cb127-362" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-365"><a href="#cb127-365" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb127-366"><a href="#cb127-366" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb127-367"><a href="#cb127-367" aria-hidden="true" tabindex="-1"></a><span class="co"># Determine the clusters for various resolutions                                </span></span>
<span id="cb127-368"><a href="#cb127-368" aria-hidden="true" tabindex="-1"></a>seurat_integrated <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(seurat_integrated,</span>
<span id="cb127-369"><a href="#cb127-369" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">resolution =</span> <span class="fu">c</span>(<span class="fl">0.01</span>, <span class="fl">0.05</span>, <span class="fl">0.1</span>, <span class="fl">0.2</span>, <span class="fl">0.3</span>, <span class="fl">0.4</span>, <span class="fl">0.5</span>, <span class="fl">0.8</span>, <span class="dv">1</span>),</span>
<span id="cb127-370"><a href="#cb127-370" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">verbose =</span> F); <span class="fu">beep</span>()</span>
<span id="cb127-371"><a href="#cb127-371" aria-hidden="true" tabindex="-1"></a><span class="co"># Explore resolutions</span></span>
<span id="cb127-372"><a href="#cb127-372" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(seurat_integrated<span class="sc">@</span>meta.data, <span class="dv">5</span>)</span>
<span id="cb127-373"><a href="#cb127-373" aria-hidden="true" tabindex="-1"></a><span class="co"># 查看各个分辨率下的细胞分群情况</span></span>
<span id="cb127-374"><a href="#cb127-374" aria-hidden="true" tabindex="-1"></a><span class="fu">select</span>(seurat_integrated<span class="sc">@</span>meta.data, <span class="fu">starts_with</span>(<span class="at">match =</span> <span class="st">"SCT_snn_res."</span>)) <span class="sc">%&gt;%</span>  </span>
<span id="cb127-375"><a href="#cb127-375" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lapply</span>(levels)</span>
<span id="cb127-376"><a href="#cb127-376" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb127-377"><a href="#cb127-377" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-378"><a href="#cb127-378" aria-hidden="true" tabindex="-1"></a>绘制聚类树展示不同分辨率下的细胞分群情况及相互关系</span>
<span id="cb127-379"><a href="#cb127-379" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-382"><a href="#cb127-382" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb127-383"><a href="#cb127-383" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb127-384"><a href="#cb127-384" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 10</span></span>
<span id="cb127-385"><a href="#cb127-385" aria-hidden="true" tabindex="-1"></a>tree <span class="ot">&lt;-</span> <span class="fu">clustree</span>(seurat_integrated<span class="sc">@</span>meta.data, </span>
<span id="cb127-386"><a href="#cb127-386" aria-hidden="true" tabindex="-1"></a>                 <span class="at">prefix =</span> <span class="st">"SCT_snn_res."</span>)<span class="co">#指定包含聚类信息的列</span></span>
<span id="cb127-387"><a href="#cb127-387" aria-hidden="true" tabindex="-1"></a>tree</span>
<span id="cb127-388"><a href="#cb127-388" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb127-389"><a href="#cb127-389" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-390"><a href="#cb127-390" aria-hidden="true" tabindex="-1"></a>通常情况下，为了决定合适的聚类分辨率，可使用以下两种策略：</span>
<span id="cb127-391"><a href="#cb127-391" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-392"><a href="#cb127-392" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>选择透明箭头出现较少的分辨率聚类结果</span>
<span id="cb127-393"><a href="#cb127-393" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-394"><a href="#cb127-394" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>基于marker基因表达选择有生物学意义的分辨率聚类结果</span>
<span id="cb127-395"><a href="#cb127-395" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-396"><a href="#cb127-396" aria-hidden="true" tabindex="-1"></a>接下来分析，按照分辨率为0.5进行</span>
<span id="cb127-397"><a href="#cb127-397" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-400"><a href="#cb127-400" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb127-401"><a href="#cb127-401" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb127-402"><a href="#cb127-402" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(seurat_integrated) <span class="ot">&lt;-</span> <span class="st">"SCT_snn_res.0.5"</span></span>
<span id="cb127-403"><a href="#cb127-403" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb127-404"><a href="#cb127-404" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-405"><a href="#cb127-405" aria-hidden="true" tabindex="-1"></a>聚类可视化</span>
<span id="cb127-406"><a href="#cb127-406" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-409"><a href="#cb127-409" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb127-410"><a href="#cb127-410" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb127-411"><a href="#cb127-411" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the UMAP</span></span>
<span id="cb127-412"><a href="#cb127-412" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(seurat_integrated,</span>
<span id="cb127-413"><a href="#cb127-413" aria-hidden="true" tabindex="-1"></a>        <span class="at">reduction =</span> <span class="st">"umap.integrated"</span>,</span>
<span id="cb127-414"><a href="#cb127-414" aria-hidden="true" tabindex="-1"></a>        <span class="at">label =</span> <span class="cn">FALSE</span>)</span>
<span id="cb127-415"><a href="#cb127-415" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb127-416"><a href="#cb127-416" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-417"><a href="#cb127-417" aria-hidden="true" tabindex="-1"></a><span class="fu">### **细胞分群质量评估**</span></span>
<span id="cb127-418"><a href="#cb127-418" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-419"><a href="#cb127-419" aria-hidden="true" tabindex="-1"></a><span class="fu">#### **分析样本类型是否影响细胞分群**</span></span>
<span id="cb127-420"><a href="#cb127-420" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-423"><a href="#cb127-423" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb127-424"><a href="#cb127-424" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb127-425"><a href="#cb127-425" aria-hidden="true" tabindex="-1"></a><span class="co"># 先简单查看不同cluster的细胞数</span></span>
<span id="cb127-426"><a href="#cb127-426" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(seurat_integrated<span class="sc">@</span>active.ident)</span>
<span id="cb127-427"><a href="#cb127-427" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-428"><a href="#cb127-428" aria-hidden="true" tabindex="-1"></a><span class="co"># 查看不同样本类型中的细胞分群情况</span></span>
<span id="cb127-429"><a href="#cb127-429" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(seurat_integrated, </span>
<span id="cb127-430"><a href="#cb127-430" aria-hidden="true" tabindex="-1"></a>        <span class="at">reduction =</span> <span class="st">"umap.integrated"</span>,</span>
<span id="cb127-431"><a href="#cb127-431" aria-hidden="true" tabindex="-1"></a>        <span class="at">label =</span> <span class="cn">TRUE</span>, </span>
<span id="cb127-432"><a href="#cb127-432" aria-hidden="true" tabindex="-1"></a>        <span class="at">split.by =</span> <span class="st">"groups"</span>) <span class="sc">+</span> </span>
<span id="cb127-433"><a href="#cb127-433" aria-hidden="true" tabindex="-1"></a>  <span class="fu">NoLegend</span>()</span>
<span id="cb127-434"><a href="#cb127-434" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb127-435"><a href="#cb127-435" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-436"><a href="#cb127-436" aria-hidden="true" tabindex="-1"></a><span class="fu">#### **分析细胞周期是否影响细胞分群**</span></span>
<span id="cb127-437"><a href="#cb127-437" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-440"><a href="#cb127-440" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb127-441"><a href="#cb127-441" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb127-442"><a href="#cb127-442" aria-hidden="true" tabindex="-1"></a><span class="co"># Explore whether clusters segregate by cell cycle phase</span></span>
<span id="cb127-443"><a href="#cb127-443" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(seurat_integrated,</span>
<span id="cb127-444"><a href="#cb127-444" aria-hidden="true" tabindex="-1"></a>        <span class="at">reduction =</span> <span class="st">"umap.integrated"</span>,</span>
<span id="cb127-445"><a href="#cb127-445" aria-hidden="true" tabindex="-1"></a>        <span class="at">label =</span> <span class="cn">TRUE</span>, </span>
<span id="cb127-446"><a href="#cb127-446" aria-hidden="true" tabindex="-1"></a>        <span class="at">split.by =</span> <span class="st">"Phase"</span>) <span class="sc">+</span> </span>
<span id="cb127-447"><a href="#cb127-447" aria-hidden="true" tabindex="-1"></a>  <span class="fu">NoLegend</span>()</span>
<span id="cb127-448"><a href="#cb127-448" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb127-449"><a href="#cb127-449" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-450"><a href="#cb127-450" aria-hidden="true" tabindex="-1"></a><span class="fu">#### **分析其他非期望变异来源是否会影响细胞分群**</span></span>
<span id="cb127-451"><a href="#cb127-451" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-454"><a href="#cb127-454" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb127-455"><a href="#cb127-455" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb127-456"><a href="#cb127-456" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 10</span></span>
<span id="cb127-457"><a href="#cb127-457" aria-hidden="true" tabindex="-1"></a><span class="co"># Determine metrics to plot present in seurat_clustered@meta.data</span></span>
<span id="cb127-458"><a href="#cb127-458" aria-hidden="true" tabindex="-1"></a>metrics <span class="ot">&lt;-</span>  <span class="fu">c</span>(<span class="st">"nCount_RNA"</span>, <span class="st">"nFeature_RNA"</span>, <span class="st">"S.Score"</span>, <span class="st">"G2M.Score"</span>, <span class="st">"mitoRatio"</span>)</span>
<span id="cb127-459"><a href="#cb127-459" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-460"><a href="#cb127-460" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>(seurat_integrated, </span>
<span id="cb127-461"><a href="#cb127-461" aria-hidden="true" tabindex="-1"></a>            <span class="at">reduction =</span> <span class="st">"umap.integrated"</span>, </span>
<span id="cb127-462"><a href="#cb127-462" aria-hidden="true" tabindex="-1"></a>            <span class="at">features =</span> metrics,</span>
<span id="cb127-463"><a href="#cb127-463" aria-hidden="true" tabindex="-1"></a>            <span class="at">pt.size =</span> <span class="fl">0.4</span>, </span>
<span id="cb127-464"><a href="#cb127-464" aria-hidden="true" tabindex="-1"></a>            <span class="at">order =</span> <span class="cn">TRUE</span>,</span>
<span id="cb127-465"><a href="#cb127-465" aria-hidden="true" tabindex="-1"></a>            <span class="at">min.cutoff =</span> <span class="st">'q10'</span>, </span>
<span id="cb127-466"><a href="#cb127-466" aria-hidden="true" tabindex="-1"></a>            <span class="at">label =</span> <span class="cn">TRUE</span>)</span>
<span id="cb127-467"><a href="#cb127-467" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb127-468"><a href="#cb127-468" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-469"><a href="#cb127-469" aria-hidden="true" tabindex="-1"></a>保存</span>
<span id="cb127-470"><a href="#cb127-470" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-473"><a href="#cb127-473" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb127-474"><a href="#cb127-474" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb127-475"><a href="#cb127-475" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(seurat_integrated, </span>
<span id="cb127-476"><a href="#cb127-476" aria-hidden="true" tabindex="-1"></a>        <span class="at">file =</span> <span class="st">"output/sc_supplementary/GSE150430_seurat_clustered.RDS"</span>); <span class="fu">beep</span>()</span>
<span id="cb127-477"><a href="#cb127-477" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb127-478"><a href="#cb127-478" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-479"><a href="#cb127-479" aria-hidden="true" tabindex="-1"></a>------------------------------------------------------------------------</span>
<span id="cb127-480"><a href="#cb127-480" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-481"><a href="#cb127-481" aria-hidden="true" tabindex="-1"></a><span class="fu">### 细胞注释</span></span>
<span id="cb127-482"><a href="#cb127-482" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-483"><a href="#cb127-483" aria-hidden="true" tabindex="-1"></a><span class="fu">#### **探索cell type markers的表达**</span></span>
<span id="cb127-484"><a href="#cb127-484" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-485"><a href="#cb127-485" aria-hidden="true" tabindex="-1"></a>这里的marker基因选用 @sec-Universal_Marker_for_initial_Cell_clustering 中的细胞初步分群通用marker。</span>
<span id="cb127-486"><a href="#cb127-486" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-489"><a href="#cb127-489" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb127-490"><a href="#cb127-490" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb127-491"><a href="#cb127-491" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 8</span></span>
<span id="cb127-492"><a href="#cb127-492" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 8</span></span>
<span id="cb127-493"><a href="#cb127-493" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(<span class="at">list =</span> <span class="fu">ls</span>())</span>
<span id="cb127-494"><a href="#cb127-494" aria-hidden="true" tabindex="-1"></a><span class="fu">gc</span>()</span>
<span id="cb127-495"><a href="#cb127-495" aria-hidden="true" tabindex="-1"></a>seurat_clustered <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"output/sc_supplementary/GSE150430_seurat_clustered.RDS"</span>)</span>
<span id="cb127-496"><a href="#cb127-496" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-497"><a href="#cb127-497" aria-hidden="true" tabindex="-1"></a>genes_to_check <span class="ot">=</span> <span class="fu">c</span>(<span class="st">'PTPRC'</span>, </span>
<span id="cb127-498"><a href="#cb127-498" aria-hidden="true" tabindex="-1"></a>                   <span class="st">"CD163"</span>,<span class="st">"AIF1"</span>, </span>
<span id="cb127-499"><a href="#cb127-499" aria-hidden="true" tabindex="-1"></a>                   <span class="st">'CD3D'</span>, <span class="st">'CD3E'</span>, <span class="st">'CD4'</span>, <span class="st">'CD8A'</span>, </span>
<span id="cb127-500"><a href="#cb127-500" aria-hidden="true" tabindex="-1"></a>                   <span class="st">'CD19'</span>, <span class="st">'CD79A'</span>, <span class="st">'MS4A1'</span>, <span class="st">"SDC1"</span>, <span class="st">"CD27"</span>, <span class="st">"CD38"</span>,</span>
<span id="cb127-501"><a href="#cb127-501" aria-hidden="true" tabindex="-1"></a>                   <span class="st">'IGHG1'</span>, <span class="st">'MZB1'</span>, <span class="st">'SDC1'</span>, <span class="st">"JCHAIN"</span>, </span>
<span id="cb127-502"><a href="#cb127-502" aria-hidden="true" tabindex="-1"></a>                   <span class="st">'CD68'</span>, <span class="st">'CD163'</span>, <span class="st">'CD14'</span>, </span>
<span id="cb127-503"><a href="#cb127-503" aria-hidden="true" tabindex="-1"></a>                   <span class="st">'S100A9'</span>, <span class="st">'S100A8'</span>, <span class="st">'MMP19'</span>,</span>
<span id="cb127-504"><a href="#cb127-504" aria-hidden="true" tabindex="-1"></a>                   <span class="st">'C1QA'</span>,  <span class="st">'C1QB'</span>, </span>
<span id="cb127-505"><a href="#cb127-505" aria-hidden="true" tabindex="-1"></a>                   <span class="st">'TPSAB1'</span>, <span class="st">'TPSB2'</span>, </span>
<span id="cb127-506"><a href="#cb127-506" aria-hidden="true" tabindex="-1"></a>                   <span class="st">'KLRB1'</span>, <span class="st">"KLRD1"</span>, <span class="st">'NCR1'</span>, <span class="st">"GNLY"</span>, <span class="st">"NKG7"</span>, </span>
<span id="cb127-507"><a href="#cb127-507" aria-hidden="true" tabindex="-1"></a>                   <span class="st">'FGF7'</span>, <span class="st">'MME'</span>, <span class="st">'ACTA2'</span>, <span class="st">"COL3A1"</span>, </span>
<span id="cb127-508"><a href="#cb127-508" aria-hidden="true" tabindex="-1"></a>                   <span class="st">'PECAM1'</span>, <span class="st">'VWF'</span>, </span>
<span id="cb127-509"><a href="#cb127-509" aria-hidden="true" tabindex="-1"></a>                   <span class="st">'EPCAM'</span>, <span class="st">'KRT19'</span>, <span class="st">'PROM1'</span>, <span class="st">'ALDH1A1'</span>) <span class="sc">|&gt;</span> <span class="fu">unique</span>()</span>
<span id="cb127-510"><a href="#cb127-510" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-511"><a href="#cb127-511" aria-hidden="true" tabindex="-1"></a><span class="fu">DotPlot</span>(seurat_clustered, </span>
<span id="cb127-512"><a href="#cb127-512" aria-hidden="true" tabindex="-1"></a>        <span class="at">features =</span> genes_to_check) <span class="sc">+</span> </span>
<span id="cb127-513"><a href="#cb127-513" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_flip</span>()</span>
<span id="cb127-514"><a href="#cb127-514" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb127-515"><a href="#cb127-515" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-516"><a href="#cb127-516" aria-hidden="true" tabindex="-1"></a><span class="fu">#### 手动注释</span></span>
<span id="cb127-517"><a href="#cb127-517" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-520"><a href="#cb127-520" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb127-521"><a href="#cb127-521" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb127-522"><a href="#cb127-522" aria-hidden="true" tabindex="-1"></a>seurat_clustered <span class="ot">&lt;-</span> <span class="fu">RenameIdents</span>(seurat_clustered, </span>
<span id="cb127-523"><a href="#cb127-523" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"0"</span> <span class="ot">=</span> <span class="st">"B cells"</span>,</span>
<span id="cb127-524"><a href="#cb127-524" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"1"</span> <span class="ot">=</span> <span class="st">"T cells"</span>,</span>
<span id="cb127-525"><a href="#cb127-525" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"2"</span> <span class="ot">=</span> <span class="st">"Epithelial/cancer"</span>,</span>
<span id="cb127-526"><a href="#cb127-526" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"3"</span> <span class="ot">=</span> <span class="st">"T cells"</span>,</span>
<span id="cb127-527"><a href="#cb127-527" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"4"</span> <span class="ot">=</span> <span class="st">"T cells"</span>,</span>
<span id="cb127-528"><a href="#cb127-528" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"5"</span> <span class="ot">=</span> <span class="st">"Macrophages"</span>,</span>
<span id="cb127-529"><a href="#cb127-529" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"6"</span> <span class="ot">=</span> <span class="st">"T cells"</span>,</span>
<span id="cb127-530"><a href="#cb127-530" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"7"</span> <span class="ot">=</span> <span class="st">"Plasma cells"</span>,</span>
<span id="cb127-531"><a href="#cb127-531" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"8"</span> <span class="ot">=</span> <span class="st">"B cells"</span>,</span>
<span id="cb127-532"><a href="#cb127-532" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"9"</span> <span class="ot">=</span> <span class="st">"Epithelial/cancer"</span>,</span>
<span id="cb127-533"><a href="#cb127-533" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"10"</span> <span class="ot">=</span> <span class="st">"B cells"</span>,</span>
<span id="cb127-534"><a href="#cb127-534" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"11"</span> <span class="ot">=</span> <span class="st">"B cells"</span>,</span>
<span id="cb127-535"><a href="#cb127-535" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"12"</span> <span class="ot">=</span> <span class="st">"Unknown"</span>,</span>
<span id="cb127-536"><a href="#cb127-536" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"13"</span> <span class="ot">=</span> <span class="st">"NK cells"</span>,</span>
<span id="cb127-537"><a href="#cb127-537" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"14"</span> <span class="ot">=</span> <span class="st">"Unknown"</span>,</span>
<span id="cb127-538"><a href="#cb127-538" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"15"</span> <span class="ot">=</span> <span class="st">"Epithelial/cancer"</span>,</span>
<span id="cb127-539"><a href="#cb127-539" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"16"</span> <span class="ot">=</span> <span class="st">"Myeloid (unspecific)"</span>,</span>
<span id="cb127-540"><a href="#cb127-540" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"17"</span> <span class="ot">=</span> <span class="st">"Epithelial/cancer"</span>,</span>
<span id="cb127-541"><a href="#cb127-541" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"18"</span> <span class="ot">=</span> <span class="st">"Epithelial/cancer"</span>,</span>
<span id="cb127-542"><a href="#cb127-542" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"19"</span> <span class="ot">=</span> <span class="st">"Epithelial/cancer"</span>,</span>
<span id="cb127-543"><a href="#cb127-543" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"20"</span> <span class="ot">=</span> <span class="st">"B cells"</span>)</span>
<span id="cb127-544"><a href="#cb127-544" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="fu">Idents</span>(seurat_clustered))</span>
<span id="cb127-545"><a href="#cb127-545" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-546"><a href="#cb127-546" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the UMAP</span></span>
<span id="cb127-547"><a href="#cb127-547" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(seurat_clustered, </span>
<span id="cb127-548"><a href="#cb127-548" aria-hidden="true" tabindex="-1"></a>        <span class="at">reduction =</span> <span class="st">"umap.integrated"</span>, </span>
<span id="cb127-549"><a href="#cb127-549" aria-hidden="true" tabindex="-1"></a>        <span class="at">label =</span> T)</span>
<span id="cb127-550"><a href="#cb127-550" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb127-551"><a href="#cb127-551" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-552"><a href="#cb127-552" aria-hidden="true" tabindex="-1"></a>在注释好的数据中再次检查marker基因的表达情况：</span>
<span id="cb127-553"><a href="#cb127-553" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-556"><a href="#cb127-556" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb127-557"><a href="#cb127-557" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb127-558"><a href="#cb127-558" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 8</span></span>
<span id="cb127-559"><a href="#cb127-559" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 10</span></span>
<span id="cb127-560"><a href="#cb127-560" aria-hidden="true" tabindex="-1"></a><span class="fu">DotPlot</span>(seurat_clustered, </span>
<span id="cb127-561"><a href="#cb127-561" aria-hidden="true" tabindex="-1"></a>        <span class="at">features =</span> genes_to_check) <span class="sc">+</span> </span>
<span id="cb127-562"><a href="#cb127-562" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_flip</span>() <span class="sc">+</span></span>
<span id="cb127-563"><a href="#cb127-563" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">30</span>,<span class="at">vjust =</span> <span class="fl">0.85</span>,<span class="at">hjust =</span> <span class="fl">0.75</span>)</span>
<span id="cb127-564"><a href="#cb127-564" aria-hidden="true" tabindex="-1"></a>) </span>
<span id="cb127-565"><a href="#cb127-565" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb127-566"><a href="#cb127-566" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-567"><a href="#cb127-567" aria-hidden="true" tabindex="-1"></a>因为接下来我们要提取髓系细胞进一步分析，所以通过几种可视化方法检查一下髓系细胞通用marker（CD163和AIF1）的表达情况：</span>
<span id="cb127-568"><a href="#cb127-568" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-571"><a href="#cb127-571" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb127-572"><a href="#cb127-572" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb127-573"><a href="#cb127-573" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 8</span></span>
<span id="cb127-574"><a href="#cb127-574" aria-hidden="true" tabindex="-1"></a><span class="fu">RidgePlot</span>(seurat_clustered, <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"CD163"</span>, <span class="st">"AIF1"</span>), <span class="at">ncol =</span> <span class="dv">1</span>)</span>
<span id="cb127-575"><a href="#cb127-575" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb127-576"><a href="#cb127-576" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-579"><a href="#cb127-579" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb127-580"><a href="#cb127-580" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb127-581"><a href="#cb127-581" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(seurat_clustered, <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"CD163"</span>, <span class="st">"AIF1"</span>))</span>
<span id="cb127-582"><a href="#cb127-582" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb127-583"><a href="#cb127-583" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-586"><a href="#cb127-586" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb127-587"><a href="#cb127-587" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb127-588"><a href="#cb127-588" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 8</span></span>
<span id="cb127-589"><a href="#cb127-589" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>(seurat_clustered, <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"CD163"</span>, <span class="st">"AIF1"</span>), <span class="at">label =</span> T)</span>
<span id="cb127-590"><a href="#cb127-590" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb127-591"><a href="#cb127-591" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-592"><a href="#cb127-592" aria-hidden="true" tabindex="-1"></a>保存</span>
<span id="cb127-593"><a href="#cb127-593" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-596"><a href="#cb127-596" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb127-597"><a href="#cb127-597" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb127-598"><a href="#cb127-598" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(seurat_clustered, </span>
<span id="cb127-599"><a href="#cb127-599" aria-hidden="true" tabindex="-1"></a>        <span class="at">file =</span> <span class="st">"output/sc_supplementary/GSE150430_seurat_annotated.RDS"</span>); <span class="fu">beep</span>()</span>
<span id="cb127-600"><a href="#cb127-600" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb127-601"><a href="#cb127-601" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-602"><a href="#cb127-602" aria-hidden="true" tabindex="-1"></a><span class="fu">### 提取髓系细胞</span></span>
<span id="cb127-603"><a href="#cb127-603" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-606"><a href="#cb127-606" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb127-607"><a href="#cb127-607" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb127-608"><a href="#cb127-608" aria-hidden="true" tabindex="-1"></a>sub_seurat <span class="ot">&lt;-</span> <span class="fu">subset</span>(seurat_clustered, </span>
<span id="cb127-609"><a href="#cb127-609" aria-hidden="true" tabindex="-1"></a>                     <span class="at">idents =</span> <span class="fu">c</span>(<span class="st">"Macrophages"</span>, </span>
<span id="cb127-610"><a href="#cb127-610" aria-hidden="true" tabindex="-1"></a>                                <span class="st">"Myeloid (unspecific)"</span>))</span>
<span id="cb127-611"><a href="#cb127-611" aria-hidden="true" tabindex="-1"></a>sub_seurat</span>
<span id="cb127-612"><a href="#cb127-612" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(sub_seurat<span class="sc">@</span>meta.data)</span>
<span id="cb127-613"><a href="#cb127-613" aria-hidden="true" tabindex="-1"></a>myeloid_seurat <span class="ot">&lt;-</span> <span class="fu">CreateSeuratObject</span>(<span class="at">counts =</span> sub_seurat<span class="sc">@</span>assays[[<span class="st">"RNA"</span>]],</span>
<span id="cb127-614"><a href="#cb127-614" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">meta.data =</span> sub_seurat<span class="sc">@</span>meta.data[,<span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">7</span>, <span class="dv">10</span><span class="sc">:</span><span class="dv">13</span>, <span class="dv">16</span>)])</span>
<span id="cb127-615"><a href="#cb127-615" aria-hidden="true" tabindex="-1"></a>myeloid_seurat</span>
<span id="cb127-616"><a href="#cb127-616" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb127-617"><a href="#cb127-617" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-618"><a href="#cb127-618" aria-hidden="true" tabindex="-1"></a>::: callout-warning</span>
<span id="cb127-619"><a href="#cb127-619" aria-hidden="true" tabindex="-1"></a>可以看到在重新创建Seurat之后，默认的“RNA” assay之中只有一个layer。为了后面执行<span class="in">`SCTransform`</span>及整合，**这里需要将其按照样本拆分成不同的layers**。</span>
<span id="cb127-620"><a href="#cb127-620" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-623"><a href="#cb127-623" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb127-624"><a href="#cb127-624" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb127-625"><a href="#cb127-625" aria-hidden="true" tabindex="-1"></a>myeloid_seurat[[<span class="st">"RNA"</span>]] <span class="ot">&lt;-</span> <span class="fu">split</span>(myeloid_seurat[[<span class="st">"RNA"</span>]], </span>
<span id="cb127-626"><a href="#cb127-626" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">f =</span> myeloid_seurat<span class="sc">$</span>samples)</span>
<span id="cb127-627"><a href="#cb127-627" aria-hidden="true" tabindex="-1"></a>myeloid_seurat</span>
<span id="cb127-628"><a href="#cb127-628" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb127-629"><a href="#cb127-629" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb127-630"><a href="#cb127-630" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-633"><a href="#cb127-633" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb127-634"><a href="#cb127-634" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb127-635"><a href="#cb127-635" aria-hidden="true" tabindex="-1"></a><span class="co"># 补充原细胞注释信息</span></span>
<span id="cb127-636"><a href="#cb127-636" aria-hidden="true" tabindex="-1"></a>myeloid_seurat<span class="sc">$</span>old_clusters <span class="ot">&lt;-</span> <span class="fu">Idents</span>(sub_seurat)</span>
<span id="cb127-637"><a href="#cb127-637" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(myeloid_seurat<span class="sc">$</span>old_clusters)</span>
<span id="cb127-638"><a href="#cb127-638" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-639"><a href="#cb127-639" aria-hidden="true" tabindex="-1"></a><span class="co"># 保存</span></span>
<span id="cb127-640"><a href="#cb127-640" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(myeloid_seurat, <span class="at">file =</span> <span class="st">"output/sc_supplementary/GSE150430_myeloid_seurat.RDS"</span>)</span>
<span id="cb127-641"><a href="#cb127-641" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb127-642"><a href="#cb127-642" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-643"><a href="#cb127-643" aria-hidden="true" tabindex="-1"></a>------------------------------------------------------------------------</span>
<span id="cb127-644"><a href="#cb127-644" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-645"><a href="#cb127-645" aria-hidden="true" tabindex="-1"></a><span class="fu">### 髓系细胞进一步归一化、整合、降维、分群</span></span>
<span id="cb127-646"><a href="#cb127-646" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-647"><a href="#cb127-647" aria-hidden="true" tabindex="-1"></a><span class="fu">#### SCTransform、PCA</span></span>
<span id="cb127-648"><a href="#cb127-648" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-651"><a href="#cb127-651" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb127-652"><a href="#cb127-652" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(<span class="at">list =</span> <span class="fu">ls</span>())</span>
<span id="cb127-653"><a href="#cb127-653" aria-hidden="true" tabindex="-1"></a><span class="fu">gc</span>()</span>
<span id="cb127-654"><a href="#cb127-654" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-655"><a href="#cb127-655" aria-hidden="true" tabindex="-1"></a>myeloid_seurat <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"output/sc_supplementary/GSE150430_myeloid_seurat.RDS"</span>)</span>
<span id="cb127-656"><a href="#cb127-656" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-657"><a href="#cb127-657" aria-hidden="true" tabindex="-1"></a><span class="co"># SCTranform</span></span>
<span id="cb127-658"><a href="#cb127-658" aria-hidden="true" tabindex="-1"></a>myeloid_seurat <span class="ot">&lt;-</span> <span class="fu">SCTransform</span>(myeloid_seurat, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb127-659"><a href="#cb127-659" aria-hidden="true" tabindex="-1"></a>myeloid_seurat</span>
<span id="cb127-660"><a href="#cb127-660" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-661"><a href="#cb127-661" aria-hidden="true" tabindex="-1"></a><span class="co"># Check which assays are stored in objects</span></span>
<span id="cb127-662"><a href="#cb127-662" aria-hidden="true" tabindex="-1"></a>myeloid_seurat<span class="sc">@</span>assays</span>
<span id="cb127-663"><a href="#cb127-663" aria-hidden="true" tabindex="-1"></a><span class="co"># 查看目前默认的assay</span></span>
<span id="cb127-664"><a href="#cb127-664" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(myeloid_seurat)</span>
<span id="cb127-665"><a href="#cb127-665" aria-hidden="true" tabindex="-1"></a><span class="co"># 查看默认assay的layers</span></span>
<span id="cb127-666"><a href="#cb127-666" aria-hidden="true" tabindex="-1"></a><span class="fu">Layers</span>(myeloid_seurat)</span>
<span id="cb127-667"><a href="#cb127-667" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-668"><a href="#cb127-668" aria-hidden="true" tabindex="-1"></a><span class="co"># 执行PCA</span></span>
<span id="cb127-669"><a href="#cb127-669" aria-hidden="true" tabindex="-1"></a>myeloid_seurat <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(myeloid_seurat)</span>
<span id="cb127-670"><a href="#cb127-670" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb127-671"><a href="#cb127-671" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-672"><a href="#cb127-672" aria-hidden="true" tabindex="-1"></a><span class="fu">#### 不进行整合时检查细胞分群情况：</span></span>
<span id="cb127-673"><a href="#cb127-673" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-676"><a href="#cb127-676" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb127-677"><a href="#cb127-677" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 15</span></span>
<span id="cb127-678"><a href="#cb127-678" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 10</span></span>
<span id="cb127-679"><a href="#cb127-679" aria-hidden="true" tabindex="-1"></a><span class="co"># 查看降维信息</span></span>
<span id="cb127-680"><a href="#cb127-680" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(myeloid_seurat<span class="sc">@</span>reductions)</span>
<span id="cb127-681"><a href="#cb127-681" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-682"><a href="#cb127-682" aria-hidden="true" tabindex="-1"></a><span class="co"># Run UMAP</span></span>
<span id="cb127-683"><a href="#cb127-683" aria-hidden="true" tabindex="-1"></a>myeloid_seurat <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(myeloid_seurat, </span>
<span id="cb127-684"><a href="#cb127-684" aria-hidden="true" tabindex="-1"></a>                          <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">40</span>, </span>
<span id="cb127-685"><a href="#cb127-685" aria-hidden="true" tabindex="-1"></a>                          <span class="at">reduction =</span> <span class="st">"pca"</span>,</span>
<span id="cb127-686"><a href="#cb127-686" aria-hidden="true" tabindex="-1"></a>                          <span class="at">reduction.name =</span> <span class="st">"umap.unintegrated"</span>)</span>
<span id="cb127-687"><a href="#cb127-687" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-688"><a href="#cb127-688" aria-hidden="true" tabindex="-1"></a><span class="co"># 分群</span></span>
<span id="cb127-689"><a href="#cb127-689" aria-hidden="true" tabindex="-1"></a><span class="co"># Determine the K-nearest neighbor graph</span></span>
<span id="cb127-690"><a href="#cb127-690" aria-hidden="true" tabindex="-1"></a>myeloid_seurat <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(myeloid_seurat, </span>
<span id="cb127-691"><a href="#cb127-691" aria-hidden="true" tabindex="-1"></a>                                <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">40</span>, </span>
<span id="cb127-692"><a href="#cb127-692" aria-hidden="true" tabindex="-1"></a>                                <span class="at">reduction =</span> <span class="st">"pca"</span>)</span>
<span id="cb127-693"><a href="#cb127-693" aria-hidden="true" tabindex="-1"></a>myeloid_seurat <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(myeloid_seurat, </span>
<span id="cb127-694"><a href="#cb127-694" aria-hidden="true" tabindex="-1"></a>                               <span class="at">cluster.name =</span> <span class="st">"unintegrated_clusters"</span>)</span>
<span id="cb127-695"><a href="#cb127-695" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-696"><a href="#cb127-696" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot UMAP</span></span>
<span id="cb127-697"><a href="#cb127-697" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(myeloid_seurat, </span>
<span id="cb127-698"><a href="#cb127-698" aria-hidden="true" tabindex="-1"></a>              <span class="at">reduction =</span> <span class="st">"umap.unintegrated"</span>,</span>
<span id="cb127-699"><a href="#cb127-699" aria-hidden="true" tabindex="-1"></a>              <span class="at">group.by =</span> <span class="st">"samples"</span>)</span>
<span id="cb127-700"><a href="#cb127-700" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(myeloid_seurat, </span>
<span id="cb127-701"><a href="#cb127-701" aria-hidden="true" tabindex="-1"></a>              <span class="at">reduction =</span> <span class="st">"umap.unintegrated"</span>,</span>
<span id="cb127-702"><a href="#cb127-702" aria-hidden="true" tabindex="-1"></a>              <span class="at">split.by =</span> <span class="st">"samples"</span>)</span>
<span id="cb127-703"><a href="#cb127-703" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_grid</span>(p1, p2, </span>
<span id="cb127-704"><a href="#cb127-704" aria-hidden="true" tabindex="-1"></a>          <span class="at">ncol =</span> <span class="dv">1</span>, <span class="at">labels =</span> <span class="st">"AUTO"</span>)</span>
<span id="cb127-705"><a href="#cb127-705" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb127-706"><a href="#cb127-706" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-707"><a href="#cb127-707" aria-hidden="true" tabindex="-1"></a><span class="fu">#### 整合</span></span>
<span id="cb127-708"><a href="#cb127-708" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-711"><a href="#cb127-711" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb127-712"><a href="#cb127-712" aria-hidden="true" tabindex="-1"></a><span class="co"># 整合</span></span>
<span id="cb127-713"><a href="#cb127-713" aria-hidden="true" tabindex="-1"></a>myeloid_integrated <span class="ot">&lt;-</span> <span class="fu">IntegrateLayers</span>(<span class="at">object =</span> myeloid_seurat,</span>
<span id="cb127-714"><a href="#cb127-714" aria-hidden="true" tabindex="-1"></a>                                      <span class="at">method =</span> HarmonyIntegration,</span>
<span id="cb127-715"><a href="#cb127-715" aria-hidden="true" tabindex="-1"></a>                                      <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb127-716"><a href="#cb127-716" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-717"><a href="#cb127-717" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-718"><a href="#cb127-718" aria-hidden="true" tabindex="-1"></a><span class="co"># 整合后合并RNA layer</span></span>
<span id="cb127-719"><a href="#cb127-719" aria-hidden="true" tabindex="-1"></a>myeloid_integrated[[<span class="st">"RNA"</span>]] <span class="ot">&lt;-</span> <span class="fu">JoinLayers</span>(myeloid_integrated[[<span class="st">"RNA"</span>]])</span>
<span id="cb127-720"><a href="#cb127-720" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-721"><a href="#cb127-721" aria-hidden="true" tabindex="-1"></a><span class="co"># 查看整合后的降维信息</span></span>
<span id="cb127-722"><a href="#cb127-722" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(myeloid_integrated<span class="sc">@</span>reductions)</span>
<span id="cb127-723"><a href="#cb127-723" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb127-724"><a href="#cb127-724" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-725"><a href="#cb127-725" aria-hidden="true" tabindex="-1"></a><span class="fu">#### **整合后检验细胞分群情况**</span></span>
<span id="cb127-726"><a href="#cb127-726" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-729"><a href="#cb127-729" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb127-730"><a href="#cb127-730" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 10</span></span>
<span id="cb127-731"><a href="#cb127-731" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 15</span></span>
<span id="cb127-732"><a href="#cb127-732" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123456</span>)</span>
<span id="cb127-733"><a href="#cb127-733" aria-hidden="true" tabindex="-1"></a><span class="co"># Run UMAP</span></span>
<span id="cb127-734"><a href="#cb127-734" aria-hidden="true" tabindex="-1"></a>myeloid_integrated <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(myeloid_integrated, </span>
<span id="cb127-735"><a href="#cb127-735" aria-hidden="true" tabindex="-1"></a>                              <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">40</span>,</span>
<span id="cb127-736"><a href="#cb127-736" aria-hidden="true" tabindex="-1"></a>                              <span class="at">reduction =</span> <span class="st">"harmony"</span>, <span class="co"># 更改降维来源为整合后的"harmony"</span></span>
<span id="cb127-737"><a href="#cb127-737" aria-hidden="true" tabindex="-1"></a>                              <span class="at">reduction.name =</span> <span class="st">"umap.integrated"</span>)</span>
<span id="cb127-738"><a href="#cb127-738" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(myeloid_integrated<span class="sc">@</span>reductions)</span>
<span id="cb127-739"><a href="#cb127-739" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-740"><a href="#cb127-740" aria-hidden="true" tabindex="-1"></a><span class="co"># 分群</span></span>
<span id="cb127-741"><a href="#cb127-741" aria-hidden="true" tabindex="-1"></a>myeloid_integrated <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(myeloid_integrated, </span>
<span id="cb127-742"><a href="#cb127-742" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">40</span>, </span>
<span id="cb127-743"><a href="#cb127-743" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">reduction =</span> <span class="st">"harmony"</span>) <span class="co">#更改降维来源为"harmony"</span></span>
<span id="cb127-744"><a href="#cb127-744" aria-hidden="true" tabindex="-1"></a>myeloid_integrated <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(myeloid_integrated, </span>
<span id="cb127-745"><a href="#cb127-745" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">cluster.name =</span> <span class="st">"integrated_clusters"</span>)</span>
<span id="cb127-746"><a href="#cb127-746" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(myeloid_integrated<span class="sc">@</span>meta.data)</span>
<span id="cb127-747"><a href="#cb127-747" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-748"><a href="#cb127-748" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot UMAP                             </span></span>
<span id="cb127-749"><a href="#cb127-749" aria-hidden="true" tabindex="-1"></a>p3 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(myeloid_integrated, </span>
<span id="cb127-750"><a href="#cb127-750" aria-hidden="true" tabindex="-1"></a>              <span class="at">reduction =</span> <span class="st">"umap.integrated"</span>, </span>
<span id="cb127-751"><a href="#cb127-751" aria-hidden="true" tabindex="-1"></a>              <span class="at">group.by =</span> <span class="st">"samples"</span>)</span>
<span id="cb127-752"><a href="#cb127-752" aria-hidden="true" tabindex="-1"></a>p4 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(myeloid_integrated, </span>
<span id="cb127-753"><a href="#cb127-753" aria-hidden="true" tabindex="-1"></a>              <span class="at">reduction =</span> <span class="st">"umap.integrated"</span>, </span>
<span id="cb127-754"><a href="#cb127-754" aria-hidden="true" tabindex="-1"></a>              <span class="at">split.by =</span> <span class="st">"samples"</span>)</span>
<span id="cb127-755"><a href="#cb127-755" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_grid</span>(p1, p3, p2, p4, </span>
<span id="cb127-756"><a href="#cb127-756" aria-hidden="true" tabindex="-1"></a>          <span class="at">ncol =</span> <span class="dv">1</span>, </span>
<span id="cb127-757"><a href="#cb127-757" aria-hidden="true" tabindex="-1"></a>          <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">"Before Harmony"</span>, <span class="st">"After Harmony"</span>, </span>
<span id="cb127-758"><a href="#cb127-758" aria-hidden="true" tabindex="-1"></a>                     <span class="st">"Before Harmony"</span>, <span class="st">"After Harmony"</span>))</span>
<span id="cb127-759"><a href="#cb127-759" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb127-760"><a href="#cb127-760" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-761"><a href="#cb127-761" aria-hidden="true" tabindex="-1"></a><span class="fu">#### 聚类</span></span>
<span id="cb127-762"><a href="#cb127-762" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-765"><a href="#cb127-765" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb127-766"><a href="#cb127-766" aria-hidden="true" tabindex="-1"></a><span class="co"># Determine the clusters for various resolutions                                </span></span>
<span id="cb127-767"><a href="#cb127-767" aria-hidden="true" tabindex="-1"></a>myeloid_integrated <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(myeloid_integrated,</span>
<span id="cb127-768"><a href="#cb127-768" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">resolution =</span> <span class="fu">c</span>(<span class="fl">0.01</span>, <span class="fl">0.05</span>, <span class="fl">0.1</span>, <span class="fl">0.2</span>, <span class="fl">0.3</span>, </span>
<span id="cb127-769"><a href="#cb127-769" aria-hidden="true" tabindex="-1"></a>                                                  <span class="fl">0.4</span>, <span class="fl">0.5</span>, <span class="fl">0.8</span>, <span class="dv">1</span>),</span>
<span id="cb127-770"><a href="#cb127-770" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">verbose =</span> F)</span>
<span id="cb127-771"><a href="#cb127-771" aria-hidden="true" tabindex="-1"></a><span class="co"># Explore resolutions</span></span>
<span id="cb127-772"><a href="#cb127-772" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(myeloid_integrated<span class="sc">@</span>meta.data, <span class="dv">5</span>)</span>
<span id="cb127-773"><a href="#cb127-773" aria-hidden="true" tabindex="-1"></a><span class="co"># 查看各个分辨率下的细胞分群情况</span></span>
<span id="cb127-774"><a href="#cb127-774" aria-hidden="true" tabindex="-1"></a><span class="fu">select</span>(myeloid_integrated<span class="sc">@</span>meta.data, </span>
<span id="cb127-775"><a href="#cb127-775" aria-hidden="true" tabindex="-1"></a>       <span class="fu">starts_with</span>(<span class="at">match =</span> <span class="st">"SCT_snn_res."</span>)) <span class="sc">%&gt;%</span>  </span>
<span id="cb127-776"><a href="#cb127-776" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lapply</span>(levels)</span>
<span id="cb127-777"><a href="#cb127-777" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb127-778"><a href="#cb127-778" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-779"><a href="#cb127-779" aria-hidden="true" tabindex="-1"></a>绘制聚类树展示不同分辨率下的细胞分群情况及相互关系</span>
<span id="cb127-780"><a href="#cb127-780" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-783"><a href="#cb127-783" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb127-784"><a href="#cb127-784" aria-hidden="true" tabindex="-1"></a>tree <span class="ot">&lt;-</span> <span class="fu">clustree</span>(myeloid_integrated<span class="sc">@</span>meta.data, </span>
<span id="cb127-785"><a href="#cb127-785" aria-hidden="true" tabindex="-1"></a>                 <span class="at">prefix =</span> <span class="st">"SCT_snn_res."</span>) <span class="co"># 指定包含聚类信息的列</span></span>
<span id="cb127-786"><a href="#cb127-786" aria-hidden="true" tabindex="-1"></a>tree</span>
<span id="cb127-787"><a href="#cb127-787" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb127-788"><a href="#cb127-788" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-789"><a href="#cb127-789" aria-hidden="true" tabindex="-1"></a>这里，选取分辨为0.2</span>
<span id="cb127-790"><a href="#cb127-790" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-793"><a href="#cb127-793" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb127-794"><a href="#cb127-794" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(myeloid_integrated) <span class="ot">&lt;-</span> <span class="st">"SCT_snn_res.0.2"</span></span>
<span id="cb127-795"><a href="#cb127-795" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb127-796"><a href="#cb127-796" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-797"><a href="#cb127-797" aria-hidden="true" tabindex="-1"></a>聚类可视化</span>
<span id="cb127-798"><a href="#cb127-798" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-801"><a href="#cb127-801" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb127-802"><a href="#cb127-802" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the UMAP</span></span>
<span id="cb127-803"><a href="#cb127-803" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(myeloid_integrated,</span>
<span id="cb127-804"><a href="#cb127-804" aria-hidden="true" tabindex="-1"></a>        <span class="at">reduction =</span> <span class="st">"umap.integrated"</span>,</span>
<span id="cb127-805"><a href="#cb127-805" aria-hidden="true" tabindex="-1"></a>        <span class="at">label =</span> <span class="cn">FALSE</span>)</span>
<span id="cb127-806"><a href="#cb127-806" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb127-807"><a href="#cb127-807" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-808"><a href="#cb127-808" aria-hidden="true" tabindex="-1"></a><span class="fu">#### **细胞分群质量评估**</span></span>
<span id="cb127-809"><a href="#cb127-809" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-810"><a href="#cb127-810" aria-hidden="true" tabindex="-1"></a>分析样本类型是否影响细胞分群</span>
<span id="cb127-811"><a href="#cb127-811" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-814"><a href="#cb127-814" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb127-815"><a href="#cb127-815" aria-hidden="true" tabindex="-1"></a><span class="co"># 先简单查看不同cluster的细胞数</span></span>
<span id="cb127-816"><a href="#cb127-816" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(myeloid_integrated<span class="sc">@</span>active.ident)</span>
<span id="cb127-817"><a href="#cb127-817" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-818"><a href="#cb127-818" aria-hidden="true" tabindex="-1"></a><span class="co"># 查看不同样本类型中的细胞分群情况</span></span>
<span id="cb127-819"><a href="#cb127-819" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(myeloid_integrated, </span>
<span id="cb127-820"><a href="#cb127-820" aria-hidden="true" tabindex="-1"></a>        <span class="at">reduction =</span> <span class="st">"umap.integrated"</span>,</span>
<span id="cb127-821"><a href="#cb127-821" aria-hidden="true" tabindex="-1"></a>        <span class="at">label =</span> <span class="cn">TRUE</span>, </span>
<span id="cb127-822"><a href="#cb127-822" aria-hidden="true" tabindex="-1"></a>        <span class="at">split.by =</span> <span class="st">"groups"</span>) <span class="sc">+</span> </span>
<span id="cb127-823"><a href="#cb127-823" aria-hidden="true" tabindex="-1"></a>  <span class="fu">NoLegend</span>()</span>
<span id="cb127-824"><a href="#cb127-824" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb127-825"><a href="#cb127-825" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-826"><a href="#cb127-826" aria-hidden="true" tabindex="-1"></a>分析细胞周期是否影响细胞分群</span>
<span id="cb127-827"><a href="#cb127-827" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-830"><a href="#cb127-830" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb127-831"><a href="#cb127-831" aria-hidden="true" tabindex="-1"></a><span class="co"># Explore whether clusters segregate by cell cycle phase</span></span>
<span id="cb127-832"><a href="#cb127-832" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(myeloid_integrated,</span>
<span id="cb127-833"><a href="#cb127-833" aria-hidden="true" tabindex="-1"></a>        <span class="at">reduction =</span> <span class="st">"umap.integrated"</span>,</span>
<span id="cb127-834"><a href="#cb127-834" aria-hidden="true" tabindex="-1"></a>        <span class="at">label =</span> <span class="cn">TRUE</span>, </span>
<span id="cb127-835"><a href="#cb127-835" aria-hidden="true" tabindex="-1"></a>        <span class="at">split.by =</span> <span class="st">"Phase"</span>) <span class="sc">+</span> </span>
<span id="cb127-836"><a href="#cb127-836" aria-hidden="true" tabindex="-1"></a>  <span class="fu">NoLegend</span>()</span>
<span id="cb127-837"><a href="#cb127-837" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb127-838"><a href="#cb127-838" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-839"><a href="#cb127-839" aria-hidden="true" tabindex="-1"></a>分析其他非期望变异来源是否会影响细胞分群</span>
<span id="cb127-840"><a href="#cb127-840" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-843"><a href="#cb127-843" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb127-844"><a href="#cb127-844" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 10</span></span>
<span id="cb127-845"><a href="#cb127-845" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 10</span></span>
<span id="cb127-846"><a href="#cb127-846" aria-hidden="true" tabindex="-1"></a><span class="co"># Determine metrics to plot present in seurat_clustered@meta.data</span></span>
<span id="cb127-847"><a href="#cb127-847" aria-hidden="true" tabindex="-1"></a>metrics <span class="ot">&lt;-</span>  <span class="fu">c</span>(<span class="st">"nCount_RNA"</span>, <span class="st">"nFeature_RNA"</span>, <span class="st">"S.Score"</span>, <span class="st">"G2M.Score"</span>, <span class="st">"mitoRatio"</span>)</span>
<span id="cb127-848"><a href="#cb127-848" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-849"><a href="#cb127-849" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>(myeloid_integrated, </span>
<span id="cb127-850"><a href="#cb127-850" aria-hidden="true" tabindex="-1"></a>            <span class="at">reduction =</span> <span class="st">"umap.integrated"</span>, </span>
<span id="cb127-851"><a href="#cb127-851" aria-hidden="true" tabindex="-1"></a>            <span class="at">features =</span> metrics,</span>
<span id="cb127-852"><a href="#cb127-852" aria-hidden="true" tabindex="-1"></a>            <span class="at">pt.size =</span> <span class="fl">0.4</span>, </span>
<span id="cb127-853"><a href="#cb127-853" aria-hidden="true" tabindex="-1"></a>            <span class="at">order =</span> <span class="cn">TRUE</span>,</span>
<span id="cb127-854"><a href="#cb127-854" aria-hidden="true" tabindex="-1"></a>            <span class="at">min.cutoff =</span> <span class="st">'q10'</span>, </span>
<span id="cb127-855"><a href="#cb127-855" aria-hidden="true" tabindex="-1"></a>            <span class="at">label =</span> <span class="cn">TRUE</span>)</span>
<span id="cb127-856"><a href="#cb127-856" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb127-857"><a href="#cb127-857" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-858"><a href="#cb127-858" aria-hidden="true" tabindex="-1"></a>保存</span>
<span id="cb127-859"><a href="#cb127-859" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-862"><a href="#cb127-862" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb127-863"><a href="#cb127-863" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(myeloid_integrated, </span>
<span id="cb127-864"><a href="#cb127-864" aria-hidden="true" tabindex="-1"></a>        <span class="at">file =</span> <span class="st">"output/sc_supplementary/GSE150430_myeloid_clustered.RDS"</span>)</span>
<span id="cb127-865"><a href="#cb127-865" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb127-866"><a href="#cb127-866" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-867"><a href="#cb127-867" aria-hidden="true" tabindex="-1"></a><span class="fu">#### 亚群细胞注释</span></span>
<span id="cb127-868"><a href="#cb127-868" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-871"><a href="#cb127-871" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb127-872"><a href="#cb127-872" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">
      <ul class="footer-items list-unstyled">
<li class="nav-item">
 This book was built with Quarto
  </li>  
</ul>
</div>   
    <div class="nav-footer-center">Copyright 2023, Du Junhong</div>
    <div class="nav-footer-right">
      <ul class="footer-items list-unstyled">
<li class="nav-item compact">
    <a class="nav-link" href="https://github.com/djhcod">
      <i class="bi bi-github" role="img">
</i> 
    </a>
  </li>  
</ul>
</div>
  </div>
</footer>


</body></html>