<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>R语言学习笔记 - 30&nbsp; 多个单细胞数据集整合分析（上）</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../single_cell/sc_supplementary/integrated_analysis_multiple_single_cell_datasets_2.html" rel="next">
<link href="../../single_cell/sc_supplementary/universal_marker.html" rel="prev">
<link href="../../images/logo.jpg" rel="icon" type="image/jpeg">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark"><script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><script type="application/json" class="js-hypothesis-config">
{
  "theme": "classic",
  "openSidebar": false,
  "showHighlights": "always",
  "enableExperimentalNewNoteButton": true
}
</script><script async="" src="https://hypothes.is/embed.js"></script>
<meta name="twitter:title" content="R语言学习笔记 - 30&nbsp; 多个单细胞数据集整合分析（上）">
<meta name="twitter:description" content="">
<meta name="twitter:image" content="https://djhcod.github.io/r-notes/single_cell/sc_supplementary/images/截屏2024-01-10 12.00.48.png">
<meta name="twitter:image-height" content="886">
<meta name="twitter:image-width" content="1446">
<meta name="twitter:card" content="summary_large_image">
</head>
<body class="nav-sidebar docked nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="navbar navbar-expand-lg navbar-dark "><div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a href="../../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../../images/logo.jpg" alt="" class="navbar-logo"></a>
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">R语言学习笔记</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
<li class="nav-item">
    <a class="nav-link active" href="../../index.html" rel="" target="" aria-current="page">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../intro.html" rel="" target="">
 <span class="menu-text">Introduction</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-parts" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Parts</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-parts">
<li>
    <a class="dropdown-item" href="../../r_basic/r_basics.html" rel="" target="">
 <span class="dropdown-text">R语言基础</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../single_cell/seurat/seurat.html" rel="" target="">
 <span class="dropdown-text">Seurat v5官方文档</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../single_cell/scRNA-seq_online/00_intro.html" rel="" target="">
 <span class="dropdown-text">scRNA-seq_online学习材料</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../single_cell/sc_supplementary/sc_supplementary_intro.html" rel="" target="">
 <span class="dropdown-text">单细胞补充内容</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../quarto_foundation/quarto_foundation.html" rel="" target="">
 <span class="dropdown-text">Quarto基础</span></a>
  </li>  
    </ul>
</li>
  <li class="nav-item">
    <a class="nav-link" href="../../references.html" rel="" target="">
 <span class="menu-text">References</span></a>
  </li>  
</ul>
<ul class="navbar-nav navbar-nav-scroll ms-auto">
<li class="nav-item">
    <a class="nav-link" href="https://bookdown.org" rel="" target=""><i class="bi bi-book-fill" role="img">
</i> 
 <span class="menu-text">Bookdown</span></a>
  </li>  
</ul>
<div class="quarto-navbar-tools tools-wide">
    <a href="https://github.com/djhcod/r-notes" rel="" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
    <div class="dropdown">
      <a href="" title="Share" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Share"><i class="bi bi-share"></i></a>
      <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="quarto-navigation-tool-dropdown-0">
<li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://twitter.com/intent/tweet?url=%7Curl%7C">
              <i class="bi bi-bi-twitter pe-1"></i>
            Twitter
            </a>
          </li>
          <li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://www.facebook.com/sharer/sharer.php?u=%7Curl%7C">
              <i class="bi bi-bi-facebook pe-1"></i>
            Facebook
            </a>
          </li>
      </ul>
</div>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../single_cell/sc_supplementary/sc_supplementary_intro.html">单细胞补充内容</a></li><li class="breadcrumb-item"><a href="../../single_cell/sc_supplementary/integrated_analysis_multiple_single_cell_datasets_1.html"><span class="chapter-number">30</span>&nbsp; <span class="chapter-title">多个单细胞数据集整合分析（上）</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto"><div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">主页</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">前言</span></a>
  </div>
</li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../r_basic/r_basics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">R语言基础</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/environment_configuration.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Rstudio环境配置</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/data_input_output.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">数据的读取与输出</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/basic_data_function.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">数据处理基本函数</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/character.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">字符的处理</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/loop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">循环</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../single_cell/seurat/seurat.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Seurat v5官方文档</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/seurat_command_list.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Seurat常用函数清单</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/seurat_tutorial.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Seurat细胞分群官方教程</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/data_visualization_methods_in_seurat.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Seurat中的数据可视化方法</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/sctransform.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">基于SCTransform的单细胞数据标准化</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/cell_cycle_regression.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">消除细胞周期的影响</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/integration.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">整合（integration）</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/integrative_analysis_in_seurat_v5.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Seurat v5单细胞数据整合分析</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/marker_gene_identification.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">寻找marker gene</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/de_vignette.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">寻找不同样本类型间同一细胞类型内的差异基因</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/mapping_and_annotating_query_datasets.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">基于参考集的细胞注释</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../single_cell/scRNA-seq_online/00_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">scRNA-seq_online学习材料</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/01_intro_to_scRNA-seq.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">单细胞测序技术介绍</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/02_SC_generation_of_count_matrix.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">单细胞RNA测序—从raw data到count matrix</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/03_SC_quality_control-setup.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">数据导入与Seurat对象构建</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/04_SC_quality_control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">质控</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/postQC_workflow.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">21</span>&nbsp; <span class="chapter-title">Single-cell RNA-seq Clustering Workflow</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/05_theory_of_PCA.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">22</span>&nbsp; <span class="chapter-title">PCA原理</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/06_SC_SCT_normalization.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">23</span>&nbsp; <span class="chapter-title">Normalization and regressing out unwanted variation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/06_integration.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">24</span>&nbsp; <span class="chapter-title">单细胞数据整合（Integration）</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/07_SC_clustering_cells_SCT.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">25</span>&nbsp; <span class="chapter-title">细胞聚类（clustering analysis）</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/08_SC_clustering_quality_control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">26</span>&nbsp; <span class="chapter-title">细胞分群质量评估</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/09_merged_SC_marker_identification.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">27</span>&nbsp; <span class="chapter-title">寻找marker基因+细胞注释</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../single_cell/sc_supplementary/sc_supplementary_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">单细胞补充内容</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/sc_supplementary/read_sc_data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">28</span>&nbsp; <span class="chapter-title">读取非标准格式的单细胞数据</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/sc_supplementary/universal_marker.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">29</span>&nbsp; <span class="chapter-title">细胞分群通用marker</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/sc_supplementary/integrated_analysis_multiple_single_cell_datasets_1.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">30</span>&nbsp; <span class="chapter-title">多个单细胞数据集整合分析（上）</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/sc_supplementary/integrated_analysis_multiple_single_cell_datasets_2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">31</span>&nbsp; <span class="chapter-title">多个单细胞数据集整合分析（下）</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/sc_supplementary/DecontX.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">32</span>&nbsp; <span class="chapter-title">预测和去除单细胞转录组的环境游离RNA污染</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../quarto_foundation/quarto_foundation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Quarto基础</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/yaml_settings.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">33</span>&nbsp; <span class="chapter-title">YAML设置</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/images_settings.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">34</span>&nbsp; <span class="chapter-title">图片设置</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/cross_references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">35</span>&nbsp; <span class="chapter-title">交叉引用</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/insert_other_content.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">36</span>&nbsp; <span class="chapter-title">插入其他内容</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/quarto_shiny.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">37</span>&nbsp; <span class="chapter-title">在Quarto中嵌入Shiny应用程序</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/quarto_books.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">38</span>&nbsp; <span class="chapter-title">Quarto Books</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/github_pages.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">39</span>&nbsp; <span class="chapter-title">发布到GitHub Pages</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="false">
 <span class="menu-text">参考文献</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active" data-toc-expanded="1"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#%E5%8A%A0%E8%BD%BD%E5%8C%85" id="toc-加载包" class="nav-link active" data-scroll-target="#%E5%8A%A0%E8%BD%BD%E5%8C%85"><span class="header-section-number">30.1</span> 加载包</a></li>
  <li>
<a href="#gse150430" id="toc-gse150430" class="nav-link" data-scroll-target="#gse150430"><span class="header-section-number">30.2</span> GSE150430</a>
  <ul class="collapse">
<li><a href="#%E6%95%B0%E6%8D%AE%E5%AF%BC%E5%85%A5" id="toc-数据导入" class="nav-link" data-scroll-target="#%E6%95%B0%E6%8D%AE%E5%AF%BC%E5%85%A5">数据导入</a></li>
  <li><a href="#%E8%B4%A8%E6%8E%A7" id="toc-质控" class="nav-link" data-scroll-target="#%E8%B4%A8%E6%8E%A7">质控</a></li>
  <li><a href="#%E5%BD%92%E4%B8%80%E5%8C%96%E5%8F%8A%E6%B6%88%E9%99%A4%E9%9D%9E%E6%9C%9F%E6%9C%9B%E5%8F%98%E5%BC%82%E6%9D%A5%E6%BA%90" id="toc-归一化及消除非期望变异来源" class="nav-link" data-scroll-target="#%E5%BD%92%E4%B8%80%E5%8C%96%E5%8F%8A%E6%B6%88%E9%99%A4%E9%9D%9E%E6%9C%9F%E6%9C%9B%E5%8F%98%E5%BC%82%E6%9D%A5%E6%BA%90">归一化及消除非期望变异来源</a></li>
  <li><a href="#%E6%95%B0%E6%8D%AE%E6%95%B4%E5%90%88" id="toc-数据整合" class="nav-link" data-scroll-target="#%E6%95%B0%E6%8D%AE%E6%95%B4%E5%90%88">数据整合</a></li>
  <li><a href="#%E8%81%9A%E7%B1%BB" id="toc-聚类" class="nav-link" data-scroll-target="#%E8%81%9A%E7%B1%BB">聚类</a></li>
  <li><a href="#%E7%BB%86%E8%83%9E%E5%88%86%E7%BE%A4%E8%B4%A8%E9%87%8F%E8%AF%84%E4%BC%B0" id="toc-细胞分群质量评估" class="nav-link" data-scroll-target="#%E7%BB%86%E8%83%9E%E5%88%86%E7%BE%A4%E8%B4%A8%E9%87%8F%E8%AF%84%E4%BC%B0"><strong>细胞分群质量评估</strong></a></li>
  <li><a href="#%E7%BB%86%E8%83%9E%E6%B3%A8%E9%87%8A" id="toc-细胞注释" class="nav-link" data-scroll-target="#%E7%BB%86%E8%83%9E%E6%B3%A8%E9%87%8A">细胞注释</a></li>
  <li><a href="#%E6%8F%90%E5%8F%96%E9%AB%93%E7%B3%BB%E7%BB%86%E8%83%9E" id="toc-提取髓系细胞" class="nav-link" data-scroll-target="#%E6%8F%90%E5%8F%96%E9%AB%93%E7%B3%BB%E7%BB%86%E8%83%9E">提取髓系细胞</a></li>
  </ul>
</li>
  </ul><div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/djhcod/r-notes/blob/main/single_cell/sc_supplementary/integrated_analysis_multiple_single_cell_datasets_1.qmd" class="toc-action">View source</a></p><p><a href="https://github.com/djhcod/r-notes/issues/new" class="toc-action">Report an issue</a></p><p><a href="https://github.com/djhcod/r-notes/edit/main/single_cell/sc_supplementary/integrated_analysis_multiple_single_cell_datasets_1.qmd" class="toc-action">Edit this page</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title"><span id="sec-multiple_single_cell_datasets_1" class="quarto-section-identifier"><span class="chapter-number">30</span>&nbsp; <span class="chapter-title">多个单细胞数据集整合分析（上）</span></span></h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><blockquote class="blockquote">
<p>参考：<a href="https://mp.weixin.qq.com/s/51PtUi7ZKp1nbrFGElU74A">单细胞多数据集整合示例</a></p>
</blockquote>
<p>有时候，为了扩大数据量和得到更加可靠的结论，我们可能获取和下载了多个单细胞数据集。如果我们只关注某一细胞类型，如T cells或B cells或髓系细胞，那么就需要合并分析这些数据集，提取感兴趣的细胞亚群。这时候有两种方法可以选择：</p>
<ul>
<li><p><strong>对各个数据集分别降维分群，然后提取各个数据集的感兴趣细胞亚群并整合到一起</strong></p></li>
<li><p><strong>在一开始分析的时候就对这些数据集进行整合，然后分群，提起感兴趣细胞群</strong></p></li>
</ul>
<p><strong>这个部分怎么分析并没有一个特定的答案或者规则，具体还是得看数据集适合以上提出的哪种分析顺序。</strong></p>
<p>例如<span class="citation" data-cites="gong2023">(<a href="../../references.html#ref-gong2023" role="doc-biblioref">Gong et al. 2023</a>)</span>就用了第二种方法。选取了三个鼻咽癌单细胞转录组数据： GSE150825, GSE150430, GSE162025。首先对三组数据进行整合，然后选取其中的一部分T细胞进行进一步分析。</p>
<p><img src="images/截屏2024-01-10 12.00.48.png" class="img-fluid" width="600"></p>
<p>这里我们介绍第一种方式，即<strong>对各数据集分别降维分群，然后提取并整合各个数据集的感兴趣细胞亚群。</strong></p>
<section id="加载包" class="level2" data-number="30.1"><h2 data-number="30.1" class="anchored" data-anchor-id="加载包">
<span class="header-section-number">30.1</span> 加载包</h2>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_1_cache/html/unnamed-chunk-1_c219338dbc87b401a126d04b4044d291">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">beepr</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/">cowplot</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/lazappi/clustree">clustree</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="gse150430" class="level2" data-number="30.2"><h2 data-number="30.2" class="anchored" data-anchor-id="gse150430">
<span class="header-section-number">30.2</span> GSE150430</h2>
<p>这里我们直接载入 <a href="read_sc_data.html#sec-reading_non_standard_large_volume_text"><span>Section&nbsp;28.4</span></a> 中构建好的Seurat对象，该案例数据的介绍及读取过程参见该章节。</p>
<section id="数据导入" class="level3"><h3 class="anchored" data-anchor-id="数据导入">数据导入</h3>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_1_cache/html/unnamed-chunk-2_6f00a42a16c4e5ac2eae67a0f5f31cb9">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">merged_seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="st">"output/sc_supplementary/GSE150430_merged_seurat.rds"</span><span class="op">)</span></span>
<span><span class="va">merged_seurat</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
24720 features across 48584 samples within 1 assay 
Active assay: RNA (24720 features, 0 variable features)
 16 layers present: counts.1, counts.2, counts.3, counts.4, counts.5, counts.6, counts.7, counts.8, counts.9, counts.10, counts.11, counts.12, counts.13, counts.14, counts.15, counts.16</code></pre>
</div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">merged_seurat</span>, <span class="fl">4</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                       orig.ident nCount_RNA nFeature_RNA samples
N01_CAGAATCAGTAATCCC.1        N01   3470.970         4173     N01
N01_TGCGGGTAGTACGTAA.1        N01   3155.725         2994     N01
N01_AAACCTGAGGGTTTCT.1        N01   2347.034         1946     N01
N01_AAACCTGAGTTCCACA.1        N01   2362.495         1993     N01</code></pre>
</div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">$</span><span class="va">samples</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "N01" "P01" "P02" "P03" "P04" "P05" "P06" "P07" "P08" "P09" "P10" "P11"
[13] "P12" "P13" "P14" "P15"</code></pre>
</div>
</div>
</section><section id="质控" class="level3"><h3 class="anchored" data-anchor-id="质控">质控</h3>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_1_cache/html/unnamed-chunk-3_6e5fc7f9148ab8af3bed26840ce5b19d">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Add number of genes per UMI for each cell to metadata</span></span>
<span><span class="va">merged_seurat</span><span class="op">$</span><span class="va">log10GenesPerUMI</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">$</span><span class="va">nFeature_RNA</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">$</span><span class="va">nCount_RNA</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">$</span><span class="va">log10GenesPerUMI</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
 0.8897  0.9488  0.9670  0.9773  0.9971  1.1429 </code></pre>
</div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Compute percent mito ratio</span></span>
<span><span class="va">merged_seurat</span><span class="op">$</span><span class="va">mitoRatio</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/PercentageFeatureSet.html">PercentageFeatureSet</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">merged_seurat</span>, </span>
<span>                                                pattern <span class="op">=</span> <span class="st">"^MT-"</span><span class="op">)</span></span>
<span><span class="va">merged_seurat</span><span class="op">$</span><span class="va">mitoRatio</span> <span class="op">&lt;-</span> <span class="va">merged_seurat</span><span class="op">$</span><span class="va">mitoRatio</span> <span class="op">/</span> <span class="fl">100</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">$</span><span class="va">mitoRatio</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     Min.   1st Qu.    Median      Mean   3rd Qu.      Max. 
0.0006859 0.0106660 0.0127552 0.0129526 0.0150420 0.0407058 </code></pre>
</div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/boxplot.html">boxplot</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">$</span><span class="va">mitoRatio</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integrated_analysis_multiple_single_cell_datasets_1_files/figure-html/unnamed-chunk-3-1.png" class="img-fluid" width="288"></p>
</div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Add cell IDs to metadata</span></span>
<span><span class="va">merged_seurat</span><span class="op">$</span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Visualize the number of cell counts per sample</span></span>
<span><span class="va">merged_seurat</span><span class="op">@</span><span class="va">meta.data</span>  <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">samples</span>, fill <span class="op">=</span> <span class="va">samples</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, vjust <span class="op">=</span> <span class="fl">1</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span>, face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"NCells"</span><span class="op">)</span></span>
<span><span class="co"># Visualize the number UMIs/transcripts per cell</span></span>
<span><span class="va">merged_seurat</span><span class="op">@</span><span class="va">meta.data</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">samples</span>, x <span class="op">=</span> <span class="va">nCount_RNA</span>, fill <span class="op">=</span> <span class="va">samples</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span><span class="op">(</span><span class="st">"Cell density"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span>
<span><span class="co"># Visualize the distribution of genes detected per cell via histogram</span></span>
<span><span class="va">merged_seurat</span><span class="op">@</span><span class="va">meta.data</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">samples</span>, x <span class="op">=</span> <span class="va">nFeature_RNA</span>, fill<span class="op">=</span> <span class="va">samples</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></span>
<span><span class="co"># Visualize the overall complexity of the gene expression by visualizing the genes detected per UMI (novelty score)</span></span>
<span><span class="va">merged_seurat</span><span class="op">@</span><span class="va">meta.data</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">log10GenesPerUMI</span>, color <span class="op">=</span> <span class="va">samples</span>, fill<span class="op">=</span><span class="va">samples</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span></span>
<span><span class="co"># Visualize the distribution of mitochondrial gene expression detected per cell</span></span>
<span><span class="va">merged_seurat</span><span class="op">@</span><span class="va">meta.data</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">samples</span>, x <span class="op">=</span> <span class="va">mitoRatio</span>, fill <span class="op">=</span> <span class="va">samples</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell quarto-layout-panel" data-hash="integrated_analysis_multiple_single_cell_datasets_1_cache/html/unnamed-chunk-4_98fd4778631183eb991ba1d86339253f">
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="cell-output-display quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img src="integrated_analysis_multiple_single_cell_datasets_1_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img src="integrated_analysis_multiple_single_cell_datasets_1_files/figure-html/unnamed-chunk-4-2.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="cell-output-display quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img src="integrated_analysis_multiple_single_cell_datasets_1_files/figure-html/unnamed-chunk-4-3.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img src="integrated_analysis_multiple_single_cell_datasets_1_files/figure-html/unnamed-chunk-4-4.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="cell-output-display quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img src="integrated_analysis_multiple_single_cell_datasets_1_files/figure-html/unnamed-chunk-4-5.png" class="img-fluid" width="672"></p>
</div>
</div>
</div>
</div>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_1_cache/html/unnamed-chunk-5_2b878f461c8310e864e70f6050e135c7">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Visualize the correlation between genes detected and number of UMIs and determine whether strong presence of cells with low numbers of genes/UMIs</span></span>
<span><span class="va">merged_seurat</span><span class="op">@</span><span class="va">meta.data</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">nCount_RNA</span>, y <span class="op">=</span> <span class="va">nFeature_RNA</span>, color <span class="op">=</span> <span class="va">mitoRatio</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html">scale_colour_gradient</a></span><span class="op">(</span>low <span class="op">=</span> <span class="st">"gray90"</span>, high <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html">stat_smooth</a></span><span class="op">(</span>method <span class="op">=</span> <span class="va">lm</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">500</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">250</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">samples</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integrated_analysis_multiple_single_cell_datasets_1_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid" width="960"></p>
</div>
</div>
<p>可以看到该数据集已经进行了质控，各项指标均在正常范围内，因此，我们可以跳过细胞/基因过滤，直接进入下面的环节。</p>
</section><section id="归一化及消除非期望变异来源" class="level3"><h3 class="anchored" data-anchor-id="归一化及消除非期望变异来源">归一化及消除非期望变异来源</h3>
<div class="callout callout-style-default callout-caution callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Caution
</div>
</div>
<div class="callout-body-container callout-body">
<p>由于这里<code>SCTransform</code>的数据量太大，在我的16GB MacBook Pro上超出了R内存分配上限，出现了“Error: vector memory exhausted (limit reached?)”报错。因此，这里在运行下面的脚本之前通过 <a href="../../r_basic/environment_configuration.html#sec-Raising_memory_limit"><span>Section&nbsp;2.8.2</span></a> 的方法进行了处理。</p>
</div>
</div>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_1_cache/html/unnamed-chunk-6_7934252e567389d743cdacd0f1ea668e">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># SCTranform</span></span>
<span><span class="va">merged_seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SCTransform.html">SCTransform</a></span><span class="op">(</span><span class="va">merged_seurat</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>; <span class="fu"><a href="https://rdrr.io/pkg/beepr/man/beep.html">beep</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/gc.html">gc</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># 释放未使用内存</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>            used   (Mb) gc trigger    (Mb) limit (Mb)   max used    (Mb)
Ncells   8269159  441.7   15696718   838.3         NA   14114610   753.9
Vcells 988418686 7541.1 3232713790 24663.7     102400 3232712447 24663.7</code></pre>
</div>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">merged_seurat</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
46357 features across 48584 samples within 2 assays 
Active assay: SCT (21637 features, 3000 variable features)
 3 layers present: counts, data, scale.data
 1 other assay present: RNA</code></pre>
</div>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Check which assays are stored in objects</span></span>
<span><span class="va">merged_seurat</span><span class="op">@</span><span class="va">assays</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$RNA
Assay (v5) data with 24720 features for 48584 cells
First 10 features:
 RP11-34P13.7, RP11-34P13.8, AL627309.1, AP006222.2, RP4-669L17.10,
RP4-669L17.2, RP5-857K21.4, RP11-206L10.3, RP11-206L10.5, RP11-206L10.2 
Layers:
 counts.1, counts.2, counts.3, counts.4, counts.5, counts.6, counts.7,
counts.8, counts.9, counts.10, counts.11, counts.12, counts.13,
counts.14, counts.15, counts.16 

$SCT
SCTAssay data with 21637 features for 48584 cells, and 16 SCTModel(s) 
Top 10 variable features:
 RNASE6, PHACTR1, NR4A3, C1orf162, HMOX1, IFIT2, CD7, CD86, CLNK, CRIP1 </code></pre>
</div>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 查看目前默认的assay</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/DefaultAssay.html">DefaultAssay</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "SCT"</code></pre>
</div>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 查看默认assay的layers</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/Layers.html">Layers</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "counts"     "data"       "scale.data"</code></pre>
</div>
</div>
<section id="评估细胞周期的影响" class="level4"><h4 class="anchored" data-anchor-id="评估细胞周期的影响">评估细胞周期的影响</h4>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_1_cache/html/unnamed-chunk-7_168d6a1fa6fdfe1a84d997ee2b24f27a">
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Load cell cycle markers</span></span>
<span><span class="va">s.genes</span> <span class="op">&lt;-</span> <span class="va">cc.genes</span><span class="op">$</span><span class="va">s.genes</span></span>
<span><span class="va">g2m.genes</span> <span class="op">&lt;-</span> <span class="va">cc.genes</span><span class="op">$</span><span class="va">g2m.genes</span></span>
<span></span>
<span><span class="co"># Score cells for cell cycle</span></span>
<span><span class="va">merged_seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/CellCycleScoring.html">CellCycleScoring</a></span><span class="op">(</span><span class="va">merged_seurat</span>, </span>
<span>                                  g2m.features <span class="op">=</span> <span class="va">g2m.genes</span>, </span>
<span>                                  s.features <span class="op">=</span> <span class="va">s.genes</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 现在的meta.data中多出了细胞周期评分“S.Score”和“G2M.Score”，以及推断的细胞所处的周期“Phase”</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                       orig.ident nCount_RNA nFeature_RNA samples
N01_CAGAATCAGTAATCCC.1        N01   3470.970         4173     N01
N01_TGCGGGTAGTACGTAA.1        N01   3155.725         2994     N01
N01_AAACCTGAGGGTTTCT.1        N01   2347.034         1946     N01
N01_AAACCTGAGTTCCACA.1        N01   2362.495         1993     N01
N01_AAACCTGGTTCTCATT.1        N01   2130.465         1459     N01
N01_AAACGGGGTCCTCCAT.1        N01   2772.010         2573     N01
                       log10GenesPerUMI   mitoRatio                  cells
N01_CAGAATCAGTAATCCC.1        1.0225953 0.008644269 N01_CAGAATCAGTAATCCC.1
N01_TGCGGGTAGTACGTAA.1        0.9934705 0.009895666 N01_TGCGGGTAGTACGTAA.1
N01_AAACCTGAGGGTTTCT.1        0.9758564 0.015183419 N01_AAACCTGAGGGTTTCT.1
N01_AAACCTGAGTTCCACA.1        0.9781039 0.014965958 N01_AAACCTGAGTTCCACA.1
N01_AAACCTGGTTCTCATT.1        0.9506023 0.017437508 N01_AAACCTGGTTCTCATT.1
N01_AAACGGGGTCCTCCAT.1        0.9906021 0.011233365 N01_AAACGGGGTCCTCCAT.1
                       nCount_SCT nFeature_SCT     S.Score    G2M.Score Phase
N01_CAGAATCAGTAATCCC.1       1840         1674 -0.05179250 -0.070444626    G1
N01_TGCGGGTAGTACGTAA.1       2840         2604 -0.04341940 -0.074302336    G1
N01_AAACCTGAGGGTTTCT.1       2443         1931 -0.02396942 -0.007239586    G1
N01_AAACCTGAGTTCCACA.1       2488         1972 -0.01294260 -0.021167805    G1
N01_AAACCTGGTTCTCATT.1       2104         1451 -0.05440456 -0.049531592    G1
N01_AAACGGGGTCCTCCAT.1       2892         2520 -0.10189203 -0.087537749    G1</code></pre>
</div>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 查看一下细胞周期的分布情况</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">$</span><span class="va">Phase</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
   G1   G2M     S 
33941  5147  9496 </code></pre>
</div>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 执行PCA</span></span>
<span><span class="va">merged_seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html">RunPCA</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot the PCA colored by cell cycle phase</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">merged_seurat</span>,</span>
<span>              reduction <span class="op">=</span> <span class="st">"pca"</span>,</span>
<span>              group.by<span class="op">=</span> <span class="st">"Phase"</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">merged_seurat</span>,</span>
<span>              reduction <span class="op">=</span> <span class="st">"pca"</span>,</span>
<span>              group.by<span class="op">=</span> <span class="st">"Phase"</span>,</span>
<span>              split.by <span class="op">=</span> <span class="st">"Phase"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">plot_grid</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span>, ncol <span class="op">=</span> <span class="fl">2</span>, labels <span class="op">=</span> <span class="st">"AUTO"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integrated_analysis_multiple_single_cell_datasets_1_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid" width="960"></p>
</div>
</div>
<p>可以看到细胞周期不是变异来源。</p>
</section><section id="评估线粒体基因的影响" class="level4"><h4 class="anchored" data-anchor-id="评估线粒体基因的影响"><strong>评估线粒体基因的影响</strong></h4>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_1_cache/html/unnamed-chunk-8_0a74e8f08964d9c86ed9894011ffee4a">
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Check quartile values</span></span>
<span><span class="va">mito_sum</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">$</span><span class="va">mitoRatio</span><span class="op">)</span></span>
<span><span class="va">mito_sum</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     Min.   1st Qu.    Median      Mean   3rd Qu.      Max. 
0.0006859 0.0106660 0.0127552 0.0129526 0.0150420 0.0407058 </code></pre>
</div>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Turn mitoRatio into categorical factor vector based on quartile values</span></span>
<span><span class="va">merged_seurat</span><span class="op">$</span><span class="va">mitoFr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cut.html">cut</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">$</span><span class="va">mitoRatio</span>, </span>
<span>                            breaks<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="cn">Inf</span>, <span class="va">mito_sum</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, <span class="va">mito_sum</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span>, <span class="va">mito_sum</span><span class="op">[</span><span class="fl">5</span><span class="op">]</span>, <span class="cn">Inf</span><span class="op">)</span>,</span>
<span>                            labels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Low"</span>, <span class="st">"Medium"</span>, <span class="st">"Medium high"</span>, <span class="st">"High"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">$</span><span class="va">mitoFr</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integrated_analysis_multiple_single_cell_datasets_1_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid" width="288"></p>
</div>
</div>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_1_cache/html/unnamed-chunk-9_a20f46c820a13427587c101ae6fc9e90">
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Plot the PCA colored by cell cycle phase</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">merged_seurat</span>,</span>
<span>              reduction <span class="op">=</span> <span class="st">"pca"</span>,</span>
<span>              group.by<span class="op">=</span> <span class="st">"mitoFr"</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">merged_seurat</span>,</span>
<span>              reduction <span class="op">=</span> <span class="st">"pca"</span>,</span>
<span>              group.by<span class="op">=</span> <span class="st">"mitoFr"</span>,</span>
<span>              split.by <span class="op">=</span> <span class="st">"mitoFr"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">plot_grid</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span>, ncol <span class="op">=</span> <span class="fl">2</span>, labels <span class="op">=</span> <span class="st">"AUTO"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integrated_analysis_multiple_single_cell_datasets_1_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid" width="1440"></p>
</div>
</div>
<p>可以看到线粒体基因比例不是变异来源。</p>
<p>由于细胞周期和线粒体基因比例都不是非期望变异来源，所以我们这里不需要再次运行SCTransform来回归这些变量。接下来，直接进入数据整合环节。</p>
</section></section><section id="数据整合" class="level3"><h3 class="anchored" data-anchor-id="数据整合">数据整合</h3>
<section id="不进行整合时检验细胞分群情况" class="level4"><h4 class="anchored" data-anchor-id="不进行整合时检验细胞分群情况"><strong>不进行整合时检验细胞分群情况</strong></h4>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_1_cache/html/unnamed-chunk-10_26470ad057188e46b9f077192d404b6a">
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 查看降维信息</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">@</span><span class="va">reductions</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "pca"</code></pre>
</div>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Run UMAP</span></span>
<span><span class="va">merged_seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html">RunUMAP</a></span><span class="op">(</span><span class="va">merged_seurat</span>, </span>
<span>                         dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">40</span>, </span>
<span>                         reduction <span class="op">=</span> <span class="st">"pca"</span>,</span>
<span>                         reduction.name <span class="op">=</span> <span class="st">"umap.unintegrated"</span><span class="op">)</span>; <span class="fu"><a href="https://rdrr.io/pkg/beepr/man/beep.html">beep</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 分群</span></span>
<span><span class="co"># Determine the K-nearest neighbor graph</span></span>
<span><span class="va">merged_seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html">FindNeighbors</a></span><span class="op">(</span><span class="va">merged_seurat</span>, </span>
<span>                               dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">40</span>, </span>
<span>                               reduction <span class="op">=</span> <span class="st">"pca"</span><span class="op">)</span></span>
<span><span class="va">merged_seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html">FindClusters</a></span><span class="op">(</span><span class="va">merged_seurat</span>, </span>
<span>                              cluster.name <span class="op">=</span> <span class="st">"unintegrated_clusters"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 48584
Number of edges: 1937300

Running Louvain algorithm...
Maximum modularity in 10 random starts: 0.9298
Number of communities: 38
Elapsed time: 6 seconds</code></pre>
</div>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Plot UMAP</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">merged_seurat</span>, </span>
<span>              reduction <span class="op">=</span> <span class="st">"umap.unintegrated"</span>,</span>
<span>              group.by <span class="op">=</span> <span class="st">"samples"</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">merged_seurat</span>, </span>
<span>              reduction <span class="op">=</span> <span class="st">"umap.unintegrated"</span>,</span>
<span>              split.by <span class="op">=</span> <span class="st">"samples"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">plot_grid</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span>, </span>
<span>          ncol <span class="op">=</span> <span class="fl">1</span>, labels <span class="op">=</span> <span class="st">"AUTO"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integrated_analysis_multiple_single_cell_datasets_1_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid" width="1440"></p>
</div>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 由于样本数较多，我们再按照样本类型“Normal” vs. “Cancer”画一下UMAP图</span></span>
<span><span class="va">merged_seurat</span><span class="op">$</span><span class="va">groups</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">$</span><span class="va">samples</span> <span class="op">==</span> <span class="st">"N01"</span>, <span class="st">"Normal"</span>, <span class="st">"Cancer"</span><span class="op">)</span></span>
<span><span class="va">p1a</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">merged_seurat</span>, </span>
<span>               reduction <span class="op">=</span> <span class="st">"umap.unintegrated"</span>,</span>
<span>               group.by <span class="op">=</span> <span class="st">"groups"</span><span class="op">)</span></span>
<span><span class="va">p2a</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">merged_seurat</span>, </span>
<span>               reduction <span class="op">=</span> <span class="st">"umap.unintegrated"</span>,</span>
<span>               split.by <span class="op">=</span> <span class="st">"groups"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">plot_grid</a></span><span class="op">(</span><span class="va">p1a</span>, <span class="va">p2a</span>, </span>
<span>          ncol <span class="op">=</span> <span class="fl">1</span>, labels <span class="op">=</span> <span class="st">"AUTO"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integrated_analysis_multiple_single_cell_datasets_1_files/figure-html/unnamed-chunk-10-2.png" class="img-fluid" width="1440"></p>
</div>
</div>
</section><section id="整合" class="level4"><h4 class="anchored" data-anchor-id="整合">整合</h4>
<p>这里我们用Harmony整合算法。</p>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_1_cache/html/unnamed-chunk-11_25080e948d9c7b2cfc82212d88fd5e53">
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 整合，比较耗时间，进度条会一直显示0%直至运算完成</span></span>
<span><span class="va">seurat_integrated</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/IntegrateLayers.html">IntegrateLayers</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">merged_seurat</span>,</span>
<span>                                     method <span class="op">=</span> <span class="va">HarmonyIntegration</span>,</span>
<span>                                     assay <span class="op">=</span> <span class="st">"SCT"</span>, <span class="co"># Integrating SCTransformed data</span></span>
<span>                                     orig.reduction <span class="op">=</span> <span class="st">"pca"</span>,</span>
<span>                                     verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>; <span class="fu"><a href="https://rdrr.io/pkg/beepr/man/beep.html">beep</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># 整合后合并RNA layer</span></span>
<span><span class="va">seurat_integrated</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/SplitLayers.html">JoinLayers</a></span><span class="op">(</span><span class="va">seurat_integrated</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 查看整合后的降维信息</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">seurat_integrated</span><span class="op">@</span><span class="va">reductions</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "pca"               "umap.unintegrated" "harmony"          </code></pre>
</div>
</div>
</section><section id="整合后检验细胞分群情况" class="level4"><h4 class="anchored" data-anchor-id="整合后检验细胞分群情况"><strong>整合后检验细胞分群情况</strong></h4>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_1_cache/html/unnamed-chunk-12_9d5c5314f37f459ebce29e3c66f7478f">
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">123456</span><span class="op">)</span></span>
<span><span class="co"># Run UMAP</span></span>
<span><span class="va">seurat_integrated</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html">RunUMAP</a></span><span class="op">(</span><span class="va">seurat_integrated</span>, </span>
<span>                             dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">40</span>,</span>
<span>                             reduction <span class="op">=</span> <span class="st">"harmony"</span>, <span class="co"># 更改降维来源为整合后的"harmony"</span></span>
<span>                             reduction.name <span class="op">=</span> <span class="st">"umap.integrated"</span><span class="op">)</span>; <span class="fu"><a href="https://rdrr.io/pkg/beepr/man/beep.html">beep</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">seurat_integrated</span><span class="op">@</span><span class="va">reductions</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "pca"               "umap.unintegrated" "harmony"          
[4] "umap.integrated"  </code></pre>
</div>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 分群</span></span>
<span><span class="va">seurat_integrated</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html">FindNeighbors</a></span><span class="op">(</span><span class="va">seurat_integrated</span>, </span>
<span>                                   dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">40</span>, </span>
<span>                                   reduction <span class="op">=</span> <span class="st">"harmony"</span><span class="op">)</span> <span class="co">#更改降维来源为"harmony"</span></span>
<span><span class="va">seurat_integrated</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html">FindClusters</a></span><span class="op">(</span><span class="va">seurat_integrated</span>, </span>
<span>                                  cluster.name <span class="op">=</span> <span class="st">"integrated_clusters"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 48584
Number of edges: 2161498

Running Louvain algorithm...
Maximum modularity in 10 random starts: 0.9121
Number of communities: 24
Elapsed time: 9 seconds</code></pre>
</div>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">seurat_integrated</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "orig.ident"            "nCount_RNA"            "nFeature_RNA"         
 [4] "samples"               "log10GenesPerUMI"      "mitoRatio"            
 [7] "cells"                 "nCount_SCT"            "nFeature_SCT"         
[10] "S.Score"               "G2M.Score"             "Phase"                
[13] "mitoFr"                "unintegrated_clusters" "seurat_clusters"      
[16] "groups"                "integrated_clusters"  </code></pre>
</div>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Plot UMAP                             </span></span>
<span><span class="va">p3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">seurat_integrated</span>, </span>
<span>              reduction <span class="op">=</span> <span class="st">"umap.integrated"</span>, </span>
<span>              group.by <span class="op">=</span> <span class="st">"samples"</span><span class="op">)</span></span>
<span><span class="va">p4</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">seurat_integrated</span>, </span>
<span>              reduction <span class="op">=</span> <span class="st">"umap.integrated"</span>, </span>
<span>              split.by <span class="op">=</span> <span class="st">"samples"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">plot_grid</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p3</span>, <span class="va">p2</span>, <span class="va">p4</span>, </span>
<span>          ncol <span class="op">=</span> <span class="fl">1</span>, </span>
<span>          labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Before Harmony"</span>, <span class="st">"After Harmony"</span>, </span>
<span>                     <span class="st">"Before Harmony"</span>, <span class="st">"After Harmony"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integrated_analysis_multiple_single_cell_datasets_1_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid" width="960"></p>
</div>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 保存</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">seurat_integrated</span>, </span>
<span>        file <span class="op">=</span> <span class="st">"output/sc_supplementary/GSE150430_seurat_integrated.RDS"</span><span class="op">)</span>; <span class="fu"><a href="https://rdrr.io/pkg/beepr/man/beep.html">beep</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr></section></section><section id="聚类" class="level3"><h3 class="anchored" data-anchor-id="聚类">聚类</h3>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_1_cache/html/unnamed-chunk-13_bf15c88dc84c6814cacd36402d7d5465">
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/rm.html">rm</a></span><span class="op">(</span>list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ls.html">ls</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/gc.html">gc</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>             used   (Mb) gc trigger    (Mb) limit (Mb)   max used    (Mb)
Ncells    8423157  449.9   15696718   838.3         NA   15696718   838.3
Vcells 1189108415 9072.2 3232713790 24663.7     102400 3232712447 24663.7</code></pre>
</div>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">seurat_integrated</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="st">"output/sc_supplementary/GSE150430_seurat_integrated.RDS"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_1_cache/html/unnamed-chunk-14_eb30832d969469850494a341cb586022">
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Determine the clusters for various resolutions                                </span></span>
<span><span class="va">seurat_integrated</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html">FindClusters</a></span><span class="op">(</span><span class="va">seurat_integrated</span>,</span>
<span>                                  resolution <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.01</span>, <span class="fl">0.05</span>, <span class="fl">0.1</span>, <span class="fl">0.2</span>, <span class="fl">0.3</span>, <span class="fl">0.4</span>, <span class="fl">0.5</span>, <span class="fl">0.8</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>                                  verbose <span class="op">=</span> <span class="cn">F</span><span class="op">)</span>; <span class="fu"><a href="https://rdrr.io/pkg/beepr/man/beep.html">beep</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co"># Explore resolutions</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">seurat_integrated</span><span class="op">@</span><span class="va">meta.data</span>, <span class="fl">5</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                       orig.ident nCount_RNA nFeature_RNA samples
N01_CAGAATCAGTAATCCC.1        N01   3470.970         4173     N01
N01_TGCGGGTAGTACGTAA.1        N01   3155.725         2994     N01
N01_AAACCTGAGGGTTTCT.1        N01   2347.034         1946     N01
N01_AAACCTGAGTTCCACA.1        N01   2362.495         1993     N01
N01_AAACCTGGTTCTCATT.1        N01   2130.465         1459     N01
                       log10GenesPerUMI   mitoRatio                  cells
N01_CAGAATCAGTAATCCC.1        1.0225953 0.008644269 N01_CAGAATCAGTAATCCC.1
N01_TGCGGGTAGTACGTAA.1        0.9934705 0.009895666 N01_TGCGGGTAGTACGTAA.1
N01_AAACCTGAGGGTTTCT.1        0.9758564 0.015183419 N01_AAACCTGAGGGTTTCT.1
N01_AAACCTGAGTTCCACA.1        0.9781039 0.014965958 N01_AAACCTGAGTTCCACA.1
N01_AAACCTGGTTCTCATT.1        0.9506023 0.017437508 N01_AAACCTGGTTCTCATT.1
                       nCount_SCT nFeature_SCT     S.Score    G2M.Score Phase
N01_CAGAATCAGTAATCCC.1       1840         1674 -0.05179250 -0.070444626    G1
N01_TGCGGGTAGTACGTAA.1       2840         2604 -0.04341940 -0.074302336    G1
N01_AAACCTGAGGGTTTCT.1       2443         1931 -0.02396942 -0.007239586    G1
N01_AAACCTGAGTTCCACA.1       2488         1972 -0.01294260 -0.021167805    G1
N01_AAACCTGGTTCTCATT.1       2104         1451 -0.05440456 -0.049531592    G1
                            mitoFr unintegrated_clusters seurat_clusters groups
N01_CAGAATCAGTAATCCC.1         Low                    20              20 Normal
N01_TGCGGGTAGTACGTAA.1         Low                    32              20 Normal
N01_AAACCTGAGGGTTTCT.1        High                     0               0 Normal
N01_AAACCTGAGTTCCACA.1 Medium high                     1               3 Normal
N01_AAACCTGGTTCTCATT.1        High                     1               8 Normal
                       integrated_clusters SCT_snn_res.0.01 SCT_snn_res.0.05
N01_CAGAATCAGTAATCCC.1                  18                2                2
N01_TGCGGGTAGTACGTAA.1                  18                2                2
N01_AAACCTGAGGGTTTCT.1                   0                1                1
N01_AAACCTGAGTTCCACA.1                   0                1                1
N01_AAACCTGGTTCTCATT.1                   7                1                1
                       SCT_snn_res.0.1 SCT_snn_res.0.2 SCT_snn_res.0.3
N01_CAGAATCAGTAATCCC.1               2               2               1
N01_TGCGGGTAGTACGTAA.1               2               2               1
N01_AAACCTGAGGGTTTCT.1               0               0               0
N01_AAACCTGAGTTCCACA.1               0               0               0
N01_AAACCTGGTTCTCATT.1               0               0               0
                       SCT_snn_res.0.4 SCT_snn_res.0.5 SCT_snn_res.0.8
N01_CAGAATCAGTAATCCC.1               1              15              18
N01_TGCGGGTAGTACGTAA.1               1              15              18
N01_AAACCTGAGGGTTTCT.1               0               0               0
N01_AAACCTGAGTTCCACA.1               0               0               0
N01_AAACCTGGTTCTCATT.1               0               8               7
                       SCT_snn_res.1
N01_CAGAATCAGTAATCCC.1            20
N01_TGCGGGTAGTACGTAA.1            20
N01_AAACCTGAGGGTTTCT.1             0
N01_AAACCTGAGTTCCACA.1             3
N01_AAACCTGGTTCTCATT.1             8</code></pre>
</div>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 查看各个分辨率下的细胞分群情况</span></span>
<span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">seurat_integrated</span><span class="op">@</span><span class="va">meta.data</span>, <span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html">starts_with</a></span><span class="op">(</span>match <span class="op">=</span> <span class="st">"SCT_snn_res."</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>  </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">levels</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$SCT_snn_res.0.01
[1] "0" "1" "2" "3" "4" "5"

$SCT_snn_res.0.05
 [1] "0"  "1"  "2"  "3"  "4"  "5"  "6"  "7"  "8"  "9"  "10"

$SCT_snn_res.0.1
 [1] "0"  "1"  "2"  "3"  "4"  "5"  "6"  "7"  "8"  "9"  "10" "11" "12"

$SCT_snn_res.0.2
 [1] "0"  "1"  "2"  "3"  "4"  "5"  "6"  "7"  "8"  "9"  "10" "11" "12"

$SCT_snn_res.0.3
 [1] "0"  "1"  "2"  "3"  "4"  "5"  "6"  "7"  "8"  "9"  "10" "11" "12" "13" "14"
[16] "15" "16"

$SCT_snn_res.0.4
 [1] "0"  "1"  "2"  "3"  "4"  "5"  "6"  "7"  "8"  "9"  "10" "11" "12" "13" "14"
[16] "15" "16" "17"

$SCT_snn_res.0.5
 [1] "0"  "1"  "2"  "3"  "4"  "5"  "6"  "7"  "8"  "9"  "10" "11" "12" "13" "14"
[16] "15" "16" "17" "18" "19" "20"

$SCT_snn_res.0.8
 [1] "0"  "1"  "2"  "3"  "4"  "5"  "6"  "7"  "8"  "9"  "10" "11" "12" "13" "14"
[16] "15" "16" "17" "18" "19" "20" "21" "22" "23"

$SCT_snn_res.1
 [1] "0"  "1"  "2"  "3"  "4"  "5"  "6"  "7"  "8"  "9"  "10" "11" "12" "13" "14"
[16] "15" "16" "17" "18" "19" "20" "21" "22" "23" "24" "25"</code></pre>
</div>
</div>
<p>绘制聚类树展示不同分辨率下的细胞分群情况及相互关系</p>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_1_cache/html/unnamed-chunk-15_3631aeb310c1201973eedb685229f643">
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://lazappi.github.io/clustree/reference/clustree.html">clustree</a></span><span class="op">(</span><span class="va">seurat_integrated</span><span class="op">@</span><span class="va">meta.data</span>, </span>
<span>                 prefix <span class="op">=</span> <span class="st">"SCT_snn_res."</span><span class="op">)</span><span class="co">#指定包含聚类信息的列</span></span>
<span><span class="va">tree</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integrated_analysis_multiple_single_cell_datasets_1_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid" width="960"></p>
</div>
</div>
<p>通常情况下，为了决定合适的聚类分辨率，可使用以下两种策略：</p>
<ol type="1">
<li><p>选择透明箭头出现较少的分辨率聚类结果</p></li>
<li><p>基于marker基因表达选择有生物学意义的分辨率聚类结果</p></li>
</ol>
<p>接下来分析，按照分辨率为0.5进行</p>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_1_cache/html/unnamed-chunk-16_6dd027a0c36c02634f557bf6940979ef">
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/Idents.html">Idents</a></span><span class="op">(</span><span class="va">seurat_integrated</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"SCT_snn_res.0.5"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>聚类可视化</p>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_1_cache/html/unnamed-chunk-17_9af1c2a0dbe4643385122ced9d9dbdb1">
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Plot the UMAP</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">seurat_integrated</span>,</span>
<span>        reduction <span class="op">=</span> <span class="st">"umap.integrated"</span>,</span>
<span>        label <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integrated_analysis_multiple_single_cell_datasets_1_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section><section id="细胞分群质量评估" class="level3"><h3 class="anchored" data-anchor-id="细胞分群质量评估"><strong>细胞分群质量评估</strong></h3>
<section id="分析样本类型是否影响细胞分群" class="level4"><h4 class="anchored" data-anchor-id="分析样本类型是否影响细胞分群"><strong>分析样本类型是否影响细胞分群</strong></h4>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_1_cache/html/unnamed-chunk-18_3865a3324a65699a58f14130d11a50dc">
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 先简单查看不同cluster的细胞数</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">seurat_integrated</span><span class="op">@</span><span class="va">active.ident</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    0     1     2     3     4     5     6     7     8     9    10    11    12 
10830  5756  5526  5321  4161  3765  2388  2365  1952  1478   986   792   759 
   13    14    15    16    17    18    19    20 
  707   398   388   232   214   198   189   179 </code></pre>
</div>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 查看不同样本类型中的细胞分群情况</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">seurat_integrated</span>, </span>
<span>        reduction <span class="op">=</span> <span class="st">"umap.integrated"</span>,</span>
<span>        label <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>        split.by <span class="op">=</span> <span class="st">"groups"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html">NoLegend</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integrated_analysis_multiple_single_cell_datasets_1_files/figure-html/unnamed-chunk-18-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section><section id="分析细胞周期是否影响细胞分群" class="level4"><h4 class="anchored" data-anchor-id="分析细胞周期是否影响细胞分群"><strong>分析细胞周期是否影响细胞分群</strong></h4>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_1_cache/html/unnamed-chunk-19_5cc8c540982056ed2d0a5bba58b888f2">
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Explore whether clusters segregate by cell cycle phase</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">seurat_integrated</span>,</span>
<span>        reduction <span class="op">=</span> <span class="st">"umap.integrated"</span>,</span>
<span>        label <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>        split.by <span class="op">=</span> <span class="st">"Phase"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html">NoLegend</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integrated_analysis_multiple_single_cell_datasets_1_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section><section id="分析其他非期望变异来源是否会影响细胞分群" class="level4"><h4 class="anchored" data-anchor-id="分析其他非期望变异来源是否会影响细胞分群"><strong>分析其他非期望变异来源是否会影响细胞分群</strong></h4>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_1_cache/html/unnamed-chunk-20_41b63448504adcd6e7c7792d471abeba">
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Determine metrics to plot present in seurat_clustered@meta.data</span></span>
<span><span class="va">metrics</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"nCount_RNA"</span>, <span class="st">"nFeature_RNA"</span>, <span class="st">"S.Score"</span>, <span class="st">"G2M.Score"</span>, <span class="st">"mitoRatio"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html">FeaturePlot</a></span><span class="op">(</span><span class="va">seurat_integrated</span>, </span>
<span>            reduction <span class="op">=</span> <span class="st">"umap.integrated"</span>, </span>
<span>            features <span class="op">=</span> <span class="va">metrics</span>,</span>
<span>            pt.size <span class="op">=</span> <span class="fl">0.4</span>, </span>
<span>            order <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            min.cutoff <span class="op">=</span> <span class="st">'q10'</span>, </span>
<span>            label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integrated_analysis_multiple_single_cell_datasets_1_files/figure-html/unnamed-chunk-20-1.png" class="img-fluid" width="960"></p>
</div>
</div>
<p>保存</p>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_1_cache/html/unnamed-chunk-21_7034a4d652c5904983cecc81471a2be7">
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">seurat_integrated</span>, </span>
<span>        file <span class="op">=</span> <span class="st">"output/sc_supplementary/GSE150430_seurat_clustered.RDS"</span><span class="op">)</span>; <span class="fu"><a href="https://rdrr.io/pkg/beepr/man/beep.html">beep</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr></section></section><section id="细胞注释" class="level3"><h3 class="anchored" data-anchor-id="细胞注释">细胞注释</h3>
<section id="探索cell-type-markers的表达" class="level4"><h4 class="anchored" data-anchor-id="探索cell-type-markers的表达"><strong>探索cell type markers的表达</strong></h4>
<p>这里的marker基因选用 <a href="universal_marker.html#sec-Universal_Marker_for_initial_Cell_clustering"><span>Section&nbsp;29.3</span></a> 中的细胞初步分群通用marker。</p>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_1_cache/html/unnamed-chunk-22_fbce6e0461732095a56b533170ee9186">
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/rm.html">rm</a></span><span class="op">(</span>list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ls.html">ls</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/gc.html">gc</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>             used   (Mb) gc trigger    (Mb) limit (Mb)   max used    (Mb)
Ncells    8509105  454.5   15696718   838.3         NA   15696718   838.3
Vcells 1006105137 7676.0 3232713790 24663.7     102400 3232712447 24663.7</code></pre>
</div>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">seurat_clustered</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="st">"output/sc_supplementary/GSE150430_seurat_clustered.RDS"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">genes_to_check</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'PTPRC'</span>, </span>
<span>                   <span class="st">"CD163"</span>,<span class="st">"AIF1"</span>, </span>
<span>                   <span class="st">'CD3D'</span>, <span class="st">'CD3E'</span>, <span class="st">'CD4'</span>, <span class="st">'CD8A'</span>, </span>
<span>                   <span class="st">'CD19'</span>, <span class="st">'CD79A'</span>, <span class="st">'MS4A1'</span>, <span class="st">"SDC1"</span>, <span class="st">"CD27"</span>, <span class="st">"CD38"</span>,</span>
<span>                   <span class="st">'IGHG1'</span>, <span class="st">'MZB1'</span>, <span class="st">'SDC1'</span>, <span class="st">"JCHAIN"</span>, </span>
<span>                   <span class="st">'CD68'</span>, <span class="st">'CD163'</span>, <span class="st">'CD14'</span>, </span>
<span>                   <span class="st">'S100A9'</span>, <span class="st">'S100A8'</span>, <span class="st">'MMP19'</span>,</span>
<span>                   <span class="st">'C1QA'</span>,  <span class="st">'C1QB'</span>, </span>
<span>                   <span class="st">'TPSAB1'</span>, <span class="st">'TPSB2'</span>, </span>
<span>                   <span class="st">'KLRB1'</span>, <span class="st">"KLRD1"</span>, <span class="st">'NCR1'</span>, <span class="st">"GNLY"</span>, <span class="st">"NKG7"</span>, </span>
<span>                   <span class="st">'FGF7'</span>, <span class="st">'MME'</span>, <span class="st">'ACTA2'</span>, <span class="st">"COL3A1"</span>, </span>
<span>                   <span class="st">'PECAM1'</span>, <span class="st">'VWF'</span>, </span>
<span>                   <span class="st">'EPCAM'</span>, <span class="st">'KRT19'</span>, <span class="st">'PROM1'</span>, <span class="st">'ALDH1A1'</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DotPlot.html">DotPlot</a></span><span class="op">(</span><span class="va">seurat_clustered</span>, </span>
<span>        features <span class="op">=</span> <span class="va">genes_to_check</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html">coord_flip</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integrated_analysis_multiple_single_cell_datasets_1_files/figure-html/unnamed-chunk-22-1.png" class="img-fluid" width="768"></p>
</div>
</div>
</section><section id="手动注释" class="level4"><h4 class="anchored" data-anchor-id="手动注释">手动注释</h4>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_1_cache/html/unnamed-chunk-23_db21479619931683b561ffe594f21bfd">
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">seurat_clustered</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/Idents.html">RenameIdents</a></span><span class="op">(</span><span class="va">seurat_clustered</span>, </span>
<span>                                 <span class="st">"0"</span> <span class="op">=</span> <span class="st">"B cells"</span>,</span>
<span>                                 <span class="st">"1"</span> <span class="op">=</span> <span class="st">"T cells"</span>,</span>
<span>                                 <span class="st">"2"</span> <span class="op">=</span> <span class="st">"Epithelial/cancer"</span>,</span>
<span>                                 <span class="st">"3"</span> <span class="op">=</span> <span class="st">"T cells"</span>,</span>
<span>                                 <span class="st">"4"</span> <span class="op">=</span> <span class="st">"T cells"</span>,</span>
<span>                                 <span class="st">"5"</span> <span class="op">=</span> <span class="st">"Macrophages"</span>,</span>
<span>                                 <span class="st">"6"</span> <span class="op">=</span> <span class="st">"T cells"</span>,</span>
<span>                                 <span class="st">"7"</span> <span class="op">=</span> <span class="st">"Plasma cells"</span>,</span>
<span>                                 <span class="st">"8"</span> <span class="op">=</span> <span class="st">"B cells"</span>,</span>
<span>                                 <span class="st">"9"</span> <span class="op">=</span> <span class="st">"Epithelial/cancer"</span>,</span>
<span>                                 <span class="st">"10"</span> <span class="op">=</span> <span class="st">"B cells"</span>,</span>
<span>                                 <span class="st">"11"</span> <span class="op">=</span> <span class="st">"B cells"</span>,</span>
<span>                                 <span class="st">"12"</span> <span class="op">=</span> <span class="st">"Unknown"</span>,</span>
<span>                                 <span class="st">"13"</span> <span class="op">=</span> <span class="st">"NK cells"</span>,</span>
<span>                                 <span class="st">"14"</span> <span class="op">=</span> <span class="st">"Unknown"</span>,</span>
<span>                                 <span class="st">"15"</span> <span class="op">=</span> <span class="st">"Epithelial/cancer"</span>,</span>
<span>                                 <span class="st">"16"</span> <span class="op">=</span> <span class="st">"Myeloid (unspecific)"</span>,</span>
<span>                                 <span class="st">"17"</span> <span class="op">=</span> <span class="st">"Epithelial/cancer"</span>,</span>
<span>                                 <span class="st">"18"</span> <span class="op">=</span> <span class="st">"Epithelial/cancer"</span>,</span>
<span>                                 <span class="st">"19"</span> <span class="op">=</span> <span class="st">"Epithelial/cancer"</span>,</span>
<span>                                 <span class="st">"20"</span> <span class="op">=</span> <span class="st">"B cells"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/Idents.html">Idents</a></span><span class="op">(</span><span class="va">seurat_clustered</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
             B cells              T cells    Epithelial/cancer 
               14739                17626                 7993 
         Macrophages         Plasma cells              Unknown 
                3765                 2365                 1157 
            NK cells Myeloid (unspecific) 
                 707                  232 </code></pre>
</div>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Plot the UMAP</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">seurat_clustered</span>, </span>
<span>        reduction <span class="op">=</span> <span class="st">"umap.integrated"</span>, </span>
<span>        label <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integrated_analysis_multiple_single_cell_datasets_1_files/figure-html/unnamed-chunk-23-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>在注释好的数据中再次检查marker基因的表达情况：</p>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_1_cache/html/unnamed-chunk-24_bb9d05608412b8da723fa06c6d30331f">
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DotPlot.html">DotPlot</a></span><span class="op">(</span><span class="va">seurat_clustered</span>, </span>
<span>        features <span class="op">=</span> <span class="va">genes_to_check</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">30</span>,vjust <span class="op">=</span> <span class="fl">0.85</span>,hjust <span class="op">=</span> <span class="fl">0.75</span><span class="op">)</span></span>
<span><span class="op">)</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integrated_analysis_multiple_single_cell_datasets_1_files/figure-html/unnamed-chunk-24-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>因为接下来我们要提取髓系细胞进一步分析，所以通过几种可视化方法检查一下髓系细胞通用marker（CD163和AIF1）的表达情况：</p>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_1_cache/html/unnamed-chunk-25_32fabb2551b2041650ed7cb583709d06">
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/RidgePlot.html">RidgePlot</a></span><span class="op">(</span><span class="va">seurat_clustered</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CD163"</span>, <span class="st">"AIF1"</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integrated_analysis_multiple_single_cell_datasets_1_files/figure-html/unnamed-chunk-25-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_1_cache/html/unnamed-chunk-26_c5a9b466492dbdf451984c631e01ac9c">
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html">VlnPlot</a></span><span class="op">(</span><span class="va">seurat_clustered</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CD163"</span>, <span class="st">"AIF1"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integrated_analysis_multiple_single_cell_datasets_1_files/figure-html/unnamed-chunk-26-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_1_cache/html/unnamed-chunk-27_4913d8903e3592fae61e5ae0750c0569">
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html">FeaturePlot</a></span><span class="op">(</span><span class="va">seurat_clustered</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CD163"</span>, <span class="st">"AIF1"</span><span class="op">)</span>, label <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integrated_analysis_multiple_single_cell_datasets_1_files/figure-html/unnamed-chunk-27-1.png" class="img-fluid" width="768"></p>
</div>
</div>
<p>保存</p>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_1_cache/html/unnamed-chunk-28_90bcf77b3f3a31701e487f2024e15dd0">
<div class="sourceCode" id="cb77"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">seurat_clustered</span>, </span>
<span>        file <span class="op">=</span> <span class="st">"output/sc_supplementary/GSE150430_seurat_annotated.RDS"</span><span class="op">)</span>; <span class="fu"><a href="https://rdrr.io/pkg/beepr/man/beep.html">beep</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section></section><section id="提取髓系细胞" class="level3"><h3 class="anchored" data-anchor-id="提取髓系细胞">提取髓系细胞</h3>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_1_cache/html/unnamed-chunk-29_b4011549ef17aaad537c77099573618a">
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">sub_seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="va">seurat_clustered</span>, </span>
<span>                     idents <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Macrophages"</span>, </span>
<span>                                <span class="st">"Myeloid (unspecific)"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sub_seurat</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
46357 features across 3997 samples within 2 assays 
Active assay: SCT (21637 features, 3000 variable features)
 3 layers present: counts, data, scale.data
 1 other assay present: RNA
 4 dimensional reductions calculated: pca, umap.unintegrated, harmony, umap.integrated</code></pre>
</div>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">sub_seurat</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "orig.ident"            "nCount_RNA"            "nFeature_RNA"         
 [4] "samples"               "log10GenesPerUMI"      "mitoRatio"            
 [7] "cells"                 "nCount_SCT"            "nFeature_SCT"         
[10] "S.Score"               "G2M.Score"             "Phase"                
[13] "mitoFr"                "unintegrated_clusters" "seurat_clusters"      
[16] "groups"                "integrated_clusters"   "SCT_snn_res.0.01"     
[19] "SCT_snn_res.0.05"      "SCT_snn_res.0.1"       "SCT_snn_res.0.2"      
[22] "SCT_snn_res.0.3"       "SCT_snn_res.0.4"       "SCT_snn_res.0.5"      
[25] "SCT_snn_res.0.8"       "SCT_snn_res.1"        </code></pre>
</div>
<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">myeloid_seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/CreateSeuratObject.html">CreateSeuratObject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">sub_seurat</span><span class="op">@</span><span class="va">assays</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>                                     meta.data <span class="op">=</span> <span class="va">sub_seurat</span><span class="op">@</span><span class="va">meta.data</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">7</span>, <span class="fl">10</span><span class="op">:</span><span class="fl">13</span>, <span class="fl">16</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">myeloid_seurat</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
24720 features across 3997 samples within 1 assay 
Active assay: RNA (24720 features, 0 variable features)
 1 layer present: counts</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Warning
</div>
</div>
<div class="callout-body-container callout-body">
<p>可以看到在重新创建Seurat之后，默认的“RNA” assay之中只有一个layer。为了后面执行<code>SCTransform</code>及整合，<strong>这里需要将其按照样本拆分成不同的layers</strong>。</p>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_1_cache/html/unnamed-chunk-30_af8e0dc4fd8dd7b73beba1db7fc21bbd">
<div class="sourceCode" id="cb84"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">myeloid_seurat</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html">split</a></span><span class="op">(</span><span class="va">myeloid_seurat</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span>, </span>
<span>                                 f <span class="op">=</span> <span class="va">myeloid_seurat</span><span class="op">$</span><span class="va">samples</span><span class="op">)</span></span>
<span><span class="va">myeloid_seurat</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
24720 features across 3997 samples within 1 assay 
Active assay: RNA (24720 features, 0 variable features)
 16 layers present: counts.N01, counts.P01, counts.P02, counts.P03, counts.P04, counts.P05, counts.P06, counts.P07, counts.P08, counts.P09, counts.P10, counts.P11, counts.P12, counts.P13, counts.P14, counts.P15</code></pre>
</div>
</div>
</div>
</div>
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_1_cache/html/unnamed-chunk-31_f6174f0498967057c31a566e67b3e594">
<div class="sourceCode" id="cb86"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 补充原细胞注释信息</span></span>
<span><span class="va">myeloid_seurat</span><span class="op">$</span><span class="va">old_clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/Idents.html">Idents</a></span><span class="op">(</span><span class="va">sub_seurat</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">myeloid_seurat</span><span class="op">$</span><span class="va">old_clusters</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
         Macrophages Myeloid (unspecific) 
                3765                  232 </code></pre>
</div>
<div class="sourceCode" id="cb88"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 保存</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">myeloid_seurat</span>, file <span class="op">=</span> <span class="st">"output/sc_supplementary/GSE150430_myeloid_seurat.RDS"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-3-contents" aria-controls="callout-3" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
如何提取并处理目标细胞亚群
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-3" class="callout-3-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>After we have completed the scRNA-seq workflow and identified the various cell types present in our samples, we might decide that for a particular cell type, we would like to identify subtypes. For example, if we have a large cluster of <strong>CD4+ Helper T cells</strong>, we may want to identify subsets of <strong>Th1, Th2, Th17, Th9, and Tfh cells</strong>. To subset the dataset, Seurat has a handy&nbsp;<code><a href="https://rdrr.io/r/base/subset.html">subset()</a></code>function（ <a href="../seurat/seurat_command_list.html#sec-subset_seurat"><span>Section&nbsp;7.3.1</span></a> ）.</p>
<p>To perform the subclustering, there are a couple of different methods you could try. The easiest would be to run the&nbsp;<code><a href="https://satijalab.org/seurat/reference/FindNeighbors.html">FindNeighbors()</a></code>&nbsp;and&nbsp;<code><a href="https://satijalab.org/seurat/reference/FindClusters.html">FindClusters()</a></code>&nbsp;on the subsetted cells, adjusting the resolution to give you the optimal clustering. However, with this approach <strong>you are not redefining the most variable genes</strong> used to find clusters, so it might not work if the genes delineating these subsets are not those driving any of the top PCs used for the clustering.</p>
<p>Alternatively, we could <strong>start over with the raw counts</strong> for this subset of cells and run&nbsp;<strong><code><a href="https://satijalab.org/seurat/reference/SCTransform.html">SCTransform()</a></code></strong>to determine the greatest sources of variation present. <strong>This would allow us to focus our clustering on the most variant genes present among our subset of cells</strong>. Hopefully, the most variant genes are those driving the various desired subsets (e.g.&nbsp;Th1, Th2, Th17, Th9, and Tfh cells). If <strong>integration</strong> is necessary, then this step would still need to be performed.</p>
<p>Since subsetting the dataset can result in a much smaller number of cells, it is important to consider the total number of cells you are looking to cluster and some of the parameters that might be affected by the small numbers. For example, if integrating, there is a ‘K’ number of cells used for determining the neighborhoods for identifying and filtering anchors. Therefore, if your integration isn’t very good for a small dataset, you might want to consider <strong>lowering the ‘K’ parameter (k.filter、k.weight)</strong> in<code>IntegrateLayers</code>. However, if ‘K’ is too small, it could also lead to poor integration.</p>
<p>（来自：<a href="https://hbctraining.github.io/scRNA-seq_online/lessons/seurat_subclustering.html"><em>Subclustering scRNA-seq datasets</em></a>）</p>
</div>
</div>
</div>
<p>下一节，针对髓系细胞进一步分群。</p>
<hr>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-4-contents" aria-controls="callout-4" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Session Info
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-4" class="callout-4-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-hash="integrated_analysis_multiple_single_cell_datasets_1_cache/html/unnamed-chunk-32_03790cfe1a3993dbb0bf41bacca6698a">
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.3.2 (2023-10-31)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Sonoma 14.2.1

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: Asia/Shanghai
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] clustree_0.5.1     ggraph_2.1.0       dplyr_1.1.4        cowplot_1.1.2     
[5] beepr_1.3          ggplot2_3.4.4      Seurat_5.0.1       SeuratObject_5.0.1
[9] sp_2.1-2          

loaded via a namespace (and not attached):
  [1] RcppAnnoy_0.0.21            splines_4.3.2              
  [3] later_1.3.2                 bitops_1.0-7               
  [5] tibble_3.2.1                polyclip_1.10-6            
  [7] fastDummies_1.7.3           lifecycle_1.0.4            
  [9] globals_0.16.2              lattice_0.22-5             
 [11] MASS_7.3-60.0.1             backports_1.4.1            
 [13] magrittr_2.0.3              plotly_4.10.4              
 [15] rmarkdown_2.25              yaml_2.3.8                 
 [17] httpuv_1.6.13               glmGamPoi_1.14.0           
 [19] sctransform_0.4.1           spam_2.10-0                
 [21] spatstat.sparse_3.0-3       reticulate_1.34.0          
 [23] pbapply_1.7-2               RColorBrewer_1.1-3         
 [25] abind_1.4-5                 zlibbioc_1.48.0            
 [27] audio_0.1-11                Rtsne_0.17                 
 [29] GenomicRanges_1.54.1        purrr_1.0.2                
 [31] BiocGenerics_0.48.1         RCurl_1.98-1.14            
 [33] tweenr_2.0.2                GenomeInfoDbData_1.2.11    
 [35] IRanges_2.36.0              S4Vectors_0.40.2           
 [37] ggrepel_0.9.5               irlba_2.3.5.1              
 [39] listenv_0.9.0               spatstat.utils_3.0-4       
 [41] goftest_1.2-3               RSpectra_0.16-1            
 [43] spatstat.random_3.2-2       fitdistrplus_1.1-11        
 [45] parallelly_1.36.0           DelayedMatrixStats_1.24.0  
 [47] leiden_0.4.3.1              codetools_0.2-19           
 [49] DelayedArray_0.28.0         ggforce_0.4.1              
 [51] tidyselect_1.2.0            farver_2.1.1               
 [53] viridis_0.6.4               matrixStats_1.2.0          
 [55] stats4_4.3.2                spatstat.explore_3.2-5     
 [57] jsonlite_1.8.8              ellipsis_0.3.2             
 [59] tidygraph_1.3.0             progressr_0.14.0           
 [61] ggridges_0.5.5              survival_3.5-7             
 [63] tools_4.3.2                 ica_1.0-3                  
 [65] Rcpp_1.0.12                 glue_1.7.0                 
 [67] gridExtra_2.3               SparseArray_1.2.3          
 [69] xfun_0.41                   mgcv_1.9-1                 
 [71] MatrixGenerics_1.14.0       GenomeInfoDb_1.38.5        
 [73] withr_3.0.0                 fastmap_1.1.1              
 [75] fansi_1.0.6                 digest_0.6.34              
 [77] R6_2.5.1                    mime_0.12                  
 [79] colorspace_2.1-0            scattermore_1.2            
 [81] tensor_1.5                  spatstat.data_3.0-4        
 [83] RhpcBLASctl_0.23-42         utf8_1.2.4                 
 [85] tidyr_1.3.0                 generics_0.1.3             
 [87] data.table_1.14.10          graphlayouts_1.0.2         
 [89] httr_1.4.7                  htmlwidgets_1.6.4          
 [91] S4Arrays_1.2.0              uwot_0.1.16                
 [93] pkgconfig_2.0.3             gtable_0.3.4               
 [95] lmtest_0.9-40               XVector_0.42.0             
 [97] htmltools_0.5.7             dotCall64_1.1-1            
 [99] scales_1.3.0                Biobase_2.62.0             
[101] png_0.1-8                   harmony_1.2.0              
[103] knitr_1.45                  rstudioapi_0.15.0          
[105] reshape2_1.4.4              checkmate_2.3.1            
[107] nlme_3.1-164                zoo_1.8-12                 
[109] stringr_1.5.1               KernSmooth_2.23-22         
[111] vipor_0.4.7                 parallel_4.3.2             
[113] miniUI_0.1.1.1              ggrastr_1.0.2              
[115] pillar_1.9.0                grid_4.3.2                 
[117] vctrs_0.6.5                 RANN_2.6.1                 
[119] promises_1.2.1              xtable_1.8-4               
[121] cluster_2.1.6               beeswarm_0.4.0             
[123] evaluate_0.23               cli_3.6.2                  
[125] compiler_4.3.2              rlang_1.1.3                
[127] crayon_1.5.2                future.apply_1.11.1        
[129] labeling_0.4.3              ggbeeswarm_0.7.2           
[131] plyr_1.8.9                  stringi_1.8.3              
[133] viridisLite_0.4.2           deldir_2.0-2               
[135] munsell_0.5.0               lazyeval_0.2.2             
[137] spatstat.geom_3.2-7         Matrix_1.6-5               
[139] RcppHNSW_0.5.0              patchwork_1.2.0            
[141] sparseMatrixStats_1.14.0    future_1.33.1              
[143] shiny_1.8.0                 SummarizedExperiment_1.32.0
[145] ROCR_1.0-11                 igraph_1.6.0               </code></pre>
</div>
</div>
</div>
</div>
</div>


<!-- -->

<div id="refs" class="references csl-bib-body hanging-indent" role="list" style="display: none">
<div id="ref-gong2023" class="csl-entry" role="listitem">
Gong, Lanqi, Jie Luo, Yu Zhang, Yuma Yang, Shanshan Li, Xiaona Fang, Baifeng Zhang, et al. 2023. <span>“Nasopharyngeal Carcinoma Cells Promote Regulatory T Cell Development and Suppressive Activity via CD70-CD27 Interaction.”</span> <em>Nature Communications</em> 14 (1). <a href="https://doi.org/10.1038/s41467-023-37614-6">https://doi.org/10.1038/s41467-023-37614-6</a>.
</div>
</div>
</section></section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  let localAlternateSentinel = 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var filterRegex = new RegExp("https:\/\/djhcod\.github\.io\/r-notes\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
          // target, if specified
          link.setAttribute("target", "_blank");
      }
    }
});
</script><script src="https://utteranc.es/client.js" repo="https://github.com/djhcod/r-notes" issue-term="pathname" theme="github-light" crossorigin="anonymous" async="">
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="../../single_cell/sc_supplementary/universal_marker.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">29</span>&nbsp; <span class="chapter-title">细胞分群通用marker</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../../single_cell/sc_supplementary/integrated_analysis_multiple_single_cell_datasets_2.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">31</span>&nbsp; <span class="chapter-title">多个单细胞数据集整合分析（下）</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb90" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a><span class="fu"># 多个单细胞数据集整合分析（上） {#sec-multiple_single_cell_datasets_1}</span></span>
<span id="cb90-2"><a href="#cb90-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-3"><a href="#cb90-3" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; 参考：</span><span class="co">[</span><span class="ot">单细胞多数据集整合示例</span><span class="co">](https://mp.weixin.qq.com/s/51PtUi7ZKp1nbrFGElU74A)</span></span>
<span id="cb90-4"><a href="#cb90-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-5"><a href="#cb90-5" aria-hidden="true" tabindex="-1"></a>有时候，为了扩大数据量和得到更加可靠的结论，我们可能获取和下载了多个单细胞数据集。如果我们只关注某一细胞类型，如T cells或B cells或髓系细胞，那么就需要合并分析这些数据集，提取感兴趣的细胞亚群。这时候有两种方法可以选择：</span>
<span id="cb90-6"><a href="#cb90-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-7"><a href="#cb90-7" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>**对各个数据集分别降维分群，然后提取各个数据集的感兴趣细胞亚群并整合到一起**</span>
<span id="cb90-8"><a href="#cb90-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-9"><a href="#cb90-9" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>**在一开始分析的时候就对这些数据集进行整合，然后分群，提起感兴趣细胞群**</span>
<span id="cb90-10"><a href="#cb90-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-11"><a href="#cb90-11" aria-hidden="true" tabindex="-1"></a>**这个部分怎么分析并没有一个特定的答案或者规则，具体还是得看数据集适合以上提出的哪种分析顺序。**</span>
<span id="cb90-12"><a href="#cb90-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-13"><a href="#cb90-13" aria-hidden="true" tabindex="-1"></a>例如<span class="co">[</span><span class="ot">@gong2023</span><span class="co">]</span>就用了第二种方法。选取了三个鼻咽癌单细胞转录组数据： GSE150825, GSE150430, GSE162025。首先对三组数据进行整合，然后选取其中的一部分T细胞进行进一步分析。</span>
<span id="cb90-14"><a href="#cb90-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-15"><a href="#cb90-15" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/截屏2024-01-10%2012.00.48.png)</span>{width="600"}</span>
<span id="cb90-16"><a href="#cb90-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-17"><a href="#cb90-17" aria-hidden="true" tabindex="-1"></a>这里我们介绍第一种方式，即**对各数据集分别降维分群，然后提取并整合各个数据集的感兴趣细胞亚群。**</span>
<span id="cb90-18"><a href="#cb90-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-19"><a href="#cb90-19" aria-hidden="true" tabindex="-1"></a><span class="fu">## 加载包</span></span>
<span id="cb90-20"><a href="#cb90-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-23"><a href="#cb90-23" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb90-24"><a href="#cb90-24" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb90-25"><a href="#cb90-25" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb90-26"><a href="#cb90-26" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(beepr)</span>
<span id="cb90-27"><a href="#cb90-27" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(cowplot)</span>
<span id="cb90-28"><a href="#cb90-28" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb90-29"><a href="#cb90-29" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(clustree)</span>
<span id="cb90-30"><a href="#cb90-30" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb90-31"><a href="#cb90-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-32"><a href="#cb90-32" aria-hidden="true" tabindex="-1"></a><span class="fu">## GSE150430</span></span>
<span id="cb90-33"><a href="#cb90-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-34"><a href="#cb90-34" aria-hidden="true" tabindex="-1"></a>这里我们直接载入 @sec-reading_non_standard_large_volume_text 中构建好的Seurat对象，该案例数据的介绍及读取过程参见该章节。</span>
<span id="cb90-35"><a href="#cb90-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-36"><a href="#cb90-36" aria-hidden="true" tabindex="-1"></a><span class="fu">### 数据导入</span></span>
<span id="cb90-37"><a href="#cb90-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-40"><a href="#cb90-40" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb90-41"><a href="#cb90-41" aria-hidden="true" tabindex="-1"></a>merged_seurat <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"output/sc_supplementary/GSE150430_merged_seurat.rds"</span>)</span>
<span id="cb90-42"><a href="#cb90-42" aria-hidden="true" tabindex="-1"></a>merged_seurat</span>
<span id="cb90-43"><a href="#cb90-43" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(merged_seurat, <span class="dv">4</span>)</span>
<span id="cb90-44"><a href="#cb90-44" aria-hidden="true" tabindex="-1"></a><span class="fu">unique</span>(merged_seurat<span class="sc">$</span>samples)</span>
<span id="cb90-45"><a href="#cb90-45" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb90-46"><a href="#cb90-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-47"><a href="#cb90-47" aria-hidden="true" tabindex="-1"></a><span class="fu">### 质控</span></span>
<span id="cb90-48"><a href="#cb90-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-51"><a href="#cb90-51" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb90-52"><a href="#cb90-52" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 3</span></span>
<span id="cb90-53"><a href="#cb90-53" aria-hidden="true" tabindex="-1"></a><span class="co"># Add number of genes per UMI for each cell to metadata</span></span>
<span id="cb90-54"><a href="#cb90-54" aria-hidden="true" tabindex="-1"></a>merged_seurat<span class="sc">$</span>log10GenesPerUMI <span class="ot">&lt;-</span> <span class="fu">log10</span>(merged_seurat<span class="sc">$</span>nFeature_RNA) <span class="sc">/</span> <span class="fu">log10</span>(merged_seurat<span class="sc">$</span>nCount_RNA)</span>
<span id="cb90-55"><a href="#cb90-55" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(merged_seurat<span class="sc">$</span>log10GenesPerUMI)</span>
<span id="cb90-56"><a href="#cb90-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-57"><a href="#cb90-57" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute percent mito ratio</span></span>
<span id="cb90-58"><a href="#cb90-58" aria-hidden="true" tabindex="-1"></a>merged_seurat<span class="sc">$</span>mitoRatio <span class="ot">&lt;-</span> <span class="fu">PercentageFeatureSet</span>(<span class="at">object =</span> merged_seurat, </span>
<span id="cb90-59"><a href="#cb90-59" aria-hidden="true" tabindex="-1"></a>                                                <span class="at">pattern =</span> <span class="st">"^MT-"</span>)</span>
<span id="cb90-60"><a href="#cb90-60" aria-hidden="true" tabindex="-1"></a>merged_seurat<span class="sc">$</span>mitoRatio <span class="ot">&lt;-</span> merged_seurat<span class="sc">$</span>mitoRatio <span class="sc">/</span> <span class="dv">100</span></span>
<span id="cb90-61"><a href="#cb90-61" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(merged_seurat<span class="sc">$</span>mitoRatio)</span>
<span id="cb90-62"><a href="#cb90-62" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(merged_seurat<span class="sc">$</span>mitoRatio)</span>
<span id="cb90-63"><a href="#cb90-63" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-64"><a href="#cb90-64" aria-hidden="true" tabindex="-1"></a><span class="co"># Add cell IDs to metadata</span></span>
<span id="cb90-65"><a href="#cb90-65" aria-hidden="true" tabindex="-1"></a>merged_seurat<span class="sc">$</span>cells <span class="ot">&lt;-</span> <span class="fu">rownames</span>(merged_seurat<span class="sc">@</span>meta.data)</span>
<span id="cb90-66"><a href="#cb90-66" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb90-67"><a href="#cb90-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-70"><a href="#cb90-70" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb90-71"><a href="#cb90-71" aria-hidden="true" tabindex="-1"></a><span class="co">#| layout-ncol: 2</span></span>
<span id="cb90-72"><a href="#cb90-72" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualize the number of cell counts per sample</span></span>
<span id="cb90-73"><a href="#cb90-73" aria-hidden="true" tabindex="-1"></a>merged_seurat<span class="sc">@</span>meta.data  <span class="sc">|&gt;</span> </span>
<span id="cb90-74"><a href="#cb90-74" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> samples, <span class="at">fill =</span> samples)) <span class="sc">+</span> </span>
<span id="cb90-75"><a href="#cb90-75" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_bar</span>() <span class="sc">+</span></span>
<span id="cb90-76"><a href="#cb90-76" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb90-77"><a href="#cb90-77" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">45</span>, <span class="at">vjust =</span> <span class="dv">1</span>, <span class="at">hjust =</span> <span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb90-78"><a href="#cb90-78" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>, <span class="at">face =</span> <span class="st">"bold"</span>)) <span class="sc">+</span></span>
<span id="cb90-79"><a href="#cb90-79" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"NCells"</span>)</span>
<span id="cb90-80"><a href="#cb90-80" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-81"><a href="#cb90-81" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualize the number UMIs/transcripts per cell</span></span>
<span id="cb90-82"><a href="#cb90-82" aria-hidden="true" tabindex="-1"></a>merged_seurat<span class="sc">@</span>meta.data <span class="sc">|&gt;</span> </span>
<span id="cb90-83"><a href="#cb90-83" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">color =</span> samples, <span class="at">x =</span> nCount_RNA, <span class="at">fill =</span> samples)) <span class="sc">+</span> </span>
<span id="cb90-84"><a href="#cb90-84" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_density</span>(<span class="at">alpha =</span> <span class="fl">0.2</span>) <span class="sc">+</span> </span>
<span id="cb90-85"><a href="#cb90-85" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_x_log10</span>() <span class="sc">+</span> </span>
<span id="cb90-86"><a href="#cb90-86" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb90-87"><a href="#cb90-87" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ylab</span>(<span class="st">"Cell density"</span>) <span class="sc">+</span></span>
<span id="cb90-88"><a href="#cb90-88" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">500</span>)</span>
<span id="cb90-89"><a href="#cb90-89" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-90"><a href="#cb90-90" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualize the distribution of genes detected per cell via histogram</span></span>
<span id="cb90-91"><a href="#cb90-91" aria-hidden="true" tabindex="-1"></a>merged_seurat<span class="sc">@</span>meta.data <span class="sc">|&gt;</span></span>
<span id="cb90-92"><a href="#cb90-92" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">color =</span> samples, <span class="at">x =</span> nFeature_RNA, <span class="at">fill=</span> samples)) <span class="sc">+</span> </span>
<span id="cb90-93"><a href="#cb90-93" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_density</span>(<span class="at">alpha =</span> <span class="fl">0.2</span>) <span class="sc">+</span> </span>
<span id="cb90-94"><a href="#cb90-94" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb90-95"><a href="#cb90-95" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_x_log10</span>() <span class="sc">+</span> </span>
<span id="cb90-96"><a href="#cb90-96" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">200</span>)</span>
<span id="cb90-97"><a href="#cb90-97" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-98"><a href="#cb90-98" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualize the overall complexity of the gene expression by visualizing the genes detected per UMI (novelty score)</span></span>
<span id="cb90-99"><a href="#cb90-99" aria-hidden="true" tabindex="-1"></a>merged_seurat<span class="sc">@</span>meta.data <span class="sc">|&gt;</span></span>
<span id="cb90-100"><a href="#cb90-100" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> log10GenesPerUMI, <span class="at">color =</span> samples, <span class="at">fill=</span>samples)) <span class="sc">+</span></span>
<span id="cb90-101"><a href="#cb90-101" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_density</span>(<span class="at">alpha =</span> <span class="fl">0.2</span>) <span class="sc">+</span></span>
<span id="cb90-102"><a href="#cb90-102" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb90-103"><a href="#cb90-103" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fl">0.8</span>)</span>
<span id="cb90-104"><a href="#cb90-104" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-105"><a href="#cb90-105" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualize the distribution of mitochondrial gene expression detected per cell</span></span>
<span id="cb90-106"><a href="#cb90-106" aria-hidden="true" tabindex="-1"></a>merged_seurat<span class="sc">@</span>meta.data <span class="sc">|&gt;</span></span>
<span id="cb90-107"><a href="#cb90-107" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">color =</span> samples, <span class="at">x =</span> mitoRatio, <span class="at">fill =</span> samples)) <span class="sc">+</span> </span>
<span id="cb90-108"><a href="#cb90-108" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_density</span>(<span class="at">alpha =</span> <span class="fl">0.2</span>) <span class="sc">+</span> </span>
<span id="cb90-109"><a href="#cb90-109" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_x_log10</span>() <span class="sc">+</span> </span>
<span id="cb90-110"><a href="#cb90-110" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb90-111"><a href="#cb90-111" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="fl">0.2</span>)</span>
<span id="cb90-112"><a href="#cb90-112" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb90-113"><a href="#cb90-113" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-116"><a href="#cb90-116" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb90-117"><a href="#cb90-117" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 10</span></span>
<span id="cb90-118"><a href="#cb90-118" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 10</span></span>
<span id="cb90-119"><a href="#cb90-119" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualize the correlation between genes detected and number of UMIs and determine whether strong presence of cells with low numbers of genes/UMIs</span></span>
<span id="cb90-120"><a href="#cb90-120" aria-hidden="true" tabindex="-1"></a>merged_seurat<span class="sc">@</span>meta.data <span class="sc">|&gt;</span> </span>
<span id="cb90-121"><a href="#cb90-121" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> nCount_RNA, <span class="at">y =</span> nFeature_RNA, <span class="at">color =</span> mitoRatio)) <span class="sc">+</span> </span>
<span id="cb90-122"><a href="#cb90-122" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>() <span class="sc">+</span> </span>
<span id="cb90-123"><a href="#cb90-123" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_colour_gradient</span>(<span class="at">low =</span> <span class="st">"gray90"</span>, <span class="at">high =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb90-124"><a href="#cb90-124" aria-hidden="true" tabindex="-1"></a>    <span class="fu">stat_smooth</span>(<span class="at">method =</span> lm) <span class="sc">+</span></span>
<span id="cb90-125"><a href="#cb90-125" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_x_log10</span>() <span class="sc">+</span> </span>
<span id="cb90-126"><a href="#cb90-126" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_y_log10</span>() <span class="sc">+</span> </span>
<span id="cb90-127"><a href="#cb90-127" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb90-128"><a href="#cb90-128" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">500</span>) <span class="sc">+</span></span>
<span id="cb90-129"><a href="#cb90-129" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">250</span>) <span class="sc">+</span></span>
<span id="cb90-130"><a href="#cb90-130" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_wrap</span>(<span class="sc">~</span>samples)</span>
<span id="cb90-131"><a href="#cb90-131" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb90-132"><a href="#cb90-132" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-133"><a href="#cb90-133" aria-hidden="true" tabindex="-1"></a>可以看到该数据集已经进行了质控，各项指标均在正常范围内，因此，我们可以跳过细胞/基因过滤，直接进入下面的环节。</span>
<span id="cb90-134"><a href="#cb90-134" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-135"><a href="#cb90-135" aria-hidden="true" tabindex="-1"></a><span class="fu">### 归一化及消除非期望变异来源</span></span>
<span id="cb90-136"><a href="#cb90-136" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-137"><a href="#cb90-137" aria-hidden="true" tabindex="-1"></a>::: callout-caution</span>
<span id="cb90-138"><a href="#cb90-138" aria-hidden="true" tabindex="-1"></a>由于这里<span class="in">`SCTransform`</span>的数据量太大，在我的16GB MacBook Pro上超出了R内存分配上限，出现了“Error: vector memory exhausted (limit reached?)”报错。因此，这里在运行下面的脚本之前通过 @sec-Raising_memory_limit 的方法进行了处理。</span>
<span id="cb90-139"><a href="#cb90-139" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb90-140"><a href="#cb90-140" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-143"><a href="#cb90-143" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb90-144"><a href="#cb90-144" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb90-145"><a href="#cb90-145" aria-hidden="true" tabindex="-1"></a><span class="co"># SCTranform</span></span>
<span id="cb90-146"><a href="#cb90-146" aria-hidden="true" tabindex="-1"></a>merged_seurat <span class="ot">&lt;-</span> <span class="fu">SCTransform</span>(merged_seurat, <span class="at">verbose =</span> <span class="cn">FALSE</span>); <span class="fu">beep</span>()</span>
<span id="cb90-147"><a href="#cb90-147" aria-hidden="true" tabindex="-1"></a><span class="fu">gc</span>() <span class="co"># 释放未使用内存</span></span>
<span id="cb90-148"><a href="#cb90-148" aria-hidden="true" tabindex="-1"></a>merged_seurat</span>
<span id="cb90-149"><a href="#cb90-149" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-150"><a href="#cb90-150" aria-hidden="true" tabindex="-1"></a><span class="co"># Check which assays are stored in objects</span></span>
<span id="cb90-151"><a href="#cb90-151" aria-hidden="true" tabindex="-1"></a>merged_seurat<span class="sc">@</span>assays</span>
<span id="cb90-152"><a href="#cb90-152" aria-hidden="true" tabindex="-1"></a><span class="co"># 查看目前默认的assay</span></span>
<span id="cb90-153"><a href="#cb90-153" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(merged_seurat)</span>
<span id="cb90-154"><a href="#cb90-154" aria-hidden="true" tabindex="-1"></a><span class="co"># 查看默认assay的layers</span></span>
<span id="cb90-155"><a href="#cb90-155" aria-hidden="true" tabindex="-1"></a><span class="fu">Layers</span>(merged_seurat)</span>
<span id="cb90-156"><a href="#cb90-156" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb90-157"><a href="#cb90-157" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-158"><a href="#cb90-158" aria-hidden="true" tabindex="-1"></a><span class="fu">#### 评估细胞周期的影响</span></span>
<span id="cb90-159"><a href="#cb90-159" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-162"><a href="#cb90-162" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb90-163"><a href="#cb90-163" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb90-164"><a href="#cb90-164" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 10</span></span>
<span id="cb90-165"><a href="#cb90-165" aria-hidden="true" tabindex="-1"></a><span class="co"># Load cell cycle markers</span></span>
<span id="cb90-166"><a href="#cb90-166" aria-hidden="true" tabindex="-1"></a>s.genes <span class="ot">&lt;-</span> cc.genes<span class="sc">$</span>s.genes</span>
<span id="cb90-167"><a href="#cb90-167" aria-hidden="true" tabindex="-1"></a>g2m.genes <span class="ot">&lt;-</span> cc.genes<span class="sc">$</span>g2m.genes</span>
<span id="cb90-168"><a href="#cb90-168" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-169"><a href="#cb90-169" aria-hidden="true" tabindex="-1"></a><span class="co"># Score cells for cell cycle</span></span>
<span id="cb90-170"><a href="#cb90-170" aria-hidden="true" tabindex="-1"></a>merged_seurat <span class="ot">&lt;-</span> <span class="fu">CellCycleScoring</span>(merged_seurat, </span>
<span id="cb90-171"><a href="#cb90-171" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">g2m.features =</span> g2m.genes, </span>
<span id="cb90-172"><a href="#cb90-172" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">s.features =</span> s.genes)</span>
<span id="cb90-173"><a href="#cb90-173" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-174"><a href="#cb90-174" aria-hidden="true" tabindex="-1"></a><span class="co"># 现在的meta.data中多出了细胞周期评分“S.Score”和“G2M.Score”，以及推断的细胞所处的周期“Phase”</span></span>
<span id="cb90-175"><a href="#cb90-175" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(merged_seurat<span class="sc">@</span>meta.data)</span>
<span id="cb90-176"><a href="#cb90-176" aria-hidden="true" tabindex="-1"></a><span class="co"># 查看一下细胞周期的分布情况</span></span>
<span id="cb90-177"><a href="#cb90-177" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(merged_seurat<span class="sc">$</span>Phase)</span>
<span id="cb90-178"><a href="#cb90-178" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-179"><a href="#cb90-179" aria-hidden="true" tabindex="-1"></a><span class="co"># 执行PCA</span></span>
<span id="cb90-180"><a href="#cb90-180" aria-hidden="true" tabindex="-1"></a>merged_seurat <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(merged_seurat)</span>
<span id="cb90-181"><a href="#cb90-181" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-182"><a href="#cb90-182" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the PCA colored by cell cycle phase</span></span>
<span id="cb90-183"><a href="#cb90-183" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(merged_seurat,</span>
<span id="cb90-184"><a href="#cb90-184" aria-hidden="true" tabindex="-1"></a>              <span class="at">reduction =</span> <span class="st">"pca"</span>,</span>
<span id="cb90-185"><a href="#cb90-185" aria-hidden="true" tabindex="-1"></a>              <span class="at">group.by=</span> <span class="st">"Phase"</span>)</span>
<span id="cb90-186"><a href="#cb90-186" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(merged_seurat,</span>
<span id="cb90-187"><a href="#cb90-187" aria-hidden="true" tabindex="-1"></a>              <span class="at">reduction =</span> <span class="st">"pca"</span>,</span>
<span id="cb90-188"><a href="#cb90-188" aria-hidden="true" tabindex="-1"></a>              <span class="at">group.by=</span> <span class="st">"Phase"</span>,</span>
<span id="cb90-189"><a href="#cb90-189" aria-hidden="true" tabindex="-1"></a>              <span class="at">split.by =</span> <span class="st">"Phase"</span>)</span>
<span id="cb90-190"><a href="#cb90-190" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_grid</span>(p1, p2, <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">labels =</span> <span class="st">"AUTO"</span>)</span>
<span id="cb90-191"><a href="#cb90-191" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb90-192"><a href="#cb90-192" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-193"><a href="#cb90-193" aria-hidden="true" tabindex="-1"></a>可以看到细胞周期不是变异来源。</span>
<span id="cb90-194"><a href="#cb90-194" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-195"><a href="#cb90-195" aria-hidden="true" tabindex="-1"></a><span class="fu">#### **评估线粒体基因的影响**</span></span>
<span id="cb90-196"><a href="#cb90-196" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-199"><a href="#cb90-199" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb90-200"><a href="#cb90-200" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb90-201"><a href="#cb90-201" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 3</span></span>
<span id="cb90-202"><a href="#cb90-202" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 3</span></span>
<span id="cb90-203"><a href="#cb90-203" aria-hidden="true" tabindex="-1"></a><span class="co"># Check quartile values</span></span>
<span id="cb90-204"><a href="#cb90-204" aria-hidden="true" tabindex="-1"></a>mito_sum <span class="ot">&lt;-</span> <span class="fu">summary</span>(merged_seurat<span class="sc">$</span>mitoRatio)</span>
<span id="cb90-205"><a href="#cb90-205" aria-hidden="true" tabindex="-1"></a>mito_sum</span>
<span id="cb90-206"><a href="#cb90-206" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-207"><a href="#cb90-207" aria-hidden="true" tabindex="-1"></a><span class="co"># Turn mitoRatio into categorical factor vector based on quartile values</span></span>
<span id="cb90-208"><a href="#cb90-208" aria-hidden="true" tabindex="-1"></a>merged_seurat<span class="sc">$</span>mitoFr <span class="ot">&lt;-</span> <span class="fu">cut</span>(merged_seurat<span class="sc">$</span>mitoRatio, </span>
<span id="cb90-209"><a href="#cb90-209" aria-hidden="true" tabindex="-1"></a>                            <span class="at">breaks=</span><span class="fu">c</span>(<span class="sc">-</span><span class="cn">Inf</span>, mito_sum[<span class="dv">2</span>], mito_sum[<span class="dv">3</span>], mito_sum[<span class="dv">5</span>], <span class="cn">Inf</span>),</span>
<span id="cb90-210"><a href="#cb90-210" aria-hidden="true" tabindex="-1"></a>                            <span class="at">labels=</span><span class="fu">c</span>(<span class="st">"Low"</span>, <span class="st">"Medium"</span>, <span class="st">"Medium high"</span>, <span class="st">"High"</span>))</span>
<span id="cb90-211"><a href="#cb90-211" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(merged_seurat<span class="sc">$</span>mitoFr)</span>
<span id="cb90-212"><a href="#cb90-212" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb90-213"><a href="#cb90-213" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-216"><a href="#cb90-216" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb90-217"><a href="#cb90-217" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb90-218"><a href="#cb90-218" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 15</span></span>
<span id="cb90-219"><a href="#cb90-219" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the PCA colored by cell cycle phase</span></span>
<span id="cb90-220"><a href="#cb90-220" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(merged_seurat,</span>
<span id="cb90-221"><a href="#cb90-221" aria-hidden="true" tabindex="-1"></a>              <span class="at">reduction =</span> <span class="st">"pca"</span>,</span>
<span id="cb90-222"><a href="#cb90-222" aria-hidden="true" tabindex="-1"></a>              <span class="at">group.by=</span> <span class="st">"mitoFr"</span>)</span>
<span id="cb90-223"><a href="#cb90-223" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(merged_seurat,</span>
<span id="cb90-224"><a href="#cb90-224" aria-hidden="true" tabindex="-1"></a>              <span class="at">reduction =</span> <span class="st">"pca"</span>,</span>
<span id="cb90-225"><a href="#cb90-225" aria-hidden="true" tabindex="-1"></a>              <span class="at">group.by=</span> <span class="st">"mitoFr"</span>,</span>
<span id="cb90-226"><a href="#cb90-226" aria-hidden="true" tabindex="-1"></a>              <span class="at">split.by =</span> <span class="st">"mitoFr"</span>)</span>
<span id="cb90-227"><a href="#cb90-227" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_grid</span>(p1, p2, <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">labels =</span> <span class="st">"AUTO"</span>)</span>
<span id="cb90-228"><a href="#cb90-228" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb90-229"><a href="#cb90-229" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-230"><a href="#cb90-230" aria-hidden="true" tabindex="-1"></a>可以看到线粒体基因比例不是变异来源。</span>
<span id="cb90-231"><a href="#cb90-231" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-232"><a href="#cb90-232" aria-hidden="true" tabindex="-1"></a>由于细胞周期和线粒体基因比例都不是非期望变异来源，所以我们这里不需要再次运行SCTransform来回归这些变量。接下来，直接进入数据整合环节。</span>
<span id="cb90-233"><a href="#cb90-233" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-234"><a href="#cb90-234" aria-hidden="true" tabindex="-1"></a><span class="fu">### 数据整合</span></span>
<span id="cb90-235"><a href="#cb90-235" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-236"><a href="#cb90-236" aria-hidden="true" tabindex="-1"></a><span class="fu">#### **不进行整合时检验细胞分群情况**</span></span>
<span id="cb90-237"><a href="#cb90-237" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-240"><a href="#cb90-240" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb90-241"><a href="#cb90-241" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb90-242"><a href="#cb90-242" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 15</span></span>
<span id="cb90-243"><a href="#cb90-243" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 15</span></span>
<span id="cb90-244"><a href="#cb90-244" aria-hidden="true" tabindex="-1"></a><span class="co"># 查看降维信息</span></span>
<span id="cb90-245"><a href="#cb90-245" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(merged_seurat<span class="sc">@</span>reductions)</span>
<span id="cb90-246"><a href="#cb90-246" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-247"><a href="#cb90-247" aria-hidden="true" tabindex="-1"></a><span class="co"># Run UMAP</span></span>
<span id="cb90-248"><a href="#cb90-248" aria-hidden="true" tabindex="-1"></a>merged_seurat <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(merged_seurat, </span>
<span id="cb90-249"><a href="#cb90-249" aria-hidden="true" tabindex="-1"></a>                         <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">40</span>, </span>
<span id="cb90-250"><a href="#cb90-250" aria-hidden="true" tabindex="-1"></a>                         <span class="at">reduction =</span> <span class="st">"pca"</span>,</span>
<span id="cb90-251"><a href="#cb90-251" aria-hidden="true" tabindex="-1"></a>                         <span class="at">reduction.name =</span> <span class="st">"umap.unintegrated"</span>); <span class="fu">beep</span>()</span>
<span id="cb90-252"><a href="#cb90-252" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-253"><a href="#cb90-253" aria-hidden="true" tabindex="-1"></a><span class="co"># 分群</span></span>
<span id="cb90-254"><a href="#cb90-254" aria-hidden="true" tabindex="-1"></a><span class="co"># Determine the K-nearest neighbor graph</span></span>
<span id="cb90-255"><a href="#cb90-255" aria-hidden="true" tabindex="-1"></a>merged_seurat <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(merged_seurat, </span>
<span id="cb90-256"><a href="#cb90-256" aria-hidden="true" tabindex="-1"></a>                               <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">40</span>, </span>
<span id="cb90-257"><a href="#cb90-257" aria-hidden="true" tabindex="-1"></a>                               <span class="at">reduction =</span> <span class="st">"pca"</span>)</span>
<span id="cb90-258"><a href="#cb90-258" aria-hidden="true" tabindex="-1"></a>merged_seurat <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(merged_seurat, </span>
<span id="cb90-259"><a href="#cb90-259" aria-hidden="true" tabindex="-1"></a>                              <span class="at">cluster.name =</span> <span class="st">"unintegrated_clusters"</span>)</span>
<span id="cb90-260"><a href="#cb90-260" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-261"><a href="#cb90-261" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot UMAP</span></span>
<span id="cb90-262"><a href="#cb90-262" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(merged_seurat, </span>
<span id="cb90-263"><a href="#cb90-263" aria-hidden="true" tabindex="-1"></a>              <span class="at">reduction =</span> <span class="st">"umap.unintegrated"</span>,</span>
<span id="cb90-264"><a href="#cb90-264" aria-hidden="true" tabindex="-1"></a>              <span class="at">group.by =</span> <span class="st">"samples"</span>)</span>
<span id="cb90-265"><a href="#cb90-265" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(merged_seurat, </span>
<span id="cb90-266"><a href="#cb90-266" aria-hidden="true" tabindex="-1"></a>              <span class="at">reduction =</span> <span class="st">"umap.unintegrated"</span>,</span>
<span id="cb90-267"><a href="#cb90-267" aria-hidden="true" tabindex="-1"></a>              <span class="at">split.by =</span> <span class="st">"samples"</span>)</span>
<span id="cb90-268"><a href="#cb90-268" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_grid</span>(p1, p2, </span>
<span id="cb90-269"><a href="#cb90-269" aria-hidden="true" tabindex="-1"></a>          <span class="at">ncol =</span> <span class="dv">1</span>, <span class="at">labels =</span> <span class="st">"AUTO"</span>)</span>
<span id="cb90-270"><a href="#cb90-270" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-271"><a href="#cb90-271" aria-hidden="true" tabindex="-1"></a><span class="co"># 由于样本数较多，我们再按照样本类型“Normal” vs. “Cancer”画一下UMAP图</span></span>
<span id="cb90-272"><a href="#cb90-272" aria-hidden="true" tabindex="-1"></a>merged_seurat<span class="sc">$</span>groups <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(merged_seurat<span class="sc">$</span>samples <span class="sc">==</span> <span class="st">"N01"</span>, <span class="st">"Normal"</span>, <span class="st">"Cancer"</span>)</span>
<span id="cb90-273"><a href="#cb90-273" aria-hidden="true" tabindex="-1"></a>p1a <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(merged_seurat, </span>
<span id="cb90-274"><a href="#cb90-274" aria-hidden="true" tabindex="-1"></a>               <span class="at">reduction =</span> <span class="st">"umap.unintegrated"</span>,</span>
<span id="cb90-275"><a href="#cb90-275" aria-hidden="true" tabindex="-1"></a>               <span class="at">group.by =</span> <span class="st">"groups"</span>)</span>
<span id="cb90-276"><a href="#cb90-276" aria-hidden="true" tabindex="-1"></a>p2a <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(merged_seurat, </span>
<span id="cb90-277"><a href="#cb90-277" aria-hidden="true" tabindex="-1"></a>               <span class="at">reduction =</span> <span class="st">"umap.unintegrated"</span>,</span>
<span id="cb90-278"><a href="#cb90-278" aria-hidden="true" tabindex="-1"></a>               <span class="at">split.by =</span> <span class="st">"groups"</span>)</span>
<span id="cb90-279"><a href="#cb90-279" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_grid</span>(p1a, p2a, </span>
<span id="cb90-280"><a href="#cb90-280" aria-hidden="true" tabindex="-1"></a>          <span class="at">ncol =</span> <span class="dv">1</span>, <span class="at">labels =</span> <span class="st">"AUTO"</span>)</span>
<span id="cb90-281"><a href="#cb90-281" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb90-282"><a href="#cb90-282" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-283"><a href="#cb90-283" aria-hidden="true" tabindex="-1"></a><span class="fu">#### 整合</span></span>
<span id="cb90-284"><a href="#cb90-284" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-285"><a href="#cb90-285" aria-hidden="true" tabindex="-1"></a>这里我们用Harmony整合算法。</span>
<span id="cb90-286"><a href="#cb90-286" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-289"><a href="#cb90-289" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb90-290"><a href="#cb90-290" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb90-291"><a href="#cb90-291" aria-hidden="true" tabindex="-1"></a><span class="co"># 整合，比较耗时间，进度条会一直显示0%直至运算完成</span></span>
<span id="cb90-292"><a href="#cb90-292" aria-hidden="true" tabindex="-1"></a>seurat_integrated <span class="ot">&lt;-</span> <span class="fu">IntegrateLayers</span>(<span class="at">object =</span> merged_seurat,</span>
<span id="cb90-293"><a href="#cb90-293" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">method =</span> HarmonyIntegration,</span>
<span id="cb90-294"><a href="#cb90-294" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">assay =</span> <span class="st">"SCT"</span>, <span class="co"># Integrating SCTransformed data</span></span>
<span id="cb90-295"><a href="#cb90-295" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">orig.reduction =</span> <span class="st">"pca"</span>,</span>
<span id="cb90-296"><a href="#cb90-296" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">verbose =</span> <span class="cn">FALSE</span>); <span class="fu">beep</span>()</span>
<span id="cb90-297"><a href="#cb90-297" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-298"><a href="#cb90-298" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-299"><a href="#cb90-299" aria-hidden="true" tabindex="-1"></a><span class="co"># 整合后合并RNA layer</span></span>
<span id="cb90-300"><a href="#cb90-300" aria-hidden="true" tabindex="-1"></a>seurat_integrated[[<span class="st">"RNA"</span>]] <span class="ot">&lt;-</span> <span class="fu">JoinLayers</span>(seurat_integrated[[<span class="st">"RNA"</span>]])</span>
<span id="cb90-301"><a href="#cb90-301" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-302"><a href="#cb90-302" aria-hidden="true" tabindex="-1"></a><span class="co"># 查看整合后的降维信息</span></span>
<span id="cb90-303"><a href="#cb90-303" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(seurat_integrated<span class="sc">@</span>reductions)</span>
<span id="cb90-304"><a href="#cb90-304" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb90-305"><a href="#cb90-305" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-306"><a href="#cb90-306" aria-hidden="true" tabindex="-1"></a><span class="fu">#### **整合后检验细胞分群情况**</span></span>
<span id="cb90-307"><a href="#cb90-307" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-310"><a href="#cb90-310" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb90-311"><a href="#cb90-311" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb90-312"><a href="#cb90-312" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 10</span></span>
<span id="cb90-313"><a href="#cb90-313" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 20</span></span>
<span id="cb90-314"><a href="#cb90-314" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123456</span>)</span>
<span id="cb90-315"><a href="#cb90-315" aria-hidden="true" tabindex="-1"></a><span class="co"># Run UMAP</span></span>
<span id="cb90-316"><a href="#cb90-316" aria-hidden="true" tabindex="-1"></a>seurat_integrated <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(seurat_integrated, </span>
<span id="cb90-317"><a href="#cb90-317" aria-hidden="true" tabindex="-1"></a>                             <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">40</span>,</span>
<span id="cb90-318"><a href="#cb90-318" aria-hidden="true" tabindex="-1"></a>                             <span class="at">reduction =</span> <span class="st">"harmony"</span>, <span class="co"># 更改降维来源为整合后的"harmony"</span></span>
<span id="cb90-319"><a href="#cb90-319" aria-hidden="true" tabindex="-1"></a>                             <span class="at">reduction.name =</span> <span class="st">"umap.integrated"</span>); <span class="fu">beep</span>()</span>
<span id="cb90-320"><a href="#cb90-320" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(seurat_integrated<span class="sc">@</span>reductions)</span>
<span id="cb90-321"><a href="#cb90-321" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-322"><a href="#cb90-322" aria-hidden="true" tabindex="-1"></a><span class="co"># 分群</span></span>
<span id="cb90-323"><a href="#cb90-323" aria-hidden="true" tabindex="-1"></a>seurat_integrated <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(seurat_integrated, </span>
<span id="cb90-324"><a href="#cb90-324" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">40</span>, </span>
<span id="cb90-325"><a href="#cb90-325" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">reduction =</span> <span class="st">"harmony"</span>) <span class="co">#更改降维来源为"harmony"</span></span>
<span id="cb90-326"><a href="#cb90-326" aria-hidden="true" tabindex="-1"></a>seurat_integrated <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(seurat_integrated, </span>
<span id="cb90-327"><a href="#cb90-327" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">cluster.name =</span> <span class="st">"integrated_clusters"</span>)</span>
<span id="cb90-328"><a href="#cb90-328" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(seurat_integrated<span class="sc">@</span>meta.data)</span>
<span id="cb90-329"><a href="#cb90-329" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-330"><a href="#cb90-330" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot UMAP                             </span></span>
<span id="cb90-331"><a href="#cb90-331" aria-hidden="true" tabindex="-1"></a>p3 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(seurat_integrated, </span>
<span id="cb90-332"><a href="#cb90-332" aria-hidden="true" tabindex="-1"></a>              <span class="at">reduction =</span> <span class="st">"umap.integrated"</span>, </span>
<span id="cb90-333"><a href="#cb90-333" aria-hidden="true" tabindex="-1"></a>              <span class="at">group.by =</span> <span class="st">"samples"</span>)</span>
<span id="cb90-334"><a href="#cb90-334" aria-hidden="true" tabindex="-1"></a>p4 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(seurat_integrated, </span>
<span id="cb90-335"><a href="#cb90-335" aria-hidden="true" tabindex="-1"></a>              <span class="at">reduction =</span> <span class="st">"umap.integrated"</span>, </span>
<span id="cb90-336"><a href="#cb90-336" aria-hidden="true" tabindex="-1"></a>              <span class="at">split.by =</span> <span class="st">"samples"</span>)</span>
<span id="cb90-337"><a href="#cb90-337" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_grid</span>(p1, p3, p2, p4, </span>
<span id="cb90-338"><a href="#cb90-338" aria-hidden="true" tabindex="-1"></a>          <span class="at">ncol =</span> <span class="dv">1</span>, </span>
<span id="cb90-339"><a href="#cb90-339" aria-hidden="true" tabindex="-1"></a>          <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">"Before Harmony"</span>, <span class="st">"After Harmony"</span>, </span>
<span id="cb90-340"><a href="#cb90-340" aria-hidden="true" tabindex="-1"></a>                     <span class="st">"Before Harmony"</span>, <span class="st">"After Harmony"</span>))</span>
<span id="cb90-341"><a href="#cb90-341" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-342"><a href="#cb90-342" aria-hidden="true" tabindex="-1"></a><span class="co"># 保存</span></span>
<span id="cb90-343"><a href="#cb90-343" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(seurat_integrated, </span>
<span id="cb90-344"><a href="#cb90-344" aria-hidden="true" tabindex="-1"></a>        <span class="at">file =</span> <span class="st">"output/sc_supplementary/GSE150430_seurat_integrated.RDS"</span>); <span class="fu">beep</span>()</span>
<span id="cb90-345"><a href="#cb90-345" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb90-346"><a href="#cb90-346" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-347"><a href="#cb90-347" aria-hidden="true" tabindex="-1"></a>------------------------------------------------------------------------</span>
<span id="cb90-348"><a href="#cb90-348" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-349"><a href="#cb90-349" aria-hidden="true" tabindex="-1"></a><span class="fu">### 聚类</span></span>
<span id="cb90-350"><a href="#cb90-350" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-353"><a href="#cb90-353" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb90-354"><a href="#cb90-354" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb90-355"><a href="#cb90-355" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(<span class="at">list =</span> <span class="fu">ls</span>())</span>
<span id="cb90-356"><a href="#cb90-356" aria-hidden="true" tabindex="-1"></a><span class="fu">gc</span>()</span>
<span id="cb90-357"><a href="#cb90-357" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-358"><a href="#cb90-358" aria-hidden="true" tabindex="-1"></a>seurat_integrated <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"output/sc_supplementary/GSE150430_seurat_integrated.RDS"</span>)</span>
<span id="cb90-359"><a href="#cb90-359" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb90-360"><a href="#cb90-360" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-363"><a href="#cb90-363" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb90-364"><a href="#cb90-364" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb90-365"><a href="#cb90-365" aria-hidden="true" tabindex="-1"></a><span class="co"># Determine the clusters for various resolutions                                </span></span>
<span id="cb90-366"><a href="#cb90-366" aria-hidden="true" tabindex="-1"></a>seurat_integrated <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(seurat_integrated,</span>
<span id="cb90-367"><a href="#cb90-367" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">resolution =</span> <span class="fu">c</span>(<span class="fl">0.01</span>, <span class="fl">0.05</span>, <span class="fl">0.1</span>, <span class="fl">0.2</span>, <span class="fl">0.3</span>, <span class="fl">0.4</span>, <span class="fl">0.5</span>, <span class="fl">0.8</span>, <span class="dv">1</span>),</span>
<span id="cb90-368"><a href="#cb90-368" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">verbose =</span> F); <span class="fu">beep</span>()</span>
<span id="cb90-369"><a href="#cb90-369" aria-hidden="true" tabindex="-1"></a><span class="co"># Explore resolutions</span></span>
<span id="cb90-370"><a href="#cb90-370" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(seurat_integrated<span class="sc">@</span>meta.data, <span class="dv">5</span>)</span>
<span id="cb90-371"><a href="#cb90-371" aria-hidden="true" tabindex="-1"></a><span class="co"># 查看各个分辨率下的细胞分群情况</span></span>
<span id="cb90-372"><a href="#cb90-372" aria-hidden="true" tabindex="-1"></a><span class="fu">select</span>(seurat_integrated<span class="sc">@</span>meta.data, <span class="fu">starts_with</span>(<span class="at">match =</span> <span class="st">"SCT_snn_res."</span>)) <span class="sc">%&gt;%</span>  </span>
<span id="cb90-373"><a href="#cb90-373" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lapply</span>(levels)</span>
<span id="cb90-374"><a href="#cb90-374" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb90-375"><a href="#cb90-375" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-376"><a href="#cb90-376" aria-hidden="true" tabindex="-1"></a>绘制聚类树展示不同分辨率下的细胞分群情况及相互关系</span>
<span id="cb90-377"><a href="#cb90-377" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-380"><a href="#cb90-380" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb90-381"><a href="#cb90-381" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb90-382"><a href="#cb90-382" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 10</span></span>
<span id="cb90-383"><a href="#cb90-383" aria-hidden="true" tabindex="-1"></a>tree <span class="ot">&lt;-</span> <span class="fu">clustree</span>(seurat_integrated<span class="sc">@</span>meta.data, </span>
<span id="cb90-384"><a href="#cb90-384" aria-hidden="true" tabindex="-1"></a>                 <span class="at">prefix =</span> <span class="st">"SCT_snn_res."</span>)<span class="co">#指定包含聚类信息的列</span></span>
<span id="cb90-385"><a href="#cb90-385" aria-hidden="true" tabindex="-1"></a>tree</span>
<span id="cb90-386"><a href="#cb90-386" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb90-387"><a href="#cb90-387" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-388"><a href="#cb90-388" aria-hidden="true" tabindex="-1"></a>通常情况下，为了决定合适的聚类分辨率，可使用以下两种策略：</span>
<span id="cb90-389"><a href="#cb90-389" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-390"><a href="#cb90-390" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>选择透明箭头出现较少的分辨率聚类结果</span>
<span id="cb90-391"><a href="#cb90-391" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-392"><a href="#cb90-392" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>基于marker基因表达选择有生物学意义的分辨率聚类结果</span>
<span id="cb90-393"><a href="#cb90-393" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-394"><a href="#cb90-394" aria-hidden="true" tabindex="-1"></a>接下来分析，按照分辨率为0.5进行</span>
<span id="cb90-395"><a href="#cb90-395" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-398"><a href="#cb90-398" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb90-399"><a href="#cb90-399" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb90-400"><a href="#cb90-400" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(seurat_integrated) <span class="ot">&lt;-</span> <span class="st">"SCT_snn_res.0.5"</span></span>
<span id="cb90-401"><a href="#cb90-401" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb90-402"><a href="#cb90-402" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-403"><a href="#cb90-403" aria-hidden="true" tabindex="-1"></a>聚类可视化</span>
<span id="cb90-404"><a href="#cb90-404" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-407"><a href="#cb90-407" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb90-408"><a href="#cb90-408" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb90-409"><a href="#cb90-409" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the UMAP</span></span>
<span id="cb90-410"><a href="#cb90-410" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(seurat_integrated,</span>
<span id="cb90-411"><a href="#cb90-411" aria-hidden="true" tabindex="-1"></a>        <span class="at">reduction =</span> <span class="st">"umap.integrated"</span>,</span>
<span id="cb90-412"><a href="#cb90-412" aria-hidden="true" tabindex="-1"></a>        <span class="at">label =</span> <span class="cn">FALSE</span>)</span>
<span id="cb90-413"><a href="#cb90-413" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb90-414"><a href="#cb90-414" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-415"><a href="#cb90-415" aria-hidden="true" tabindex="-1"></a><span class="fu">### **细胞分群质量评估**</span></span>
<span id="cb90-416"><a href="#cb90-416" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-417"><a href="#cb90-417" aria-hidden="true" tabindex="-1"></a><span class="fu">#### **分析样本类型是否影响细胞分群**</span></span>
<span id="cb90-418"><a href="#cb90-418" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-421"><a href="#cb90-421" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb90-422"><a href="#cb90-422" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb90-423"><a href="#cb90-423" aria-hidden="true" tabindex="-1"></a><span class="co"># 先简单查看不同cluster的细胞数</span></span>
<span id="cb90-424"><a href="#cb90-424" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(seurat_integrated<span class="sc">@</span>active.ident)</span>
<span id="cb90-425"><a href="#cb90-425" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-426"><a href="#cb90-426" aria-hidden="true" tabindex="-1"></a><span class="co"># 查看不同样本类型中的细胞分群情况</span></span>
<span id="cb90-427"><a href="#cb90-427" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(seurat_integrated, </span>
<span id="cb90-428"><a href="#cb90-428" aria-hidden="true" tabindex="-1"></a>        <span class="at">reduction =</span> <span class="st">"umap.integrated"</span>,</span>
<span id="cb90-429"><a href="#cb90-429" aria-hidden="true" tabindex="-1"></a>        <span class="at">label =</span> <span class="cn">TRUE</span>, </span>
<span id="cb90-430"><a href="#cb90-430" aria-hidden="true" tabindex="-1"></a>        <span class="at">split.by =</span> <span class="st">"groups"</span>) <span class="sc">+</span> </span>
<span id="cb90-431"><a href="#cb90-431" aria-hidden="true" tabindex="-1"></a>  <span class="fu">NoLegend</span>()</span>
<span id="cb90-432"><a href="#cb90-432" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb90-433"><a href="#cb90-433" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-434"><a href="#cb90-434" aria-hidden="true" tabindex="-1"></a><span class="fu">#### **分析细胞周期是否影响细胞分群**</span></span>
<span id="cb90-435"><a href="#cb90-435" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-438"><a href="#cb90-438" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb90-439"><a href="#cb90-439" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb90-440"><a href="#cb90-440" aria-hidden="true" tabindex="-1"></a><span class="co"># Explore whether clusters segregate by cell cycle phase</span></span>
<span id="cb90-441"><a href="#cb90-441" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(seurat_integrated,</span>
<span id="cb90-442"><a href="#cb90-442" aria-hidden="true" tabindex="-1"></a>        <span class="at">reduction =</span> <span class="st">"umap.integrated"</span>,</span>
<span id="cb90-443"><a href="#cb90-443" aria-hidden="true" tabindex="-1"></a>        <span class="at">label =</span> <span class="cn">TRUE</span>, </span>
<span id="cb90-444"><a href="#cb90-444" aria-hidden="true" tabindex="-1"></a>        <span class="at">split.by =</span> <span class="st">"Phase"</span>) <span class="sc">+</span> </span>
<span id="cb90-445"><a href="#cb90-445" aria-hidden="true" tabindex="-1"></a>  <span class="fu">NoLegend</span>()</span>
<span id="cb90-446"><a href="#cb90-446" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb90-447"><a href="#cb90-447" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-448"><a href="#cb90-448" aria-hidden="true" tabindex="-1"></a><span class="fu">#### **分析其他非期望变异来源是否会影响细胞分群**</span></span>
<span id="cb90-449"><a href="#cb90-449" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-452"><a href="#cb90-452" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb90-453"><a href="#cb90-453" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb90-454"><a href="#cb90-454" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 10</span></span>
<span id="cb90-455"><a href="#cb90-455" aria-hidden="true" tabindex="-1"></a><span class="co"># Determine metrics to plot present in seurat_clustered@meta.data</span></span>
<span id="cb90-456"><a href="#cb90-456" aria-hidden="true" tabindex="-1"></a>metrics <span class="ot">&lt;-</span>  <span class="fu">c</span>(<span class="st">"nCount_RNA"</span>, <span class="st">"nFeature_RNA"</span>, <span class="st">"S.Score"</span>, <span class="st">"G2M.Score"</span>, <span class="st">"mitoRatio"</span>)</span>
<span id="cb90-457"><a href="#cb90-457" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-458"><a href="#cb90-458" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>(seurat_integrated, </span>
<span id="cb90-459"><a href="#cb90-459" aria-hidden="true" tabindex="-1"></a>            <span class="at">reduction =</span> <span class="st">"umap.integrated"</span>, </span>
<span id="cb90-460"><a href="#cb90-460" aria-hidden="true" tabindex="-1"></a>            <span class="at">features =</span> metrics,</span>
<span id="cb90-461"><a href="#cb90-461" aria-hidden="true" tabindex="-1"></a>            <span class="at">pt.size =</span> <span class="fl">0.4</span>, </span>
<span id="cb90-462"><a href="#cb90-462" aria-hidden="true" tabindex="-1"></a>            <span class="at">order =</span> <span class="cn">TRUE</span>,</span>
<span id="cb90-463"><a href="#cb90-463" aria-hidden="true" tabindex="-1"></a>            <span class="at">min.cutoff =</span> <span class="st">'q10'</span>, </span>
<span id="cb90-464"><a href="#cb90-464" aria-hidden="true" tabindex="-1"></a>            <span class="at">label =</span> <span class="cn">TRUE</span>)</span>
<span id="cb90-465"><a href="#cb90-465" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb90-466"><a href="#cb90-466" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-467"><a href="#cb90-467" aria-hidden="true" tabindex="-1"></a>保存</span>
<span id="cb90-468"><a href="#cb90-468" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-471"><a href="#cb90-471" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb90-472"><a href="#cb90-472" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb90-473"><a href="#cb90-473" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(seurat_integrated, </span>
<span id="cb90-474"><a href="#cb90-474" aria-hidden="true" tabindex="-1"></a>        <span class="at">file =</span> <span class="st">"output/sc_supplementary/GSE150430_seurat_clustered.RDS"</span>); <span class="fu">beep</span>()</span>
<span id="cb90-475"><a href="#cb90-475" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb90-476"><a href="#cb90-476" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-477"><a href="#cb90-477" aria-hidden="true" tabindex="-1"></a>------------------------------------------------------------------------</span>
<span id="cb90-478"><a href="#cb90-478" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-479"><a href="#cb90-479" aria-hidden="true" tabindex="-1"></a><span class="fu">### 细胞注释</span></span>
<span id="cb90-480"><a href="#cb90-480" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-481"><a href="#cb90-481" aria-hidden="true" tabindex="-1"></a><span class="fu">#### **探索cell type markers的表达**</span></span>
<span id="cb90-482"><a href="#cb90-482" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-483"><a href="#cb90-483" aria-hidden="true" tabindex="-1"></a>这里的marker基因选用 @sec-Universal_Marker_for_initial_Cell_clustering 中的细胞初步分群通用marker。</span>
<span id="cb90-484"><a href="#cb90-484" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-487"><a href="#cb90-487" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb90-488"><a href="#cb90-488" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb90-489"><a href="#cb90-489" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 8</span></span>
<span id="cb90-490"><a href="#cb90-490" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 8</span></span>
<span id="cb90-491"><a href="#cb90-491" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(<span class="at">list =</span> <span class="fu">ls</span>())</span>
<span id="cb90-492"><a href="#cb90-492" aria-hidden="true" tabindex="-1"></a><span class="fu">gc</span>()</span>
<span id="cb90-493"><a href="#cb90-493" aria-hidden="true" tabindex="-1"></a>seurat_clustered <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"output/sc_supplementary/GSE150430_seurat_clustered.RDS"</span>)</span>
<span id="cb90-494"><a href="#cb90-494" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-495"><a href="#cb90-495" aria-hidden="true" tabindex="-1"></a>genes_to_check <span class="ot">=</span> <span class="fu">c</span>(<span class="st">'PTPRC'</span>, </span>
<span id="cb90-496"><a href="#cb90-496" aria-hidden="true" tabindex="-1"></a>                   <span class="st">"CD163"</span>,<span class="st">"AIF1"</span>, </span>
<span id="cb90-497"><a href="#cb90-497" aria-hidden="true" tabindex="-1"></a>                   <span class="st">'CD3D'</span>, <span class="st">'CD3E'</span>, <span class="st">'CD4'</span>, <span class="st">'CD8A'</span>, </span>
<span id="cb90-498"><a href="#cb90-498" aria-hidden="true" tabindex="-1"></a>                   <span class="st">'CD19'</span>, <span class="st">'CD79A'</span>, <span class="st">'MS4A1'</span>, <span class="st">"SDC1"</span>, <span class="st">"CD27"</span>, <span class="st">"CD38"</span>,</span>
<span id="cb90-499"><a href="#cb90-499" aria-hidden="true" tabindex="-1"></a>                   <span class="st">'IGHG1'</span>, <span class="st">'MZB1'</span>, <span class="st">'SDC1'</span>, <span class="st">"JCHAIN"</span>, </span>
<span id="cb90-500"><a href="#cb90-500" aria-hidden="true" tabindex="-1"></a>                   <span class="st">'CD68'</span>, <span class="st">'CD163'</span>, <span class="st">'CD14'</span>, </span>
<span id="cb90-501"><a href="#cb90-501" aria-hidden="true" tabindex="-1"></a>                   <span class="st">'S100A9'</span>, <span class="st">'S100A8'</span>, <span class="st">'MMP19'</span>,</span>
<span id="cb90-502"><a href="#cb90-502" aria-hidden="true" tabindex="-1"></a>                   <span class="st">'C1QA'</span>,  <span class="st">'C1QB'</span>, </span>
<span id="cb90-503"><a href="#cb90-503" aria-hidden="true" tabindex="-1"></a>                   <span class="st">'TPSAB1'</span>, <span class="st">'TPSB2'</span>, </span>
<span id="cb90-504"><a href="#cb90-504" aria-hidden="true" tabindex="-1"></a>                   <span class="st">'KLRB1'</span>, <span class="st">"KLRD1"</span>, <span class="st">'NCR1'</span>, <span class="st">"GNLY"</span>, <span class="st">"NKG7"</span>, </span>
<span id="cb90-505"><a href="#cb90-505" aria-hidden="true" tabindex="-1"></a>                   <span class="st">'FGF7'</span>, <span class="st">'MME'</span>, <span class="st">'ACTA2'</span>, <span class="st">"COL3A1"</span>, </span>
<span id="cb90-506"><a href="#cb90-506" aria-hidden="true" tabindex="-1"></a>                   <span class="st">'PECAM1'</span>, <span class="st">'VWF'</span>, </span>
<span id="cb90-507"><a href="#cb90-507" aria-hidden="true" tabindex="-1"></a>                   <span class="st">'EPCAM'</span>, <span class="st">'KRT19'</span>, <span class="st">'PROM1'</span>, <span class="st">'ALDH1A1'</span>) <span class="sc">|&gt;</span> <span class="fu">unique</span>()</span>
<span id="cb90-508"><a href="#cb90-508" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-509"><a href="#cb90-509" aria-hidden="true" tabindex="-1"></a><span class="fu">DotPlot</span>(seurat_clustered, </span>
<span id="cb90-510"><a href="#cb90-510" aria-hidden="true" tabindex="-1"></a>        <span class="at">features =</span> genes_to_check) <span class="sc">+</span> </span>
<span id="cb90-511"><a href="#cb90-511" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_flip</span>()</span>
<span id="cb90-512"><a href="#cb90-512" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb90-513"><a href="#cb90-513" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-514"><a href="#cb90-514" aria-hidden="true" tabindex="-1"></a><span class="fu">#### 手动注释</span></span>
<span id="cb90-515"><a href="#cb90-515" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-518"><a href="#cb90-518" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb90-519"><a href="#cb90-519" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb90-520"><a href="#cb90-520" aria-hidden="true" tabindex="-1"></a>seurat_clustered <span class="ot">&lt;-</span> <span class="fu">RenameIdents</span>(seurat_clustered, </span>
<span id="cb90-521"><a href="#cb90-521" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"0"</span> <span class="ot">=</span> <span class="st">"B cells"</span>,</span>
<span id="cb90-522"><a href="#cb90-522" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"1"</span> <span class="ot">=</span> <span class="st">"T cells"</span>,</span>
<span id="cb90-523"><a href="#cb90-523" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"2"</span> <span class="ot">=</span> <span class="st">"Epithelial/cancer"</span>,</span>
<span id="cb90-524"><a href="#cb90-524" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"3"</span> <span class="ot">=</span> <span class="st">"T cells"</span>,</span>
<span id="cb90-525"><a href="#cb90-525" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"4"</span> <span class="ot">=</span> <span class="st">"T cells"</span>,</span>
<span id="cb90-526"><a href="#cb90-526" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"5"</span> <span class="ot">=</span> <span class="st">"Macrophages"</span>,</span>
<span id="cb90-527"><a href="#cb90-527" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"6"</span> <span class="ot">=</span> <span class="st">"T cells"</span>,</span>
<span id="cb90-528"><a href="#cb90-528" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"7"</span> <span class="ot">=</span> <span class="st">"Plasma cells"</span>,</span>
<span id="cb90-529"><a href="#cb90-529" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"8"</span> <span class="ot">=</span> <span class="st">"B cells"</span>,</span>
<span id="cb90-530"><a href="#cb90-530" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"9"</span> <span class="ot">=</span> <span class="st">"Epithelial/cancer"</span>,</span>
<span id="cb90-531"><a href="#cb90-531" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"10"</span> <span class="ot">=</span> <span class="st">"B cells"</span>,</span>
<span id="cb90-532"><a href="#cb90-532" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"11"</span> <span class="ot">=</span> <span class="st">"B cells"</span>,</span>
<span id="cb90-533"><a href="#cb90-533" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"12"</span> <span class="ot">=</span> <span class="st">"Unknown"</span>,</span>
<span id="cb90-534"><a href="#cb90-534" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"13"</span> <span class="ot">=</span> <span class="st">"NK cells"</span>,</span>
<span id="cb90-535"><a href="#cb90-535" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"14"</span> <span class="ot">=</span> <span class="st">"Unknown"</span>,</span>
<span id="cb90-536"><a href="#cb90-536" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"15"</span> <span class="ot">=</span> <span class="st">"Epithelial/cancer"</span>,</span>
<span id="cb90-537"><a href="#cb90-537" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"16"</span> <span class="ot">=</span> <span class="st">"Myeloid (unspecific)"</span>,</span>
<span id="cb90-538"><a href="#cb90-538" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"17"</span> <span class="ot">=</span> <span class="st">"Epithelial/cancer"</span>,</span>
<span id="cb90-539"><a href="#cb90-539" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"18"</span> <span class="ot">=</span> <span class="st">"Epithelial/cancer"</span>,</span>
<span id="cb90-540"><a href="#cb90-540" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"19"</span> <span class="ot">=</span> <span class="st">"Epithelial/cancer"</span>,</span>
<span id="cb90-541"><a href="#cb90-541" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"20"</span> <span class="ot">=</span> <span class="st">"B cells"</span>)</span>
<span id="cb90-542"><a href="#cb90-542" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="fu">Idents</span>(seurat_clustered))</span>
<span id="cb90-543"><a href="#cb90-543" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-544"><a href="#cb90-544" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the UMAP</span></span>
<span id="cb90-545"><a href="#cb90-545" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(seurat_clustered, </span>
<span id="cb90-546"><a href="#cb90-546" aria-hidden="true" tabindex="-1"></a>        <span class="at">reduction =</span> <span class="st">"umap.integrated"</span>, </span>
<span id="cb90-547"><a href="#cb90-547" aria-hidden="true" tabindex="-1"></a>        <span class="at">label =</span> T)</span>
<span id="cb90-548"><a href="#cb90-548" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb90-549"><a href="#cb90-549" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-550"><a href="#cb90-550" aria-hidden="true" tabindex="-1"></a>在注释好的数据中再次检查marker基因的表达情况：</span>
<span id="cb90-551"><a href="#cb90-551" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-554"><a href="#cb90-554" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb90-555"><a href="#cb90-555" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb90-556"><a href="#cb90-556" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 8</span></span>
<span id="cb90-557"><a href="#cb90-557" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 10</span></span>
<span id="cb90-558"><a href="#cb90-558" aria-hidden="true" tabindex="-1"></a><span class="fu">DotPlot</span>(seurat_clustered, </span>
<span id="cb90-559"><a href="#cb90-559" aria-hidden="true" tabindex="-1"></a>        <span class="at">features =</span> genes_to_check) <span class="sc">+</span> </span>
<span id="cb90-560"><a href="#cb90-560" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_flip</span>() <span class="sc">+</span></span>
<span id="cb90-561"><a href="#cb90-561" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">30</span>,<span class="at">vjust =</span> <span class="fl">0.85</span>,<span class="at">hjust =</span> <span class="fl">0.75</span>)</span>
<span id="cb90-562"><a href="#cb90-562" aria-hidden="true" tabindex="-1"></a>) </span>
<span id="cb90-563"><a href="#cb90-563" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb90-564"><a href="#cb90-564" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-565"><a href="#cb90-565" aria-hidden="true" tabindex="-1"></a>因为接下来我们要提取髓系细胞进一步分析，所以通过几种可视化方法检查一下髓系细胞通用marker（CD163和AIF1）的表达情况：</span>
<span id="cb90-566"><a href="#cb90-566" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-569"><a href="#cb90-569" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb90-570"><a href="#cb90-570" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb90-571"><a href="#cb90-571" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 8</span></span>
<span id="cb90-572"><a href="#cb90-572" aria-hidden="true" tabindex="-1"></a><span class="fu">RidgePlot</span>(seurat_clustered, <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"CD163"</span>, <span class="st">"AIF1"</span>), <span class="at">ncol =</span> <span class="dv">1</span>)</span>
<span id="cb90-573"><a href="#cb90-573" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb90-574"><a href="#cb90-574" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-577"><a href="#cb90-577" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb90-578"><a href="#cb90-578" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb90-579"><a href="#cb90-579" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(seurat_clustered, <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"CD163"</span>, <span class="st">"AIF1"</span>))</span>
<span id="cb90-580"><a href="#cb90-580" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb90-581"><a href="#cb90-581" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-584"><a href="#cb90-584" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb90-585"><a href="#cb90-585" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb90-586"><a href="#cb90-586" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 8</span></span>
<span id="cb90-587"><a href="#cb90-587" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>(seurat_clustered, <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"CD163"</span>, <span class="st">"AIF1"</span>), <span class="at">label =</span> T)</span>
<span id="cb90-588"><a href="#cb90-588" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb90-589"><a href="#cb90-589" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-590"><a href="#cb90-590" aria-hidden="true" tabindex="-1"></a>保存</span>
<span id="cb90-591"><a href="#cb90-591" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-594"><a href="#cb90-594" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb90-595"><a href="#cb90-595" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb90-596"><a href="#cb90-596" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(seurat_clustered, </span>
<span id="cb90-597"><a href="#cb90-597" aria-hidden="true" tabindex="-1"></a>        <span class="at">file =</span> <span class="st">"output/sc_supplementary/GSE150430_seurat_annotated.RDS"</span>); <span class="fu">beep</span>()</span>
<span id="cb90-598"><a href="#cb90-598" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb90-599"><a href="#cb90-599" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-600"><a href="#cb90-600" aria-hidden="true" tabindex="-1"></a><span class="fu">### 提取髓系细胞</span></span>
<span id="cb90-601"><a href="#cb90-601" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-604"><a href="#cb90-604" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb90-605"><a href="#cb90-605" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb90-606"><a href="#cb90-606" aria-hidden="true" tabindex="-1"></a>sub_seurat <span class="ot">&lt;-</span> <span class="fu">subset</span>(seurat_clustered, </span>
<span id="cb90-607"><a href="#cb90-607" aria-hidden="true" tabindex="-1"></a>                     <span class="at">idents =</span> <span class="fu">c</span>(<span class="st">"Macrophages"</span>, </span>
<span id="cb90-608"><a href="#cb90-608" aria-hidden="true" tabindex="-1"></a>                                <span class="st">"Myeloid (unspecific)"</span>))</span>
<span id="cb90-609"><a href="#cb90-609" aria-hidden="true" tabindex="-1"></a>sub_seurat</span>
<span id="cb90-610"><a href="#cb90-610" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(sub_seurat<span class="sc">@</span>meta.data)</span>
<span id="cb90-611"><a href="#cb90-611" aria-hidden="true" tabindex="-1"></a>myeloid_seurat <span class="ot">&lt;-</span> <span class="fu">CreateSeuratObject</span>(<span class="at">counts =</span> sub_seurat<span class="sc">@</span>assays[[<span class="st">"RNA"</span>]],</span>
<span id="cb90-612"><a href="#cb90-612" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">meta.data =</span> sub_seurat<span class="sc">@</span>meta.data[,<span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">7</span>, <span class="dv">10</span><span class="sc">:</span><span class="dv">13</span>, <span class="dv">16</span>)])</span>
<span id="cb90-613"><a href="#cb90-613" aria-hidden="true" tabindex="-1"></a>myeloid_seurat</span>
<span id="cb90-614"><a href="#cb90-614" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb90-615"><a href="#cb90-615" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-616"><a href="#cb90-616" aria-hidden="true" tabindex="-1"></a>::: callout-warning</span>
<span id="cb90-617"><a href="#cb90-617" aria-hidden="true" tabindex="-1"></a>可以看到在重新创建Seurat之后，默认的“RNA” assay之中只有一个layer。为了后面执行<span class="in">`SCTransform`</span>及整合，**这里需要将其按照样本拆分成不同的layers**。</span>
<span id="cb90-618"><a href="#cb90-618" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-621"><a href="#cb90-621" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb90-622"><a href="#cb90-622" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb90-623"><a href="#cb90-623" aria-hidden="true" tabindex="-1"></a>myeloid_seurat[[<span class="st">"RNA"</span>]] <span class="ot">&lt;-</span> <span class="fu">split</span>(myeloid_seurat[[<span class="st">"RNA"</span>]], </span>
<span id="cb90-624"><a href="#cb90-624" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">f =</span> myeloid_seurat<span class="sc">$</span>samples)</span>
<span id="cb90-625"><a href="#cb90-625" aria-hidden="true" tabindex="-1"></a>myeloid_seurat</span>
<span id="cb90-626"><a href="#cb90-626" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb90-627"><a href="#cb90-627" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb90-628"><a href="#cb90-628" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-631"><a href="#cb90-631" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb90-632"><a href="#cb90-632" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb90-633"><a href="#cb90-633" aria-hidden="true" tabindex="-1"></a><span class="co"># 补充原细胞注释信息</span></span>
<span id="cb90-634"><a href="#cb90-634" aria-hidden="true" tabindex="-1"></a>myeloid_seurat<span class="sc">$</span>old_clusters <span class="ot">&lt;-</span> <span class="fu">Idents</span>(sub_seurat)</span>
<span id="cb90-635"><a href="#cb90-635" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(myeloid_seurat<span class="sc">$</span>old_clusters)</span>
<span id="cb90-636"><a href="#cb90-636" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-637"><a href="#cb90-637" aria-hidden="true" tabindex="-1"></a><span class="co"># 保存</span></span>
<span id="cb90-638"><a href="#cb90-638" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(myeloid_seurat, <span class="at">file =</span> <span class="st">"output/sc_supplementary/GSE150430_myeloid_seurat.RDS"</span>)</span>
<span id="cb90-639"><a href="#cb90-639" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb90-640"><a href="#cb90-640" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-641"><a href="#cb90-641" aria-hidden="true" tabindex="-1"></a>::: {.callout-tip collapse="true"}</span>
<span id="cb90-642"><a href="#cb90-642" aria-hidden="true" tabindex="-1"></a><span class="fu">###### 如何提取并处理目标细胞亚群</span></span>
<span id="cb90-643"><a href="#cb90-643" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-644"><a href="#cb90-644" aria-hidden="true" tabindex="-1"></a>After we have completed the scRNA-seq workflow and identified the various cell types present in our samples, we might decide that for a particular cell type, we would like to identify subtypes. For example, if we have a large cluster of **CD4+ Helper T cells**, we may want to identify subsets of **Th1, Th2, Th17, Th9, and Tfh cells**. To subset the dataset, Seurat has a handy&nbsp;<span class="in">`subset()`</span>function（ @sec-subset_seurat ）.</span>
<span id="cb90-645"><a href="#cb90-645" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-646"><a href="#cb90-646" aria-hidden="true" tabindex="-1"></a>To perform the subclustering, there are a couple of different methods you could try. The easiest would be to run the&nbsp;<span class="in">`FindNeighbors()`</span>&nbsp;and&nbsp;<span class="in">`FindClusters()`</span>&nbsp;on the subsetted cells, adjusting the resolution to give you the optimal clustering. However, with this approach **you are not redefining the most variable genes** used to find clusters, so it might not work if the genes delineating these subsets are not those driving any of the top PCs used for the clustering.</span>
<span id="cb90-647"><a href="#cb90-647" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-648"><a href="#cb90-648" aria-hidden="true" tabindex="-1"></a>Alternatively, we could **start over with the raw counts** for this subset of cells and run&nbsp;**`SCTransform()`**to determine the greatest sources of variation present. **This would allow us to focus our clustering on the most variant genes present among our subset of cells**. Hopefully, the most variant genes are those driving the various desired subsets (e.g. Th1, Th2, Th17, Th9, and Tfh cells). If **integration** is necessary, then this step would still need to be performed.</span>
<span id="cb90-649"><a href="#cb90-649" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-650"><a href="#cb90-650" aria-hidden="true" tabindex="-1"></a>Since subsetting the dataset can result in a much smaller number of cells, it is important to consider the total number of cells you are looking to cluster and some of the parameters that might be affected by the small numbers. For example, if integrating, there is a ‘K’ number of cells used for determining the neighborhoods for identifying and filtering anchors. Therefore, if your integration isn’t very good for a small dataset, you might want to consider **lowering the ‘K’ parameter (k.filter、k.weight)** in<span class="in">`IntegrateLayers`</span>. However, if ‘K’ is too small, it could also lead to poor integration.</span>
<span id="cb90-651"><a href="#cb90-651" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-652"><a href="#cb90-652" aria-hidden="true" tabindex="-1"></a>（来自：<span class="co">[</span><span class="ot">*Subclustering scRNA-seq datasets*</span><span class="co">](https://hbctraining.github.io/scRNA-seq_online/lessons/seurat_subclustering.html)</span>）</span>
<span id="cb90-653"><a href="#cb90-653" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb90-654"><a href="#cb90-654" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-655"><a href="#cb90-655" aria-hidden="true" tabindex="-1"></a>下一节，针对髓系细胞进一步分群。</span>
<span id="cb90-656"><a href="#cb90-656" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-657"><a href="#cb90-657" aria-hidden="true" tabindex="-1"></a>------------------------------------------------------------------------</span>
<span id="cb90-658"><a href="#cb90-658" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-659"><a href="#cb90-659" aria-hidden="true" tabindex="-1"></a>::: {.callout-note collapse="true" icon="false"}</span>
<span id="cb90-660"><a href="#cb90-660" aria-hidden="true" tabindex="-1"></a><span class="fu">## Session Info</span></span>
<span id="cb90-661"><a href="#cb90-661" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-664"><a href="#cb90-664" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb90-665"><a href="#cb90-665" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb90-666"><a href="#cb90-666" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span>
<span id="cb90-667"><a href="#cb90-667" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb90-668"><a href="#cb90-668" aria-hidden="true" tabindex="-1"></a>:::</span>
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">
      <ul class="footer-items list-unstyled">
<li class="nav-item">
 This book was built with Quarto
  </li>  
</ul>
</div>   
    <div class="nav-footer-center">Copyright 2023, Du Junhong</div>
    <div class="nav-footer-right">
      <ul class="footer-items list-unstyled">
<li class="nav-item compact">
    <a class="nav-link" href="https://github.com/djhcod">
      <i class="bi bi-github" role="img">
</i> 
    </a>
  </li>  
</ul>
</div>
  </div>
</footer>


</body></html>