<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>R语言学习笔记 - 28&nbsp; 读取非标准格式的单细胞数据</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../single_cell/sc_supplementary/universal_marker.html" rel="next">
<link href="../../single_cell/scRNA-seq_online/09_merged_SC_marker_identification.html" rel="prev">
<link href="../../images/logo.jpg" rel="icon" type="image/jpeg">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark"><script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><script type="application/json" class="js-hypothesis-config">
{
  "theme": "classic",
  "openSidebar": false,
  "showHighlights": "always",
  "enableExperimentalNewNoteButton": true
}
</script><script async="" src="https://hypothes.is/embed.js"></script>
<meta name="twitter:title" content="R语言学习笔记 - 28&nbsp; 读取非标准格式的单细胞数据">
<meta name="twitter:description" content="">
<meta name="twitter:image" content="https://djhcod.github.io/r-notes/single_cell/sc_supplementary/images/640-01.png">
<meta name="twitter:image-height" content="344">
<meta name="twitter:image-width" content="1080">
<meta name="twitter:card" content="summary_large_image">
</head>
<body class="nav-sidebar docked nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="navbar navbar-expand-lg navbar-dark "><div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a href="../../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../../images/logo.jpg" alt="" class="navbar-logo"></a>
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">R语言学习笔记</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
<li class="nav-item">
    <a class="nav-link active" href="../../index.html" rel="" target="" aria-current="page">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../intro.html" rel="" target="">
 <span class="menu-text">Introduction</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-parts" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Parts</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-parts">
<li>
    <a class="dropdown-item" href="../../r_basic/r_basics.html" rel="" target="">
 <span class="dropdown-text">R语言基础</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../single_cell/seurat/seurat.html" rel="" target="">
 <span class="dropdown-text">Seurat v5官方文档</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../single_cell/scRNA-seq_online/00_intro.html" rel="" target="">
 <span class="dropdown-text">scRNA-seq_online学习材料</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../quarto_foundation/quarto_foundation.html" rel="" target="">
 <span class="dropdown-text">Quarto基础</span></a>
  </li>  
    </ul>
</li>
  <li class="nav-item">
    <a class="nav-link" href="../../references.html" rel="" target="">
 <span class="menu-text">References</span></a>
  </li>  
</ul>
<ul class="navbar-nav navbar-nav-scroll ms-auto">
<li class="nav-item">
    <a class="nav-link" href="https://bookdown.org" rel="" target=""><i class="bi bi-book-fill" role="img">
</i> 
 <span class="menu-text">Bookdown</span></a>
  </li>  
</ul>
<div class="quarto-navbar-tools tools-wide">
    <a href="https://github.com/djhcod/r-notes" rel="" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
    <div class="dropdown">
      <a href="" title="Share" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Share"><i class="bi bi-share"></i></a>
      <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="quarto-navigation-tool-dropdown-0">
<li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://twitter.com/intent/tweet?url=%7Curl%7C">
              <i class="bi bi-bi-twitter pe-1"></i>
            Twitter
            </a>
          </li>
          <li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://www.facebook.com/sharer/sharer.php?u=%7Curl%7C">
              <i class="bi bi-bi-facebook pe-1"></i>
            Facebook
            </a>
          </li>
      </ul>
</div>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../single_cell/sc_supplementary/read_sc_data.html">单细胞补充内容</a></li><li class="breadcrumb-item"><a href="../../single_cell/sc_supplementary/read_sc_data.html"><span class="chapter-number">28</span>&nbsp; <span class="chapter-title">**读取非标准格式的单细胞数据**</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto"><div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">主页</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">前言</span></a>
  </div>
</li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../r_basic/r_basics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">R语言基础</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/environment_configuration.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Rstudio环境配置</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/data_input_output.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">数据的读取与输出</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/basic_data_function.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">数据处理基本函数</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/character.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">字符的处理</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/loop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">循环</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../single_cell/seurat/seurat.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Seurat v5官方文档</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/seurat_command_list.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Seurat常用函数清单</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/seurat_tutorial.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Seurat细胞分群官方教程</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/data_visualization_methods_in_seurat.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Seurat中的数据可视化方法</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/sctransform.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">基于SCTransform的单细胞数据标准化</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/cell_cycle_regression.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">消除细胞周期的影响</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/integration.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">整合（integration）</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/integrative_analysis_in_seurat_v5.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Seurat v5单细胞数据整合分析</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/marker_gene_identification.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">寻找marker gene</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/de_vignette.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">寻找不同样本类型间同一细胞类型内的差异基因</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/mapping_and_annotating_query_datasets.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">基于参考集的细胞注释</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../single_cell/scRNA-seq_online/00_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">scRNA-seq_online学习材料</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/01_intro_to_scRNA-seq.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">单细胞测序技术介绍</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/02_SC_generation_of_count_matrix.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">单细胞RNA测序—从raw data到count matrix</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/03_SC_quality_control-setup.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">数据导入与Seurat对象构建</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/04_SC_quality_control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">质控</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/postQC_workflow.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">21</span>&nbsp; <span class="chapter-title">Single-cell RNA-seq Clustering Workflow</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/05_theory_of_PCA.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">22</span>&nbsp; <span class="chapter-title">PCA原理</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/06_SC_SCT_normalization.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">23</span>&nbsp; <span class="chapter-title">Normalization and regressing out unwanted variation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/06_integration.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">24</span>&nbsp; <span class="chapter-title">单细胞数据整合（Integration）</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/07_SC_clustering_cells_SCT.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">25</span>&nbsp; <span class="chapter-title">细胞聚类（clustering analysis）</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/08_SC_clustering_quality_control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">26</span>&nbsp; <span class="chapter-title">细胞分群质量评估</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/09_merged_SC_marker_identification.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">27</span>&nbsp; <span class="chapter-title">寻找marker基因+细胞注释</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
 <span class="menu-text">单细胞补充内容</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/sc_supplementary/read_sc_data.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">28</span>&nbsp; <span class="chapter-title"><strong>读取非标准格式的单细胞数据</strong></span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/sc_supplementary/universal_marker.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">29</span>&nbsp; <span class="chapter-title">细胞分群通用marker</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/sc_supplementary/integrated_analysis_multiple_single_cell_datasets.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">30</span>&nbsp; <span class="chapter-title">多个单细胞数据集整合分析</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../quarto_foundation/quarto_foundation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Quarto基础</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/yaml_settings.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">31</span>&nbsp; <span class="chapter-title">YAML设置</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/images_settings.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">32</span>&nbsp; <span class="chapter-title">图片设置</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/cross_references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">33</span>&nbsp; <span class="chapter-title">交叉引用</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/insert_other_content.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">34</span>&nbsp; <span class="chapter-title">插入其他内容</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/quarto_books.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">35</span>&nbsp; <span class="chapter-title">Quarto Books</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/github_pages.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">36</span>&nbsp; <span class="chapter-title">发布到GitHub Pages</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="false">
 <span class="menu-text">参考文献</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active" data-toc-expanded="1"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#sec-Read_non-standard_10x_files" id="toc-sec-Read_non-standard_10x_files" class="nav-link active" data-scroll-target="#sec-Read_non-standard_10x_files"><span class="header-section-number">28.1</span> 读取非标准10X格式文件</a></li>
  <li>
<a href="#%E8%AF%BB%E5%8F%96raw_counts%E5%BD%A2%E5%BC%8F%E7%9A%84%E6%95%B0%E6%8D%AE" id="toc-读取raw_counts形式的数据" class="nav-link" data-scroll-target="#%E8%AF%BB%E5%8F%96raw_counts%E5%BD%A2%E5%BC%8F%E7%9A%84%E6%95%B0%E6%8D%AE"><span class="header-section-number">28.2</span> 读取raw_counts形式的数据</a>
  <ul class="collapse">
<li><a href="#%E8%AF%BB%E5%8F%96raw_counts%E7%9F%A9%E9%98%B5" id="toc-读取raw_counts矩阵" class="nav-link" data-scroll-target="#%E8%AF%BB%E5%8F%96raw_counts%E7%9F%A9%E9%98%B5">读取raw_counts矩阵</a></li>
  <li><a href="#%E6%B7%BB%E5%8A%A0%E6%A0%B7%E6%9C%AC%E4%BF%A1%E6%81%AF" id="toc-添加样本信息" class="nav-link" data-scroll-target="#%E6%B7%BB%E5%8A%A0%E6%A0%B7%E6%9C%AC%E4%BF%A1%E6%81%AF">添加样本信息</a></li>
  </ul>
</li>
  <li><a href="#%E8%AF%BB%E5%8F%96txttsvcsv%E7%AD%89%E6%96%87%E6%9C%AC%E6%96%87%E4%BB%B6%E8%A1%A8%E8%BE%BE%E9%87%8F%E7%9F%A9%E9%98%B5%E4%BF%A1%E6%81%AF" id="toc-读取txttsvcsv等文本文件表达量矩阵信息" class="nav-link" data-scroll-target="#%E8%AF%BB%E5%8F%96txttsvcsv%E7%AD%89%E6%96%87%E6%9C%AC%E6%96%87%E4%BB%B6%E8%A1%A8%E8%BE%BE%E9%87%8F%E7%9F%A9%E9%98%B5%E4%BF%A1%E6%81%AF"><span class="header-section-number">28.3</span> 读取txt/tsv/csv等文本文件表达量矩阵信息</a></li>
  </ul><div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/djhcod/r-notes/blob/main/single_cell/sc_supplementary/read_sc_data.qmd" class="toc-action">View source</a></p><p><a href="https://github.com/djhcod/r-notes/issues/new" class="toc-action">Report an issue</a></p><p><a href="https://github.com/djhcod/r-notes/edit/main/single_cell/sc_supplementary/read_sc_data.qmd" class="toc-action">Edit this page</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title">
<span class="chapter-number">28</span>&nbsp; <span class="chapter-title"><strong>读取非标准格式的单细胞数据</strong></span>
</h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><section id="sec-Read_non-standard_10x_files" class="level2" data-number="28.1"><h2 data-number="28.1" class="anchored" data-anchor-id="sec-Read_non-standard_10x_files">
<span class="header-section-number">28.1</span> 读取非标准10X格式文件</h2>
<p>该案例来自文献 <span class="citation" data-cites="xu2022">(<a href="../../references.html#ref-xu2022" role="doc-biblioref">Xu et al. 2022</a>)</span>。该研究为了系统地研究高级别浆膜上皮卵巢癌（HGSOC）的肿瘤内异质性，采用深层单细胞RNA测序技术(scRNA-seq)<strong>对7例初治HGSOC早期和晚期患者及5例年龄匹配的非恶性卵巢组织标本进行肿瘤分析。</strong>共获得59,324个HGSOC和非恶性卵巢组织的单细胞，其中，33264个细胞（56%）来⾃HGSOC肿瘤患者组织，26060个（44%）来⾃⾮恶性卵巢组织标本。原始数据：<a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE184880">GSE184880</a>。</p>
<p><img src="images/640-01.png" class="img-fluid"></p>
<p>可以看到原始数据以10X格式提供:</p>
<p><img src="images/截屏2024-01-08 11.17.00.png" class="img-fluid" width="465"></p>
<p>但是，<strong>没有按照标准的10X格式的三个文件命名，并且没有按照样本分别建立文件夹</strong>。所以需要对其进行整理：</p>
<ul>
<li><p>首先给每个样本建立文件夹</p></li>
<li><p>然后将对应的10X标准格式的三个文件分别复制到各个样本文件夹内</p></li>
<li><p>最后将文件重命名为10X格式要求的标准形式，即<strong>“barcodes.tsv.gz”、“features.tsv.gz”、“matrix.mtx.gz”</strong></p></li>
</ul>
<div class="cell" data-hash="read_sc_data_cache/html/unnamed-chunk-1_39129bf8b407dd8fb905293b98f9e57b">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://stringr.tidyverse.org">stringr</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-hash="read_sc_data_cache/html/unnamed-chunk-2_e5a051a0afb0f5adbb71a37ea9e65e67">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 给每个样本建立文件夹</span></span>
<span><span class="va">files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="st">"data/sc_supplementary/GSE184880_RAW"</span><span class="op">)</span></span>
<span><span class="va">files</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "GSM5599220_Norm1"                   "GSM5599220_Norm1.barcodes.tsv.gz"  
 [3] "GSM5599220_Norm1.genes.tsv.gz"      "GSM5599220_Norm1.matrix.mtx.gz"    
 [5] "GSM5599221_Norm2"                   "GSM5599221_Norm2.barcodes.tsv.gz"  
 [7] "GSM5599221_Norm2.genes.tsv.gz"      "GSM5599221_Norm2.matrix.mtx.gz"    
 [9] "GSM5599222_Norm3"                   "GSM5599222_Norm3.barcodes.tsv.gz"  
[11] "GSM5599222_Norm3.genes.tsv.gz"      "GSM5599222_Norm3.matrix.mtx.gz"    
[13] "GSM5599223_Norm4"                   "GSM5599223_Norm4.barcodes.tsv.gz"  
[15] "GSM5599223_Norm4.genes.tsv.gz"      "GSM5599223_Norm4.matrix.mtx.gz"    
[17] "GSM5599224_Norm5"                   "GSM5599224_Norm5.barcodes.tsv.gz"  
[19] "GSM5599224_Norm5.genes.tsv.gz"      "GSM5599224_Norm5.matrix.mtx.gz"    
[21] "GSM5599225_Cancer1"                 "GSM5599225_Cancer1.barcodes.tsv.gz"
[23] "GSM5599225_Cancer1.genes.tsv.gz"    "GSM5599225_Cancer1.matrix.mtx.gz"  
[25] "GSM5599226_Cancer2"                 "GSM5599226_Cancer2.barcodes.tsv.gz"
[27] "GSM5599226_Cancer2.genes.tsv.gz"    "GSM5599226_Cancer2.matrix.mtx.gz"  
[29] "GSM5599227_Cancer3"                 "GSM5599227_Cancer3.barcodes.tsv.gz"
[31] "GSM5599227_Cancer3.genes.tsv.gz"    "GSM5599227_Cancer3.matrix.mtx.gz"  
[33] "GSM5599228_Cancer4"                 "GSM5599228_Cancer4.barcodes.tsv.gz"
[35] "GSM5599228_Cancer4.genes.tsv.gz"    "GSM5599228_Cancer4.matrix.mtx.gz"  
[37] "GSM5599229_Cancer5"                 "GSM5599229_Cancer5.barcodes.tsv.gz"
[39] "GSM5599229_Cancer5.genes.tsv.gz"    "GSM5599229_Cancer5.matrix.mtx.gz"  
[41] "GSM5599230_Cancer6"                 "GSM5599230_Cancer6.barcodes.tsv.gz"
[43] "GSM5599230_Cancer6.genes.tsv.gz"    "GSM5599230_Cancer6.matrix.mtx.gz"  
[45] "GSM5599231_Cancer7"                 "GSM5599231_Cancer7.barcodes.tsv.gz"
[47] "GSM5599231_Cancer7.genes.tsv.gz"    "GSM5599231_Cancer7.matrix.mtx.gz"  </code></pre>
</div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">dirnames</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">".barcodes.tsv.gz|.genes.tsv.gz|.matrix.mtx.gz"</span>, </span>
<span>                 replacement <span class="op">=</span> <span class="st">""</span>, </span>
<span>                 x <span class="op">=</span> <span class="va">files</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>  </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"data/sc_supplementary/GSE184880_RAW/"</span>, <span class="va">.</span><span class="op">)</span></span>
<span><span class="va">dirnames</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "data/sc_supplementary/GSE184880_RAW/GSM5599220_Norm1"  
 [2] "data/sc_supplementary/GSE184880_RAW/GSM5599221_Norm2"  
 [3] "data/sc_supplementary/GSE184880_RAW/GSM5599222_Norm3"  
 [4] "data/sc_supplementary/GSE184880_RAW/GSM5599223_Norm4"  
 [5] "data/sc_supplementary/GSE184880_RAW/GSM5599224_Norm5"  
 [6] "data/sc_supplementary/GSE184880_RAW/GSM5599225_Cancer1"
 [7] "data/sc_supplementary/GSE184880_RAW/GSM5599226_Cancer2"
 [8] "data/sc_supplementary/GSE184880_RAW/GSM5599227_Cancer3"
 [9] "data/sc_supplementary/GSE184880_RAW/GSM5599228_Cancer4"
[10] "data/sc_supplementary/GSE184880_RAW/GSM5599229_Cancer5"
[11] "data/sc_supplementary/GSE184880_RAW/GSM5599230_Cancer6"
[12] "data/sc_supplementary/GSE184880_RAW/GSM5599231_Cancer7"</code></pre>
</div>
</div>
<div class="cell" data-hash="read_sc_data_cache/html/unnamed-chunk-3_e77e40c1ce7cb3876b69097005c719be">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">dirnames</span>, <span class="va">dir.create</span>, showWarnings <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-hash="read_sc_data_cache/html/unnamed-chunk-4_35aaa4f9fef8c2fa0c493d136d7b0bd0">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">files</span>, <span class="kw">function</span><span class="op">(</span><span class="va">files</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="co"># 复制相应的10X文件到各个样本文件夹内</span></span>
<span>  <span class="va">new_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">".barcodes.tsv.gz|.genes.tsv.gz|.matrix.mtx.gz"</span>, </span>
<span>                  replacement <span class="op">=</span> <span class="st">""</span>, </span>
<span>                  x <span class="op">=</span> <span class="va">files</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>  </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"data/sc_supplementary/GSE184880_RAW/"</span>, <span class="va">.</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/files.html">file.copy</a></span><span class="op">(</span>from <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"data/sc_supplementary/GSE184880_RAW/"</span>, <span class="va">files</span><span class="op">)</span>, </span>
<span>            to <span class="op">=</span> <span class="va">new_dir</span><span class="op">)</span></span>
<span>  <span class="co"># 重命名3个10X文件</span></span>
<span>  <span class="va">new_filename</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"barcodes"</span>, x <span class="op">=</span> <span class="va">files</span><span class="op">)</span>, <span class="st">"barcodes.tsv.gz"</span>, </span>
<span>                         <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"genes"</span>, x <span class="op">=</span> <span class="va">files</span><span class="op">)</span>, <span class="st">"features.tsv.gz"</span>, </span>
<span>                         <span class="st">"matrix.mtx.gz"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/files.html">file.rename</a></span><span class="op">(</span>from <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="va">new_dir</span>, <span class="va">files</span>, sep  <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span>,</span>
<span>              to <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="va">new_dir</span>, <span class="va">new_filename</span>, sep <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">}</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
用setwd简化对文件路径的定义
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>由于在渲染本文档时，knitr会将工作目录恢复为Rproject根目录，所以不能使用getwd指定工作目录。在实际应用中，可以设置工作目录到目标数据文件夹下以简化对文件路径的定义：</p>
<div class="cell" data-hash="read_sc_data_cache/html/unnamed-chunk-5_c015b490890eef7bec9871b6accefe00">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">root_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/getwd.html">getwd</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html">setwd</a></span><span class="op">(</span><span class="st">"data/sc_supplementary/GSE184880_RAW"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-hash="read_sc_data_cache/html/unnamed-chunk-6_9984d215989ba9e5935c412354671db2">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 给每个样本建立文件夹</span></span>
<span><span class="va">files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">files</span></span>
<span><span class="va">dirnames</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">".barcodes.tsv.gz|.genes.tsv.gz|.matrix.mtx.gz"</span>, </span>
<span>                 replacement <span class="op">=</span> <span class="st">""</span>, </span>
<span>                 x <span class="op">=</span> <span class="va">files</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">dirnames</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">dirnames</span>, <span class="va">dir.create</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">files</span>, <span class="kw">function</span><span class="op">(</span><span class="va">files</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="co"># 复制相应的10X文件到各个样本文件夹内</span></span>
<span>  <span class="va">new_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">".barcodes.tsv.gz|.genes.tsv.gz|.matrix.mtx.gz"</span>, </span>
<span>                      replacement <span class="op">=</span> <span class="st">""</span>, </span>
<span>                      x <span class="op">=</span> <span class="va">files</span><span class="op">)</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/files.html">file.copy</a></span><span class="op">(</span>from <span class="op">=</span> <span class="va">files</span>, </span>
<span>            to <span class="op">=</span> <span class="va">new_dir</span><span class="op">)</span></span>
<span>  <span class="co"># 重命名3个10X文件</span></span>
<span>  <span class="va">new_filename</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"barcodes"</span>, x <span class="op">=</span> <span class="va">files</span><span class="op">)</span>, <span class="st">"barcodes.tsv.gz"</span>, </span>
<span>                         <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"genes"</span>, x <span class="op">=</span> <span class="va">files</span><span class="op">)</span>, <span class="st">"features.tsv.gz"</span>, </span>
<span>                         <span class="st">"matrix.mtx.gz"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/files.html">file.rename</a></span><span class="op">(</span>from <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="va">new_dir</span>, <span class="va">files</span>, sep  <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span>,</span>
<span>              to <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="va">new_dir</span>, <span class="va">new_filename</span>, sep <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">}</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-hash="read_sc_data_cache/html/unnamed-chunk-7_37831b1e1c02eabbea2c60c853841bde">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">seurat_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">dirnames</span>, <span class="kw">function</span><span class="op">(</span><span class="va">dirnames</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">dirnames</span><span class="op">)</span></span>
<span>  <span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/CreateSeuratObject.html">CreateSeuratObject</a></span><span class="op">(</span>counts <span class="op">=</span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/Read10X.html">Read10X</a></span><span class="op">(</span><span class="va">dirnames</span><span class="op">)</span>,</span>
<span>                            project <span class="op">=</span>  <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_split.html">str_split</a></span><span class="op">(</span>string <span class="op">=</span> <span class="va">dirnames</span>, </span>
<span>                                                 pattern <span class="op">=</span> <span class="st">"_"</span>, </span>
<span>                                                 simplify <span class="op">=</span> <span class="cn">T</span><span class="op">)</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span>,</span>
<span>                            min.cells <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                            min.features <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="va">merged_seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html">merge</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">seurat_list</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,</span>
<span>                       y <span class="op">=</span> <span class="va">seurat_list</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>                       add.cell.ids <span class="op">=</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_split.html">str_split</a></span><span class="op">(</span>string <span class="op">=</span> <span class="va">dirnames</span>, </span>
<span>                                                pattern <span class="op">=</span> <span class="st">"_"</span>, </span>
<span>                                                simplify <span class="op">=</span> <span class="cn">T</span><span class="op">)</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="va">merged_seurat</span></span>
<span><span class="va">merged_seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/SplitLayers.html">JoinLayers</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">)</span></span>
<span><span class="va">merged_seurat</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">tail</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">$</span><span class="va">orig.ident</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 将样本信息添加到meta.data的新的一列“sample”中:</span></span>
<span><span class="va">merged_seurat</span><span class="op">$</span><span class="va">sample</span> <span class="op">&lt;-</span> <span class="va">merged_seurat</span><span class="op">$</span><span class="va">orig.ident</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>后续如有需要可以随时将工作目录恢复为Rproject根目录：</p>
<div class="cell" data-hash="read_sc_data_cache/html/unnamed-chunk-8_1f82070a950845cff9476979a7bd515f">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html">setwd</a></span><span class="op">(</span><span class="va">root_dir</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<p>接下来，就可以进行常规的Seurat对象构建：</p>
<div class="cell" data-hash="read_sc_data_cache/html/unnamed-chunk-9_2276ec70dba789e7e29f337262db2ed9">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">seurat_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">dirnames</span>, <span class="kw">function</span><span class="op">(</span><span class="va">dirnames</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">dirnames</span><span class="op">)</span></span>
<span>  <span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/CreateSeuratObject.html">CreateSeuratObject</a></span><span class="op">(</span>counts <span class="op">=</span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/Read10X.html">Read10X</a></span><span class="op">(</span><span class="va">dirnames</span><span class="op">)</span>,</span>
<span>                            project <span class="op">=</span>  <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_split.html">str_split</a></span><span class="op">(</span>string <span class="op">=</span> <span class="va">dirnames</span>, </span>
<span>                                                 pattern <span class="op">=</span> <span class="st">"_"</span>, </span>
<span>                                                 simplify <span class="op">=</span> <span class="cn">T</span><span class="op">)</span><span class="op">[</span>,<span class="fl">4</span><span class="op">]</span>,</span>
<span>                            min.cells <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                            min.features <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "data/sc_supplementary/GSE184880_RAW/GSM5599220_Norm1"
[1] "data/sc_supplementary/GSE184880_RAW/GSM5599221_Norm2"
[1] "data/sc_supplementary/GSE184880_RAW/GSM5599222_Norm3"
[1] "data/sc_supplementary/GSE184880_RAW/GSM5599223_Norm4"
[1] "data/sc_supplementary/GSE184880_RAW/GSM5599224_Norm5"
[1] "data/sc_supplementary/GSE184880_RAW/GSM5599225_Cancer1"
[1] "data/sc_supplementary/GSE184880_RAW/GSM5599226_Cancer2"
[1] "data/sc_supplementary/GSE184880_RAW/GSM5599227_Cancer3"
[1] "data/sc_supplementary/GSE184880_RAW/GSM5599228_Cancer4"
[1] "data/sc_supplementary/GSE184880_RAW/GSM5599229_Cancer5"
[1] "data/sc_supplementary/GSE184880_RAW/GSM5599230_Cancer6"
[1] "data/sc_supplementary/GSE184880_RAW/GSM5599231_Cancer7"</code></pre>
</div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">merged_seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html">merge</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">seurat_list</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,</span>
<span>                       y <span class="op">=</span> <span class="va">seurat_list</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>                       add.cell.ids <span class="op">=</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_split.html">str_split</a></span><span class="op">(</span>string <span class="op">=</span> <span class="va">dirnames</span>, </span>
<span>                                                pattern <span class="op">=</span> <span class="st">"_"</span>, </span>
<span>                                                simplify <span class="op">=</span> <span class="cn">T</span><span class="op">)</span><span class="op">[</span>,<span class="fl">4</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="va">merged_seurat</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
21454 features across 60886 samples within 1 assay 
Active assay: RNA (21454 features, 0 variable features)
 12 layers present: counts.Norm1, counts.Norm2, counts.Norm3, counts.Norm4, counts.Norm5, counts.Cancer1, counts.Cancer2, counts.Cancer3, counts.Cancer4, counts.Cancer5, counts.Cancer6, counts.Cancer7</code></pre>
</div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">merged_seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/SplitLayers.html">JoinLayers</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">)</span></span>
<span><span class="va">merged_seurat</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
21454 features across 60886 samples within 1 assay 
Active assay: RNA (21454 features, 0 variable features)
 1 layer present: counts</code></pre>
</div>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                         orig.ident nCount_RNA nFeature_RNA
Norm1_AAACCCAAGGACAGCT-1      Norm1       8191         2218
Norm1_AAACCCAAGGTATCTC-1      Norm1       8720         2396
Norm1_AAACCCACAAGTGCAG-1      Norm1       8554         2518
Norm1_AAACCCACACCCTCTA-1      Norm1       7650         2697
Norm1_AAACCCACACTCCGGA-1      Norm1       9231         2370
Norm1_AAACCCACAGAGGTTG-1      Norm1       5701         1619
Norm1_AAACCCAGTACTTCCC-1      Norm1      10002         2421
Norm1_AAACCCATCTACCACC-1      Norm1       2934         1135
Norm1_AAACGAAAGTAGGCCA-1      Norm1       6955         2198
Norm1_AAACGAAAGTCACACT-1      Norm1       7742         1891</code></pre>
</div>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">tail</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                           orig.ident nCount_RNA nFeature_RNA
Cancer7_TTTGGAGTCGTGAGAG-1    Cancer7       3568         1525
Cancer7_TTTGGTTAGGATCATA-1    Cancer7       3720         1402
Cancer7_TTTGGTTAGTTGGCTT-1    Cancer7       2591         1112
Cancer7_TTTGGTTGTCACCCTT-1    Cancer7      27107         5367
Cancer7_TTTGGTTGTTCCTACC-1    Cancer7      55055         7556
Cancer7_TTTGGTTGTTTCGCTC-1    Cancer7      66262         7369
Cancer7_TTTGGTTTCCACAAGT-1    Cancer7       2342         1143
Cancer7_TTTGTTGCATCCTAAG-1    Cancer7       3498         1440
Cancer7_TTTGTTGGTGACCTGC-1    Cancer7      74316         8182
Cancer7_TTTGTTGGTGATCATC-1    Cancer7       4245          508</code></pre>
</div>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">$</span><span class="va">orig.ident</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "Norm1"   "Norm2"   "Norm3"   "Norm4"   "Norm5"   "Cancer1" "Cancer2"
 [8] "Cancer3" "Cancer4" "Cancer5" "Cancer6" "Cancer7"</code></pre>
</div>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 将样本信息添加到meta.data的新的一列“sample”中:</span></span>
<span><span class="va">merged_seurat</span><span class="op">$</span><span class="va">sample</span> <span class="op">&lt;-</span> <span class="va">merged_seurat</span><span class="op">$</span><span class="va">orig.ident</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr></section><section id="读取raw_counts形式的数据" class="level2" data-number="28.2"><h2 data-number="28.2" class="anchored" data-anchor-id="读取raw_counts形式的数据">
<span class="header-section-number">28.2</span> 读取raw_counts形式的数据</h2>
<p>案例来自<span class="citation" data-cites="Bill2023">(<a href="../../references.html#ref-Bill2023" role="doc-biblioref">Bill et al. 2023</a>)</span>。样本来自头颈部鳞状细胞癌（HNSCC）患者，总共52例标本，包括24例原发肿瘤、16例局部复发和12例多器官远处转移。原发部位包括口腔、喉/下咽、口咽和鼻咽。其中15份样本HPV阳性。原始数据：<a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE234933">GSE234933</a>。</p>
<p>可以看到该数据集没有提供标准的10X格式的3个文件。</p>
<p><img src="images/截屏2024-01-03 20.56.56.png" class="img-fluid"></p>
<p>其中的“raw_counts”文件解压后是52个RDS文件，对应了每个样本的单细胞测序数据，是一个已经包含了表达量、基因名和细胞barcode的矩阵，所以可以通过批量读取这些文件建立Seurat对象。</p>
<p><img src="images/截屏2024-01-09 11.10.35.png" class="img-fluid" width="492"></p>
<p>“sample_annotation.txt”文件包含了对每个样本的详细临床病理资料的注释，因此可以将其添加到meta.data中。</p>
<p><img src="images/截屏2024-01-09 11.11.45.png" class="img-fluid"></p>
<p>“cell_annotation.txt”文件则是原作者对细胞类型的注释，可以不添加。</p>
<p><img src="images/截屏2024-01-09 11.12.14.png" class="img-fluid"></p>
<section id="读取raw_counts矩阵" class="level3"><h3 class="anchored" data-anchor-id="读取raw_counts矩阵">读取raw_counts矩阵</h3>
<div class="cell" data-hash="read_sc_data_cache/html/unnamed-chunk-10_4aedaea4c2d8a1fdb858f1c1988faafd">
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/rm.html">rm</a></span><span class="op">(</span>list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ls.html">ls</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://stringr.tidyverse.org">stringr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tibble.tidyverse.org/">tibble</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-hash="read_sc_data_cache/html/unnamed-chunk-11_d786fed8f4578b526d59e9aab408df15">
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 获取所有rds文件的列表</span></span>
<span><span class="va">file_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="st">"data/sc_supplementary/GSE234933/GSE234933_MGH_HNSCC_gex_raw_counts"</span>, pattern <span class="op">=</span> <span class="st">".rds"</span><span class="op">)</span></span>
<span><span class="va">file_list</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "HN1_gex_raw_counts.rds"       "HN13_gex_raw_counts.rds"     
 [3] "HN14_gex_raw_counts.rds"      "HN17_gex_raw_counts.rds"     
 [5] "HN2_gex_raw_counts.rds"       "HN20_gex_raw_counts.rds"     
 [7] "HN21_gex_raw_counts.rds"      "HN22_gex_raw_counts.rds"     
 [9] "HN23_gex_raw_counts.rds"      "HN25_gex_raw_counts.rds"     
[11] "HN26_gex_raw_counts.rds"      "HN27_gex_raw_counts.rds"     
[13] "HN28_gex_raw_counts.rds"      "HN29_gex_raw_counts.rds"     
[15] "HN30_gex_raw_counts.rds"      "HN31TS_gex_raw_counts.rds"   
[17] "HN32T_gex_raw_counts.rds"     "HN33_gex_raw_counts.rds"     
[19] "HN34_gex_raw_counts.rds"      "HN35_gex_raw_counts.rds"     
[21] "HN37_gex_raw_counts.rds"      "HN38_gex_raw_counts.rds"     
[23] "HN39_gex_raw_counts.rds"      "HN40-3-05_gex_raw_counts.rds"
[25] "HN42_gex_raw_counts.rds"      "HN43_gex_raw_counts.rds"     
[27] "HN45_gex_raw_counts.rds"      "HN46_gex_raw_counts.rds"     
[29] "HN49_gex_raw_counts.rds"      "HN50_gex_raw_counts.rds"     
[31] "HN52_gex_raw_counts.rds"      "HN55_gex_raw_counts.rds"     
[33] "HN57-1_gex_raw_counts.rds"    "HN58_gex_raw_counts.rds"     
[35] "HN59_gex_raw_counts.rds"      "HN60_gex_raw_counts.rds"     
[37] "HN61_gex_raw_counts.rds"      "HN63_gex_raw_counts.rds"     
[39] "HN64_gex_raw_counts.rds"      "HN66_gex_raw_counts.rds"     
[41] "HN67_gex_raw_counts.rds"      "HN68_gex_raw_counts.rds"     
[43] "HN7_gex_raw_counts.rds"       "HN70_gex_raw_counts.rds"     
[45] "HN71_gex_raw_counts.rds"      "HN72_gex_raw_counts.rds"     
[47] "HN73_gex_raw_counts.rds"      "HN74_gex_raw_counts.rds"     
[49] "HN75_gex_raw_counts.rds"      "HN76_gex_raw_counts.rds"     
[51] "HN77_gex_raw_counts.rds"      "HN8_gex_raw_counts.rds"      </code></pre>
</div>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 循环读取每个rds文件的数据并创建Seurat对象</span></span>
<span><span class="va">seurat_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">file_list</span>, <span class="kw">function</span><span class="op">(</span><span class="va">file</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="co"># 拼接文件路径</span></span>
<span>  <span class="va">data.path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"data/sc_supplementary/GSE234933/GSE234933_MGH_HNSCC_gex_raw_counts/"</span>, <span class="va">file</span><span class="op">)</span></span>
<span>  <span class="co"># 创建Seurat对象</span></span>
<span>  <span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/CreateSeuratObject.html">CreateSeuratObject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="va">data.path</span><span class="op">)</span>,</span>
<span>                                   min.features <span class="op">=</span> <span class="fl">200</span>,</span>
<span>                                   min.cells <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="va">seurat_obj</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 合并Seurat对象</span></span>
<span><span class="va">seurat_combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html">merge</a></span><span class="op">(</span><span class="va">seurat_list</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,</span>
<span>                         y <span class="op">=</span> <span class="va">seurat_list</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>                         add.cell.ids <span class="op">=</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_split.html">str_split</a></span><span class="op">(</span>string <span class="op">=</span> <span class="va">file_list</span>, </span>
<span>                                                  pattern <span class="op">=</span> <span class="st">"_"</span>, </span>
<span>                                                  simplify <span class="op">=</span> <span class="cn">T</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">seurat_combined</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
26734 features across 330385 samples within 1 assay 
Active assay: RNA (26734 features, 0 variable features)
 52 layers present: counts.1, counts.2, counts.3, counts.4, counts.5, counts.6, counts.7, counts.8, counts.9, counts.10, counts.11, counts.12, counts.13, counts.14, counts.15, counts.16, counts.17, counts.18, counts.19, counts.20, counts.21, counts.22, counts.23, counts.24, counts.25, counts.26, counts.27, counts.28, counts.29, counts.30, counts.31, counts.32, counts.33, counts.34, counts.35, counts.36, counts.37, counts.38, counts.39, counts.40, counts.41, counts.42, counts.43, counts.44, counts.45, counts.46, counts.47, counts.48, counts.49, counts.50, counts.51, counts.52</code></pre>
</div>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">seurat_combined</span><span class="op">)</span> <span class="co"># 可以看到这个案例实际上不需要添加cell.ids</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                           orig.ident nCount_RNA nFeature_RNA
HN1_HN1_AAACCTGAGTATTGGA-1        HN1        948          574
HN1_HN1_AAACCTGAGTTGCAGG-1        HN1       6745          257
HN1_HN1_AAACCTGCAATAGCAA-1        HN1        677          320
HN1_HN1_AAACCTGCATTTCACT-1        HN1       4540         1129
HN1_HN1_AAACCTGTCGGCGGTT-1        HN1        909          444
HN1_HN1_AAACCTGTCTCGTTTA-1        HN1        633          352
HN1_HN1_AAACGGGAGCAGCCTC-1        HN1       1262          633
HN1_HN1_AAACGGGAGCCCAGCT-1        HN1        513          275
HN1_HN1_AAACGGGAGTACGCGA-1        HN1       1152          416
HN1_HN1_AAACGGGCACATTTCT-1        HN1        426          249</code></pre>
</div>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">seurat_combined</span><span class="op">$</span><span class="va">orig.ident</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "HN1"       "HN13"      "HN14"      "HN17"      "HN2"       "HN20"     
 [7] "HN21"      "HN22"      "HN23"      "HN25"      "HN26"      "HN27"     
[13] "HN28"      "HN29"      "HN30"      "HN31TS"    "HN32T"     "HN33"     
[19] "HN34"      "HN35"      "HN37"      "HN38"      "HN39"      "HN40-3-05"
[25] "HN42"      "HN43"      "HN45"      "HN46"      "HN49"      "HN50"     
[31] "HN52"      "HN55"      "HN57-1"    "HN58"      "HN59"      "HN60"     
[37] "HN61"      "HN63"      "HN64"      "HN66"      "HN67"      "HN68"     
[43] "HN7"       "HN70"      "HN71"      "HN72"      "HN73"      "HN74"     
[49] "HN75"      "HN76"      "HN77"      "HN8"      </code></pre>
</div>
</div>
</section><section id="添加样本信息" class="level3"><h3 class="anchored" data-anchor-id="添加样本信息">添加样本信息</h3>
<div class="cell" data-hash="read_sc_data_cache/html/unnamed-chunk-12_b2f8b2ea2a01a9021c2c16a1d026e987">
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">sample_anno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span><span class="op">(</span><span class="st">"data/sc_supplementary/GSE234933/GSE234933_MGH_HNSCC_sample_annotation.txt"</span>, </span>
<span>                          sep <span class="op">=</span> <span class="st">"\t"</span>, </span>
<span>                          header <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">sample_anno</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Sample    Sex Age HPV.Status Smoking.history Original.anatomic.site
1    HN1   Male  37   Negative             Yes            Oral cavity
2    HN2   Male  55   Negative             Yes            Oral cavity
3    HN7   Male  90   Negative              No            Oral cavity
4    HN8   Male  89   Negative             Yes     Larynx/Hypopharynx
5   HN13 Female  64   Negative             Yes     Larynx/Hypopharynx
6   HN14 Female  53   Negative             Yes             Oropharynx
  Anatomic.location.of.scRNA.seq.specimen
1                                 Primary
2               Distant metastasis (Lung)
3                                 Primary
4            Distant metastasis (sternum)
5               Distant metastasis (skin)
6             Distant metastasis (pleura)</code></pre>
</div>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">sample_anno</span><span class="op">$</span><span class="va">Sample</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "HN1"       "HN2"       "HN7"       "HN8"       "HN13"      "HN14"     
 [7] "HN17"      "HN20"      "HN21"      "HN22"      "HN23"      "HN25"     
[13] "HN26"      "HN27"      "HN28"      "HN29"      "HN30"      "HN31TS"   
[19] "HN32T"     "HN33"      "HN34"      "HN35"      "HN37"      "HN38"     
[25] "HN39"      "HN40-3-05" "HN42"      "HN43"      "HN45"      "HN46"     
[31] "HN49"      "HN50"      "HN52"      "HN55"      "HN57-1"    "HN58"     
[37] "HN59"      "HN60"      "HN61"      "HN63"      "HN64"      "HN66"     
[43] "HN67"      "HN68"      "HN70"      "HN71"      "HN72"      "HN73"     
[49] "HN74"      "HN75"      "HN76"      "HN77"     </code></pre>
</div>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">old_meta.data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames_to_column</a></span><span class="op">(</span><span class="va">seurat_combined</span><span class="op">@</span><span class="va">meta.data</span>, var <span class="op">=</span> <span class="st">"cell_id"</span><span class="op">)</span></span>
<span><span class="va">new_meta.data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html">merge</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">old_meta.data</span>,</span>
<span>                       y <span class="op">=</span> <span class="va">sample_anno</span>,</span>
<span>                       by.x <span class="op">=</span> <span class="st">"orig.ident"</span>,</span>
<span>                       by.y <span class="op">=</span> <span class="st">"Sample"</span>,</span>
<span>                       all.x <span class="op">=</span> <span class="cn">T</span><span class="op">)</span> </span>
<span><span class="va">new_meta.data</span><span class="op">$</span><span class="va">sample</span> <span class="op">&lt;-</span> <span class="va">new_meta.data</span><span class="op">$</span><span class="va">orig.ident</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">new_meta.data</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  orig.ident                    cell_id nCount_RNA nFeature_RNA  Sex Age
1        HN1 HN1_HN1_AAACCTGAGTATTGGA-1        948          574 Male  37
2        HN1 HN1_HN1_AAACCTGAGTTGCAGG-1       6745          257 Male  37
3        HN1 HN1_HN1_AAACCTGCAATAGCAA-1        677          320 Male  37
4        HN1 HN1_HN1_AAACCTGCATTTCACT-1       4540         1129 Male  37
5        HN1 HN1_HN1_AAACCTGTCGGCGGTT-1        909          444 Male  37
6        HN1 HN1_HN1_AAACCTGTCTCGTTTA-1        633          352 Male  37
  HPV.Status Smoking.history Original.anatomic.site
1   Negative             Yes            Oral cavity
2   Negative             Yes            Oral cavity
3   Negative             Yes            Oral cavity
4   Negative             Yes            Oral cavity
5   Negative             Yes            Oral cavity
6   Negative             Yes            Oral cavity
  Anatomic.location.of.scRNA.seq.specimen sample
1                                 Primary    HN1
2                                 Primary    HN1
3                                 Primary    HN1
4                                 Primary    HN1
5                                 Primary    HN1
6                                 Primary    HN1</code></pre>
</div>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">new_meta.data</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">new_meta.data</span><span class="op">$</span><span class="va">cell_id</span></span>
<span><span class="va">new_meta.data</span><span class="op">$</span><span class="va">cell_id</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span><span class="va">new_meta.data</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">6</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">new_meta.data</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">5</span>, <span class="fl">7</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span><span class="op">]</span>, <span class="va">as.factor</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">new_meta.data</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  orig.ident          nCount_RNA      nFeature_RNA      Sex        
 Length:330385      Min.   :   215   Min.   : 184   Female: 65737  
 Class :character   1st Qu.:   897   1st Qu.: 471   Male  :264648  
 Mode  :character   Median :  1886   Median : 845                  
                    Mean   :  3439   Mean   :1107                  
                    3rd Qu.:  3537   3rd Qu.:1348                  
                    Max.   :126539   Max.   :8839                  
                                                                   
      Age          HPV.Status     Smoking.history        Original.anatomic.site
 Min.   :31.00   64     : 29808   No :177238      Larynx/Hypopharynx: 68470    
 1st Qu.:56.00   76     : 25982   Yes:153147      Nasopharynx       : 30573    
 Median :64.00   70     : 23193                   Oral cavity       :107833    
 Mean   :63.42   44     : 21985                   Oropharynx        :115022    
 3rd Qu.:71.00   67     : 21053                   Unknown           :  8487    
 Max.   :91.00   55     : 19302                                                
                 (Other):189062                                                
                 Anatomic.location.of.scRNA.seq.specimen     sample      
 Primary                             :190573             HN63   : 25501  
 Locoregional recurrence             : 59176             HN76   : 15798  
 Distant metastasis (Lung)           : 34541             HN68   : 14851  
 Distant metastasis (Liver)          : 23083             HN77   : 13892  
 Unknown primary of the head and neck:  8487             HN42   : 12062  
 Distant metastasis (skin)           :  7073             HN74   : 11863  
 (Other)                             :  7452             (Other):236418  </code></pre>
</div>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">seurat_combined</span><span class="op">@</span><span class="va">meta.data</span> <span class="op">&lt;-</span> <span class="va">new_meta.data</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section></section><section id="读取txttsvcsv等文本文件表达量矩阵信息" class="level2" data-number="28.3"><h2 data-number="28.3" class="anchored" data-anchor-id="读取txttsvcsv等文本文件表达量矩阵信息">
<span class="header-section-number">28.3</span> 读取txt/tsv/csv等文本文件表达量矩阵信息</h2>
<p>以2020的文章：《Single-Cell Transcriptome Analysis Reveals Dynamic Cell Populations and Differential Gene Expression Patterns in Control and Aneurysmal Human Aortic Tissue》<span class="citation" data-cites="Li2020">(<a href="../../references.html#ref-Li2020" role="doc-biblioref">Li et al. 2020</a>)</span> 举例说明。数据集链接：<a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE155468"><strong>GSE155468</strong></a><strong>。</strong></p>
<p><img src="images/截屏2024-01-10 11.02.14.png" class="img-fluid" width="565"></p>
<p>该案例提供了11个人体主动脉组织样品的单细胞转录组测序数据，其中8名患者为升胸主动脉瘤（ATAA）and 3名患者作为对照。每个样品都是一个独立的txt文本文件，蕴藏着其表达量矩阵信息。</p>
<p>首先我们读取其中一个文件看一下其形式：</p>
<div class="cell" data-hash="read_sc_data_cache/html/unnamed-chunk-13_e7d3fb864e3215f20d9c0d4900b1418a">
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/rm.html">rm</a></span><span class="op">(</span>list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ls.html">ls</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com">data.table</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-hash="read_sc_data_cache/html/unnamed-chunk-14_299eedbcad06d5daca3874247374bf99">
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 列出所有的文件名</span></span>
<span><span class="va">files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="st">"data/sc_supplementary/GSE155468_RAW"</span><span class="op">)</span></span>
<span><span class="va">files</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "GSM4704931_Con4.txt.gz" "GSM4704932_Con6.txt.gz" "GSM4704933_Con9.txt.gz"
 [4] "GSM4704934_TAA1.txt.gz" "GSM4704935_TAA2.txt.gz" "GSM4704936_TAA3.txt.gz"
 [7] "GSM4704937_TAA4.txt.gz" "GSM4704938_TAA5.txt.gz" "GSM4704939_TAA6.txt.gz"
[10] "GSM4704940_TAA7.txt.gz" "GSM4704941_TAA8.txt.gz"</code></pre>
</div>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 简化文件名用作样本名</span></span>
<span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"GSM.*._|.txt.gz"</span>, replacement <span class="op">=</span> <span class="st">""</span>, x <span class="op">=</span> <span class="va">files</span><span class="op">)</span></span>
<span><span class="va">samples</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "Con4" "Con6" "Con9" "TAA1" "TAA2" "TAA3" "TAA4" "TAA5" "TAA6" "TAA7"
[11] "TAA8"</code></pre>
</div>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 试读取其中一个文件</span></span>
<span><span class="va">test</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html">fread</a></span><span class="op">(</span><span class="st">"data/sc_supplementary/GSE155468_RAW/GSM4704931_Con4.txt.gz"</span>, data.table <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span><span class="va">test</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          V1 AAACCCAAGCGGACAT.1 AAACCCAAGCGTCAAG.1
1 AL627309.1                  0                  0
2 AL669831.5                  0                  0
3     FAM87B                  0                  0</code></pre>
</div>
</div>
<p>可以看到，该表达矩阵的问题是行名不是基因名没有作为行名，而是以单独的一列显示，因此需要进行预处理后再用上面类似的方法批量读取和创建Seurat对象：</p>
<div class="cell" data-hash="read_sc_data_cache/html/unnamed-chunk-15_7406ee77160bec6e8cd07f047423b86f">
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">seurat_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">files</span>, <span class="kw">function</span><span class="op">(</span><span class="va">file</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">filepath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"data/sc_supplementary/GSE155468_RAW/"</span>, <span class="va">file</span><span class="op">)</span></span>
<span>  <span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html">fread</a></span><span class="op">(</span><span class="va">filepath</span>, data.table <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">counts</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span></span>
<span>  <span class="va">counts</span> <span class="op">&lt;-</span> <span class="va">counts</span><span class="op">[</span>,<span class="op">-</span><span class="fl">1</span><span class="op">]</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/CreateSeuratObject.html">CreateSeuratObject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">counts</span>,</span>
<span>                     project <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"GSM.*._|.txt.gz"</span>, </span>
<span>                                    replacement <span class="op">=</span> <span class="st">""</span>, </span>
<span>                                    x <span class="op">=</span> <span class="va">file</span><span class="op">)</span>,</span>
<span>                     min.cells <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                     min.features <span class="op">=</span> <span class="fl">300</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">merged_seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html">merge</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">seurat_list</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, </span>
<span>                       y <span class="op">=</span> <span class="va">seurat_list</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span>, </span>
<span>                       add.cell.ids <span class="op">=</span> <span class="va">samples</span><span class="op">)</span></span>
<span><span class="va">merged_seurat</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
21210 features across 47881 samples within 1 assay 
Active assay: RNA (21210 features, 0 variable features)
 11 layers present: counts.Con4, counts.Con6, counts.Con9, counts.TAA1, counts.TAA2, counts.TAA3, counts.TAA4, counts.TAA5, counts.TAA6, counts.TAA7, counts.TAA8</code></pre>
</div>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                        orig.ident nCount_RNA nFeature_RNA
Con4_AAACCCAAGCGGACAT.1       Con4       5700         2067
Con4_AAACCCAAGCGTCAAG.1       Con4      12694         2850
Con4_AAACCCACATTCTCTA.1       Con4      14713         2946
Con4_AAACCCAGTCGACTTA.1       Con4       3000          546
Con4_AAACCCATCCATCAGA.1       Con4      15045         4106
Con4_AAACGAAAGGAACGCT.1       Con4       9877         2793
Con4_AAACGAAAGGTAGATT.1       Con4       6860         2050
Con4_AAACGAAAGTCAATCC.1       Con4       5628         1434
Con4_AAACGAACACGGCACT.1       Con4       9226         2217
Con4_AAACGAAGTTCAAAGA.1       Con4       6064         1743</code></pre>
</div>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 新增样本列</span></span>
<span><span class="va">merged_seurat</span><span class="op">$</span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="va">merged_seurat</span><span class="op">$</span><span class="va">orig.ident</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>简单查看一下细胞/基因数：</p>
<div>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Visualize the number of cell counts per sample</span></span>
<span><span class="va">merged_seurat</span><span class="op">@</span><span class="va">meta.data</span>  <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">samples</span>, fill <span class="op">=</span> <span class="va">samples</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, vjust <span class="op">=</span> <span class="fl">1</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span>, face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"NCells"</span><span class="op">)</span></span>
<span><span class="co"># Visualize the number UMIs/transcripts per cell</span></span>
<span><span class="va">merged_seurat</span><span class="op">@</span><span class="va">meta.data</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">samples</span>, x <span class="op">=</span> <span class="va">nCount_RNA</span>, fill <span class="op">=</span> <span class="va">samples</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span><span class="op">(</span><span class="st">"Cell density"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell quarto-layout-panel" data-hash="read_sc_data_cache/html/unnamed-chunk-16_a7b81f43cc3ee99dc02927afee9b7931">
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="cell-output-display quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img src="read_sc_data_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img src="read_sc_data_files/figure-html/unnamed-chunk-16-2.png" class="img-fluid" width="672"></p>
</div>
</div>
</div>
</div>
<hr>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-2-contents" aria-controls="callout-2" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Session Info
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-2" class="callout-2-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-hash="read_sc_data_cache/html/unnamed-chunk-17_1de9cf55c73c3094263f0433cc2d3119">
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.3.2 (2023-10-31)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Sonoma 14.2.1

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: Asia/Shanghai
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] data.table_1.14.10 stringr_1.5.1      ggplot2_3.4.4      tibble_3.2.1      
[5] Seurat_5.0.1       SeuratObject_5.0.1 sp_2.1-2          

loaded via a namespace (and not attached):
  [1] deldir_2.0-2           pbapply_1.7-2          gridExtra_2.3         
  [4] rlang_1.1.2            magrittr_2.0.3         RcppAnnoy_0.0.21      
  [7] spatstat.geom_3.2-7    matrixStats_1.2.0      ggridges_0.5.5        
 [10] compiler_4.3.2         png_0.1-8              vctrs_0.6.5           
 [13] reshape2_1.4.4         pkgconfig_2.0.3        fastmap_1.1.1         
 [16] ellipsis_0.3.2         labeling_0.4.3         utf8_1.2.4            
 [19] promises_1.2.1         rmarkdown_2.25         purrr_1.0.2           
 [22] xfun_0.41              jsonlite_1.8.8         goftest_1.2-3         
 [25] later_1.3.2            spatstat.utils_3.0-4   irlba_2.3.5.1         
 [28] parallel_4.3.2         cluster_2.1.6          R6_2.5.1              
 [31] ica_1.0-3              stringi_1.8.3          RColorBrewer_1.1-3    
 [34] spatstat.data_3.0-3    reticulate_1.34.0      parallelly_1.36.0     
 [37] lmtest_0.9-40          scattermore_1.2        Rcpp_1.0.11           
 [40] knitr_1.45             tensor_1.5             future.apply_1.11.0   
 [43] zoo_1.8-12             R.utils_2.12.3         sctransform_0.4.1     
 [46] httpuv_1.6.13          Matrix_1.6-4           splines_4.3.2         
 [49] igraph_1.6.0           tidyselect_1.2.0       abind_1.4-5           
 [52] rstudioapi_0.15.0      yaml_2.3.8             spatstat.random_3.2-2 
 [55] codetools_0.2-19       miniUI_0.1.1.1         spatstat.explore_3.2-5
 [58] listenv_0.9.0          lattice_0.22-5         plyr_1.8.9            
 [61] withr_2.5.2            shiny_1.8.0            ROCR_1.0-11           
 [64] evaluate_0.23          Rtsne_0.17             future_1.33.0         
 [67] fastDummies_1.7.3      survival_3.5-7         polyclip_1.10-6       
 [70] fitdistrplus_1.1-11    pillar_1.9.0           KernSmooth_2.23-22    
 [73] plotly_4.10.3          generics_0.1.3         RcppHNSW_0.5.0        
 [76] munsell_0.5.0          scales_1.3.0           globals_0.16.2        
 [79] xtable_1.8-4           glue_1.6.2             lazyeval_0.2.2        
 [82] tools_4.3.2            RSpectra_0.16-1        RANN_2.6.1            
 [85] leiden_0.4.3.1         dotCall64_1.1-1        cowplot_1.1.2         
 [88] grid_4.3.2             tidyr_1.3.0            colorspace_2.1-0      
 [91] nlme_3.1-164           patchwork_1.1.3        cli_3.6.2             
 [94] spatstat.sparse_3.0-3  spam_2.10-0            fansi_1.0.6           
 [97] viridisLite_0.4.2      dplyr_1.1.4            uwot_0.1.16           
[100] gtable_0.3.4           R.methodsS3_1.8.2      digest_0.6.33         
[103] progressr_0.14.0       ggrepel_0.9.4          farver_2.1.1          
[106] htmlwidgets_1.6.4      R.oo_1.25.0            htmltools_0.5.7       
[109] lifecycle_1.0.4        httr_1.4.7             mime_0.12             
[112] MASS_7.3-60           </code></pre>
</div>
</div>
</div>
</div>
</div>


<!-- -->

<div id="refs" class="references csl-bib-body hanging-indent" role="list" style="display: none">
<div id="ref-Bill2023" class="csl-entry" role="listitem">
Bill, Ruben, Pratyaksha Wirapati, Marius Messemaker, Whijae Roh, Beatrice Zitti, Florent Duval, Máté Kiss, et al. 2023. <span>“<span><em>CXCL9:SPP1</em></span> Macrophage Polarity Identifies a Network of Cellular Programs That Control Human Cancers.”</span> <em>Science</em> 381 (6657): 515–24. <a href="https://doi.org/10.1126/science.ade2292">https://doi.org/10.1126/science.ade2292</a>.
</div>
<div id="ref-Li2020" class="csl-entry" role="listitem">
Li, Yanming, Pingping Ren, Ashley Dawson, Hernan G. Vasquez, Waleed Ageedi, Chen Zhang, Wei Luo, et al. 2020. <span>“Single-Cell Transcriptome Analysis Reveals Dynamic Cell Populations and Differential Gene Expression Patterns in Control and Aneurysmal Human Aortic Tissue.”</span> <em>Circulation</em> 142 (14): 1374–88. <a href="https://doi.org/10.1161/circulationaha.120.046528">https://doi.org/10.1161/circulationaha.120.046528</a>.
</div>
<div id="ref-xu2022" class="csl-entry" role="listitem">
Xu, Junfen, Yifeng Fang, Kelie Chen, Sen Li, Sangsang Tang, Yan Ren, Yixuan Cen, et al. 2022. <span>“Single-Cell RNA Sequencing Reveals the Tissue Architecture in Human High-Grade Serous Ovarian Cancer.”</span> <em>Clinical Cancer Research: An Official Journal of the American Association for Cancer Research</em> 28 (16): 3590–3602. <a href="https://doi.org/10.1158/1078-0432.CCR-22-0296">https://doi.org/10.1158/1078-0432.CCR-22-0296</a>.
</div>
</div>
</section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  let localAlternateSentinel = 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var filterRegex = new RegExp("https:\/\/djhcod\.github\.io\/r-notes\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
          // target, if specified
          link.setAttribute("target", "_blank");
      }
    }
});
</script><script src="https://utteranc.es/client.js" repo="https://github.com/djhcod/r-notes" issue-term="pathname" theme="github-light" crossorigin="anonymous" async="">
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="../../single_cell/scRNA-seq_online/09_merged_SC_marker_identification.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">27</span>&nbsp; <span class="chapter-title">寻找marker基因+细胞注释</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../../single_cell/sc_supplementary/universal_marker.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">29</span>&nbsp; <span class="chapter-title">细胞分群通用marker</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb57" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="fu"># **读取非标准格式的单细胞数据**</span></span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a><span class="fu">## 读取非标准10X格式文件 {#sec-Read_non-standard_10x_files}</span></span>
<span id="cb57-4"><a href="#cb57-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-5"><a href="#cb57-5" aria-hidden="true" tabindex="-1"></a>该案例来自文献 <span class="co">[</span><span class="ot">@xu2022</span><span class="co">]</span>。该研究为了系统地研究高级别浆膜上皮卵巢癌（HGSOC）的肿瘤内异质性，采用深层单细胞RNA测序技术(scRNA-seq)**对7例初治HGSOC早期和晚期患者及5例年龄匹配的非恶性卵巢组织标本进行肿瘤分析。**共获得59,324个HGSOC和非恶性卵巢组织的单细胞，其中，33264个细胞（56%）来⾃HGSOC肿瘤患者组织，26060个（44%）来⾃⾮恶性卵巢组织标本。原始数据：<span class="co">[</span><span class="ot">GSE184880</span><span class="co">](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE184880)</span>。</span>
<span id="cb57-6"><a href="#cb57-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-7"><a href="#cb57-7" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/640-01.png)</span></span>
<span id="cb57-8"><a href="#cb57-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-9"><a href="#cb57-9" aria-hidden="true" tabindex="-1"></a>可以看到原始数据以10X格式提供:</span>
<span id="cb57-10"><a href="#cb57-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-11"><a href="#cb57-11" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/截屏2024-01-08%2011.17.00.png)</span>{width="465"}</span>
<span id="cb57-12"><a href="#cb57-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-13"><a href="#cb57-13" aria-hidden="true" tabindex="-1"></a>但是，**没有按照标准的10X格式的三个文件命名，并且没有按照样本分别建立文件夹**。所以需要对其进行整理：</span>
<span id="cb57-14"><a href="#cb57-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-15"><a href="#cb57-15" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>首先给每个样本建立文件夹</span>
<span id="cb57-16"><a href="#cb57-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-17"><a href="#cb57-17" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>然后将对应的10X标准格式的三个文件分别复制到各个样本文件夹内</span>
<span id="cb57-18"><a href="#cb57-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-19"><a href="#cb57-19" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>最后将文件重命名为10X格式要求的标准形式，即**“barcodes.tsv.gz”、"features.tsv.gz"、"matrix.mtx.gz"**</span>
<span id="cb57-20"><a href="#cb57-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-23"><a href="#cb57-23" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-24"><a href="#cb57-24" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb57-25"><a href="#cb57-25" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(stringr)</span>
<span id="cb57-26"><a href="#cb57-26" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-27"><a href="#cb57-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-30"><a href="#cb57-30" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-31"><a href="#cb57-31" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb57-32"><a href="#cb57-32" aria-hidden="true" tabindex="-1"></a><span class="co"># 给每个样本建立文件夹</span></span>
<span id="cb57-33"><a href="#cb57-33" aria-hidden="true" tabindex="-1"></a>files <span class="ot">&lt;-</span> <span class="fu">list.files</span>(<span class="st">"data/sc_supplementary/GSE184880_RAW"</span>)</span>
<span id="cb57-34"><a href="#cb57-34" aria-hidden="true" tabindex="-1"></a>files</span>
<span id="cb57-35"><a href="#cb57-35" aria-hidden="true" tabindex="-1"></a>dirnames <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="at">pattern =</span> <span class="st">".barcodes.tsv.gz|.genes.tsv.gz|.matrix.mtx.gz"</span>, </span>
<span id="cb57-36"><a href="#cb57-36" aria-hidden="true" tabindex="-1"></a>                 <span class="at">replacement =</span> <span class="st">""</span>, </span>
<span id="cb57-37"><a href="#cb57-37" aria-hidden="true" tabindex="-1"></a>                 <span class="at">x =</span> files) <span class="sc">%&gt;%</span>  </span>
<span id="cb57-38"><a href="#cb57-38" aria-hidden="true" tabindex="-1"></a>  <span class="fu">unique</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb57-39"><a href="#cb57-39" aria-hidden="true" tabindex="-1"></a>  <span class="fu">paste0</span>(<span class="st">"data/sc_supplementary/GSE184880_RAW/"</span>, .)</span>
<span id="cb57-40"><a href="#cb57-40" aria-hidden="true" tabindex="-1"></a>dirnames</span>
<span id="cb57-41"><a href="#cb57-41" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-42"><a href="#cb57-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-45"><a href="#cb57-45" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-46"><a href="#cb57-46" aria-hidden="true" tabindex="-1"></a><span class="co">#| output: false</span></span>
<span id="cb57-47"><a href="#cb57-47" aria-hidden="true" tabindex="-1"></a><span class="fu">lapply</span>(dirnames, dir.create, <span class="at">showWarnings =</span> <span class="cn">FALSE</span>)</span>
<span id="cb57-48"><a href="#cb57-48" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-49"><a href="#cb57-49" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-52"><a href="#cb57-52" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-53"><a href="#cb57-53" aria-hidden="true" tabindex="-1"></a><span class="co">#| output: false</span></span>
<span id="cb57-54"><a href="#cb57-54" aria-hidden="true" tabindex="-1"></a><span class="fu">lapply</span>(files, <span class="cf">function</span>(files) {</span>
<span id="cb57-55"><a href="#cb57-55" aria-hidden="true" tabindex="-1"></a>  <span class="co"># 复制相应的10X文件到各个样本文件夹内</span></span>
<span id="cb57-56"><a href="#cb57-56" aria-hidden="true" tabindex="-1"></a>  new_dir <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="at">pattern =</span> <span class="st">".barcodes.tsv.gz|.genes.tsv.gz|.matrix.mtx.gz"</span>, </span>
<span id="cb57-57"><a href="#cb57-57" aria-hidden="true" tabindex="-1"></a>                  <span class="at">replacement =</span> <span class="st">""</span>, </span>
<span id="cb57-58"><a href="#cb57-58" aria-hidden="true" tabindex="-1"></a>                  <span class="at">x =</span> files) <span class="sc">%&gt;%</span>  </span>
<span id="cb57-59"><a href="#cb57-59" aria-hidden="true" tabindex="-1"></a>    <span class="fu">paste0</span>(<span class="st">"data/sc_supplementary/GSE184880_RAW/"</span>, .)</span>
<span id="cb57-60"><a href="#cb57-60" aria-hidden="true" tabindex="-1"></a>  <span class="fu">file.copy</span>(<span class="at">from =</span> <span class="fu">paste0</span>(<span class="st">"data/sc_supplementary/GSE184880_RAW/"</span>, files), </span>
<span id="cb57-61"><a href="#cb57-61" aria-hidden="true" tabindex="-1"></a>            <span class="at">to =</span> new_dir)</span>
<span id="cb57-62"><a href="#cb57-62" aria-hidden="true" tabindex="-1"></a>  <span class="co"># 重命名3个10X文件</span></span>
<span id="cb57-63"><a href="#cb57-63" aria-hidden="true" tabindex="-1"></a>  new_filename <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(<span class="fu">grepl</span>(<span class="at">pattern =</span> <span class="st">"barcodes"</span>, <span class="at">x =</span> files), <span class="st">"barcodes.tsv.gz"</span>, </span>
<span id="cb57-64"><a href="#cb57-64" aria-hidden="true" tabindex="-1"></a>                         <span class="fu">ifelse</span>(<span class="fu">grepl</span>(<span class="at">pattern =</span> <span class="st">"genes"</span>, <span class="at">x =</span> files), <span class="st">"features.tsv.gz"</span>, </span>
<span id="cb57-65"><a href="#cb57-65" aria-hidden="true" tabindex="-1"></a>                         <span class="st">"matrix.mtx.gz"</span>))</span>
<span id="cb57-66"><a href="#cb57-66" aria-hidden="true" tabindex="-1"></a>  <span class="fu">file.rename</span>(<span class="at">from =</span> <span class="fu">paste</span>(new_dir, files, <span class="at">sep  =</span> <span class="st">"/"</span>),</span>
<span id="cb57-67"><a href="#cb57-67" aria-hidden="true" tabindex="-1"></a>              <span class="at">to =</span> <span class="fu">paste</span>(new_dir, new_filename, <span class="at">sep =</span> <span class="st">"/"</span>))</span>
<span id="cb57-68"><a href="#cb57-68" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb57-69"><a href="#cb57-69" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-70"><a href="#cb57-70" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-71"><a href="#cb57-71" aria-hidden="true" tabindex="-1"></a>::: {.callout-tip collapse="true"}</span>
<span id="cb57-72"><a href="#cb57-72" aria-hidden="true" tabindex="-1"></a><span class="fu">###### 用setwd简化对文件路径的定义</span></span>
<span id="cb57-73"><a href="#cb57-73" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-74"><a href="#cb57-74" aria-hidden="true" tabindex="-1"></a>由于在渲染本文档时，knitr会将工作目录恢复为Rproject根目录，所以不能使用getwd指定工作目录。在实际应用中，可以设置工作目录到目标数据文件夹下以简化对文件路径的定义：</span>
<span id="cb57-75"><a href="#cb57-75" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-78"><a href="#cb57-78" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-79"><a href="#cb57-79" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb57-80"><a href="#cb57-80" aria-hidden="true" tabindex="-1"></a>root_dir <span class="ot">&lt;-</span> <span class="fu">getwd</span>()</span>
<span id="cb57-81"><a href="#cb57-81" aria-hidden="true" tabindex="-1"></a><span class="fu">setwd</span>(<span class="st">"data/sc_supplementary/GSE184880_RAW"</span>)</span>
<span id="cb57-82"><a href="#cb57-82" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-83"><a href="#cb57-83" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-86"><a href="#cb57-86" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-87"><a href="#cb57-87" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb57-88"><a href="#cb57-88" aria-hidden="true" tabindex="-1"></a><span class="co"># 给每个样本建立文件夹</span></span>
<span id="cb57-89"><a href="#cb57-89" aria-hidden="true" tabindex="-1"></a>files <span class="ot">&lt;-</span> <span class="fu">list.files</span>()</span>
<span id="cb57-90"><a href="#cb57-90" aria-hidden="true" tabindex="-1"></a>files</span>
<span id="cb57-91"><a href="#cb57-91" aria-hidden="true" tabindex="-1"></a>dirnames <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="at">pattern =</span> <span class="st">".barcodes.tsv.gz|.genes.tsv.gz|.matrix.mtx.gz"</span>, </span>
<span id="cb57-92"><a href="#cb57-92" aria-hidden="true" tabindex="-1"></a>                 <span class="at">replacement =</span> <span class="st">""</span>, </span>
<span id="cb57-93"><a href="#cb57-93" aria-hidden="true" tabindex="-1"></a>                 <span class="at">x =</span> files) <span class="sc">|&gt;</span> <span class="fu">unique</span>()</span>
<span id="cb57-94"><a href="#cb57-94" aria-hidden="true" tabindex="-1"></a>dirnames</span>
<span id="cb57-95"><a href="#cb57-95" aria-hidden="true" tabindex="-1"></a><span class="fu">lapply</span>(dirnames, dir.create)</span>
<span id="cb57-96"><a href="#cb57-96" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-97"><a href="#cb57-97" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-98"><a href="#cb57-98" aria-hidden="true" tabindex="-1"></a><span class="fu">lapply</span>(files, <span class="cf">function</span>(files) {</span>
<span id="cb57-99"><a href="#cb57-99" aria-hidden="true" tabindex="-1"></a>  <span class="co"># 复制相应的10X文件到各个样本文件夹内</span></span>
<span id="cb57-100"><a href="#cb57-100" aria-hidden="true" tabindex="-1"></a>  new_dir <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="at">pattern =</span> <span class="st">".barcodes.tsv.gz|.genes.tsv.gz|.matrix.mtx.gz"</span>, </span>
<span id="cb57-101"><a href="#cb57-101" aria-hidden="true" tabindex="-1"></a>                      <span class="at">replacement =</span> <span class="st">""</span>, </span>
<span id="cb57-102"><a href="#cb57-102" aria-hidden="true" tabindex="-1"></a>                      <span class="at">x =</span> files) </span>
<span id="cb57-103"><a href="#cb57-103" aria-hidden="true" tabindex="-1"></a>  <span class="fu">file.copy</span>(<span class="at">from =</span> files, </span>
<span id="cb57-104"><a href="#cb57-104" aria-hidden="true" tabindex="-1"></a>            <span class="at">to =</span> new_dir)</span>
<span id="cb57-105"><a href="#cb57-105" aria-hidden="true" tabindex="-1"></a>  <span class="co"># 重命名3个10X文件</span></span>
<span id="cb57-106"><a href="#cb57-106" aria-hidden="true" tabindex="-1"></a>  new_filename <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(<span class="fu">grepl</span>(<span class="at">pattern =</span> <span class="st">"barcodes"</span>, <span class="at">x =</span> files), <span class="st">"barcodes.tsv.gz"</span>, </span>
<span id="cb57-107"><a href="#cb57-107" aria-hidden="true" tabindex="-1"></a>                         <span class="fu">ifelse</span>(<span class="fu">grepl</span>(<span class="at">pattern =</span> <span class="st">"genes"</span>, <span class="at">x =</span> files), <span class="st">"features.tsv.gz"</span>, </span>
<span id="cb57-108"><a href="#cb57-108" aria-hidden="true" tabindex="-1"></a>                         <span class="st">"matrix.mtx.gz"</span>))</span>
<span id="cb57-109"><a href="#cb57-109" aria-hidden="true" tabindex="-1"></a>  <span class="fu">file.rename</span>(<span class="at">from =</span> <span class="fu">paste</span>(new_dir, files, <span class="at">sep  =</span> <span class="st">"/"</span>),</span>
<span id="cb57-110"><a href="#cb57-110" aria-hidden="true" tabindex="-1"></a>              <span class="at">to =</span> <span class="fu">paste</span>(new_dir, new_filename, <span class="at">sep =</span> <span class="st">"/"</span>))</span>
<span id="cb57-111"><a href="#cb57-111" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb57-112"><a href="#cb57-112" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-113"><a href="#cb57-113" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-116"><a href="#cb57-116" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-117"><a href="#cb57-117" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb57-118"><a href="#cb57-118" aria-hidden="true" tabindex="-1"></a>seurat_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(dirnames, <span class="cf">function</span>(dirnames) {</span>
<span id="cb57-119"><a href="#cb57-119" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(dirnames)</span>
<span id="cb57-120"><a href="#cb57-120" aria-hidden="true" tabindex="-1"></a>  sce <span class="ot">&lt;-</span> <span class="fu">CreateSeuratObject</span>(<span class="at">counts =</span>  <span class="fu">Read10X</span>(dirnames),</span>
<span id="cb57-121"><a href="#cb57-121" aria-hidden="true" tabindex="-1"></a>                            <span class="at">project =</span>  <span class="fu">str_split</span>(<span class="at">string =</span> dirnames, </span>
<span id="cb57-122"><a href="#cb57-122" aria-hidden="true" tabindex="-1"></a>                                                 <span class="at">pattern =</span> <span class="st">"_"</span>, </span>
<span id="cb57-123"><a href="#cb57-123" aria-hidden="true" tabindex="-1"></a>                                                 <span class="at">simplify =</span> T)[,<span class="dv">2</span>],</span>
<span id="cb57-124"><a href="#cb57-124" aria-hidden="true" tabindex="-1"></a>                            <span class="at">min.cells =</span> <span class="dv">5</span>,</span>
<span id="cb57-125"><a href="#cb57-125" aria-hidden="true" tabindex="-1"></a>                            <span class="at">min.features =</span> <span class="dv">500</span>)</span>
<span id="cb57-126"><a href="#cb57-126" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(sce)</span>
<span id="cb57-127"><a href="#cb57-127" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb57-128"><a href="#cb57-128" aria-hidden="true" tabindex="-1"></a>merged_seurat <span class="ot">&lt;-</span> <span class="fu">merge</span>(<span class="at">x =</span> seurat_list[[<span class="dv">1</span>]],</span>
<span id="cb57-129"><a href="#cb57-129" aria-hidden="true" tabindex="-1"></a>                       <span class="at">y =</span> seurat_list[<span class="sc">-</span><span class="dv">1</span>],</span>
<span id="cb57-130"><a href="#cb57-130" aria-hidden="true" tabindex="-1"></a>                       <span class="at">add.cell.ids =</span> <span class="fu">str_split</span>(<span class="at">string =</span> dirnames, </span>
<span id="cb57-131"><a href="#cb57-131" aria-hidden="true" tabindex="-1"></a>                                                <span class="at">pattern =</span> <span class="st">"_"</span>, </span>
<span id="cb57-132"><a href="#cb57-132" aria-hidden="true" tabindex="-1"></a>                                                <span class="at">simplify =</span> T)[,<span class="dv">2</span>])</span>
<span id="cb57-133"><a href="#cb57-133" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-134"><a href="#cb57-134" aria-hidden="true" tabindex="-1"></a>merged_seurat</span>
<span id="cb57-135"><a href="#cb57-135" aria-hidden="true" tabindex="-1"></a>merged_seurat <span class="ot">&lt;-</span> <span class="fu">JoinLayers</span>(merged_seurat)</span>
<span id="cb57-136"><a href="#cb57-136" aria-hidden="true" tabindex="-1"></a>merged_seurat</span>
<span id="cb57-137"><a href="#cb57-137" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(merged_seurat)</span>
<span id="cb57-138"><a href="#cb57-138" aria-hidden="true" tabindex="-1"></a><span class="fu">tail</span>(merged_seurat)</span>
<span id="cb57-139"><a href="#cb57-139" aria-hidden="true" tabindex="-1"></a><span class="fu">unique</span>(merged_seurat<span class="sc">$</span>orig.ident)</span>
<span id="cb57-140"><a href="#cb57-140" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-141"><a href="#cb57-141" aria-hidden="true" tabindex="-1"></a><span class="co"># 将样本信息添加到meta.data的新的一列“sample”中:</span></span>
<span id="cb57-142"><a href="#cb57-142" aria-hidden="true" tabindex="-1"></a>merged_seurat<span class="sc">$</span>sample <span class="ot">&lt;-</span> merged_seurat<span class="sc">$</span>orig.ident</span>
<span id="cb57-143"><a href="#cb57-143" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-144"><a href="#cb57-144" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-145"><a href="#cb57-145" aria-hidden="true" tabindex="-1"></a>后续如有需要可以随时将工作目录恢复为Rproject根目录：</span>
<span id="cb57-146"><a href="#cb57-146" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-149"><a href="#cb57-149" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-150"><a href="#cb57-150" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb57-151"><a href="#cb57-151" aria-hidden="true" tabindex="-1"></a><span class="fu">setwd</span>(root_dir)</span>
<span id="cb57-152"><a href="#cb57-152" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-153"><a href="#cb57-153" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb57-154"><a href="#cb57-154" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-155"><a href="#cb57-155" aria-hidden="true" tabindex="-1"></a>接下来，就可以进行常规的Seurat对象构建：</span>
<span id="cb57-156"><a href="#cb57-156" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-159"><a href="#cb57-159" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-160"><a href="#cb57-160" aria-hidden="true" tabindex="-1"></a>seurat_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(dirnames, <span class="cf">function</span>(dirnames) {</span>
<span id="cb57-161"><a href="#cb57-161" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(dirnames)</span>
<span id="cb57-162"><a href="#cb57-162" aria-hidden="true" tabindex="-1"></a>  sce <span class="ot">&lt;-</span> <span class="fu">CreateSeuratObject</span>(<span class="at">counts =</span>  <span class="fu">Read10X</span>(dirnames),</span>
<span id="cb57-163"><a href="#cb57-163" aria-hidden="true" tabindex="-1"></a>                            <span class="at">project =</span>  <span class="fu">str_split</span>(<span class="at">string =</span> dirnames, </span>
<span id="cb57-164"><a href="#cb57-164" aria-hidden="true" tabindex="-1"></a>                                                 <span class="at">pattern =</span> <span class="st">"_"</span>, </span>
<span id="cb57-165"><a href="#cb57-165" aria-hidden="true" tabindex="-1"></a>                                                 <span class="at">simplify =</span> T)[,<span class="dv">4</span>],</span>
<span id="cb57-166"><a href="#cb57-166" aria-hidden="true" tabindex="-1"></a>                            <span class="at">min.cells =</span> <span class="dv">5</span>,</span>
<span id="cb57-167"><a href="#cb57-167" aria-hidden="true" tabindex="-1"></a>                            <span class="at">min.features =</span> <span class="dv">500</span>)</span>
<span id="cb57-168"><a href="#cb57-168" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(sce)</span>
<span id="cb57-169"><a href="#cb57-169" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb57-170"><a href="#cb57-170" aria-hidden="true" tabindex="-1"></a>merged_seurat <span class="ot">&lt;-</span> <span class="fu">merge</span>(<span class="at">x =</span> seurat_list[[<span class="dv">1</span>]],</span>
<span id="cb57-171"><a href="#cb57-171" aria-hidden="true" tabindex="-1"></a>                       <span class="at">y =</span> seurat_list[<span class="sc">-</span><span class="dv">1</span>],</span>
<span id="cb57-172"><a href="#cb57-172" aria-hidden="true" tabindex="-1"></a>                       <span class="at">add.cell.ids =</span> <span class="fu">str_split</span>(<span class="at">string =</span> dirnames, </span>
<span id="cb57-173"><a href="#cb57-173" aria-hidden="true" tabindex="-1"></a>                                                <span class="at">pattern =</span> <span class="st">"_"</span>, </span>
<span id="cb57-174"><a href="#cb57-174" aria-hidden="true" tabindex="-1"></a>                                                <span class="at">simplify =</span> T)[,<span class="dv">4</span>])</span>
<span id="cb57-175"><a href="#cb57-175" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-176"><a href="#cb57-176" aria-hidden="true" tabindex="-1"></a>merged_seurat</span>
<span id="cb57-177"><a href="#cb57-177" aria-hidden="true" tabindex="-1"></a>merged_seurat <span class="ot">&lt;-</span> <span class="fu">JoinLayers</span>(merged_seurat)</span>
<span id="cb57-178"><a href="#cb57-178" aria-hidden="true" tabindex="-1"></a>merged_seurat</span>
<span id="cb57-179"><a href="#cb57-179" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(merged_seurat)</span>
<span id="cb57-180"><a href="#cb57-180" aria-hidden="true" tabindex="-1"></a><span class="fu">tail</span>(merged_seurat)</span>
<span id="cb57-181"><a href="#cb57-181" aria-hidden="true" tabindex="-1"></a><span class="fu">unique</span>(merged_seurat<span class="sc">$</span>orig.ident)</span>
<span id="cb57-182"><a href="#cb57-182" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-183"><a href="#cb57-183" aria-hidden="true" tabindex="-1"></a><span class="co"># 将样本信息添加到meta.data的新的一列“sample”中:</span></span>
<span id="cb57-184"><a href="#cb57-184" aria-hidden="true" tabindex="-1"></a>merged_seurat<span class="sc">$</span>sample <span class="ot">&lt;-</span> merged_seurat<span class="sc">$</span>orig.ident</span>
<span id="cb57-185"><a href="#cb57-185" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-186"><a href="#cb57-186" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-187"><a href="#cb57-187" aria-hidden="true" tabindex="-1"></a>------------------------------------------------------------------------</span>
<span id="cb57-188"><a href="#cb57-188" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-189"><a href="#cb57-189" aria-hidden="true" tabindex="-1"></a><span class="fu">## 读取raw_counts形式的数据</span></span>
<span id="cb57-190"><a href="#cb57-190" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-191"><a href="#cb57-191" aria-hidden="true" tabindex="-1"></a>案例来自<span class="co">[</span><span class="ot">@Bill2023</span><span class="co">]</span>。样本来自头颈部鳞状细胞癌（HNSCC）患者，总共52例标本，包括24例原发肿瘤、16例局部复发和12例多器官远处转移。原发部位包括口腔、喉/下咽、口咽和鼻咽。其中15份样本HPV阳性。原始数据：<span class="co">[</span><span class="ot">GSE234933</span><span class="co">](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE234933)</span>。</span>
<span id="cb57-192"><a href="#cb57-192" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-193"><a href="#cb57-193" aria-hidden="true" tabindex="-1"></a>可以看到该数据集没有提供标准的10X格式的3个文件。</span>
<span id="cb57-194"><a href="#cb57-194" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-195"><a href="#cb57-195" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/截屏2024-01-03%2020.56.56.png)</span></span>
<span id="cb57-196"><a href="#cb57-196" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-197"><a href="#cb57-197" aria-hidden="true" tabindex="-1"></a>其中的“raw_counts”文件解压后是52个RDS文件，对应了每个样本的单细胞测序数据，是一个已经包含了表达量、基因名和细胞barcode的矩阵，所以可以通过批量读取这些文件建立Seurat对象。</span>
<span id="cb57-198"><a href="#cb57-198" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-199"><a href="#cb57-199" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/截屏2024-01-09%2011.10.35.png)</span>{width="492"}</span>
<span id="cb57-200"><a href="#cb57-200" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-201"><a href="#cb57-201" aria-hidden="true" tabindex="-1"></a>“sample_annotation.txt”文件包含了对每个样本的详细临床病理资料的注释，因此可以将其添加到meta.data中。</span>
<span id="cb57-202"><a href="#cb57-202" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-203"><a href="#cb57-203" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/截屏2024-01-09%2011.11.45.png)</span></span>
<span id="cb57-204"><a href="#cb57-204" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-205"><a href="#cb57-205" aria-hidden="true" tabindex="-1"></a>“cell_annotation.txt”文件则是原作者对细胞类型的注释，可以不添加。</span>
<span id="cb57-206"><a href="#cb57-206" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-207"><a href="#cb57-207" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/截屏2024-01-09%2011.12.14.png)</span></span>
<span id="cb57-208"><a href="#cb57-208" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-209"><a href="#cb57-209" aria-hidden="true" tabindex="-1"></a><span class="fu">### 读取raw_counts矩阵</span></span>
<span id="cb57-210"><a href="#cb57-210" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-213"><a href="#cb57-213" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-214"><a href="#cb57-214" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(<span class="at">list =</span> <span class="fu">ls</span>())</span>
<span id="cb57-215"><a href="#cb57-215" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb57-216"><a href="#cb57-216" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(stringr)</span>
<span id="cb57-217"><a href="#cb57-217" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tibble)</span>
<span id="cb57-218"><a href="#cb57-218" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-219"><a href="#cb57-219" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-222"><a href="#cb57-222" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-223"><a href="#cb57-223" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb57-224"><a href="#cb57-224" aria-hidden="true" tabindex="-1"></a><span class="co"># 获取所有rds文件的列表</span></span>
<span id="cb57-225"><a href="#cb57-225" aria-hidden="true" tabindex="-1"></a>file_list <span class="ot">&lt;-</span> <span class="fu">list.files</span>(<span class="st">"data/sc_supplementary/GSE234933/GSE234933_MGH_HNSCC_gex_raw_counts"</span>, <span class="at">pattern =</span> <span class="st">".rds"</span>)</span>
<span id="cb57-226"><a href="#cb57-226" aria-hidden="true" tabindex="-1"></a>file_list</span>
<span id="cb57-227"><a href="#cb57-227" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-228"><a href="#cb57-228" aria-hidden="true" tabindex="-1"></a><span class="co"># 循环读取每个rds文件的数据并创建Seurat对象</span></span>
<span id="cb57-229"><a href="#cb57-229" aria-hidden="true" tabindex="-1"></a>seurat_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(file_list, <span class="cf">function</span>(file) {</span>
<span id="cb57-230"><a href="#cb57-230" aria-hidden="true" tabindex="-1"></a>  <span class="co"># 拼接文件路径</span></span>
<span id="cb57-231"><a href="#cb57-231" aria-hidden="true" tabindex="-1"></a>  data.path <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"data/sc_supplementary/GSE234933/GSE234933_MGH_HNSCC_gex_raw_counts/"</span>, file)</span>
<span id="cb57-232"><a href="#cb57-232" aria-hidden="true" tabindex="-1"></a>  <span class="co"># 创建Seurat对象</span></span>
<span id="cb57-233"><a href="#cb57-233" aria-hidden="true" tabindex="-1"></a>  seurat_obj <span class="ot">&lt;-</span> <span class="fu">CreateSeuratObject</span>(<span class="at">counts =</span> <span class="fu">readRDS</span>(data.path),</span>
<span id="cb57-234"><a href="#cb57-234" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">min.features =</span> <span class="dv">200</span>,</span>
<span id="cb57-235"><a href="#cb57-235" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">min.cells =</span> <span class="dv">3</span>)</span>
<span id="cb57-236"><a href="#cb57-236" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(seurat_obj)</span>
<span id="cb57-237"><a href="#cb57-237" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb57-238"><a href="#cb57-238" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-239"><a href="#cb57-239" aria-hidden="true" tabindex="-1"></a><span class="co"># 合并Seurat对象</span></span>
<span id="cb57-240"><a href="#cb57-240" aria-hidden="true" tabindex="-1"></a>seurat_combined <span class="ot">&lt;-</span> <span class="fu">merge</span>(seurat_list[[<span class="dv">1</span>]],</span>
<span id="cb57-241"><a href="#cb57-241" aria-hidden="true" tabindex="-1"></a>                         <span class="at">y =</span> seurat_list[<span class="sc">-</span><span class="dv">1</span>],</span>
<span id="cb57-242"><a href="#cb57-242" aria-hidden="true" tabindex="-1"></a>                         <span class="at">add.cell.ids =</span> <span class="fu">str_split</span>(<span class="at">string =</span> file_list, </span>
<span id="cb57-243"><a href="#cb57-243" aria-hidden="true" tabindex="-1"></a>                                                  <span class="at">pattern =</span> <span class="st">"_"</span>, </span>
<span id="cb57-244"><a href="#cb57-244" aria-hidden="true" tabindex="-1"></a>                                                  <span class="at">simplify =</span> T)[,<span class="dv">1</span>])</span>
<span id="cb57-245"><a href="#cb57-245" aria-hidden="true" tabindex="-1"></a>seurat_combined</span>
<span id="cb57-246"><a href="#cb57-246" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(seurat_combined) <span class="co"># 可以看到这个案例实际上不需要添加cell.ids</span></span>
<span id="cb57-247"><a href="#cb57-247" aria-hidden="true" tabindex="-1"></a><span class="fu">unique</span>(seurat_combined<span class="sc">$</span>orig.ident)</span>
<span id="cb57-248"><a href="#cb57-248" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-249"><a href="#cb57-249" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-250"><a href="#cb57-250" aria-hidden="true" tabindex="-1"></a><span class="fu">### 添加样本信息</span></span>
<span id="cb57-251"><a href="#cb57-251" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-254"><a href="#cb57-254" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-255"><a href="#cb57-255" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb57-256"><a href="#cb57-256" aria-hidden="true" tabindex="-1"></a>sample_anno <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"data/sc_supplementary/GSE234933/GSE234933_MGH_HNSCC_sample_annotation.txt"</span>, </span>
<span id="cb57-257"><a href="#cb57-257" aria-hidden="true" tabindex="-1"></a>                          <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, </span>
<span id="cb57-258"><a href="#cb57-258" aria-hidden="true" tabindex="-1"></a>                          <span class="at">header =</span> T)</span>
<span id="cb57-259"><a href="#cb57-259" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(sample_anno)</span>
<span id="cb57-260"><a href="#cb57-260" aria-hidden="true" tabindex="-1"></a>sample_anno<span class="sc">$</span>Sample</span>
<span id="cb57-261"><a href="#cb57-261" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-262"><a href="#cb57-262" aria-hidden="true" tabindex="-1"></a>old_meta.data <span class="ot">&lt;-</span> <span class="fu">rownames_to_column</span>(seurat_combined<span class="sc">@</span>meta.data, <span class="at">var =</span> <span class="st">"cell_id"</span>)</span>
<span id="cb57-263"><a href="#cb57-263" aria-hidden="true" tabindex="-1"></a>new_meta.data <span class="ot">&lt;-</span> <span class="fu">merge</span>(<span class="at">x =</span> old_meta.data,</span>
<span id="cb57-264"><a href="#cb57-264" aria-hidden="true" tabindex="-1"></a>                       <span class="at">y =</span> sample_anno,</span>
<span id="cb57-265"><a href="#cb57-265" aria-hidden="true" tabindex="-1"></a>                       <span class="at">by.x =</span> <span class="st">"orig.ident"</span>,</span>
<span id="cb57-266"><a href="#cb57-266" aria-hidden="true" tabindex="-1"></a>                       <span class="at">by.y =</span> <span class="st">"Sample"</span>,</span>
<span id="cb57-267"><a href="#cb57-267" aria-hidden="true" tabindex="-1"></a>                       <span class="at">all.x =</span> T) </span>
<span id="cb57-268"><a href="#cb57-268" aria-hidden="true" tabindex="-1"></a>new_meta.data<span class="sc">$</span>sample <span class="ot">&lt;-</span> new_meta.data<span class="sc">$</span>orig.ident</span>
<span id="cb57-269"><a href="#cb57-269" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(new_meta.data)</span>
<span id="cb57-270"><a href="#cb57-270" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(new_meta.data) <span class="ot">&lt;-</span> new_meta.data<span class="sc">$</span>cell_id</span>
<span id="cb57-271"><a href="#cb57-271" aria-hidden="true" tabindex="-1"></a>new_meta.data<span class="sc">$</span>cell_id <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb57-272"><a href="#cb57-272" aria-hidden="true" tabindex="-1"></a>new_meta.data[,<span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">6</span><span class="sc">:</span><span class="dv">10</span>)] <span class="ot">&lt;-</span> <span class="fu">lapply</span>(new_meta.data[,<span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">5</span>, <span class="dv">7</span><span class="sc">:</span><span class="dv">10</span>)], as.factor)</span>
<span id="cb57-273"><a href="#cb57-273" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(new_meta.data)</span>
<span id="cb57-274"><a href="#cb57-274" aria-hidden="true" tabindex="-1"></a>seurat_combined<span class="sc">@</span>meta.data <span class="ot">&lt;-</span> new_meta.data</span>
<span id="cb57-275"><a href="#cb57-275" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-276"><a href="#cb57-276" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-277"><a href="#cb57-277" aria-hidden="true" tabindex="-1"></a><span class="fu">## 读取txt/tsv/csv等文本文件表达量矩阵信息</span></span>
<span id="cb57-278"><a href="#cb57-278" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-279"><a href="#cb57-279" aria-hidden="true" tabindex="-1"></a>以2020的文章：《Single-Cell Transcriptome Analysis Reveals Dynamic Cell Populations and Differential Gene Expression Patterns in Control and Aneurysmal Human Aortic Tissue》<span class="co">[</span><span class="ot">@Li2020</span><span class="co">]</span> 举例说明。数据集链接：<span class="co">[</span><span class="ot">**GSE155468**</span><span class="co">](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE155468)</span>**。**</span>
<span id="cb57-280"><a href="#cb57-280" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-281"><a href="#cb57-281" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/截屏2024-01-10%2011.02.14.png)</span>{width="565"}</span>
<span id="cb57-282"><a href="#cb57-282" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-283"><a href="#cb57-283" aria-hidden="true" tabindex="-1"></a>该案例提供了11个人体主动脉组织样品的单细胞转录组测序数据，其中8名患者为升胸主动脉瘤（ATAA）and 3名患者作为对照。每个样品都是一个独立的txt文本文件，蕴藏着其表达量矩阵信息。</span>
<span id="cb57-284"><a href="#cb57-284" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-285"><a href="#cb57-285" aria-hidden="true" tabindex="-1"></a>首先我们读取其中一个文件看一下其形式：</span>
<span id="cb57-286"><a href="#cb57-286" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-289"><a href="#cb57-289" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-290"><a href="#cb57-290" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(<span class="at">list =</span> <span class="fu">ls</span>())</span>
<span id="cb57-291"><a href="#cb57-291" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(data.table)</span>
<span id="cb57-292"><a href="#cb57-292" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb57-293"><a href="#cb57-293" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb57-294"><a href="#cb57-294" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-295"><a href="#cb57-295" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-298"><a href="#cb57-298" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-299"><a href="#cb57-299" aria-hidden="true" tabindex="-1"></a><span class="co"># 列出所有的文件名</span></span>
<span id="cb57-300"><a href="#cb57-300" aria-hidden="true" tabindex="-1"></a>files <span class="ot">&lt;-</span> <span class="fu">list.files</span>(<span class="st">"data/sc_supplementary/GSE155468_RAW"</span>)</span>
<span id="cb57-301"><a href="#cb57-301" aria-hidden="true" tabindex="-1"></a>files</span>
<span id="cb57-302"><a href="#cb57-302" aria-hidden="true" tabindex="-1"></a><span class="co"># 简化文件名用作样本名</span></span>
<span id="cb57-303"><a href="#cb57-303" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="at">pattern =</span> <span class="st">"GSM.*._|.txt.gz"</span>, <span class="at">replacement =</span> <span class="st">""</span>, <span class="at">x =</span> files)</span>
<span id="cb57-304"><a href="#cb57-304" aria-hidden="true" tabindex="-1"></a>samples</span>
<span id="cb57-305"><a href="#cb57-305" aria-hidden="true" tabindex="-1"></a><span class="co"># 试读取其中一个文件</span></span>
<span id="cb57-306"><a href="#cb57-306" aria-hidden="true" tabindex="-1"></a>test <span class="ot">&lt;-</span> <span class="fu">fread</span>(<span class="st">"data/sc_supplementary/GSE155468_RAW/GSM4704931_Con4.txt.gz"</span>, <span class="at">data.table =</span> F)</span>
<span id="cb57-307"><a href="#cb57-307" aria-hidden="true" tabindex="-1"></a>test[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]</span>
<span id="cb57-308"><a href="#cb57-308" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-309"><a href="#cb57-309" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-310"><a href="#cb57-310" aria-hidden="true" tabindex="-1"></a>可以看到，该表达矩阵的问题是行名不是基因名没有作为行名，而是以单独的一列显示，因此需要进行预处理后再用上面类似的方法批量读取和创建Seurat对象：</span>
<span id="cb57-311"><a href="#cb57-311" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-314"><a href="#cb57-314" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-315"><a href="#cb57-315" aria-hidden="true" tabindex="-1"></a>seurat_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(files, <span class="cf">function</span>(file) {</span>
<span id="cb57-316"><a href="#cb57-316" aria-hidden="true" tabindex="-1"></a>  filepath <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"data/sc_supplementary/GSE155468_RAW/"</span>, file)</span>
<span id="cb57-317"><a href="#cb57-317" aria-hidden="true" tabindex="-1"></a>  counts <span class="ot">&lt;-</span> <span class="fu">fread</span>(filepath, <span class="at">data.table =</span> F)</span>
<span id="cb57-318"><a href="#cb57-318" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames</span>(counts) <span class="ot">&lt;-</span> counts[,<span class="dv">1</span>]</span>
<span id="cb57-319"><a href="#cb57-319" aria-hidden="true" tabindex="-1"></a>  counts <span class="ot">&lt;-</span> counts[,<span class="sc">-</span><span class="dv">1</span>]</span>
<span id="cb57-320"><a href="#cb57-320" aria-hidden="true" tabindex="-1"></a>  <span class="fu">CreateSeuratObject</span>(<span class="at">counts =</span> counts,</span>
<span id="cb57-321"><a href="#cb57-321" aria-hidden="true" tabindex="-1"></a>                     <span class="at">project =</span> <span class="fu">gsub</span>(<span class="at">pattern =</span> <span class="st">"GSM.*._|.txt.gz"</span>, </span>
<span id="cb57-322"><a href="#cb57-322" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">replacement =</span> <span class="st">""</span>, </span>
<span id="cb57-323"><a href="#cb57-323" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">x =</span> file),</span>
<span id="cb57-324"><a href="#cb57-324" aria-hidden="true" tabindex="-1"></a>                     <span class="at">min.cells =</span> <span class="dv">5</span>,</span>
<span id="cb57-325"><a href="#cb57-325" aria-hidden="true" tabindex="-1"></a>                     <span class="at">min.features =</span> <span class="dv">300</span>)</span>
<span id="cb57-326"><a href="#cb57-326" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb57-327"><a href="#cb57-327" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-328"><a href="#cb57-328" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-329"><a href="#cb57-329" aria-hidden="true" tabindex="-1"></a>merged_seurat <span class="ot">&lt;-</span> <span class="fu">merge</span>(<span class="at">x =</span> seurat_list[[<span class="dv">1</span>]], </span>
<span id="cb57-330"><a href="#cb57-330" aria-hidden="true" tabindex="-1"></a>                       <span class="at">y =</span> seurat_list[<span class="sc">-</span><span class="dv">1</span>], </span>
<span id="cb57-331"><a href="#cb57-331" aria-hidden="true" tabindex="-1"></a>                       <span class="at">add.cell.ids =</span> samples)</span>
<span id="cb57-332"><a href="#cb57-332" aria-hidden="true" tabindex="-1"></a>merged_seurat</span>
<span id="cb57-333"><a href="#cb57-333" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(merged_seurat)</span>
<span id="cb57-334"><a href="#cb57-334" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-335"><a href="#cb57-335" aria-hidden="true" tabindex="-1"></a><span class="co"># 新增样本列</span></span>
<span id="cb57-336"><a href="#cb57-336" aria-hidden="true" tabindex="-1"></a>merged_seurat<span class="sc">$</span>samples <span class="ot">&lt;-</span> merged_seurat<span class="sc">$</span>orig.ident</span>
<span id="cb57-337"><a href="#cb57-337" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-338"><a href="#cb57-338" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-339"><a href="#cb57-339" aria-hidden="true" tabindex="-1"></a>简单查看一下细胞/基因数：</span>
<span id="cb57-340"><a href="#cb57-340" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-343"><a href="#cb57-343" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-344"><a href="#cb57-344" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 7</span></span>
<span id="cb57-345"><a href="#cb57-345" aria-hidden="true" tabindex="-1"></a><span class="co">#| layout-ncol: 2</span></span>
<span id="cb57-346"><a href="#cb57-346" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualize the number of cell counts per sample</span></span>
<span id="cb57-347"><a href="#cb57-347" aria-hidden="true" tabindex="-1"></a>merged_seurat<span class="sc">@</span>meta.data  <span class="sc">|&gt;</span> </span>
<span id="cb57-348"><a href="#cb57-348" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> samples, <span class="at">fill =</span> samples)) <span class="sc">+</span> </span>
<span id="cb57-349"><a href="#cb57-349" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_bar</span>() <span class="sc">+</span></span>
<span id="cb57-350"><a href="#cb57-350" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb57-351"><a href="#cb57-351" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">45</span>, <span class="at">vjust =</span> <span class="dv">1</span>, <span class="at">hjust =</span> <span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb57-352"><a href="#cb57-352" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>, <span class="at">face =</span> <span class="st">"bold"</span>)) <span class="sc">+</span></span>
<span id="cb57-353"><a href="#cb57-353" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"NCells"</span>)</span>
<span id="cb57-354"><a href="#cb57-354" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-355"><a href="#cb57-355" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualize the number UMIs/transcripts per cell</span></span>
<span id="cb57-356"><a href="#cb57-356" aria-hidden="true" tabindex="-1"></a>merged_seurat<span class="sc">@</span>meta.data <span class="sc">|&gt;</span> </span>
<span id="cb57-357"><a href="#cb57-357" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">color =</span> samples, <span class="at">x =</span> nCount_RNA, <span class="at">fill =</span> samples)) <span class="sc">+</span> </span>
<span id="cb57-358"><a href="#cb57-358" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_density</span>(<span class="at">alpha =</span> <span class="fl">0.2</span>) <span class="sc">+</span> </span>
<span id="cb57-359"><a href="#cb57-359" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_x_log10</span>() <span class="sc">+</span> </span>
<span id="cb57-360"><a href="#cb57-360" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb57-361"><a href="#cb57-361" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ylab</span>(<span class="st">"Cell density"</span>) <span class="sc">+</span></span>
<span id="cb57-362"><a href="#cb57-362" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">500</span>)</span>
<span id="cb57-363"><a href="#cb57-363" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-364"><a href="#cb57-364" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-365"><a href="#cb57-365" aria-hidden="true" tabindex="-1"></a>------------------------------------------------------------------------</span>
<span id="cb57-366"><a href="#cb57-366" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-367"><a href="#cb57-367" aria-hidden="true" tabindex="-1"></a>::: {.callout-note collapse="true" icon="false"}</span>
<span id="cb57-368"><a href="#cb57-368" aria-hidden="true" tabindex="-1"></a><span class="fu">## Session Info</span></span>
<span id="cb57-369"><a href="#cb57-369" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-372"><a href="#cb57-372" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb57-373"><a href="#cb57-373" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb57-374"><a href="#cb57-374" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span>
<span id="cb57-375"><a href="#cb57-375" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb57-376"><a href="#cb57-376" aria-hidden="true" tabindex="-1"></a>:::</span>
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">
      <ul class="footer-items list-unstyled">
<li class="nav-item">
 This book was built with Quarto
  </li>  
</ul>
</div>   
    <div class="nav-footer-center">Copyright 2023, Du Junhong</div>
    <div class="nav-footer-right">
      <ul class="footer-items list-unstyled">
<li class="nav-item compact">
    <a class="nav-link" href="https://github.com/djhcod">
      <i class="bi bi-github" role="img">
</i> 
    </a>
  </li>  
</ul>
</div>
  </div>
</footer>


</body></html>