<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.549">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>R语言学习笔记：从数据清洗到高级统计学和生物信息学 - 读取非标准格式的单细胞数据</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../single_cell/sc_supplementary/universal_marker.html" rel="next">
<link href="../../single_cell/sc_supplementary/sc_supplementary_intro.html" rel="prev">
<link href="../../images/logo.jpg" rel="icon" type="image/jpeg">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark"><script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><script async="" src="https://www.googletagmanager.com/gtag/js?id=G-G2HZEXHNCE"></script><script type="text/javascript">

window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', 'G-G2HZEXHNCE', { 'anonymize_ip': true});
</script><script type="application/json" class="js-hypothesis-config">
{
  "theme": "classic",
  "openSidebar": false,
  "showHighlights": "always",
  "enableExperimentalNewNoteButton": true
}
</script><script async="" src="https://hypothes.is/embed.js"></script><script>
  window.document.addEventListener("DOMContentLoaded", function (_event) {
    document.body.classList.add('hypothesis-enabled');
  });
</script>
<meta property="og:title" content="R语言学习笔记：从数据清洗到高级统计学和生物信息学 - 读取非标准格式的单细胞数据">
<meta property="og:description" content="">
<meta property="og:site_name" content="R语言学习笔记：从数据清洗到高级统计学和生物信息学">
<meta name="twitter:title" content="R语言学习笔记：从数据清洗到高级统计学和生物信息学 - 读取非标准格式的单细胞数据">
<meta name="twitter:description" content="">
<meta name="twitter:card" content="summary">
</head>
<body class="nav-sidebar docked nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="navbar navbar-expand-lg " data-bs-theme="dark"><div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="../../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../../images/logo.jpg" alt="" class="navbar-logo"></a>
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">R语言学习笔记</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
<li class="nav-item">
    <a class="nav-link active" href="../../index.html" aria-current="page"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../r_basic/r_basics.html"> 
<span class="menu-text">R语言基础</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">单细胞数据分析</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-">
<li>
    <a class="dropdown-item" href="../../single_cell/single_cell_intro.html">
 <span class="dropdown-text">单细胞数据分析介绍</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../single_cell/seurat/seurat_intro.html">
 <span class="dropdown-text">Seurat v5官方文档</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../single_cell/scRNA-seq_online/00_intro.html">
 <span class="dropdown-text">scRNA-seq_online学习材料</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../single_cell/sc_supplementary/sc_supplementary_intro.html">
 <span class="dropdown-text">单细胞补充内容</span></a>
  </li>  
    </ul>
</li>
  <li class="nav-item">
    <a class="nav-link" href="../../quarto_foundation/quarto_foundation.html"> 
<span class="menu-text">Quarto基础</span></a>
  </li>  
</ul>
<ul class="navbar-nav navbar-nav-scroll ms-auto">
<li class="nav-item">
    <a class="nav-link" href="https://bookdown.org"> <i class="bi bi-book-fill" role="img">
</i> 
<span class="menu-text">Bookdown</span></a>
  </li>  
</ul>
</div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools tools-wide">
    <a href="https://github.com/djhcod/r-notes" title="" class="quarto-navigation-tool px-1" aria-label=""><i class="bi bi-github"></i></a>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
  <a href="" class="quarto-reader-toggle quarto-navigation-tool px-1" onclick="window.quartoToggleReader(); return false;" title="Toggle reader mode">
  <div class="quarto-reader-toggle-btn">
  <i class="bi"></i>
  </div>
</a>
</div>
      </div> <!-- /container-fluid -->
    </nav><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../single_cell/single_cell_intro.html">单细胞数据分析</a></li><li class="breadcrumb-item"><a href="../../single_cell/sc_supplementary/sc_supplementary_intro.html">单细胞补充内容</a></li><li class="breadcrumb-item"><a href="../../single_cell/sc_supplementary/read_sc_data.html">读取非标准格式的单细胞数据</a></li></ol></nav>
        <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto"><div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">主页</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../r_basic/r_basics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">R语言基础</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/environment_configuration.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Rstudio环境配置</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/data_input_output.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">数据的读取与输出</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/basic_data_function.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">数据处理基本函数</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/dplyr.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">基于dplyr包的数据整理</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/tidyr.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">基于tidyr的长/宽数据转换</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/data_visualization.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">数据可视化</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/character.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">字符的处理</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../r_basic/loop.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">循环</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../single_cell/single_cell_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">单细胞数据分析</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">单细胞测序技术介绍</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/generation_of_count_matrix.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">单细胞RNA测序—从raw data到count matrix</span></a>
  </div>
</li>
          <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../single_cell/seurat/seurat_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Seurat v5官方文档</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth2 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/seurat_command_list.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Seurat常用函数清单</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/seurat_tutorial.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Seurat细胞分群官方教程</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/data_visualization_methods_in_seurat.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Seurat中的数据可视化方法</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/sctransform.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">基于SCTransform的单细胞数据标准化</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/cell_cycle_regression.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">消除细胞周期的影响</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/integration.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">整合（integration）</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/integrative_analysis_in_seurat_v5.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Seurat v5单细胞数据整合分析</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/marker_gene_identification.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">寻找marker gene</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/de_vignette.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">寻找不同样本类型间同一细胞类型内的差异基因</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/seurat/mapping_and_annotating_query_datasets.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">基于参考集的细胞注释</span></a>
  </div>
</li>
      </ul>
</li>
          <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../single_cell/scRNA-seq_online/00_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">scRNA-seq_online学习材料</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth2 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/03_SC_quality_control-setup.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">数据导入与Seurat对象构建</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/04_SC_quality_control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">质控</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/postQC_workflow.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Single-cell RNA-seq Clustering Workflow</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/05_theory_of_PCA.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">PCA原理</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/06_SC_SCT_normalization.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Normalization and regressing out unwanted variation</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/06_integration.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">单细胞数据整合（Integration）</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/07_SC_clustering_cells_SCT.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">细胞聚类（clustering analysis）</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/08_SC_clustering_quality_control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">细胞分群质量评估</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/scRNA-seq_online/09_merged_SC_marker_identification.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">寻找marker基因+细胞注释</span></a>
  </div>
</li>
      </ul>
</li>
          <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../single_cell/sc_supplementary/sc_supplementary_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">单细胞补充内容</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth2 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/sc_supplementary/read_sc_data.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">读取非标准格式的单细胞数据</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/sc_supplementary/universal_marker.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">细胞分群通用marker</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/sc_supplementary/integrated_analysis_multiple_single_cell_datasets_1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">多个单细胞数据集整合分析（上）</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/sc_supplementary/integrated_analysis_multiple_single_cell_datasets_2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">多个单细胞数据集整合分析（下）</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../single_cell/sc_supplementary/DecontX.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">预测和去除单细胞转录组的环境游离RNA污染</span></a>
  </div>
</li>
      </ul>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../../quarto_foundation/quarto_foundation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Quarto基础</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/yaml_settings.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">YAML设置</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/images_settings.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">图片设置</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/cross_references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">交叉引用</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/insert_other_content.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">插入其他内容</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/quarto_shiny.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">在Quarto中嵌入Shiny应用程序</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/quarto_books.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Quarto Books</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../quarto_foundation/github_pages.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">发布到GitHub Pages</span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active" data-toc-expanded="1"><h2 id="toc-title">On this page</h2>
   
  <ul>
<li><a href="#Read_non-standard_10x_files" id="toc-Read_non-standard_10x_files" class="nav-link active" data-scroll-target="#Read_non-standard_10x_files"><span class="header-section-number">1</span> 读取非标准10X格式文件</a></li>
  <li>
<a href="#%E8%AF%BB%E5%8F%96raw_counts%E5%BD%A2%E5%BC%8F%E7%9A%84%E6%95%B0%E6%8D%AE" id="toc-读取raw_counts形式的数据" class="nav-link" data-scroll-target="#%E8%AF%BB%E5%8F%96raw_counts%E5%BD%A2%E5%BC%8F%E7%9A%84%E6%95%B0%E6%8D%AE"><span class="header-section-number">2</span> 读取raw_counts形式的数据</a>
  <ul class="collapse">
<li><a href="#sec-read_raw_counts" id="toc-sec-read_raw_counts" class="nav-link" data-scroll-target="#sec-read_raw_counts"><span class="header-section-number">2.1</span> 读取raw_counts矩阵</a></li>
  <li><a href="#%E6%B7%BB%E5%8A%A0%E6%A0%B7%E6%9C%AC%E4%BF%A1%E6%81%AF" id="toc-添加样本信息" class="nav-link" data-scroll-target="#%E6%B7%BB%E5%8A%A0%E6%A0%B7%E6%9C%AC%E4%BF%A1%E6%81%AF"><span class="header-section-number">2.2</span> 添加样本信息</a></li>
  </ul>
</li>
  <li><a href="#%E8%AF%BB%E5%8F%96txttsvcsv%E7%AD%89%E6%96%87%E6%9C%AC%E6%96%87%E4%BB%B6%E8%A1%A8%E8%BE%BE%E9%87%8F%E7%9F%A9%E9%98%B5%E4%BF%A1%E6%81%AF" id="toc-读取txttsvcsv等文本文件表达量矩阵信息" class="nav-link" data-scroll-target="#%E8%AF%BB%E5%8F%96txttsvcsv%E7%AD%89%E6%96%87%E6%9C%AC%E6%96%87%E4%BB%B6%E8%A1%A8%E8%BE%BE%E9%87%8F%E7%9F%A9%E9%98%B5%E4%BF%A1%E6%81%AF"><span class="header-section-number">3</span> 读取txt/tsv/csv等文本文件表达量矩阵信息</a></li>
  <li>
<a href="#sec-reading_non_standard_large_volume_text" id="toc-sec-reading_non_standard_large_volume_text" class="nav-link" data-scroll-target="#sec-reading_non_standard_large_volume_text"><span class="header-section-number">4</span> 读取非标准大体积文本文件单细胞数据</a>
  <ul class="collapse">
<li><a href="#%E5%8A%A0%E8%BD%BD%E5%8C%85" id="toc-加载包" class="nav-link" data-scroll-target="#%E5%8A%A0%E8%BD%BD%E5%8C%85"><span class="header-section-number">4.1</span> 加载包</a></li>
  <li><a href="#%E6%95%B0%E6%8D%AE%E8%AF%BB%E5%8F%96" id="toc-数据读取" class="nav-link" data-scroll-target="#%E6%95%B0%E6%8D%AE%E8%AF%BB%E5%8F%96"><span class="header-section-number">4.2</span> 数据读取</a></li>
  <li><a href="#%E6%95%B0%E6%8D%AE%E6%95%B4%E7%90%86" id="toc-数据整理" class="nav-link" data-scroll-target="#%E6%95%B0%E6%8D%AE%E6%95%B4%E7%90%86"><span class="header-section-number">4.3</span> 数据整理</a></li>
  <li><a href="#%E6%9E%84%E5%BB%BAseurat%E5%AF%B9%E8%B1%A1" id="toc-构建seurat对象" class="nav-link" data-scroll-target="#%E6%9E%84%E5%BB%BAseurat%E5%AF%B9%E8%B1%A1"><span class="header-section-number">4.4</span> 构建Seurat对象</a></li>
  </ul>
</li>
  <li><a href="#%E8%AF%BB%E5%8F%96smart-seq%E6%B5%8B%E5%BA%8F%E6%95%B0%E6%8D%AE" id="toc-读取smart-seq测序数据" class="nav-link" data-scroll-target="#%E8%AF%BB%E5%8F%96smart-seq%E6%B5%8B%E5%BA%8F%E6%95%B0%E6%8D%AE"><span class="header-section-number">5</span> 读取SMART-seq测序数据</a></li>
  <li><a href="#%E5%8D%95%E7%8B%AC%E6%8F%90%E4%BE%9B%E6%AF%8F%E4%B8%AA%E7%BB%86%E8%83%9E%E6%B5%8B%E5%BA%8F%E7%BB%93%E6%9E%9C%E7%9A%84%E6%95%B0%E6%8D%AE%E7%9A%84%E8%AF%BB%E5%8F%96" id="toc-单独提供每个细胞测序结果的数据的读取" class="nav-link" data-scroll-target="#%E5%8D%95%E7%8B%AC%E6%8F%90%E4%BE%9B%E6%AF%8F%E4%B8%AA%E7%BB%86%E8%83%9E%E6%B5%8B%E5%BA%8F%E7%BB%93%E6%9E%9C%E7%9A%84%E6%95%B0%E6%8D%AE%E7%9A%84%E8%AF%BB%E5%8F%96"><span class="header-section-number">6</span> 单独提供每个细胞测序结果的数据的读取</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul><div class="toc-actions"><ul><li><a href="https://github.com/djhcod/r-notes/blob/main/single_cell/sc_supplementary/read_sc_data.qmd" class="toc-action"><i class="bi bi-github"></i>View source</a></li><li><a href="https://github.com/djhcod/r-notes/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../single_cell/single_cell_intro.html">单细胞数据分析</a></li><li class="breadcrumb-item"><a href="../../single_cell/sc_supplementary/sc_supplementary_intro.html">单细胞补充内容</a></li><li class="breadcrumb-item"><a href="../../single_cell/sc_supplementary/read_sc_data.html">读取非标准格式的单细胞数据</a></li></ol></nav><div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title">读取非标准格式的单细胞数据</h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header><section id="Read_non-standard_10x_files" class="level1" data-number="1"><h1 data-number="1">
<span class="header-section-number">1</span> 读取非标准10X格式文件</h1>
<p>该案例来自文献 <span class="citation" data-cites="xu2022">(<a href="#ref-xu2022" role="doc-biblioref">Xu et al. 2022</a>)</span>。该研究为了系统地研究高级别浆膜上皮卵巢癌（HGSOC）的肿瘤内异质性，采用深层单细胞RNA测序技术(scRNA-seq)<strong>对7例初治HGSOC早期和晚期患者及5例年龄匹配的非恶性卵巢组织标本进行肿瘤分析。</strong>共获得59,324个HGSOC和非恶性卵巢组织的单细胞，其中，33264个细胞（56%）来⾃HGSOC肿瘤患者组织，26060个（44%）来⾃⾮恶性卵巢组织标本。原始数据：<a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE184880">GSE184880</a>。</p>
<p><img src="images/640-01.png" class="img-fluid"></p>
<p>可以看到原始数据以10X格式提供:</p>
<p><img src="images/截屏2024-01-08 11.17.00.png" class="img-fluid" width="465"></p>
<p>但是，<strong>没有按照标准的10X格式的三个文件命名，并且没有按照样本分别建立文件夹</strong>。所以需要对其进行整理：</p>
<ul>
<li><p>首先给每个样本建立文件夹</p></li>
<li><p>然后将对应的10X标准格式的三个文件分别复制到各个样本文件夹内</p></li>
<li><p>最后将文件重命名为10X格式要求的标准形式，即<strong>“barcodes.tsv.gz”、“features.tsv.gz”、“matrix.mtx.gz”</strong></p></li>
</ul>
<div class="cell" data-hash="read_sc_data_cache/html/unnamed-chunk-1_39129bf8b407dd8fb905293b98f9e57b">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://stringr.tidyverse.org">stringr</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-hash="read_sc_data_cache/html/unnamed-chunk-2_e5a051a0afb0f5adbb71a37ea9e65e67">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 给每个样本建立文件夹</span></span>
<span><span class="va">files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="st">"data/sc_supplementary/GSE184880_RAW"</span><span class="op">)</span></span>
<span><span class="va">files</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "GSM5599220_Norm1"                   "GSM5599220_Norm1.barcodes.tsv.gz"  
 [3] "GSM5599220_Norm1.genes.tsv.gz"      "GSM5599220_Norm1.matrix.mtx.gz"    
 [5] "GSM5599221_Norm2"                   "GSM5599221_Norm2.barcodes.tsv.gz"  
 [7] "GSM5599221_Norm2.genes.tsv.gz"      "GSM5599221_Norm2.matrix.mtx.gz"    
 [9] "GSM5599222_Norm3"                   "GSM5599222_Norm3.barcodes.tsv.gz"  
[11] "GSM5599222_Norm3.genes.tsv.gz"      "GSM5599222_Norm3.matrix.mtx.gz"    
[13] "GSM5599223_Norm4"                   "GSM5599223_Norm4.barcodes.tsv.gz"  
[15] "GSM5599223_Norm4.genes.tsv.gz"      "GSM5599223_Norm4.matrix.mtx.gz"    
[17] "GSM5599224_Norm5"                   "GSM5599224_Norm5.barcodes.tsv.gz"  
[19] "GSM5599224_Norm5.genes.tsv.gz"      "GSM5599224_Norm5.matrix.mtx.gz"    
[21] "GSM5599225_Cancer1"                 "GSM5599225_Cancer1.barcodes.tsv.gz"
[23] "GSM5599225_Cancer1.genes.tsv.gz"    "GSM5599225_Cancer1.matrix.mtx.gz"  
[25] "GSM5599226_Cancer2"                 "GSM5599226_Cancer2.barcodes.tsv.gz"
[27] "GSM5599226_Cancer2.genes.tsv.gz"    "GSM5599226_Cancer2.matrix.mtx.gz"  
[29] "GSM5599227_Cancer3"                 "GSM5599227_Cancer3.barcodes.tsv.gz"
[31] "GSM5599227_Cancer3.genes.tsv.gz"    "GSM5599227_Cancer3.matrix.mtx.gz"  
[33] "GSM5599228_Cancer4"                 "GSM5599228_Cancer4.barcodes.tsv.gz"
[35] "GSM5599228_Cancer4.genes.tsv.gz"    "GSM5599228_Cancer4.matrix.mtx.gz"  
[37] "GSM5599229_Cancer5"                 "GSM5599229_Cancer5.barcodes.tsv.gz"
[39] "GSM5599229_Cancer5.genes.tsv.gz"    "GSM5599229_Cancer5.matrix.mtx.gz"  
[41] "GSM5599230_Cancer6"                 "GSM5599230_Cancer6.barcodes.tsv.gz"
[43] "GSM5599230_Cancer6.genes.tsv.gz"    "GSM5599230_Cancer6.matrix.mtx.gz"  
[45] "GSM5599231_Cancer7"                 "GSM5599231_Cancer7.barcodes.tsv.gz"
[47] "GSM5599231_Cancer7.genes.tsv.gz"    "GSM5599231_Cancer7.matrix.mtx.gz"  </code></pre>
</div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">dirnames</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">".barcodes.tsv.gz|.genes.tsv.gz|.matrix.mtx.gz"</span>, </span>
<span>                 replacement <span class="op">=</span> <span class="st">""</span>, </span>
<span>                 x <span class="op">=</span> <span class="va">files</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>  </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"data/sc_supplementary/GSE184880_RAW/"</span>, <span class="va">.</span><span class="op">)</span></span>
<span><span class="va">dirnames</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "data/sc_supplementary/GSE184880_RAW/GSM5599220_Norm1"  
 [2] "data/sc_supplementary/GSE184880_RAW/GSM5599221_Norm2"  
 [3] "data/sc_supplementary/GSE184880_RAW/GSM5599222_Norm3"  
 [4] "data/sc_supplementary/GSE184880_RAW/GSM5599223_Norm4"  
 [5] "data/sc_supplementary/GSE184880_RAW/GSM5599224_Norm5"  
 [6] "data/sc_supplementary/GSE184880_RAW/GSM5599225_Cancer1"
 [7] "data/sc_supplementary/GSE184880_RAW/GSM5599226_Cancer2"
 [8] "data/sc_supplementary/GSE184880_RAW/GSM5599227_Cancer3"
 [9] "data/sc_supplementary/GSE184880_RAW/GSM5599228_Cancer4"
[10] "data/sc_supplementary/GSE184880_RAW/GSM5599229_Cancer5"
[11] "data/sc_supplementary/GSE184880_RAW/GSM5599230_Cancer6"
[12] "data/sc_supplementary/GSE184880_RAW/GSM5599231_Cancer7"</code></pre>
</div>
</div>
<div class="cell" data-hash="read_sc_data_cache/html/unnamed-chunk-3_e77e40c1ce7cb3876b69097005c719be">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">dirnames</span>, <span class="va">dir.create</span>, showWarnings <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-hash="read_sc_data_cache/html/unnamed-chunk-4_35aaa4f9fef8c2fa0c493d136d7b0bd0">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">files</span>, <span class="kw">function</span><span class="op">(</span><span class="va">files</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="co"># 复制相应的10X文件到各个样本文件夹内</span></span>
<span>  <span class="va">new_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">".barcodes.tsv.gz|.genes.tsv.gz|.matrix.mtx.gz"</span>, </span>
<span>                  replacement <span class="op">=</span> <span class="st">""</span>, </span>
<span>                  x <span class="op">=</span> <span class="va">files</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span>  </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"data/sc_supplementary/GSE184880_RAW/"</span>, <span class="va">.</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/files.html">file.copy</a></span><span class="op">(</span>from <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"data/sc_supplementary/GSE184880_RAW/"</span>, <span class="va">files</span><span class="op">)</span>, </span>
<span>            to <span class="op">=</span> <span class="va">new_dir</span><span class="op">)</span></span>
<span>  <span class="co"># 重命名3个10X文件</span></span>
<span>  <span class="va">new_filename</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"barcodes"</span>, x <span class="op">=</span> <span class="va">files</span><span class="op">)</span>, <span class="st">"barcodes.tsv.gz"</span>, </span>
<span>                         <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"genes"</span>, x <span class="op">=</span> <span class="va">files</span><span class="op">)</span>, <span class="st">"features.tsv.gz"</span>, </span>
<span>                         <span class="st">"matrix.mtx.gz"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/files.html">file.rename</a></span><span class="op">(</span>from <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="va">new_dir</span>, <span class="va">files</span>, sep  <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span>,</span>
<span>              to <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="va">new_dir</span>, <span class="va">new_filename</span>, sep <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">}</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
用setwd简化对文件路径的定义
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>由于在渲染本文档时，knitr会将工作目录恢复为Rproject根目录，所以不能使用getwd指定工作目录。在实际应用中，可以设置工作目录到目标数据文件夹下以简化对文件路径的定义：</p>
<div class="cell" data-hash="read_sc_data_cache/html/unnamed-chunk-5_c015b490890eef7bec9871b6accefe00">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">root_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/getwd.html">getwd</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html">setwd</a></span><span class="op">(</span><span class="st">"data/sc_supplementary/GSE184880_RAW"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-hash="read_sc_data_cache/html/unnamed-chunk-6_9984d215989ba9e5935c412354671db2">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 给每个样本建立文件夹</span></span>
<span><span class="va">files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">files</span></span>
<span><span class="va">dirnames</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">".barcodes.tsv.gz|.genes.tsv.gz|.matrix.mtx.gz"</span>, </span>
<span>                 replacement <span class="op">=</span> <span class="st">""</span>, </span>
<span>                 x <span class="op">=</span> <span class="va">files</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">dirnames</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">dirnames</span>, <span class="va">dir.create</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">files</span>, <span class="kw">function</span><span class="op">(</span><span class="va">files</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="co"># 复制相应的10X文件到各个样本文件夹内</span></span>
<span>  <span class="va">new_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">".barcodes.tsv.gz|.genes.tsv.gz|.matrix.mtx.gz"</span>, </span>
<span>                      replacement <span class="op">=</span> <span class="st">""</span>, </span>
<span>                      x <span class="op">=</span> <span class="va">files</span><span class="op">)</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/files.html">file.copy</a></span><span class="op">(</span>from <span class="op">=</span> <span class="va">files</span>, </span>
<span>            to <span class="op">=</span> <span class="va">new_dir</span><span class="op">)</span></span>
<span>  <span class="co"># 重命名3个10X文件</span></span>
<span>  <span class="va">new_filename</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"barcodes"</span>, x <span class="op">=</span> <span class="va">files</span><span class="op">)</span>, <span class="st">"barcodes.tsv.gz"</span>, </span>
<span>                         <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"genes"</span>, x <span class="op">=</span> <span class="va">files</span><span class="op">)</span>, <span class="st">"features.tsv.gz"</span>, </span>
<span>                         <span class="st">"matrix.mtx.gz"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/files.html">file.rename</a></span><span class="op">(</span>from <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="va">new_dir</span>, <span class="va">files</span>, sep  <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span>,</span>
<span>              to <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="va">new_dir</span>, <span class="va">new_filename</span>, sep <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">}</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-hash="read_sc_data_cache/html/unnamed-chunk-7_37831b1e1c02eabbea2c60c853841bde">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">seurat_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">dirnames</span>, <span class="kw">function</span><span class="op">(</span><span class="va">dirnames</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">dirnames</span><span class="op">)</span></span>
<span>  <span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/CreateSeuratObject.html">CreateSeuratObject</a></span><span class="op">(</span>counts <span class="op">=</span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/Read10X.html">Read10X</a></span><span class="op">(</span><span class="va">dirnames</span><span class="op">)</span>,</span>
<span>                            project <span class="op">=</span>  <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_split.html">str_split</a></span><span class="op">(</span>string <span class="op">=</span> <span class="va">dirnames</span>, </span>
<span>                                                 pattern <span class="op">=</span> <span class="st">"_"</span>, </span>
<span>                                                 simplify <span class="op">=</span> <span class="cn">T</span><span class="op">)</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span>,</span>
<span>                            min.cells <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                            min.features <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="va">merged_seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html">merge</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">seurat_list</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,</span>
<span>                       y <span class="op">=</span> <span class="va">seurat_list</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>                       add.cell.ids <span class="op">=</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_split.html">str_split</a></span><span class="op">(</span>string <span class="op">=</span> <span class="va">dirnames</span>, </span>
<span>                                                pattern <span class="op">=</span> <span class="st">"_"</span>, </span>
<span>                                                simplify <span class="op">=</span> <span class="cn">T</span><span class="op">)</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="va">merged_seurat</span></span>
<span><span class="va">merged_seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/SplitLayers.html">JoinLayers</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">)</span></span>
<span><span class="va">merged_seurat</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">tail</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">$</span><span class="va">orig.ident</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 将样本信息添加到meta.data的新的一列“sample”中:</span></span>
<span><span class="va">merged_seurat</span><span class="op">$</span><span class="va">sample</span> <span class="op">&lt;-</span> <span class="va">merged_seurat</span><span class="op">$</span><span class="va">orig.ident</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>后续如有需要可以随时将工作目录恢复为Rproject根目录：</p>
<div class="cell" data-hash="read_sc_data_cache/html/unnamed-chunk-8_1f82070a950845cff9476979a7bd515f">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html">setwd</a></span><span class="op">(</span><span class="va">root_dir</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<p>接下来，就可以进行常规的Seurat对象构建：</p>
<div class="cell" data-hash="read_sc_data_cache/html/unnamed-chunk-9_2276ec70dba789e7e29f337262db2ed9">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">seurat_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">dirnames</span>, <span class="kw">function</span><span class="op">(</span><span class="va">dirnames</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">dirnames</span><span class="op">)</span></span>
<span>  <span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/CreateSeuratObject.html">CreateSeuratObject</a></span><span class="op">(</span>counts <span class="op">=</span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/Read10X.html">Read10X</a></span><span class="op">(</span><span class="va">dirnames</span><span class="op">)</span>,</span>
<span>                            project <span class="op">=</span>  <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_split.html">str_split</a></span><span class="op">(</span>string <span class="op">=</span> <span class="va">dirnames</span>, </span>
<span>                                                 pattern <span class="op">=</span> <span class="st">"_"</span>, </span>
<span>                                                 simplify <span class="op">=</span> <span class="cn">T</span><span class="op">)</span><span class="op">[</span>,<span class="fl">4</span><span class="op">]</span>,</span>
<span>                            min.cells <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                            min.features <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "data/sc_supplementary/GSE184880_RAW/GSM5599220_Norm1"
[1] "data/sc_supplementary/GSE184880_RAW/GSM5599221_Norm2"
[1] "data/sc_supplementary/GSE184880_RAW/GSM5599222_Norm3"
[1] "data/sc_supplementary/GSE184880_RAW/GSM5599223_Norm4"
[1] "data/sc_supplementary/GSE184880_RAW/GSM5599224_Norm5"
[1] "data/sc_supplementary/GSE184880_RAW/GSM5599225_Cancer1"
[1] "data/sc_supplementary/GSE184880_RAW/GSM5599226_Cancer2"
[1] "data/sc_supplementary/GSE184880_RAW/GSM5599227_Cancer3"
[1] "data/sc_supplementary/GSE184880_RAW/GSM5599228_Cancer4"
[1] "data/sc_supplementary/GSE184880_RAW/GSM5599229_Cancer5"
[1] "data/sc_supplementary/GSE184880_RAW/GSM5599230_Cancer6"
[1] "data/sc_supplementary/GSE184880_RAW/GSM5599231_Cancer7"</code></pre>
</div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">merged_seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html">merge</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">seurat_list</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,</span>
<span>                       y <span class="op">=</span> <span class="va">seurat_list</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>                       add.cell.ids <span class="op">=</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_split.html">str_split</a></span><span class="op">(</span>string <span class="op">=</span> <span class="va">dirnames</span>, </span>
<span>                                                pattern <span class="op">=</span> <span class="st">"_"</span>, </span>
<span>                                                simplify <span class="op">=</span> <span class="cn">T</span><span class="op">)</span><span class="op">[</span>,<span class="fl">4</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="va">merged_seurat</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
21454 features across 60886 samples within 1 assay 
Active assay: RNA (21454 features, 0 variable features)
 12 layers present: counts.Norm1, counts.Norm2, counts.Norm3, counts.Norm4, counts.Norm5, counts.Cancer1, counts.Cancer2, counts.Cancer3, counts.Cancer4, counts.Cancer5, counts.Cancer6, counts.Cancer7</code></pre>
</div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">merged_seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/SplitLayers.html">JoinLayers</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">)</span></span>
<span><span class="va">merged_seurat</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
21454 features across 60886 samples within 1 assay 
Active assay: RNA (21454 features, 0 variable features)
 1 layer present: counts</code></pre>
</div>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                         orig.ident nCount_RNA nFeature_RNA
Norm1_AAACCCAAGGACAGCT-1      Norm1       8191         2218
Norm1_AAACCCAAGGTATCTC-1      Norm1       8720         2396
Norm1_AAACCCACAAGTGCAG-1      Norm1       8554         2518
Norm1_AAACCCACACCCTCTA-1      Norm1       7650         2697
Norm1_AAACCCACACTCCGGA-1      Norm1       9231         2370
Norm1_AAACCCACAGAGGTTG-1      Norm1       5701         1619
Norm1_AAACCCAGTACTTCCC-1      Norm1      10002         2421
Norm1_AAACCCATCTACCACC-1      Norm1       2934         1135
Norm1_AAACGAAAGTAGGCCA-1      Norm1       6955         2198
Norm1_AAACGAAAGTCACACT-1      Norm1       7742         1891</code></pre>
</div>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">tail</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                           orig.ident nCount_RNA nFeature_RNA
Cancer7_TTTGGAGTCGTGAGAG-1    Cancer7       3568         1525
Cancer7_TTTGGTTAGGATCATA-1    Cancer7       3720         1402
Cancer7_TTTGGTTAGTTGGCTT-1    Cancer7       2591         1112
Cancer7_TTTGGTTGTCACCCTT-1    Cancer7      27107         5367
Cancer7_TTTGGTTGTTCCTACC-1    Cancer7      55055         7556
Cancer7_TTTGGTTGTTTCGCTC-1    Cancer7      66262         7369
Cancer7_TTTGGTTTCCACAAGT-1    Cancer7       2342         1143
Cancer7_TTTGTTGCATCCTAAG-1    Cancer7       3498         1440
Cancer7_TTTGTTGGTGACCTGC-1    Cancer7      74316         8182
Cancer7_TTTGTTGGTGATCATC-1    Cancer7       4245          508</code></pre>
</div>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">$</span><span class="va">orig.ident</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "Norm1"   "Norm2"   "Norm3"   "Norm4"   "Norm5"   "Cancer1" "Cancer2"
 [8] "Cancer3" "Cancer4" "Cancer5" "Cancer6" "Cancer7"</code></pre>
</div>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 将样本信息添加到meta.data的新的一列“sample”中:</span></span>
<span><span class="va">merged_seurat</span><span class="op">$</span><span class="va">sample</span> <span class="op">&lt;-</span> <span class="va">merged_seurat</span><span class="op">$</span><span class="va">orig.ident</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr></section><section id="读取raw_counts形式的数据" class="level1" data-number="2"><h1 data-number="2">
<span class="header-section-number">2</span> 读取raw_counts形式的数据</h1>
<p>案例来自<span class="citation" data-cites="Bill2023">(<a href="#ref-Bill2023" role="doc-biblioref">Bill et al. 2023</a>)</span>。样本来自头颈部鳞状细胞癌（HNSCC）患者，总共52例标本，包括24例原发肿瘤、16例局部复发和12例多器官远处转移。原发部位包括口腔、喉/下咽、口咽和鼻咽。其中15份样本HPV阳性。原始数据：<a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE234933">GSE234933</a>。</p>
<p>可以看到该数据集没有提供标准的10X格式的3个文件。</p>
<p><img src="images/截屏2024-01-03 20.56.56.png" class="img-fluid"></p>
<p>其中的“raw_counts”文件解压后是52个RDS文件，对应了每个样本的单细胞测序数据，是一个已经包含了表达量、基因名和细胞barcode的矩阵，所以可以通过批量读取这些文件建立Seurat对象。</p>
<p><img src="images/截屏2024-01-09 11.10.35.png" class="img-fluid" width="492"></p>
<p>“sample_annotation.txt”文件包含了对每个样本的详细临床病理资料的注释，因此可以将其添加到meta.data中。</p>
<p><img src="images/截屏2024-01-09 11.11.45.png" class="img-fluid"></p>
<p>“cell_annotation.txt”文件则是原作者对细胞类型的注释，可以不添加。</p>
<p><img src="images/截屏2024-01-09 11.12.14.png" class="img-fluid"></p>
<section id="sec-read_raw_counts" class="level2" data-number="2.1"><h2 data-number="2.1" class="anchored" data-anchor-id="sec-read_raw_counts">
<span class="header-section-number">2.1</span> 读取raw_counts矩阵</h2>
<div class="cell" data-hash="read_sc_data_cache/html/unnamed-chunk-10_4aedaea4c2d8a1fdb858f1c1988faafd">
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/rm.html">rm</a></span><span class="op">(</span>list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ls.html">ls</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://stringr.tidyverse.org">stringr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tibble.tidyverse.org/">tibble</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-hash="read_sc_data_cache/html/unnamed-chunk-11_d786fed8f4578b526d59e9aab408df15">
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 获取所有rds文件的列表</span></span>
<span><span class="va">file_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="st">"data/sc_supplementary/GSE234933/GSE234933_MGH_HNSCC_gex_raw_counts"</span>, pattern <span class="op">=</span> <span class="st">".rds"</span><span class="op">)</span></span>
<span><span class="va">file_list</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "HN1_gex_raw_counts.rds"       "HN13_gex_raw_counts.rds"     
 [3] "HN14_gex_raw_counts.rds"      "HN17_gex_raw_counts.rds"     
 [5] "HN2_gex_raw_counts.rds"       "HN20_gex_raw_counts.rds"     
 [7] "HN21_gex_raw_counts.rds"      "HN22_gex_raw_counts.rds"     
 [9] "HN23_gex_raw_counts.rds"      "HN25_gex_raw_counts.rds"     
[11] "HN26_gex_raw_counts.rds"      "HN27_gex_raw_counts.rds"     
[13] "HN28_gex_raw_counts.rds"      "HN29_gex_raw_counts.rds"     
[15] "HN30_gex_raw_counts.rds"      "HN31TS_gex_raw_counts.rds"   
[17] "HN32T_gex_raw_counts.rds"     "HN33_gex_raw_counts.rds"     
[19] "HN34_gex_raw_counts.rds"      "HN35_gex_raw_counts.rds"     
[21] "HN37_gex_raw_counts.rds"      "HN38_gex_raw_counts.rds"     
[23] "HN39_gex_raw_counts.rds"      "HN40-3-05_gex_raw_counts.rds"
[25] "HN42_gex_raw_counts.rds"      "HN43_gex_raw_counts.rds"     
[27] "HN45_gex_raw_counts.rds"      "HN46_gex_raw_counts.rds"     
[29] "HN49_gex_raw_counts.rds"      "HN50_gex_raw_counts.rds"     
[31] "HN52_gex_raw_counts.rds"      "HN55_gex_raw_counts.rds"     
[33] "HN57-1_gex_raw_counts.rds"    "HN58_gex_raw_counts.rds"     
[35] "HN59_gex_raw_counts.rds"      "HN60_gex_raw_counts.rds"     
[37] "HN61_gex_raw_counts.rds"      "HN63_gex_raw_counts.rds"     
[39] "HN64_gex_raw_counts.rds"      "HN66_gex_raw_counts.rds"     
[41] "HN67_gex_raw_counts.rds"      "HN68_gex_raw_counts.rds"     
[43] "HN7_gex_raw_counts.rds"       "HN70_gex_raw_counts.rds"     
[45] "HN71_gex_raw_counts.rds"      "HN72_gex_raw_counts.rds"     
[47] "HN73_gex_raw_counts.rds"      "HN74_gex_raw_counts.rds"     
[49] "HN75_gex_raw_counts.rds"      "HN76_gex_raw_counts.rds"     
[51] "HN77_gex_raw_counts.rds"      "HN8_gex_raw_counts.rds"      </code></pre>
</div>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 循环读取每个rds文件的数据并创建Seurat对象</span></span>
<span><span class="va">seurat_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">file_list</span>, <span class="kw">function</span><span class="op">(</span><span class="va">file</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="co"># 拼接文件路径</span></span>
<span>  <span class="va">data.path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"data/sc_supplementary/GSE234933/GSE234933_MGH_HNSCC_gex_raw_counts/"</span>, <span class="va">file</span><span class="op">)</span></span>
<span>  <span class="co"># 创建Seurat对象</span></span>
<span>  <span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/CreateSeuratObject.html">CreateSeuratObject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="va">data.path</span><span class="op">)</span>,</span>
<span>                                   min.features <span class="op">=</span> <span class="fl">200</span>,</span>
<span>                                   min.cells <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="va">seurat_obj</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 合并Seurat对象</span></span>
<span><span class="va">seurat_combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html">merge</a></span><span class="op">(</span><span class="va">seurat_list</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,</span>
<span>                         y <span class="op">=</span> <span class="va">seurat_list</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>                         add.cell.ids <span class="op">=</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_split.html">str_split</a></span><span class="op">(</span>string <span class="op">=</span> <span class="va">file_list</span>, </span>
<span>                                                  pattern <span class="op">=</span> <span class="st">"_"</span>, </span>
<span>                                                  simplify <span class="op">=</span> <span class="cn">T</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">seurat_combined</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
26734 features across 330385 samples within 1 assay 
Active assay: RNA (26734 features, 0 variable features)
 52 layers present: counts.1, counts.2, counts.3, counts.4, counts.5, counts.6, counts.7, counts.8, counts.9, counts.10, counts.11, counts.12, counts.13, counts.14, counts.15, counts.16, counts.17, counts.18, counts.19, counts.20, counts.21, counts.22, counts.23, counts.24, counts.25, counts.26, counts.27, counts.28, counts.29, counts.30, counts.31, counts.32, counts.33, counts.34, counts.35, counts.36, counts.37, counts.38, counts.39, counts.40, counts.41, counts.42, counts.43, counts.44, counts.45, counts.46, counts.47, counts.48, counts.49, counts.50, counts.51, counts.52</code></pre>
</div>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">seurat_combined</span><span class="op">)</span> <span class="co"># 可以看到这个案例实际上不需要添加cell.ids</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                           orig.ident nCount_RNA nFeature_RNA
HN1_HN1_AAACCTGAGTATTGGA-1        HN1        948          574
HN1_HN1_AAACCTGAGTTGCAGG-1        HN1       6745          257
HN1_HN1_AAACCTGCAATAGCAA-1        HN1        677          320
HN1_HN1_AAACCTGCATTTCACT-1        HN1       4540         1129
HN1_HN1_AAACCTGTCGGCGGTT-1        HN1        909          444
HN1_HN1_AAACCTGTCTCGTTTA-1        HN1        633          352
HN1_HN1_AAACGGGAGCAGCCTC-1        HN1       1262          633
HN1_HN1_AAACGGGAGCCCAGCT-1        HN1        513          275
HN1_HN1_AAACGGGAGTACGCGA-1        HN1       1152          416
HN1_HN1_AAACGGGCACATTTCT-1        HN1        426          249</code></pre>
</div>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">seurat_combined</span><span class="op">$</span><span class="va">orig.ident</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "HN1"       "HN13"      "HN14"      "HN17"      "HN2"       "HN20"     
 [7] "HN21"      "HN22"      "HN23"      "HN25"      "HN26"      "HN27"     
[13] "HN28"      "HN29"      "HN30"      "HN31TS"    "HN32T"     "HN33"     
[19] "HN34"      "HN35"      "HN37"      "HN38"      "HN39"      "HN40-3-05"
[25] "HN42"      "HN43"      "HN45"      "HN46"      "HN49"      "HN50"     
[31] "HN52"      "HN55"      "HN57-1"    "HN58"      "HN59"      "HN60"     
[37] "HN61"      "HN63"      "HN64"      "HN66"      "HN67"      "HN68"     
[43] "HN7"       "HN70"      "HN71"      "HN72"      "HN73"      "HN74"     
[49] "HN75"      "HN76"      "HN77"      "HN8"      </code></pre>
</div>
</div>
</section><section id="添加样本信息" class="level2" data-number="2.2"><h2 data-number="2.2" class="anchored" data-anchor-id="添加样本信息">
<span class="header-section-number">2.2</span> 添加样本信息</h2>
<div class="cell" data-hash="read_sc_data_cache/html/unnamed-chunk-12_b2f8b2ea2a01a9021c2c16a1d026e987">
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">sample_anno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span><span class="op">(</span><span class="st">"data/sc_supplementary/GSE234933/GSE234933_MGH_HNSCC_sample_annotation.txt"</span>, </span>
<span>                          sep <span class="op">=</span> <span class="st">"\t"</span>, </span>
<span>                          header <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">sample_anno</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Sample    Sex Age HPV.Status Smoking.history Original.anatomic.site
1    HN1   Male  37   Negative             Yes            Oral cavity
2    HN2   Male  55   Negative             Yes            Oral cavity
3    HN7   Male  90   Negative              No            Oral cavity
4    HN8   Male  89   Negative             Yes     Larynx/Hypopharynx
5   HN13 Female  64   Negative             Yes     Larynx/Hypopharynx
6   HN14 Female  53   Negative             Yes             Oropharynx
  Anatomic.location.of.scRNA.seq.specimen
1                                 Primary
2               Distant metastasis (Lung)
3                                 Primary
4            Distant metastasis (sternum)
5               Distant metastasis (skin)
6             Distant metastasis (pleura)</code></pre>
</div>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">sample_anno</span><span class="op">$</span><span class="va">Sample</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "HN1"       "HN2"       "HN7"       "HN8"       "HN13"      "HN14"     
 [7] "HN17"      "HN20"      "HN21"      "HN22"      "HN23"      "HN25"     
[13] "HN26"      "HN27"      "HN28"      "HN29"      "HN30"      "HN31TS"   
[19] "HN32T"     "HN33"      "HN34"      "HN35"      "HN37"      "HN38"     
[25] "HN39"      "HN40-3-05" "HN42"      "HN43"      "HN45"      "HN46"     
[31] "HN49"      "HN50"      "HN52"      "HN55"      "HN57-1"    "HN58"     
[37] "HN59"      "HN60"      "HN61"      "HN63"      "HN64"      "HN66"     
[43] "HN67"      "HN68"      "HN70"      "HN71"      "HN72"      "HN73"     
[49] "HN74"      "HN75"      "HN76"      "HN77"     </code></pre>
</div>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">old_meta.data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames_to_column</a></span><span class="op">(</span><span class="va">seurat_combined</span><span class="op">@</span><span class="va">meta.data</span>, var <span class="op">=</span> <span class="st">"cell_id"</span><span class="op">)</span></span>
<span><span class="va">new_meta.data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html">merge</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">old_meta.data</span>,</span>
<span>                       y <span class="op">=</span> <span class="va">sample_anno</span>,</span>
<span>                       by.x <span class="op">=</span> <span class="st">"orig.ident"</span>,</span>
<span>                       by.y <span class="op">=</span> <span class="st">"Sample"</span>,</span>
<span>                       all.x <span class="op">=</span> <span class="cn">T</span><span class="op">)</span> </span>
<span><span class="va">new_meta.data</span><span class="op">$</span><span class="va">sample</span> <span class="op">&lt;-</span> <span class="va">new_meta.data</span><span class="op">$</span><span class="va">orig.ident</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">new_meta.data</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  orig.ident                    cell_id nCount_RNA nFeature_RNA  Sex Age
1        HN1 HN1_HN1_AAACCTGAGTATTGGA-1        948          574 Male  37
2        HN1 HN1_HN1_AAACCTGAGTTGCAGG-1       6745          257 Male  37
3        HN1 HN1_HN1_AAACCTGCAATAGCAA-1        677          320 Male  37
4        HN1 HN1_HN1_AAACCTGCATTTCACT-1       4540         1129 Male  37
5        HN1 HN1_HN1_AAACCTGTCGGCGGTT-1        909          444 Male  37
6        HN1 HN1_HN1_AAACCTGTCTCGTTTA-1        633          352 Male  37
  HPV.Status Smoking.history Original.anatomic.site
1   Negative             Yes            Oral cavity
2   Negative             Yes            Oral cavity
3   Negative             Yes            Oral cavity
4   Negative             Yes            Oral cavity
5   Negative             Yes            Oral cavity
6   Negative             Yes            Oral cavity
  Anatomic.location.of.scRNA.seq.specimen sample
1                                 Primary    HN1
2                                 Primary    HN1
3                                 Primary    HN1
4                                 Primary    HN1
5                                 Primary    HN1
6                                 Primary    HN1</code></pre>
</div>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">new_meta.data</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">new_meta.data</span><span class="op">$</span><span class="va">cell_id</span></span>
<span><span class="va">new_meta.data</span><span class="op">$</span><span class="va">cell_id</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span><span class="va">new_meta.data</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">6</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">new_meta.data</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">5</span>, <span class="fl">7</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span><span class="op">]</span>, <span class="va">as.factor</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">new_meta.data</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  orig.ident          nCount_RNA      nFeature_RNA      Sex        
 Length:330385      Min.   :   215   Min.   : 184   Female: 65737  
 Class :character   1st Qu.:   897   1st Qu.: 471   Male  :264648  
 Mode  :character   Median :  1886   Median : 845                  
                    Mean   :  3439   Mean   :1107                  
                    3rd Qu.:  3537   3rd Qu.:1348                  
                    Max.   :126539   Max.   :8839                  
                                                                   
      Age          HPV.Status     Smoking.history        Original.anatomic.site
 Min.   :31.00   64     : 29808   No :177238      Larynx/Hypopharynx: 68470    
 1st Qu.:56.00   76     : 25982   Yes:153147      Nasopharynx       : 30573    
 Median :64.00   70     : 23193                   Oral cavity       :107833    
 Mean   :63.42   44     : 21985                   Oropharynx        :115022    
 3rd Qu.:71.00   67     : 21053                   Unknown           :  8487    
 Max.   :91.00   55     : 19302                                                
                 (Other):189062                                                
                 Anatomic.location.of.scRNA.seq.specimen     sample      
 Primary                             :190573             HN63   : 25501  
 Locoregional recurrence             : 59176             HN76   : 15798  
 Distant metastasis (Lung)           : 34541             HN68   : 14851  
 Distant metastasis (Liver)          : 23083             HN77   : 13892  
 Unknown primary of the head and neck:  8487             HN42   : 12062  
 Distant metastasis (skin)           :  7073             HN74   : 11863  
 (Other)                             :  7452             (Other):236418  </code></pre>
</div>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">seurat_combined</span><span class="op">@</span><span class="va">meta.data</span> <span class="op">&lt;-</span> <span class="va">new_meta.data</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section></section><section id="读取txttsvcsv等文本文件表达量矩阵信息" class="level1" data-number="3"><h1 data-number="3">
<span class="header-section-number">3</span> 读取txt/tsv/csv等文本文件表达量矩阵信息</h1>
<p>以2020的文章：《Single-Cell Transcriptome Analysis Reveals Dynamic Cell Populations and Differential Gene Expression Patterns in Control and Aneurysmal Human Aortic Tissue》<span class="citation" data-cites="Li2020">(<a href="#ref-Li2020" role="doc-biblioref">Li et al. 2020</a>)</span> 举例说明。数据集链接：<a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE155468"><strong>GSE155468</strong></a><strong>。</strong></p>
<p><img src="images/截屏2024-01-10 11.02.14.png" class="img-fluid" width="565"></p>
<p>该案例提供了11个人体主动脉组织样品的单细胞转录组测序数据，其中8名患者为升胸主动脉瘤（ATAA）and 3名患者作为对照。每个样品都是一个独立的txt文本文件，蕴藏着其表达量矩阵信息。</p>
<p>首先我们读取其中一个文件看一下其形式：</p>
<div class="cell" data-hash="read_sc_data_cache/html/unnamed-chunk-13_e7d3fb864e3215f20d9c0d4900b1418a">
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/rm.html">rm</a></span><span class="op">(</span>list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ls.html">ls</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com">data.table</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-hash="read_sc_data_cache/html/unnamed-chunk-14_299eedbcad06d5daca3874247374bf99">
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 列出所有的文件名</span></span>
<span><span class="va">files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="st">"data/sc_supplementary/GSE155468_RAW"</span><span class="op">)</span></span>
<span><span class="va">files</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "GSM4704931_Con4.txt.gz" "GSM4704932_Con6.txt.gz" "GSM4704933_Con9.txt.gz"
 [4] "GSM4704934_TAA1.txt.gz" "GSM4704935_TAA2.txt.gz" "GSM4704936_TAA3.txt.gz"
 [7] "GSM4704937_TAA4.txt.gz" "GSM4704938_TAA5.txt.gz" "GSM4704939_TAA6.txt.gz"
[10] "GSM4704940_TAA7.txt.gz" "GSM4704941_TAA8.txt.gz"</code></pre>
</div>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 简化文件名用作样本名</span></span>
<span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"GSM.*._|.txt.gz"</span>, replacement <span class="op">=</span> <span class="st">""</span>, x <span class="op">=</span> <span class="va">files</span><span class="op">)</span></span>
<span><span class="va">samples</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "Con4" "Con6" "Con9" "TAA1" "TAA2" "TAA3" "TAA4" "TAA5" "TAA6" "TAA7"
[11] "TAA8"</code></pre>
</div>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 试读取其中一个文件</span></span>
<span><span class="va">test</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html">fread</a></span><span class="op">(</span><span class="st">"data/sc_supplementary/GSE155468_RAW/GSM4704931_Con4.txt.gz"</span>, data.table <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span><span class="va">test</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          V1 AAACCCAAGCGGACAT.1 AAACCCAAGCGTCAAG.1
1 AL627309.1                  0                  0
2 AL669831.5                  0                  0
3     FAM87B                  0                  0</code></pre>
</div>
</div>
<p>可以看到，该表达矩阵的问题是行名不是基因名没有作为行名，而是以单独的一列显示，因此需要进行预处理后再用上面类似的方法批量读取和创建Seurat对象：</p>
<div class="cell" data-hash="read_sc_data_cache/html/unnamed-chunk-15_7406ee77160bec6e8cd07f047423b86f">
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">seurat_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">files</span>, <span class="kw">function</span><span class="op">(</span><span class="va">file</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">filepath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"data/sc_supplementary/GSE155468_RAW/"</span>, <span class="va">file</span><span class="op">)</span></span>
<span>  <span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html">fread</a></span><span class="op">(</span><span class="va">filepath</span>, data.table <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">counts</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span></span>
<span>  <span class="va">counts</span> <span class="op">&lt;-</span> <span class="va">counts</span><span class="op">[</span>,<span class="op">-</span><span class="fl">1</span><span class="op">]</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/CreateSeuratObject.html">CreateSeuratObject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">counts</span>,</span>
<span>                     project <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"GSM.*._|.txt.gz"</span>, </span>
<span>                                    replacement <span class="op">=</span> <span class="st">""</span>, </span>
<span>                                    x <span class="op">=</span> <span class="va">file</span><span class="op">)</span>,</span>
<span>                     min.cells <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                     min.features <span class="op">=</span> <span class="fl">300</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">merged_seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html">merge</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">seurat_list</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, </span>
<span>                       y <span class="op">=</span> <span class="va">seurat_list</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span>, </span>
<span>                       add.cell.ids <span class="op">=</span> <span class="va">samples</span><span class="op">)</span></span>
<span><span class="va">merged_seurat</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
21210 features across 47881 samples within 1 assay 
Active assay: RNA (21210 features, 0 variable features)
 11 layers present: counts.Con4, counts.Con6, counts.Con9, counts.TAA1, counts.TAA2, counts.TAA3, counts.TAA4, counts.TAA5, counts.TAA6, counts.TAA7, counts.TAA8</code></pre>
</div>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                        orig.ident nCount_RNA nFeature_RNA
Con4_AAACCCAAGCGGACAT.1       Con4       5700         2067
Con4_AAACCCAAGCGTCAAG.1       Con4      12694         2850
Con4_AAACCCACATTCTCTA.1       Con4      14713         2946
Con4_AAACCCAGTCGACTTA.1       Con4       3000          546
Con4_AAACCCATCCATCAGA.1       Con4      15045         4106
Con4_AAACGAAAGGAACGCT.1       Con4       9877         2793
Con4_AAACGAAAGGTAGATT.1       Con4       6860         2050
Con4_AAACGAAAGTCAATCC.1       Con4       5628         1434
Con4_AAACGAACACGGCACT.1       Con4       9226         2217
Con4_AAACGAAGTTCAAAGA.1       Con4       6064         1743</code></pre>
</div>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 新增样本列</span></span>
<span><span class="va">merged_seurat</span><span class="op">$</span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="va">merged_seurat</span><span class="op">$</span><span class="va">orig.ident</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>简单查看一下细胞/基因数：</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Visualize the number of cell counts per sample</span></span>
<span><span class="va">merged_seurat</span><span class="op">@</span><span class="va">meta.data</span>  <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">samples</span>, fill <span class="op">=</span> <span class="va">samples</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, vjust <span class="op">=</span> <span class="fl">1</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span>, face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"NCells"</span><span class="op">)</span></span>
<span><span class="co"># Visualize the number UMIs/transcripts per cell</span></span>
<span><span class="va">merged_seurat</span><span class="op">@</span><span class="va">meta.data</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">samples</span>, x <span class="op">=</span> <span class="va">nCount_RNA</span>, fill <span class="op">=</span> <span class="va">samples</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span><span class="op">(</span><span class="st">"Cell density"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell quarto-layout-panel" data-hash="read_sc_data_cache/html/unnamed-chunk-16_a7b81f43cc3ee99dc02927afee9b7931" data-layout-ncol="2">
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img src="read_sc_data_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid" width="672"></p>
</div>
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<p><img src="read_sc_data_files/figure-html/unnamed-chunk-16-2.png" class="img-fluid" width="672"></p>
</div>
</div>
</div>
<hr></section><section id="sec-reading_non_standard_large_volume_text" class="level1" data-number="4"><h1 data-number="4">
<span class="header-section-number">4</span> 读取非标准大体积文本文件单细胞数据</h1>
<p>案例数据来自2020年的文章《<em>Single-cell transcriptomics reveals regulators underlying immune cell diversity and immune subtypes associated with prognosis in nasopharyngeal carcinoma</em>》<span class="citation" data-cites="chen2020">(<a href="#ref-chen2020" role="doc-biblioref">Chen et al. 2020</a>)</span>。该研究以15例鼻咽癌患者作为研究对象，1例慢性鼻咽炎患者的正常鼻咽上皮组织作为对照。所有样品均采用10x genomics V2 3’单细胞RNA测序。数据链接：<a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE150430"><strong>GSE150430</strong></a><strong>。</strong></p>
<p><img src="images/截屏2024-01-10 12.07.54.png" class="img-fluid"></p>
<section id="加载包" class="level2" data-number="4.1"><h2 data-number="4.1" class="anchored" data-anchor-id="加载包">
<span class="header-section-number">4.1</span> 加载包</h2>
<div class="cell" data-hash="read_sc_data_cache/html/unnamed-chunk-17_c478eaef9efdd609703b4ebe54dacaa5">
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/rm.html">rm</a></span><span class="op">(</span>list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ls.html">ls</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com">data.table</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://stringr.tidyverse.org">stringr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://purrr.tidyverse.org/">purrr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="数据读取" class="level2" data-number="4.2"><h2 data-number="4.2" class="anchored" data-anchor-id="数据读取">
<span class="header-section-number">4.2</span> 数据读取</h2>
<p>下载数据之后，首先按照通常做法通过<code>data.table</code>包的<code><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html">fread()</a></code>函数尝试读取之后，发现无法正确读取列名，即cell barcode：</p>
<div class="cell" data-hash="read_sc_data_cache/html/unnamed-chunk-18_89e930c14747255d03bef0c6c034c49d">
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html">fread</a></span><span class="op">(</span><span class="st">"data/sc_supplementary/GSE150430_npc_scRNA_hg19_processed_data.txt"</span>,</span>
<span>                data.table <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><img src="images/截屏2024-01-10 17.28.25.png" class="img-fluid"></p>
<p>而用<code><a href="https://rdrr.io/r/utils/read.table.html">read.table()</a></code>函数读取这个文件是可以正确读取列名的:</p>
<div class="cell" data-hash="read_sc_data_cache/html/unnamed-chunk-19_b7aeeb91b80f705149c7cb1dbee12dff">
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span><span class="op">(</span><span class="st">"data/sc_supplementary/GSE150430_npc_scRNA_hg19_processed_data.txt"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>读取后的数据如下：</p>
<p><img src="images/截屏2024-01-10 15.53.59.png" class="img-fluid"></p>
<p>可以发现该表达矩阵行名为基因，列为cell barcode，同时有代表样本的前缀。但是<strong>第一行为作者注释的细胞类型</strong>，因此，可能是由于这样的表头结构导致<code>fread</code>不能正确的读取这个数据。但是用基础的<code>read.table</code>读取这种大型数据非常耗时间，因此下面<strong>首先用<code>read.table</code>读取前两行的数据，即cell barcode和细胞类型；然后通过<code>fread</code>读取剩余的表达矩阵；最后将cell barcode和表达矩阵进行合并以构建Seurat对象</strong>。</p>
</section><section id="数据整理" class="level2" data-number="4.3"><h2 data-number="4.3" class="anchored" data-anchor-id="数据整理">
<span class="header-section-number">4.3</span> 数据整理</h2>
<div class="cell" data-hash="read_sc_data_cache/html/unnamed-chunk-20_511178f2883f2009e109ac6d4e09c49a">
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 单独读取表头</span></span>
<span><span class="va">head</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span><span class="op">(</span><span class="st">"data/sc_supplementary/GSE150430_npc_scRNA_hg19_processed_data.txt"</span>,</span>
<span>                   nrows <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                   header <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><img src="images/截屏2024-01-11 20.32.16.png" class="img-fluid"></p>
<div class="cell" data-hash="read_sc_data_cache/html/unnamed-chunk-21_8334736786fb10e09159e328dbd1e6c5">
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 读取表达矩阵</span></span>
<span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html">fread</a></span><span class="op">(</span><span class="st">"data/sc_supplementary/GSE150430_npc_scRNA_hg19_processed_data.txt"</span>,</span>
<span>                data.table <span class="op">=</span> <span class="cn">F</span>, </span>
<span>                skip <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="co"># 跳过前两行</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><img src="images/截屏2024-01-10 17.43.49.png" class="img-fluid"></p>
<div class="cell" data-hash="read_sc_data_cache/html/unnamed-chunk-22_e86a417ae35b0fd1d2c9e160b2786d9d">
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 将第一列（基因名）作为行名</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">counts</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="va">counts</span><span class="op">[</span>,<span class="op">-</span><span class="fl">1</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># 添加列名（cell barcode）</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">head</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="quarto-figure quarto-figure-left">
<figure class="figure"><p><img src="images/截屏2024-01-10 17.45.43.png" class="img-fluid figure-img"></p>
<figcaption>此时，我们就得到了一个满足Seurat对象构建的counts矩阵了。但是，当我们直接用<code><a href="https://rdrr.io/pkg/SeuratObject/man/CreateSeuratObject.html">CreateSeuratObject()</a></code>去创建Seurat对象时，由于该数据包含了所有样本的单细胞测序数据，导致其体积过大超出了R的内存上限。因此，接下来，我们<strong>将这个数据根据样本拆分成不同的RDS文件</strong>，然后再用<a href="#sec-read_raw_counts">上面</a>类似的方法依次读取这些RDS文件，避免一次性读取导致的内存溢出。</figcaption></figure>
</div>
<div class="cell" data-hash="read_sc_data_cache/html/unnamed-chunk-23_4f9a30be7580189d38248d5152a9af73">
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 提取样本标签</span></span>
<span><span class="va">samples</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_split.html">str_split</a></span><span class="op">(</span>string <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>, </span>
<span>                      pattern <span class="op">=</span> <span class="st">"_"</span>, </span>
<span>                      simplify <span class="op">=</span> <span class="cn">T</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">samples</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-hash="read_sc_data_cache/html/unnamed-chunk-24_dc8914488386038af10611732cc2418e">
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># [1] "N01" "P01" "P03" "P05" "P06" "P09" "P11" "P12" "P13" "P14" "P15" "P02"</span></span>
<span><span class="co"># [13] "P04" "P07" "P08" "P10"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-hash="read_sc_data_cache/html/unnamed-chunk-25_3f36d1f33765a91f6047578ce768a456">
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 逐一提取不同样本的测序数据，分别导出为RDS文件</span></span>
<span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span><span class="op">(</span><span class="va">samples</span>, <span class="kw">function</span><span class="op">(</span><span class="va">sample</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">sample</span><span class="op">)</span></span>
<span>  <span class="va">sub_exp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="va">sample</span>, </span>
<span>                  x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>    <span class="va">counts</span><span class="op">[</span>,<span class="va">.</span><span class="op">]</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">sub_exp</span>, </span>
<span>          file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"output/sc_supplementary/GSE150430/"</span>, <span class="va">sample</span>, <span class="st">".rds"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="构建seurat对象" class="level2" data-number="4.4"><h2 data-number="4.4" class="anchored" data-anchor-id="构建seurat对象">
<span class="header-section-number">4.4</span> 构建Seurat对象</h2>
<div class="cell" data-hash="read_sc_data_cache/html/unnamed-chunk-26_d743796d5636187abd8451be54cddccc">
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">filenames</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="st">"output/sc_supplementary/GSE150430"</span><span class="op">)</span></span>
<span><span class="va">filenames</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-hash="read_sc_data_cache/html/unnamed-chunk-27_41f62d5d7dba151fae738a3bbd089e48">
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">#  [1] "N01.rds" "P01.rds" "P02.rds" "P03.rds" "P04.rds" "P05.rds" "P06.rds"</span></span>
<span><span class="co">#  [8] "P07.rds" "P08.rds" "P09.rds" "P10.rds" "P11.rds" "P12.rds" "P13.rds"</span></span>
<span><span class="co"># [15] "P14.rds" "P15.rds"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-hash="read_sc_data_cache/html/unnamed-chunk-28_9f6896b26d362af867f761a9e64ce848">
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 循环读取每个rds文件的数据并创建Seurat对象</span></span>
<span><span class="va">seurat_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">filenames</span>, <span class="kw">function</span><span class="op">(</span><span class="va">file</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="co"># 拼接文件路径</span></span>
<span>  <span class="va">data.path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"output/sc_supplementary/GSE150430/"</span>, <span class="va">file</span><span class="op">)</span></span>
<span>  <span class="co"># 创建Seurat对象</span></span>
<span>  <span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/CreateSeuratObject.html">CreateSeuratObject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="va">data.path</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="va">seurat_obj</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 合并Seurat对象</span></span>
<span><span class="va">merged_seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html">merge</a></span><span class="op">(</span><span class="va">seurat_list</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,</span>
<span>                       y <span class="op">=</span> <span class="va">seurat_list</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">merged_seurat</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-hash="read_sc_data_cache/html/unnamed-chunk-29_f1d00827d04b0987744050237fd065dd">
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># An object of class Seurat </span></span>
<span><span class="co"># 24720 features across 48584 samples within 1 assay </span></span>
<span><span class="co"># Active assay: RNA (24720 features, 0 variable features)</span></span>
<span><span class="co">#  16 layers present: counts.1, counts.2, counts.3, counts.4, counts.5, counts.6, counts.7, counts.8, counts.9, counts.10, counts.11, counts.12, counts.13, counts.14, counts.15, counts.16</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-hash="read_sc_data_cache/html/unnamed-chunk-30_e535ed63b8adc75131283787f8cd937e">
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">merged_seurat</span><span class="op">)</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><img src="images/截屏2024-01-11 20.36.33.png" class="img-fluid" width="434"></p>
<div class="cell" data-hash="read_sc_data_cache/html/unnamed-chunk-31_05f4bb873ef272c0240916b4a0388b43">
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 添加样本列</span></span>
<span><span class="va">merged_seurat</span><span class="op">$</span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="va">merged_seurat</span><span class="op">$</span><span class="va">orig.ident</span></span>
<span></span>
<span><span class="co"># 保存</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">merged_seurat</span>, file <span class="op">=</span> <span class="st">"output/sc_supplementary/GSE150430_merged_seurat.rds"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr></section></section><section id="读取smart-seq测序数据" class="level1" data-number="5"><h1 data-number="5">
<span class="header-section-number">5</span> 读取SMART-seq测序数据</h1>
<p>数据链接：<a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE140228"><strong>GSE140228</strong></a><strong>。</strong></p>
<p><img src="images/截屏2024-01-18 14.41.29.png" class="img-fluid"></p>
<p>这个数据集用了两种测序技术，即Drop-seq和SMART-seq，分别属于测<strong>3′ 或5′ 端（3′或5′-end sequencing）</strong>的技术和测<strong>全长转录本（full-length transcript sequencing）</strong>的技术。这里我们下载SMART-seq测序数据。</p>
<p>读取barcodes、features和matrix数据:</p>
<div class="cell" data-hash="read_sc_data_cache/html/unnamed-chunk-32_8c6073da8d13f3422d226c1e4bedb229">
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/rm.html">rm</a></span><span class="op">(</span>list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ls.html">ls</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com">data.table</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="st">"data/sc_supplementary/GSE140228"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "GSE140228_cell_info_Smartseq2.tsv.gz"  
[2] "GSE140228_gene_info_Smartseq2.tsv.gz"  
[3] "GSE140228_read_counts_Smartseq2.csv"   
[4] "GSE140228_read_counts_Smartseq2.csv.gz"</code></pre>
</div>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">barcodes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html">fread</a></span><span class="op">(</span></span>
<span>  <span class="st">"data/sc_supplementary/GSE140228/GSE140228_cell_info_Smartseq2.tsv.gz"</span>,</span>
<span>  data.table <span class="op">=</span> <span class="cn">F</span> <span class="co"># 返回data.frame型数据</span></span>
<span>  <span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">barcodes</span>, <span class="fl">5</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          Barcode     Donor Tissue celltype_sub                 Sample
1 NL-plus-19-0222 D20170222 Normal  NK-C7-CD160 D20170222-Normal-CD45+
2 NL-plus-20-0222 D20170222 Normal  CD8-C6-GZMK D20170222-Normal-CD45+
3  NL-plus-2-0222 D20170222 Normal CD8-C8-PDCD1 D20170222-Normal-CD45+
4 NL-plus-22-0222 D20170222 Normal  CD8-C6-GZMK D20170222-Normal-CD45+
5 NL-plus-23-0222 D20170222 Normal CD8-C8-PDCD1 D20170222-Normal-CD45+
  Histology Tissue_sub
1       HCC     Normal
2       HCC     Normal
3       HCC     Normal
4       HCC     Normal
5       HCC     Normal</code></pre>
</div>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">features</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html">fread</a></span><span class="op">(</span></span>
<span>  <span class="st">"data/sc_supplementary/GSE140228/GSE140228_gene_info_Smartseq2.tsv.gz"</span>,</span>
<span>  data.table <span class="op">=</span> <span class="cn">F</span></span>
<span>  <span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">features</span>, <span class="fl">5</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          ENSEMBL    SYMBOL CHR START   END STRAND
1 ENSG00000223972   DDX11L1   1 11869 14409      +
2 ENSG00000278267 MIR6859-1   1 17369 17436      -
3 ENSG00000284332 MIR1302-2   1 30366 30503      +
4 ENSG00000237613   FAM138A   1 34554 36081      -
5 ENSG00000268020    OR4G4P   1 52473 53312      +
                             BIOTYPE
1 transcribed_unprocessed_pseudogene
2                              miRNA
3                              miRNA
4                            lincRNA
5             unprocessed_pseudogene</code></pre>
</div>
<div class="sourceCode" id="cb77"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html">fread</a></span><span class="op">(</span></span>
<span>  <span class="st">"data/sc_supplementary/GSE140228/GSE140228_read_counts_Smartseq2.csv.gz"</span>,</span>
<span>  data.table <span class="op">=</span> <span class="cn">F</span>,</span>
<span>  <span class="op">)</span></span>
<span><span class="va">matrix</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       gene NL-plus-19-0222 NL-plus-20-0222 NL-plus-2-0222 NL-plus-22-0222
1   DDX11L1               0               0              0               0
2 MIR6859-1               0               0              0               0
3 MIR1302-2               0               0              0               0
4   FAM138A               0               0              0               0
5    OR4G4P               0               0              0               0</code></pre>
</div>
</div>
<p>这个<code>matrix</code>第一列为基因名，列名为barcodes。<code>features</code>和<code>barcodes</code>是对每个基因和细胞的补充信息。所以，接下来我们可以直接用这个<code>matrix</code>文件构建counts矩阵。后续如有需要可以将<code>features</code>和<code>barcodes</code>中的信息补充上去。</p>
<div class="cell" data-hash="read_sc_data_cache/html/unnamed-chunk-33_873a0ea3817c4ed83364014107c41732">
<div class="sourceCode" id="cb79"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 构建counts矩阵</span></span>
<span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="va">matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/row.names.html">row.names</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">counts</span><span class="op">$</span><span class="va">gene</span></span>
<span><span class="va">counts</span><span class="op">$</span><span class="va">gene</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span></span>
<span><span class="co"># 创建Seurat对象</span></span>
<span><span class="va">merged_seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/CreateSeuratObject.html">CreateSeuratObject</a></span><span class="op">(</span></span>
<span>  counts <span class="op">=</span> <span class="va">counts</span>,</span>
<span>  project <span class="op">=</span> <span class="st">"HCC"</span>,</span>
<span>  min.cells <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  min.features <span class="op">=</span> <span class="fl">200</span></span>
<span><span class="op">)</span></span>
<span><span class="va">merged_seurat</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
44417 features across 7074 samples within 1 assay 
Active assay: RNA (44417 features, 0 variable features)
 1 layer present: counts</code></pre>
</div>
<div class="sourceCode" id="cb81"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">merged_seurat</span>, <span class="fl">5</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                orig.ident nCount_RNA nFeature_RNA
NL-plus-19-0222        HCC   685649.3         4313
NL-plus-20-0222        HCC   932696.0         4746
NL-plus-2-0222         HCC   692261.5         4567
NL-plus-22-0222        HCC   765533.3         4790
NL-plus-23-0222        HCC   574235.8         3549</code></pre>
</div>
</div>
<hr></section><section id="单独提供每个细胞测序结果的数据的读取" class="level1" data-number="6"><h1 data-number="6">
<span class="header-section-number">6</span> 单独提供每个细胞测序结果的数据的读取</h1>
<p>数据链接：<a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE84465"><strong>GSE84465</strong></a><strong>。</strong></p>
<p><img src="images/截屏2024-01-18 15.47.01.png" class="img-fluid" width="418"><img src="images/截屏2024-01-18 15.47.10.png" class="img-fluid" width="591"></p>
<p>可以看到，该数据集提供每一个细胞的测序结果。对于这样的数据，可以参考<a href="https://mp.weixin.qq.com/s?__biz=Mzg2MTExNTkwNA==&amp;mid=2247510395&amp;idx=1&amp;sn=f2a2c0e3d2d75656f1dcfceaf708beb8&amp;chksm=ce1eea14f969630295e8988209df55b61e2062178d43b22d9eb41a93badf23772a312830ab49&amp;mpshare=1&amp;scene=1&amp;srcid=1117TnC4m9nO1EleC5e8JIOl&amp;sharer_sharetime=1668645592942&amp;sharer_shareid=568f49cb24905bf546294e6985a632bf#rd"><em>单细胞数据下载和Seurat对象的构建（上）</em></a>中的方法来读取。</p>
<hr>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-2-contents" aria-controls="callout-2" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Session Info
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-2" class="callout-2-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-hash="read_sc_data_cache/html/unnamed-chunk-34_134fa4df9da792e145655b5dcfd69144">
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.3.2 (2023-10-31)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Sonoma 14.3

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: Asia/Shanghai
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] ggplot2_3.4.4      data.table_1.14.10 tibble_3.2.1       stringr_1.5.1     
[5] Seurat_5.0.1       SeuratObject_5.0.1 sp_2.1-2          

loaded via a namespace (and not attached):
  [1] deldir_2.0-2           pbapply_1.7-2          gridExtra_2.3         
  [4] rlang_1.1.3            magrittr_2.0.3         RcppAnnoy_0.0.21      
  [7] spatstat.geom_3.2-7    matrixStats_1.2.0      ggridges_0.5.5        
 [10] compiler_4.3.2         png_0.1-8              vctrs_0.6.5           
 [13] reshape2_1.4.4         pkgconfig_2.0.3        fastmap_1.1.1         
 [16] ellipsis_0.3.2         labeling_0.4.3         utf8_1.2.4            
 [19] promises_1.2.1         rmarkdown_2.25         purrr_1.0.2           
 [22] xfun_0.41              jsonlite_1.8.8         goftest_1.2-3         
 [25] later_1.3.2            spatstat.utils_3.0-4   irlba_2.3.5.1         
 [28] parallel_4.3.2         cluster_2.1.6          R6_2.5.1              
 [31] ica_1.0-3              stringi_1.8.3          RColorBrewer_1.1-3    
 [34] spatstat.data_3.0-4    reticulate_1.34.0      parallelly_1.36.0     
 [37] lmtest_0.9-40          scattermore_1.2        Rcpp_1.0.12           
 [40] knitr_1.45             tensor_1.5             future.apply_1.11.1   
 [43] zoo_1.8-12             R.utils_2.12.3         sctransform_0.4.1     
 [46] httpuv_1.6.13          Matrix_1.6-5           splines_4.3.2         
 [49] igraph_1.6.0           tidyselect_1.2.0       abind_1.4-5           
 [52] rstudioapi_0.15.0      yaml_2.3.8             spatstat.random_3.2-2 
 [55] codetools_0.2-19       miniUI_0.1.1.1         spatstat.explore_3.2-5
 [58] listenv_0.9.0          lattice_0.22-5         plyr_1.8.9            
 [61] withr_3.0.0            shiny_1.8.0            ROCR_1.0-11           
 [64] evaluate_0.23          Rtsne_0.17             future_1.33.1         
 [67] fastDummies_1.7.3      survival_3.5-7         polyclip_1.10-6       
 [70] fitdistrplus_1.1-11    pillar_1.9.0           KernSmooth_2.23-22    
 [73] plotly_4.10.4          generics_0.1.3         RcppHNSW_0.5.0        
 [76] munsell_0.5.0          scales_1.3.0           globals_0.16.2        
 [79] xtable_1.8-4           glue_1.7.0             lazyeval_0.2.2        
 [82] tools_4.3.2            RSpectra_0.16-1        RANN_2.6.1            
 [85] leiden_0.4.3.1         dotCall64_1.1-1        cowplot_1.1.2         
 [88] grid_4.3.2             tidyr_1.3.0            colorspace_2.1-0      
 [91] nlme_3.1-164           patchwork_1.2.0        cli_3.6.2             
 [94] spatstat.sparse_3.0-3  spam_2.10-0            fansi_1.0.6           
 [97] viridisLite_0.4.2      dplyr_1.1.4            uwot_0.1.16           
[100] gtable_0.3.4           R.methodsS3_1.8.2      digest_0.6.34         
[103] progressr_0.14.0       ggrepel_0.9.5          farver_2.1.1          
[106] htmlwidgets_1.6.4      R.oo_1.25.0            htmltools_0.5.7       
[109] lifecycle_1.0.4        httr_1.4.7             mime_0.12             
[112] MASS_7.3-60.0.1       </code></pre>
</div>
</div>
</div>
</div>
</div>
</section><section id="references" class="level1 unnumbered"><h1 class="unnumbered">References</h1>
<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-Bill2023" class="csl-entry" role="listitem">
Bill, Ruben, Pratyaksha Wirapati, Marius Messemaker, Whijae Roh, Beatrice Zitti, Florent Duval, Máté Kiss, et al. 2023. <span>“<span><em>CXCL9:SPP1</em></span> Macrophage Polarity Identifies a Network of Cellular Programs That Control Human Cancers.”</span> <em>Science</em> 381 (6657): 515–24. <a href="https://doi.org/10.1126/science.ade2292">https://doi.org/10.1126/science.ade2292</a>.
</div>
<div id="ref-chen2020" class="csl-entry" role="listitem">
Chen, Yu-Pei, Jian-Hua Yin, Wen-Fei Li, Han-Jie Li, Dong-Ping Chen, Cui-Juan Zhang, Jia-Wei Lv, et al. 2020. <span>“Single-Cell Transcriptomics Reveals Regulators Underlying Immune Cell Diversity and Immune Subtypes Associated with Prognosis in Nasopharyngeal Carcinoma.”</span> <em>Cell Research</em> 30 (11): 1024–42. <a href="https://doi.org/10.1038/s41422-020-0374-x">https://doi.org/10.1038/s41422-020-0374-x</a>.
</div>
<div id="ref-Li2020" class="csl-entry" role="listitem">
Li, Yanming, Pingping Ren, Ashley Dawson, Hernan G. Vasquez, Waleed Ageedi, Chen Zhang, Wei Luo, et al. 2020. <span>“Single-Cell Transcriptome Analysis Reveals Dynamic Cell Populations and Differential Gene Expression Patterns in Control and Aneurysmal Human Aortic Tissue.”</span> <em>Circulation</em> 142 (14): 1374–88. <a href="https://doi.org/10.1161/circulationaha.120.046528">https://doi.org/10.1161/circulationaha.120.046528</a>.
</div>
<div id="ref-xu2022" class="csl-entry" role="listitem">
Xu, Junfen, Yifeng Fang, Kelie Chen, Sen Li, Sangsang Tang, Yan Ren, Yixuan Cen, et al. 2022. <span>“Single-Cell RNA Sequencing Reveals the Tissue Architecture in Human High-Grade Serous Ovarian Cancer.”</span> <em>Clinical Cancer Research: An Official Journal of the American Association for Cancer Research</em> 28 (16): 3590–3602. <a href="https://doi.org/10.1158/1078-0432.CCR-22-0296">https://doi.org/10.1158/1078-0432.CCR-22-0296</a>.
</div>
</div>


<!-- -->

</section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/djhcod\.github\.io\/r-notes\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
          // target, if specified
          link.setAttribute("target", "_blank");
          if (link.getAttribute("rel") === null) {
            link.setAttribute("rel", "noopener");
          }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><script src="https://giscus.app/client.js" data-repo="djhcod/r-notes" data-repo-id="R_kgDOKw7GMg" data-category="General" data-category-id="DIC_kwDOKw7GMs4Cbz4B" data-mapping="title" data-reactions-enabled="1" data-emit-metadata="0" data-input-position="top" data-theme="light" data-lang="en" crossorigin="anonymous" async="">
</script><input type="hidden" id="giscus-base-theme" value="light"><input type="hidden" id="giscus-alt-theme" value="dark"><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="../../single_cell/sc_supplementary/sc_supplementary_intro.html" class="pagination-link  aria-label=">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">单细胞补充内容</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../../single_cell/sc_supplementary/universal_marker.html" class="pagination-link" aria-label="细胞分群通用marker">
        <span class="nav-page-text">细胞分群通用marker</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb84" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb84-2"><a href="#cb84-2" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> "读取非标准格式的单细胞数据"</span></span>
<span id="cb84-3"><a href="#cb84-3" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb84-4"><a href="#cb84-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-5"><a href="#cb84-5" aria-hidden="true" tabindex="-1"></a><span class="fu"># 读取非标准10X格式文件 {#Read_non-standard_10x_files}</span></span>
<span id="cb84-6"><a href="#cb84-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-7"><a href="#cb84-7" aria-hidden="true" tabindex="-1"></a>该案例来自文献 <span class="co">[</span><span class="ot">@xu2022</span><span class="co">]</span>。该研究为了系统地研究高级别浆膜上皮卵巢癌（HGSOC）的肿瘤内异质性，采用深层单细胞RNA测序技术(scRNA-seq)**对7例初治HGSOC早期和晚期患者及5例年龄匹配的非恶性卵巢组织标本进行肿瘤分析。**共获得59,324个HGSOC和非恶性卵巢组织的单细胞，其中，33264个细胞（56%）来⾃HGSOC肿瘤患者组织，26060个（44%）来⾃⾮恶性卵巢组织标本。原始数据：<span class="co">[</span><span class="ot">GSE184880</span><span class="co">](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE184880)</span>。</span>
<span id="cb84-8"><a href="#cb84-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-9"><a href="#cb84-9" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/640-01.png)</span></span>
<span id="cb84-10"><a href="#cb84-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-11"><a href="#cb84-11" aria-hidden="true" tabindex="-1"></a>可以看到原始数据以10X格式提供:</span>
<span id="cb84-12"><a href="#cb84-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-13"><a href="#cb84-13" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/截屏2024-01-08%2011.17.00.png)</span>{width="465"}</span>
<span id="cb84-14"><a href="#cb84-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-15"><a href="#cb84-15" aria-hidden="true" tabindex="-1"></a>但是，**没有按照标准的10X格式的三个文件命名，并且没有按照样本分别建立文件夹**。所以需要对其进行整理：</span>
<span id="cb84-16"><a href="#cb84-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-17"><a href="#cb84-17" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>首先给每个样本建立文件夹</span>
<span id="cb84-18"><a href="#cb84-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-19"><a href="#cb84-19" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>然后将对应的10X标准格式的三个文件分别复制到各个样本文件夹内</span>
<span id="cb84-20"><a href="#cb84-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-21"><a href="#cb84-21" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>最后将文件重命名为10X格式要求的标准形式，即**“barcodes.tsv.gz”、"features.tsv.gz"、"matrix.mtx.gz"**</span>
<span id="cb84-22"><a href="#cb84-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-25"><a href="#cb84-25" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb84-26"><a href="#cb84-26" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb84-27"><a href="#cb84-27" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(stringr)</span>
<span id="cb84-28"><a href="#cb84-28" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb84-29"><a href="#cb84-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-32"><a href="#cb84-32" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb84-33"><a href="#cb84-33" aria-hidden="true" tabindex="-1"></a><span class="co">#| warning: false</span></span>
<span id="cb84-34"><a href="#cb84-34" aria-hidden="true" tabindex="-1"></a><span class="co"># 给每个样本建立文件夹</span></span>
<span id="cb84-35"><a href="#cb84-35" aria-hidden="true" tabindex="-1"></a>files <span class="ot">&lt;-</span> <span class="fu">list.files</span>(<span class="st">"data/sc_supplementary/GSE184880_RAW"</span>)</span>
<span id="cb84-36"><a href="#cb84-36" aria-hidden="true" tabindex="-1"></a>files</span>
<span id="cb84-37"><a href="#cb84-37" aria-hidden="true" tabindex="-1"></a>dirnames <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="at">pattern =</span> <span class="st">".barcodes.tsv.gz|.genes.tsv.gz|.matrix.mtx.gz"</span>, </span>
<span id="cb84-38"><a href="#cb84-38" aria-hidden="true" tabindex="-1"></a>                 <span class="at">replacement =</span> <span class="st">""</span>, </span>
<span id="cb84-39"><a href="#cb84-39" aria-hidden="true" tabindex="-1"></a>                 <span class="at">x =</span> files) <span class="sc">%&gt;%</span>  </span>
<span id="cb84-40"><a href="#cb84-40" aria-hidden="true" tabindex="-1"></a>  <span class="fu">unique</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb84-41"><a href="#cb84-41" aria-hidden="true" tabindex="-1"></a>  <span class="fu">paste0</span>(<span class="st">"data/sc_supplementary/GSE184880_RAW/"</span>, .)</span>
<span id="cb84-42"><a href="#cb84-42" aria-hidden="true" tabindex="-1"></a>dirnames</span>
<span id="cb84-43"><a href="#cb84-43" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb84-44"><a href="#cb84-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-47"><a href="#cb84-47" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb84-48"><a href="#cb84-48" aria-hidden="true" tabindex="-1"></a><span class="co">#| output: false</span></span>
<span id="cb84-49"><a href="#cb84-49" aria-hidden="true" tabindex="-1"></a><span class="fu">lapply</span>(dirnames, dir.create, <span class="at">showWarnings =</span> <span class="cn">FALSE</span>)</span>
<span id="cb84-50"><a href="#cb84-50" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb84-51"><a href="#cb84-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-54"><a href="#cb84-54" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb84-55"><a href="#cb84-55" aria-hidden="true" tabindex="-1"></a><span class="co">#| output: false</span></span>
<span id="cb84-56"><a href="#cb84-56" aria-hidden="true" tabindex="-1"></a><span class="fu">lapply</span>(files, <span class="cf">function</span>(files) {</span>
<span id="cb84-57"><a href="#cb84-57" aria-hidden="true" tabindex="-1"></a>  <span class="co"># 复制相应的10X文件到各个样本文件夹内</span></span>
<span id="cb84-58"><a href="#cb84-58" aria-hidden="true" tabindex="-1"></a>  new_dir <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="at">pattern =</span> <span class="st">".barcodes.tsv.gz|.genes.tsv.gz|.matrix.mtx.gz"</span>, </span>
<span id="cb84-59"><a href="#cb84-59" aria-hidden="true" tabindex="-1"></a>                  <span class="at">replacement =</span> <span class="st">""</span>, </span>
<span id="cb84-60"><a href="#cb84-60" aria-hidden="true" tabindex="-1"></a>                  <span class="at">x =</span> files) <span class="sc">%&gt;%</span>  </span>
<span id="cb84-61"><a href="#cb84-61" aria-hidden="true" tabindex="-1"></a>    <span class="fu">paste0</span>(<span class="st">"data/sc_supplementary/GSE184880_RAW/"</span>, .)</span>
<span id="cb84-62"><a href="#cb84-62" aria-hidden="true" tabindex="-1"></a>  <span class="fu">file.copy</span>(<span class="at">from =</span> <span class="fu">paste0</span>(<span class="st">"data/sc_supplementary/GSE184880_RAW/"</span>, files), </span>
<span id="cb84-63"><a href="#cb84-63" aria-hidden="true" tabindex="-1"></a>            <span class="at">to =</span> new_dir)</span>
<span id="cb84-64"><a href="#cb84-64" aria-hidden="true" tabindex="-1"></a>  <span class="co"># 重命名3个10X文件</span></span>
<span id="cb84-65"><a href="#cb84-65" aria-hidden="true" tabindex="-1"></a>  new_filename <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(<span class="fu">grepl</span>(<span class="at">pattern =</span> <span class="st">"barcodes"</span>, <span class="at">x =</span> files), <span class="st">"barcodes.tsv.gz"</span>, </span>
<span id="cb84-66"><a href="#cb84-66" aria-hidden="true" tabindex="-1"></a>                         <span class="fu">ifelse</span>(<span class="fu">grepl</span>(<span class="at">pattern =</span> <span class="st">"genes"</span>, <span class="at">x =</span> files), <span class="st">"features.tsv.gz"</span>, </span>
<span id="cb84-67"><a href="#cb84-67" aria-hidden="true" tabindex="-1"></a>                         <span class="st">"matrix.mtx.gz"</span>))</span>
<span id="cb84-68"><a href="#cb84-68" aria-hidden="true" tabindex="-1"></a>  <span class="fu">file.rename</span>(<span class="at">from =</span> <span class="fu">paste</span>(new_dir, files, <span class="at">sep  =</span> <span class="st">"/"</span>),</span>
<span id="cb84-69"><a href="#cb84-69" aria-hidden="true" tabindex="-1"></a>              <span class="at">to =</span> <span class="fu">paste</span>(new_dir, new_filename, <span class="at">sep =</span> <span class="st">"/"</span>))</span>
<span id="cb84-70"><a href="#cb84-70" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb84-71"><a href="#cb84-71" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb84-72"><a href="#cb84-72" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-73"><a href="#cb84-73" aria-hidden="true" tabindex="-1"></a>::: {.callout-tip collapse="true"}</span>
<span id="cb84-74"><a href="#cb84-74" aria-hidden="true" tabindex="-1"></a><span class="fu">###### 用setwd简化对文件路径的定义</span></span>
<span id="cb84-75"><a href="#cb84-75" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-76"><a href="#cb84-76" aria-hidden="true" tabindex="-1"></a>由于在渲染本文档时，knitr会将工作目录恢复为Rproject根目录，所以不能使用getwd指定工作目录。在实际应用中，可以设置工作目录到目标数据文件夹下以简化对文件路径的定义：</span>
<span id="cb84-77"><a href="#cb84-77" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-80"><a href="#cb84-80" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb84-81"><a href="#cb84-81" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb84-82"><a href="#cb84-82" aria-hidden="true" tabindex="-1"></a>root_dir <span class="ot">&lt;-</span> <span class="fu">getwd</span>()</span>
<span id="cb84-83"><a href="#cb84-83" aria-hidden="true" tabindex="-1"></a><span class="fu">setwd</span>(<span class="st">"data/sc_supplementary/GSE184880_RAW"</span>)</span>
<span id="cb84-84"><a href="#cb84-84" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb84-85"><a href="#cb84-85" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-88"><a href="#cb84-88" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb84-89"><a href="#cb84-89" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb84-90"><a href="#cb84-90" aria-hidden="true" tabindex="-1"></a><span class="co"># 给每个样本建立文件夹</span></span>
<span id="cb84-91"><a href="#cb84-91" aria-hidden="true" tabindex="-1"></a>files <span class="ot">&lt;-</span> <span class="fu">list.files</span>()</span>
<span id="cb84-92"><a href="#cb84-92" aria-hidden="true" tabindex="-1"></a>files</span>
<span id="cb84-93"><a href="#cb84-93" aria-hidden="true" tabindex="-1"></a>dirnames <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="at">pattern =</span> <span class="st">".barcodes.tsv.gz|.genes.tsv.gz|.matrix.mtx.gz"</span>, </span>
<span id="cb84-94"><a href="#cb84-94" aria-hidden="true" tabindex="-1"></a>                 <span class="at">replacement =</span> <span class="st">""</span>, </span>
<span id="cb84-95"><a href="#cb84-95" aria-hidden="true" tabindex="-1"></a>                 <span class="at">x =</span> files) <span class="sc">|&gt;</span> <span class="fu">unique</span>()</span>
<span id="cb84-96"><a href="#cb84-96" aria-hidden="true" tabindex="-1"></a>dirnames</span>
<span id="cb84-97"><a href="#cb84-97" aria-hidden="true" tabindex="-1"></a><span class="fu">lapply</span>(dirnames, dir.create)</span>
<span id="cb84-98"><a href="#cb84-98" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-99"><a href="#cb84-99" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-100"><a href="#cb84-100" aria-hidden="true" tabindex="-1"></a><span class="fu">lapply</span>(files, <span class="cf">function</span>(files) {</span>
<span id="cb84-101"><a href="#cb84-101" aria-hidden="true" tabindex="-1"></a>  <span class="co"># 复制相应的10X文件到各个样本文件夹内</span></span>
<span id="cb84-102"><a href="#cb84-102" aria-hidden="true" tabindex="-1"></a>  new_dir <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="at">pattern =</span> <span class="st">".barcodes.tsv.gz|.genes.tsv.gz|.matrix.mtx.gz"</span>, </span>
<span id="cb84-103"><a href="#cb84-103" aria-hidden="true" tabindex="-1"></a>                      <span class="at">replacement =</span> <span class="st">""</span>, </span>
<span id="cb84-104"><a href="#cb84-104" aria-hidden="true" tabindex="-1"></a>                      <span class="at">x =</span> files) </span>
<span id="cb84-105"><a href="#cb84-105" aria-hidden="true" tabindex="-1"></a>  <span class="fu">file.copy</span>(<span class="at">from =</span> files, </span>
<span id="cb84-106"><a href="#cb84-106" aria-hidden="true" tabindex="-1"></a>            <span class="at">to =</span> new_dir)</span>
<span id="cb84-107"><a href="#cb84-107" aria-hidden="true" tabindex="-1"></a>  <span class="co"># 重命名3个10X文件</span></span>
<span id="cb84-108"><a href="#cb84-108" aria-hidden="true" tabindex="-1"></a>  new_filename <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(<span class="fu">grepl</span>(<span class="at">pattern =</span> <span class="st">"barcodes"</span>, <span class="at">x =</span> files), <span class="st">"barcodes.tsv.gz"</span>, </span>
<span id="cb84-109"><a href="#cb84-109" aria-hidden="true" tabindex="-1"></a>                         <span class="fu">ifelse</span>(<span class="fu">grepl</span>(<span class="at">pattern =</span> <span class="st">"genes"</span>, <span class="at">x =</span> files), <span class="st">"features.tsv.gz"</span>, </span>
<span id="cb84-110"><a href="#cb84-110" aria-hidden="true" tabindex="-1"></a>                         <span class="st">"matrix.mtx.gz"</span>))</span>
<span id="cb84-111"><a href="#cb84-111" aria-hidden="true" tabindex="-1"></a>  <span class="fu">file.rename</span>(<span class="at">from =</span> <span class="fu">paste</span>(new_dir, files, <span class="at">sep  =</span> <span class="st">"/"</span>),</span>
<span id="cb84-112"><a href="#cb84-112" aria-hidden="true" tabindex="-1"></a>              <span class="at">to =</span> <span class="fu">paste</span>(new_dir, new_filename, <span class="at">sep =</span> <span class="st">"/"</span>))</span>
<span id="cb84-113"><a href="#cb84-113" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb84-114"><a href="#cb84-114" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb84-115"><a href="#cb84-115" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-118"><a href="#cb84-118" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb84-119"><a href="#cb84-119" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb84-120"><a href="#cb84-120" aria-hidden="true" tabindex="-1"></a>seurat_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(dirnames, <span class="cf">function</span>(dirnames) {</span>
<span id="cb84-121"><a href="#cb84-121" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(dirnames)</span>
<span id="cb84-122"><a href="#cb84-122" aria-hidden="true" tabindex="-1"></a>  sce <span class="ot">&lt;-</span> <span class="fu">CreateSeuratObject</span>(<span class="at">counts =</span>  <span class="fu">Read10X</span>(dirnames),</span>
<span id="cb84-123"><a href="#cb84-123" aria-hidden="true" tabindex="-1"></a>                            <span class="at">project =</span>  <span class="fu">str_split</span>(<span class="at">string =</span> dirnames, </span>
<span id="cb84-124"><a href="#cb84-124" aria-hidden="true" tabindex="-1"></a>                                                 <span class="at">pattern =</span> <span class="st">"_"</span>, </span>
<span id="cb84-125"><a href="#cb84-125" aria-hidden="true" tabindex="-1"></a>                                                 <span class="at">simplify =</span> T)[,<span class="dv">2</span>],</span>
<span id="cb84-126"><a href="#cb84-126" aria-hidden="true" tabindex="-1"></a>                            <span class="at">min.cells =</span> <span class="dv">5</span>,</span>
<span id="cb84-127"><a href="#cb84-127" aria-hidden="true" tabindex="-1"></a>                            <span class="at">min.features =</span> <span class="dv">500</span>)</span>
<span id="cb84-128"><a href="#cb84-128" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(sce)</span>
<span id="cb84-129"><a href="#cb84-129" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb84-130"><a href="#cb84-130" aria-hidden="true" tabindex="-1"></a>merged_seurat <span class="ot">&lt;-</span> <span class="fu">merge</span>(<span class="at">x =</span> seurat_list[[<span class="dv">1</span>]],</span>
<span id="cb84-131"><a href="#cb84-131" aria-hidden="true" tabindex="-1"></a>                       <span class="at">y =</span> seurat_list[<span class="sc">-</span><span class="dv">1</span>],</span>
<span id="cb84-132"><a href="#cb84-132" aria-hidden="true" tabindex="-1"></a>                       <span class="at">add.cell.ids =</span> <span class="fu">str_split</span>(<span class="at">string =</span> dirnames, </span>
<span id="cb84-133"><a href="#cb84-133" aria-hidden="true" tabindex="-1"></a>                                                <span class="at">pattern =</span> <span class="st">"_"</span>, </span>
<span id="cb84-134"><a href="#cb84-134" aria-hidden="true" tabindex="-1"></a>                                                <span class="at">simplify =</span> T)[,<span class="dv">2</span>])</span>
<span id="cb84-135"><a href="#cb84-135" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-136"><a href="#cb84-136" aria-hidden="true" tabindex="-1"></a>merged_seurat</span>
<span id="cb84-137"><a href="#cb84-137" aria-hidden="true" tabindex="-1"></a>merged_seurat <span class="ot">&lt;-</span> <span class="fu">JoinLayers</span>(merged_seurat)</span>
<span id="cb84-138"><a href="#cb84-138" aria-hidden="true" tabindex="-1"></a>merged_seurat</span>
<span id="cb84-139"><a href="#cb84-139" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(merged_seurat)</span>
<span id="cb84-140"><a href="#cb84-140" aria-hidden="true" tabindex="-1"></a><span class="fu">tail</span>(merged_seurat)</span>
<span id="cb84-141"><a href="#cb84-141" aria-hidden="true" tabindex="-1"></a><span class="fu">unique</span>(merged_seurat<span class="sc">$</span>orig.ident)</span>
<span id="cb84-142"><a href="#cb84-142" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-143"><a href="#cb84-143" aria-hidden="true" tabindex="-1"></a><span class="co"># 将样本信息添加到meta.data的新的一列“sample”中:</span></span>
<span id="cb84-144"><a href="#cb84-144" aria-hidden="true" tabindex="-1"></a>merged_seurat<span class="sc">$</span>sample <span class="ot">&lt;-</span> merged_seurat<span class="sc">$</span>orig.ident</span>
<span id="cb84-145"><a href="#cb84-145" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb84-146"><a href="#cb84-146" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-147"><a href="#cb84-147" aria-hidden="true" tabindex="-1"></a>后续如有需要可以随时将工作目录恢复为Rproject根目录：</span>
<span id="cb84-148"><a href="#cb84-148" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-151"><a href="#cb84-151" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb84-152"><a href="#cb84-152" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb84-153"><a href="#cb84-153" aria-hidden="true" tabindex="-1"></a><span class="fu">setwd</span>(root_dir)</span>
<span id="cb84-154"><a href="#cb84-154" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb84-155"><a href="#cb84-155" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb84-156"><a href="#cb84-156" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-157"><a href="#cb84-157" aria-hidden="true" tabindex="-1"></a>接下来，就可以进行常规的Seurat对象构建：</span>
<span id="cb84-158"><a href="#cb84-158" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-161"><a href="#cb84-161" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb84-162"><a href="#cb84-162" aria-hidden="true" tabindex="-1"></a>seurat_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(dirnames, <span class="cf">function</span>(dirnames) {</span>
<span id="cb84-163"><a href="#cb84-163" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(dirnames)</span>
<span id="cb84-164"><a href="#cb84-164" aria-hidden="true" tabindex="-1"></a>  sce <span class="ot">&lt;-</span> <span class="fu">CreateSeuratObject</span>(<span class="at">counts =</span>  <span class="fu">Read10X</span>(dirnames),</span>
<span id="cb84-165"><a href="#cb84-165" aria-hidden="true" tabindex="-1"></a>                            <span class="at">project =</span>  <span class="fu">str_split</span>(<span class="at">string =</span> dirnames, </span>
<span id="cb84-166"><a href="#cb84-166" aria-hidden="true" tabindex="-1"></a>                                                 <span class="at">pattern =</span> <span class="st">"_"</span>, </span>
<span id="cb84-167"><a href="#cb84-167" aria-hidden="true" tabindex="-1"></a>                                                 <span class="at">simplify =</span> T)[,<span class="dv">4</span>],</span>
<span id="cb84-168"><a href="#cb84-168" aria-hidden="true" tabindex="-1"></a>                            <span class="at">min.cells =</span> <span class="dv">5</span>,</span>
<span id="cb84-169"><a href="#cb84-169" aria-hidden="true" tabindex="-1"></a>                            <span class="at">min.features =</span> <span class="dv">500</span>)</span>
<span id="cb84-170"><a href="#cb84-170" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(sce)</span>
<span id="cb84-171"><a href="#cb84-171" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb84-172"><a href="#cb84-172" aria-hidden="true" tabindex="-1"></a>merged_seurat <span class="ot">&lt;-</span> <span class="fu">merge</span>(<span class="at">x =</span> seurat_list[[<span class="dv">1</span>]],</span>
<span id="cb84-173"><a href="#cb84-173" aria-hidden="true" tabindex="-1"></a>                       <span class="at">y =</span> seurat_list[<span class="sc">-</span><span class="dv">1</span>],</span>
<span id="cb84-174"><a href="#cb84-174" aria-hidden="true" tabindex="-1"></a>                       <span class="at">add.cell.ids =</span> <span class="fu">str_split</span>(<span class="at">string =</span> dirnames, </span>
<span id="cb84-175"><a href="#cb84-175" aria-hidden="true" tabindex="-1"></a>                                                <span class="at">pattern =</span> <span class="st">"_"</span>, </span>
<span id="cb84-176"><a href="#cb84-176" aria-hidden="true" tabindex="-1"></a>                                                <span class="at">simplify =</span> T)[,<span class="dv">4</span>])</span>
<span id="cb84-177"><a href="#cb84-177" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-178"><a href="#cb84-178" aria-hidden="true" tabindex="-1"></a>merged_seurat</span>
<span id="cb84-179"><a href="#cb84-179" aria-hidden="true" tabindex="-1"></a>merged_seurat <span class="ot">&lt;-</span> <span class="fu">JoinLayers</span>(merged_seurat)</span>
<span id="cb84-180"><a href="#cb84-180" aria-hidden="true" tabindex="-1"></a>merged_seurat</span>
<span id="cb84-181"><a href="#cb84-181" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(merged_seurat)</span>
<span id="cb84-182"><a href="#cb84-182" aria-hidden="true" tabindex="-1"></a><span class="fu">tail</span>(merged_seurat)</span>
<span id="cb84-183"><a href="#cb84-183" aria-hidden="true" tabindex="-1"></a><span class="fu">unique</span>(merged_seurat<span class="sc">$</span>orig.ident)</span>
<span id="cb84-184"><a href="#cb84-184" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-185"><a href="#cb84-185" aria-hidden="true" tabindex="-1"></a><span class="co"># 将样本信息添加到meta.data的新的一列“sample”中:</span></span>
<span id="cb84-186"><a href="#cb84-186" aria-hidden="true" tabindex="-1"></a>merged_seurat<span class="sc">$</span>sample <span class="ot">&lt;-</span> merged_seurat<span class="sc">$</span>orig.ident</span>
<span id="cb84-187"><a href="#cb84-187" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb84-188"><a href="#cb84-188" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-189"><a href="#cb84-189" aria-hidden="true" tabindex="-1"></a>------------------------------------------------------------------------</span>
<span id="cb84-190"><a href="#cb84-190" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-191"><a href="#cb84-191" aria-hidden="true" tabindex="-1"></a><span class="fu"># 读取raw_counts形式的数据</span></span>
<span id="cb84-192"><a href="#cb84-192" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-193"><a href="#cb84-193" aria-hidden="true" tabindex="-1"></a>案例来自<span class="co">[</span><span class="ot">@Bill2023</span><span class="co">]</span>。样本来自头颈部鳞状细胞癌（HNSCC）患者，总共52例标本，包括24例原发肿瘤、16例局部复发和12例多器官远处转移。原发部位包括口腔、喉/下咽、口咽和鼻咽。其中15份样本HPV阳性。原始数据：<span class="co">[</span><span class="ot">GSE234933</span><span class="co">](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE234933)</span>。</span>
<span id="cb84-194"><a href="#cb84-194" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-195"><a href="#cb84-195" aria-hidden="true" tabindex="-1"></a>可以看到该数据集没有提供标准的10X格式的3个文件。</span>
<span id="cb84-196"><a href="#cb84-196" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-197"><a href="#cb84-197" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/截屏2024-01-03%2020.56.56.png)</span></span>
<span id="cb84-198"><a href="#cb84-198" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-199"><a href="#cb84-199" aria-hidden="true" tabindex="-1"></a>其中的“raw_counts”文件解压后是52个RDS文件，对应了每个样本的单细胞测序数据，是一个已经包含了表达量、基因名和细胞barcode的矩阵，所以可以通过批量读取这些文件建立Seurat对象。</span>
<span id="cb84-200"><a href="#cb84-200" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-201"><a href="#cb84-201" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/截屏2024-01-09%2011.10.35.png)</span>{width="492"}</span>
<span id="cb84-202"><a href="#cb84-202" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-203"><a href="#cb84-203" aria-hidden="true" tabindex="-1"></a>“sample_annotation.txt”文件包含了对每个样本的详细临床病理资料的注释，因此可以将其添加到meta.data中。</span>
<span id="cb84-204"><a href="#cb84-204" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-205"><a href="#cb84-205" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/截屏2024-01-09%2011.11.45.png)</span></span>
<span id="cb84-206"><a href="#cb84-206" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-207"><a href="#cb84-207" aria-hidden="true" tabindex="-1"></a>“cell_annotation.txt”文件则是原作者对细胞类型的注释，可以不添加。</span>
<span id="cb84-208"><a href="#cb84-208" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-209"><a href="#cb84-209" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/截屏2024-01-09%2011.12.14.png)</span></span>
<span id="cb84-210"><a href="#cb84-210" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-211"><a href="#cb84-211" aria-hidden="true" tabindex="-1"></a><span class="fu">## 读取raw_counts矩阵 {#sec-read_raw_counts}</span></span>
<span id="cb84-212"><a href="#cb84-212" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-215"><a href="#cb84-215" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb84-216"><a href="#cb84-216" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(<span class="at">list =</span> <span class="fu">ls</span>())</span>
<span id="cb84-217"><a href="#cb84-217" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb84-218"><a href="#cb84-218" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(stringr)</span>
<span id="cb84-219"><a href="#cb84-219" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tibble)</span>
<span id="cb84-220"><a href="#cb84-220" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb84-221"><a href="#cb84-221" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-224"><a href="#cb84-224" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb84-225"><a href="#cb84-225" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb84-226"><a href="#cb84-226" aria-hidden="true" tabindex="-1"></a><span class="co"># 获取所有rds文件的列表</span></span>
<span id="cb84-227"><a href="#cb84-227" aria-hidden="true" tabindex="-1"></a>file_list <span class="ot">&lt;-</span> <span class="fu">list.files</span>(<span class="st">"data/sc_supplementary/GSE234933/GSE234933_MGH_HNSCC_gex_raw_counts"</span>, <span class="at">pattern =</span> <span class="st">".rds"</span>)</span>
<span id="cb84-228"><a href="#cb84-228" aria-hidden="true" tabindex="-1"></a>file_list</span>
<span id="cb84-229"><a href="#cb84-229" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-230"><a href="#cb84-230" aria-hidden="true" tabindex="-1"></a><span class="co"># 循环读取每个rds文件的数据并创建Seurat对象</span></span>
<span id="cb84-231"><a href="#cb84-231" aria-hidden="true" tabindex="-1"></a>seurat_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(file_list, <span class="cf">function</span>(file) {</span>
<span id="cb84-232"><a href="#cb84-232" aria-hidden="true" tabindex="-1"></a>  <span class="co"># 拼接文件路径</span></span>
<span id="cb84-233"><a href="#cb84-233" aria-hidden="true" tabindex="-1"></a>  data.path <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"data/sc_supplementary/GSE234933/GSE234933_MGH_HNSCC_gex_raw_counts/"</span>, file)</span>
<span id="cb84-234"><a href="#cb84-234" aria-hidden="true" tabindex="-1"></a>  <span class="co"># 创建Seurat对象</span></span>
<span id="cb84-235"><a href="#cb84-235" aria-hidden="true" tabindex="-1"></a>  seurat_obj <span class="ot">&lt;-</span> <span class="fu">CreateSeuratObject</span>(<span class="at">counts =</span> <span class="fu">readRDS</span>(data.path),</span>
<span id="cb84-236"><a href="#cb84-236" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">min.features =</span> <span class="dv">200</span>,</span>
<span id="cb84-237"><a href="#cb84-237" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">min.cells =</span> <span class="dv">3</span>)</span>
<span id="cb84-238"><a href="#cb84-238" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(seurat_obj)</span>
<span id="cb84-239"><a href="#cb84-239" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb84-240"><a href="#cb84-240" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-241"><a href="#cb84-241" aria-hidden="true" tabindex="-1"></a><span class="co"># 合并Seurat对象</span></span>
<span id="cb84-242"><a href="#cb84-242" aria-hidden="true" tabindex="-1"></a>seurat_combined <span class="ot">&lt;-</span> <span class="fu">merge</span>(seurat_list[[<span class="dv">1</span>]],</span>
<span id="cb84-243"><a href="#cb84-243" aria-hidden="true" tabindex="-1"></a>                         <span class="at">y =</span> seurat_list[<span class="sc">-</span><span class="dv">1</span>],</span>
<span id="cb84-244"><a href="#cb84-244" aria-hidden="true" tabindex="-1"></a>                         <span class="at">add.cell.ids =</span> <span class="fu">str_split</span>(<span class="at">string =</span> file_list, </span>
<span id="cb84-245"><a href="#cb84-245" aria-hidden="true" tabindex="-1"></a>                                                  <span class="at">pattern =</span> <span class="st">"_"</span>, </span>
<span id="cb84-246"><a href="#cb84-246" aria-hidden="true" tabindex="-1"></a>                                                  <span class="at">simplify =</span> T)[,<span class="dv">1</span>])</span>
<span id="cb84-247"><a href="#cb84-247" aria-hidden="true" tabindex="-1"></a>seurat_combined</span>
<span id="cb84-248"><a href="#cb84-248" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(seurat_combined) <span class="co"># 可以看到这个案例实际上不需要添加cell.ids</span></span>
<span id="cb84-249"><a href="#cb84-249" aria-hidden="true" tabindex="-1"></a><span class="fu">unique</span>(seurat_combined<span class="sc">$</span>orig.ident)</span>
<span id="cb84-250"><a href="#cb84-250" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb84-251"><a href="#cb84-251" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-252"><a href="#cb84-252" aria-hidden="true" tabindex="-1"></a><span class="fu">## 添加样本信息</span></span>
<span id="cb84-253"><a href="#cb84-253" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-256"><a href="#cb84-256" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb84-257"><a href="#cb84-257" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb84-258"><a href="#cb84-258" aria-hidden="true" tabindex="-1"></a>sample_anno <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"data/sc_supplementary/GSE234933/GSE234933_MGH_HNSCC_sample_annotation.txt"</span>, </span>
<span id="cb84-259"><a href="#cb84-259" aria-hidden="true" tabindex="-1"></a>                          <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, </span>
<span id="cb84-260"><a href="#cb84-260" aria-hidden="true" tabindex="-1"></a>                          <span class="at">header =</span> T)</span>
<span id="cb84-261"><a href="#cb84-261" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(sample_anno)</span>
<span id="cb84-262"><a href="#cb84-262" aria-hidden="true" tabindex="-1"></a>sample_anno<span class="sc">$</span>Sample</span>
<span id="cb84-263"><a href="#cb84-263" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-264"><a href="#cb84-264" aria-hidden="true" tabindex="-1"></a>old_meta.data <span class="ot">&lt;-</span> <span class="fu">rownames_to_column</span>(seurat_combined<span class="sc">@</span>meta.data, <span class="at">var =</span> <span class="st">"cell_id"</span>)</span>
<span id="cb84-265"><a href="#cb84-265" aria-hidden="true" tabindex="-1"></a>new_meta.data <span class="ot">&lt;-</span> <span class="fu">merge</span>(<span class="at">x =</span> old_meta.data,</span>
<span id="cb84-266"><a href="#cb84-266" aria-hidden="true" tabindex="-1"></a>                       <span class="at">y =</span> sample_anno,</span>
<span id="cb84-267"><a href="#cb84-267" aria-hidden="true" tabindex="-1"></a>                       <span class="at">by.x =</span> <span class="st">"orig.ident"</span>,</span>
<span id="cb84-268"><a href="#cb84-268" aria-hidden="true" tabindex="-1"></a>                       <span class="at">by.y =</span> <span class="st">"Sample"</span>,</span>
<span id="cb84-269"><a href="#cb84-269" aria-hidden="true" tabindex="-1"></a>                       <span class="at">all.x =</span> T) </span>
<span id="cb84-270"><a href="#cb84-270" aria-hidden="true" tabindex="-1"></a>new_meta.data<span class="sc">$</span>sample <span class="ot">&lt;-</span> new_meta.data<span class="sc">$</span>orig.ident</span>
<span id="cb84-271"><a href="#cb84-271" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(new_meta.data)</span>
<span id="cb84-272"><a href="#cb84-272" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(new_meta.data) <span class="ot">&lt;-</span> new_meta.data<span class="sc">$</span>cell_id</span>
<span id="cb84-273"><a href="#cb84-273" aria-hidden="true" tabindex="-1"></a>new_meta.data<span class="sc">$</span>cell_id <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb84-274"><a href="#cb84-274" aria-hidden="true" tabindex="-1"></a>new_meta.data[,<span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">6</span><span class="sc">:</span><span class="dv">10</span>)] <span class="ot">&lt;-</span> <span class="fu">lapply</span>(new_meta.data[,<span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">5</span>, <span class="dv">7</span><span class="sc">:</span><span class="dv">10</span>)], as.factor)</span>
<span id="cb84-275"><a href="#cb84-275" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(new_meta.data)</span>
<span id="cb84-276"><a href="#cb84-276" aria-hidden="true" tabindex="-1"></a>seurat_combined<span class="sc">@</span>meta.data <span class="ot">&lt;-</span> new_meta.data</span>
<span id="cb84-277"><a href="#cb84-277" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb84-278"><a href="#cb84-278" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-279"><a href="#cb84-279" aria-hidden="true" tabindex="-1"></a><span class="fu"># 读取txt/tsv/csv等文本文件表达量矩阵信息</span></span>
<span id="cb84-280"><a href="#cb84-280" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-281"><a href="#cb84-281" aria-hidden="true" tabindex="-1"></a>以2020的文章：《Single-Cell Transcriptome Analysis Reveals Dynamic Cell Populations and Differential Gene Expression Patterns in Control and Aneurysmal Human Aortic Tissue》<span class="co">[</span><span class="ot">@Li2020</span><span class="co">]</span> 举例说明。数据集链接：<span class="co">[</span><span class="ot">**GSE155468**</span><span class="co">](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE155468)</span>**。**</span>
<span id="cb84-282"><a href="#cb84-282" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-283"><a href="#cb84-283" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/截屏2024-01-10%2011.02.14.png)</span>{width="565"}</span>
<span id="cb84-284"><a href="#cb84-284" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-285"><a href="#cb84-285" aria-hidden="true" tabindex="-1"></a>该案例提供了11个人体主动脉组织样品的单细胞转录组测序数据，其中8名患者为升胸主动脉瘤（ATAA）and 3名患者作为对照。每个样品都是一个独立的txt文本文件，蕴藏着其表达量矩阵信息。</span>
<span id="cb84-286"><a href="#cb84-286" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-287"><a href="#cb84-287" aria-hidden="true" tabindex="-1"></a>首先我们读取其中一个文件看一下其形式：</span>
<span id="cb84-288"><a href="#cb84-288" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-291"><a href="#cb84-291" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb84-292"><a href="#cb84-292" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(<span class="at">list =</span> <span class="fu">ls</span>())</span>
<span id="cb84-293"><a href="#cb84-293" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(data.table)</span>
<span id="cb84-294"><a href="#cb84-294" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb84-295"><a href="#cb84-295" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb84-296"><a href="#cb84-296" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb84-297"><a href="#cb84-297" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-300"><a href="#cb84-300" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb84-301"><a href="#cb84-301" aria-hidden="true" tabindex="-1"></a><span class="co"># 列出所有的文件名</span></span>
<span id="cb84-302"><a href="#cb84-302" aria-hidden="true" tabindex="-1"></a>files <span class="ot">&lt;-</span> <span class="fu">list.files</span>(<span class="st">"data/sc_supplementary/GSE155468_RAW"</span>)</span>
<span id="cb84-303"><a href="#cb84-303" aria-hidden="true" tabindex="-1"></a>files</span>
<span id="cb84-304"><a href="#cb84-304" aria-hidden="true" tabindex="-1"></a><span class="co"># 简化文件名用作样本名</span></span>
<span id="cb84-305"><a href="#cb84-305" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="at">pattern =</span> <span class="st">"GSM.*._|.txt.gz"</span>, <span class="at">replacement =</span> <span class="st">""</span>, <span class="at">x =</span> files)</span>
<span id="cb84-306"><a href="#cb84-306" aria-hidden="true" tabindex="-1"></a>samples</span>
<span id="cb84-307"><a href="#cb84-307" aria-hidden="true" tabindex="-1"></a><span class="co"># 试读取其中一个文件</span></span>
<span id="cb84-308"><a href="#cb84-308" aria-hidden="true" tabindex="-1"></a>test <span class="ot">&lt;-</span> <span class="fu">fread</span>(<span class="st">"data/sc_supplementary/GSE155468_RAW/GSM4704931_Con4.txt.gz"</span>, <span class="at">data.table =</span> F)</span>
<span id="cb84-309"><a href="#cb84-309" aria-hidden="true" tabindex="-1"></a>test[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]</span>
<span id="cb84-310"><a href="#cb84-310" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb84-311"><a href="#cb84-311" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-312"><a href="#cb84-312" aria-hidden="true" tabindex="-1"></a>可以看到，该表达矩阵的问题是行名不是基因名没有作为行名，而是以单独的一列显示，因此需要进行预处理后再用上面类似的方法批量读取和创建Seurat对象：</span>
<span id="cb84-313"><a href="#cb84-313" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-316"><a href="#cb84-316" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb84-317"><a href="#cb84-317" aria-hidden="true" tabindex="-1"></a>seurat_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(files, <span class="cf">function</span>(file) {</span>
<span id="cb84-318"><a href="#cb84-318" aria-hidden="true" tabindex="-1"></a>  filepath <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"data/sc_supplementary/GSE155468_RAW/"</span>, file)</span>
<span id="cb84-319"><a href="#cb84-319" aria-hidden="true" tabindex="-1"></a>  counts <span class="ot">&lt;-</span> <span class="fu">fread</span>(filepath, <span class="at">data.table =</span> F)</span>
<span id="cb84-320"><a href="#cb84-320" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames</span>(counts) <span class="ot">&lt;-</span> counts[,<span class="dv">1</span>]</span>
<span id="cb84-321"><a href="#cb84-321" aria-hidden="true" tabindex="-1"></a>  counts <span class="ot">&lt;-</span> counts[,<span class="sc">-</span><span class="dv">1</span>]</span>
<span id="cb84-322"><a href="#cb84-322" aria-hidden="true" tabindex="-1"></a>  <span class="fu">CreateSeuratObject</span>(<span class="at">counts =</span> counts,</span>
<span id="cb84-323"><a href="#cb84-323" aria-hidden="true" tabindex="-1"></a>                     <span class="at">project =</span> <span class="fu">gsub</span>(<span class="at">pattern =</span> <span class="st">"GSM.*._|.txt.gz"</span>, </span>
<span id="cb84-324"><a href="#cb84-324" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">replacement =</span> <span class="st">""</span>, </span>
<span id="cb84-325"><a href="#cb84-325" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">x =</span> file),</span>
<span id="cb84-326"><a href="#cb84-326" aria-hidden="true" tabindex="-1"></a>                     <span class="at">min.cells =</span> <span class="dv">5</span>,</span>
<span id="cb84-327"><a href="#cb84-327" aria-hidden="true" tabindex="-1"></a>                     <span class="at">min.features =</span> <span class="dv">300</span>)</span>
<span id="cb84-328"><a href="#cb84-328" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb84-329"><a href="#cb84-329" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-330"><a href="#cb84-330" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-331"><a href="#cb84-331" aria-hidden="true" tabindex="-1"></a>merged_seurat <span class="ot">&lt;-</span> <span class="fu">merge</span>(<span class="at">x =</span> seurat_list[[<span class="dv">1</span>]], </span>
<span id="cb84-332"><a href="#cb84-332" aria-hidden="true" tabindex="-1"></a>                       <span class="at">y =</span> seurat_list[<span class="sc">-</span><span class="dv">1</span>], </span>
<span id="cb84-333"><a href="#cb84-333" aria-hidden="true" tabindex="-1"></a>                       <span class="at">add.cell.ids =</span> samples)</span>
<span id="cb84-334"><a href="#cb84-334" aria-hidden="true" tabindex="-1"></a>merged_seurat</span>
<span id="cb84-335"><a href="#cb84-335" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(merged_seurat)</span>
<span id="cb84-336"><a href="#cb84-336" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-337"><a href="#cb84-337" aria-hidden="true" tabindex="-1"></a><span class="co"># 新增样本列</span></span>
<span id="cb84-338"><a href="#cb84-338" aria-hidden="true" tabindex="-1"></a>merged_seurat<span class="sc">$</span>samples <span class="ot">&lt;-</span> merged_seurat<span class="sc">$</span>orig.ident</span>
<span id="cb84-339"><a href="#cb84-339" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb84-340"><a href="#cb84-340" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-341"><a href="#cb84-341" aria-hidden="true" tabindex="-1"></a>简单查看一下细胞/基因数：</span>
<span id="cb84-342"><a href="#cb84-342" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-345"><a href="#cb84-345" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb84-346"><a href="#cb84-346" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 7</span></span>
<span id="cb84-347"><a href="#cb84-347" aria-hidden="true" tabindex="-1"></a><span class="co">#| layout-ncol: 2</span></span>
<span id="cb84-348"><a href="#cb84-348" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualize the number of cell counts per sample</span></span>
<span id="cb84-349"><a href="#cb84-349" aria-hidden="true" tabindex="-1"></a>merged_seurat<span class="sc">@</span>meta.data  <span class="sc">|&gt;</span> </span>
<span id="cb84-350"><a href="#cb84-350" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> samples, <span class="at">fill =</span> samples)) <span class="sc">+</span> </span>
<span id="cb84-351"><a href="#cb84-351" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_bar</span>() <span class="sc">+</span></span>
<span id="cb84-352"><a href="#cb84-352" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb84-353"><a href="#cb84-353" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">45</span>, <span class="at">vjust =</span> <span class="dv">1</span>, <span class="at">hjust =</span> <span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb84-354"><a href="#cb84-354" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>, <span class="at">face =</span> <span class="st">"bold"</span>)) <span class="sc">+</span></span>
<span id="cb84-355"><a href="#cb84-355" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"NCells"</span>)</span>
<span id="cb84-356"><a href="#cb84-356" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-357"><a href="#cb84-357" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualize the number UMIs/transcripts per cell</span></span>
<span id="cb84-358"><a href="#cb84-358" aria-hidden="true" tabindex="-1"></a>merged_seurat<span class="sc">@</span>meta.data <span class="sc">|&gt;</span> </span>
<span id="cb84-359"><a href="#cb84-359" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">color =</span> samples, <span class="at">x =</span> nCount_RNA, <span class="at">fill =</span> samples)) <span class="sc">+</span> </span>
<span id="cb84-360"><a href="#cb84-360" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_density</span>(<span class="at">alpha =</span> <span class="fl">0.2</span>) <span class="sc">+</span> </span>
<span id="cb84-361"><a href="#cb84-361" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_x_log10</span>() <span class="sc">+</span> </span>
<span id="cb84-362"><a href="#cb84-362" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb84-363"><a href="#cb84-363" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ylab</span>(<span class="st">"Cell density"</span>) <span class="sc">+</span></span>
<span id="cb84-364"><a href="#cb84-364" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">500</span>)</span>
<span id="cb84-365"><a href="#cb84-365" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb84-366"><a href="#cb84-366" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-367"><a href="#cb84-367" aria-hidden="true" tabindex="-1"></a>------------------------------------------------------------------------</span>
<span id="cb84-368"><a href="#cb84-368" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-369"><a href="#cb84-369" aria-hidden="true" tabindex="-1"></a><span class="fu"># 读取非标准大体积文本文件单细胞数据 {#sec-reading_non_standard_large_volume_text}</span></span>
<span id="cb84-370"><a href="#cb84-370" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-371"><a href="#cb84-371" aria-hidden="true" tabindex="-1"></a>案例数据来自2020年的文章《*Single-cell transcriptomics reveals regulators underlying immune cell diversity and immune subtypes associated with prognosis in nasopharyngeal carcinoma*》[@chen2020]。该研究以15例鼻咽癌患者作为研究对象，1例慢性鼻咽炎患者的正常鼻咽上皮组织作为对照。所有样品均采用10x genomics V2 3'单细胞RNA测序。数据链接：[**GSE150430**](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE150430)**。**</span>
<span id="cb84-372"><a href="#cb84-372" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-373"><a href="#cb84-373" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/截屏2024-01-10%2012.07.54.png)</span></span>
<span id="cb84-374"><a href="#cb84-374" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-375"><a href="#cb84-375" aria-hidden="true" tabindex="-1"></a><span class="fu">## 加载包</span></span>
<span id="cb84-376"><a href="#cb84-376" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-379"><a href="#cb84-379" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb84-380"><a href="#cb84-380" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb84-381"><a href="#cb84-381" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(<span class="at">list =</span> <span class="fu">ls</span>())</span>
<span id="cb84-382"><a href="#cb84-382" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(data.table)</span>
<span id="cb84-383"><a href="#cb84-383" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(stringr)</span>
<span id="cb84-384"><a href="#cb84-384" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb84-385"><a href="#cb84-385" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(purrr)</span>
<span id="cb84-386"><a href="#cb84-386" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb84-387"><a href="#cb84-387" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb84-388"><a href="#cb84-388" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-389"><a href="#cb84-389" aria-hidden="true" tabindex="-1"></a><span class="fu">## 数据读取</span></span>
<span id="cb84-390"><a href="#cb84-390" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-391"><a href="#cb84-391" aria-hidden="true" tabindex="-1"></a>下载数据之后，首先按照通常做法通过<span class="in">`data.table`</span>包的<span class="in">`fread()`</span>函数尝试读取之后，发现无法正确读取列名，即cell barcode：</span>
<span id="cb84-392"><a href="#cb84-392" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-395"><a href="#cb84-395" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb84-396"><a href="#cb84-396" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb84-397"><a href="#cb84-397" aria-hidden="true" tabindex="-1"></a>counts <span class="ot">&lt;-</span> <span class="fu">fread</span>(<span class="st">"data/sc_supplementary/GSE150430_npc_scRNA_hg19_processed_data.txt"</span>,</span>
<span id="cb84-398"><a href="#cb84-398" aria-hidden="true" tabindex="-1"></a>                <span class="at">data.table =</span> F)</span>
<span id="cb84-399"><a href="#cb84-399" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb84-400"><a href="#cb84-400" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-401"><a href="#cb84-401" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/截屏2024-01-10%2017.28.25.png)</span></span>
<span id="cb84-402"><a href="#cb84-402" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-403"><a href="#cb84-403" aria-hidden="true" tabindex="-1"></a>而用<span class="in">`read.table()`</span>函数读取这个文件是可以正确读取列名的:</span>
<span id="cb84-404"><a href="#cb84-404" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-407"><a href="#cb84-407" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb84-408"><a href="#cb84-408" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb84-409"><a href="#cb84-409" aria-hidden="true" tabindex="-1"></a>counts <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"data/sc_supplementary/GSE150430_npc_scRNA_hg19_processed_data.txt"</span>)</span>
<span id="cb84-410"><a href="#cb84-410" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb84-411"><a href="#cb84-411" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-412"><a href="#cb84-412" aria-hidden="true" tabindex="-1"></a>读取后的数据如下：</span>
<span id="cb84-413"><a href="#cb84-413" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-414"><a href="#cb84-414" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/截屏2024-01-10%2015.53.59.png)</span></span>
<span id="cb84-415"><a href="#cb84-415" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-416"><a href="#cb84-416" aria-hidden="true" tabindex="-1"></a>可以发现该表达矩阵行名为基因，列为cell barcode，同时有代表样本的前缀。但是**第一行为作者注释的细胞类型**，因此，可能是由于这样的表头结构导致`fread`不能正确的读取这个数据。但是用基础的`read.table`读取这种大型数据非常耗时间，因此下面**首先用`read.table`读取前两行的数据，即cell barcode和细胞类型；然后通过`fread`读取剩余的表达矩阵；最后将cell barcode和表达矩阵进行合并以构建Seurat对象**。</span>
<span id="cb84-417"><a href="#cb84-417" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-418"><a href="#cb84-418" aria-hidden="true" tabindex="-1"></a><span class="fu">## 数据整理</span></span>
<span id="cb84-419"><a href="#cb84-419" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-422"><a href="#cb84-422" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb84-423"><a href="#cb84-423" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb84-424"><a href="#cb84-424" aria-hidden="true" tabindex="-1"></a><span class="co"># 单独读取表头</span></span>
<span id="cb84-425"><a href="#cb84-425" aria-hidden="true" tabindex="-1"></a>head <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"data/sc_supplementary/GSE150430_npc_scRNA_hg19_processed_data.txt"</span>,</span>
<span id="cb84-426"><a href="#cb84-426" aria-hidden="true" tabindex="-1"></a>                   <span class="at">nrows =</span> <span class="dv">1</span>,</span>
<span id="cb84-427"><a href="#cb84-427" aria-hidden="true" tabindex="-1"></a>                   <span class="at">header =</span> T)</span>
<span id="cb84-428"><a href="#cb84-428" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb84-429"><a href="#cb84-429" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-430"><a href="#cb84-430" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/截屏2024-01-11%2020.32.16.png)</span></span>
<span id="cb84-431"><a href="#cb84-431" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-434"><a href="#cb84-434" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb84-435"><a href="#cb84-435" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb84-436"><a href="#cb84-436" aria-hidden="true" tabindex="-1"></a><span class="co"># 读取表达矩阵</span></span>
<span id="cb84-437"><a href="#cb84-437" aria-hidden="true" tabindex="-1"></a>counts <span class="ot">&lt;-</span> <span class="fu">fread</span>(<span class="st">"data/sc_supplementary/GSE150430_npc_scRNA_hg19_processed_data.txt"</span>,</span>
<span id="cb84-438"><a href="#cb84-438" aria-hidden="true" tabindex="-1"></a>                <span class="at">data.table =</span> F, </span>
<span id="cb84-439"><a href="#cb84-439" aria-hidden="true" tabindex="-1"></a>                <span class="at">skip =</span> <span class="dv">2</span>) <span class="co"># 跳过前两行</span></span>
<span id="cb84-440"><a href="#cb84-440" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb84-441"><a href="#cb84-441" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-442"><a href="#cb84-442" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/截屏2024-01-10%2017.43.49.png)</span></span>
<span id="cb84-443"><a href="#cb84-443" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-446"><a href="#cb84-446" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb84-447"><a href="#cb84-447" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb84-448"><a href="#cb84-448" aria-hidden="true" tabindex="-1"></a><span class="co"># 将第一列（基因名）作为行名</span></span>
<span id="cb84-449"><a href="#cb84-449" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(counts) <span class="ot">&lt;-</span> counts[,<span class="dv">1</span>]</span>
<span id="cb84-450"><a href="#cb84-450" aria-hidden="true" tabindex="-1"></a>counts <span class="ot">&lt;-</span> counts[,<span class="sc">-</span><span class="dv">1</span>]</span>
<span id="cb84-451"><a href="#cb84-451" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-452"><a href="#cb84-452" aria-hidden="true" tabindex="-1"></a><span class="co"># 添加列名（cell barcode）</span></span>
<span id="cb84-453"><a href="#cb84-453" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(counts) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(head)</span>
<span id="cb84-454"><a href="#cb84-454" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb84-455"><a href="#cb84-455" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-456"><a href="#cb84-456" aria-hidden="true" tabindex="-1"></a>!<span class="co">[</span><span class="ot">此时，我们就得到了一个满足Seurat对象构建的counts矩阵了。但是，当我们直接用`CreateSeuratObject()`去创建Seurat对象时，由于该数据包含了所有样本的单细胞测序数据，导致其体积过大超出了R的内存上限。因此，接下来，我们**将这个数据根据样本拆分成不同的RDS文件**，然后再用[上面](#sec-read_raw_counts)类似的方法依次读取这些RDS文件，避免一次性读取导致的内存溢出。</span><span class="co">](images/截屏2024-01-10%2017.45.43.png)</span></span>
<span id="cb84-457"><a href="#cb84-457" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-460"><a href="#cb84-460" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb84-461"><a href="#cb84-461" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb84-462"><a href="#cb84-462" aria-hidden="true" tabindex="-1"></a><span class="co"># 提取样本标签</span></span>
<span id="cb84-463"><a href="#cb84-463" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span>  <span class="fu">str_split</span>(<span class="at">string =</span> <span class="fu">colnames</span>(counts), </span>
<span id="cb84-464"><a href="#cb84-464" aria-hidden="true" tabindex="-1"></a>                      <span class="at">pattern =</span> <span class="st">"_"</span>, </span>
<span id="cb84-465"><a href="#cb84-465" aria-hidden="true" tabindex="-1"></a>                      <span class="at">simplify =</span> T)[,<span class="dv">1</span>] <span class="sc">%&gt;%</span> </span>
<span id="cb84-466"><a href="#cb84-466" aria-hidden="true" tabindex="-1"></a>  <span class="fu">unique</span>()</span>
<span id="cb84-467"><a href="#cb84-467" aria-hidden="true" tabindex="-1"></a>samples</span>
<span id="cb84-468"><a href="#cb84-468" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb84-469"><a href="#cb84-469" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-472"><a href="#cb84-472" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb84-473"><a href="#cb84-473" aria-hidden="true" tabindex="-1"></a><span class="co"># [1] "N01" "P01" "P03" "P05" "P06" "P09" "P11" "P12" "P13" "P14" "P15" "P02"</span></span>
<span id="cb84-474"><a href="#cb84-474" aria-hidden="true" tabindex="-1"></a><span class="co"># [13] "P04" "P07" "P08" "P10"</span></span>
<span id="cb84-475"><a href="#cb84-475" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb84-476"><a href="#cb84-476" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-479"><a href="#cb84-479" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb84-480"><a href="#cb84-480" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb84-481"><a href="#cb84-481" aria-hidden="true" tabindex="-1"></a><span class="co"># 逐一提取不同样本的测序数据，分别导出为RDS文件</span></span>
<span id="cb84-482"><a href="#cb84-482" aria-hidden="true" tabindex="-1"></a><span class="fu">map</span>(samples, <span class="cf">function</span>(sample) {</span>
<span id="cb84-483"><a href="#cb84-483" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(sample)</span>
<span id="cb84-484"><a href="#cb84-484" aria-hidden="true" tabindex="-1"></a>  sub_exp <span class="ot">&lt;-</span> <span class="fu">grep</span>(<span class="at">pattern =</span> sample, </span>
<span id="cb84-485"><a href="#cb84-485" aria-hidden="true" tabindex="-1"></a>                  <span class="at">x =</span> <span class="fu">colnames</span>(counts)) <span class="sc">%&gt;%</span> </span>
<span id="cb84-486"><a href="#cb84-486" aria-hidden="true" tabindex="-1"></a>    counts[,.]</span>
<span id="cb84-487"><a href="#cb84-487" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(sub_exp, </span>
<span id="cb84-488"><a href="#cb84-488" aria-hidden="true" tabindex="-1"></a>          <span class="at">file =</span> <span class="fu">paste0</span>(<span class="st">"output/sc_supplementary/GSE150430/"</span>, sample, <span class="st">".rds"</span>))</span>
<span id="cb84-489"><a href="#cb84-489" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb84-490"><a href="#cb84-490" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb84-491"><a href="#cb84-491" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-492"><a href="#cb84-492" aria-hidden="true" tabindex="-1"></a><span class="fu">## 构建Seurat对象</span></span>
<span id="cb84-493"><a href="#cb84-493" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-496"><a href="#cb84-496" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb84-497"><a href="#cb84-497" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb84-498"><a href="#cb84-498" aria-hidden="true" tabindex="-1"></a>filenames <span class="ot">&lt;-</span> <span class="fu">list.files</span>(<span class="st">"output/sc_supplementary/GSE150430"</span>)</span>
<span id="cb84-499"><a href="#cb84-499" aria-hidden="true" tabindex="-1"></a>filenames</span>
<span id="cb84-500"><a href="#cb84-500" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb84-501"><a href="#cb84-501" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-504"><a href="#cb84-504" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb84-505"><a href="#cb84-505" aria-hidden="true" tabindex="-1"></a><span class="co">#  [1] "N01.rds" "P01.rds" "P02.rds" "P03.rds" "P04.rds" "P05.rds" "P06.rds"</span></span>
<span id="cb84-506"><a href="#cb84-506" aria-hidden="true" tabindex="-1"></a><span class="co">#  [8] "P07.rds" "P08.rds" "P09.rds" "P10.rds" "P11.rds" "P12.rds" "P13.rds"</span></span>
<span id="cb84-507"><a href="#cb84-507" aria-hidden="true" tabindex="-1"></a><span class="co"># [15] "P14.rds" "P15.rds"</span></span>
<span id="cb84-508"><a href="#cb84-508" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb84-509"><a href="#cb84-509" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-512"><a href="#cb84-512" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb84-513"><a href="#cb84-513" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb84-514"><a href="#cb84-514" aria-hidden="true" tabindex="-1"></a><span class="co"># 循环读取每个rds文件的数据并创建Seurat对象</span></span>
<span id="cb84-515"><a href="#cb84-515" aria-hidden="true" tabindex="-1"></a>seurat_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(filenames, <span class="cf">function</span>(file) {</span>
<span id="cb84-516"><a href="#cb84-516" aria-hidden="true" tabindex="-1"></a>  <span class="co"># 拼接文件路径</span></span>
<span id="cb84-517"><a href="#cb84-517" aria-hidden="true" tabindex="-1"></a>  data.path <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"output/sc_supplementary/GSE150430/"</span>, file)</span>
<span id="cb84-518"><a href="#cb84-518" aria-hidden="true" tabindex="-1"></a>  <span class="co"># 创建Seurat对象</span></span>
<span id="cb84-519"><a href="#cb84-519" aria-hidden="true" tabindex="-1"></a>  seurat_obj <span class="ot">&lt;-</span> <span class="fu">CreateSeuratObject</span>(<span class="at">counts =</span> <span class="fu">readRDS</span>(data.path))</span>
<span id="cb84-520"><a href="#cb84-520" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(seurat_obj)</span>
<span id="cb84-521"><a href="#cb84-521" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb84-522"><a href="#cb84-522" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-523"><a href="#cb84-523" aria-hidden="true" tabindex="-1"></a><span class="co"># 合并Seurat对象</span></span>
<span id="cb84-524"><a href="#cb84-524" aria-hidden="true" tabindex="-1"></a>merged_seurat <span class="ot">&lt;-</span> <span class="fu">merge</span>(seurat_list[[<span class="dv">1</span>]],</span>
<span id="cb84-525"><a href="#cb84-525" aria-hidden="true" tabindex="-1"></a>                       <span class="at">y =</span> seurat_list[<span class="sc">-</span><span class="dv">1</span>])</span>
<span id="cb84-526"><a href="#cb84-526" aria-hidden="true" tabindex="-1"></a>merged_seurat</span>
<span id="cb84-527"><a href="#cb84-527" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb84-528"><a href="#cb84-528" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-531"><a href="#cb84-531" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb84-532"><a href="#cb84-532" aria-hidden="true" tabindex="-1"></a><span class="co"># An object of class Seurat </span></span>
<span id="cb84-533"><a href="#cb84-533" aria-hidden="true" tabindex="-1"></a><span class="co"># 24720 features across 48584 samples within 1 assay </span></span>
<span id="cb84-534"><a href="#cb84-534" aria-hidden="true" tabindex="-1"></a><span class="co"># Active assay: RNA (24720 features, 0 variable features)</span></span>
<span id="cb84-535"><a href="#cb84-535" aria-hidden="true" tabindex="-1"></a><span class="co">#  16 layers present: counts.1, counts.2, counts.3, counts.4, counts.5, counts.6, counts.7, counts.8, counts.9, counts.10, counts.11, counts.12, counts.13, counts.14, counts.15, counts.16</span></span>
<span id="cb84-536"><a href="#cb84-536" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb84-537"><a href="#cb84-537" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-540"><a href="#cb84-540" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb84-541"><a href="#cb84-541" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb84-542"><a href="#cb84-542" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(merged_seurat) </span>
<span id="cb84-543"><a href="#cb84-543" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb84-544"><a href="#cb84-544" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-545"><a href="#cb84-545" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/截屏2024-01-11%2020.36.33.png)</span>{width="434"}</span>
<span id="cb84-546"><a href="#cb84-546" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-549"><a href="#cb84-549" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb84-550"><a href="#cb84-550" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb84-551"><a href="#cb84-551" aria-hidden="true" tabindex="-1"></a><span class="co"># 添加样本列</span></span>
<span id="cb84-552"><a href="#cb84-552" aria-hidden="true" tabindex="-1"></a>merged_seurat<span class="sc">$</span>samples <span class="ot">&lt;-</span> merged_seurat<span class="sc">$</span>orig.ident</span>
<span id="cb84-553"><a href="#cb84-553" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-554"><a href="#cb84-554" aria-hidden="true" tabindex="-1"></a><span class="co"># 保存</span></span>
<span id="cb84-555"><a href="#cb84-555" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(merged_seurat, <span class="at">file =</span> <span class="st">"output/sc_supplementary/GSE150430_merged_seurat.rds"</span>)</span>
<span id="cb84-556"><a href="#cb84-556" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb84-557"><a href="#cb84-557" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-558"><a href="#cb84-558" aria-hidden="true" tabindex="-1"></a>------------------------------------------------------------------------</span>
<span id="cb84-559"><a href="#cb84-559" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-560"><a href="#cb84-560" aria-hidden="true" tabindex="-1"></a><span class="fu"># 读取SMART-seq测序数据</span></span>
<span id="cb84-561"><a href="#cb84-561" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-562"><a href="#cb84-562" aria-hidden="true" tabindex="-1"></a>数据链接：<span class="co">[</span><span class="ot">**GSE140228**</span><span class="co">](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE140228)</span>**。**</span>
<span id="cb84-563"><a href="#cb84-563" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-564"><a href="#cb84-564" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/截屏2024-01-18%2014.41.29.png)</span></span>
<span id="cb84-565"><a href="#cb84-565" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-566"><a href="#cb84-566" aria-hidden="true" tabindex="-1"></a>这个数据集用了两种测序技术，即Drop-seq和SMART-seq，分别属于测**3′ 或5′ 端（3′或5′-end sequencing）**的技术和测**全长转录本（full-length transcript sequencing）**的技术。这里我们下载SMART-seq测序数据。</span>
<span id="cb84-567"><a href="#cb84-567" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-568"><a href="#cb84-568" aria-hidden="true" tabindex="-1"></a>读取barcodes、features和matrix数据:</span>
<span id="cb84-569"><a href="#cb84-569" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-572"><a href="#cb84-572" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb84-573"><a href="#cb84-573" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb84-574"><a href="#cb84-574" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(<span class="at">list =</span> <span class="fu">ls</span>())</span>
<span id="cb84-575"><a href="#cb84-575" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb84-576"><a href="#cb84-576" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(data.table)</span>
<span id="cb84-577"><a href="#cb84-577" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-578"><a href="#cb84-578" aria-hidden="true" tabindex="-1"></a><span class="fu">list.files</span>(<span class="st">"data/sc_supplementary/GSE140228"</span>)</span>
<span id="cb84-579"><a href="#cb84-579" aria-hidden="true" tabindex="-1"></a>barcodes <span class="ot">&lt;-</span> <span class="fu">fread</span>(</span>
<span id="cb84-580"><a href="#cb84-580" aria-hidden="true" tabindex="-1"></a>  <span class="st">"data/sc_supplementary/GSE140228/GSE140228_cell_info_Smartseq2.tsv.gz"</span>,</span>
<span id="cb84-581"><a href="#cb84-581" aria-hidden="true" tabindex="-1"></a>  <span class="at">data.table =</span> F <span class="co"># 返回data.frame型数据</span></span>
<span id="cb84-582"><a href="#cb84-582" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb84-583"><a href="#cb84-583" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(barcodes, <span class="dv">5</span>)</span>
<span id="cb84-584"><a href="#cb84-584" aria-hidden="true" tabindex="-1"></a>features <span class="ot">&lt;-</span> <span class="fu">fread</span>(</span>
<span id="cb84-585"><a href="#cb84-585" aria-hidden="true" tabindex="-1"></a>  <span class="st">"data/sc_supplementary/GSE140228/GSE140228_gene_info_Smartseq2.tsv.gz"</span>,</span>
<span id="cb84-586"><a href="#cb84-586" aria-hidden="true" tabindex="-1"></a>  <span class="at">data.table =</span> F</span>
<span id="cb84-587"><a href="#cb84-587" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb84-588"><a href="#cb84-588" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(features, <span class="dv">5</span>)</span>
<span id="cb84-589"><a href="#cb84-589" aria-hidden="true" tabindex="-1"></a>matrix <span class="ot">&lt;-</span> <span class="fu">fread</span>(</span>
<span id="cb84-590"><a href="#cb84-590" aria-hidden="true" tabindex="-1"></a>  <span class="st">"data/sc_supplementary/GSE140228/GSE140228_read_counts_Smartseq2.csv.gz"</span>,</span>
<span id="cb84-591"><a href="#cb84-591" aria-hidden="true" tabindex="-1"></a>  <span class="at">data.table =</span> F,</span>
<span id="cb84-592"><a href="#cb84-592" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb84-593"><a href="#cb84-593" aria-hidden="true" tabindex="-1"></a>matrix[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span>
<span id="cb84-594"><a href="#cb84-594" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb84-595"><a href="#cb84-595" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-596"><a href="#cb84-596" aria-hidden="true" tabindex="-1"></a>这个<span class="in">`matrix`</span>第一列为基因名，列名为barcodes。<span class="in">`features`</span>和<span class="in">`barcodes`</span>是对每个基因和细胞的补充信息。所以，接下来我们可以直接用这个<span class="in">`matrix`</span>文件构建counts矩阵。后续如有需要可以将<span class="in">`features`</span>和<span class="in">`barcodes`</span>中的信息补充上去。</span>
<span id="cb84-597"><a href="#cb84-597" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-600"><a href="#cb84-600" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb84-601"><a href="#cb84-601" aria-hidden="true" tabindex="-1"></a><span class="co">#| cache-lazy: false</span></span>
<span id="cb84-602"><a href="#cb84-602" aria-hidden="true" tabindex="-1"></a><span class="co"># 构建counts矩阵</span></span>
<span id="cb84-603"><a href="#cb84-603" aria-hidden="true" tabindex="-1"></a>counts <span class="ot">&lt;-</span> matrix</span>
<span id="cb84-604"><a href="#cb84-604" aria-hidden="true" tabindex="-1"></a><span class="fu">row.names</span>(counts) <span class="ot">&lt;-</span> counts<span class="sc">$</span>gene</span>
<span id="cb84-605"><a href="#cb84-605" aria-hidden="true" tabindex="-1"></a>counts<span class="sc">$</span>gene <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb84-606"><a href="#cb84-606" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-607"><a href="#cb84-607" aria-hidden="true" tabindex="-1"></a><span class="co"># 创建Seurat对象</span></span>
<span id="cb84-608"><a href="#cb84-608" aria-hidden="true" tabindex="-1"></a>merged_seurat <span class="ot">&lt;-</span> <span class="fu">CreateSeuratObject</span>(</span>
<span id="cb84-609"><a href="#cb84-609" aria-hidden="true" tabindex="-1"></a>  <span class="at">counts =</span> counts,</span>
<span id="cb84-610"><a href="#cb84-610" aria-hidden="true" tabindex="-1"></a>  <span class="at">project =</span> <span class="st">"HCC"</span>,</span>
<span id="cb84-611"><a href="#cb84-611" aria-hidden="true" tabindex="-1"></a>  <span class="at">min.cells =</span> <span class="dv">3</span>,</span>
<span id="cb84-612"><a href="#cb84-612" aria-hidden="true" tabindex="-1"></a>  <span class="at">min.features =</span> <span class="dv">200</span></span>
<span id="cb84-613"><a href="#cb84-613" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb84-614"><a href="#cb84-614" aria-hidden="true" tabindex="-1"></a>merged_seurat</span>
<span id="cb84-615"><a href="#cb84-615" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(merged_seurat, <span class="dv">5</span>)</span>
<span id="cb84-616"><a href="#cb84-616" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb84-617"><a href="#cb84-617" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-618"><a href="#cb84-618" aria-hidden="true" tabindex="-1"></a>------------------------------------------------------------------------</span>
<span id="cb84-619"><a href="#cb84-619" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-620"><a href="#cb84-620" aria-hidden="true" tabindex="-1"></a><span class="fu"># 单独提供每个细胞测序结果的数据的读取</span></span>
<span id="cb84-621"><a href="#cb84-621" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-622"><a href="#cb84-622" aria-hidden="true" tabindex="-1"></a>数据链接：<span class="co">[</span><span class="ot">**GSE84465**</span><span class="co">](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE84465)</span>**。**</span>
<span id="cb84-623"><a href="#cb84-623" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-624"><a href="#cb84-624" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/截屏2024-01-18%2015.47.01.png)</span>{width="418"}<span class="al">![](images/截屏2024-01-18%2015.47.10.png)</span>{width="591"}</span>
<span id="cb84-625"><a href="#cb84-625" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-626"><a href="#cb84-626" aria-hidden="true" tabindex="-1"></a>可以看到，该数据集提供每一个细胞的测序结果。对于这样的数据，可以参考<span class="co">[</span><span class="ot">*单细胞数据下载和Seurat对象的构建（上）*</span><span class="co">](https://mp.weixin.qq.com/s?__biz=Mzg2MTExNTkwNA==&amp;mid=2247510395&amp;idx=1&amp;sn=f2a2c0e3d2d75656f1dcfceaf708beb8&amp;chksm=ce1eea14f969630295e8988209df55b61e2062178d43b22d9eb41a93badf23772a312830ab49&amp;mpshare=1&amp;scene=1&amp;srcid=1117TnC4m9nO1EleC5e8JIOl&amp;sharer_sharetime=1668645592942&amp;sharer_shareid=568f49cb24905bf546294e6985a632bf#rd)</span>中的方法来读取。</span>
<span id="cb84-627"><a href="#cb84-627" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-628"><a href="#cb84-628" aria-hidden="true" tabindex="-1"></a>------------------------------------------------------------------------</span>
<span id="cb84-629"><a href="#cb84-629" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-630"><a href="#cb84-630" aria-hidden="true" tabindex="-1"></a>::: {.callout-note collapse="true" icon="false"}</span>
<span id="cb84-631"><a href="#cb84-631" aria-hidden="true" tabindex="-1"></a><span class="fu">## Session Info</span></span>
<span id="cb84-632"><a href="#cb84-632" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-635"><a href="#cb84-635" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb84-636"><a href="#cb84-636" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb84-637"><a href="#cb84-637" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span>
<span id="cb84-638"><a href="#cb84-638" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb84-639"><a href="#cb84-639" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb84-640"><a href="#cb84-640" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-641"><a href="#cb84-641" aria-hidden="true" tabindex="-1"></a><span class="fu"># References {.unnumbered}</span></span>
<span id="cb84-642"><a href="#cb84-642" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-643"><a href="#cb84-643" aria-hidden="true" tabindex="-1"></a>::: {#refs}</span>
<span id="cb84-644"><a href="#cb84-644" aria-hidden="true" tabindex="-1"></a>:::</span>
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">
      <ul class="footer-items list-unstyled">
<li class="nav-item">
 This website was built with Quarto
  </li>  
</ul>
</div>   
    <div class="nav-footer-center">
<p>Copyright 2024, Du Junhong</p>
<div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/djhcod/r-notes/blob/main/single_cell/sc_supplementary/read_sc_data.qmd" class="toc-action"><i class="bi bi-github"></i>View source</a></li><li><a href="https://github.com/djhcod/r-notes/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></div>
    <div class="nav-footer-right">
      <ul class="footer-items list-unstyled">
<li class="nav-item compact">
    <a class="nav-link" href="https://github.com/djhcod">
      <i class="bi bi-github" role="img">
</i> 
    </a>
  </li>  
</ul>
</div>
  </div>
</footer>


</body></html>