<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.433">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>R语言学习笔记 - 8&nbsp; 单细胞数据整合（integration）</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../single_cell/integrative_analysis_in_seurat_v5.html" rel="next">
<link href="../single_cell/sctransform.html" rel="prev">
<link href="../images/logo.jpg" rel="icon" type="image/jpeg">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark"><script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><script type="application/json" class="js-hypothesis-config">
{
  "theme": "classic",
  "openSidebar": false,
  "showHighlights": "always",
  "enableExperimentalNewNoteButton": true
}
</script><script async="" src="https://hypothes.is/embed.js"></script>
<meta name="twitter:title" content="R语言学习笔记 - 8&nbsp; 单细胞数据整合（integration）">
<meta name="twitter:description" content="">
<meta name="twitter:image" content="https://djhcod.github.io/r-notes/single_cell/images/截屏2023-11-27 12.49.37.png">
<meta name="twitter:image-height" content="330">
<meta name="twitter:image-width" content="1732">
<meta name="twitter:card" content="summary_large_image">
</head>
<body class="nav-sidebar docked nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="navbar navbar-expand-lg navbar-dark "><div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a href="../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../images/logo.jpg" alt="" class="navbar-logo"></a>
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">R语言学习笔记</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
<li class="nav-item">
    <a class="nav-link active" href="../index.html" rel="" target="" aria-current="page">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../intro.html" rel="" target="">
 <span class="menu-text">Introduction</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-parts" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Parts</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-parts">
<li>
    <a class="dropdown-item" href="../r_basic/r_basics.html" rel="" target="">
 <span class="dropdown-text">R语言基础</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../single_cell/seurat.html" rel="" target="">
 <span class="dropdown-text">Seurat v5官方文档</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../quarto_foundation/quarto_foundation.html" rel="" target="">
 <span class="dropdown-text">Quarto基础</span></a>
  </li>  
    </ul>
</li>
  <li class="nav-item">
    <a class="nav-link" href="../references.html" rel="" target="">
 <span class="menu-text">References</span></a>
  </li>  
</ul>
<ul class="navbar-nav navbar-nav-scroll ms-auto">
<li class="nav-item">
    <a class="nav-link" href="https://bookdown.org" rel="" target=""><i class="bi bi-book-fill" role="img">
</i> 
 <span class="menu-text">Bookdown</span></a>
  </li>  
</ul>
<div class="quarto-navbar-tools tools-wide">
    <a href="https://github.com/djhcod/r-notes" rel="" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
    <div class="dropdown">
      <a href="" title="Share" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Share"><i class="bi bi-share"></i></a>
      <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="quarto-navigation-tool-dropdown-0">
<li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://twitter.com/intent/tweet?url=%7Curl%7C">
              <i class="bi bi-bi-twitter pe-1"></i>
            Twitter
            </a>
          </li>
          <li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://www.facebook.com/sharer/sharer.php?u=%7Curl%7C">
              <i class="bi bi-bi-facebook pe-1"></i>
            Facebook
            </a>
          </li>
      </ul>
</div>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../single_cell/seurat.html">Seurat v5官方文档</a></li><li class="breadcrumb-item"><a href="../single_cell/integration.html"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">单细胞数据整合（integration）</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto"><div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">主页</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">前言</span></a>
  </div>
</li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../r_basic/r_basics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">R语言基础</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../r_basic/environment_configuration.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Rstudio环境配置</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../r_basic/data_input_output.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">数据的读取与输出</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../single_cell/seurat.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Seurat v5官方文档</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../single_cell/seurat_command_list.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Seurat常用函数清单</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../single_cell/seurat_tutorial.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Seurat细胞分群官方教程</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../single_cell/data_visualization_methods_in_seurat.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Seurat中的数据可视化方法</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../single_cell/sctransform.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">基于SCTransform的单细胞数据标准化</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../single_cell/integration.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">单细胞数据整合（integration）</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../single_cell/integrative_analysis_in_seurat_v5.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Seurat v5单细胞数据整合分析</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../single_cell/mapping_and_annotating_query_datasets.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">基于参考集的细胞注释</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../single_cell/cell_cycle_regression.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">消除细胞周期的影响</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../single_cell/de_vignette.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">差异表达分析</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="../quarto_foundation/quarto_foundation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Quarto基础</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../quarto_foundation/yaml_settings.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">YAML设置</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../quarto_foundation/images_settings.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">图片设置</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../quarto_foundation/cross_references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">交叉引用</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../quarto_foundation/insert_other_content.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">插入其他内容</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../quarto_foundation/quarto_books.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">Quarto Books</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../quarto_foundation/github_pages.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">发布到GitHub Pages</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="false">
 <span class="menu-text">参考文献</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 ">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active" data-toc-expanded="1"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#%E6%95%B0%E6%8D%AE%E8%AF%BB%E5%8F%96%E5%92%8C%E5%88%86%E5%B1%82" id="toc-数据读取和分层" class="nav-link active" data-scroll-target="#%E6%95%B0%E6%8D%AE%E8%AF%BB%E5%8F%96%E5%92%8C%E5%88%86%E5%B1%82"><span class="header-section-number">8.1</span> 数据读取和分层</a></li>
  <li><a href="#%E4%B8%8D%E8%BF%9B%E8%A1%8C%E6%95%B4%E5%90%88%E7%9A%84%E6%83%85%E5%86%B5%E4%B8%8B%E7%9A%84%E6%95%B0%E6%8D%AE%E5%A4%84%E7%90%86" id="toc-不进行整合的情况下的数据处理" class="nav-link" data-scroll-target="#%E4%B8%8D%E8%BF%9B%E8%A1%8C%E6%95%B4%E5%90%88%E7%9A%84%E6%83%85%E5%86%B5%E4%B8%8B%E7%9A%84%E6%95%B0%E6%8D%AE%E5%A4%84%E7%90%86"><span class="header-section-number">8.2</span> 不进行整合的情况下的数据处理</a></li>
  <li>
<a href="#%E8%BF%9B%E8%A1%8C%E6%95%B0%E6%8D%AE%E6%95%B4%E5%90%88" id="toc-进行数据整合" class="nav-link" data-scroll-target="#%E8%BF%9B%E8%A1%8C%E6%95%B0%E6%8D%AE%E6%95%B4%E5%90%88"><span class="header-section-number">8.3</span> 进行数据整合</a>
  <ul class="collapse">
<li><a href="#%E6%95%B4%E5%90%88%E5%90%8E%E9%87%8D%E6%96%B0%E8%81%9A%E7%B1%BB%E9%99%8D%E7%BB%B4" id="toc-整合后重新聚类降维" class="nav-link" data-scroll-target="#%E6%95%B4%E5%90%88%E5%90%8E%E9%87%8D%E6%96%B0%E8%81%9A%E7%B1%BB%E9%99%8D%E7%BB%B4">整合后重新聚类、降维</a></li>
  </ul>
</li>
  <li>
<a href="#%E9%89%B4%E5%AE%9A%E4%BF%9D%E5%AE%88%E7%9A%84cell-marker" id="toc-鉴定保守的cell-marker" class="nav-link" data-scroll-target="#%E9%89%B4%E5%AE%9A%E4%BF%9D%E5%AE%88%E7%9A%84cell-marker"><span class="header-section-number">8.4</span> 鉴定保守的cell marker</a>
  <ul class="collapse">
<li><a href="#%E5%8F%AF%E8%A7%86%E5%8C%96cell-markers%E7%9A%84%E8%A1%A8%E8%BE%BE" id="toc-可视化cell-markers的表达" class="nav-link" data-scroll-target="#%E5%8F%AF%E8%A7%86%E5%8C%96cell-markers%E7%9A%84%E8%A1%A8%E8%BE%BE">可视化cell markers的表达</a></li>
  </ul>
</li>
  <li>
<a href="#sec-%E8%AF%86%E5%88%AB%E4%B8%8D%E5%90%8C%E6%A0%B7%E6%9C%AC%E7%B1%BB%E5%9E%8B%E9%97%B4%E7%9A%84%E5%B7%AE%E5%BC%82%E5%9F%BA%E5%9B%A0" id="toc-sec-识别不同样本类型间的差异基因" class="nav-link" data-scroll-target="#sec-%E8%AF%86%E5%88%AB%E4%B8%8D%E5%90%8C%E6%A0%B7%E6%9C%AC%E7%B1%BB%E5%9E%8B%E9%97%B4%E7%9A%84%E5%B7%AE%E5%BC%82%E5%9F%BA%E5%9B%A0"><span class="header-section-number">8.5</span> 识别不同样本类型间的差异基因</a>
  <ul class="collapse">
<li><a href="#sec-%E6%AD%A3%E5%BC%8F%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90" id="toc-sec-正式差异分析" class="nav-link" data-scroll-target="#sec-%E6%AD%A3%E5%BC%8F%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90">正式差异分析</a></li>
  </ul>
</li>
  <li>
<a href="#%E6%89%A7%E8%A1%8Csctransform%E6%A0%87%E5%87%86%E5%8C%96%E6%B5%81%E7%A8%8B%E4%B9%8B%E5%90%8E%E7%9A%84%E6%95%B4%E5%90%88" id="toc-执行sctransform标准化流程之后的整合" class="nav-link" data-scroll-target="#%E6%89%A7%E8%A1%8Csctransform%E6%A0%87%E5%87%86%E5%8C%96%E6%B5%81%E7%A8%8B%E4%B9%8B%E5%90%8E%E7%9A%84%E6%95%B4%E5%90%88"><span class="header-section-number">8.6</span> 执行<code>SCTransform</code>标准化流程之后的整合</a>
  <ul class="collapse">
<li><a href="#%E4%B8%8D%E8%BF%9B%E8%A1%8C%E6%95%B4%E5%90%88%E7%9A%84%E6%83%85%E5%86%B5%E4%B8%8B%E7%9A%84%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90" id="toc-不进行整合的情况下的数据分析" class="nav-link" data-scroll-target="#%E4%B8%8D%E8%BF%9B%E8%A1%8C%E6%95%B4%E5%90%88%E7%9A%84%E6%83%85%E5%86%B5%E4%B8%8B%E7%9A%84%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90">不进行整合的情况下的数据分析</a></li>
  <li><a href="#sec-integration_after_sct" id="toc-sec-integration_after_sct" class="nav-link" data-scroll-target="#sec-integration_after_sct">进行整合</a></li>
  <li><a href="#%E6%95%B4%E5%90%88%E5%90%8E%E8%81%9A%E7%B1%BB" id="toc-整合后聚类" class="nav-link" data-scroll-target="#%E6%95%B4%E5%90%88%E5%90%8E%E8%81%9A%E7%B1%BB">整合后聚类</a></li>
  <li><a href="#%E5%B7%AE%E5%BC%82%E8%A1%A8%E8%BE%BE%E5%88%86%E6%9E%90" id="toc-差异表达分析" class="nav-link" data-scroll-target="#%E5%B7%AE%E5%BC%82%E8%A1%A8%E8%BE%BE%E5%88%86%E6%9E%90">差异表达分析</a></li>
  </ul>
</li>
  </ul><div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/djhcod/r-notes/blob/main/single_cell/integration.qmd" class="toc-action">View source</a></p><p><a href="https://github.com/djhcod/r-notes/issues/new" class="toc-action">Report an issue</a></p><p><a href="https://github.com/djhcod/r-notes/edit/main/single_cell/integration.qmd" class="toc-action">Edit this page</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title"><span id="sec-单细胞数据整合" class="quarto-section-identifier"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">单细胞数据整合（integration）</span></span></h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><p>原文：<a href="https://satijalab.org/seurat/articles/integration_introduction"><em>Introduction to scRNA-seq integration</em></a></p>
<p>原文发布日期：2023年10月31日</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>本篇主要介绍来自不同样本类型的单细胞数据的整合。对于如何整合不同测序技术的单细胞数据集，参考Seurat官方文档：<a href="https://satijalab.org/seurat/articles/seurat5_integration"><em>Integrative analysis in Seurat v5</em></a>。</p>
</div>
</div>
<p>Integration of single-cell sequencing datasets, for example across experimental batches, donors, or conditions, is often an important step in scRNA-seq workflows. Integrative analysis can help to <strong>match shared cell types and states across datasets</strong>, which can boost statistical power, and most importantly, facilitate accurate comparative analysis across datasets. In previous versions of Seurat we introduced methods for integrative analysis, including our ‘anchor-based’ integration workflow. Many labs have also published powerful and pioneering methods, including&nbsp;<a href="https://portals.broadinstitute.org/harmony/">Harmony</a>&nbsp;and&nbsp;<a href="https://docs.scvi-tools.org/en/stable/index.html">scVI</a>, for integrative analysis.</p>
<p>数据整合的目标：</p>
<p>The following tutorial is designed to give you an overview of the kinds of comparative analyses on complex cell types that are possible using the Seurat integration procedure. Here, we address a few key goals:</p>
<ul>
<li><p>Identify <strong>cell subpopulations</strong> that are present in both datasets</p></li>
<li><p>Obtain <strong>cell type markers</strong> that are conserved in both control and stimulated cells</p></li>
<li><p><strong>Compare the datasets</strong> to find cell-type specific responses to stimulation</p></li>
</ul>
<section id="数据读取和分层" class="level2" data-number="8.1"><h2 data-number="8.1" class="anchored" data-anchor-id="数据读取和分层">
<span class="header-section-number">8.1</span> 数据读取和分层</h2>
<div class="callout callout-style-simple callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-2-contents" aria-controls="callout-2" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
在线读取（可能需要全局代理）
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-2" class="callout-2-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-hash="integration_cache/html/unnamed-chunk-1_4259f900996d24c302bead7d4bd5c2b3">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SeuratData)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">InstallData</span>(<span class="st">"ifnb"</span>)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>ifnb <span class="ot">&lt;-</span> <span class="fu">LoadData</span>(<span class="st">"ifnb"</span>)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<p>从本地下载好的数据读取：</p>
<div class="cell" data-hash="integration_cache/html/unnamed-chunk-2_df8775f921397a3373e989a627d65007">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="st">"data/pbmc_ifnb.rds"</span><span class="op">)</span></span>
<span><span class="va">ifnb</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
14053 features across 13999 samples within 1 assay 
Active assay: RNA (14053 features, 0 variable features)
 2 layers present: counts, data</code></pre>
</div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">ifnb</span><span class="op">@</span><span class="va">meta.data</span>, <span class="fl">5</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                  orig.ident nCount_RNA nFeature_RNA stim seurat_annotations
AAACATACATTTCC.1 IMMUNE_CTRL       3017          877 CTRL          CD14 Mono
AAACATACCAGAAA.1 IMMUNE_CTRL       2481          713 CTRL          CD14 Mono
AAACATACCTCGCT.1 IMMUNE_CTRL       3420          850 CTRL          CD14 Mono
AAACATACCTGGTA.1 IMMUNE_CTRL       3156         1109 CTRL                pDC
AAACATACGATGAA.1 IMMUNE_CTRL       1868          634 CTRL       CD4 Memory T</code></pre>
</div>
</div>
<p><img src="images/截屏2023-11-27 12.49.37.png" class="img-fluid"></p>
<p>The object contains data from human PBMC from two conditions, <strong>interferon-stimulated</strong> and <strong>control</strong> cells (stored in the&nbsp;<code>stim</code> column in the object metadata). We will aim to integrate the two conditions together, so that we can jointly identify cell subpopulations across datasets, and then explore how each group differs across conditions</p>
<p>In previous versions of Seurat, we would require the data to be represented as two different Seurat objects. <strong>In Seurat v5, we keep all the data in one object, but simply split it into multiple ‘layers’</strong>. To learn more about layers, check out our&nbsp;<a href="https://satijalab.org/seurat/articles/interaction_vignette">Seurat object interaction vignette</a>.</p>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Important
</div>
</div>
<div class="callout-body-container callout-body">
<p>Seurat v5 assays store data in <strong>layers</strong>. These layers can store:</p>
<ul>
<li><p>raw, un-normalized counts (<code>layer='counts'</code>)</p></li>
<li><p>normalized data (<code>layer='data'</code>)</p></li>
<li><p>z-scored/variance-stabilized data (<code>layer='scale.data'</code>).</p></li>
</ul>
</div>
</div>
<p><strong>split the RNA measurements into two layers one for control cells, one for stimulated cells:</strong></p>
<div class="cell" data-hash="integration_cache/html/unnamed-chunk-3_cdf052a9c3657bcf3f255f18204908cd">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span></span>
<span><span class="va">ifnb</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html">split</a></span><span class="op">(</span><span class="va">ifnb</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span>, </span>
<span>                       f <span class="op">=</span> <span class="va">ifnb</span><span class="op">$</span><span class="va">stim</span><span class="op">)</span> <span class="co"># 按照meta.data中的“stim”列进行分割</span></span>
<span><span class="va">ifnb</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
14053 features across 13999 samples within 1 assay 
Active assay: RNA (14053 features, 0 variable features)
 4 layers present: counts.CTRL, counts.STIM, data.CTRL, data.STIM</code></pre>
</div>
</div>
<p>现在可以发现<code>ifnb</code>被分为了4个layer，此前是2个layer（<code>counts</code>和<code>data</code>）：</p>
<p><img src="images/截屏2023-11-27 11.55.48.png" class="img-fluid"></p>
</section><section id="不进行整合的情况下的数据处理" class="level2" data-number="8.2"><h2 data-number="8.2" class="anchored" data-anchor-id="不进行整合的情况下的数据处理">
<span class="header-section-number">8.2</span> 不进行整合的情况下的数据处理</h2>
<p>进行标准的数据处理流程：</p>
<div class="cell" data-hash="integration_cache/html/unnamed-chunk-4_217249d68077f9b90661a1346f559aed">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/NormalizeData.html">NormalizeData</a></span><span class="op">(</span><span class="va">ifnb</span><span class="op">)</span></span>
<span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindVariableFeatures.html">FindVariableFeatures</a></span><span class="op">(</span><span class="va">ifnb</span><span class="op">)</span></span>
<span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/ScaleData.html">ScaleData</a></span><span class="op">(</span><span class="va">ifnb</span><span class="op">)</span></span>
<span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/RunPCA.html">RunPCA</a></span><span class="op">(</span><span class="va">ifnb</span><span class="op">)</span></span>
<span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindNeighbors.html">FindNeighbors</a></span><span class="op">(</span><span class="va">ifnb</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>, reduction <span class="op">=</span> <span class="st">"pca"</span><span class="op">)</span></span>
<span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindClusters.html">FindClusters</a></span><span class="op">(</span><span class="va">ifnb</span>, </span>
<span>                     resolution <span class="op">=</span> <span class="fl">2</span>, </span>
<span>                     cluster.name <span class="op">=</span> <span class="st">"unintegrated_clusters"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 13999
Number of edges: 555146

Running Louvain algorithm...
Maximum modularity in 10 random starts: 0.8153
Number of communities: 26
Elapsed time: 1 seconds</code></pre>
</div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/RunUMAP.html">RunUMAP</a></span><span class="op">(</span><span class="va">ifnb</span>, </span>
<span>                dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>, </span>
<span>                reduction <span class="op">=</span> <span class="st">"pca"</span>, </span>
<span>                reduction.name <span class="op">=</span> <span class="st">"umap.unintegrated"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>分别按照<strong>样本分组（“stim”</strong>）<strong>和</strong>细胞聚类情况（“seurat_clusters”<strong>）</strong>着色绘制UMAP图：</p>
<div class="cell" data-hash="integration_cache/html/fig-未整合_01ef1bf20181b2d65d4ac058a26a015b">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">ifnb</span>, </span>
<span>        reduction <span class="op">=</span> <span class="st">"umap.unintegrated"</span>, </span>
<span>        group.by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"stim"</span>, <span class="st">"seurat_clusters"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-未整合" class="quarto-figure quarto-figure-left anchored">
<figure class="figure"><p><img src="integration_files/figure-html/fig-未整合-1.png" class="img-fluid figure-img" width="960"></p>
<figcaption class="figure-caption">Figure&nbsp;8.1: 未整合时的细胞分群情况（左：按照刺激条件着色；右：按照细胞聚类情况着色）</figcaption></figure>
</div>
</div>
</div>
<p>可以发现：The resulting clusters are defined <strong>both by cell type</strong> <strong>and stimulation condition</strong>, which creates challenges for downstream analysis.</p>
</section><section id="进行数据整合" class="level2" data-number="8.3"><h2 data-number="8.3" class="anchored" data-anchor-id="进行数据整合">
<span class="header-section-number">8.3</span> 进行数据整合</h2>
<p>We now aim to integrate data from the two conditions, so that cells from the same cell type/subpopulation will cluster together.</p>
<p>We often refer to this procedure as intergration/alignment. When aligning two genome sequences together, identification of shared/homologous regions can help to interpret differences between the sequences as well. Similarly for scRNA-seq integration, our goal is <strong>not to remove biological differences across conditions</strong>, <strong>but to learn shared cell types/states in an initial step-specifically</strong> because that will enable us to compare control stimulated and control profiles for these individual cell types.</p>
<p>The Seurat v5 integration procedure aims to return a single dimensional reduction that captures the shared sources of variance across multiple layers, so that cells in a similar biological state will cluster. The method returns a dimensional reduction (i.e.&nbsp;<code>integrated.cca</code>) which can be used for visualization and unsupervised clustering analysis. For evaluating performance, we can use cell type labels that are pre-loaded in the&nbsp;<code>seurat_annotations</code>&nbsp;metadata column.</p>
<div class="cell" data-hash="integration_cache/html/unnamed-chunk-6_e388dce510f11afc50da791a628099d2">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 整合，比较耗时间，进度条会一直显示0%直至运算完成</span></span>
<span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/IntegrateLayers.html">IntegrateLayers</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">ifnb</span>, </span>
<span>                        method <span class="op">=</span> <span class="va">CCAIntegration</span>, </span>
<span>                        orig.reduction <span class="op">=</span> <span class="st">"pca"</span>, </span>
<span>                        new.reduction <span class="op">=</span> <span class="st">"integrated.cca"</span>, <span class="co"># 整合后新的降维数据的名称</span></span>
<span>                        verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 整合后重新合并layer</span></span>
<span><span class="va">ifnb</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/SplitLayers.html">JoinLayers</a></span><span class="op">(</span><span class="va">ifnb</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>可以看到经过整合的Seurat对象的降维（“reduction”）中多出了整合后的降维（“integrated.cca”）：</p>
<p><img src="images/截屏2023-11-28 09.57.11.png" class="img-fluid"></p>
<section id="整合后重新聚类降维" class="level3"><h3 class="anchored" data-anchor-id="整合后重新聚类降维">整合后重新聚类、降维</h3>
<div class="cell" data-hash="integration_cache/html/fig-整合后_1c5afaaa91bf8e3aef304b71d286b505">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 重新聚类</span></span>
<span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindNeighbors.html">FindNeighbors</a></span><span class="op">(</span><span class="va">ifnb</span>, </span>
<span>                      reduction <span class="op">=</span> <span class="st">"integrated.cca"</span>, <span class="co">#更改降维来源为"integrated.cca"</span></span>
<span>                      dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span></span>
<span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindClusters.html">FindClusters</a></span><span class="op">(</span><span class="va">ifnb</span>, resolution <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 13999
Number of edges: 590406

Running Louvain algorithm...
Maximum modularity in 10 random starts: 0.8448
Number of communities: 18
Elapsed time: 1 seconds</code></pre>
</div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 重新降维</span></span>
<span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/RunUMAP.html">RunUMAP</a></span><span class="op">(</span><span class="va">ifnb</span>, </span>
<span>                dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>, </span>
<span>                reduction <span class="op">=</span> <span class="st">"integrated.cca"</span><span class="op">)</span> <span class="co">#更改降维来源为"integrated.cca"</span></span>
<span></span>
<span><span class="co"># Visualization：</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">ifnb</span>, </span>
<span>        reduction <span class="op">=</span> <span class="st">"umap"</span>, </span>
<span>        group.by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"stim"</span>, <span class="st">"seurat_annotations"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-整合后" class="quarto-figure quarto-figure-left anchored">
<figure class="figure"><p><img src="integration_files/figure-html/fig-整合后-1.png" class="img-fluid figure-img" width="960"></p>
<figcaption class="figure-caption">Figure&nbsp;8.2: 整合后的细胞分群情况（左：按照刺激条件着色；右：按照细胞聚类情况着色）</figcaption></figure>
</div>
</div>
</div>
<p>可以看到和 <a href="#fig-%E6%9C%AA%E6%95%B4%E5%90%88">Figure&nbsp;<span>8.1</span></a> 相比，在整合后，细胞就只按照细胞类型进行聚类了。</p>
<p>也可以按照刺激条件（“stim”）绘制分面图，分别展示刺激组和对照组的细胞分群情况：</p>
<div class="cell" data-hash="integration_cache/html/unnamed-chunk-8_de61e09ee1a74b4439e0139f4e70c939">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">ifnb</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, split.by <span class="op">=</span> <span class="st">"stim"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integration_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid" width="960"></p>
</div>
</div>
<p>可以看到，和上面的结论一致，两种条件下的细胞分群基本一致。</p>
</section></section><section id="鉴定保守的cell-marker" class="level2" data-number="8.4"><h2 data-number="8.4" class="anchored" data-anchor-id="鉴定保守的cell-marker">
<span class="header-section-number">8.4</span> 鉴定保守的cell marker</h2>
<p>To <strong>identify canonical cell type marker genes that are conserved across conditions</strong>, we provide the&nbsp;<code><a href="https://rdrr.io/pkg/Seurat/man/FindConservedMarkers.html">FindConservedMarkers()</a></code> function. This function performs <strong>differential gene expression testing</strong> for each dataset/group and combines the p-values using meta-analysis methods from the MetaDE R package. For example, <strong>we can calculated the genes that are conserved markers irrespective of stimulation condition in cluster 6 (NK cells).</strong></p>
<p><code>FindConservedMarkers</code>函数会调用<code>metap</code>包，<code>metap</code>包需要<code>multtest</code>包，所以需要先安装这两个依赖包：</p>
<div class="cell" data-hash="integration_cache/html/unnamed-chunk-9_da601d45ab67c89910aad22336a07da1">
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">'multtest'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">'metap'</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-hash="integration_cache/html/unnamed-chunk-10_0b97bc94d0d7d331dde44610087bcb23">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 这里的meta.data已经提前注释好了细胞类型（储存在"seurat_annotations"列中）。</span></span>
<span><span class="co"># 将细胞类型注释指定为"seurat_annotations"</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/Idents.html">Idents</a></span><span class="op">(</span><span class="va">ifnb</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"seurat_annotations"</span></span>
<span></span>
<span><span class="va">nk.markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindConservedMarkers.html">FindConservedMarkers</a></span><span class="op">(</span><span class="va">ifnb</span>, </span>
<span>                                   ident.1 <span class="op">=</span> <span class="st">"NK"</span>, </span>
<span>                                   grouping.var <span class="op">=</span> <span class="st">"stim"</span>, </span>
<span>                                   verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">nk.markers</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      CTRL_p_val CTRL_avg_log2FC CTRL_pct.1 CTRL_pct.2 CTRL_p_val_adj
GNLY           0        6.854586      0.943      0.046              0
NKG7           0        5.358881      0.953      0.085              0
GZMB           0        5.078135      0.839      0.044              0
CLIC3          0        5.765314      0.601      0.024              0
CTSW           0        5.307246      0.537      0.030              0
KLRD1          0        5.261553      0.507      0.019              0
      STIM_p_val STIM_avg_log2FC STIM_pct.1 STIM_pct.2 STIM_p_val_adj max_pval
GNLY           0        6.435910      0.956      0.059              0        0
NKG7           0        4.971397      0.950      0.081              0        0
GZMB           0        5.151924      0.897      0.060              0        0
CLIC3          0        5.505208      0.623      0.031              0        0
CTSW           0        5.240729      0.592      0.035              0        0
KLRD1          0        4.852457      0.555      0.027              0        0
      minimump_p_val
GNLY               0
NKG7               0
GZMB               0
CLIC3              0
CTSW               0
KLRD1              0</code></pre>
</div>
</div>
<p>在实际分析中，鉴定这些保守的cell marker主要用来辅助对cluster的注释：you can perform these same analysis on the unsupervised clustering results (stored in&nbsp;<code>seurat_clusters</code>), and <strong>use these conserved markers to annotate cell types in your dataset</strong>.</p>
<section id="可视化cell-markers的表达" class="level3"><h3 class="anchored" data-anchor-id="可视化cell-markers的表达">可视化cell markers的表达</h3>
<p>The&nbsp;<code><a href="https://rdrr.io/pkg/Seurat/man/DotPlot.html">DotPlot()</a></code>&nbsp;function with the&nbsp;<code>split.by</code>&nbsp;parameter can be useful for viewing conserved cell type markers across conditions, showing both the expression level and the percentage of cells in a cluster expressing any given gene. Here we plot 2-3 strong marker genes for each of our 14 clusters.</p>
<div class="cell" data-hash="integration_cache/html/unnamed-chunk-11_4782e702756fa60264204717c0345dca">
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># NEEDS TO BE FIXED AND SET ORDER CORRECTLY</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/Idents.html">Idents</a></span><span class="op">(</span><span class="va">ifnb</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/Idents.html">Idents</a></span><span class="op">(</span><span class="va">ifnb</span><span class="op">)</span>, </span>
<span>                       levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"pDC"</span>, <span class="st">"Eryth"</span>, <span class="st">"Mk"</span>, <span class="st">"DC"</span>, <span class="st">"CD14 Mono"</span>, <span class="st">"CD16 Mono"</span>, </span>
<span>                                  <span class="st">"B Activated"</span>, <span class="st">"B"</span>, <span class="st">"CD8 T"</span>, <span class="st">"NK"</span>, <span class="st">"T activated"</span>, </span>
<span>                                  <span class="st">"CD4 Naive T"</span>, <span class="st">"CD4 Memory T"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">markers.to.plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CD3D"</span>, <span class="st">"CREM"</span>, <span class="st">"HSPH1"</span>, <span class="st">"SELL"</span>, <span class="st">"GIMAP5"</span>, <span class="st">"CACYBP"</span>, <span class="st">"GNLY"</span>, <span class="st">"NKG7"</span>,</span>
<span>                     <span class="st">"CCL5"</span>, <span class="st">"CD8A"</span>, <span class="st">"MS4A1"</span>, <span class="st">"CD79A"</span>, <span class="st">"MIR155HG"</span>, <span class="st">"NME1"</span>, <span class="st">"FCGR3A"</span>, </span>
<span>                     <span class="st">"VMO1"</span>, <span class="st">"CCL2"</span>, <span class="st">"S100A9"</span>, <span class="st">"HLA-DQA1"</span>, <span class="st">"GPR183"</span>, <span class="st">"PPBP"</span>, <span class="st">"GNG11"</span>,</span>
<span>                     <span class="st">"HBA2"</span>, <span class="st">"HBB"</span>, <span class="st">"TSPAN13"</span>, <span class="st">"IL3RA"</span>, <span class="st">"IGJ"</span>, <span class="st">"PRSS57"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/DotPlot.html">DotPlot</a></span><span class="op">(</span><span class="va">ifnb</span>, </span>
<span>        features <span class="op">=</span> <span class="va">markers.to.plot</span>, </span>
<span>        cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"blue"</span>, <span class="st">"red"</span><span class="op">)</span>, </span>
<span>        dot.scale <span class="op">=</span> <span class="fl">8</span>, </span>
<span>        split.by <span class="op">=</span> <span class="st">"stim"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/SeuratTheme.html">RotatedAxis</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integration_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid" width="960"></p>
</div>
</div>
</section></section><section id="sec-识别不同样本类型间的差异基因" class="level2" data-number="8.5"><h2 data-number="8.5" class="anchored" data-anchor-id="sec-识别不同样本类型间的差异基因">
<span class="header-section-number">8.5</span> 识别不同样本类型间的差异基因</h2>
<p><strong>Now that we’ve aligned the stimulated and control cells, we can start to do comparative analyses and look at the differences induced by stimulation.</strong></p>
<p>We can aggregate cells of a similar type and condition together to create “pseudobulk” profiles using the&nbsp;<code>AggregateExpression</code> command（通过<code>AggregateExpression</code>命令将同一类型的细胞按照不同的处理条件合并起来，形成一个假的组织水平的测序数据。本例中，细胞被注释为13种细胞类型，而处理条件为”STIM”和”CTRL”，因此总共会被合并成13*2=26个类别，将每一个类别看作是一个样本，这样就形成了一个所谓的假的组织水平的测序数据）.</p>
<div class="cell" data-hash="integration_cache/html/unnamed-chunk-12_16672390297b0fe43a0b11ab0c171d0b">
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">aggregate_ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/AggregateExpression.html">AggregateExpression</a></span><span class="op">(</span><span class="va">ifnb</span>, </span>
<span>                                      group.by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"seurat_annotations"</span>, <span class="st">"stim"</span><span class="op">)</span>, </span>
<span>                                      return.seurat <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">aggregate_ifnb</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
14053 features across 26 samples within 1 assay 
Active assay: RNA (14053 features, 0 variable features)
 3 layers present: counts, data, scale.data</code></pre>
</div>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">aggregate_ifnb</span><span class="op">@</span><span class="va">meta.data</span>, <span class="fl">5</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                         orig.ident seurat_annotations stim
CD14 Mono_CTRL       CD14 Mono_CTRL          CD14 Mono CTRL
CD14 Mono_STIM       CD14 Mono_STIM          CD14 Mono STIM
CD4 Naive T_CTRL   CD4 Naive T_CTRL        CD4 Naive T CTRL
CD4 Naive T_STIM   CD4 Naive T_STIM        CD4 Naive T STIM
CD4 Memory T_CTRL CD4 Memory T_CTRL       CD4 Memory T CTRL</code></pre>
</div>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">aggregate_ifnb</span><span class="op">)</span> <span class="co"># 可以看到现在的表达矩阵的列（即样本）为细胞类型+处理条件</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "CD14 Mono_CTRL"    "CD14 Mono_STIM"    "CD4 Naive T_CTRL" 
 [4] "CD4 Naive T_STIM"  "CD4 Memory T_CTRL" "CD4 Memory T_STIM"
 [7] "CD16 Mono_CTRL"    "CD16 Mono_STIM"    "B_CTRL"           
[10] "B_STIM"            "CD8 T_CTRL"        "CD8 T_STIM"       
[13] "T activated_CTRL"  "T activated_STIM"  "NK_CTRL"          
[16] "NK_STIM"           "DC_CTRL"           "DC_STIM"          
[19] "B Activated_CTRL"  "B Activated_STIM"  "Mk_CTRL"          
[22] "Mk_STIM"           "pDC_CTRL"          "pDC_STIM"         
[25] "Eryth_CTRL"        "Eryth_STIM"       </code></pre>
</div>
</div>
<p><img src="images/截屏2023-11-27 15.59.00.png" class="img-fluid"></p>
<p>As an initial exploratory analysis, we can <strong>compare pseudobulk profiles of two cell types (naive CD4 T cells, and CD14 monocytes)</strong>, and compare their gene expression profiles before and after stimulation. We highlight genes that exhibit dramatic responses to interferon stimulation.</p>
<div>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/">cowplot</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/theme_get.html">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/cowplot/man/theme_cowplot.html">theme_cowplot</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># genes that exhibit dramatic responses to interferon stimulation</span></span>
<span><span class="va">genes.to.label</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ISG15"</span>, <span class="st">"LY6E"</span>, <span class="st">"IFI6"</span>, <span class="st">"ISG20"</span>, <span class="st">"MX1"</span>, <span class="st">"IFIT2"</span>, <span class="st">"IFIT1"</span>, <span class="st">"CXCL10"</span>,</span>
<span>                   <span class="st">"CCL8"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/CellScatter.html">CellScatter</a></span><span class="op">(</span><span class="va">aggregate_ifnb</span>, </span>
<span>                  <span class="st">"CD14 Mono_CTRL"</span>, <span class="st">"CD14 Mono_STIM"</span>, </span>
<span>                  highlight <span class="op">=</span> <span class="va">genes.to.label</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/LabelPoints.html">LabelPoints</a></span><span class="op">(</span>plot <span class="op">=</span> <span class="va">p1</span>, </span>
<span>                  points <span class="op">=</span> <span class="va">genes.to.label</span>, </span>
<span>                  repel <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/CellScatter.html">CellScatter</a></span><span class="op">(</span><span class="va">aggregate_ifnb</span>, </span>
<span>                  <span class="st">"CD4 Naive T_CTRL"</span>, <span class="st">"CD4 Naive T_STIM"</span>, </span>
<span>                  highlight <span class="op">=</span> <span class="va">genes.to.label</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/LabelPoints.html">LabelPoints</a></span><span class="op">(</span>plot <span class="op">=</span> <span class="va">p3</span>, </span>
<span>                  points <span class="op">=</span> <span class="va">genes.to.label</span>, </span>
<span>                  repel <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="fig-cellscater" class="cell quarto-layout-panel" data-hash="integration_cache/html/fig-cellscater_bca40ba392859d225e39425cd9940881">
<figure class="figure"><div class="quarto-layout-row quarto-layout-valign-top">
<div class="cell-output-display quarto-layout-cell quarto-layout-cell-subref" style="flex-basis: 50.0%;justify-content: center;">
<div id="fig-cellscater-1" class="quarto-figure quarto-figure-left anchored">
<figure class="figure"><p><img src="integration_files/figure-html/fig-cellscater-1.png" class="img-fluid figure-img" data-ref-parent="fig-cellscater" width="672"></p>
<figcaption class="figure-caption">(A) CD14 Mono细胞中的基因在对照组和刺激组之间的表达量散点图</figcaption></figure>
</div>
</div>
<div class="cell-output-display quarto-layout-cell quarto-layout-cell-subref" style="flex-basis: 50.0%;justify-content: center;">
<div id="fig-cellscater-2" class="quarto-figure quarto-figure-left anchored">
<figure class="figure"><p><img src="integration_files/figure-html/fig-cellscater-2.png" class="img-fluid figure-img" data-ref-parent="fig-cellscater" width="672"></p>
<figcaption class="figure-caption">(B) CD4 Naive T细胞中的基因在对照组和刺激组之间的表达量散点图</figcaption></figure>
</div>
</div>
</div>
<p></p><figcaption class="figure-caption">Figure&nbsp;8.3: CD14 Mono和CD4 Naive T细胞中的基因在对照组和刺激组之间的表达量散点图</figcaption><p></p>
</figure>
</div>
</div>
<blockquote class="blockquote">
<p>As you can see, many of the same genes are <strong>upregulated</strong> (位于对角线上方) in both of these cell types and <strong>likely represent a conserved interferon response pathway</strong>, though CD14 monocytes exhibit a stronger transcriptional response.</p>
</blockquote>
<section id="sec-正式差异分析" class="level3"><h3 class="anchored" data-anchor-id="sec-正式差异分析">正式差异分析</h3>
<p>We can now ask <strong>what genes change in different conditions for cells of the same type</strong>.</p>
<ol type="1">
<li><p>First, we create a column in the meta.data slot to hold both the cell type and stimulation information and switch the current ident to that column.</p></li>
<li><p>Then we use&nbsp;<a href="https://satijalab.org/seurat/reference/findmarkers"><code>FindMarkers()</code></a>&nbsp;to find the genes that are different <strong>between stimulated and control B cells</strong>. Notice that many of the top genes that show up here are the same as the ones we plotted earlier as core interferon response genes. Additionally, genes like CXCL10 which we saw were specific to monocyte and B cell interferon response show up as highly significant in this list as well.</p></li>
</ol>
<div class="cell" data-hash="integration_cache/html/unnamed-chunk-14_a955786577d95a182326add668db5e5b">
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">ifnb</span><span class="op">$</span><span class="va">celltype.stim</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="va">ifnb</span><span class="op">$</span><span class="va">seurat_annotations</span>, <span class="va">ifnb</span><span class="op">$</span><span class="va">stim</span>, sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/Idents.html">Idents</a></span><span class="op">(</span><span class="va">ifnb</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"celltype.stim"</span></span>
<span></span>
<span><span class="co"># 寻找对照组和刺激组之间在B细胞中的差异基因</span></span>
<span><span class="va">b.interferon.response</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindMarkers.html">FindMarkers</a></span><span class="op">(</span><span class="va">ifnb</span>, </span>
<span>                                     ident.1 <span class="op">=</span> <span class="st">"B_STIM"</span>, </span>
<span>                                     ident.2 <span class="op">=</span> <span class="st">"B_CTRL"</span>, </span>
<span>                                     verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">b.interferon.response</span>, n <span class="op">=</span> <span class="fl">15</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                p_val avg_log2FC pct.1 pct.2     p_val_adj
ISG15   5.387767e-159  5.0588481 0.998 0.233 7.571429e-155
IFIT3   1.945114e-154  6.1124940 0.965 0.052 2.733468e-150
IFI6    2.503565e-152  5.4933132 0.965 0.076 3.518260e-148
ISG20   6.492570e-150  3.0549593 1.000 0.668 9.124009e-146
IFIT1   1.951022e-139  6.2320388 0.907 0.029 2.741772e-135
MX1     6.897626e-123  3.9798482 0.905 0.115 9.693234e-119
LY6E    2.825649e-120  3.7907800 0.898 0.150 3.970885e-116
TNFSF10 4.007285e-112  6.5802175 0.786 0.020 5.631437e-108
IFIT2   2.672552e-108  5.5525558 0.786 0.037 3.755738e-104
B2M      5.283684e-98  0.6104044 1.000 1.000  7.425161e-94
PLSCR1   4.634658e-96  3.8010721 0.793 0.113  6.513085e-92
IRF7     2.411149e-94  3.1992949 0.835 0.187  3.388388e-90
CXCL10   3.708508e-86  8.0906108 0.651 0.010  5.211566e-82
UBE2L6   5.547472e-83  2.5167981 0.851 0.297  7.795863e-79
PSMB9    1.716262e-77  1.7715351 0.937 0.568  2.411863e-73</code></pre>
</div>
</div>
<p>Please note that p-values obtained from this analysis should be interpreted with caution, as <strong>these tests treat each cell as an independent replicate, and ignore inherent correlations between cells originating from the same sample</strong>. As discussed&nbsp;here <span class="citation" data-cites="crowell2020">(<a href="../references.html#ref-crowell2020" role="doc-biblioref">Crowell et al. 2020</a>)</span>, DE tests across multiple conditions should expressly utilize multiple samples/replicates, and can be performed after aggregating (‘pseudobulking’) cells from the same sample and subpopulation together. We do not perform this analysis here, as there is a single replicate in the data, but please see our&nbsp;<a href="https://satijalab.org/seurat/articles/parsebio_sketch_integration">vignette comparing healthy and diabetic samples</a>&nbsp;as an example for how to perform DE analysis across conditions.</p>
<p>Another useful way to visualize these changes in gene expression is with the&nbsp;<code>split.by</code>&nbsp;option to the&nbsp;<code><a href="https://rdrr.io/pkg/Seurat/man/FeaturePlot.html">FeaturePlot()</a></code>&nbsp;or&nbsp;<code><a href="https://rdrr.io/pkg/Seurat/man/VlnPlot.html">VlnPlot()</a></code>&nbsp;function. This will display FeaturePlots of the list of given genes, split by a grouping variable (stimulation condition here).</p>
<div class="cell" data-hash="integration_cache/html/unnamed-chunk-15_98d8ec3222f1df091da45997b968c6d5">
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FeaturePlot.html">FeaturePlot</a></span><span class="op">(</span><span class="va">ifnb</span>, </span>
<span>            features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CD3D"</span>, <span class="st">"GNLY"</span>, <span class="st">"IFI6"</span>, <span class="st">"ISG15"</span>, <span class="st">"CD14"</span>, <span class="st">"CXCL10"</span><span class="op">)</span>, </span>
<span>            split.by <span class="op">=</span> <span class="st">"stim"</span>, </span>
<span>            max.cutoff <span class="op">=</span> <span class="fl">3</span>, </span>
<span>            cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"grey"</span>, <span class="st">"red"</span><span class="op">)</span>, </span>
<span>            reduction <span class="op">=</span> <span class="st">"umap"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integration_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell" data-hash="integration_cache/html/unnamed-chunk-16_c510ad0382f0a414118e9d87bc657675">
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/VlnPlot.html">VlnPlot</a></span><span class="op">(</span><span class="va">ifnb</span>,</span>
<span>        features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CD3D"</span>, <span class="st">"GNLY"</span>, <span class="st">"IFI6"</span>, <span class="st">"ISG15"</span>, <span class="st">"CD14"</span>, <span class="st">"CXCL10"</span>, <span class="st">"LYZ"</span><span class="op">)</span>,</span>
<span>        split.by <span class="op">=</span> <span class="st">"stim"</span>,</span>
<span>        group.by <span class="op">=</span> <span class="st">"seurat_annotations"</span>,</span>
<span>        pt.size <span class="op">=</span> <span class="fl">0</span>,</span>
<span>        combine <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="co"># 由于VlnPlot绘制组图时没有图例，所以这里取消绘制组图</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com">patchwork</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/patchwork/man/wrap_plots.html">wrap_plots</a></span><span class="op">(</span>plots <span class="op">=</span> <span class="va">plots</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="co"># 将plots列表组合成组图</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integration_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid" width="960"></p>
</div>
</div>
<blockquote class="blockquote">
<ul>
<li><p><em>Genes such as <strong>CD3D</strong> and <strong>GNLY</strong> are canonical cell type markers (for T cells and NK/CD8 T cells) that are virtually <strong>unaffected by interferon stimulation</strong> and <strong>display similar gene expression patterns in the control and stimulated group</strong>.</em></p></li>
<li><p><em><strong>IFI6</strong> and <strong>ISG15</strong>, on the other hand, are <strong>core interferon response genes</strong> and are <strong>upregulated</strong> accordingly <strong>in all cell types</strong>.</em></p></li>
<li>
<p><em><strong>CD14</strong> and <strong>CXCL10</strong> are genes that show a <strong>cell type specific interferon response</strong>.</em></p>
<ul>
<li><p><em><strong>CD14</strong> expression <strong>decreases</strong> after stimulation in <strong>CD14 monocytes</strong>, <u>which could lead to misclassification in a supervised analysis framework, underscoring the value of integrated analysis.</u><strong>如果用于识别细胞类型的marker本身在不同的样本类型（处理 vs.&nbsp;对照、恶性组织 vs.&nbsp;正常组织）中存在表达量的差异，那么就会导致对细胞类型判断的错误。而本篇的数据整合则能够避免出现这种情况。</strong></em></p></li>
<li><p><em><strong>CXCL10</strong> shows a distinct <strong>upregulation</strong> in <strong>monocytes</strong> and <strong>B cells</strong> after interferon stimulation but not in other cell types.</em></p></li>
</ul>
</li>
</ul>
</blockquote>
</section></section><section id="执行sctransform标准化流程之后的整合" class="level2" data-number="8.6"><h2 data-number="8.6" class="anchored" data-anchor-id="执行sctransform标准化流程之后的整合">
<span class="header-section-number">8.6</span> 执行<code>SCTransform</code>标准化流程之后的整合</h2>
<p>As an alternative to log-normalization, Seurat also includes support for preprocessing of scRNA-seq using the&nbsp;<code>SCTransform</code> workflow（ <a href="sctransform.html"><span>Chapter&nbsp;7</span></a> ）. The&nbsp;<code>IntegrateLayers</code>&nbsp;function also supports SCTransform-normalized data, by setting the&nbsp;<code>normalization.method</code> parameter, as shown below.</p>
<section id="不进行整合的情况下的数据分析" class="level3"><h3 class="anchored" data-anchor-id="不进行整合的情况下的数据分析">不进行整合的情况下的数据分析</h3>
<div class="cell" data-hash="integration_cache/html/fig-sctransform_未整合_444b0268a7062ac06e0880f21735ad19">
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 重新载入原始的Seurat对象ifnb</span></span>
<span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="st">"data/pbmc_ifnb.rds"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 同样先拆分数据集，然后进行无整合情况下的降维</span></span>
<span><span class="va">ifnb</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html">split</a></span><span class="op">(</span><span class="va">ifnb</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span>, f <span class="op">=</span> <span class="va">ifnb</span><span class="op">$</span><span class="va">stim</span><span class="op">)</span></span>
<span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/SCTransform.html">SCTransform</a></span><span class="op">(</span><span class="va">ifnb</span><span class="op">)</span></span>
<span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/RunPCA.html">RunPCA</a></span><span class="op">(</span><span class="va">ifnb</span><span class="op">)</span></span>
<span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/RunUMAP.html">RunUMAP</a></span><span class="op">(</span><span class="va">ifnb</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">ifnb</span>, </span>
<span>        reduction <span class="op">=</span> <span class="st">"umap"</span>, </span>
<span>        group.by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"stim"</span>, <span class="st">"seurat_annotations"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-sctransform_未整合" class="quarto-figure quarto-figure-left anchored">
<figure class="figure"><p><img src="integration_files/figure-html/fig-sctransform_未整合-1.png" class="img-fluid figure-img" width="960"></p>
<figcaption class="figure-caption">Figure&nbsp;8.4: 未整合时的细胞分群情况（左：按照刺激条件着色；右：按照细胞聚类情况着色）</figcaption></figure>
</div>
</div>
</div>
<p>可以看到，如果不进行整合，不同样本（STIM vs.&nbsp;STIM）的细胞类型差异很大。</p>
</section><section id="sec-integration_after_sct" class="level3"><h3 class="anchored" data-anchor-id="sec-integration_after_sct">进行整合</h3>
<p>同样通过<code>IntegrateLayers</code>函数进行数据整合，只不过需要将默认的标准化方法由”LogNormalize”指定为”SCT”（<code>normalization.method = "SCT"</code>）：</p>
<div class="cell" data-hash="integration_cache/html/unnamed-chunk-18_634d1f8b55e6e636141ffeb5c3c8079f">
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/IntegrateLayers.html">IntegrateLayers</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">ifnb</span>, </span>
<span>                        method <span class="op">=</span> <span class="va">CCAIntegration</span>, </span>
<span>                        normalization.method <span class="op">=</span> <span class="st">"SCT"</span>, </span>
<span>                        verbose <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><img src="images/截屏2023-11-28 09.16.10.png" class="img-fluid"></p>
<p>可以看到经过整合的现在的Seurat对象中除了”RNA”的assay还由”SCT”的assay。同时，降维（“reduction”）中多出了整合后的降维（“integrated.dr”）。</p>
</section><section id="整合后聚类" class="level3"><h3 class="anchored" data-anchor-id="整合后聚类">整合后聚类</h3>
<div class="cell" data-hash="integration_cache/html/unnamed-chunk-19_0fb028fe7c65a8a3f4bbbb128b7b6975">
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindNeighbors.html">FindNeighbors</a></span><span class="op">(</span><span class="va">ifnb</span>, </span>
<span>                      reduction <span class="op">=</span> <span class="st">"integrated.dr"</span>, <span class="co">#更改降维来源为"integrated.dr"</span></span>
<span>                      dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span></span>
<span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindClusters.html">FindClusters</a></span><span class="op">(</span><span class="va">ifnb</span>, </span>
<span>                     resolution <span class="op">=</span> <span class="fl">0.6</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 13999
Number of edges: 527905

Running Louvain algorithm...
Maximum modularity in 10 random starts: 0.9058
Number of communities: 19
Elapsed time: 1 seconds</code></pre>
</div>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/RunUMAP.html">RunUMAP</a></span><span class="op">(</span><span class="va">ifnb</span>, </span>
<span>                dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>, </span>
<span>                reduction <span class="op">=</span> <span class="st">"integrated.dr"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">ifnb</span>, </span>
<span>        reduction <span class="op">=</span> <span class="st">"umap"</span>, </span>
<span>        group.by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"stim"</span>, <span class="st">"seurat_annotations"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integration_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid" width="960"></p>
</div>
</div>
<p>可以看到和 <a href="#fig-sctransform_%E6%9C%AA%E6%95%B4%E5%90%88">Figure&nbsp;<span>8.4</span></a> 相比，整合后在样本间的细胞类型基本均匀分布。</p>
</section><section id="差异表达分析" class="level3"><h3 class="anchored" data-anchor-id="差异表达分析">差异表达分析</h3>
<p>对于<code>SCTransform</code>处理的数据首先要通过PrepSCTFindMarkers函数来预处理，然后再进行差异分析，基本内容和 <a href="#sec-%E6%AD%A3%E5%BC%8F%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90"><span>Section&nbsp;8.5.1</span></a> 一致。</p>
<div class="cell" data-hash="integration_cache/html/unnamed-chunk-20_be71e54a06b47d31f8095a6517204348">
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">ifnb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/PrepSCTFindMarkers.html">PrepSCTFindMarkers</a></span><span class="op">(</span><span class="va">ifnb</span><span class="op">)</span></span>
<span><span class="va">ifnb</span><span class="op">$</span><span class="va">celltype.stim</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="va">ifnb</span><span class="op">$</span><span class="va">seurat_annotations</span>, </span>
<span>                            <span class="va">ifnb</span><span class="op">$</span><span class="va">stim</span>, </span>
<span>                            sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/Idents.html">Idents</a></span><span class="op">(</span><span class="va">ifnb</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"celltype.stim"</span></span>
<span><span class="va">b.interferon.response</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindMarkers.html">FindMarkers</a></span><span class="op">(</span><span class="va">ifnb</span>, </span>
<span>                                     ident.1 <span class="op">=</span> <span class="st">"B_STIM"</span>, </span>
<span>                                     ident.2 <span class="op">=</span> <span class="st">"B_CTRL"</span>, </span>
<span>                                     verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">b.interferon.response</span>, n <span class="op">=</span> <span class="fl">15</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                p_val avg_log2FC pct.1 pct.2     p_val_adj
ISG15   1.505650e-159  5.1597242 0.998 0.229 2.003417e-155
IFIT3   4.128835e-154  6.2281506 0.961 0.052 5.493827e-150
IFI6    2.493020e-153  5.6458391 0.965 0.076 3.317212e-149
ISG20   9.385626e-152  3.1715979 1.000 0.666 1.248851e-147
IFIT1   2.447118e-139  6.2614957 0.904 0.029 3.256136e-135
MX1     2.111944e-124  4.1490465 0.900 0.115 2.810153e-120
LY6E    2.930414e-122  3.9278136 0.898 0.150 3.899209e-118
TNFSF10 1.104024e-112  6.8254288 0.785 0.020 1.469014e-108
IFIT2   3.491368e-108  5.5668205 0.783 0.037 4.645615e-104
B2M      3.405403e-98  0.6074989 1.000 1.000  4.531229e-94
IRF7     1.114291e-96  3.3721350 0.834 0.187  1.482675e-92
PLSCR1   3.364901e-96  4.0217697 0.783 0.111  4.477338e-92
UBE2L6   1.155610e-85  2.6732717 0.849 0.295  1.537655e-81
CXCL10   5.689834e-84  8.0923962 0.639 0.010  7.570893e-80
PSMB9    2.304426e-81  1.8684260 0.937 0.568  3.066269e-77</code></pre>
</div>
</div>
<div class="cell" data-hash="integration_cache/html/unnamed-chunk-21_4c88edec45ec5e87d8e6020b82774d63">
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FeaturePlot.html">FeaturePlot</a></span><span class="op">(</span><span class="va">ifnb</span>, </span>
<span>            features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CD3D"</span>, <span class="st">"GNLY"</span>, <span class="st">"IFI6"</span>, <span class="st">"ISG15"</span>, <span class="st">"CD14"</span>, <span class="st">"CXCL10"</span><span class="op">)</span>, </span>
<span>            split.by <span class="op">=</span> <span class="st">"stim"</span>, </span>
<span>            max.cutoff <span class="op">=</span> <span class="fl">3</span>, </span>
<span>            cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"grey"</span>, <span class="st">"red"</span><span class="op">)</span>, </span>
<span>            reduction <span class="op">=</span> <span class="st">"umap"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="integration_files/figure-html/unnamed-chunk-21-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-4-contents" aria-controls="callout-4" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Session Info
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-4" class="callout-4-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-hash="integration_cache/html/unnamed-chunk-22_9cad0e98af70231dc7bc62e0b03fb548">
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.3.2 (2023-10-31)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Sonoma 14.1.1

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: Asia/Shanghai
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] patchwork_1.1.3    cowplot_1.1.1      ggplot2_3.4.4      Seurat_5.0.1      
[5] SeuratObject_5.0.1 sp_2.1-1          

loaded via a namespace (and not attached):
  [1] mathjaxr_1.6-0              RColorBrewer_1.1-3         
  [3] rstudioapi_0.15.0           jsonlite_1.8.7             
  [5] magrittr_2.0.3              TH.data_1.1-2              
  [7] spatstat.utils_3.0-4        farver_2.1.1               
  [9] rmarkdown_2.25              zlibbioc_1.48.0            
 [11] vctrs_0.6.4                 multtest_2.58.0            
 [13] ROCR_1.0-11                 DelayedMatrixStats_1.24.0  
 [15] spatstat.explore_3.2-5      RCurl_1.98-1.13            
 [17] S4Arrays_1.2.0              htmltools_0.5.7            
 [19] plotrix_3.8-4               SparseArray_1.2.2          
 [21] sctransform_0.4.1           parallelly_1.36.0          
 [23] KernSmooth_2.23-22          htmlwidgets_1.6.3          
 [25] ica_1.0-3                   sandwich_3.0-2             
 [27] plyr_1.8.9                  plotly_4.10.3              
 [29] zoo_1.8-12                  igraph_1.5.1               
 [31] mime_0.12                   lifecycle_1.0.4            
 [33] pkgconfig_2.0.3             Matrix_1.6-3               
 [35] R6_2.5.1                    fastmap_1.1.1              
 [37] GenomeInfoDbData_1.2.11     MatrixGenerics_1.14.0      
 [39] rbibutils_2.2.16            fitdistrplus_1.1-11        
 [41] future_1.33.0               shiny_1.8.0                
 [43] digest_0.6.33               numDeriv_2016.8-1.1        
 [45] colorspace_2.1-0            S4Vectors_0.40.1           
 [47] tensor_1.5                  RSpectra_0.16-1            
 [49] irlba_2.3.5.1               GenomicRanges_1.54.1       
 [51] labeling_0.4.3              progressr_0.14.0           
 [53] fansi_1.0.5                 spatstat.sparse_3.0-3      
 [55] httr_1.4.7                  TFisher_0.2.0              
 [57] polyclip_1.10-6             abind_1.4-5                
 [59] compiler_4.3.2              withr_2.5.2                
 [61] mutoss_0.1-13               fastDummies_1.7.3          
 [63] MASS_7.3-60                 DelayedArray_0.28.0        
 [65] tools_4.3.2                 lmtest_0.9-40              
 [67] metap_1.9                   httpuv_1.6.12              
 [69] future.apply_1.11.0         qqconf_1.3.2               
 [71] goftest_1.2-3               glmGamPoi_1.14.0           
 [73] glue_1.6.2                  nlme_3.1-163               
 [75] promises_1.2.1              grid_4.3.2                 
 [77] Rtsne_0.16                  cluster_2.1.4              
 [79] reshape2_1.4.4              generics_0.1.3             
 [81] gtable_0.3.4                spatstat.data_3.0-3        
 [83] tidyr_1.3.0                 sn_2.1.1                   
 [85] data.table_1.14.8           XVector_0.42.0             
 [87] utf8_1.2.4                  BiocGenerics_0.48.1        
 [89] spatstat.geom_3.2-7         RcppAnnoy_0.0.21           
 [91] ggrepel_0.9.4               RANN_2.6.1                 
 [93] pillar_1.9.0                stringr_1.5.1              
 [95] spam_2.10-0                 RcppHNSW_0.5.0             
 [97] later_1.3.1                 splines_4.3.2              
 [99] dplyr_1.1.4                 lattice_0.22-5             
[101] survival_3.5-7              deldir_2.0-2               
[103] tidyselect_1.2.0            miniUI_0.1.1.1             
[105] pbapply_1.7-2               knitr_1.45                 
[107] gridExtra_2.3               IRanges_2.36.0             
[109] SummarizedExperiment_1.32.0 scattermore_1.2            
[111] stats4_4.3.2                xfun_0.41                  
[113] Biobase_2.62.0              matrixStats_1.1.0          
[115] stringi_1.8.2               lazyeval_0.2.2             
[117] yaml_2.3.7                  evaluate_0.23              
[119] codetools_0.2-19            tibble_3.2.1               
[121] cli_3.6.1                   uwot_0.1.16                
[123] xtable_1.8-4                reticulate_1.34.0          
[125] Rdpack_2.6                  munsell_0.5.0              
[127] GenomeInfoDb_1.38.1         Rcpp_1.0.11                
[129] globals_0.16.2              spatstat.random_3.2-1      
[131] png_0.1-8                   parallel_4.3.2             
[133] ellipsis_0.3.2              dotCall64_1.1-0            
[135] sparseMatrixStats_1.14.0    bitops_1.0-7               
[137] listenv_0.9.0               viridisLite_0.4.2          
[139] mvtnorm_1.2-3               scales_1.2.1               
[141] ggridges_0.5.4              crayon_1.5.2               
[143] leiden_0.4.3.1              purrr_1.0.2                
[145] rlang_1.1.2                 multcomp_1.4-25            
[147] mnormt_2.1.1               </code></pre>
</div>
</div>
</div>
</div>
</div>


<!-- -->

<div id="refs" class="references csl-bib-body hanging-indent" role="list" style="display: none">
<div id="ref-crowell2020" class="csl-entry" role="listitem">
Crowell, Helena L., Charlotte Soneson, Pierre-Luc Germain, Daniela Calini, Ludovic Collin, Catarina Raposo, Dheeraj Malhotra, and Mark D. Robinson. 2020. <span>“Muscat Detects Subpopulation-Specific State Transitions from Multi-Sample Multi-Condition Single-Cell Transcriptomics Data.”</span> <em>Nature Communications</em> 11 (1). <a href="https://doi.org/10.1038/s41467-020-19894-4">https://doi.org/10.1038/s41467-020-19894-4</a>.
</div>
</div>
</section></section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  let localAlternateSentinel = 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var filterRegex = new RegExp("https:\/\/djhcod\.github\.io\/r-notes\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
          // target, if specified
          link.setAttribute("target", "_blank");
      }
    }
});
</script><script src="https://utteranc.es/client.js" repo="https://github.com/djhcod/r-notes" issue-term="pathname" theme="github-light" crossorigin="anonymous" async="">
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="../single_cell/sctransform.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">基于SCTransform的单细胞数据标准化</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../single_cell/integrative_analysis_in_seurat_v5.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Seurat v5单细胞数据整合分析</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb42" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="fu"># 单细胞数据整合（integration） {#sec-单细胞数据整合}</span></span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a>原文：<span class="co">[</span><span class="ot">*Introduction to scRNA-seq integration*</span><span class="co">](https://satijalab.org/seurat/articles/integration_introduction)</span></span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a>原文发布日期：2023年10月31日</span>
<span id="cb42-6"><a href="#cb42-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-7"><a href="#cb42-7" aria-hidden="true" tabindex="-1"></a>::: callout-tip</span>
<span id="cb42-8"><a href="#cb42-8" aria-hidden="true" tabindex="-1"></a>本篇主要介绍来自不同样本类型的单细胞数据的整合。对于如何整合不同测序技术的单细胞数据集，参考Seurat官方文档：<span class="co">[</span><span class="ot">*Integrative analysis in Seurat v5*</span><span class="co">](https://satijalab.org/seurat/articles/seurat5_integration)</span>。</span>
<span id="cb42-9"><a href="#cb42-9" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb42-10"><a href="#cb42-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-11"><a href="#cb42-11" aria-hidden="true" tabindex="-1"></a>Integration of single-cell sequencing datasets, for example across experimental batches, donors, or conditions, is often an important step in scRNA-seq workflows. Integrative analysis can help to **match shared cell types and states across datasets**, which can boost statistical power, and most importantly, facilitate accurate comparative analysis across datasets. In previous versions of Seurat we introduced methods for integrative analysis, including our 'anchor-based' integration workflow. Many labs have also published powerful and pioneering methods, including&nbsp;<span class="co">[</span><span class="ot">Harmony</span><span class="co">](https://portals.broadinstitute.org/harmony/)</span>&nbsp;and&nbsp;<span class="co">[</span><span class="ot">scVI</span><span class="co">](https://docs.scvi-tools.org/en/stable/index.html)</span>, for integrative analysis.</span>
<span id="cb42-12"><a href="#cb42-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-13"><a href="#cb42-13" aria-hidden="true" tabindex="-1"></a>数据整合的目标：</span>
<span id="cb42-14"><a href="#cb42-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-15"><a href="#cb42-15" aria-hidden="true" tabindex="-1"></a>The following tutorial is designed to give you an overview of the kinds of comparative analyses on complex cell types that are possible using the Seurat integration procedure. Here, we address a few key goals:</span>
<span id="cb42-16"><a href="#cb42-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-17"><a href="#cb42-17" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Identify **cell subpopulations** that are present in both datasets</span>
<span id="cb42-18"><a href="#cb42-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-19"><a href="#cb42-19" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Obtain **cell type markers** that are conserved in both control and stimulated cells</span>
<span id="cb42-20"><a href="#cb42-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-21"><a href="#cb42-21" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>**Compare the datasets** to find cell-type specific responses to stimulation</span>
<span id="cb42-22"><a href="#cb42-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-23"><a href="#cb42-23" aria-hidden="true" tabindex="-1"></a><span class="fu">## 数据读取和分层</span></span>
<span id="cb42-24"><a href="#cb42-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-25"><a href="#cb42-25" aria-hidden="true" tabindex="-1"></a>::: {.callout-note collapse="true" appearance="minimal" icon="false"}</span>
<span id="cb42-26"><a href="#cb42-26" aria-hidden="true" tabindex="-1"></a><span class="fu">### 在线读取（可能需要全局代理）</span></span>
<span id="cb42-27"><a href="#cb42-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-30"><a href="#cb42-30" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb42-31"><a href="#cb42-31" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb42-32"><a href="#cb42-32" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: fenced</span></span>
<span id="cb42-33"><a href="#cb42-33" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb42-34"><a href="#cb42-34" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SeuratData)</span>
<span id="cb42-35"><a href="#cb42-35" aria-hidden="true" tabindex="-1"></a><span class="fu">InstallData</span>(<span class="st">"ifnb"</span>)</span>
<span id="cb42-36"><a href="#cb42-36" aria-hidden="true" tabindex="-1"></a>ifnb <span class="ot">&lt;-</span> <span class="fu">LoadData</span>(<span class="st">"ifnb"</span>)</span>
<span id="cb42-37"><a href="#cb42-37" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb42-38"><a href="#cb42-38" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb42-39"><a href="#cb42-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-40"><a href="#cb42-40" aria-hidden="true" tabindex="-1"></a>从本地下载好的数据读取：</span>
<span id="cb42-41"><a href="#cb42-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-44"><a href="#cb42-44" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb42-45"><a href="#cb42-45" aria-hidden="true" tabindex="-1"></a>ifnb <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"data/pbmc_ifnb.rds"</span>)</span>
<span id="cb42-46"><a href="#cb42-46" aria-hidden="true" tabindex="-1"></a>ifnb</span>
<span id="cb42-47"><a href="#cb42-47" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(ifnb<span class="sc">@</span>meta.data, <span class="dv">5</span>)</span>
<span id="cb42-48"><a href="#cb42-48" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb42-49"><a href="#cb42-49" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-50"><a href="#cb42-50" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/截屏2023-11-27%2012.49.37.png)</span></span>
<span id="cb42-51"><a href="#cb42-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-52"><a href="#cb42-52" aria-hidden="true" tabindex="-1"></a>The object contains data from human PBMC from two conditions, **interferon-stimulated** and **control** cells (stored in the&nbsp;<span class="in">`stim`</span> column in the object metadata). We will aim to integrate the two conditions together, so that we can jointly identify cell subpopulations across datasets, and then explore how each group differs across conditions</span>
<span id="cb42-53"><a href="#cb42-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-54"><a href="#cb42-54" aria-hidden="true" tabindex="-1"></a>In previous versions of Seurat, we would require the data to be represented as two different Seurat objects. **In Seurat v5, we keep all the data in one object, but simply split it into multiple 'layers'**. To learn more about layers, check out our&nbsp;<span class="co">[</span><span class="ot">Seurat object interaction vignette</span><span class="co">](https://satijalab.org/seurat/articles/interaction_vignette)</span>.</span>
<span id="cb42-55"><a href="#cb42-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-56"><a href="#cb42-56" aria-hidden="true" tabindex="-1"></a>::: callout-important</span>
<span id="cb42-57"><a href="#cb42-57" aria-hidden="true" tabindex="-1"></a>Seurat v5 assays store data in **layers**. These layers can store:</span>
<span id="cb42-58"><a href="#cb42-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-59"><a href="#cb42-59" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>raw, un-normalized counts (<span class="in">`layer='counts'`</span>)</span>
<span id="cb42-60"><a href="#cb42-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-61"><a href="#cb42-61" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>normalized data (<span class="in">`layer='data'`</span>)</span>
<span id="cb42-62"><a href="#cb42-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-63"><a href="#cb42-63" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>z-scored/variance-stabilized data (<span class="in">`layer='scale.data'`</span>).</span>
<span id="cb42-64"><a href="#cb42-64" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb42-65"><a href="#cb42-65" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-66"><a href="#cb42-66" aria-hidden="true" tabindex="-1"></a>**split the RNA measurements into two layers one for control cells, one for stimulated cells:**</span>
<span id="cb42-67"><a href="#cb42-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-70"><a href="#cb42-70" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb42-71"><a href="#cb42-71" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb42-72"><a href="#cb42-72" aria-hidden="true" tabindex="-1"></a>ifnb[[<span class="st">"RNA"</span>]] <span class="ot">&lt;-</span> <span class="fu">split</span>(ifnb[[<span class="st">"RNA"</span>]], </span>
<span id="cb42-73"><a href="#cb42-73" aria-hidden="true" tabindex="-1"></a>                       <span class="at">f =</span> ifnb<span class="sc">$</span>stim) <span class="co"># 按照meta.data中的“stim”列进行分割</span></span>
<span id="cb42-74"><a href="#cb42-74" aria-hidden="true" tabindex="-1"></a>ifnb</span>
<span id="cb42-75"><a href="#cb42-75" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb42-76"><a href="#cb42-76" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-77"><a href="#cb42-77" aria-hidden="true" tabindex="-1"></a>现在可以发现<span class="in">`ifnb`</span>被分为了4个layer，此前是2个layer（<span class="in">`counts`</span>和<span class="in">`data`</span>）：</span>
<span id="cb42-78"><a href="#cb42-78" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-79"><a href="#cb42-79" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/截屏2023-11-27%2011.55.48.png)</span></span>
<span id="cb42-80"><a href="#cb42-80" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-81"><a href="#cb42-81" aria-hidden="true" tabindex="-1"></a><span class="fu">## 不进行整合的情况下的数据处理</span></span>
<span id="cb42-82"><a href="#cb42-82" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-83"><a href="#cb42-83" aria-hidden="true" tabindex="-1"></a>进行标准的数据处理流程：</span>
<span id="cb42-84"><a href="#cb42-84" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-87"><a href="#cb42-87" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb42-88"><a href="#cb42-88" aria-hidden="true" tabindex="-1"></a>ifnb <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(ifnb)</span>
<span id="cb42-89"><a href="#cb42-89" aria-hidden="true" tabindex="-1"></a>ifnb <span class="ot">&lt;-</span> <span class="fu">FindVariableFeatures</span>(ifnb)</span>
<span id="cb42-90"><a href="#cb42-90" aria-hidden="true" tabindex="-1"></a>ifnb <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(ifnb)</span>
<span id="cb42-91"><a href="#cb42-91" aria-hidden="true" tabindex="-1"></a>ifnb <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(ifnb)</span>
<span id="cb42-92"><a href="#cb42-92" aria-hidden="true" tabindex="-1"></a>ifnb <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(ifnb, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>, <span class="at">reduction =</span> <span class="st">"pca"</span>)</span>
<span id="cb42-93"><a href="#cb42-93" aria-hidden="true" tabindex="-1"></a>ifnb <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(ifnb, </span>
<span id="cb42-94"><a href="#cb42-94" aria-hidden="true" tabindex="-1"></a>                     <span class="at">resolution =</span> <span class="dv">2</span>, </span>
<span id="cb42-95"><a href="#cb42-95" aria-hidden="true" tabindex="-1"></a>                     <span class="at">cluster.name =</span> <span class="st">"unintegrated_clusters"</span>)</span>
<span id="cb42-96"><a href="#cb42-96" aria-hidden="true" tabindex="-1"></a>ifnb <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(ifnb, </span>
<span id="cb42-97"><a href="#cb42-97" aria-hidden="true" tabindex="-1"></a>                <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>, </span>
<span id="cb42-98"><a href="#cb42-98" aria-hidden="true" tabindex="-1"></a>                <span class="at">reduction =</span> <span class="st">"pca"</span>, </span>
<span id="cb42-99"><a href="#cb42-99" aria-hidden="true" tabindex="-1"></a>                <span class="at">reduction.name =</span> <span class="st">"umap.unintegrated"</span>)</span>
<span id="cb42-100"><a href="#cb42-100" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb42-101"><a href="#cb42-101" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-102"><a href="#cb42-102" aria-hidden="true" tabindex="-1"></a>分别按照**样本分组（"stim"**）**和**细胞聚类情况（"seurat_clusters"**）**着色绘制UMAP图：</span>
<span id="cb42-103"><a href="#cb42-103" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-106"><a href="#cb42-106" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb42-107"><a href="#cb42-107" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-未整合</span></span>
<span id="cb42-108"><a href="#cb42-108" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: 未整合时的细胞分群情况（左：按照刺激条件着色；右：按照细胞聚类情况着色）</span></span>
<span id="cb42-109"><a href="#cb42-109" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 10</span></span>
<span id="cb42-110"><a href="#cb42-110" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(ifnb, </span>
<span id="cb42-111"><a href="#cb42-111" aria-hidden="true" tabindex="-1"></a>        <span class="at">reduction =</span> <span class="st">"umap.unintegrated"</span>, </span>
<span id="cb42-112"><a href="#cb42-112" aria-hidden="true" tabindex="-1"></a>        <span class="at">group.by =</span> <span class="fu">c</span>(<span class="st">"stim"</span>, <span class="st">"seurat_clusters"</span>))</span>
<span id="cb42-113"><a href="#cb42-113" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb42-114"><a href="#cb42-114" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-115"><a href="#cb42-115" aria-hidden="true" tabindex="-1"></a>可以发现：The resulting clusters are defined **both by cell type** **and stimulation condition**, which creates challenges for downstream analysis.</span>
<span id="cb42-116"><a href="#cb42-116" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-117"><a href="#cb42-117" aria-hidden="true" tabindex="-1"></a><span class="fu">## 进行数据整合</span></span>
<span id="cb42-118"><a href="#cb42-118" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-119"><a href="#cb42-119" aria-hidden="true" tabindex="-1"></a>We now aim to integrate data from the two conditions, so that cells from the same cell type/subpopulation will cluster together.</span>
<span id="cb42-120"><a href="#cb42-120" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-121"><a href="#cb42-121" aria-hidden="true" tabindex="-1"></a>We often refer to this procedure as intergration/alignment. When aligning two genome sequences together, identification of shared/homologous regions can help to interpret differences between the sequences as well. Similarly for scRNA-seq integration, our goal is **not to remove biological differences across conditions**, **but to learn shared cell types/states in an initial step-specifically** because that will enable us to compare control stimulated and control profiles for these individual cell types.</span>
<span id="cb42-122"><a href="#cb42-122" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-123"><a href="#cb42-123" aria-hidden="true" tabindex="-1"></a>The Seurat v5 integration procedure aims to return a single dimensional reduction that captures the shared sources of variance across multiple layers, so that cells in a similar biological state will cluster. The method returns a dimensional reduction (i.e.&nbsp;<span class="in">`integrated.cca`</span>) which can be used for visualization and unsupervised clustering analysis. For evaluating performance, we can use cell type labels that are pre-loaded in the&nbsp;<span class="in">`seurat_annotations`</span>&nbsp;metadata column.</span>
<span id="cb42-124"><a href="#cb42-124" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-127"><a href="#cb42-127" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb42-128"><a href="#cb42-128" aria-hidden="true" tabindex="-1"></a><span class="co"># 整合，比较耗时间，进度条会一直显示0%直至运算完成</span></span>
<span id="cb42-129"><a href="#cb42-129" aria-hidden="true" tabindex="-1"></a>ifnb <span class="ot">&lt;-</span> <span class="fu">IntegrateLayers</span>(<span class="at">object =</span> ifnb, </span>
<span id="cb42-130"><a href="#cb42-130" aria-hidden="true" tabindex="-1"></a>                        <span class="at">method =</span> CCAIntegration, </span>
<span id="cb42-131"><a href="#cb42-131" aria-hidden="true" tabindex="-1"></a>                        <span class="at">orig.reduction =</span> <span class="st">"pca"</span>, </span>
<span id="cb42-132"><a href="#cb42-132" aria-hidden="true" tabindex="-1"></a>                        <span class="at">new.reduction =</span> <span class="st">"integrated.cca"</span>, <span class="co"># 整合后新的降维数据的名称</span></span>
<span id="cb42-133"><a href="#cb42-133" aria-hidden="true" tabindex="-1"></a>                        <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb42-134"><a href="#cb42-134" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-135"><a href="#cb42-135" aria-hidden="true" tabindex="-1"></a><span class="co"># 整合后重新合并layer</span></span>
<span id="cb42-136"><a href="#cb42-136" aria-hidden="true" tabindex="-1"></a>ifnb[[<span class="st">"RNA"</span>]] <span class="ot">&lt;-</span> <span class="fu">JoinLayers</span>(ifnb[[<span class="st">"RNA"</span>]])</span>
<span id="cb42-137"><a href="#cb42-137" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb42-138"><a href="#cb42-138" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-139"><a href="#cb42-139" aria-hidden="true" tabindex="-1"></a>可以看到经过整合的Seurat对象的降维（"reduction"）中多出了整合后的降维（"integrated.cca"）：</span>
<span id="cb42-140"><a href="#cb42-140" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-141"><a href="#cb42-141" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/截屏2023-11-28%2009.57.11.png)</span></span>
<span id="cb42-142"><a href="#cb42-142" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-143"><a href="#cb42-143" aria-hidden="true" tabindex="-1"></a><span class="fu">### 整合后重新聚类、降维</span></span>
<span id="cb42-144"><a href="#cb42-144" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-147"><a href="#cb42-147" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb42-148"><a href="#cb42-148" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-整合后</span></span>
<span id="cb42-149"><a href="#cb42-149" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: 整合后的细胞分群情况（左：按照刺激条件着色；右：按照细胞聚类情况着色）</span></span>
<span id="cb42-150"><a href="#cb42-150" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 10</span></span>
<span id="cb42-151"><a href="#cb42-151" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-152"><a href="#cb42-152" aria-hidden="true" tabindex="-1"></a><span class="co"># 重新聚类</span></span>
<span id="cb42-153"><a href="#cb42-153" aria-hidden="true" tabindex="-1"></a>ifnb <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(ifnb, </span>
<span id="cb42-154"><a href="#cb42-154" aria-hidden="true" tabindex="-1"></a>                      <span class="at">reduction =</span> <span class="st">"integrated.cca"</span>, <span class="co">#更改降维来源为"integrated.cca"</span></span>
<span id="cb42-155"><a href="#cb42-155" aria-hidden="true" tabindex="-1"></a>                      <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>)</span>
<span id="cb42-156"><a href="#cb42-156" aria-hidden="true" tabindex="-1"></a>ifnb <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(ifnb, <span class="at">resolution =</span> <span class="dv">1</span>)</span>
<span id="cb42-157"><a href="#cb42-157" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-158"><a href="#cb42-158" aria-hidden="true" tabindex="-1"></a><span class="co"># 重新降维</span></span>
<span id="cb42-159"><a href="#cb42-159" aria-hidden="true" tabindex="-1"></a>ifnb <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(ifnb, </span>
<span id="cb42-160"><a href="#cb42-160" aria-hidden="true" tabindex="-1"></a>                <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>, </span>
<span id="cb42-161"><a href="#cb42-161" aria-hidden="true" tabindex="-1"></a>                <span class="at">reduction =</span> <span class="st">"integrated.cca"</span>) <span class="co">#更改降维来源为"integrated.cca"</span></span>
<span id="cb42-162"><a href="#cb42-162" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-163"><a href="#cb42-163" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualization：</span></span>
<span id="cb42-164"><a href="#cb42-164" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(ifnb, </span>
<span id="cb42-165"><a href="#cb42-165" aria-hidden="true" tabindex="-1"></a>        <span class="at">reduction =</span> <span class="st">"umap"</span>, </span>
<span id="cb42-166"><a href="#cb42-166" aria-hidden="true" tabindex="-1"></a>        <span class="at">group.by =</span> <span class="fu">c</span>(<span class="st">"stim"</span>, <span class="st">"seurat_annotations"</span>))</span>
<span id="cb42-167"><a href="#cb42-167" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb42-168"><a href="#cb42-168" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-169"><a href="#cb42-169" aria-hidden="true" tabindex="-1"></a>可以看到和 @fig-未整合 相比，在整合后，细胞就只按照细胞类型进行聚类了。</span>
<span id="cb42-170"><a href="#cb42-170" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-171"><a href="#cb42-171" aria-hidden="true" tabindex="-1"></a>也可以按照刺激条件（"stim"）绘制分面图，分别展示刺激组和对照组的细胞分群情况：</span>
<span id="cb42-172"><a href="#cb42-172" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-175"><a href="#cb42-175" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb42-176"><a href="#cb42-176" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 10</span></span>
<span id="cb42-177"><a href="#cb42-177" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(ifnb, <span class="at">reduction =</span> <span class="st">"umap"</span>, <span class="at">split.by =</span> <span class="st">"stim"</span>)</span>
<span id="cb42-178"><a href="#cb42-178" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb42-179"><a href="#cb42-179" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-180"><a href="#cb42-180" aria-hidden="true" tabindex="-1"></a>可以看到，和上面的结论一致，两种条件下的细胞分群基本一致。</span>
<span id="cb42-181"><a href="#cb42-181" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-182"><a href="#cb42-182" aria-hidden="true" tabindex="-1"></a><span class="fu">## 鉴定保守的cell marker</span></span>
<span id="cb42-183"><a href="#cb42-183" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-184"><a href="#cb42-184" aria-hidden="true" tabindex="-1"></a>To **identify canonical cell type marker genes that are conserved across conditions**, we provide the&nbsp;`FindConservedMarkers()` function. This function performs **differential gene expression testing** for each dataset/group and combines the p-values using meta-analysis methods from the MetaDE R package. For example, **we can calculated the genes that are conserved markers irrespective of stimulation condition in cluster 6 (NK cells).**</span>
<span id="cb42-185"><a href="#cb42-185" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-186"><a href="#cb42-186" aria-hidden="true" tabindex="-1"></a><span class="in">`FindConservedMarkers`</span>函数会调用<span class="in">`metap`</span>包，<span class="in">`metap`</span>包需要<span class="in">`multtest`</span>包，所以需要先安装这两个依赖包：</span>
<span id="cb42-187"><a href="#cb42-187" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-190"><a href="#cb42-190" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb42-191"><a href="#cb42-191" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb42-192"><a href="#cb42-192" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: true</span></span>
<span id="cb42-193"><a href="#cb42-193" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">'multtest'</span>)</span>
<span id="cb42-194"><a href="#cb42-194" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">'metap'</span>)</span>
<span id="cb42-195"><a href="#cb42-195" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb42-196"><a href="#cb42-196" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-199"><a href="#cb42-199" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb42-200"><a href="#cb42-200" aria-hidden="true" tabindex="-1"></a><span class="co"># 这里的meta.data已经提前注释好了细胞类型（储存在"seurat_annotations"列中）。</span></span>
<span id="cb42-201"><a href="#cb42-201" aria-hidden="true" tabindex="-1"></a><span class="co"># 将细胞类型注释指定为"seurat_annotations"</span></span>
<span id="cb42-202"><a href="#cb42-202" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(ifnb) <span class="ot">&lt;-</span> <span class="st">"seurat_annotations"</span></span>
<span id="cb42-203"><a href="#cb42-203" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-204"><a href="#cb42-204" aria-hidden="true" tabindex="-1"></a>nk.markers <span class="ot">&lt;-</span> <span class="fu">FindConservedMarkers</span>(ifnb, </span>
<span id="cb42-205"><a href="#cb42-205" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">ident.1 =</span> <span class="st">"NK"</span>, </span>
<span id="cb42-206"><a href="#cb42-206" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">grouping.var =</span> <span class="st">"stim"</span>, </span>
<span id="cb42-207"><a href="#cb42-207" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb42-208"><a href="#cb42-208" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(nk.markers)</span>
<span id="cb42-209"><a href="#cb42-209" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb42-210"><a href="#cb42-210" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-211"><a href="#cb42-211" aria-hidden="true" tabindex="-1"></a>在实际分析中，鉴定这些保守的cell marker主要用来辅助对cluster的注释：you can perform these same analysis on the unsupervised clustering results (stored in&nbsp;<span class="in">`seurat_clusters`</span>), and **use these conserved markers to annotate cell types in your dataset**.</span>
<span id="cb42-212"><a href="#cb42-212" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-213"><a href="#cb42-213" aria-hidden="true" tabindex="-1"></a><span class="fu">### 可视化cell markers的表达</span></span>
<span id="cb42-214"><a href="#cb42-214" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-215"><a href="#cb42-215" aria-hidden="true" tabindex="-1"></a>The&nbsp;<span class="in">`DotPlot()`</span>&nbsp;function with the&nbsp;<span class="in">`split.by`</span>&nbsp;parameter can be useful for viewing conserved cell type markers across conditions, showing both the expression level and the percentage of cells in a cluster expressing any given gene. Here we plot 2-3 strong marker genes for each of our 14 clusters.</span>
<span id="cb42-216"><a href="#cb42-216" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-219"><a href="#cb42-219" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb42-220"><a href="#cb42-220" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 10</span></span>
<span id="cb42-221"><a href="#cb42-221" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 10</span></span>
<span id="cb42-222"><a href="#cb42-222" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-223"><a href="#cb42-223" aria-hidden="true" tabindex="-1"></a><span class="co"># NEEDS TO BE FIXED AND SET ORDER CORRECTLY</span></span>
<span id="cb42-224"><a href="#cb42-224" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(ifnb) <span class="ot">&lt;-</span> <span class="fu">factor</span>(<span class="fu">Idents</span>(ifnb), </span>
<span id="cb42-225"><a href="#cb42-225" aria-hidden="true" tabindex="-1"></a>                       <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"pDC"</span>, <span class="st">"Eryth"</span>, <span class="st">"Mk"</span>, <span class="st">"DC"</span>, <span class="st">"CD14 Mono"</span>, <span class="st">"CD16 Mono"</span>, </span>
<span id="cb42-226"><a href="#cb42-226" aria-hidden="true" tabindex="-1"></a>                                  <span class="st">"B Activated"</span>, <span class="st">"B"</span>, <span class="st">"CD8 T"</span>, <span class="st">"NK"</span>, <span class="st">"T activated"</span>, </span>
<span id="cb42-227"><a href="#cb42-227" aria-hidden="true" tabindex="-1"></a>                                  <span class="st">"CD4 Naive T"</span>, <span class="st">"CD4 Memory T"</span>))</span>
<span id="cb42-228"><a href="#cb42-228" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-229"><a href="#cb42-229" aria-hidden="true" tabindex="-1"></a>markers.to.plot <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"CD3D"</span>, <span class="st">"CREM"</span>, <span class="st">"HSPH1"</span>, <span class="st">"SELL"</span>, <span class="st">"GIMAP5"</span>, <span class="st">"CACYBP"</span>, <span class="st">"GNLY"</span>, <span class="st">"NKG7"</span>,</span>
<span id="cb42-230"><a href="#cb42-230" aria-hidden="true" tabindex="-1"></a>                     <span class="st">"CCL5"</span>, <span class="st">"CD8A"</span>, <span class="st">"MS4A1"</span>, <span class="st">"CD79A"</span>, <span class="st">"MIR155HG"</span>, <span class="st">"NME1"</span>, <span class="st">"FCGR3A"</span>, </span>
<span id="cb42-231"><a href="#cb42-231" aria-hidden="true" tabindex="-1"></a>                     <span class="st">"VMO1"</span>, <span class="st">"CCL2"</span>, <span class="st">"S100A9"</span>, <span class="st">"HLA-DQA1"</span>, <span class="st">"GPR183"</span>, <span class="st">"PPBP"</span>, <span class="st">"GNG11"</span>,</span>
<span id="cb42-232"><a href="#cb42-232" aria-hidden="true" tabindex="-1"></a>                     <span class="st">"HBA2"</span>, <span class="st">"HBB"</span>, <span class="st">"TSPAN13"</span>, <span class="st">"IL3RA"</span>, <span class="st">"IGJ"</span>, <span class="st">"PRSS57"</span>)</span>
<span id="cb42-233"><a href="#cb42-233" aria-hidden="true" tabindex="-1"></a><span class="fu">DotPlot</span>(ifnb, </span>
<span id="cb42-234"><a href="#cb42-234" aria-hidden="true" tabindex="-1"></a>        <span class="at">features =</span> markers.to.plot, </span>
<span id="cb42-235"><a href="#cb42-235" aria-hidden="true" tabindex="-1"></a>        <span class="at">cols =</span> <span class="fu">c</span>(<span class="st">"blue"</span>, <span class="st">"red"</span>), </span>
<span id="cb42-236"><a href="#cb42-236" aria-hidden="true" tabindex="-1"></a>        <span class="at">dot.scale =</span> <span class="dv">8</span>, </span>
<span id="cb42-237"><a href="#cb42-237" aria-hidden="true" tabindex="-1"></a>        <span class="at">split.by =</span> <span class="st">"stim"</span>) <span class="sc">+</span></span>
<span id="cb42-238"><a href="#cb42-238" aria-hidden="true" tabindex="-1"></a>  <span class="fu">RotatedAxis</span>()</span>
<span id="cb42-239"><a href="#cb42-239" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb42-240"><a href="#cb42-240" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-241"><a href="#cb42-241" aria-hidden="true" tabindex="-1"></a><span class="fu">## 识别不同样本类型间的差异基因 {#sec-识别不同样本类型间的差异基因}</span></span>
<span id="cb42-242"><a href="#cb42-242" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-243"><a href="#cb42-243" aria-hidden="true" tabindex="-1"></a>**Now that we've aligned the stimulated and control cells, we can start to do comparative analyses and look at the differences induced by stimulation.**</span>
<span id="cb42-244"><a href="#cb42-244" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-245"><a href="#cb42-245" aria-hidden="true" tabindex="-1"></a>We can aggregate cells of a similar type and condition together to create "pseudobulk" profiles using the&nbsp;<span class="in">`AggregateExpression`</span> command（通过<span class="in">`AggregateExpression`</span>命令将同一类型的细胞按照不同的处理条件合并起来，形成一个假的组织水平的测序数据。本例中，细胞被注释为13种细胞类型，而处理条件为"STIM"和"CTRL"，因此总共会被合并成13<span class="sc">\*</span>2=26个类别，将每一个类别看作是一个样本，这样就形成了一个所谓的假的组织水平的测序数据）.</span>
<span id="cb42-246"><a href="#cb42-246" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-249"><a href="#cb42-249" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb42-250"><a href="#cb42-250" aria-hidden="true" tabindex="-1"></a>aggregate_ifnb <span class="ot">&lt;-</span> <span class="fu">AggregateExpression</span>(ifnb, </span>
<span id="cb42-251"><a href="#cb42-251" aria-hidden="true" tabindex="-1"></a>                                      <span class="at">group.by =</span> <span class="fu">c</span>(<span class="st">"seurat_annotations"</span>, <span class="st">"stim"</span>), </span>
<span id="cb42-252"><a href="#cb42-252" aria-hidden="true" tabindex="-1"></a>                                      <span class="at">return.seurat =</span> <span class="cn">TRUE</span>)</span>
<span id="cb42-253"><a href="#cb42-253" aria-hidden="true" tabindex="-1"></a>aggregate_ifnb</span>
<span id="cb42-254"><a href="#cb42-254" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(aggregate_ifnb<span class="sc">@</span>meta.data, <span class="dv">5</span>)</span>
<span id="cb42-255"><a href="#cb42-255" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(aggregate_ifnb) <span class="co"># 可以看到现在的表达矩阵的列（即样本）为细胞类型+处理条件</span></span>
<span id="cb42-256"><a href="#cb42-256" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb42-257"><a href="#cb42-257" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-258"><a href="#cb42-258" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/截屏2023-11-27%2015.59.00.png)</span></span>
<span id="cb42-259"><a href="#cb42-259" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-260"><a href="#cb42-260" aria-hidden="true" tabindex="-1"></a>As an initial exploratory analysis, we can **compare pseudobulk profiles of two cell types (naive CD4 T cells, and CD14 monocytes)**, and compare their gene expression profiles before and after stimulation. We highlight genes that exhibit dramatic responses to interferon stimulation.</span>
<span id="cb42-261"><a href="#cb42-261" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-264"><a href="#cb42-264" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb42-265"><a href="#cb42-265" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-cellscater</span></span>
<span id="cb42-266"><a href="#cb42-266" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: CD14 Mono和CD4 Naive T细胞中的基因在对照组和刺激组之间的表达量散点图</span></span>
<span id="cb42-267"><a href="#cb42-267" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-subcap: </span></span>
<span id="cb42-268"><a href="#cb42-268" aria-hidden="true" tabindex="-1"></a><span class="co">#|   - CD14 Mono细胞中的基因在对照组和刺激组之间的表达量散点图</span></span>
<span id="cb42-269"><a href="#cb42-269" aria-hidden="true" tabindex="-1"></a><span class="co">#|   - CD4 Naive T细胞中的基因在对照组和刺激组之间的表达量散点图</span></span>
<span id="cb42-270"><a href="#cb42-270" aria-hidden="true" tabindex="-1"></a><span class="co">#| layout-ncol: 2</span></span>
<span id="cb42-271"><a href="#cb42-271" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb42-272"><a href="#cb42-272" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(cowplot)</span>
<span id="cb42-273"><a href="#cb42-273" aria-hidden="true" tabindex="-1"></a><span class="fu">theme_set</span>(<span class="fu">theme_cowplot</span>())</span>
<span id="cb42-274"><a href="#cb42-274" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-275"><a href="#cb42-275" aria-hidden="true" tabindex="-1"></a><span class="co"># genes that exhibit dramatic responses to interferon stimulation</span></span>
<span id="cb42-276"><a href="#cb42-276" aria-hidden="true" tabindex="-1"></a>genes.to.label <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"ISG15"</span>, <span class="st">"LY6E"</span>, <span class="st">"IFI6"</span>, <span class="st">"ISG20"</span>, <span class="st">"MX1"</span>, <span class="st">"IFIT2"</span>, <span class="st">"IFIT1"</span>, <span class="st">"CXCL10"</span>,</span>
<span id="cb42-277"><a href="#cb42-277" aria-hidden="true" tabindex="-1"></a>                   <span class="st">"CCL8"</span>)</span>
<span id="cb42-278"><a href="#cb42-278" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-279"><a href="#cb42-279" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">CellScatter</span>(aggregate_ifnb, </span>
<span id="cb42-280"><a href="#cb42-280" aria-hidden="true" tabindex="-1"></a>                  <span class="st">"CD14 Mono_CTRL"</span>, <span class="st">"CD14 Mono_STIM"</span>, </span>
<span id="cb42-281"><a href="#cb42-281" aria-hidden="true" tabindex="-1"></a>                  <span class="at">highlight =</span> genes.to.label)</span>
<span id="cb42-282"><a href="#cb42-282" aria-hidden="true" tabindex="-1"></a><span class="fu">LabelPoints</span>(<span class="at">plot =</span> p1, </span>
<span id="cb42-283"><a href="#cb42-283" aria-hidden="true" tabindex="-1"></a>                  <span class="at">points =</span> genes.to.label, </span>
<span id="cb42-284"><a href="#cb42-284" aria-hidden="true" tabindex="-1"></a>                  <span class="at">repel =</span> <span class="cn">TRUE</span>)</span>
<span id="cb42-285"><a href="#cb42-285" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-286"><a href="#cb42-286" aria-hidden="true" tabindex="-1"></a>p3 <span class="ot">&lt;-</span> <span class="fu">CellScatter</span>(aggregate_ifnb, </span>
<span id="cb42-287"><a href="#cb42-287" aria-hidden="true" tabindex="-1"></a>                  <span class="st">"CD4 Naive T_CTRL"</span>, <span class="st">"CD4 Naive T_STIM"</span>, </span>
<span id="cb42-288"><a href="#cb42-288" aria-hidden="true" tabindex="-1"></a>                  <span class="at">highlight =</span> genes.to.label)</span>
<span id="cb42-289"><a href="#cb42-289" aria-hidden="true" tabindex="-1"></a><span class="fu">LabelPoints</span>(<span class="at">plot =</span> p3, </span>
<span id="cb42-290"><a href="#cb42-290" aria-hidden="true" tabindex="-1"></a>                  <span class="at">points =</span> genes.to.label, </span>
<span id="cb42-291"><a href="#cb42-291" aria-hidden="true" tabindex="-1"></a>                  <span class="at">repel =</span> <span class="cn">TRUE</span>)</span>
<span id="cb42-292"><a href="#cb42-292" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb42-293"><a href="#cb42-293" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-294"><a href="#cb42-294" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; As you can see, many of the same genes are **upregulated** (位于对角线上方) in both of these cell types and **likely represent a conserved interferon response pathway**, though CD14 monocytes exhibit a stronger transcriptional response.</span></span>
<span id="cb42-295"><a href="#cb42-295" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-296"><a href="#cb42-296" aria-hidden="true" tabindex="-1"></a><span class="fu">### 正式差异分析 {#sec-正式差异分析}</span></span>
<span id="cb42-297"><a href="#cb42-297" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-298"><a href="#cb42-298" aria-hidden="true" tabindex="-1"></a>We can now ask **what genes change in different conditions for cells of the same type**.</span>
<span id="cb42-299"><a href="#cb42-299" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-300"><a href="#cb42-300" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>First, we create a column in the meta.data slot to hold both the cell type and stimulation information and switch the current ident to that column.</span>
<span id="cb42-301"><a href="#cb42-301" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-302"><a href="#cb42-302" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>Then we use&nbsp;<span class="co">[</span><span class="ot">`FindMarkers()`</span><span class="co">](https://satijalab.org/seurat/reference/findmarkers)</span>&nbsp;to find the genes that are different **between stimulated and control B cells**. Notice that many of the top genes that show up here are the same as the ones we plotted earlier as core interferon response genes. Additionally, genes like CXCL10 which we saw were specific to monocyte and B cell interferon response show up as highly significant in this list as well.</span>
<span id="cb42-303"><a href="#cb42-303" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-306"><a href="#cb42-306" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb42-307"><a href="#cb42-307" aria-hidden="true" tabindex="-1"></a>ifnb<span class="sc">$</span>celltype.stim <span class="ot">&lt;-</span> <span class="fu">paste</span>(ifnb<span class="sc">$</span>seurat_annotations, ifnb<span class="sc">$</span>stim, <span class="at">sep =</span> <span class="st">"_"</span>)</span>
<span id="cb42-308"><a href="#cb42-308" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(ifnb) <span class="ot">&lt;-</span> <span class="st">"celltype.stim"</span></span>
<span id="cb42-309"><a href="#cb42-309" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-310"><a href="#cb42-310" aria-hidden="true" tabindex="-1"></a><span class="co"># 寻找对照组和刺激组之间在B细胞中的差异基因</span></span>
<span id="cb42-311"><a href="#cb42-311" aria-hidden="true" tabindex="-1"></a>b.interferon.response <span class="ot">&lt;-</span> <span class="fu">FindMarkers</span>(ifnb, </span>
<span id="cb42-312"><a href="#cb42-312" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">ident.1 =</span> <span class="st">"B_STIM"</span>, </span>
<span id="cb42-313"><a href="#cb42-313" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">ident.2 =</span> <span class="st">"B_CTRL"</span>, </span>
<span id="cb42-314"><a href="#cb42-314" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb42-315"><a href="#cb42-315" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(b.interferon.response, <span class="at">n =</span> <span class="dv">15</span>)</span>
<span id="cb42-316"><a href="#cb42-316" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb42-317"><a href="#cb42-317" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-318"><a href="#cb42-318" aria-hidden="true" tabindex="-1"></a>Please note that p-values obtained from this analysis should be interpreted with caution, as **these tests treat each cell as an independent replicate, and ignore inherent correlations between cells originating from the same sample**. As discussed&nbsp;here <span class="co">[</span><span class="ot">@crowell2020</span><span class="co">]</span>, DE tests across multiple conditions should expressly utilize multiple samples/replicates, and can be performed after aggregating ('pseudobulking') cells from the same sample and subpopulation together. We do not perform this analysis here, as there is a single replicate in the data, but please see our&nbsp;<span class="co">[</span><span class="ot">vignette comparing healthy and diabetic samples</span><span class="co">](https://satijalab.org/seurat/articles/parsebio_sketch_integration)</span>&nbsp;as an example for how to perform DE analysis across conditions.</span>
<span id="cb42-319"><a href="#cb42-319" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-320"><a href="#cb42-320" aria-hidden="true" tabindex="-1"></a>Another useful way to visualize these changes in gene expression is with the&nbsp;<span class="in">`split.by`</span>&nbsp;option to the&nbsp;<span class="in">`FeaturePlot()`</span>&nbsp;or&nbsp;<span class="in">`VlnPlot()`</span>&nbsp;function. This will display FeaturePlots of the list of given genes, split by a grouping variable (stimulation condition here).</span>
<span id="cb42-321"><a href="#cb42-321" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-324"><a href="#cb42-324" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb42-325"><a href="#cb42-325" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 12</span></span>
<span id="cb42-326"><a href="#cb42-326" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>(ifnb, </span>
<span id="cb42-327"><a href="#cb42-327" aria-hidden="true" tabindex="-1"></a>            <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"CD3D"</span>, <span class="st">"GNLY"</span>, <span class="st">"IFI6"</span>, <span class="st">"ISG15"</span>, <span class="st">"CD14"</span>, <span class="st">"CXCL10"</span>), </span>
<span id="cb42-328"><a href="#cb42-328" aria-hidden="true" tabindex="-1"></a>            <span class="at">split.by =</span> <span class="st">"stim"</span>, </span>
<span id="cb42-329"><a href="#cb42-329" aria-hidden="true" tabindex="-1"></a>            <span class="at">max.cutoff =</span> <span class="dv">3</span>, </span>
<span id="cb42-330"><a href="#cb42-330" aria-hidden="true" tabindex="-1"></a>            <span class="at">cols =</span> <span class="fu">c</span>(<span class="st">"grey"</span>, <span class="st">"red"</span>), </span>
<span id="cb42-331"><a href="#cb42-331" aria-hidden="true" tabindex="-1"></a>            <span class="at">reduction =</span> <span class="st">"umap"</span>)</span>
<span id="cb42-332"><a href="#cb42-332" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb42-333"><a href="#cb42-333" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-336"><a href="#cb42-336" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb42-337"><a href="#cb42-337" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 10</span></span>
<span id="cb42-338"><a href="#cb42-338" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 15</span></span>
<span id="cb42-339"><a href="#cb42-339" aria-hidden="true" tabindex="-1"></a>plots <span class="ot">&lt;-</span> <span class="fu">VlnPlot</span>(ifnb,</span>
<span id="cb42-340"><a href="#cb42-340" aria-hidden="true" tabindex="-1"></a>        <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"CD3D"</span>, <span class="st">"GNLY"</span>, <span class="st">"IFI6"</span>, <span class="st">"ISG15"</span>, <span class="st">"CD14"</span>, <span class="st">"CXCL10"</span>, <span class="st">"LYZ"</span>),</span>
<span id="cb42-341"><a href="#cb42-341" aria-hidden="true" tabindex="-1"></a>        <span class="at">split.by =</span> <span class="st">"stim"</span>,</span>
<span id="cb42-342"><a href="#cb42-342" aria-hidden="true" tabindex="-1"></a>        <span class="at">group.by =</span> <span class="st">"seurat_annotations"</span>,</span>
<span id="cb42-343"><a href="#cb42-343" aria-hidden="true" tabindex="-1"></a>        <span class="at">pt.size =</span> <span class="dv">0</span>,</span>
<span id="cb42-344"><a href="#cb42-344" aria-hidden="true" tabindex="-1"></a>        <span class="at">combine =</span> <span class="cn">FALSE</span>) <span class="co"># 由于VlnPlot绘制组图时没有图例，所以这里取消绘制组图</span></span>
<span id="cb42-345"><a href="#cb42-345" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb42-346"><a href="#cb42-346" aria-hidden="true" tabindex="-1"></a><span class="fu">wrap_plots</span>(<span class="at">plots =</span> plots, <span class="at">ncol =</span> <span class="dv">2</span>) <span class="co"># 将plots列表组合成组图</span></span>
<span id="cb42-347"><a href="#cb42-347" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb42-348"><a href="#cb42-348" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-349"><a href="#cb42-349" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; -   *Genes such as **CD3D** and **GNLY** are canonical cell type markers (for T cells and NK/CD8 T cells) that are virtually **unaffected by interferon stimulation** and **display similar gene expression patterns in the control and stimulated group**.*</span></span>
<span id="cb42-350"><a href="#cb42-350" aria-hidden="true" tabindex="-1"></a><span class="at">&gt;</span></span>
<span id="cb42-351"><a href="#cb42-351" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; -   ***IFI6** and **ISG15**, on the other hand, are **core interferon response genes** and are **upregulated** accordingly **in all cell types**.*</span></span>
<span id="cb42-352"><a href="#cb42-352" aria-hidden="true" tabindex="-1"></a><span class="at">&gt;</span></span>
<span id="cb42-353"><a href="#cb42-353" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; -   ***CD14** and **CXCL10** are genes that show a **cell type specific interferon response**.*</span></span>
<span id="cb42-354"><a href="#cb42-354" aria-hidden="true" tabindex="-1"></a><span class="at">&gt;</span></span>
<span id="cb42-355"><a href="#cb42-355" aria-hidden="true" tabindex="-1"></a><span class="at">&gt;     -   ***CD14** expression **decreases** after stimulation in **CD14 monocytes**, [which could lead to misclassification in a supervised analysis framework, underscoring the value of integrated analysis.]{.underline}**如果用于识别细胞类型的marker本身在不同的样本类型（处理 vs. 对照、恶性组织 vs. 正常组织）中存在表达量的差异，那么就会导致对细胞类型判断的错误。而本篇的数据整合则能够避免出现这种情况。***</span></span>
<span id="cb42-356"><a href="#cb42-356" aria-hidden="true" tabindex="-1"></a><span class="at">&gt;</span></span>
<span id="cb42-357"><a href="#cb42-357" aria-hidden="true" tabindex="-1"></a><span class="at">&gt;     -   ***CXCL10** shows a distinct **upregulation** in **monocytes** and **B cells** after interferon stimulation but not in other cell types.*</span></span>
<span id="cb42-358"><a href="#cb42-358" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-359"><a href="#cb42-359" aria-hidden="true" tabindex="-1"></a><span class="fu">## 执行`SCTransform`标准化流程之后的整合</span></span>
<span id="cb42-360"><a href="#cb42-360" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-361"><a href="#cb42-361" aria-hidden="true" tabindex="-1"></a>As an alternative to log-normalization, Seurat also includes support for preprocessing of scRNA-seq using the&nbsp;<span class="in">`SCTransform`</span> workflow（ @sec-sctransform ）. The&nbsp;<span class="in">`IntegrateLayers`</span>&nbsp;function also supports SCTransform-normalized data, by setting the&nbsp;<span class="in">`normalization.method`</span> parameter, as shown below.</span>
<span id="cb42-362"><a href="#cb42-362" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-363"><a href="#cb42-363" aria-hidden="true" tabindex="-1"></a><span class="fu">### 不进行整合的情况下的数据分析</span></span>
<span id="cb42-364"><a href="#cb42-364" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-367"><a href="#cb42-367" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb42-368"><a href="#cb42-368" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-sctransform_未整合</span></span>
<span id="cb42-369"><a href="#cb42-369" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: 未整合时的细胞分群情况（左：按照刺激条件着色；右：按照细胞聚类情况着色）</span></span>
<span id="cb42-370"><a href="#cb42-370" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 10</span></span>
<span id="cb42-371"><a href="#cb42-371" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-372"><a href="#cb42-372" aria-hidden="true" tabindex="-1"></a><span class="co"># 重新载入原始的Seurat对象ifnb</span></span>
<span id="cb42-373"><a href="#cb42-373" aria-hidden="true" tabindex="-1"></a>ifnb <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"data/pbmc_ifnb.rds"</span>)</span>
<span id="cb42-374"><a href="#cb42-374" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-375"><a href="#cb42-375" aria-hidden="true" tabindex="-1"></a><span class="co"># 同样先拆分数据集，然后进行无整合情况下的降维</span></span>
<span id="cb42-376"><a href="#cb42-376" aria-hidden="true" tabindex="-1"></a>ifnb[[<span class="st">"RNA"</span>]] <span class="ot">&lt;-</span> <span class="fu">split</span>(ifnb[[<span class="st">"RNA"</span>]], <span class="at">f =</span> ifnb<span class="sc">$</span>stim)</span>
<span id="cb42-377"><a href="#cb42-377" aria-hidden="true" tabindex="-1"></a>ifnb <span class="ot">&lt;-</span> <span class="fu">SCTransform</span>(ifnb)</span>
<span id="cb42-378"><a href="#cb42-378" aria-hidden="true" tabindex="-1"></a>ifnb <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(ifnb)</span>
<span id="cb42-379"><a href="#cb42-379" aria-hidden="true" tabindex="-1"></a>ifnb <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(ifnb, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>)</span>
<span id="cb42-380"><a href="#cb42-380" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(ifnb, </span>
<span id="cb42-381"><a href="#cb42-381" aria-hidden="true" tabindex="-1"></a>        <span class="at">reduction =</span> <span class="st">"umap"</span>, </span>
<span id="cb42-382"><a href="#cb42-382" aria-hidden="true" tabindex="-1"></a>        <span class="at">group.by =</span> <span class="fu">c</span>(<span class="st">"stim"</span>, <span class="st">"seurat_annotations"</span>))</span>
<span id="cb42-383"><a href="#cb42-383" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb42-384"><a href="#cb42-384" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-385"><a href="#cb42-385" aria-hidden="true" tabindex="-1"></a>可以看到，如果不进行整合，不同样本（STIM vs. STIM）的细胞类型差异很大。</span>
<span id="cb42-386"><a href="#cb42-386" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-387"><a href="#cb42-387" aria-hidden="true" tabindex="-1"></a><span class="fu">### 进行整合 {#sec-integration_after_sct}</span></span>
<span id="cb42-388"><a href="#cb42-388" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-389"><a href="#cb42-389" aria-hidden="true" tabindex="-1"></a>同样通过<span class="in">`IntegrateLayers`</span>函数进行数据整合，只不过需要将默认的标准化方法由"LogNormalize"指定为"SCT"（<span class="in">`normalization.method = "SCT"`</span>）：</span>
<span id="cb42-390"><a href="#cb42-390" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-393"><a href="#cb42-393" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb42-394"><a href="#cb42-394" aria-hidden="true" tabindex="-1"></a>ifnb <span class="ot">&lt;-</span> <span class="fu">IntegrateLayers</span>(<span class="at">object =</span> ifnb, </span>
<span id="cb42-395"><a href="#cb42-395" aria-hidden="true" tabindex="-1"></a>                        <span class="at">method =</span> CCAIntegration, </span>
<span id="cb42-396"><a href="#cb42-396" aria-hidden="true" tabindex="-1"></a>                        <span class="at">normalization.method =</span> <span class="st">"SCT"</span>, </span>
<span id="cb42-397"><a href="#cb42-397" aria-hidden="true" tabindex="-1"></a>                        <span class="at">verbose =</span> F)</span>
<span id="cb42-398"><a href="#cb42-398" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb42-399"><a href="#cb42-399" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-400"><a href="#cb42-400" aria-hidden="true" tabindex="-1"></a><span class="al">![](images/截屏2023-11-28%2009.16.10.png)</span></span>
<span id="cb42-401"><a href="#cb42-401" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-402"><a href="#cb42-402" aria-hidden="true" tabindex="-1"></a>可以看到经过整合的现在的Seurat对象中除了"RNA"的assay还由"SCT"的assay。同时，降维（"reduction"）中多出了整合后的降维（"integrated.dr"）。</span>
<span id="cb42-403"><a href="#cb42-403" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-404"><a href="#cb42-404" aria-hidden="true" tabindex="-1"></a><span class="fu">### 整合后聚类</span></span>
<span id="cb42-405"><a href="#cb42-405" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-408"><a href="#cb42-408" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb42-409"><a href="#cb42-409" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 10</span></span>
<span id="cb42-410"><a href="#cb42-410" aria-hidden="true" tabindex="-1"></a>ifnb <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(ifnb, </span>
<span id="cb42-411"><a href="#cb42-411" aria-hidden="true" tabindex="-1"></a>                      <span class="at">reduction =</span> <span class="st">"integrated.dr"</span>, <span class="co">#更改降维来源为"integrated.dr"</span></span>
<span id="cb42-412"><a href="#cb42-412" aria-hidden="true" tabindex="-1"></a>                      <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>)</span>
<span id="cb42-413"><a href="#cb42-413" aria-hidden="true" tabindex="-1"></a>ifnb <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(ifnb, </span>
<span id="cb42-414"><a href="#cb42-414" aria-hidden="true" tabindex="-1"></a>                     <span class="at">resolution =</span> <span class="fl">0.6</span>)</span>
<span id="cb42-415"><a href="#cb42-415" aria-hidden="true" tabindex="-1"></a>ifnb <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(ifnb, </span>
<span id="cb42-416"><a href="#cb42-416" aria-hidden="true" tabindex="-1"></a>                <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>, </span>
<span id="cb42-417"><a href="#cb42-417" aria-hidden="true" tabindex="-1"></a>                <span class="at">reduction =</span> <span class="st">"integrated.dr"</span>)</span>
<span id="cb42-418"><a href="#cb42-418" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(ifnb, </span>
<span id="cb42-419"><a href="#cb42-419" aria-hidden="true" tabindex="-1"></a>        <span class="at">reduction =</span> <span class="st">"umap"</span>, </span>
<span id="cb42-420"><a href="#cb42-420" aria-hidden="true" tabindex="-1"></a>        <span class="at">group.by =</span> <span class="fu">c</span>(<span class="st">"stim"</span>, <span class="st">"seurat_annotations"</span>))</span>
<span id="cb42-421"><a href="#cb42-421" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb42-422"><a href="#cb42-422" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-423"><a href="#cb42-423" aria-hidden="true" tabindex="-1"></a>可以看到和 @fig-sctransform_未整合 相比，整合后在样本间的细胞类型基本均匀分布。</span>
<span id="cb42-424"><a href="#cb42-424" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-425"><a href="#cb42-425" aria-hidden="true" tabindex="-1"></a><span class="fu">### 差异表达分析</span></span>
<span id="cb42-426"><a href="#cb42-426" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-427"><a href="#cb42-427" aria-hidden="true" tabindex="-1"></a>对于<span class="in">`SCTransform`</span>处理的数据首先要通过PrepSCTFindMarkers函数来预处理，然后再进行差异分析，基本内容和 @sec-正式差异分析 一致。</span>
<span id="cb42-428"><a href="#cb42-428" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-431"><a href="#cb42-431" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb42-432"><a href="#cb42-432" aria-hidden="true" tabindex="-1"></a>ifnb <span class="ot">&lt;-</span> <span class="fu">PrepSCTFindMarkers</span>(ifnb)</span>
<span id="cb42-433"><a href="#cb42-433" aria-hidden="true" tabindex="-1"></a>ifnb<span class="sc">$</span>celltype.stim <span class="ot">&lt;-</span> <span class="fu">paste</span>(ifnb<span class="sc">$</span>seurat_annotations, </span>
<span id="cb42-434"><a href="#cb42-434" aria-hidden="true" tabindex="-1"></a>                            ifnb<span class="sc">$</span>stim, </span>
<span id="cb42-435"><a href="#cb42-435" aria-hidden="true" tabindex="-1"></a>                            <span class="at">sep =</span> <span class="st">"_"</span>)</span>
<span id="cb42-436"><a href="#cb42-436" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(ifnb) <span class="ot">&lt;-</span> <span class="st">"celltype.stim"</span></span>
<span id="cb42-437"><a href="#cb42-437" aria-hidden="true" tabindex="-1"></a>b.interferon.response <span class="ot">&lt;-</span> <span class="fu">FindMarkers</span>(ifnb, </span>
<span id="cb42-438"><a href="#cb42-438" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">ident.1 =</span> <span class="st">"B_STIM"</span>, </span>
<span id="cb42-439"><a href="#cb42-439" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">ident.2 =</span> <span class="st">"B_CTRL"</span>, </span>
<span id="cb42-440"><a href="#cb42-440" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb42-441"><a href="#cb42-441" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(b.interferon.response, <span class="at">n =</span> <span class="dv">15</span>)</span>
<span id="cb42-442"><a href="#cb42-442" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb42-443"><a href="#cb42-443" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-446"><a href="#cb42-446" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb42-447"><a href="#cb42-447" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 12</span></span>
<span id="cb42-448"><a href="#cb42-448" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>(ifnb, </span>
<span id="cb42-449"><a href="#cb42-449" aria-hidden="true" tabindex="-1"></a>            <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"CD3D"</span>, <span class="st">"GNLY"</span>, <span class="st">"IFI6"</span>, <span class="st">"ISG15"</span>, <span class="st">"CD14"</span>, <span class="st">"CXCL10"</span>), </span>
<span id="cb42-450"><a href="#cb42-450" aria-hidden="true" tabindex="-1"></a>            <span class="at">split.by =</span> <span class="st">"stim"</span>, </span>
<span id="cb42-451"><a href="#cb42-451" aria-hidden="true" tabindex="-1"></a>            <span class="at">max.cutoff =</span> <span class="dv">3</span>, </span>
<span id="cb42-452"><a href="#cb42-452" aria-hidden="true" tabindex="-1"></a>            <span class="at">cols =</span> <span class="fu">c</span>(<span class="st">"grey"</span>, <span class="st">"red"</span>), </span>
<span id="cb42-453"><a href="#cb42-453" aria-hidden="true" tabindex="-1"></a>            <span class="at">reduction =</span> <span class="st">"umap"</span>)</span>
<span id="cb42-454"><a href="#cb42-454" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb42-455"><a href="#cb42-455" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-456"><a href="#cb42-456" aria-hidden="true" tabindex="-1"></a>::: {.callout-note collapse="true" icon="false"}</span>
<span id="cb42-457"><a href="#cb42-457" aria-hidden="true" tabindex="-1"></a><span class="fu">## Session Info</span></span>
<span id="cb42-458"><a href="#cb42-458" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-461"><a href="#cb42-461" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb42-462"><a href="#cb42-462" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span>
<span id="cb42-463"><a href="#cb42-463" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb42-464"><a href="#cb42-464" aria-hidden="true" tabindex="-1"></a>:::</span>
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">
      <ul class="footer-items list-unstyled">
<li class="nav-item">
 This book was built with Quarto
  </li>  
</ul>
</div>   
    <div class="nav-footer-center">Copyright 2023, Du Junhong</div>
    <div class="nav-footer-right">
      <ul class="footer-items list-unstyled">
<li class="nav-item compact">
    <a class="nav-link" href="https://github.com/djhcod">
      <i class="bi bi-github" role="img">
</i> 
    </a>
  </li>  
</ul>
</div>
  </div>
</footer>


</body></html>